<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en" sourcelanguage="en">
<context>
    <name>GTest_UHMM3SearchCompare</name>
    <message>
        <source>cannot_parse_double_number_from_string:%1</source>
        <translation type="obsolete">Internal error (cannot parse float number from string &apos;%1&apos;)</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="41"/>
        <source>Writing HMM profile file failed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="302"/>
        <source>HMM files are read only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="350"/>
        <source>HMM reader error occurred: </source>
        <translation type="unfinished">HMM reader error:</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="351"/>
        <source>Reading file failed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="78"/>
        <source>HMM3 Profile</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GTest_CompareHmmFiles</name>
    <message>
        <source>file1_not_given</source>
        <translation type="obsolete">File #1 not set</translation>
    </message>
    <message>
        <source>file2_not_given</source>
        <translation type="obsolete">File #2 not set</translation>
    </message>
    <message>
        <source>cannot_create_io_adapter_for_1_file</source>
        <translation type="obsolete">Error creating ioadapter for first file</translation>
    </message>
    <message>
        <source>cannot_open_1_file</source>
        <translation type="obsolete">Error opening 1 file</translation>
    </message>
    <message>
        <source>cannot_create_io_adapter_for_2_file</source>
        <translation type="obsolete">Error creating ioadapter for second file</translation>
    </message>
    <message>
        <source>cannot_open_2_file</source>
        <translation type="obsolete">Error opening second file</translation>
    </message>
    <message>
        <source>names_of_alignments_not_similar</source>
        <translation type="obsolete">Names of aligments not matched</translation>
    </message>
    <message>
        <source>strings_written_from_files:different_length</source>
        <translation type="obsolete">Comparing files length not matched</translation>
    </message>
    <message>
        <source>files_not_equal: &quot;%1&quot; and &quot;%2&quot;</source>
        <translation type="obsolete">Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="304"/>
        <source>File #1 not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="310"/>
        <source>File #2 not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="370"/>
        <source>Error creating ioadapter for first file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="374"/>
        <source>Error opening 1 file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="381"/>
        <source>Error creating ioadapter for second file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="385"/>
        <source>Error opening second file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="410"/>
        <source>Names of aligments not matched</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="414"/>
        <source>Comparing files length not matched</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="421"/>
        <source>Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMM3Search</name>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="247"/>
        <source>Sequence is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMM3SearchCompare</name>
    <message>
        <source>cannot_parse_double_number_from_string:%1</source>
        <translation type="obsolete">Internal error (cannot parse float number from string &apos;%1&apos;)</translation>
    </message>
    <message>
        <source>cannot_find_search_task_in_context</source>
        <translation type="obsolete">No search task in test context</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="374"/>
        <source>Internal error (cannot parse float number from string &apos;%1&apos;)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="387"/>
        <source>Can&apos;t parse significance:%1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="602"/>
        <source>No search task in test context</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMMER3Build</name>
    <message>
        <source>input_file_not_given</source>
        <translation type="obsolete">No input file given</translation>
    </message>
    <message>
        <source>output_file_not_given</source>
        <translation type="obsolete">No output file given</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="228"/>
        <source>No input file given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="234"/>
        <source>No output file given</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3BuildPrompter</name>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="109"/>
        <source>For each MSA from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="111"/>
        <source>%1 builds a HMM3 profile.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3BuildWorker</name>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="61"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="62"/>
        <source>Input multiple sequence alignment for building statistical model.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="63"/>
        <source>HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="63"/>
        <source>Produced HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="73"/>
        <source>Random seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="73"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="77"/>
        <source>HMM3 Build</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="77"/>
        <source>Builds a HMM3 profile from a multiple sequence alignment.&lt;p&gt;The HMM3 profile is a statistical model which captures position-specific information about how conserved each column of the alignment is, and which residues are likely.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="182"/>
        <source>Built HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Lib</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="72"/>
        <source>HMM3 Profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="80"/>
        <source>HMMER3 Tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="142"/>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="159"/>
        <source>HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="142"/>
        <source>Input HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="143"/>
        <source>Location</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="143"/>
        <source>Location hint for the target file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="153"/>
        <source>Write HMM3 Profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="153"/>
        <source>Saves all input HMM3 profiles to specified location.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="159"/>
        <source>Loaded HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="166"/>
        <source>Read HMM3 Profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="166"/>
        <source>Reads HMM3 profiles from file(s). The files can be local or Internet URLs.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3ReadPrompter</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="178"/>
        <source>Read HMM3 profile(s) from %1.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Reader</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="253"/>
        <source>Loaded HMM3 profile(s) from %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3SearchPrompter</name>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="136"/>
        <source>For each sequence from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="137"/>
        <source>using all profiles provided by &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="141"/>
        <source>%1 search HMM3 signals %2. &lt;br&gt;Output the list of found regions annotated as &lt;u&gt;%4&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3SearchWorker</name>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="71"/>
        <source>HMM3 profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="71"/>
        <source>HMM3 profile(s) to search with.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="72"/>
        <source>Input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="73"/>
        <source>An input sequence (nucleotide or protein) to search in.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="74"/>
        <source>HMM3 annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="75"/>
        <source>Annotations marking found similar sequence regions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="89"/>
        <source>Result annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="89"/>
        <source>A name of the result annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="90"/>
        <source>Seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="90"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="91"/>
        <source>Filter by high E-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="91"/>
        <source>Report domains with e-value less than.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="92"/>
        <source>Filter by low score</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="92"/>
        <source>Report domains with score greater than.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="100"/>
        <source>HMM3 Search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="101"/>
        <source>Searches each input sequence for significantly similar sequence matches to all specified HMM profiles. In case several profiles were supplied, searches with all profiles one by one and outputs united set of annotations for each sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="168"/>
        <source>Power of e-value must be less or equal to zero. Using default value: 1e-1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="175"/>
        <source>Score must be greater than zero. Using default value: 0.01</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="181"/>
        <source>Value for attribute name is empty, default name used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="227"/>
        <source>Search HMM3 signals in %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="231"/>
        <source>Bad sequence supplied to input: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="260"/>
        <source>Found %1 HMM3 signals</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3WritePrompter</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="184"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="188"/>
        <source>Save HMM3 profile(s) from &lt;u&gt;%1&lt;/u&gt; to &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Writer</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="277"/>
        <source>Empty HMM3 passed for writing to %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="277"/>
        <source>Unspecified URL for writing HMM3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="293"/>
        <source>Writing HMM3 profile to %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3ADVContext</name>
    <message>
        <source>Search HMM signals with HMMER3</source>
        <translation type="obsolete">Search HMM signals with HMMER3</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="247"/>
        <source>Search HMM signals with HMMER3...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="257"/>
        <source>error</source>
        <translation type="unfinished">Error</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="257"/>
        <source>No sequence in focus found</source>
        <translation type="unfinished">No sequence in focus found</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Build</name>
    <message>
        <source>cannot_convert_alphabet</source>
        <translation type="obsolete">UGENE cannot determine alphabet of alignment</translation>
    </message>
    <message>
        <source>no_memory: cannot_create_alphabet</source>
        <translation type="obsolete">Run out of memory (creating alphabet failed)</translation>
    </message>
    <message>
        <source>no_memory: cannot_create_null_model</source>
        <translation type="obsolete">Run out of memory (creating null model failed)</translation>
    </message>
    <message>
        <source>no_memory: cannot_create_builder</source>
        <translation type="obsolete">Run out of memory (creating builder failed)</translation>
    </message>
    <message>
        <source>no_memory: cannot_convert_msa</source>
        <translation type="obsolete">Run out of memory (creating multiple alignment failed)</translation>
    </message>
    <message>
        <source>no_memory: cannot_digitize_msa</source>
        <translation type="obsolete">Run out of memory (digitizing of alignment failed)</translation>
    </message>
    <message>
        <source>builder_error_occurred</source>
        <translation type="obsolete">Model building failed</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="65"/>
        <source>UGENE cannot determine alphabet of alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="70"/>
        <source>Run out of memory (creating alphabet failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="76"/>
        <source>Run out of memory (creating null model failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="81"/>
        <source>Run out of memory (creating builder failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="87"/>
        <source>Run out of memory (creating multiple alignment failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="92"/>
        <source>Run out of memory (digitizing of alignment failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="100"/>
        <source>Model building failed</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildDialogImpl</name>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="69"/>
        <source>Build</source>
        <translation type="unfinished">Build</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="70"/>
        <source>Cancel</source>
        <translation type="unfinished">Cancel</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="109"/>
        <source>Select multiple alignment file</source>
        <translation type="unfinished">Select multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="118"/>
        <source>Select hmm file to create</source>
        <translation type="unfinished">Select HMM file to create</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="184"/>
        <source>input file is empty</source>
        <translation type="unfinished">Input file is empty</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="187"/>
        <source>output hmm file is empty</source>
        <translation type="unfinished">Output HMM profile file is empty</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="196"/>
        <source>Error: bad arguments!</source>
        <translation type="unfinished">Error: bad arguments!</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildTask</name>
    <message>
        <source>Build HMM profile</source>
        <translation type="obsolete">Build HMM profile</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="57"/>
        <source>Build HMM profile from %1 alignment</source>
        <translation type="unfinished">Build HMM profile from %1 alignment</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="71"/>
        <source>Given multiple alignment has no sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="74"/>
        <source>Given multiple alignment is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>multiple_alignment_is_empty</source>
        <translation type="obsolete">Given multiple alignment has no sequences</translation>
    </message>
    <message>
        <source>multiple_alignment_is_0_len</source>
        <translation type="obsolete">Given multiple alignment is empty</translation>
    </message>
    <message>
        <source>multiple_alignment_is_not_consistent</source>
        <translation type="obsolete">Given multiple alignment is not normalized</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildToFileTask</name>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="126"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="146"/>
        <source>Build HMM profile to &apos;%1&apos;</source>
        <translation type="unfinished">Build HMM profile to &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_output_file_given</source>
        <translation type="obsolete">Output file is not given</translation>
    </message>
    <message>
        <source>empty_msa_list_given</source>
        <translation type="obsolete">No multiple alignments given</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="163"/>
        <source>Build HMM profile &apos;%1&apos; -&gt; &apos;%2&apos;</source>
        <translation type="unfinished">Build HMM profile &apos;%1&apos; -&gt; &apos;%2&apos;</translation>
    </message>
    <message>
        <source>no_input_file_given</source>
        <translation type="obsolete">Input file is not given</translation>
    </message>
    <message>
        <source>input_format_error</source>
        <translation type="obsolete">Unrecognized input alignment file format</translation>
    </message>
    <message>
        <source>cannot_create_io_adapter_for_%1_file</source>
        <translation type="obsolete">Error opening &apos;%1&apos; file</translation>
    </message>
    <message>
        <source>alignment_objects_not_found_in_document</source>
        <translation type="obsolete">No multiple alignments found in input file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="129"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="149"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="170"/>
        <source>Output file is not given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="134"/>
        <source>No multiple alignments given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="166"/>
        <source>Input file is not given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="181"/>
        <source>Unrecognized input alignment file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="188"/>
        <source>Error opening &apos;%1&apos; file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="229"/>
        <source>No multiple alignments found in input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="276"/>
        <source>Source alignment</source>
        <translation type="unfinished">Source alignment</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="278"/>
        <source>Profile name</source>
        <translation type="unfinished">Profile name</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="282"/>
        <source>Options:</source>
        <translation type="unfinished">Options:</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="283"/>
        <source>Model construction strategies</source>
        <translation type="unfinished">Model construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="291"/>
        <source>Relative model construction strategies</source>
        <translation type="unfinished">Relative model construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="293"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation type="unfinished">Gerstein/Sonnhammer/Chothia tree weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="294"/>
        <source>Henikoff simple filter weights</source>
        <translation type="unfinished">Henikoff simple filter weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="295"/>
        <source>Henikoff position-based weights</source>
        <translation type="unfinished">Henikoff position-based weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="296"/>
        <source>No relative weighting; set all to 1</source>
        <translation type="unfinished">No relative weighting; set all to 1</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="297"/>
        <source>Weights given in MSA file</source>
        <translation type="unfinished">Weights given in MSA file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="302"/>
        <source>Effective sequence weighting strategies</source>
        <translation type="unfinished">Effective sequence weighting strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="304"/>
        <source>adjust effective sequence number to achieve relative entropy target</source>
        <translation type="unfinished">Adjust effective sequence number to achieve relative entropy target</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="305"/>
        <source>effective sequence number is number of single linkage clusters</source>
        <translation type="unfinished">Effective sequence number is number of single linkage clusters</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="306"/>
        <source>no effective sequence number weighting: just use number of sequences</source>
        <translation type="unfinished">No effective sequence number weighting: just use number of sequences</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="307"/>
        <source>set effective sequence number for all models to: %1</source>
        <translation type="unfinished">Set effective sequence number for all models to: %1</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="313"/>
        <source>Task finished with error: &apos;%1&apos;</source>
        <translation type="unfinished">Task finished with error: &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3LoadProfileAndSearchTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="514"/>
        <source>HMM search with &apos;%1&apos; HMM profile file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3MSAEditorContext</name>
    <message>
        <source>Build HMMER3 model</source>
        <translation type="obsolete">Build HMMER3 model</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="202"/>
        <source>Build HMMER3 profile</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Phmmer</name>
    <message>
        <source>query_sequence_not_given</source>
        <translation type="obsolete">No input query sequence given</translation>
    </message>
    <message>
        <source>sequence_to_search_in_not_given</source>
        <translation type="obsolete">Database sequence to search in is not given</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_alphabet</source>
        <translation type="obsolete">Run out of memory (creating alphabet failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_null_model</source>
        <translation type="obsolete">Run out of memory (creating null model failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_builder</source>
        <translation type="obsolete">Run out of memory (creating builder failed)</translation>
    </message>
    <message>
        <source>failed_to_set_scoring_system_with_error:%1</source>
        <translation type="obsolete">Setting scoring system failed with error: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_convert_sequence:sequence_to_search_in</source>
        <translation type="obsolete">Error with digitizing sequence to search in</translation>
    </message>
    <message>
        <source>cannot_convert_sequence:query_sequence</source>
        <translation type="obsolete">Error digitizing query sequence</translation>
    </message>
    <message>
        <source>cannot_create_optimized_profile_for_query_sequence</source>
        <translation type="obsolete">Error with creating HMM profile for query sequence</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_top_hits_list</source>
        <translation type="obsolete">Run out of memory (creating top hits list failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_pipeline</source>
        <translation type="obsolete">Run out of memory (creating pipeline failed)</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="168"/>
        <source>No input query sequence given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="172"/>
        <source>Database sequence to search in is not given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="179"/>
        <source>Run out of memory (creating alphabet failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="184"/>
        <source>Run out of memory (creating null model failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="191"/>
        <source>Run out of memory (creating builder failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="198"/>
        <source>Setting scoring system failed with error: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="205"/>
        <source>Error with digitizing sequence to search in</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="212"/>
        <source>Error digitizing query sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="222"/>
        <source>Error with creating HMM profile for query sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="228"/>
        <source>Run out of memory (creating top hits list failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="234"/>
        <source>Run out of memory (creating pipeline failed)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerDialogImpl</name>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="58"/>
        <source>Search</source>
        <translation type="unfinished">Search</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="59"/>
        <source>Cancel</source>
        <translation type="unfinished">Cancel</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="117"/>
        <source>Select query sequence file</source>
        <translation type="unfinished">Select query sequence file</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="165"/>
        <source>Query sequence file path is empty</source>
        <translation type="unfinished">Query sequence file path is empty</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="170"/>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="181"/>
        <source>Error: bad arguments!</source>
        <translation type="unfinished">Error: bad arguments!</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="186"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="186"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerTask</name>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="61"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="79"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="101"/>
        <source>HMM Phmmer task</source>
        <translation type="unfinished">HMM Phmmer task</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="69"/>
        <source>Input query sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="70"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="108"/>
        <source>Database sequence to search in</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="84"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="107"/>
        <source>Query sequence file path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="85"/>
        <source>Database sequence file path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="90"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="113"/>
        <source>Error opening query sequence file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="94"/>
        <source>Error opening database sequence file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="128"/>
        <source>Error loading sequence document:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="131"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="134"/>
        <source>No dna sequence objects found in document</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="138"/>
        <source>Empty sequence loaded from document</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="150"/>
        <source> query sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="154"/>
        <source> db sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>query_sequence</source>
        <translation type="obsolete">Input query sequence</translation>
    </message>
    <message>
        <source>sequence_to_search_in</source>
        <translation type="obsolete">Database sequence to search in</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="72"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="110"/>
        <source>HMM Phmmer search %1 sequence in %2 database</source>
        <translation type="unfinished">HMM Phmmer search %1 sequence in %2 databas</translation>
    </message>
    <message>
        <source>query_sequence_filename</source>
        <translation type="obsolete">Query sequence file path</translation>
    </message>
    <message>
        <source>db_sequence_to_search_in</source>
        <translation type="obsolete">Database sequence file path</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="87"/>
        <source>HMM Phmmer search %1 sequence with %2 database</source>
        <translation type="unfinished">HMM Phmmer search %1 sequence with %2 database</translation>
    </message>
    <message>
        <source>cannot_create_load_query_doc_task</source>
        <translation type="obsolete">Error opening query sequence file</translation>
    </message>
    <message>
        <source>cannot_create_load_db_doc_task</source>
        <translation type="obsolete">Error opening database sequence file</translation>
    </message>
    <message>
        <source>cannot load document from:</source>
        <translation type="obsolete">Error loading sequence document</translation>
    </message>
    <message>
        <source>no_dna_sequence_objects_in_document</source>
        <translation type="obsolete">No dna sequence objects found in document</translation>
    </message>
    <message>
        <source>cannot_cast_to_dna_object</source>
        <translation type="obsolete">No dna sequence objects found in document</translation>
    </message>
    <message>
        <source>empty_sequence_given</source>
        <translation type="obsolete">Empty sequence loaded from document</translation>
    </message>
    <message>
        <source>query sequence</source>
        <translation type="obsolete">Query sequence</translation>
    </message>
    <message>
        <source>db sequence</source>
        <translation type="obsolete">Database sequence</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerToAnnotationsTask</name>
    <message>
        <source>query sequence file path</source>
        <translation type="obsolete">Query sequence file path</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="375"/>
        <source>querySeq sequence file path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="380"/>
        <source>db sequence</source>
        <translation type="unfinished">Database sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="385"/>
        <source>annotation object</source>
        <translation type="unfinished">Annotation object</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="390"/>
        <source>annotation name</source>
        <translation type="unfinished">Annotation name</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="395"/>
        <source>annotation group</source>
        <translation type="unfinished">Annotation group</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="408"/>
        <source>HMM Phmmer task</source>
        <translation type="unfinished">HMM Phmmer task</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="422"/>
        <source>HMM Phmmer search %1 sequence with %2 database</source>
        <translation type="unfinished">Phmmer search %1 sequence with %2 database</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="432"/>
        <source>Annotation object was removed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Annotation object removed</source>
        <translation type="obsolete">Selected annotation object removed</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="456"/>
        <source>Query sequence</source>
        <translation type="unfinished">Query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="459"/>
        <source>Task was not finished</source>
        <translation type="unfinished">Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="464"/>
        <source>Result annotation table</source>
        <translation type="unfinished">Result annotation table</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="465"/>
        <source>Result annotation group</source>
        <translation type="unfinished">Result annotation group</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="466"/>
        <source>Result annotation name</source>
        <translation type="unfinished">Result annotation name</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="469"/>
        <source>Results count</source>
        <translation type="unfinished">Results count</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Plugin</name>
    <message>
        <source>hmm3_plugin_name</source>
        <translation type="obsolete">HMM3</translation>
    </message>
    <message>
        <source>hmm3_plugin_desc</source>
        <translation type="obsolete">HMM profile tools. Plugin is based on HMMER 3.0b3 package: freely distributable implementation of profile HMM software for protein sequence analysis. Home page of project: http://hmmer.janelia.org/</translation>
    </message>
    <message>
        <source>HMM3 tools</source>
        <translation type="obsolete">HMM3 (Hidden Markov models)</translation>
    </message>
    <message>
        <source>build profile</source>
        <translation type="obsolete">Build profile</translation>
    </message>
    <message>
        <source>search HMM signals</source>
        <translation type="obsolete">Search HMM signals</translation>
    </message>
    <message>
        <source>phmmer search</source>
        <translation type="obsolete">Phmmer search</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="70"/>
        <source>HMM3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="70"/>
        <source>HMM profile tools. Plugin is based on HMMER 3.0b3 package: freely distributable implementation of profile HMM software for protein sequence analysis. Home page of project: http://hmmer.janelia.org/</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="100"/>
        <source>Build HMM3 profile...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="105"/>
        <source>Search with HMM3...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="110"/>
        <source>Search with HMM3 phmmer...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="167"/>
        <location filename="../src/uHMM3Plugin.cpp" line="178"/>
        <source>Error!</source>
        <translation type="unfinished">Error!</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="167"/>
        <location filename="../src/uHMM3Plugin.cpp" line="178"/>
        <source>Target sequence not selected: no opened annotated dna view</source>
        <translation type="unfinished">No target sequence!</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3QDActor</name>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="73"/>
        <location filename="../src/search/uhmm3QDActor.cpp" line="167"/>
        <source>HMM3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="84"/>
        <source>QD HMM3 search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="168"/>
        <source>Searches HMM signals in a sequence with one or more profile HMM and saves the results as annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="174"/>
        <source>Profile HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="175"/>
        <source>Semicolon-separated list of input HMM files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="178"/>
        <source>Min Length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="179"/>
        <source>Minimum length of a result region.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="182"/>
        <source>Max Length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="183"/>
        <source>Maximum length of a result region.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="188"/>
        <source>Use E-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="189"/>
        <source>Filters by E-value if true. Otherwise filters by score.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="192"/>
        <source>Filter by High E-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="193"/>
        <source>Reports domains &amp;lt;= this E-value threshold in output.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="197"/>
        <source>Filter by Low Score</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="198"/>
        <source>Reports domains &amp;gt;= this score cutoff in output.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="203"/>
        <source>Max</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="204"/>
        <source>Turns off all acceleration heuristic filters. This increases sensitivity somewhat, at a large cost in speed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="209"/>
        <source>MSV Filter Threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="210"/>
        <source>P-value threshold for the MSV filter step of the acceleration pipeline.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="214"/>
        <source>Viterbi Filter Threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="215"/>
        <source>P-value threshold for the Viterbi filter step of the acceleration pipeline.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="219"/>
        <source>Forward Filter Threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="220"/>
        <source>P-value threshold for the Forward filter step of the acceleration pipeline.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="224"/>
        <source>No Bias Filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="225"/>
        <source>Turns off composition bias filter. This increases sensitivity somewhat, but can come at a high cost in speed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="232"/>
        <source>No Null2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="233"/>
        <source>Turns off the null2 score corrections for biased composition.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="237"/>
        <source>Number of Sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="238"/>
        <source>Specifies number of significant sequences. It is used for domain E-value calculations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="242"/>
        <source>Seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="243"/>
        <source>Random number seed. The default is to use a fixed seed(42), so that results are exactly reproducible. Any other positive integer will give different (but also reproducible) results. A choice of 0 uses a randomly chosen seed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>HMM profile</source>
        <translation type="obsolete">HMM profile</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3RemoteSearchToAnnotationsTask</name>
    <message>
        <source>HMMER3 search task on remote machine</source>
        <translation type="obsolete">HMMER3 search task on remote machine</translation>
    </message>
    <message>
        <source>HMMER3 search task on remote machine. Profile: &apos;%1&apos;, sequence: &apos;%2&apos;</source>
        <translation type="obsolete">HMMER3 search task on remote machine. Profile: &apos;%1&apos;, sequence: &apos;%2&apos;</translation>
    </message>
    <message>
        <source>hmm profile path</source>
        <translation type="obsolete">hmm profile path</translation>
    </message>
    <message>
        <source>sequence to search in</source>
        <translation type="obsolete">sequence to search in</translation>
    </message>
    <message>
        <source>remote machine settings</source>
        <translation type="obsolete">remote machine settings</translation>
    </message>
    <message>
        <source>annotation object</source>
        <translation type="obsolete">Annotation object</translation>
    </message>
    <message>
        <source>annotations group name</source>
        <translation type="obsolete">Annotations group name</translation>
    </message>
    <message>
        <source>annotations name</source>
        <translation type="obsolete">Annotations name</translation>
    </message>
    <message>
        <source>HMM search task runned on remote machine.</source>
        <translation type="obsolete">HMM search task runned on remote machine.</translation>
    </message>
    <message>
        <source>HMM profile used</source>
        <translation type="obsolete">HMM profile used</translation>
    </message>
    <message>
        <source>Sequence</source>
        <translation type="obsolete">Sequence</translation>
    </message>
    <message>
        <source>Task was not finished</source>
        <translation type="obsolete">Task was not finished</translation>
    </message>
    <message>
        <source>Result annotation table</source>
        <translation type="obsolete">Result annotation table</translation>
    </message>
    <message>
        <source>Result annotation group</source>
        <translation type="obsolete">Result annotation group</translation>
    </message>
    <message>
        <source>Result annotation name</source>
        <translation type="obsolete">Result annotation name</translation>
    </message>
    <message>
        <source>Results count</source>
        <translation type="obsolete">Results count</translation>
    </message>
    <message>
        <source>remote task didn&apos;t create results</source>
        <translation type="obsolete">remote task didn&apos;t create results</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWPhmmerTask</name>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="216"/>
        <source>HMM Phmmer search %1 sequence in %2 database</source>
        <translation type="unfinished">HMM Phmmer search %1 sequence in %2 databas</translation>
    </message>
    <message>
        <source>cannot_create_load_query_doc_task</source>
        <translation type="obsolete">Error opening query sequence file</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="222"/>
        <source>Can not create load query doc task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="232"/>
        <source> querySeq sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="245"/>
        <source>Invalid db sequence alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="251"/>
        <source>Invalid query sequence alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="274"/>
        <source>Cannot search for nucleic query in amino sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="302"/>
        <source>HMMER3 phmmer sequence walker search task</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWSearchTask</name>
    <message>
        <source>Sequence_walker_HMM_search_task</source>
        <translation type="obsolete">HMM search task with amino and complement translations</translation>
    </message>
    <message>
        <source>Sequence_walker_hmm_search_with_&apos;%1&apos;</source>
        <translation type="obsolete">HMM search task with amino and complement translations using &apos;%1&apos; profile HMM</translation>
    </message>
    <message>
        <source>sequence_walker_hmmer3_search_task</source>
        <translation type="obsolete">HMM search task with amino and complement translations</translation>
    </message>
    <message>
        <source>unknown_alphabet_type</source>
        <translation type="obsolete">Unrecognized alphabet of sequence</translation>
    </message>
    <message>
        <source>invalid_sequence_alphabet_type</source>
        <translation type="obsolete">Invalid alphabet of sequence</translation>
    </message>
    <message>
        <source>cannot_search_for_nucleic_hmm_in_amino_sequence</source>
        <translation type="obsolete">Cannot search for nucleic HMM profile in amino sequence</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="86"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="106"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="147"/>
        <source>HMM search task with amino and complement translations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="92"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="110"/>
        <source>HMM search task with amino and complement translations using &apos;%1&apos; profile HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="365"/>
        <source>Unrecognized alphabet of sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="369"/>
        <source>Invalid alphabet of sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="375"/>
        <source>Cannot search for nucleic HMM profile in amino sequence</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWSearchToAnnotationsTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="557"/>
        <source>hmm profile filename</source>
        <translation type="unfinished">Profile HMM file path</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="561"/>
        <source>annotation object</source>
        <translation type="unfinished">Annotation object</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="565"/>
        <source>annotations group name</source>
        <translation type="unfinished">Annotations group name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="569"/>
        <source>annotations name</source>
        <translation type="unfinished">Annotations name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="586"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="612"/>
        <source>HMMER3 search task</source>
        <translation type="unfinished">HMM3 search task</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="589"/>
        <source>dna sequence</source>
        <translation type="unfinished">Sequence to search in</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="594"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="620"/>
        <source>HMMER3 search task with &apos;%1&apos; profile</source>
        <translation type="unfinished">HMMER3 search task with &apos;%1&apos; profile</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="615"/>
        <source>Sequence file</source>
        <translation type="unfinished">Sequence file path</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="634"/>
        <source>HMM profile used</source>
        <translation type="unfinished">HMM profile used</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="637"/>
        <source>Task was not finished</source>
        <translation type="unfinished">Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="642"/>
        <source>Result annotation table</source>
        <translation type="unfinished">Result annotation table</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="643"/>
        <source>Result annotation group</source>
        <translation type="unfinished">Result annotation group</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="644"/>
        <source>Result annotation name</source>
        <translation type="unfinished">Result annotation name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="647"/>
        <source>Results count</source>
        <translation type="unfinished">Results count</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="657"/>
        <source>Cannot load sequence document</source>
        <translation type="unfinished">Cannot load sequence document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="662"/>
        <source>No sequence objects loaded</source>
        <translation type="unfinished">No sequence objects found in document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="666"/>
        <source>Unknown sequence type loaded</source>
        <translation type="unfinished">No sequence objects found in document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="670"/>
        <source>Empty sequence loaded</source>
        <translation type="unfinished">Empty sequence loaded</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="686"/>
        <source>Annotation object removed</source>
        <translation type="unfinished">Choosed annotation object removed</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Search</name>
    <message>
        <source>no_hmm_given</source>
        <translation type="obsolete">Bad HMM profile given</translation>
    </message>
    <message>
        <source>no_sequence_given</source>
        <translation type="obsolete">Empty sequence given</translation>
    </message>
    <message>
        <source>no_memory:cannot_convert_sequence</source>
        <translation type="obsolete">Run out of memory (creation of sequence failed)</translation>
    </message>
    <message>
        <source>sequence_and_hmm_alphabets_not_match</source>
        <translation type="obsolete">Profile HMM and sequence alphabets no matched</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_alphabet</source>
        <translation type="obsolete">Run out of memory (creation of alphabet failed)</translation>
    </message>
    <message>
        <source>cannot_digitize_sequence</source>
        <translation type="obsolete">Run out of memory (digitizing of sequence failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_null_model</source>
        <translation type="obsolete">Run out of memory (creation of null model failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_profile</source>
        <translation type="obsolete">Run out of memory</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_optimized_profile</source>
        <translation type="obsolete">Run out of memory (creation of optimized profile failed)</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_pipeline</source>
        <translation type="obsolete">Run out of memory</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_tophits</source>
        <translation type="obsolete">Run out of memory (top hits list creation failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="61"/>
        <source>Bad HMM profile given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="65"/>
        <source>Empty sequence given</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="73"/>
        <source>Run out of memory (creation of sequence failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="79"/>
        <source>Profile HMM and sequence alphabets no matched</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="85"/>
        <source>Run out of memory (creation of alphabet failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="90"/>
        <source>Run out of memory (digitizing of sequence failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="96"/>
        <source>Run out of memory (creation of null model failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="102"/>
        <location filename="../src/search/uhmm3search.cpp" line="116"/>
        <source>Run out of memory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="107"/>
        <source>Run out of memory (creation of optimized profile failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="121"/>
        <source>Run out of memory (top hits list creation failed)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SearchDialogImpl</name>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="58"/>
        <source>Run</source>
        <translation type="unfinished">Run</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="59"/>
        <source>Cancel</source>
        <translation type="unfinished">Cancel</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="160"/>
        <source>HMM profile file path is empty</source>
        <translation type="unfinished">HMM profile file path is empty</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="175"/>
        <source>Error: bad arguments!</source>
        <translation type="unfinished">Error: bad arguments!</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="182"/>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="186"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="186"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Selecting machines error!</source>
        <translation type="obsolete">Selecting machines error!</translation>
    </message>
    <message>
        <source>You didn&apos;t select a machine to run remote task!</source>
        <translation type="obsolete">You didn&apos;t select a machine to run remote task!</translation>
    </message>
    <message>
        <source>Distributed run on many machines is not supported yet. Select 1 machine</source>
        <translation type="obsolete">Distributed run on many machines is not supported yet. Select 1 machine</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="236"/>
        <source>Select query HMM profile</source>
        <translation type="unfinished">Select query HMM profile</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SearchLocalTask</name>
    <message>
        <source>HMM3 search local task</source>
        <translation type="obsolete">HMM3 search local task</translation>
    </message>
    <message>
        <source>No settings given</source>
        <translation type="obsolete">No settings given</translation>
    </message>
    <message>
        <source>HMM3 search local task with amino and complement translations. Sequence: &apos;%1&apos;. HMM profile:&apos;%2&apos;</source>
        <translation type="obsolete">HMM3 search local task with amino and complement translations. Sequence: &apos;%1&apos;. HMM profile:&apos;%2&apos;</translation>
    </message>
    <message>
        <source>Alien HMM search task from remote machine.</source>
        <translation type="obsolete">Alien HMM search task from remote machine.</translation>
    </message>
    <message>
        <source>HMM profile</source>
        <translation type="obsolete">HMM profile</translation>
    </message>
    <message>
        <source>Sequence</source>
        <translation type="obsolete">Sequence</translation>
    </message>
    <message>
        <source>Task finished with error</source>
        <translation type="obsolete">Task finished with error</translation>
    </message>
    <message>
        <source>Results count</source>
        <translation type="obsolete">Results count</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SearchTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="469"/>
        <source>HMM search task</source>
        <translation type="unfinished">HMM search task</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="478"/>
        <source>HMM search with %1 profiles</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>no_hmm_given</source>
        <translation type="obsolete">HMM profile not given</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="476"/>
        <source>HMM search with &apos;%1&apos;</source>
        <translation type="unfinished">HMM search with &apos;%1&apos;</translation>
    </message>
    <message>
        <source>empty_sequence_given</source>
        <translation type="obsolete">Sequence to search in not given</translation>
    </message>
    <message>
        <source>no_hmm_file_given</source>
        <translation type="obsolete">HMM profile file not given</translation>
    </message>
    <message>
        <source>HMM search with &apos;%1&apos; file</source>
        <translation type="obsolete">HMM search with &apos;%1&apos; file</translation>
    </message>
</context>
<context>
    <name>U2::UHMMFormat</name>
    <message>
        <source>write_to_file_failed</source>
        <translation type="obsolete">Writing HMM profile file failed</translation>
    </message>
    <message>
        <source>unknown_error_occurred</source>
        <translation type="obsolete">Unknown error occurred</translation>
    </message>
    <message>
        <source>hmm_files_are_read_only</source>
        <translation type="obsolete">HMM files are read only</translation>
    </message>
    <message>
        <source>hmmer_format</source>
        <translation type="obsolete">Profile HMM format</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="279"/>
        <source>Unknown error occurred</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="305"/>
        <source>Profile HMM format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="306"/>
        <source>hmm is a format for storing hmm profiles</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>cannot_create_io_adapter</source>
        <translation type="obsolete">Internal error (cannot create ioadapter)</translation>
    </message>
</context>
<context>
    <name>U2::UHMMFormatReader</name>
    <message>
        <source>bad_line_in_header_section:%1</source>
        <translation type="obsolete">Bad line in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_value_found_in_line:%1</source>
        <translation type="obsolete">Expected value after tag in: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="117"/>
        <source>Number expected in NULE line. %1 found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="140"/>
        <source>Bad line in header section: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="147"/>
        <source>Empty value in header line:%1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>cannot_parse_integer_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_float_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse float number from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_2_float_numbers_in_str:%1</source>
        <translation type="obsolete">Cannot parse 2 float numbers from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_uinteger_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse unsigned integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_yes/no_value_from_string:%1</source>
        <translation type="obsolete">Cannot parse y/n value from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_memory:cannot_allocate_hmm_date</source>
        <translation type="obsolete">Run out of memory (date allocation failed)</translation>
    </message>
    <message>
        <source>bad_stats_line:%1</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.&quot;LOCAL&quot; word is not found</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.%2 is not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>bad_stats_line:%1. %2 not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>unexpected_end_of_line</source>
        <translation type="obsolete">Unexpected end of file</translation>
    </message>
    <message>
        <source>HMM reader error occurred: </source>
        <translation type="obsolete">HMM reader error:</translation>
    </message>
    <message>
        <source>reading file failed</source>
        <translation type="obsolete">Reading failed</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="387"/>
        <source>Input file made by unknown version of HMMER or is not HMM profile file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>no_memory:cannot_create_hmm_shell</source>
        <translation type="obsolete">Run out of memory (allocation of HMM shell failed)</translation>
    </message>
    <message>
        <source>length_of_a_model_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Length of HMM model should be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_alphabet_type:%1</source>
        <translation type="obsolete">Unrecognized alphabet type: %1</translation>
    </message>
    <message>
        <source>nseq_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Number of sequences should be positive. Found: %1</translation>
    </message>
    <message>
        <source>effn_should_be_positive.we_have:%1</source>
        <translation type="obsolete">EFFN shoold be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_tag_in_header_section:%1</source>
        <translation type="obsolete">Unrecognized tag in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>missing_one_or_more_STATS_parameter_lines</source>
        <translation type="obsolete">Missing one or more STATS parameters</translation>
    </message>
    <message>
        <source>failed_to_create_alphabet</source>
        <translation type="obsolete">Run out of memory (failed to create alphabet)</translation>
    </message>
    <message>
        <source>failed_to_allocate_body_of_hmm</source>
        <translation type="obsolete">Run out of memory (allocation of HMM body failed)</translation>
    </message>
    <message>
        <source>expected_line_to_start_with:%1.it_starts_with:%2</source>
        <translation type="obsolete">Line was excpected to start with %1. Found: %2</translation>
    </message>
    <message>
        <source>expected_closing_//.found_%1_instead</source>
        <translation type="obsolete">Excpected to find closing &apos;//&apos;. Found %1 instead</translation>
    </message>
    <message>
        <source>no_name_found_in_hmm</source>
        <translation type="obsolete">HMM name not found</translation>
    </message>
    <message>
        <source>allocation failure, HMM shell</source>
        <translation type="obsolete">Failed to allocate HMM shell</translation>
    </message>
    <message>
        <source>unrecognized_alphabet_type:%1
</source>
        <translation type="obsolete">Unrecognized alphabet type: %1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="750"/>
        <source>ALPH must precede NULE in HMMER2 save files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="769"/>
        <source>ALPH section must precede HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="774"/>
        <source>Unknown alphabet</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="797"/>
        <source>Failed to allocate body of the new HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="157"/>
        <source>Cannot parse integer from string: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="168"/>
        <source>Cannot parse float number from string: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="184"/>
        <source>Cannot parse 2 float numbers from string: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="195"/>
        <source>Cannot parse unsigned integer from string: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="207"/>
        <source>Cannot parse y/n value from string: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="216"/>
        <source>Run out of memory (date allocation failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="231"/>
        <source>Bad STATS line: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="235"/>
        <source>Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="263"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="286"/>
        <source>Bad STATS line: %1. %2 not recognized</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="299"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="409"/>
        <source>Run out of memory (allocation of HMM shell failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="439"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="697"/>
        <source>Length of HMM model should be positive. Found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="447"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="707"/>
        <source>Unrecognized alphabet type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="470"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="731"/>
        <source>Number of sequences should be positive. Found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="477"/>
        <source>EFFN shoold be positive. Found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="512"/>
        <source>Unrecognized tag in header section: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="523"/>
        <source>Missing one or more STATS parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="528"/>
        <source>Run out of memory (failed to create alphabet)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="533"/>
        <source>Run out of memory (allocation of HMM body failed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="574"/>
        <source>Line was excpected to start with %1. Found: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="616"/>
        <source>Excpected to find closing &apos;//&apos;. Found %1 instead</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="633"/>
        <source>HMM name not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="659"/>
        <source>Allocation failure, HMM shell</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="800"/>
        <source>Failed to create background model</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="830"/>
        <source>Expected match line to start with %1. saw %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="884"/>
        <source>Expected closing //. found %1 instead</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="897"/>
        <source>No NAME found for HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="900"/>
        <source>No LENG found for HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="903"/>
        <source>No ALPH found for HMM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="909"/>
        <source>Failed to calibrate HMMER2 model after input conversion</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>UHMM3BuildDialog</name>
    <message>
        <source>Dialog</source>
        <translation type="obsolete">HMM3 build</translation>
    </message>
    <message>
        <source>input and output</source>
        <translation type="obsolete">Input and output</translation>
    </message>
    <message>
        <source>Alignment file:</source>
        <translation type="obsolete">Input alignment file</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="59"/>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="76"/>
        <source>...</source>
        <translation type="unfinished">...</translation>
    </message>
    <message>
        <source>Build to profile:</source>
        <translation type="obsolete">Build to profile</translation>
    </message>
    <message>
        <source>Build</source>
        <translation type="obsolete">Build</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Cancel</translation>
    </message>
    <message>
        <source>model construction strategies</source>
        <translation type="obsolete">Construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="123"/>
        <source>assign cols w/ &gt;= symfrac residues as consensus</source>
        <translation type="unfinished">assign cols w/ &gt;= symfrac residues as consensus</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="126"/>
        <source>fast</source>
        <translation type="unfinished">Fast</translation>
    </message>
    <message>
        <source>sym fraction:</source>
        <translation type="obsolete">Sym fraction</translation>
    </message>
    <message>
        <source>manual construction</source>
        <translation type="obsolete">Manual construction</translation>
    </message>
    <message>
        <source>hand (requires reference annotation)</source>
        <translation type="obsolete">Hand (requires reference annotation)</translation>
    </message>
    <message>
        <source>tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</source>
        <translation type="obsolete">Tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</translation>
    </message>
    <message>
        <source>relative sequence weighting strategies</source>
        <translation type="obsolete">Relative weighting</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="20"/>
        <source>Hmm3 Build</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="39"/>
        <source>Input and output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="49"/>
        <source>Input alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="66"/>
        <source>Build to profile</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="101"/>
        <source>Construction strategies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="139"/>
        <source>Sym fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="161"/>
        <source>Manual construction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="164"/>
        <source>Hand (requires reference annotation)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="182"/>
        <source>Tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="223"/>
        <source>Relative weighting</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="231"/>
        <source>Henikoff position-based weights</source>
        <translation type="unfinished">Henikoff position-based weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="247"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation type="unfinished">Gerstein/Sonnhammer/Chothia tree weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="271"/>
        <source>Henikoff simple filter weights</source>
        <translation type="unfinished">Henikoff simple filter weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="281"/>
        <source>Identity cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="306"/>
        <source>Skip relative weighting; set all to 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="313"/>
        <source>Use weights given in alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="336"/>
        <source>Effective weighting</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="344"/>
        <source>Adjust effective sequence number to achieve relative entropy target</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="378"/>
        <source>Minimum relative entropy/position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="430"/>
        <source>Sigma parameter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="463"/>
        <source>Use number of single linkage clusters as effective</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="473"/>
        <source>Fractional identity cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="498"/>
        <source>Use number of sequences as effective</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="513"/>
        <source>Effective sequence number for all models to</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="551"/>
        <source>E-value calibration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="570"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="606"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="639"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="669"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="699"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="729"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="759"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="800"/>
        <source>Other</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="816"/>
        <source>Random generator seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="859"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>identity cutoff:</source>
        <translation type="obsolete">Identity cutoff</translation>
    </message>
    <message>
        <source>don&apos;t do any relative weighting; set all to 1</source>
        <translation type="obsolete">Skip relative weighting; set all to 1</translation>
    </message>
    <message>
        <source>use weights as given in MSA file</source>
        <translation type="obsolete">Use weights given in alignment file</translation>
    </message>
    <message>
        <source>effective sequence weighting strategies</source>
        <translation type="obsolete">Effective weighting</translation>
    </message>
    <message>
        <source>adjust effective sequence number to achieve relative entropy target</source>
        <translation type="obsolete">Adjust effective sequence number to achieve relative entropy target</translation>
    </message>
    <message>
        <source>set minimum relative entropy/position to:</source>
        <translation type="obsolete">Minimum relative entropy/position</translation>
    </message>
    <message>
        <source>set sigma parameter to:</source>
        <translation type="obsolete">Sigma parameter</translation>
    </message>
    <message>
        <source>effective sequence number is number of single linkage clusters</source>
        <translation type="obsolete">Use number of single linkage clusters as effective</translation>
    </message>
    <message>
        <source>fractional identity cutoff:</source>
        <translation type="obsolete">Fractional identity cutoff</translation>
    </message>
    <message>
        <source>no effective sequence number weighting: just use number of sequences</source>
        <translation type="obsolete">Use number of sequences as effective</translation>
    </message>
    <message>
        <source>set effective sequence number for all models to:</source>
        <translation type="obsolete">Effective sequence number for all models to</translation>
    </message>
    <message>
        <source>e-value calibration</source>
        <translation type="obsolete">E-value calibration</translation>
    </message>
    <message>
        <source>length of sequences for MSV Gumbel mu fit:</source>
        <translation type="obsolete">Length of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <source>number of sequences for MSV Gumbel mu fit:</source>
        <translation type="obsolete">Number of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <source>length of sequences for Viterbi Gumbel mu fit:</source>
        <translation type="obsolete">Length of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <source>number of sequences for Viterbi Gumbel mu fit:</source>
        <translation type="obsolete">Number of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <source>length of sequences for Forward exp tail mu fit:</source>
        <translation type="obsolete">Length of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <source>number of sequences for Forward exp tail mu fit:</source>
        <translation type="obsolete">Number of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <source>tail mass for Forward exponential tail mu fit:</source>
        <translation type="obsolete">Tail mass for Forward exponential tail mu fit</translation>
    </message>
    <message>
        <source>other</source>
        <translation type="obsolete">Other</translation>
    </message>
    <message>
        <source>random generator seed:</source>
        <translation type="obsolete">Random generator seed</translation>
    </message>
    <message>
        <source>* if 0: one-time arbitrary seed used</source>
        <translation type="obsolete">* 0 means that one-time arbitrary seed will be used</translation>
    </message>
</context>
<context>
    <name>UHMM3PhmmerDialog</name>
    <message>
        <source>Dialog</source>
        <translation type="obsolete">Phmmer search</translation>
    </message>
    <message>
        <source>input and output</source>
        <translation type="obsolete">Input and output</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="20"/>
        <source>Phmmer Search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="36"/>
        <source>Input and output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="50"/>
        <source>Query sequence file:</source>
        <translation type="unfinished">Query sequence file:</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="60"/>
        <source>...</source>
        <translation type="unfinished">...</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="87"/>
        <source>Reporting tresholds</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="103"/>
        <source>Report domains with E-value less than</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="142"/>
        <source>Report domains with score greater than</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="166"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="201"/>
        <source>Scoring system</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="217"/>
        <source>Gap open probability</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="247"/>
        <source>Gap extend probability</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="285"/>
        <source>Acceleration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="297"/>
        <source>Turn off composition bias filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="310"/>
        <source>Turn off biased composition score corrections</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="338"/>
        <source>Viterbi filter treshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="368"/>
        <source>MSV filter treshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="401"/>
        <source>Forward filter treshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="437"/>
        <source>E-value calibration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="456"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="492"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="525"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="555"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="585"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="615"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="645"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="686"/>
        <source>Other</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="702"/>
        <source>Random generator seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="745"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>search</source>
        <translation type="obsolete">Search</translation>
    </message>
    <message>
        <source>cancel</source>
        <translation type="obsolete">Cancel</translation>
    </message>
    <message>
        <source>reporting tresholds</source>
        <translation type="obsolete">Reporting tresholds</translation>
    </message>
    <message>
        <source>report domains &lt;= this E-value threshold:</source>
        <translation type="obsolete">Report domains with E-value less than</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="113"/>
        <source>1E+</source>
        <translation type="unfinished">1E+</translation>
    </message>
    <message>
        <source>report domains &gt;= this score treshold:</source>
        <translation type="obsolete">Report domains with score greater than</translation>
    </message>
    <message>
        <source> set number of significant seqs, for domain E-value calculation:</source>
        <translation type="obsolete">Number of significant sequences for domain E-value calculation</translation>
    </message>
    <message>
        <source>scoring system</source>
        <translation type="obsolete">Scoring system</translation>
    </message>
    <message>
        <source>gap open probability:</source>
        <translation type="obsolete">Gap open probability</translation>
    </message>
    <message>
        <source>gap extend probability:</source>
        <translation type="obsolete">Gap extend probability</translation>
    </message>
    <message>
        <source>acceleration heuristics</source>
        <translation type="obsolete">Acceleration</translation>
    </message>
    <message>
        <source> turn off composition bias filter</source>
        <translation type="obsolete">Turn off composition bias filter</translation>
    </message>
    <message>
        <source>turn off biased composition score corrections</source>
        <translation type="obsolete">Turn off biased composition score corrections</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="323"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation type="unfinished">Turn all heuristic filters off (less speed, more power)</translation>
    </message>
    <message>
        <source>Viterbi filter treshold:</source>
        <translation type="obsolete">Viterbi filter treshold</translation>
    </message>
    <message>
        <source>MSV filter treshold:</source>
        <translation type="obsolete">MSV filter treshold</translation>
    </message>
    <message>
        <source>Forward filter treshold:</source>
        <translation type="obsolete">Forward filter treshold</translation>
    </message>
    <message>
        <source>e-value calibration</source>
        <translation type="obsolete">E-value calibration</translation>
    </message>
    <message>
        <source>length of sequences for MSV Gumbel mu fit:</source>
        <translation type="obsolete">Length of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <source>number of sequences for MSV Gumbel mu fit:</source>
        <translation type="obsolete">Number of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <source>length of sequences for Viterbi Gumbel mu fit:</source>
        <translation type="obsolete">Length of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <source>number of sequences for Viterbi Gumbel mu fit:</source>
        <translation type="obsolete">Number of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <source>length of sequences for Forward exp tail mu fit:</source>
        <translation type="obsolete">Length of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <source>number of sequences for Forward exp tail mu fit:</source>
        <translation type="obsolete">Number of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <source>tail mass for Forward exponential tail mu fit:</source>
        <translation type="obsolete">Tail mass for Forward exponential tail mu fit</translation>
    </message>
    <message>
        <source>other</source>
        <translation type="obsolete">Other</translation>
    </message>
    <message>
        <source>random generator seed:</source>
        <translation type="obsolete">Random generator seed</translation>
    </message>
    <message>
        <source>* if 0: one-time arbitrary seed used</source>
        <translation type="obsolete">* 0 means that one-time arbitrary seed will be used</translation>
    </message>
</context>
<context>
    <name>UHMM3SWSearchTask</name>
    <message>
        <source>no_hmm_found_in_file</source>
        <translation type="obsolete">HMM profile not found in doucment</translation>
    </message>
    <message>
        <source>cannot_cast_to_hmm_object</source>
        <translation type="obsolete">HMM profile cannot be cast to HMM object</translation>
    </message>
</context>
<context>
    <name>UHMM3SearchDialog</name>
    <message>
        <source>Dialog</source>
        <translation type="obsolete">HMM3 search</translation>
    </message>
    <message>
        <source>input and output</source>
        <translation type="obsolete">Input and output</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="65"/>
        <source>Query HMM file:</source>
        <translation type="unfinished">Query profile HMM file:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="75"/>
        <source>...</source>
        <translation type="unfinished">...</translation>
    </message>
    <message>
        <source>Search</source>
        <translation type="obsolete">Search</translation>
    </message>
    <message>
        <source>Remote run...</source>
        <translation type="obsolete">Remote run...</translation>
    </message>
    <message>
        <source>Run</source>
        <translation type="obsolete">Run</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Cancel</translation>
    </message>
    <message>
        <source>reporting tresholds</source>
        <translation type="obsolete">Reporting tresholds</translation>
    </message>
    <message>
        <source>report domains &lt;= this E-value threshold:</source>
        <translation type="obsolete">Report domains with E-value less than</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="119"/>
        <source>1E+</source>
        <translation type="unfinished">1E+</translation>
    </message>
    <message>
        <source>report domains &gt;= this score treshold:</source>
        <translatorcomment>Report domains with score greater than</translatorcomment>
        <translation type="obsolete">Report domains with score greater than</translation>
    </message>
    <message>
        <source>score treshold:</source>
        <translation type="obsolete">Score treshold</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="54"/>
        <source>Input and output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="92"/>
        <source>Reporting thresholds</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="109"/>
        <source>Report domains with E-value less than</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="146"/>
        <source>Report domains with score greater than</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>use profile&apos;s GA gathering cutoffs to set score treshold</source>
        <translation type="obsolete">Use profile&apos;s GA gathering cutoffs</translation>
    </message>
    <message>
        <source>use profile&apos;s NC noise cutoffs to set score treshold</source>
        <translation type="obsolete">use profile&apos;s NC noise cutoffs</translation>
    </message>
    <message>
        <source>use profile&apos;s TC trusted cutoffs to set score treshold</source>
        <translation type="obsolete">use profile&apos;s TC trusted cutoffs</translation>
    </message>
    <message>
        <source> set number of significant seqs, for domain E-value calculation:</source>
        <translation type="obsolete">Number of significant sequences for domain E-value calculation</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="20"/>
        <source>HMM3 Search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="180"/>
        <source>Score threshold:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="229"/>
        <source>Use profile&apos;s GA gathering cutoffs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="265"/>
        <source>Use profile&apos;s NC noise cutoffs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="295"/>
        <source>Use profile&apos;s TC trusted cutoffs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="312"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="345"/>
        <source>Acceleration heuristics</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="360"/>
        <source>Turn off composition bias filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="373"/>
        <source>Turn off biased composition score corrections</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="401"/>
        <source>MSV filter threshold:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="434"/>
        <source>Viterbi filter threshold:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="464"/>
        <source>Forward filter threshold:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="500"/>
        <source>Other</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="522"/>
        <source>Random generator seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="565"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>acceleration heuristics</source>
        <translation type="obsolete">Acceleration</translation>
    </message>
    <message>
        <source> turn off composition bias filter</source>
        <translation type="obsolete">Turn off composition bias filter</translation>
    </message>
    <message>
        <source>turn off biased composition score corrections</source>
        <translation type="obsolete">Turn off biased composition score corrections</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="386"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation type="unfinished">Turn all heuristic filters off (less speed, more power)</translation>
    </message>
    <message>
        <source>MSV filter treshold:</source>
        <translation type="obsolete">MSV filter treshold</translation>
    </message>
    <message>
        <source>Viterbi filter treshold:</source>
        <translation type="obsolete">Viterbi filter treshold</translation>
    </message>
    <message>
        <source>Forward filter treshold:</source>
        <translation type="obsolete">Forward filter treshold</translation>
    </message>
    <message>
        <source>other</source>
        <translation type="obsolete">Other</translation>
    </message>
    <message>
        <source>random generator seed:</source>
        <translation type="obsolete">Random generator seed</translation>
    </message>
    <message>
        <source>* if 0: one-time arbitrary seed used</source>
        <translation type="obsolete">* 0 means that one-time arbitrary seed will be used</translation>
    </message>
</context>
<context>
    <name>UHMMFormat</name>
    <message>
        <source>reading_file_failed!</source>
        <translation type="obsolete">Reading HMM profile file failed!</translation>
    </message>
    <message>
        <source>write_to_file_failed</source>
        <translation type="obsolete">Writing HMM profile file failed</translation>
    </message>
    <message>
        <source>bad_line_in_header_section:%1</source>
        <translation type="obsolete">Bad line in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_value_found_in_line:%1</source>
        <translation type="obsolete">Expected value after tag in: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_integer_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_float_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse float number from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_2_float_numbers_in_str:%1</source>
        <translation type="obsolete">Cannot parse 2 float numbers from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_uinteger_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse unsigned integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_yes/no_value_from_string:%1</source>
        <translation type="obsolete">Cannot parse y/n value from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_memory:cannot_allocate_hmm_date</source>
        <translation type="obsolete">Run out of memory (date allocation failed)</translation>
    </message>
    <message>
        <source>bad_stats_line:%1</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.&quot;LOCAL&quot; word is not found</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.%2 is not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>bad_stats_line:%1. %2 not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>unexpected_end_of_line</source>
        <translation type="obsolete">Unexpected end of file</translation>
    </message>
    <message>
        <source>invalid_header_line</source>
        <translation type="obsolete">Invalid header line</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_hmm_shell</source>
        <translation type="obsolete">Run out of memory (allocation of HMM shell failed)</translation>
    </message>
    <message>
        <source>length_of_a_model_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Length of HMM model should be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_alphabet_type:%1</source>
        <translation type="obsolete">Unrecognized alphabet type: %1</translation>
    </message>
    <message>
        <source>nseq_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Number of sequences should be positive. Found: %1</translation>
    </message>
    <message>
        <source>effn_should_be_positive.we_have:%1</source>
        <translation type="obsolete">EFFN shoold be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_tag_in_header_section:%1</source>
        <translation type="obsolete">Unrecognized tag in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>missing_one_or_more_STATS_parameter_lines</source>
        <translation type="obsolete">Missing one or more STATS parameters</translation>
    </message>
    <message>
        <source>failed_to_create_alphabet</source>
        <translation type="obsolete">Run out of memory (failed to create alphabet)</translation>
    </message>
    <message>
        <source>failed_to_allocate_body_of_hmm</source>
        <translation type="obsolete">Run out of memory (allocation of HMM body failed)</translation>
    </message>
    <message>
        <source>expected_line_to_start_with:%1.it_starts_with:%2</source>
        <translation type="obsolete">Line was excpected to start with %1. Found: %2</translation>
    </message>
    <message>
        <source>expected_closing_//.found_%1_instead</source>
        <translation type="obsolete">Excpected to find closing &apos;//&apos;. Found %1 instead</translation>
    </message>
    <message>
        <source>no_name_found_in_hmm</source>
        <translation type="obsolete">HMM name not found</translation>
    </message>
    <message>
        <source>unknown_error_occurred</source>
        <translation type="obsolete">Unknown error occurred</translation>
    </message>
    <message>
        <source>hmm_files_are_read_only</source>
        <translation type="obsolete">HMM files are read only</translation>
    </message>
</context>
<context>
    <name>UHMMFormatReader</name>
    <message>
        <source>bad_line_in_header_section:%1</source>
        <translation type="obsolete">Bad line in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_value_found_in_line:%1</source>
        <translation type="obsolete">Expected value after tag in: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_integer_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_float_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse float number from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_2_float_numbers_in_str:%1</source>
        <translation type="obsolete">Cannot parse 2 float numbers from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_uinteger_number_from_string:%1</source>
        <translation type="obsolete">Cannot parse unsigned integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>cannot_parse_yes/no_value_from_string:%1</source>
        <translation type="obsolete">Cannot parse y/n value from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>no_memory:cannot_allocate_hmm_date</source>
        <translation type="obsolete">Run out of memory (date allocation failed)</translation>
    </message>
    <message>
        <source>bad_stats_line:%1</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.&quot;LOCAL&quot; word is not found</source>
        <translation type="obsolete">Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</translation>
    </message>
    <message>
        <source>bad_stats_line:%1.%2 is not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>bad_stats_line:%1. %2 not recognized</source>
        <translation type="obsolete">Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <source>unexpected_end_of_line</source>
        <translation type="obsolete">Unexpected end of file</translation>
    </message>
    <message>
        <source>HMM reader error occurred: </source>
        <translation type="obsolete">HMM reader error:</translation>
    </message>
    <message>
        <source>reading file failed</source>
        <translation type="obsolete">Reading failed</translation>
    </message>
    <message>
        <source>no_memory:cannot_create_hmm_shell</source>
        <translation type="obsolete">Run out of memory (allocation of HMM shell failed)</translation>
    </message>
    <message>
        <source>length_of_a_model_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Length of HMM model should be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_alphabet_type:%1</source>
        <translation type="obsolete">Unrecognized alphabet type: %1</translation>
    </message>
    <message>
        <source>nseq_should_be_positive.we_have:%1</source>
        <translation type="obsolete">Number of sequences should be positive. Found: %1</translation>
    </message>
    <message>
        <source>effn_should_be_positive.we_have:%1</source>
        <translation type="obsolete">EFFN shoold be positive. Found: %1</translation>
    </message>
    <message>
        <source>unrecognized_tag_in_header_section:%1</source>
        <translation type="obsolete">Unrecognized tag in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <source>missing_one_or_more_STATS_parameter_lines</source>
        <translation type="obsolete">Missing one or more STATS parameters</translation>
    </message>
    <message>
        <source>failed_to_create_alphabet</source>
        <translation type="obsolete">Run out of memory (failed to create alphabet)</translation>
    </message>
    <message>
        <source>failed_to_allocate_body_of_hmm</source>
        <translation type="obsolete">Run out of memory (allocation of HMM body failed)</translation>
    </message>
    <message>
        <source>expected_line_to_start_with:%1.it_starts_with:%2</source>
        <translation type="obsolete">Line was excpected to start with %1. Found: %2</translation>
    </message>
    <message>
        <source>expected_closing_//.found_%1_instead</source>
        <translation type="obsolete">Excpected to find closing &apos;//&apos;. Found %1 instead</translation>
    </message>
    <message>
        <source>no_name_found_in_hmm</source>
        <translation type="obsolete">HMM name not found</translation>
    </message>
</context>
</TS>
