<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en" sourcelanguage="en">
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="350"/>
        <source>HMM reader error occurred: </source>
        <translation>HMM reader error:</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="351"/>
        <source>Reading file failed</source>
        <translation>Reading file failed</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="78"/>
        <source>HMM3 Profile</source>
        <translation>HMM3 Profile</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="41"/>
        <source>Writing HMM profile file failed</source>
        <translation>Writing HMM profile file failed</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="302"/>
        <source>HMM files are read only</source>
        <translation>HMM files are read only</translation>
    </message>
</context>
<context>
    <name>U2::GTest_CompareHmmFiles</name>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="304"/>
        <source>File #1 not set</source>
        <translation>File #1 not set</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="310"/>
        <source>File #2 not set</source>
        <translation>File #2 not set</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="370"/>
        <source>Error creating ioadapter for first file</source>
        <translation>Error creating ioadapter for first file</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="374"/>
        <source>Error opening 1 file</source>
        <translation>Error opening 1 file</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="381"/>
        <source>Error creating ioadapter for second file</source>
        <translation>Error creating ioadapter for second file</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="385"/>
        <source>Error opening second file</source>
        <translation>Error opening second file</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="410"/>
        <source>Names of aligments not matched</source>
        <translation>Names of aligments not matched</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="414"/>
        <source>Comparing files length not matched</source>
        <translation>Comparing files length not matched</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="421"/>
        <source>Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</source>
        <translation>Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMM3Search</name>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="247"/>
        <source>Sequence is empty</source>
        <translation>Sequence is empty</translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMM3SearchCompare</name>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="374"/>
        <source>Internal error (cannot parse float number from string &apos;%1&apos;)</source>
        <translation>Internal error (cannot parse float number from string &apos;%1&apos;)</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="387"/>
        <source>Can&apos;t parse significance:%1</source>
        <translation>Can&apos;t parse significance:%1</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3SearchTests.cpp" line="602"/>
        <source>No search task in test context</source>
        <translation>No search task in test context</translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMMER3Build</name>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="228"/>
        <source>No input file given</source>
        <translation>No input file given</translation>
    </message>
    <message>
        <location filename="../src/tests/uhmmer3BuildTests.cpp" line="234"/>
        <source>No output file given</source>
        <translation>No output file given</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3BuildPrompter</name>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="109"/>
        <source>For each MSA from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>For each MSA from &lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="111"/>
        <source>%1 builds a HMM3 profile.</source>
        <translation>%1 builds a HMM3 profile.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3BuildWorker</name>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="61"/>
        <source>Input MSA</source>
        <translation>Input MSA</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="62"/>
        <source>Input multiple sequence alignment for building statistical model.</source>
        <translation>Input multiple sequence alignment for building statistical model.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="63"/>
        <source>HMM3 profile</source>
        <translation>HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="63"/>
        <source>Produced HMM3 profile</source>
        <translation>Produced HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="73"/>
        <source>Random seed</source>
        <translation>Random seed</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="73"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation>Random generator seed. 0 - means that one-time arbitrary seed will be used.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="77"/>
        <source>HMM3 Build</source>
        <translation>HMM3 Build</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="77"/>
        <source>Builds a HMM3 profile from a multiple sequence alignment.&lt;p&gt;The HMM3 profile is a statistical model which captures position-specific information about how conserved each column of the alignment is, and which residues are likely.</source>
        <translation>Builds a HMM3 profile from a multiple sequence alignment.&lt;p&gt;The HMM3 profile is a statistical model which captures position-specific information about how conserved each column of the alignment is, and which residues are likely.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3BuildWorker.cpp" line="182"/>
        <source>Built HMM3 profile</source>
        <translation>Built HMM3 profile</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Lib</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="72"/>
        <source>HMM3 Profile</source>
        <translation>HMM3 Profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="80"/>
        <source>HMMER3 Tools</source>
        <translation>HMMER3 Tools</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="142"/>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="159"/>
        <source>HMM3 profile</source>
        <translation>HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="142"/>
        <source>Input HMM3 profile</source>
        <translation>Input HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="143"/>
        <source>Location</source>
        <translation>Location</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="143"/>
        <source>Location hint for the target file.</source>
        <translation>Location hint for the target file.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="153"/>
        <source>Write HMM3 Profile</source>
        <translation>Write HMM3 Profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="153"/>
        <source>Saves all input HMM3 profiles to specified location.</source>
        <translation>Saves all input HMM3 profiles to specified location.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="159"/>
        <source>Loaded HMM3 profile</source>
        <translation>Loaded HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="166"/>
        <source>Read HMM3 Profile</source>
        <translation>Read HMM3 Profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="166"/>
        <source>Reads HMM3 profiles from file(s). The files can be local or Internet URLs.</source>
        <translation>Reads HMM3 profiles from file(s). The files can be local or Internet URLs.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3ReadPrompter</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="178"/>
        <source>Read HMM3 profile(s) from %1.</source>
        <translation>Read HMM3 profile(s) from %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Reader</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="253"/>
        <source>Loaded HMM3 profile(s) from %1</source>
        <translation>Loaded HMM3 profile(s) from %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3SearchPrompter</name>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="136"/>
        <source>For each sequence from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>For each sequence from &lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="137"/>
        <source>using all profiles provided by &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>using all profiles provided by &lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="141"/>
        <source>%1 search HMM3 signals %2. &lt;br&gt;Output the list of found regions annotated as &lt;u&gt;%4&lt;/u&gt;.</source>
        <translation>%1 search HMM3 signals %2. &lt;br&gt;Output the list of found regions annotated as &lt;u&gt;%4&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3SearchWorker</name>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="71"/>
        <source>HMM3 profile</source>
        <translation>HMM3 profile</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="71"/>
        <source>HMM3 profile(s) to search with.</source>
        <translation>HMM3 profile(s) to search with.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="72"/>
        <source>Input sequence</source>
        <translation>Input sequence</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="73"/>
        <source>An input sequence (nucleotide or protein) to search in.</source>
        <translation>An input sequence (nucleotide or protein) to search in.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="74"/>
        <source>HMM3 annotations</source>
        <translation>HMM3 annotations</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="75"/>
        <source>Annotations marking found similar sequence regions.</source>
        <translation>Annotations marking found similar sequence regions.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="89"/>
        <source>Result annotation</source>
        <translation>Result annotation</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="89"/>
        <source>A name of the result annotations.</source>
        <translation>A name of the result annotations.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="90"/>
        <source>Seed</source>
        <translation>Seed</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="90"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation>Random generator seed. 0 - means that one-time arbitrary seed will be used.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="91"/>
        <source>Filter by high E-value</source>
        <translation>Filter by high E-value</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="91"/>
        <source>Report domains with e-value less than.</source>
        <translation>Report domains with e-value less than.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="92"/>
        <source>Filter by low score</source>
        <translation>Filter by low score</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="92"/>
        <source>Report domains with score greater than.</source>
        <translation>Report domains with score greater than.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="100"/>
        <source>HMM3 Search</source>
        <translation>HMM3 Search</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="101"/>
        <source>Searches each input sequence for significantly similar sequence matches to all specified HMM profiles. In case several profiles were supplied, searches with all profiles one by one and outputs united set of annotations for each sequence.</source>
        <translation>Searches each input sequence for significantly similar sequence matches to all specified HMM profiles. In case several profiles were supplied, searches with all profiles one by one and outputs united set of annotations for each sequence.</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="168"/>
        <source>Power of e-value must be less or equal to zero. Using default value: 1e-1</source>
        <translation>Power of e-value must be less or equal to zero. Using default value: 1e-1</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="175"/>
        <source>Score must be greater than zero. Using default value: 0.01</source>
        <translation>Score must be greater than zero. Using default value: 0.01</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="181"/>
        <source>Value for attribute name is empty, default name used</source>
        <translation>Value for attribute name is empty, default name used</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="227"/>
        <source>Search HMM3 signals in %1</source>
        <translation>Search HMM3 signals in %1</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="231"/>
        <source>Bad sequence supplied to input: %1</source>
        <translation>Bad sequence supplied to input: %1</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3SearchWorker.cpp" line="260"/>
        <source>Found %1 HMM3 signals</source>
        <translation>Found %1 HMM3 signals</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3WritePrompter</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="184"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="188"/>
        <source>Save HMM3 profile(s) from &lt;u&gt;%1&lt;/u&gt; to &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation>Save HMM3 profile(s) from &lt;u&gt;%1&lt;/u&gt; to &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HMM3Writer</name>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="277"/>
        <source>Empty HMM3 passed for writing to %1</source>
        <translation>Empty HMM3 passed for writing to %1</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="277"/>
        <source>Unspecified URL for writing HMM3</source>
        <translation>Unspecified URL for writing HMM3</translation>
    </message>
    <message>
        <location filename="../src/workers/HMM3IOWorker.cpp" line="293"/>
        <source>Writing HMM3 profile to %1</source>
        <translation>Writing HMM3 profile to %1</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3ADVContext</name>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="247"/>
        <source>Search HMM signals with HMMER3...</source>
        <translation>Search HMM signals with HMMER3...</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="257"/>
        <source>error</source>
        <translation>Error</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="257"/>
        <source>No sequence in focus found</source>
        <translation>No sequence in focus found</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Build</name>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="65"/>
        <source>UGENE cannot determine alphabet of alignment</source>
        <translation>UGENE cannot determine alphabet of alignment</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="70"/>
        <source>Run out of memory (creating alphabet failed)</source>
        <translation>Run out of memory (creating alphabet failed)</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="76"/>
        <source>Run out of memory (creating null model failed)</source>
        <translation>Run out of memory (creating null model failed)</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="81"/>
        <source>Run out of memory (creating builder failed)</source>
        <translation>Run out of memory (creating builder failed)</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="87"/>
        <source>Run out of memory (creating multiple alignment failed)</source>
        <translation>Run out of memory (creating multiple alignment failed)</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="92"/>
        <source>Run out of memory (digitizing of alignment failed)</source>
        <translation>Run out of memory (digitizing of alignment failed)</translation>
    </message>
    <message>
        <location filename="../src/build/uhmm3build.cpp" line="100"/>
        <source>Model building failed</source>
        <translation>Model building failed</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildDialogImpl</name>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="69"/>
        <source>Build</source>
        <translation>Build</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="70"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="109"/>
        <source>Select multiple alignment file</source>
        <translation>Select multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="118"/>
        <source>Select hmm file to create</source>
        <translation>Select HMM file to create</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="184"/>
        <source>input file is empty</source>
        <translation>Input file is empty</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="187"/>
        <source>output hmm file is empty</source>
        <translation>Output HMM profile file is empty</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildDialogImpl.cpp" line="196"/>
        <source>Error: bad arguments!</source>
        <translation>Error: bad arguments!</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildTask</name>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="57"/>
        <source>Build HMM profile from %1 alignment</source>
        <translation>Build HMM profile from %1 alignment</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="71"/>
        <source>Given multiple alignment has no sequences</source>
        <translation>Given multiple alignment has no sequences</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="74"/>
        <source>Given multiple alignment is empty</source>
        <translation>Given multiple alignment is empty</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3BuildToFileTask</name>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="126"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="146"/>
        <source>Build HMM profile to &apos;%1&apos;</source>
        <translation>Build HMM profile to &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="163"/>
        <source>Build HMM profile &apos;%1&apos; -&gt; &apos;%2&apos;</source>
        <translation>Build HMM profile &apos;%1&apos; -&gt; &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="129"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="149"/>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="170"/>
        <source>Output file is not given</source>
        <translation>Output file is not given</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="134"/>
        <source>No multiple alignments given</source>
        <translation>No multiple alignments given</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="166"/>
        <source>Input file is not given</source>
        <translation>Input file is not given</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="181"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="188"/>
        <source>Error opening &apos;%1&apos; file</source>
        <translation>Error opening &apos;%1&apos; file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="229"/>
        <source>No multiple alignments found in input file</source>
        <translation>No multiple alignments found in input file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="276"/>
        <source>Source alignment</source>
        <translation>Source alignment</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="278"/>
        <source>Profile name</source>
        <translation>Profile name</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="282"/>
        <source>Options:</source>
        <translation>Options:</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="283"/>
        <source>Model construction strategies</source>
        <translation>Model construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="291"/>
        <source>Relative model construction strategies</source>
        <translation>Relative model construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="293"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation>Gerstein/Sonnhammer/Chothia tree weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="294"/>
        <source>Henikoff simple filter weights</source>
        <translation>Henikoff simple filter weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="295"/>
        <source>Henikoff position-based weights</source>
        <translation>Henikoff position-based weights</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="296"/>
        <source>No relative weighting; set all to 1</source>
        <translation>No relative weighting; set all to 1</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="297"/>
        <source>Weights given in MSA file</source>
        <translation>Weights given in MSA file</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="302"/>
        <source>Effective sequence weighting strategies</source>
        <translation>Effective sequence weighting strategies</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="304"/>
        <source>adjust effective sequence number to achieve relative entropy target</source>
        <translation>Adjust effective sequence number to achieve relative entropy target</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="305"/>
        <source>effective sequence number is number of single linkage clusters</source>
        <translation>Effective sequence number is number of single linkage clusters</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="306"/>
        <source>no effective sequence number weighting: just use number of sequences</source>
        <translation>No effective sequence number weighting: just use number of sequences</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="307"/>
        <source>set effective sequence number for all models to: %1</source>
        <translation>Set effective sequence number for all models to: %1</translation>
    </message>
    <message>
        <location filename="../src/build/uHMM3BuildTask.cpp" line="313"/>
        <source>Task finished with error: &apos;%1&apos;</source>
        <translation>Task finished with error: &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3LoadProfileAndSearchTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="514"/>
        <source>HMM search with &apos;%1&apos; HMM profile file</source>
        <translation>HMM search with &apos;%1&apos; HMM profile file</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3MSAEditorContext</name>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="202"/>
        <source>Build HMMER3 profile</source>
        <translation>Build HMMER3 profile</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Phmmer</name>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="168"/>
        <source>No input query sequence given</source>
        <translation>No input query sequence given</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="172"/>
        <source>Database sequence to search in is not given</source>
        <translation>Database sequence to search in is not given</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="179"/>
        <source>Run out of memory (creating alphabet failed)</source>
        <translation>Run out of memory (creating alphabet failed)</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="184"/>
        <source>Run out of memory (creating null model failed)</source>
        <translation>Run out of memory (creating null model failed)</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="191"/>
        <source>Run out of memory (creating builder failed)</source>
        <translation>Run out of memory (creating builder failed)</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="198"/>
        <source>Setting scoring system failed with error: &apos;%1&apos;</source>
        <translation>Setting scoring system failed with error: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="205"/>
        <source>Error with digitizing sequence to search in</source>
        <translation>Error with digitizing sequence to search in</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="212"/>
        <source>Error digitizing query sequence</source>
        <translation>Error digitizing query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="222"/>
        <source>Error with creating HMM profile for query sequence</source>
        <translation>Error with creating HMM profile for query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="228"/>
        <source>Run out of memory (creating top hits list failed)</source>
        <translation>Run out of memory (creating top hits list failed)</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3phmmer.cpp" line="234"/>
        <source>Run out of memory (creating pipeline failed)</source>
        <translation>Run out of memory (creating pipeline failed)</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerDialogImpl</name>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="58"/>
        <source>Search</source>
        <translation>Search</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="59"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="117"/>
        <source>Select query sequence file</source>
        <translation>Select query sequence file</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="165"/>
        <source>Query sequence file path is empty</source>
        <translation>Query sequence file path is empty</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="170"/>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="181"/>
        <source>Error: bad arguments!</source>
        <translation>Error: bad arguments!</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="186"/>
        <source>Error</source>
        <translation>Error</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uHMM3PhmmerDialogImpl.cpp" line="186"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Cannot create an annotation object. Please check settings</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerTask</name>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="61"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="79"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="101"/>
        <source>HMM Phmmer task</source>
        <translation>HMM Phmmer task</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="72"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="110"/>
        <source>HMM Phmmer search %1 sequence in %2 database</source>
        <translation>HMM Phmmer search %1 sequence in %2 databas</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="87"/>
        <source>HMM Phmmer search %1 sequence with %2 database</source>
        <translation>HMM Phmmer search %1 sequence with %2 database</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="69"/>
        <source>Input query sequence</source>
        <translation>Input query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="70"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="108"/>
        <source>Database sequence to search in</source>
        <translation>Database sequence to search in</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="84"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="107"/>
        <source>Query sequence file path</source>
        <translation>Query sequence file path</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="85"/>
        <source>Database sequence file path</source>
        <translation>Database sequence file path</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="90"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="113"/>
        <source>Error opening query sequence file</source>
        <translation>Error opening query sequence file</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="94"/>
        <source>Error opening database sequence file</source>
        <translation>Error opening database sequence file</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="128"/>
        <source>Error loading sequence document:</source>
        <translation>Error loading sequence document:</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="131"/>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="134"/>
        <source>No dna sequence objects found in document</source>
        <translation>No dna sequence objects found in document</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="138"/>
        <source>Empty sequence loaded from document</source>
        <translation>Empty sequence loaded from document</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="150"/>
        <source> query sequence</source>
        <translation> query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="154"/>
        <source> db sequence</source>
        <translation> db sequence</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3PhmmerToAnnotationsTask</name>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="375"/>
        <source>querySeq sequence file path</source>
        <translation>querySeq sequence file path</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="380"/>
        <source>db sequence</source>
        <translation>Database sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="385"/>
        <source>annotation object</source>
        <translation>Annotation object</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="390"/>
        <source>annotation name</source>
        <translation>Annotation name</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="395"/>
        <source>annotation group</source>
        <translation>Annotation group</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="408"/>
        <source>HMM Phmmer task</source>
        <translation>HMM Phmmer task</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="422"/>
        <source>HMM Phmmer search %1 sequence with %2 database</source>
        <translation>Phmmer search %1 sequence with %2 database</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="432"/>
        <source>Annotation object was removed</source>
        <translation>Annotation object was removed</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="456"/>
        <source>Query sequence</source>
        <translation>Query sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="459"/>
        <source>Task was not finished</source>
        <translation>Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="464"/>
        <source>Result annotation table</source>
        <translation>Result annotation table</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="465"/>
        <source>Result annotation group</source>
        <translation>Result annotation group</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="466"/>
        <source>Result annotation name</source>
        <translation>Result annotation name</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="469"/>
        <source>Results count</source>
        <translation>Results count</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Plugin</name>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="70"/>
        <source>HMM3</source>
        <translation>HMM3</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="70"/>
        <source>HMM profile tools. Plugin is based on HMMER 3.0b3 package: freely distributable implementation of profile HMM software for protein sequence analysis. Home page of project: http://hmmer.janelia.org/</source>
        <translation>HMM profile tools. Plugin is based on HMMER 3.0b3 package: freely distributable implementation of profile HMM software for protein sequence analysis. Home page of project: http://hmmer.janelia.org/</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="100"/>
        <source>Build HMM3 profile...</source>
        <translation>Build HMM3 profile...</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="105"/>
        <source>Search with HMM3...</source>
        <translation>Search with HMM3...</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="110"/>
        <source>Search with HMM3 phmmer...</source>
        <translation>Search with HMM3 phmmer...</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="167"/>
        <location filename="../src/uHMM3Plugin.cpp" line="178"/>
        <source>Error!</source>
        <translation>Error!</translation>
    </message>
    <message>
        <location filename="../src/uHMM3Plugin.cpp" line="167"/>
        <location filename="../src/uHMM3Plugin.cpp" line="178"/>
        <source>Target sequence not selected: no opened annotated dna view</source>
        <translation>No target sequence</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3QDActor</name>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="73"/>
        <location filename="../src/search/uhmm3QDActor.cpp" line="167"/>
        <source>HMM3</source>
        <translation>HMM3</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="84"/>
        <source>QD HMM3 search</source>
        <translation>QD HMM3 search</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="168"/>
        <source>Searches HMM signals in a sequence with one or more profile HMM and saves the results as annotations.</source>
        <translation>Searches HMM signals in a sequence with one or more profile HMM and saves the results as annotations.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="174"/>
        <source>Profile HMM</source>
        <translation>Profile HMM</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="175"/>
        <source>Semicolon-separated list of input HMM files.</source>
        <translation>Semicolon-separated list of input HMM files.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="178"/>
        <source>Min Length</source>
        <translation>Min Length</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="179"/>
        <source>Minimum length of a result region.</source>
        <translation>Minimum length of a result region.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="182"/>
        <source>Max Length</source>
        <translation>Max Length</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="183"/>
        <source>Maximum length of a result region.</source>
        <translation>Maximum length of a result region.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="188"/>
        <source>Use E-value</source>
        <translation>Use E-value</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="189"/>
        <source>Filters by E-value if true. Otherwise filters by score.</source>
        <translation>Filters by E-value if true. Otherwise filters by score.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="192"/>
        <source>Filter by High E-value</source>
        <translation>Filter by High E-value</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="193"/>
        <source>Reports domains &amp;lt;= this E-value threshold in output.</source>
        <translation>Reports domains &amp;lt;= this E-value threshold in output.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="197"/>
        <source>Filter by Low Score</source>
        <translation>Filter by Low Score</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="198"/>
        <source>Reports domains &amp;gt;= this score cutoff in output.</source>
        <translation>Reports domains &amp;gt;= this score cutoff in output.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="203"/>
        <source>Max</source>
        <translation>Max</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="204"/>
        <source>Turns off all acceleration heuristic filters. This increases sensitivity somewhat, at a large cost in speed.</source>
        <translation>Turns off all acceleration heuristic filters. This increases sensitivity somewhat, at a large cost in speed.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="209"/>
        <source>MSV Filter Threshold</source>
        <translation>MSV Filter Threshold</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="210"/>
        <source>P-value threshold for the MSV filter step of the acceleration pipeline.</source>
        <translation>P-value threshold for the MSV filter step of the acceleration pipeline.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="214"/>
        <source>Viterbi Filter Threshold</source>
        <translation>Viterbi Filter Threshold</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="215"/>
        <source>P-value threshold for the Viterbi filter step of the acceleration pipeline.</source>
        <translation>P-value threshold for the Viterbi filter step of the acceleration pipeline.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="219"/>
        <source>Forward Filter Threshold</source>
        <translation>Forward Filter Threshold</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="220"/>
        <source>P-value threshold for the Forward filter step of the acceleration pipeline.</source>
        <translation>P-value threshold for the Forward filter step of the acceleration pipeline.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="224"/>
        <source>No Bias Filter</source>
        <translation>No Bias Filter</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="225"/>
        <source>Turns off composition bias filter. This increases sensitivity somewhat, but can come at a high cost in speed.</source>
        <translation>Turns off composition bias filter. This increases sensitivity somewhat, but can come at a high cost in speed.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="232"/>
        <source>No Null2</source>
        <translation>No Null2</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="233"/>
        <source>Turns off the null2 score corrections for biased composition.</source>
        <translation>Turns off the null2 score corrections for biased composition.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="237"/>
        <source>Number of Sequences</source>
        <translation>Number of Sequences</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="238"/>
        <source>Specifies number of significant sequences. It is used for domain E-value calculations.</source>
        <translation>Specifies number of significant sequences. It is used for domain E-value calculations.</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="242"/>
        <source>Seed</source>
        <translation>Seed</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3QDActor.cpp" line="243"/>
        <source>Random number seed. The default is to use a fixed seed(42), so that results are exactly reproducible. Any other positive integer will give different (but also reproducible) results. A choice of 0 uses a randomly chosen seed.</source>
        <translation>Random number seed. The default is to use a fixed seed(42), so that results are exactly reproducible. Any other positive integer will give different (but also reproducible) results. A choice of 0 uses a randomly chosen seed.</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWPhmmerTask</name>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="216"/>
        <source>HMM Phmmer search %1 sequence in %2 database</source>
        <translation>HMM Phmmer search %1 sequence in %2 databas</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="222"/>
        <source>Can not create load query doc task</source>
        <translation>Can not create load query doc task</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="232"/>
        <source> querySeq sequence</source>
        <translation> querySeq sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="245"/>
        <source>Invalid db sequence alphabet: %1</source>
        <translation>Invalid db sequence alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="251"/>
        <source>Invalid query sequence alphabet: %1</source>
        <translation>Invalid query sequence alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="274"/>
        <source>Cannot search for nucleic query in amino sequence</source>
        <translation>Cannot search for nucleic query in amino sequence</translation>
    </message>
    <message>
        <location filename="../src/phmmer/uhmm3PhmmerTask.cpp" line="302"/>
        <source>HMMER3 phmmer sequence walker search task</source>
        <translation>HMMER3 phmmer sequence walker search task</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWSearchTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="86"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="106"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="147"/>
        <source>HMM search task with amino and complement translations</source>
        <translation>HMM search task with amino and complement translations</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="92"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="110"/>
        <source>HMM search task with amino and complement translations using &apos;%1&apos; profile HMM</source>
        <translation>HMM search task with amino and complement translations using &apos;%1&apos; profile HMM</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="365"/>
        <source>Unrecognized alphabet of sequence</source>
        <translation>Unrecognized alphabet of sequence</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="369"/>
        <source>Invalid alphabet of sequence</source>
        <translation>Invalid alphabet of sequence</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="375"/>
        <source>Cannot search for nucleic HMM profile in amino sequence</source>
        <translation>Cannot search for nucleic HMM profile in amino sequence</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SWSearchToAnnotationsTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="557"/>
        <source>hmm profile filename</source>
        <translation>Profile HMM file path</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="561"/>
        <source>annotation object</source>
        <translation>Annotation object</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="565"/>
        <source>annotations group name</source>
        <translation>Annotations group name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="569"/>
        <source>annotations name</source>
        <translation>Annotations name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="586"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="612"/>
        <source>HMMER3 search task</source>
        <translation>HMM3 search task</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="589"/>
        <source>dna sequence</source>
        <translation>Sequence to search in</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="594"/>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="620"/>
        <source>HMMER3 search task with &apos;%1&apos; profile</source>
        <translation>HMMER3 search task with &apos;%1&apos; profile</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="615"/>
        <source>Sequence file</source>
        <translation>Sequence file path</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="634"/>
        <source>HMM profile used</source>
        <translation>HMM profile used</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="637"/>
        <source>Task was not finished</source>
        <translation>Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="642"/>
        <source>Result annotation table</source>
        <translation>Result annotation table</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="643"/>
        <source>Result annotation group</source>
        <translation>Result annotation group</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="644"/>
        <source>Result annotation name</source>
        <translation>Result annotation name</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="647"/>
        <source>Results count</source>
        <translation>Results count</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="657"/>
        <source>Cannot load sequence document</source>
        <translation>Cannot load sequence document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="662"/>
        <source>No sequence objects loaded</source>
        <translation>No sequence objects found in document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="666"/>
        <source>Unknown sequence type loaded</source>
        <translation>No sequence objects found in document</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="670"/>
        <source>Empty sequence loaded</source>
        <translation>Empty sequence loaded</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="686"/>
        <source>Annotation object removed</source>
        <translation>Choosed annotation object removed</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3Search</name>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="61"/>
        <source>Bad HMM profile given</source>
        <translation>Bad HMM profile given</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="65"/>
        <source>Empty sequence given</source>
        <translation>Empty sequence given</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="73"/>
        <source>Run out of memory (creation of sequence failed)</source>
        <translation>Run out of memory (creation of sequence failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="79"/>
        <source>Profile HMM and sequence alphabets no matched</source>
        <translation>Profile HMM and sequence alphabets no matched</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="85"/>
        <source>Run out of memory (creation of alphabet failed)</source>
        <translation>Run out of memory (creation of alphabet failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="90"/>
        <source>Run out of memory (digitizing of sequence failed)</source>
        <translation>Run out of memory (digitizing of sequence failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="96"/>
        <source>Run out of memory (creation of null model failed)</source>
        <translation>Run out of memory (creation of null model failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="102"/>
        <location filename="../src/search/uhmm3search.cpp" line="116"/>
        <source>Run out of memory</source>
        <translation>Run out of memory</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="107"/>
        <source>Run out of memory (creation of optimized profile failed)</source>
        <translation>Run out of memory (creation of optimized profile failed)</translation>
    </message>
    <message>
        <location filename="../src/search/uhmm3search.cpp" line="121"/>
        <source>Run out of memory (top hits list creation failed)</source>
        <translation>Run out of memory (top hits list creation failed)</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SearchDialogImpl</name>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="58"/>
        <source>Run</source>
        <translation>Run</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="59"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="160"/>
        <source>HMM profile file path is empty</source>
        <translation>HMM profile file path is empty</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="175"/>
        <source>Error: bad arguments!</source>
        <translation>Error: bad arguments!</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="182"/>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="186"/>
        <source>Error</source>
        <translation>Error</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="186"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Cannot create an annotation object. Please check settings</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchDialogImpl.cpp" line="236"/>
        <source>Select query HMM profile</source>
        <translation>Select query HMM profile</translation>
    </message>
</context>
<context>
    <name>U2::UHMM3SearchTask</name>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="469"/>
        <source>HMM search task</source>
        <translation>HMM search task</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="478"/>
        <source>HMM search with %1 profiles</source>
        <translation>HMM search with %1 profiles</translation>
    </message>
    <message>
        <location filename="../src/search/uHMM3SearchTask.cpp" line="476"/>
        <source>HMM search with &apos;%1&apos;</source>
        <translation>HMM search with &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::UHMMFormat</name>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="279"/>
        <source>Unknown error occurred</source>
        <translation>Unknown error occurred</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="305"/>
        <source>Profile HMM format</source>
        <translation>Profile HMM format</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormat.cpp" line="306"/>
        <source>hmm is a format for storing hmm profiles</source>
        <translation>hmm is a format for storing hmm profiles</translation>
    </message>
</context>
<context>
    <name>U2::UHMMFormatReader</name>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="117"/>
        <source>Number expected in NULE line. %1 found</source>
        <translation>Number expected in NULE line. %1 found</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="140"/>
        <source>Bad line in header section: &apos;%1&apos;</source>
        <translation>Bad line in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="147"/>
        <source>Empty value in header line:%1</source>
        <translation>Empty value in header line:%1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="387"/>
        <source>Input file made by unknown version of HMMER or is not HMM profile file</source>
        <translation>Input file made by unknown version of HMMER or is not HMM profile file</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="750"/>
        <source>ALPH must precede NULE in HMMER2 save files</source>
        <translation>ALPH must precede NULE in HMMER2 save files</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="769"/>
        <source>ALPH section must precede HMM</source>
        <translation>ALPH section must precede HMM</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="774"/>
        <source>Unknown alphabet</source>
        <translation>Unknown alphabet</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="797"/>
        <source>Failed to allocate body of the new HMM</source>
        <translation>Failed to allocate body of the new HMM</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="157"/>
        <source>Cannot parse integer from string: &apos;%1&apos;</source>
        <translation>Cannot parse integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="168"/>
        <source>Cannot parse float number from string: &apos;%1&apos;</source>
        <translation>Cannot parse float number from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="184"/>
        <source>Cannot parse 2 float numbers from string: &apos;%1&apos;</source>
        <translation>Cannot parse 2 float numbers from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="195"/>
        <source>Cannot parse unsigned integer from string: &apos;%1&apos;</source>
        <translation>Cannot parse unsigned integer from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="207"/>
        <source>Cannot parse y/n value from string: &apos;%1&apos;</source>
        <translation>Cannot parse y/n value from string: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="216"/>
        <source>Run out of memory (date allocation failed)</source>
        <translation>Run out of memory (date allocation failed)</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="231"/>
        <source>Bad STATS line: &apos;%1&apos;</source>
        <translation>Bad STATS line: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="235"/>
        <source>Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</source>
        <translation>Bad STATS line: &apos;%1&apos;/  LOCAL word was not found</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="263"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="286"/>
        <source>Bad STATS line: %1. %2 not recognized</source>
        <translation>Bad STATS line: %1. %2 not recognized</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="299"/>
        <source>Unexpected end of file</source>
        <translation>Unexpected end of file</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="409"/>
        <source>Run out of memory (allocation of HMM shell failed)</source>
        <translation>Run out of memory (allocation of HMM shell failed)</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="439"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="697"/>
        <source>Length of HMM model should be positive. Found: %1</source>
        <translation>Length of HMM model should be positive. Found: %1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="447"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="707"/>
        <source>Unrecognized alphabet type: %1</source>
        <translation>Unrecognized alphabet type: %1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="470"/>
        <location filename="../src/format/uHMMFormatReader.cpp" line="731"/>
        <source>Number of sequences should be positive. Found: %1</source>
        <translation>Number of sequences should be positive. Found: %1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="477"/>
        <source>EFFN shoold be positive. Found: %1</source>
        <translation>EFFN shoold be positive. Found: %1</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="512"/>
        <source>Unrecognized tag in header section: &apos;%1&apos;</source>
        <translation>Unrecognized tag in header section: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="523"/>
        <source>Missing one or more STATS parameters</source>
        <translation>Missing one or more STATS parameters</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="528"/>
        <source>Run out of memory (failed to create alphabet)</source>
        <translation>Run out of memory (failed to create alphabet)</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="533"/>
        <source>Run out of memory (allocation of HMM body failed)</source>
        <translation>Run out of memory (allocation of HMM body failed)</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="574"/>
        <source>Line was excpected to start with %1. Found: %2</source>
        <translation>Line was excpected to start with %1. Found: %2</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="616"/>
        <source>Excpected to find closing &apos;//&apos;. Found %1 instead</source>
        <translation>Excpected to find closing &apos;//&apos;. Found %1 instead</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="633"/>
        <source>HMM name not found</source>
        <translation>HMM name not found</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="659"/>
        <source>Allocation failure, HMM shell</source>
        <translation>Allocation failure, HMM shell</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="800"/>
        <source>Failed to create background model</source>
        <translation>Failed to create background model</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="830"/>
        <source>Expected match line to start with %1. saw %2</source>
        <translation>Expected match line to start with %1. saw %2</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="884"/>
        <source>Expected closing //. found %1 instead</source>
        <translation>Expected closing //. found %1 instead</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="897"/>
        <source>No NAME found for HMM</source>
        <translation>No NAME found for HMM</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="900"/>
        <source>No LENG found for HMM</source>
        <translation>No LENG found for HMM</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="903"/>
        <source>No ALPH found for HMM</source>
        <translation>No ALPH found for HMM</translation>
    </message>
    <message>
        <location filename="../src/format/uHMMFormatReader.cpp" line="909"/>
        <source>Failed to calibrate HMMER2 model after input conversion</source>
        <translation>Failed to calibrate HMMER2 model after input conversion</translation>
    </message>
</context>
<context>
    <name>UHMM3BuildDialog</name>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="59"/>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="76"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="123"/>
        <source>assign cols w/ &gt;= symfrac residues as consensus</source>
        <translation>assign cols w/ &gt;= symfrac residues as consensus</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="126"/>
        <source>fast</source>
        <translation>Fast</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="20"/>
        <source>Hmm3 Build</source>
        <translation>Hmm3 Build</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="39"/>
        <source>Input and output</source>
        <translation>Input and output</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="49"/>
        <source>Input alignment file</source>
        <translation>Input alignment file</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="66"/>
        <source>Build to profile</source>
        <translation>Build to profile</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="101"/>
        <source>Construction strategies</source>
        <translation>Construction strategies</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="139"/>
        <source>Sym fraction</source>
        <translation>Sym fraction</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="161"/>
        <source>Manual construction</source>
        <translation>Manual construction</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="164"/>
        <source>Hand (requires reference annotation)</source>
        <translation>Hand (requires reference annotation)</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="182"/>
        <source>Tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</source>
        <translation>Tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="223"/>
        <source>Relative weighting</source>
        <translation>Relative weighting</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="231"/>
        <source>Henikoff position-based weights</source>
        <translation>Henikoff position-based weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="247"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation>Gerstein/Sonnhammer/Chothia tree weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="271"/>
        <source>Henikoff simple filter weights</source>
        <translation>Henikoff simple filter weights</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="281"/>
        <source>Identity cutoff</source>
        <translation>Identity cutoff</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="306"/>
        <source>Skip relative weighting; set all to 1</source>
        <translation>Skip relative weighting; set all to 1</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="313"/>
        <source>Use weights given in alignment file</source>
        <translation>Use weights given in alignment file</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="336"/>
        <source>Effective weighting</source>
        <translation>Effective weighting</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="344"/>
        <source>Adjust effective sequence number to achieve relative entropy target</source>
        <translation>Adjust effective sequence number to achieve relative entropy target</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="378"/>
        <source>Minimum relative entropy/position</source>
        <translation>Minimum relative entropy/position</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="430"/>
        <source>Sigma parameter</source>
        <translation>Sigma parameter</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="463"/>
        <source>Use number of single linkage clusters as effective</source>
        <translation>Use number of single linkage clusters as effective</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="473"/>
        <source>Fractional identity cutoff</source>
        <translation>Fractional identity cutoff</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="498"/>
        <source>Use number of sequences as effective</source>
        <translation>Use number of sequences as effective</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="513"/>
        <source>Effective sequence number for all models to</source>
        <translation>Effective sequence number for all models to</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="551"/>
        <source>E-value calibration</source>
        <translation>E-value calibration</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="570"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation>Length of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="606"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation>Number of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="639"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation>Length of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="669"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation>Number of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="699"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation>Length of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="729"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation>Number of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="759"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation>Tail mass for Forward exponential tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="800"/>
        <source>Other</source>
        <translation>Other</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="816"/>
        <source>Random generator seed</source>
        <translation>Random generator seed</translation>
    </message>
    <message>
        <location filename="../src/build/ui/UHMM3BuildDialog.ui" line="859"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 means that one-time arbitrary seed will be used</translation>
    </message>
</context>
<context>
    <name>UHMM3PhmmerDialog</name>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="20"/>
        <source>Phmmer Search</source>
        <translation>Phmmer Search</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="36"/>
        <source>Input and output</source>
        <translation>Input and output</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="50"/>
        <source>Query sequence file:</source>
        <translation>Query sequence file:</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="60"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="87"/>
        <source>Reporting tresholds</source>
        <translation>Reporting tresholds</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="103"/>
        <source>Report domains with E-value less than</source>
        <translation>Report domains with E-value less than</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="142"/>
        <source>Report domains with score greater than</source>
        <translation>Report domains with score greater than</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="166"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation>Number of significant sequences for domain E-value calculation</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="201"/>
        <source>Scoring system</source>
        <translation>Scoring system</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="217"/>
        <source>Gap open probability</source>
        <translation>Gap open probability</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="247"/>
        <source>Gap extend probability</source>
        <translation>Gap extend probability</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="285"/>
        <source>Acceleration</source>
        <translation>Acceleration</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="297"/>
        <source>Turn off composition bias filter</source>
        <translation>Turn off composition bias filter</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="310"/>
        <source>Turn off biased composition score corrections</source>
        <translation>Turn off biased composition score corrections</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="338"/>
        <source>Viterbi filter treshold</source>
        <translation>Viterbi filter treshold</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="368"/>
        <source>MSV filter treshold</source>
        <translation>MSV filter treshold</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="401"/>
        <source>Forward filter treshold</source>
        <translation>Forward filter treshold</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="437"/>
        <source>E-value calibration</source>
        <translation>E-value calibration</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="456"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation>Length of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="492"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation>Number of sequences for MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="525"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation>Length of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="555"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation>Number of sequences for Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="585"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation>Length of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="615"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation>Number of sequences for Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="645"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation>Tail mass for Forward exponential tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="686"/>
        <source>Other</source>
        <translation>Other</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="702"/>
        <source>Random generator seed</source>
        <translation>Random generator seed</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="745"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 means that one-time arbitrary seed will be used</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="113"/>
        <source>1E+</source>
        <translation>1E+</translation>
    </message>
    <message>
        <location filename="../src/phmmer/ui/UHMM3PhmmerDialog.ui" line="323"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation>Turn all heuristic filters off (less speed, more power)</translation>
    </message>
</context>
<context>
    <name>UHMM3SearchDialog</name>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="65"/>
        <source>Query HMM file:</source>
        <translation>Query profile HMM file:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="75"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="119"/>
        <source>1E+</source>
        <translation>1E+</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="54"/>
        <source>Input and output</source>
        <translation>Input and output</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="92"/>
        <source>Reporting thresholds</source>
        <translation>Reporting thresholds</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="109"/>
        <source>Report domains with E-value less than</source>
        <translation>Report domains with E-value less than</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="146"/>
        <source>Report domains with score greater than</source>
        <translation>Report domains with score greater than</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="20"/>
        <source>HMM3 Search</source>
        <translation>HMM3 Search</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="180"/>
        <source>Score threshold:</source>
        <translation>Score threshold:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="229"/>
        <source>Use profile&apos;s GA gathering cutoffs</source>
        <translation>Use profile&apos;s GA gathering cutoffs</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="265"/>
        <source>Use profile&apos;s NC noise cutoffs</source>
        <translation>Use profile&apos;s NC noise cutoffs</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="295"/>
        <source>Use profile&apos;s TC trusted cutoffs</source>
        <translation>Use profile&apos;s TC trusted cutoffs</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="312"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation>Number of significant sequences for domain E-value calculation</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="345"/>
        <source>Acceleration heuristics</source>
        <translation>Acceleration heuristics</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="360"/>
        <source>Turn off composition bias filter</source>
        <translation>Turn off composition bias filter</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="373"/>
        <source>Turn off biased composition score corrections</source>
        <translation>Turn off biased composition score corrections</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="401"/>
        <source>MSV filter threshold:</source>
        <translation>MSV filter threshold:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="434"/>
        <source>Viterbi filter threshold:</source>
        <translation>Viterbi filter threshold:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="464"/>
        <source>Forward filter threshold:</source>
        <translation>Forward filter threshold:</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="500"/>
        <source>Other</source>
        <translation>Other</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="522"/>
        <source>Random generator seed</source>
        <translation>Random generator seed</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="565"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 means that one-time arbitrary seed will be used</translation>
    </message>
    <message>
        <location filename="../src/search/ui/UHMM3SearchDialog.ui" line="386"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation>Turn all heuristic filters off (less speed, more power)</translation>
    </message>
</context>
</TS>
