<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>PhyTreeGeneratorLauncherTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="45"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="54"/>
        <source>Tree construction algorithm %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>PhyTreeGeneratorTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="32"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="64"/>
        <source>Loading OPENCL driver library</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="68"/>
        <source>Cannot load OpenCL library. Error while loading %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="75"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="82"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="89"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="96"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="105"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="112"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="119"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="126"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="133"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="140"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="147"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="154"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="161"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="168"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="175"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="182"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="189"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="196"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="203"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="210"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="217"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="224"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="231"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="238"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="245"/>
        <source>Cannot resolve symbol %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="260"/>
        <source>Cannot load library: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="263"/>
        <source>Some errors occurs in library: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusUtils.cpp" line="109"/>
        <source>Gaps:</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="70"/>
        <source>clear OpenCL resources</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="169"/>
        <source>GPU memory usage: %1 Mb</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="183"/>
        <source>Device&apos;s preferred work group size multiple is %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLUtils.cpp" line="72"/>
        <source>OPENCL: BUILD LOG 
 ******************** 
 %1 
 ********************</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLUtils.cpp" line="73"/>
        <source>OPENCL: Program::build() failed. (%1)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AlignInAminoFormTask</name>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="141"/>
        <source>Can not find row %1 in original alignment.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="39"/>
        <source>Default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="43"/>
        <source>Returns simply the most frequent base and &apos;N&apos; are no reads intersecting this position</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactorySamtools</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="45"/>
        <source>SAMtools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="49"/>
        <source>Uses SAMtools to calculate consensus with regard to quality of reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="71"/>
        <source>Fetching reads from database and converting to SAMtools format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="74"/>
        <source>Sorting reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="77"/>
        <source>Calculating consensus</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CreateSubalignmentTask</name>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="40"/>
        <source>Create sub-alignment: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="71"/>
        <source>Document is locked: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DnaAssemblyMultiTask</name>
    <message>
        <location filename="../src/misc/DnaAssemblyMultiTask.cpp" line="81"/>
        <source>The short reads can&apos;t be mapped to the reference sequence!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FindAlgorithmTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="43"/>
        <source>Find in sequence task</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LoadPatternsFileTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="125"/>
        <source>Detecting format error for file %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryClustal</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmClustal.cpp" line="30"/>
        <source>Emulates ClustalW program and file format behavior.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmClustal.cpp" line="34"/>
        <source>ClustalW</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmDefault.cpp" line="37"/>
        <source>Based on JalView algorithm. Returns &apos;+&apos; if there are 2 characters with high frequency. Returns symbol in lower case if the symbol content in a row is lower than the threshold specified.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmDefault.cpp" line="41"/>
        <source>Default</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryLevitsky</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmLevitsky.cpp" line="36"/>
        <source>The algorithm proposed by Victor Levitsky to work with DNA alignments.
Collects global alignment frequency for every symbol using extended (15 symbols) DNA alphabet first.
For every column selects the most rare symbol in the whole alignment with percentage in the column greater or equals to the threshold value.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmLevitsky.cpp" line="44"/>
        <source>Levitsky</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryStrict</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmStrict.cpp" line="39"/>
        <source>The algorithm returns gap character (&apos;-&apos;) if symbol frequency in a column is lower than threshold specified.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmStrict.cpp" line="43"/>
        <source>Strict</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithm</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithm.cpp" line="63"/>
        <source>MSA distance algorithm &quot;%1&quot; task</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHamming</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHamming.cpp" line="36"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHamming.cpp" line="40"/>
        <source>Hamming dissimilarity</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="42"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="46"/>
        <source>Hamming reverse-complement</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactorySimilarity</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmSimilarity.cpp" line="36"/>
        <source>Based on similarity distance between two sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmSimilarity.cpp" line="40"/>
        <source>Simple identity</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="75"/>
        <source>An unexpected error has occurred during running the Hamming reverse-complement algorithm.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MolecularSurfaceCalcTask</name>
    <message>
        <location filename="../src/molecular_geometry/MolecularSurface.cpp" line="123"/>
        <source>Molecular surface calculation</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ORFFindTask</name>
    <message>
        <location filename="../src/util_orf/ORFAlgorithmTask.cpp" line="31"/>
        <source>ORF find</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryBVH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmBVH.cpp" line="42"/>
        <source>Berg and von Hippel weight function</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryLOD</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="38"/>
        <source>log-odds</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="42"/>
        <source>log-odds weight function</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryMCH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="37"/>
        <source>Match</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="41"/>
        <source>Match weight function</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryNLG</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="37"/>
        <source>NLG</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="41"/>
        <source>NLG weight function</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SecStructPredictTask</name>
    <message>
        <location filename="../src/registry/SecStructPredictTask.cpp" line="27"/>
        <source>Secondary structure predict</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackAnnotImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="62"/>
        <source>Annotation object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="66"/>
        <source>Annotation table is read-only</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackMAImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="178"/>
        <source>SmithWatermanReportCallback failed to create new MA document</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="188"/>
        <source>Failed to add a reference subsequence row.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="192"/>
        <source>Failed to add a pattern subsequence row.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="195"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="276"/>
        <source>Failed to create an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="224"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="298"/>
        <source>Smith-Waterman pairwise alignement: no results</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="231"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="303"/>
        <source>Sequences aren`t stored in the source msa dbi. Reimplement this code.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="243"/>
        <source>SmithWatermanReportCallback failed to create new MA document.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="246"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="248"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="310"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="312"/>
        <source>Failed to get the sequence object.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="262"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="264"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="326"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="328"/>
        <source>Failed to get the sequence data.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="271"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="273"/>
        <source>Failed to add row to result msa.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="334"/>
        <source>Failed to get msa from dbi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="355"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="357"/>
        <source>Failed to update row gap model</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SplicedAlignmentTask</name>
    <message>
        <location filename="../src/registry/SplicedAlignmentTask.cpp" line="27"/>
        <source>SplicedAlignmentTask</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SubstMatrixRegistry</name>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="104"/>
        <source>Reading substitution matrix from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="112"/>
        <source>Error weight matrix file &apos;%1&apos; : %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="120"/>
        <source>Error opening file for read: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="127"/>
        <source>Error reading file: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="160"/>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="184"/>
        <source>Invalid character token &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="165"/>
        <source>Duplicate character &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="172"/>
        <source>Alphabet is neither nucleic nor protein!</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="176"/>
        <source>Matrix doesn&apos;t contains score for default alphabet character &apos;%1&apos;</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="189"/>
        <source>Invalid character row &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="193"/>
        <source>Duplicate character mapping &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="198"/>
        <source>Invalid number of columns &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="206"/>
        <source>Can&apos;t parse numeric value &apos;%1&apos;, line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="215"/>
        <source>Unexpected end of file!</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::TranslateMSA2AminoTask</name>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="48"/>
        <source>Unable to find suitable translation for %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
