<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="64"/>
        <source>Loading OPENCL driver library</source>
        <translation>Loading OPENCL driver library</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="68"/>
        <source>Cannot load OpenCL library. Error while loading %1</source>
        <translation>Cannot load OpenCL library. Error while loading %1</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="75"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="82"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="89"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="96"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="105"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="112"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="119"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="126"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="133"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="140"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="147"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="154"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="161"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="168"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="175"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="182"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="189"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="196"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="203"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="210"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="217"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="224"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="231"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="238"/>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="245"/>
        <source>Cannot resolve symbol %1</source>
        <translation>Cannot resolve symbol %1</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="260"/>
        <source>Cannot load library: %1</source>
        <translation>Cannot load library: %1</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLHelper.cpp" line="263"/>
        <source>Some errors occurs in library: %1</source>
        <translation>Some errors occurs in library: %1</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusUtils.cpp" line="109"/>
        <source>Gaps:</source>
        <translation>Gaps:</translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="70"/>
        <source>clear OpenCL resources</source>
        <translation>clear OpenCL resources</translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="169"/>
        <source>GPU memory usage: %1 Mb</source>
        <translation>GPU memory usage: %1 Mb</translation>
    </message>
    <message>
        <location filename="../src/misc/BinaryFindOpenCL.cpp" line="183"/>
        <source>Device&apos;s preferred work group size multiple is %1</source>
        <translation>Device&apos;s preferred work group size multiple is %1</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLUtils.cpp" line="72"/>
        <source>OPENCL: BUILD LOG 
 ******************** 
 %1 
 ********************</source>
        <translation>OPENCL: BUILD LOG 
 ******************** 
 %1 
 ********************</translation>
    </message>
    <message>
        <location filename="../src/util_gpu/opencl/OpenCLUtils.cpp" line="73"/>
        <source>OPENCL: Program::build() failed. (%1)</source>
        <translation>OPENCL: Program::build() failed. (%1)</translation>
    </message>
</context>
<context>
    <name>U2::AlignInAminoFormTask</name>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="99"/>
        <source>Align in amino form</source>
        <translation>Align in amino form</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="109"/>
        <source>Invalid MSA object detected</source>
        <translation>Invalid MSA object detected</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="110"/>
        <source>AlignInAminoFormTask: Input alphabet is not nucleic!</source>
        <translation>AlignInAminoFormTask: Input alphabet is not nucleic!</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="111"/>
        <source>AlignInAminoFormTask: Input alignment is empty!</source>
        <translation>AlignInAminoFormTask: Input alignment is empty!</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="118"/>
        <source>Invalid applications settings detected</source>
        <translation>Invalid applications settings detected</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="121"/>
        <source>Invalid users applications settings detected</source>
        <translation>Invalid users applications settings detected</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="128"/>
        <source>Invalid MSA document detected</source>
        <translation>Invalid MSA document detected</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="148"/>
        <source>NULL clonedObj in AlignInAminoFormTask::prepare!</source>
        <translation>NULL clonedObj in AlignInAminoFormTask::prepare!</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="158"/>
        <source>Can not find row %1 in original alignment.</source>
        <translation>Can not find row %1 in original alignment.</translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="39"/>
        <source>Default</source>
        <translation>Default</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="43"/>
        <source>Returns simply the most frequent base and &apos;N&apos; are no reads intersecting this position</source>
        <translation>Returns simply the most frequent base and &apos;N&apos; are no reads intersecting this position</translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactorySamtools</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="45"/>
        <source>SAMtools</source>
        <translation>SAMtools</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="49"/>
        <source>Uses SAMtools to calculate consensus with regard to quality of reads</source>
        <translation>Uses SAMtools to calculate consensus with regard to quality of reads</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="71"/>
        <source>Fetching reads from database and converting to SAMtools format</source>
        <translation>Fetching reads from database and converting to SAMtools format</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="74"/>
        <source>Sorting reads</source>
        <translation>Sorting reads</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="77"/>
        <source>Calculating consensus</source>
        <translation>Calculating consensus</translation>
    </message>
</context>
<context>
    <name>U2::BestPositionFindTask</name>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="105"/>
        <source>Best position find task</source>
        <translation>Best position find task</translation>
    </message>
</context>
<context>
    <name>U2::CreateSubalignmentTask</name>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="55"/>
        <source>Create sub-alignment: %1</source>
        <translation>Create sub-alignment: %1</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="89"/>
        <source>Document is locked: %1</source>
        <translation>Document is locked: %1</translation>
    </message>
</context>
<context>
    <name>U2::DnaAssemblyMultiTask</name>
    <message>
        <location filename="../src/misc/DnaAssemblyMultiTask.cpp" line="86"/>
        <source>The short reads can&apos;t be mapped to the reference sequence!</source>
        <translation>The short reads can&apos;t be mapped to the reference sequence!</translation>
    </message>
</context>
<context>
    <name>U2::DnaAssemblyToReferenceTask</name>
    <message>
        <location filename="../src/registry/DnaAssemblyTask.cpp" line="30"/>
        <source>Align short reads</source>
        <translation>Align short reads</translation>
    </message>
</context>
<context>
    <name>U2::FindAlgorithmTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="44"/>
        <source>Find in sequence task</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAssemblyMultiTask</name>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="86"/>
        <source>Assembly cannot be performed.</source>
        <translation>Assembly cannot be performed.</translation>
    </message>
</context>
<context>
    <name>U2::LoadPatternsFileTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="107"/>
        <source>Load pattern from file</source>
        <translation>Load pattern from file</translation>
    </message>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="128"/>
        <source>Detecting format error for file %1</source>
        <translation>Detecting format error for file %1</translation>
    </message>
</context>
<context>
    <name>U2::MSAColorSchemeRegistry</name>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1003"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1018"/>
        <source>No colors</source>
        <translation>No colors</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1011"/>
        <source>Jalview</source>
        <translation>Jalview</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1013"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1052"/>
        <source>Percentage Identity</source>
        <translation>Percentage Identity</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1014"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1053"/>
        <source>Percentage Identity (gray)</source>
        <translation>Percentage Identity (gray)</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1026"/>
        <source>Zappo</source>
        <translation>Zappo</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1030"/>
        <source>Tailor</source>
        <translation>Tailor</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1034"/>
        <source>Hydrophobicity</source>
        <translation>Hydrophobicity</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1038"/>
        <source>Helix propensity</source>
        <translation>Helix propensity</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1042"/>
        <source>Strand propensity</source>
        <translation>Strand propensity</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1046"/>
        <source>Turn propensity</source>
        <translation>Turn propensity</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1050"/>
        <source>Buried index</source>
        <translation>Buried index</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1055"/>
        <source>Clustal X</source>
        <translation>Clustal X</translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryClustal</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmClustal.cpp" line="30"/>
        <source>Emulates ClustalW program and file format behavior.</source>
        <translation>Emulates ClustalW program and file format behavior.</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmClustal.cpp" line="34"/>
        <source>ClustalW</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmDefault.cpp" line="37"/>
        <source>Based on JalView algorithm. Returns &apos;+&apos; if there are 2 characters with high frequency. Returns symbol in lower case if the symbol content in a row is lower than the threshold specified.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmDefault.cpp" line="41"/>
        <source>Default</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryLevitsky</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmLevitsky.cpp" line="36"/>
        <source>The algorithm proposed by Victor Levitsky to work with DNA alignments.
Collects global alignment frequency for every symbol using extended (15 symbols) DNA alphabet first.
For every column selects the most rare symbol in the whole alignment with percentage in the column greater or equals to the threshold value.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmLevitsky.cpp" line="44"/>
        <source>Levitsky</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryStrict</name>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmStrict.cpp" line="39"/>
        <source>The algorithm returns gap character (&apos;-&apos;) if symbol frequency in a column is lower than threshold specified.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MSAConsensusAlgorithmStrict.cpp" line="43"/>
        <source>Strict</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithm</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithm.cpp" line="60"/>
        <source>MSA distance algorithm &quot;%1&quot; task</source>
        <translation>MSA distance algorithm &quot;%1&quot; task</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHamming</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHamming.cpp" line="36"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation>Based on Hamming distance between two sequences</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHamming.cpp" line="40"/>
        <source>Hamming dissimilarity</source>
        <translation>Hamming dissimilarity</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="42"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation>Based on Hamming distance between two sequences</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="46"/>
        <source>Hamming reverse-complement</source>
        <translation>Hamming reverse-complement</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactorySimilarity</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmSimilarity.cpp" line="36"/>
        <source>Based on similarity distance between two sequences</source>
        <translation>Based on similarity distance between two sequences</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmSimilarity.cpp" line="40"/>
        <source>Simple identity</source>
        <translation>Simple identity</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MSADistanceAlgorithmHammingRevCompl.cpp" line="78"/>
        <source>An unexpected error has occurred during running the Hamming reverse-complement algorithm.</source>
        <translation>An unexpected error has occurred during running the Hamming reverse-complement algorithm.</translation>
    </message>
</context>
<context>
    <name>U2::MSAHighlightingSchemeRegistry</name>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1062"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1063"/>
        <source>No highlighting</source>
        <translation>No highlighting</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1065"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1066"/>
        <source>Agreements</source>
        <translation>Agreements</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1068"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1069"/>
        <source>Disagreements</source>
        <translation>Disagreements</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1071"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1072"/>
        <source>Gaps</source>
        <translation>Gaps</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1074"/>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1075"/>
        <source>Conservation level</source>
        <translation>Conservation level</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1077"/>
        <source>Transitions</source>
        <translation>Transitions</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MSAColorScheme.cpp" line="1078"/>
        <source>Transversions</source>
        <translation>Transversions</translation>
    </message>
</context>
<context>
    <name>U2::MolecularSurfaceCalcTask</name>
    <message>
        <location filename="../src/molecular_geometry/MolecularSurface.cpp" line="122"/>
        <source>Molecular surface calculation</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ORFFindTask</name>
    <message>
        <location filename="../src/util_orf/ORFAlgorithmTask.cpp" line="31"/>
        <source>ORF find</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryBVH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmBVH.cpp" line="42"/>
        <source>Berg and von Hippel weight function</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryLOD</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="38"/>
        <source>log-odds</source>
        <translation>log-odds</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="42"/>
        <source>log-odds weight function</source>
        <translation>log-odds weight function</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryMCH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="37"/>
        <source>Match</source>
        <translation>Match</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="41"/>
        <source>Match weight function</source>
        <translation>Match weight function</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryNLG</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="37"/>
        <source>NLG</source>
        <translation>NLG</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="41"/>
        <source>NLG weight function</source>
        <translation>NLG weight function</translation>
    </message>
</context>
<context>
    <name>U2::PairwiseAlignmentTask</name>
    <message>
        <location filename="../src/pairwise_alignment/PairwiseAlignmentTask.cpp" line="48"/>
        <source>Pairwise alignment task</source>
        <translation>Pairwise alignment task</translation>
    </message>
</context>
<context>
    <name>U2::PhyTreeGeneratorLauncherTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="45"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation>Calculating Phylogenetic Tree</translation>
    </message>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="54"/>
        <source>Tree construction algorithm %1 not found</source>
        <translation>Tree construction algorithm %1 not found</translation>
    </message>
</context>
<context>
    <name>U2::PhyTreeGeneratorTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="32"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation>Calculating Phylogenetic Tree</translation>
    </message>
</context>
<context>
    <name>U2::SecStructPredictTask</name>
    <message>
        <location filename="../src/registry/SecStructPredictTask.cpp" line="27"/>
        <source>Secondary structure predict</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SequenceContentFilterTask</name>
    <message>
        <location filename="../src/misc/SequenceContentFilterTask.cpp" line="39"/>
        <source>Sequence content</source>
        <translation>Sequence content</translation>
    </message>
</context>
<context>
    <name>U2::SimpleAddToAlignmentTask</name>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="49"/>
        <source>Align sequences to an existing alignment by UGENE started</source>
        <translation>Align sequences to an existing alignment by UGENE started</translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackAnnotImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="67"/>
        <source>Annotation object not found.</source>
        <translation>Annotation object not found.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="71"/>
        <source>Annotation table is read-only</source>
        <translation>Annotation table is read-only</translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackMAImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="205"/>
        <source>SmithWatermanReportCallback failed to create new MA document</source>
        <translation>SmithWatermanReportCallback failed to create new MA document</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="215"/>
        <source>Failed to add a reference subsequence row.</source>
        <translation>Failed to add a reference subsequence row.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="219"/>
        <source>Failed to add a pattern subsequence row.</source>
        <translation>Failed to add a pattern subsequence row.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="222"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="308"/>
        <source>Failed to create an alignment.</source>
        <translation>Failed to create an alignment.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="251"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="328"/>
        <source>Smith-Waterman pairwise alignment: no results</source>
        <translation>Smith-Waterman pairwise alignment: no results</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="258"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="334"/>
        <source>Sequences aren&apos;t stored in the source msa dbi</source>
        <translation>Sequences aren&apos;t stored in the source msa dbi</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="272"/>
        <source>SmithWatermanReportCallback failed to create new MA document.</source>
        <translation>SmithWatermanReportCallback failed to create new MA document.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="278"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="280"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="344"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="346"/>
        <source>Failed to get the sequence object.</source>
        <translation>Failed to get the sequence object.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="294"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="296"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="360"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="363"/>
        <source>Failed to get the sequence data.</source>
        <translation>Failed to get the sequence data.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="303"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="305"/>
        <source>Failed to add row to result msa.</source>
        <translation>Failed to add row to result msa.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="372"/>
        <source>Failed to get msa from dbi</source>
        <translation>Failed to get msa from dbi</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="393"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="395"/>
        <source>Failed to update row gap model</source>
        <translation>Failed to update row gap model</translation>
    </message>
</context>
<context>
    <name>U2::SplicedAlignmentTask</name>
    <message>
        <location filename="../src/registry/SplicedAlignmentTask.cpp" line="27"/>
        <source>SplicedAlignmentTask</source>
        <translation>SplicedAlignmentTask</translation>
    </message>
</context>
<context>
    <name>U2::SubstMatrixRegistry</name>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="104"/>
        <source>Reading substitution matrix from %1</source>
        <translation>Reading substitution matrix from %1</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="112"/>
        <source>Error weight matrix file &apos;%1&apos; : %2</source>
        <translation>Error weight matrix file &apos;%1&apos; : %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="120"/>
        <source>Error opening file for read: %1</source>
        <translation>Error opening file for read: %1</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="127"/>
        <source>Error reading file: %1</source>
        <translation>Error reading file: %1</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="160"/>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="184"/>
        <source>Invalid character token &apos;%1&apos; , line %2</source>
        <translation>Invalid character token &apos;%1&apos; , line %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="165"/>
        <source>Duplicate character &apos;%1&apos; , line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="172"/>
        <source>Alphabet is neither nucleic nor protein!</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="176"/>
        <source>Matrix doesn&apos;t contains score for default alphabet character &apos;%1&apos;</source>
        <translation>Matrix doesn&apos;t contains score for default alphabet character &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="189"/>
        <source>Invalid character row &apos;%1&apos; , line %2</source>
        <translation>Invalid character row &apos;%1&apos; , line %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="193"/>
        <source>Duplicate character mapping &apos;%1&apos; , line %2</source>
        <translation>Duplicate character mapping &apos;%1&apos; , line %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="198"/>
        <source>Invalid number of columns &apos;%1&apos; , line %2</source>
        <translation>Invalid number of columns &apos;%1&apos; , line %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="206"/>
        <source>Can&apos;t parse numeric value &apos;%1&apos;, line %2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="215"/>
        <source>Unexpected end of file!</source>
        <translation>Unexpected end of file!</translation>
    </message>
</context>
<context>
    <name>U2::TranslateMSA2AminoTask</name>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="48"/>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="61"/>
        <source>Translate nucleic alignment to amino</source>
        <translation>Translate nucleic alignment to amino</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="50"/>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="63"/>
        <source>Invalid MSA object detected</source>
        <translation>Invalid MSA object detected</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="51"/>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="64"/>
        <source>Multiple alignment already has amino-acid alphabet</source>
        <translation>Multiple alignment already has amino-acid alphabet</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="55"/>
        <source>Unable to find suitable translation for %1</source>
        <translation>Unable to find suitable translation for %1</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MAlignmentUtilTasks.cpp" line="71"/>
        <source>Invalid translation object</source>
        <translation>Invalid translation object</translation>
    </message>
</context>
</TS>
