<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="zh_CN">
<context>
    <name>Document</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="205"/>
        <location filename="../src/PDWFormat.cpp" line="130"/>
        <source>Document is empty.</source>
        <translation>文档是空的。</translation>
    </message>
</context>
<context>
    <name>SQLiteL10n</name>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="127"/>
        <source>Failed to detect assembly storage format: %1</source>
        <translation type="unfinished">检测数据集存储类型 %1 失败</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="152"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="156"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="161"/>
        <source>Failed to parse packed row range info %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="155"/>
        <source>Attribute not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="210"/>
        <source>Unsupported attribute type: %1</source>
        <translation>不支持属性类型：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="183"/>
        <source>Not a %1 SQLite database: %2</source>
        <translation>%2 不是一个 %1 类型的 SQLite 数据库</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="189"/>
        <source>Warning! Database of version %1 was created with a newer %2 version: %3. Not all database features are supported!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="49"/>
        <source>ABIF</source>
        <translation>ABIF 格式</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="94"/>
        <source>Not a valid ABIF file: %1</source>
        <translation>不是一个合法的 ABIF 文件：%1</translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ACEFormat.cpp" line="43"/>
        <source>ACE</source>
        <translation>ACE 格式</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="209"/>
        <source>Line is too long</source>
        <translation>某一行太长</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="407"/>
        <source>First line is not an ace header</source>
        <translation>第一行不是 ace 的头</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="412"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished">标头行里没有 contig 的计数标志</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="426"/>
        <source>Must be CO keyword</source>
        <translation>必须是 CO 关键字</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="205"/>
        <source>Unexpected end of file</source>
        <translation>未知的文件结尾</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="440"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="221"/>
        <location filename="../src/ACEFormat.cpp" line="274"/>
        <source>There is no AF note</source>
        <translation>没有 AF 注释</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="225"/>
        <location filename="../src/ACEFormat.cpp" line="294"/>
        <source>A name is duplicated</source>
        <translation>有名称重复</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="233"/>
        <source>No consensus</source>
        <translation type="unfinished">不一致</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="244"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation>没有找到 BQ 关键字</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="249"/>
        <source>Bad consensus data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="281"/>
        <location filename="../src/ACEFormat.cpp" line="287"/>
        <source>Bad AF note</source>
        <translation>AF 注释错误</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="324"/>
        <source>There is no read note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="332"/>
        <source>No sequence</source>
        <translation>没有序列</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="343"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation>没有找到 QA 关键字</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="351"/>
        <location filename="../src/ACEFormat.cpp" line="357"/>
        <source>QA error no clear range</source>
        <translation>QA 错误：范围不清</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="363"/>
        <source>QA error bad range</source>
        <translation>QA 错误：范围错误</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="369"/>
        <source>Bad sequence data</source>
        <translation type="unfinished">序列质量欠佳</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="374"/>
        <source>A name is not match with AF names</source>
        <translation>有名称与 AF 名称不匹配</translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="485"/>
        <source>Alphabet unknown</source>
        <translation>序列的类型未知</translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="50"/>
        <source>MMDB</source>
        <translation>MMDB</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="74"/>
        <source>Standard residue dictionary not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="216"/>
        <location filename="../src/ASNFormat.cpp" line="615"/>
        <source>Unknown error occurred</source>
        <translation>发生未知错误</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="599"/>
        <source>no root element</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="605"/>
        <source>states stack is not empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="632"/>
        <source>First line is too long</source>
        <translation>首行过长</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="146"/>
        <source>read error occurred</source>
        <translation>发生读取错误</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="154"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation>导入 biostruct3d 对象时发生错误：%1</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="50"/>
        <source>CLUSTALW</source>
        <translation>CLUSTALW 格式</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="73"/>
        <source>Illegal header line</source>
        <translation>非法的标头行</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="93"/>
        <source>Error parsing file</source>
        <translation>解析文件时发生错误</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="105"/>
        <source>Invalid alignment format</source>
        <translation>非法的比对格式</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="146"/>
        <source>Incorrect number of sequences in block</source>
        <translation>区块中的序列数量有误</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="152"/>
        <source>Sequence names are not matched</source>
        <translation>序列名称不对应</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="176"/>
        <source>Alphabet is unknown</source>
        <translation>序列的类型未知</translation>
    </message>
</context>
<context>
    <name>U2::Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="211"/>
        <source>Document is empty.</source>
        <translation>文档是空的。</translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormatUtils</name>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="67"/>
        <source>Maximum number of objects per document limit reached. Try different options for opening the document!</source>
        <translation>已经超过每个文档允许的对象数目了，请尝试其它选项以打开该文档!</translation>
    </message>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="74"/>
        <source>Undefined sequence alphabet</source>
        <translation>无法确定序列类型</translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="120"/>
        <source>Reading entry header</source>
        <translation>正在读取标头行</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="167"/>
        <source>Merge error: found annotations without sequence</source>
        <translation>合并时发生错误：有一些注释没有对应的序列</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="366"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="295"/>
        <source>Annotation name is empty</source>
        <translation>注释的名称为空</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="374"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="380"/>
        <source>Error parsing location</source>
        <translation>解析位置时发生错误</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="406"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="349"/>
        <source>Unexpected line format</source>
        <translation>位置的行格式</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="451"/>
        <source>Reading sequence %1</source>
        <translation>正在读取序列 %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="473"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>解析序列时发生错误：存在未知的空行</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="500"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation>读取序列时发生错误：序列格式不合法</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="515"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>读取序列时发生错误：分配内存失败</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="522"/>
        <source>Sequence is truncated</source>
        <translation>序列已被截短</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="529"/>
        <source>Reading annotations %1</source>
        <translation>正在读取注释 %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="539"/>
        <source>Invalid format of feature table</source>
        <translation>特征列表的格式不合法</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="581"/>
        <source>Line is too long.</source>
        <translation>某一行过长。</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="583"/>
        <source>IO error.</source>
        <translation>IO（输入/输出）错误。</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="222"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="211"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="201"/>
        <source>Record is truncated.</source>
        <translation>记录已被截短。</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="304"/>
        <source>Annotation start position is empty</source>
        <translation>某个注释的起始位置为空</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="309"/>
        <source>Annotation end position is empty</source>
        <translation>某个注释的终止位置为空</translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="43"/>
        <source>EMBL</source>
        <translation>EMBL 格式</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="46"/>
        <source>EMBL Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation>EMBL 平文本文件格式是一种丰富型的序列格式，可存储序列以及与其相关的注释</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="83"/>
        <source>ID is not the first line</source>
        <translation>第一行不是序列的 ID 号</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="90"/>
        <source>Error parsing ID line</source>
        <translation>解析 ID 行时发生错误</translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="47"/>
        <source>FASTA</source>
        <translation>FASTA 格式</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="50"/>
        <source>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</source>
        <translation>FASTA 是一种基于文本的核酸序列或蛋白质序列格式，每个碱基或氨基酸都使用单个字母来表示。该格式还可以在序列前面添加必要的名称和注释等。</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="146"/>
        <location filename="../src/FastaFormat.cpp" line="313"/>
        <source>Line is too long</source>
        <translation>某一行太长</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="150"/>
        <location filename="../src/FastaFormat.cpp" line="317"/>
        <source>First line is not a FASTA header</source>
        <translation>第一行不是 FASTA 格式的头</translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <location filename="../src/FastqFormat.cpp" line="41"/>
        <source>FASTQ</source>
        <translation>FASTQ 格式</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="44"/>
        <source>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data, but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</source>
        <translation>FASTQ 是一种基于文本的序列格式，其存储的信息包括生物序列(一般是核酸序列)以及对应的测序质量评价得分。为了简单起见，序列的每个碱基和质量得分都分别以一个 ASCII 字符来表示。这种格式最初来源于 Wellcome Trust Sanger Institute，用于绑定 FASTA 序列与其对应的质量评价数据，但现在已经成为存储高通量测序结果的事实标准。</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="101"/>
        <source>Unexpected end of file</source>
        <translation>未知的文件结尾</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="174"/>
        <source>Not a valid FASTQ file: %1. The @ identifier is not found.</source>
        <translation>不是一个合法的 FASTQ 文件：%1，因为没有找到 @ 标识符。</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="198"/>
        <source>Not a valid FASTQ file: %1</source>
        <translation>不是一个合法的 FASTQ 文件：%1</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="210"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation>不是一个合法的 FASTQ 文件：%1，质量评分有问题，与序列的长度不一致。</translation>
    </message>
</context>
<context>
    <name>U2::GBFeatureUtils</name>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="75"/>
        <source>Sequence related to transcription termination</source>
        <translation>与转录终止相关的序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="76"/>
        <source>Span of the C immunological feature</source>
        <translation>C-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="77"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation>真核基因启动子中的 CAAT 盒(CAAT box)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="78"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation>蛋白质编码区(包括终止密码)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="80"/>
        <source>Independent sequence determinations differ</source>
        <translation>不同测定结果所得的差异序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="81"/>
        <source>Displacement loop</source>
        <translation>D 环</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="82"/>
        <source>Span of the D immunological feature</source>
        <translation>D-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="83"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation>顺式作用增强子，强化启动子功能</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="84"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation>外显子，成熟 mRNA 的一部分</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="85"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation>基因区域，包括上游启动子、增强子和下游控制区。</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="87"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation>真核生物启动子中 GC 盒</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="88"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation>重组引入的 DNA 插入区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="89"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation>内含子，在 mRNA 剪切过程中被切除的转录区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="90"/>
        <source>Span of the J immunological feature</source>
        <translation>J-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="91"/>
        <source>Long terminal repeat</source>
        <translation>长终止重复序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="92"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation>成熟肽编码区(不包括终止密码)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="93"/>
        <source>Miscellaneous binding site</source>
        <translation>其它结合位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="94"/>
        <source>Miscellaneous difference feature</source>
        <translation>其它特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="95"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation>其它尚未确定的重要生物功能区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="97"/>
        <source>Miscellaneous, recombination feature</source>
        <translation>其它重组特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="98"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation>其它尚未确定的转录特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="99"/>
        <source>Miscellaneous signal</source>
        <translation>其它信号区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="100"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation>其它 DNA 或 RNA 结构</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="101"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation>修饰碱基</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="102"/>
        <source>Messenger RNA</source>
        <translation>信使 RNA 区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="103"/>
        <source>Span of the N immunological feature</source>
        <translation>N-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="104"/>
        <source>Presented sequence revises a previous version</source>
        <translation>旧版本序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="105"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation>剪切 &amp; 多聚 A 信号区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="106"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation>mRNA 的多聚 A 添加位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="107"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation>前体 RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="108"/>
        <source>Primary (unprocessed) transcript</source>
        <translation>初始(未处理)转录产物</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="109"/>
        <source>Primer binding region used with PCR</source>
        <translation>PCR 引物结合位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="110"/>
        <source>Non-covalent primer binding site</source>
        <translation>引物非共价结合位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="111"/>
        <source>A region involved in transcription initiation</source>
        <translation>启动子区域，参与转录起始过程</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="112"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation>蛋白质非共价结合位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="113"/>
        <source>Ribosome binding site</source>
        <translation>核糖体结合位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="114"/>
        <source>Replication origin for duplex DNA</source>
        <translation>DNA 双链复制起始区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="115"/>
        <source>Sequence containing repeated subsequences</source>
        <translation>重复序列区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="116"/>
        <source>One repeated unit of a repeat_region</source>
        <translation>重复序列区域的重复单位</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="117"/>
        <source>Ribosomal RNA</source>
        <translation>核糖体 RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="118"/>
        <source>Span of the S immunological feature</source>
        <translation>S-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="119"/>
        <source>Satellite repeated sequence</source>
        <translation>卫星 DNA 重复序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="120"/>
        <source>Small cytoplasmic RNA</source>
        <translation>胞浆内小 RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="121"/>
        <source>Signal peptide coding region</source>
        <translation>信号肽编码区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="122"/>
        <source>Small nuclear RNA</source>
        <translation>核内小 RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="123"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation>该序列的来源</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="124"/>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation>DNA 或 RNA 中的发夹环结构</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="125"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation>序列标签位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="126"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation>真核生物启动子中 TATA 盒(TATA box)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="127"/>
        <source>Sequence causing transcription termination</source>
        <translation>转录终止位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="128"/>
        <source>Transit peptide coding region</source>
        <translation>转运肽编码区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="129"/>
        <source>Transposable element (TN)</source>
        <translation>转座子(TN)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="130"/>
        <source>Transfer RNA</source>
        <translation>转移 RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="131"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation>未确定区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="132"/>
        <source>Span of the V immunological feature</source>
        <translation>V-免疫特征区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="133"/>
        <source>A related population contains stable mutation</source>
        <translation>变异区</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="134"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation>原核生物启动子中的 -10 位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="135"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation>原核生物启动子中的 -35 位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="136"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation>前体 mRNA 在转录后加工时切除的 3’端区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="137"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation>3’ 端不翻译区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="138"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation>前体 mRNA 在转录后加工时切除的 5’端区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="139"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation>5’ 端不翻译区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="140"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation>蛋白质</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="141"/>
        <source>&apos;Region&apos; feature key</source>
        <translation>区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="142"/>
        <source>&apos;Site&apos; feature key</source>
        <translation>位点</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="161"/>
        <source>Genes</source>
        <translation>基因</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="176"/>
        <source>Signals</source>
        <translation>信号</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="195"/>
        <source>Binding</source>
        <translation>绑定</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="200"/>
        <source>Variation</source>
        <translation>变异</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="209"/>
        <source>Repeats</source>
        <translation>重复序列</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="216"/>
        <source>RNA</source>
        <translation>RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="224"/>
        <source>Misc</source>
        <translation>未知区域</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="238"/>
        <source>Spans</source>
        <translation>序列范围</translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="45"/>
        <source>GFF</source>
        <translation>GFF 格式</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="72"/>
        <source>Parsing error: invalid header</source>
        <translation>解析错误：不合法的标头</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="76"/>
        <source>Parsing error: file does not contain version header</source>
        <translation>解析错误：文件头部没有包含版本号</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="80"/>
        <source>Parsing error: format version is not an integer</source>
        <translation>解析错误：版本号出现了非整数值</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="84"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation>解析错误：不支持 GFF 版本 %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="160"/>
        <source>Parsing error: sequence in FASTA sequence has whitespaces at line %1</source>
        <translation>解析错误：FASTA 序列的第 %1 行出现了空格</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="167"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation>解析错误：第 %1 行的条目太少</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="173"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation>解析错误：第 %1 行的起始位置不是一个整数</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="179"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation>解析错误：第 %1 行的结束位置不是一个整数</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="184"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation>解析错误：第 %1 行的注释区域不正确</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="203"/>
        <source>Parsing error: one of the qualifiers in attributes are incorrect at line %1</source>
        <translation>解析错误：第 %1 行的属性值有错误</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="218"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation>第 %1 行在连接注释时位置错误，已跳过该行。</translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="51"/>
        <source>Genbank</source>
        <translation>Genbank 格式</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="53"/>
        <source>GenBank Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation>GenBank 平文本文件格式是一种丰富型的序列格式，可存储序列以及与其相关的注释</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="86"/>
        <source>LOCUS is not the first line</source>
        <translation>第一行不是 LOCUS 项</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="93"/>
        <source>Error parsing LOCUS line</source>
        <translation>解析 LOCUS 这一行时发生错误</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="155"/>
        <source>incomplete SOURCE record</source>
        <translation>SOURCE(来源) 记录不完整</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="253"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="477"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="490"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="501"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="528"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="537"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="544"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="553"/>
        <source>Error writing document</source>
        <translation>写入文档时发生错误</translation>
    </message>
</context>
<context>
    <name>U2::IndexFormat</name>
    <message>
        <location filename="../src/IndexFormat.cpp" line="127"/>
        <source>Empty io identifier in item section</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="129"/>
        <source>Empty document format in item section </source>
        <translation>条目栏中文件格式为空</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="131"/>
        <source>Empty io adapter identifier in io section</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="133"/>
        <source>Empty url in io section</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="135"/>
        <source>Bad offset in item section</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="137"/>
        <source>Unrecognized io section id in item section</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="140"/>
        <source>Unknown index error occurred</source>
        <translation>发生未知的索引错误</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="256"/>
        <location filename="../src/IndexFormat.cpp" line="262"/>
        <source>Bad value in item section: tag</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="268"/>
        <location filename="../src/IndexFormat.cpp" line="272"/>
        <source>Bad value in item section: val</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="350"/>
        <source>invalid header</source>
        <translation>非法的头</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="395"/>
        <source>Unknown section found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="404"/>
        <source>no items found in index file</source>
        <translation>索引文件中没有找到条目</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="484"/>
        <source>Index</source>
        <translation>索引</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="504"/>
        <location filename="../src/IndexFormat.cpp" line="522"/>
        <source>Unknown error occurred</source>
        <translation>发生未知错误</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.h" line="66"/>
        <source>Read error occurred</source>
        <translation>发生读取错误</translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.h" line="74"/>
        <source>Write error occurred</source>
        <translation>发生写入错误</translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="56"/>
        <source>MSF</source>
        <translation>MSF 格式</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="58"/>
        <source>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</source>
        <translation>MSF 是用于存储多序列比对的一种格式。文件包含序列的名称和序列本身，并且与文件中的其它序列进行了比对。</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="118"/>
        <source>Incorrect format</source>
        <translation>错误的格式</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="136"/>
        <source>Unexpected end of file</source>
        <translation>未知的文件结尾</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="161"/>
        <location filename="../src/MSFFormat.cpp" line="200"/>
        <source>Check sum test failed</source>
        <translation>未能通过求和测试</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="209"/>
        <source>Alphabet unknown</source>
        <translation>序列的类型未知</translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="44"/>
        <source>Mega</source>
        <translation>Mega 格式</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="95"/>
        <source>Found sequences of different sizes</source>
        <translation>有若干条序列长度不等</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="101"/>
        <source>Model is of zero size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="159"/>
        <source>Bad name of sequence</source>
        <translation>序列名称非法</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="179"/>
        <source>Unexpected # in comments</source>
        <translation>注释中出现了未知的 # 号</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="188"/>
        <source>A comment has not end</source>
        <translation>某条注释没有结尾</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="210"/>
        <source>Unexpected symbol between comments</source>
        <translation>注释之间出现了未知符号</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="270"/>
        <source>Incorrect format</source>
        <translation>错误的格式</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="286"/>
        <source>Identical symbol at the first sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="308"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation>序列名称的顺序不正确</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="327"/>
        <source>Alphabet unknown</source>
        <translation>序列的类型未知</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="392"/>
        <location filename="../src/MegaFormat.cpp" line="404"/>
        <source>No header</source>
        <translation>没有头</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="397"/>
        <source>No # before header</source>
        <translation>头的前面没有 # 符号</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="409"/>
        <source>Not MEGA-header</source>
        <translation>不是 MEGA-头</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="432"/>
        <location filename="../src/MegaFormat.cpp" line="443"/>
        <location filename="../src/MegaFormat.cpp" line="470"/>
        <source>No data in file</source>
        <translation>文件中没有数据</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="452"/>
        <location filename="../src/MegaFormat.cpp" line="458"/>
        <source>Incorrect title</source>
        <translation>标题不正确</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="502"/>
        <source>Sequence has empty part</source>
        <translation>序列有空白区</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="522"/>
        <source>Bad symbols in a sequence</source>
        <translation>某条序列中有非法字符</translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="43"/>
        <source>NEXUS</source>
        <translation>NEXUS 格式</translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="627"/>
        <source>#NEXUS header missing</source>
        <translation>#NEXUS 标头缺失</translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="37"/>
        <source>Newick Standard</source>
        <translation>标准 Newick 格式</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="38"/>
        <source>Newick is a simple format used to write out trees in a text file</source>
        <translation>Newick 是一种在文本文件中描述进化树的简单格式</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="200"/>
        <source>Unexpected weight: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="205"/>
        <source>Error parsing weight: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="221"/>
        <source>Unexpected weight start token: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="229"/>
        <source>Unexpected new sibling %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="242"/>
        <source>Unexpected closing bracket :%1</source>
        <translation>未知的闭合括号：%1</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="250"/>
        <location filename="../src/NewickFormat.cpp" line="272"/>
        <source>Unexpected end of file</source>
        <translation>未知的文件结尾</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="284"/>
        <source>Empty file</source>
        <translation>文件为空</translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <location filename="../src/PDBFormat.cpp" line="53"/>
        <source>PDB</source>
        <translation>PDB 格式</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="54"/>
        <source>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</source>
        <translation>PDB(蛋白质数据库)格式提供了蛋白分子结构的标准，多来源于 X 射线衍射或者 NMR 研究。</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="191"/>
        <source>Line is too long</source>
        <translation>某一行太长</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="227"/>
        <source>SPLIT recored is not supported</source>
        <translation>不支持 SPLIT 记录</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="249"/>
        <source>Some mandaroty records are absent</source>
        <translation>缺失一些必需的记录</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="358"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation>PDB 警告：未知的残基名称：%1</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="455"/>
        <source>Invalid secondary structure record</source>
        <translation>存在非法的二级结构记录</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="596"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation>存在非法的 SEQRES：少于 24 个氨基酸</translation>
    </message>
</context>
<context>
    <name>U2::PDWFormat</name>
    <message>
        <location filename="../src/PDWFormat.cpp" line="55"/>
        <source>pDRAW</source>
        <translation>pDRAW 格式</translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="92"/>
        <location filename="../src/PDWFormat.cpp" line="180"/>
        <source>Line is too long</source>
        <translation>某一行太长</translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="38"/>
        <source>Plain text</source>
        <translation>平文本格式</translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="40"/>
        <source>A simple plain text file.</source>
        <translation>简单的平文本文件。</translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="42"/>
        <source>Raw sequence</source>
        <translation>原始序列格式</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="45"/>
        <source>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</source>
        <translation>原始序列文件 - 文件的所有内容被视为单独的一条核酸序列或多肽序列，UGENE 会移除所有非字母的字符</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="85"/>
        <source>Sequence is empty</source>
        <translation>序列为空</translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="80"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation>栏目“%1”的模式“%2”没有匹配，应该匹配“%3”</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="89"/>
        <source>SAM</source>
        <translation>SAM 格式</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="207"/>
        <source>Unsupported file version &quot;%1&quot;</source>
        <translation>不支持版本“%1”</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="261"/>
        <source>Unexpected end of file</source>
        <translation>未知的文件结尾</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="292"/>
        <source>Can&apos;t find alphabet for sequence &quot;%1&quot;</source>
        <translation>无法找到序列“%1”的类型</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="297"/>
        <source>Can&apos;t translation for alphabet &quot;%1&quot;</source>
        <translation>无法翻译序列“%1”</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="330"/>
        <location filename="../src/SAMFormat.cpp" line="340"/>
        <source>Alphabet is unknown</source>
        <translation>序列的类型未知</translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="45"/>
        <source>SCF</source>
        <translation>SCF 格式</translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="85"/>
        <source>Failed to parse SCF file: %1</source>
        <translation>解析 SCF 格式时发生错误：%1</translation>
    </message>
</context>
<context>
    <name>U2::SQLiteL10N</name>
    <message>
        <location filename="../src/sqlite_dbi/assembly/RTreeAssemblyAdapter.cpp" line="65"/>
        <source>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="82"/>
        <source>There is no assembly object with the specified id.</source>
        <translation>对于该 id 号没有合适的对象。</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="98"/>
        <source>Unsupported reads storage type: %1</source>
        <translation>不支持读取存储类型 %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="119"/>
        <source>Assembly object not found.</source>
        <translation>未能找到合适的对象。</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="330"/>
        <source>Packing method is not supported: %1</source>
        <translation>不支持该种打包方式：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="381"/>
        <source>Packed data is empty!</source>
        <translation>打包的数据为空!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="386"/>
        <source>Packing method prefix is not supported: %1</source>
        <translation type="unfinished">不支持该种打包方式索引：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="392"/>
        <source>Data is corrupted, no name end marker found: %1</source>
        <translation>数据有损坏，未能找到名称结束标记：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="400"/>
        <source>Data is corrupted, no sequence end marker found: %1</source>
        <translation>数据有损坏，未能找到序列结束标记：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="408"/>
        <source>Data is corrupted, no CIGAR end marker found: %1</source>
        <translation>数据有损坏，未能找到 CIGAR 结束标记：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="142"/>
        <source>Error checking SQLite database: %1!</source>
        <translation>检查 SQLite 数据库时发生错误：%1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="155"/>
        <source>Error creating table: %1, error: %2</source>
        <translation>创建表格 %1 时发生错误，错误类型：%2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="233"/>
        <source>Database is already opened!</source>
        <translation>数据库已经打开!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="237"/>
        <source>Illegal database state: %1</source>
        <translation>非法的数据库状态：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="243"/>
        <source>URL is not specified</source>
        <translation>未指定 URL 地址</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="257"/>
        <source>Error opening SQLite database: %1!</source>
        <translation>打开 SQLite 数据库时发生错误：%1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="302"/>
        <source>Database is already closed!</source>
        <translation>数据库已经关闭!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="306"/>
        <source>Illegal database state %1!</source>
        <translation>非法的数据库状态 %1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="323"/>
        <source>Failed to close database: %1, err: %2</source>
        <translation>关闭数据库 %1 时发生错误，错误：%2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="129"/>
        <source>Not an object! Id: %1, type: %2</source>
        <translation>ID:%1，类型：%2 不是一个对象！</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="144"/>
        <source>Object: %1 not found in folder: %2</source>
        <translation>没有在目录 %2 中找到对象 %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="197"/>
        <source>Unknown object type! Id: %1, type: %2</source>
        <translation>ID:%1，类型：%2 对象类型未知！</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="439"/>
        <source>Object not found.</source>
        <translation>没有找到对象。</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="462"/>
        <source>Folder not found :%1</source>
        <translation>没有找到文件夹：%1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSequenceDbi.cpp" line="64"/>
        <source>Sequence object not found.</source>
        <translation>没有找到序列对象。</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="74"/>
        <source>Sequence object is not set!</source>
        <translation>没有设置序列对象!</translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="462"/>
        <source>invalid file: bad header line</source>
        <translation>文件非法：标头行有错误</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="492"/>
        <source>invalid file: empty sequence name</source>
        <translation>文件非法：缺失序列名字</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="495"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation>文件非法：某个区块内出现了相同的序列名字</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="503"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation>文件非法：序列名字在各区块之间不一致</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="514"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation>文件非法：区块内的序列不等长</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="521"/>
        <source>invalid file: empty sequence alignment</source>
        <translation>文件非法：存在空的序列比对</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="525"/>
        <source>invalid file: unknown alphabet</source>
        <translation>文件非法：未知的序列类型</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="674"/>
        <source>Stockholm</source>
        <translation>Stockholm 格式</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="699"/>
        <location filename="../src/StockholmFormat.cpp" line="718"/>
        <source>unknown error occurred</source>
        <translation>发生了未知错误</translation>
    </message>
</context>
<context>
    <name>U2::SwissProtPlainTextFormat</name>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="44"/>
        <source>Swiss-Prot</source>
        <translation>Swiss-Prot 格式</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="81"/>
        <source>ID is not the first line</source>
        <translation>第一行不是序列的 ID 号</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="88"/>
        <source>Error parsing ID line</source>
        <translation>解析 ID 行时发生错误</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="98"/>
        <source>Error parsing ID line. Not found sequence length</source>
        <translation>解析 ID 行时发生错误：未找到序列长度</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="210"/>
        <source>Reading sequence %1</source>
        <translation>正在读取序列 %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="230"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>解析序列时发生错误：存在未知的空行</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="249"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>读取序列时发生错误：分配内存失败</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="256"/>
        <source>Sequence is truncated</source>
        <translation>序列已被截短</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="262"/>
        <source>Reading annotations %1</source>
        <translation>正在读取注释 %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="269"/>
        <source>Invalid format of feature table</source>
        <translation>特征列表的格式不合法</translation>
    </message>
</context>
</TS>
