<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="cs_CZ">
<context>
    <name>Document</name>
    <message>
        <source>Document is empty.</source>
        <translation type="vanished">Dokument je prázdný.</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/BAMUtils.cpp" line="99"/>
        <source>Fail to open &quot;%1&quot; for reading</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="103"/>
        <source>Fail to read the header from the file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="107"/>
        <source>Can not build the fasta index for the file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="111"/>
        <source>Error parsing the reads from the file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="115"/>
        <source>Truncated file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="84"/>
        <source>BED</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="76"/>
        <source>FPKM Tracking Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="84"/>
        <source>GTF</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="473"/>
        <source>Invalid database user permissions set, so UGENE unable to use this database. Connect to your system administrator to fix the issue.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="766"/>
        <source>Unexpected number of modified objects. Expected: 1, actual: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteBlobInputStream.cpp" line="100"/>
        <source>Can not read data. The database is closed or the data were changed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteBlobOutputStream.cpp" line="50"/>
        <source>Can not write data. The database is closed or the data were changed.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="214"/>
        <source>File &quot;%1&quot; contains too many sequences to be displayed. However, you can process these data using instruments from the menu &lt;i&gt;Tools -&gt; NGS data analysis&lt;/i&gt; or pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="50"/>
        <source>ABIF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="51"/>
        <source>A chromatogram file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="90"/>
        <source>Not a valid ABIF file: %1</source>
        <translation>Není platným souborem ABIF: %1</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="118"/>
        <source>Failed to load sequence from ABI file %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="50"/>
        <source>ACE</source>
        <translation>ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="219"/>
        <source>Line is too long</source>
        <translation>Řádek je příliš dlouhý</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="384"/>
        <source>A name is not match with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="427"/>
        <source>First line is not an ace header</source>
        <translation>První řádek není hlavička ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="432"/>
        <source>No contig count tag in the header line</source>
        <translation>Žádná počtu contig v řádku s hlavičkou</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="446"/>
        <source>Must be CO keyword</source>
        <translation>Musí být klíčové slovo CO</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="215"/>
        <source>Unexpected end of file</source>
        <translation>Neočekávaný konec souboru</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="460"/>
        <source>There is no note about reads count</source>
        <translation>Není tu žádná poznámka o počtu čtení</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="231"/>
        <location filename="../src/ace/AceFormat.cpp" line="284"/>
        <source>There is no AF note</source>
        <translation>Není tu žádná poznámka AF</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="51"/>
        <source>ACE is a format used for storing information about genomic confgurations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="235"/>
        <location filename="../src/ace/AceFormat.cpp" line="304"/>
        <source>A name is duplicated</source>
        <translation>Název je zdvojen</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="243"/>
        <source>No consensus</source>
        <translation>Žádná shoda</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="254"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation>Klíčové slovo BQ nebylo nalezeno</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="259"/>
        <source>Bad consensus data</source>
        <translation>Špatná data shody</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="291"/>
        <location filename="../src/ace/AceFormat.cpp" line="297"/>
        <source>Bad AF note</source>
        <translation>Špatná poznámka AF</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="334"/>
        <source>There is no read note</source>
        <translation>Není tu žádná poznámka o čtení</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="342"/>
        <source>No sequence</source>
        <translation>Žádná sekvence</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="353"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation>Klíčové slovo QA nebylo nalezeno</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="361"/>
        <location filename="../src/ace/AceFormat.cpp" line="367"/>
        <source>QA error no clear range</source>
        <translation>Chyba QA žádný jasný rozsah</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="373"/>
        <source>QA error bad range</source>
        <translation>Chyba QA špatný rozsah</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="379"/>
        <source>Bad sequence data</source>
        <translation>Špatná data sekvence</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="539"/>
        <source>File doesn&apos;t contain any msa objects</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>A name is not mutch with AF names</source>
        <translation type="vanished">Název není hodně s názvy AF</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="513"/>
        <source>Alphabet unknown</source>
        <translation>Neznámá abeceda</translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="50"/>
        <source>MMDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="147"/>
        <source>read error occurred</source>
        <translation>Vyskytla se chyba čtení</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="155"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation>Chyba při nahrávání biostruct3d obj: %1</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="51"/>
        <source>ASN is a format used my the Molecular Modeling Database (MMDB)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="206"/>
        <location filename="../src/ASNFormat.cpp" line="605"/>
        <source>Unknown error occurred</source>
        <translation>Vyskytla se neznámá chyba</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="589"/>
        <source>no root element</source>
        <translation>žádný kořenový prvek</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="595"/>
        <source>states stack is not empty</source>
        <translation>Zásobník stavů není prázdný</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="622"/>
        <source>First line is too long</source>
        <translation>První řádek je příliš dlouhý</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="74"/>
        <source>Standard residue dictionary not found</source>
        <translation>Běžný slovník reziduí nenalezen</translation>
    </message>
</context>
<context>
    <name>U2::AbstractVariationFormat</name>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="61"/>
        <source>SNP formats are used to store single-nucleotide polymorphism data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="119"/>
        <location filename="../src/AbstractVariationFormat.cpp" line="132"/>
        <source>Line %1: There are too few columns in this line. The line was skipped.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AceImporter</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="161"/>
        <source>ACE file importer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="165"/>
        <source>ACE files importer is used to convert conventional ACE files into UGENE database format.Having ACE file converted into UGENE DB format you get an fast and efficient interface to your data with an option to change the content</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AceImporterTask</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="51"/>
        <source>ACE file import: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="86"/>
        <source>Can&apos;t create a temporary database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="146"/>
        <source>Failed to get load task for : %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BAMUtils</name>
    <message>
        <location filename="../src/BAMUtils.cpp" line="178"/>
        <source>A SAM file has not a header [%1]. The list of references will be generated.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="336"/>
        <source>Sort bam file: &quot;%1&quot; using %2 Mb of memory. Result sorted file is: &quot;%3&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="347"/>
        <source>Merging BAM files: &quot;%1&quot;. Resulting merged file is: &quot;%2&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="364"/>
        <source>Remove PCR duplicate in BAM file: &quot;%1&quot;. Resulting  file is: &quot;%2&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="451"/>
        <source>Build index for bam file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="712"/>
        <source>Wrong line in a SAM file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BedFormat</name>
    <message>
        <location filename="../src/BedFormat.cpp" line="103"/>
        <source>The BED (Browser Extensible Data) format was developed by UCSC for displaying transcript structures in the genome browser.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="139"/>
        <source>File &quot;%1&quot; contains too many annotation tables to be displayed. However, you can process these data using pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="614"/>
        <source>BED parsing error: incorrect format of the &apos;track&apos; header line!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="647"/>
        <source>BED parsing error: unexpected number of fields in the first annotations line!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="655"/>
        <source>BED parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="661"/>
        <source>BED parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="667"/>
        <source>BED parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="724"/>
        <source>BED parsing error: incorrect score value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="729"/>
        <source>BED parsing error: incorrect strand value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="734"/>
        <source>BED parsing error: incorrect thick coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="738"/>
        <source>BED parsing error: incorrect itemRgb value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="743"/>
        <source>BED parsing error: incorrect value of the block parameters at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="772"/>
        <source>Starting BED saving: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="783"/>
        <source>Can not convert GObject to AnnotationTableObject</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="797"/>
        <source>Can not detect chromosome name. &apos;Chr&apos; name will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="814"/>
        <source>You are trying to save joined annotation to BED format! The joining will be lost</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="878"/>
        <source>BED saving error: incorrect thick coordinates in the first annotation!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="892"/>
        <source>BED saving error: incorrect block fields in the first annotation!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="901"/>
        <source>BED saving: detected %1 fields per line for file &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="908"/>
        <source>BED saving error: an annotation is expected to have &apos;%1&apos; qualifier, but it is absent! Skipping the annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="968"/>
        <source>BED saving error: an annotation is expected to have the block qualifiers! Skipping the annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="991"/>
        <source>Finished BED saving: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BgzipTask</name>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="43"/>
        <source>Bgzip Compression task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="50"/>
        <source>Start bgzip compression &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="52"/>
        <source>IOAdapterRegistry is NULL!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="55"/>
        <source>IOAdapterFactory is NULL!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="57"/>
        <source>Can not create IOAdapter!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="61"/>
        <source>Can not open input file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="71"/>
        <source>Can not open output file &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="85"/>
        <source>Error reading file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="90"/>
        <source>Error writing to file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="97"/>
        <source>Bgzip compression finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="102"/>
        <source>Bgzip compression task was finished with an error: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="104"/>
        <source>Bgzip compression task was finished. A new bgzf file is: &lt;a href=&quot;%1&quot;&gt;%2&lt;/a&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CloneAssemblyWithReferenceToDbiTask</name>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="43"/>
        <source>Clone assembly object to the destination database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="52"/>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="53"/>
        <source>Invalid assembly ID</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="54"/>
        <source>Invalid source database reference</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="55"/>
        <source>Invalid destination database reference</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="73"/>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="79"/>
        <source>Can&apos;t get the cloned object</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="75"/>
        <source>Unexpected result object: expect AssemblyObject, got %1 object</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="81"/>
        <source>Unexpected result object: expect U2SequenceObject, got %1 object</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <source>found_sequences_of_different_size</source>
        <translation type="obsolete">Found sequences of different sizes</translation>
    </message>
    <message>
        <source>model_of_0_size</source>
        <translation type="obsolete">Model is of zero size</translation>
    </message>
    <message>
        <source>illegal_header_line</source>
        <translation type="obsolete">Illegal header line</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="60"/>
        <source>Clustalw is a format for storing multiple sequence alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="84"/>
        <source>Illegal header line</source>
        <translation>Neplatný řádek záhlaví</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="104"/>
        <source>Error parsing file</source>
        <translation>Chyba při zpracování souboru</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="116"/>
        <source>Invalid alignment format</source>
        <translation>Neplatný formát zarovnání</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="157"/>
        <source>Incorrect number of sequences in block</source>
        <translation>Nesprávný počet sekvencí v bloku</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="163"/>
        <source>Sequence names are not matched</source>
        <translation>Názvy sekvencí se neshodují</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="187"/>
        <source>Alphabet is unknown</source>
        <translation>Abeceda je neznámá</translation>
    </message>
    <message>
        <source>error_parsing_file</source>
        <translation type="obsolete">Error parsing file</translation>
    </message>
    <message>
        <source>invalid_alignment_format</source>
        <translation type="obsolete">Invalid alignment format</translation>
    </message>
    <message>
        <source>incorrect_num_seqs_in_block</source>
        <translation type="obsolete">Incorrect number of sequences in block</translation>
    </message>
    <message>
        <source>seq_name_not_matched</source>
        <translation type="obsolete">Sequence names are not matched</translation>
    </message>
    <message>
        <source>alphabet_unknown</source>
        <translation type="obsolete">Alphabet is unknown</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="59"/>
        <source>CLUSTALW</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ConvertAceToSqliteTask</name>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="47"/>
        <source>Convert ACE to UGENE database (%1)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="59"/>
        <source>Converting assembly from %1 to %2 started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="67"/>
        <source>IOAdapterFactory is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="71"/>
        <source>Can&apos;t open file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="84"/>
        <source>DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="86"/>
        <source>Object DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="89"/>
        <source>Importing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="137"/>
        <source>Sequence DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="156"/>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="199"/>
        <source>Invalid source file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="215"/>
        <source>Assembly DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="221"/>
        <source>Packing reads for assembly &apos;%1&apos; (%2 of %3)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="247"/>
        <source>Attribute DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="276"/>
        <source>Warning: incorrect maxProw == %1, probably packing was not done! Attribute was not set</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConvertAssemblyToSamTask</name>
    <message>
        <location filename="../src/tasks/ConvertAssemblyToSamTask.cpp" line="94"/>
        <source>Given file is not valid UGENE database file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DNAQualityIOUtils</name>
    <message>
        <location filename="../src/DNAQualityIOUtils.cpp" line="67"/>
        <source>No IO adapter found for URL: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Database</name>
    <message>
        <location filename="../src/Database.cpp" line="48"/>
        <source>Not a valid S3-database file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/Database.cpp" line="58"/>
        <source>File already exists: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DefaultConvertFileTask</name>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="121"/>
        <source>The formats are not compatible: %1 and %2</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DifferentialFormat</name>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="44"/>
        <source>Differential</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="46"/>
        <source>Differential format is a text-based format for representing Cuffdiff differential output files: expression, splicing, promoters and cds.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="289"/>
        <source>Required column is missed: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="258"/>
        <location filename="../src/FastaFormat.cpp" line="273"/>
        <location filename="../src/FastqFormat.cpp" line="339"/>
        <location filename="../src/PDWFormat.cpp" line="156"/>
        <source>Document is empty.</source>
        <translation>Dokument je prázdný.</translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormat</name>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="50"/>
        <source>Database connection</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="51"/>
        <source>A fake format that was added to implement shared database connection within existing document model.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="90"/>
        <source>You have no permissions to modify the content of this database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="110"/>
        <source>Empty object name</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormatUtils</name>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="232"/>
        <source>Undefined sequence alphabet</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="136"/>
        <source>First line is not an ace header</source>
        <translation type="unfinished">První řádek není hlavička ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="140"/>
        <source>There is no assemblies in input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="156"/>
        <source>There are not enough assemblies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="212"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="466"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished">Neočekávaný konec souboru</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="213"/>
        <source>Line is too long</source>
        <translation type="unfinished">Řádek je příliš dlouhý</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="220"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished">Žádná počtu contig v řádku s hlavičkou</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="232"/>
        <source>Not enough parameters in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="243"/>
        <source>Parameter is not a digit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="251"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished">Není tu žádná poznámka o počtu čtení</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="263"/>
        <source>A name is duplicated</source>
        <translation type="unfinished">Název je zdvojen</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="270"/>
        <source>No consensus</source>
        <translation type="unfinished">Žádná shoda</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="280"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation type="unfinished">Klíčové slovo BQ nebylo nalezeno</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="283"/>
        <source>Unexpected symbols in consensus data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="294"/>
        <source>Can&apos;t find a sequence name in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="304"/>
        <source>An empty sequence name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="356"/>
        <source>Invalid AF tag</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="373"/>
        <source>A name is duplicated: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="379"/>
        <source>Not all reads were found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="388"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="392"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="424"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="436"/>
        <source>Bad AF note</source>
        <translation type="unfinished">Špatná poznámka AF</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="462"/>
        <source>There is no read note</source>
        <translation type="unfinished">Není tu žádná poznámka o čtení</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="475"/>
        <source>Invalid RD part</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="476"/>
        <source>Can&apos;t find the RD tag</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="485"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation type="unfinished">Klíčové slovo QA nebylo nalezeno</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="494"/>
        <source>QA error bad range</source>
        <translation type="unfinished">Chyba QA špatný rozsah</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="497"/>
        <source>Unexpected symbols in sequence data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="499"/>
        <source>A name is not match with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="506"/>
        <source>Can&apos;t find clear range start in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="507"/>
        <source>Clear range start is invalid</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="514"/>
        <source>Can&apos;t find clear range end in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="515"/>
        <source>Clear range end is invalid</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="539"/>
        <source>There is no next element</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="56"/>
        <source>Conversion file from %1 to %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/MergeBamTask.cpp" line="40"/>
        <source>Merge BAM files with SAMTools merge</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <source>annotation_name_is_empty</source>
        <translation type="obsolete">Annotation name is empty</translation>
    </message>
    <message>
        <source>annotation_location_is_empty</source>
        <translation type="obsolete">Error parsing location</translation>
    </message>
    <message>
        <source>error_reading_sequence_abnormal_term</source>
        <translation type="obsolete">Error parsing sequence: unexpected empty line</translation>
    </message>
    <message>
        <source>error_reading_sequence_invalid_format</source>
        <translation type="obsolete">Error reading sequence: invalid sequence format</translation>
    </message>
    <message>
        <source>erorr_reading_sequence_enable_to_add_more_data</source>
        <translation type="obsolete">Error reading sequence: memory allocation failed</translation>
    </message>
    <message>
        <source>line_is_too_long_or_unexpected_oef</source>
        <translation type="obsolete">Unexpected line format</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="122"/>
        <source>File &quot;%1&quot; contains too many sequences to be displayed. However, you can process these data using instruments from the menu &lt;i&gt;Tools -&gt; NGS data analysis&lt;/i&gt; or pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="139"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="308"/>
        <source>Reading entry header</source>
        <translation>Čte se záhlaví záznamu</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="213"/>
        <source>Merge error: found annotations without sequence</source>
        <translation>Chyba při slučování: nalezeny poznámky bez sekvence</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="464"/>
        <source>The file contains an incorrect data that describes a qualifier value. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="465"/>
        <source>The value cannot contain a single quote character. The qualifier is &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="535"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="316"/>
        <source>Annotation name is empty</source>
        <translation>Název poznámky je prázdný</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="544"/>
        <source>Error parsing location</source>
        <translation>Chyba při zpracování umístění poznámky</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="572"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="370"/>
        <source>Unexpected line format</source>
        <translation>Neočekávaný formát řádku</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="626"/>
        <source>Reading sequence %1</source>
        <translation>Čte se sekvence %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="647"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>Chyba při zpracování sekvence: neočekávaný prázdný řádek</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="674"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation>Chyba při čtení sekvence: neplatný formát sekvence</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="692"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>Chyba při čtení sekvence: přidělení paměti selhalo</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="707"/>
        <source>Sequence is truncated</source>
        <translation>Sekvence je zkrácena</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="714"/>
        <source>Reading annotations %1</source>
        <translation>Čtou se poznámky %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="724"/>
        <source>Invalid format of feature table</source>
        <translation>Neplatný formát nebo tabulka funkcí</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="787"/>
        <source>Line is too long.</source>
        <translation>Řádek je příliš dlouhý.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="789"/>
        <source>IO error.</source>
        <translation>Chyba I/O.</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="226"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="253"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="207"/>
        <source>Record is truncated.</source>
        <translation>Záznam je zkrácen.</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="325"/>
        <source>Annotation start position is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="330"/>
        <source>Annotation end position is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="43"/>
        <source>EMBL</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="46"/>
        <source>EMBL Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="83"/>
        <source>ID is not the first line</source>
        <translation>ID není na prvním řádku</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="90"/>
        <source>Error parsing ID line</source>
        <translation>Chyba při zpracování řádku s ID</translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <source>line_is_too_long</source>
        <translation type="obsolete">Line is too long</translation>
    </message>
    <message>
        <source>first_line_is_not_a_fasta_header</source>
        <translation type="obsolete">First line is not a FASTA header</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="57"/>
        <source>FASTA</source>
        <translation>FASTA</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="60"/>
        <source>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="180"/>
        <location filename="../src/FastaFormat.cpp" line="410"/>
        <source>Line is too long</source>
        <translation>Řádek je příliš dlouhý</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="183"/>
        <location filename="../src/FastaFormat.cpp" line="412"/>
        <source>First line is not a FASTA header</source>
        <translation>První řádek není hlavičkou FASTA</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="241"/>
        <source>File &quot;%1&quot; contains too many sequences to be displayed. However, you can process these data using instruments from the menu &lt;i&gt;Tools -&gt; NGS data analysis&lt;/i&gt; or pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="279"/>
        <source>Loaded sequences: %1.
</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="280"/>
        <source>Skipped sequences: %1.
</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="281"/>
        <source>The following sequences are empty:
%1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="478"/>
        <source>Unreferenced sequence in the beginning of patterns: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <source>Unexpected end of file</source>
        <translation type="vanished">Neočekávaný konec souboru</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="50"/>
        <source>FASTQ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="53"/>
        <source>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores.         Both the sequence letter and quality score are encoded with a single ASCII character for brevity.         It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data,         but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="136"/>
        <source>Error while trying to find sequence name start</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="174"/>
        <location filename="../src/FastqFormat.cpp" line="200"/>
        <source>Error while reading sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="292"/>
        <source>Not a valid FASTQ file: %1, sequence name differs from quality scores name: %2 and %3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="313"/>
        <source>File &quot;%1&quot; contains too many sequences to be displayed. However, you can process these data using instruments from the menu &lt;i&gt;Tools -&gt; NGS data analysis&lt;/i&gt; or pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="491"/>
        <source>Not a valid FASTQ file, sequence name differs from quality scores name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="500"/>
        <source>Not a valid FASTQ file. Bad quality scores: inconsistent size.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Not a valid FASTQ file: %1. The @ identifier is not found.</source>
        <translation type="vanished">Neplatný soubor FASTQ: %1. Identifikátor @ nenalezen.</translation>
    </message>
    <message>
        <source>Not a valid FASTQ file: %1</source>
        <translation type="vanished">Neplatný soubor FASTQ: %1</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="302"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation>Neplatný soubor FASTQ: %1. Špatné výsledky u kvality: odporující si velikost.</translation>
    </message>
</context>
<context>
    <name>U2::FpkmTrackingFormat</name>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="94"/>
        <source>The FPKM (fragments per kilobase of exon model per million mapped fragments) Tracking Format is a native Cufflinks format to output estimated expression values.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="129"/>
        <source>Internal error: qualifier with name &apos;%1&apos; and &apos;%2&apos; can&apos;t be added</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="160"/>
        <source>FPKM Tracking Format parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="165"/>
        <source>FPKM Tracking Format parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="170"/>
        <source>FPKM Tracking Format parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="182"/>
        <source>FPKM Tracking Format parsing error: different sequence names were detected in an input file. Sequence name &apos;%1&apos; is used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="198"/>
        <source>FPKM Tracking Format parsing error: tracking ID value is empty at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="221"/>
        <source>FPKM Tracking Format parsing error: incorrect length value at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="226"/>
        <source>FPKM Tracking Format parsing error: incorrect coverage value at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="516"/>
        <source>Skipped qualifier &apos;%1&apos; while saving a FPKM header.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="586"/>
        <source>FPKM Tracking Format saving error: tracking ID shouldn&apos;t be empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="611"/>
        <source>FPKM Tracking Format saving error: failed to parse locus qualifier &apos;%1&apos;, writing it to the output file anyway!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="618"/>
        <source>FPKM Tracking Format saving error: an annotation region (%1, %2) differs from the information stored in the &apos;locus&apos; qualifier (%3, %4). Writing the &apos;locus&apos; qualifier to output!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="652"/>
        <source>FPKM Tracking Format saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GBFeatureUtils</name>
    <message>
        <source>Sequence related to transcription termination</source>
        <translation type="vanished">Sekvence mající spojitost s ukončením přepisu (transkripce)</translation>
    </message>
    <message>
        <source>Span of the C immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu C</translation>
    </message>
    <message>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation type="vanished">`CAAT box&apos; v eukaryotických pořadatelích</translation>
    </message>
    <message>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation type="vanished">Kódování sekvence pro aminokyseliny v bílkovině (zahrnuje zastavovací kodon)</translation>
    </message>
    <message>
        <source>Independent sequence determinations differ</source>
        <translation type="vanished">Odlišení určení nezávislé sekvence</translation>
    </message>
    <message>
        <source>Displacement loop</source>
        <translation type="vanished">Smyčka přemístění</translation>
    </message>
    <message>
        <source>Span of the D immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu D</translation>
    </message>
    <message>
        <source>Cis-acting enhancer of promoter function</source>
        <translation type="vanished">Cis-sloužící vylepšovač funkce pořadatele</translation>
    </message>
    <message>
        <source>Region that codes for part of spliced mRNA</source>
        <translation type="vanished">Oblast, která kóduje pro část stočené mRNA</translation>
    </message>
    <message>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation type="vanished">Oblast stanovující funkční gen, možná včetně ovládacích prvků proti proudu (pořadatel, vylepšovač atd.) a po proudu, a kterému byl přiřazen název.</translation>
    </message>
    <message>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation type="vanished">`GC box&apos; v eukaryotických pořadatelích</translation>
    </message>
    <message>
        <source>Intervening DNA eliminated by recombination</source>
        <translation type="vanished">Mezilehlá DNA odstraněná rekombinací</translation>
    </message>
    <message>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation type="vanished">Přepsaná oblast vynechaná spojováním mRNA</translation>
    </message>
    <message>
        <source>Span of the J immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu J</translation>
    </message>
    <message>
        <source>Long terminal repeat</source>
        <translation type="vanished">Dlouhé konečné opakování</translation>
    </message>
    <message>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation type="vanished">Oblast kódování zralého peptidu (nezahrnuje stavěcí kodon)</translation>
    </message>
    <message>
        <source>Miscellaneous binding site</source>
        <translation type="vanished">Různé vázající místo</translation>
    </message>
    <message>
        <source>Miscellaneous difference feature</source>
        <translation type="vanished">Různý rozlišovací znak</translation>
    </message>
    <message>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation type="vanished">Oblast s biologickým významem, kterou nelze popsat žádným jiným znakem</translation>
    </message>
    <message>
        <source>Miscellaneous, recombination feature</source>
        <translation type="vanished">Různé, znak rekombinace</translation>
    </message>
    <message>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation type="vanished">Různý znak přepisu (transkripce) nevymezený v jiných klíčích RNA</translation>
    </message>
    <message>
        <source>Miscellaneous signal</source>
        <translation type="vanished">Různá znamení</translation>
    </message>
    <message>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation type="vanished">Různá stavba DNA nebo RNA</translation>
    </message>
    <message>
        <source>The indicated base is a modified nucleotide</source>
        <translation type="vanished">Naznačený základ je pozměněným nukleotidem</translation>
    </message>
    <message>
        <source>Messenger RNA</source>
        <translation type="vanished">RNA nositel biologické informace</translation>
    </message>
    <message>
        <source>Span of the N immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu N</translation>
    </message>
    <message>
        <source>Presented sequence revises a previous version</source>
        <translation type="vanished">Předložená sekvence poopravuje předchozí verzi</translation>
    </message>
    <message>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation type="vanished">Znamení pro buněčné dělení &amp; polyadenylaci</translation>
    </message>
    <message>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation type="vanished">Místo, na kterém je polyadenin přidán do mRNA</translation>
    </message>
    <message>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation type="vanished">Jakýkoli druh RNA, který ještě není zralým produktem RNA</translation>
    </message>
    <message>
        <source>Primary (unprocessed) transcript</source>
        <translation type="vanished">Hlavní (nezpracovaný) přepis (transkripce)</translation>
    </message>
    <message>
        <source>Primer binding region used with PCR</source>
        <translation type="vanished">Základová vázající oblast použitá s PCR</translation>
    </message>
    <message>
        <source>Non-covalent primer binding site</source>
        <translation type="vanished">Nekovalentní základové vázající místo</translation>
    </message>
    <message>
        <source>A region involved in transcription initiation</source>
        <translation type="vanished">Oblast zapojená do zahájení přepisu (transkripce)</translation>
    </message>
    <message>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation type="vanished">Nekovalentní bílkovinné vázající místo na DNA nebo RNA</translation>
    </message>
    <message>
        <source>Ribosome binding site</source>
        <translation type="vanished">Ribozomové vázající místo</translation>
    </message>
    <message>
        <source>Replication origin for duplex DNA</source>
        <translation type="vanished">Původ replikace pro zdvojenou DNA</translation>
    </message>
    <message>
        <source>Sequence containing repeated subsequences</source>
        <translation type="vanished">Sekvence obsahující opakující se podsekvence</translation>
    </message>
    <message>
        <source>One repeated unit of a repeat_region</source>
        <translation type="vanished">Jedna opakující se jednotka oblasti opakování</translation>
    </message>
    <message>
        <source>Ribosomal RNA</source>
        <translation type="vanished">Ribozomová RNA</translation>
    </message>
    <message>
        <source>Span of the S immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu S</translation>
    </message>
    <message>
        <source>Satellite repeated sequence</source>
        <translation type="vanished">Přidružená opakující se sekvence</translation>
    </message>
    <message>
        <source>Small cytoplasmic RNA</source>
        <translation type="vanished">Malá cytoplazmatická RNA</translation>
    </message>
    <message>
        <source>Signal peptide coding region</source>
        <translation type="vanished">Oblast kódování významného peptidu</translation>
    </message>
    <message>
        <source>Small nuclear RNA</source>
        <translation type="vanished">Malá jaderná RNA</translation>
    </message>
    <message>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation type="vanished">Určuje biologický zdroj daného rozpětí sekvence</translation>
    </message>
    <message>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation type="vanished">Vlásenková stavba smyčky v DNA nebo RNA</translation>
    </message>
    <message>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation type="vanished">Označené místo sekvence; provozně jedinečná sekvence, která určuje spojení rozpětí základové barvy použitých ve zkoušce PCR</translation>
    </message>
    <message>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation type="vanished">`TATA box&apos; v eukaryotických předkladatelích</translation>
    </message>
    <message>
        <source>Sequence causing transcription termination</source>
        <translation type="vanished">Sekvence způsobující ukončení přepisu (transkripce)</translation>
    </message>
    <message>
        <source>Transit peptide coding region</source>
        <translation type="vanished">Oblast kódování průjezdního peptidu</translation>
    </message>
    <message>
        <source>Transposable element (TN)</source>
        <translation type="vanished">Transpozon (přemístitelná část genetického materiálu, TN)</translation>
    </message>
    <message>
        <source>Transfer RNA</source>
        <translation type="vanished">Přenést RNA</translation>
    </message>
    <message>
        <source>Authors are unsure about the sequence in this region</source>
        <translation type="vanished">Autoři si sekvencí v této oblasti nejsou jisti</translation>
    </message>
    <message>
        <source>Span of the V immunological feature</source>
        <translation type="vanished">Rozpětí imunologického rysu V</translation>
    </message>
    <message>
        <source>A related population contains stable mutation</source>
        <translation type="vanished">Související populace obsahuje stálou mutaci</translation>
    </message>
    <message>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation type="vanished">`Pribnow box&apos; v prokaryotických předkladatelích</translation>
    </message>
    <message>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation type="vanished">`-35 box&apos; v prokaryotických předkladatelích</translation>
    </message>
    <message>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation type="vanished">Oblast 3&apos;-most přepisu (transkripce) předchůdce odstraněna při zpracování</translation>
    </message>
    <message>
        <source>3&apos; untranslated region (trailer)</source>
        <translation type="vanished">3&apos; nepřeložená oblast (na konci)</translation>
    </message>
    <message>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation type="vanished">Oblast 5&apos;-most přepisu (transkripce) předchůdce odstraněna při zpracování</translation>
    </message>
    <message>
        <source>5&apos; untranslated region (leader)</source>
        <translation type="vanished">5&apos; nepřeložená oblast (na začátku)</translation>
    </message>
    <message>
        <source>&apos;Protein&apos; feature key</source>
        <translation type="vanished">Klíč vlastnosti &apos;Protein&apos; </translation>
    </message>
    <message>
        <source>&apos;Region&apos; feature key</source>
        <translation type="vanished">Klíč vlastnosti &apos;Region&apos;</translation>
    </message>
    <message>
        <source>&apos;Site&apos; feature key</source>
        <translation type="vanished">Klíč vlastnosti &apos;Site&apos;</translation>
    </message>
    <message>
        <source>Genes</source>
        <translation type="vanished">Geny</translation>
    </message>
    <message>
        <source>Signals</source>
        <translation type="vanished">Signály</translation>
    </message>
    <message>
        <source>Binding</source>
        <translation type="vanished">Vazba</translation>
    </message>
    <message>
        <source>Variation</source>
        <translation type="vanished">Obměna</translation>
    </message>
    <message>
        <source>Repeats</source>
        <translation type="vanished">Opakování</translation>
    </message>
    <message>
        <source>Misc</source>
        <translation type="vanished">Různé</translation>
    </message>
    <message>
        <source>Spans</source>
        <translation type="vanished">Rozpětí</translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="50"/>
        <source>GFF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="51"/>
        <source>GFF is a format used for storing features and annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="86"/>
        <source>Parsing error: invalid header</source>
        <translation>Chyba zpracování: neplatná hlavička</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="90"/>
        <source>Parsing error: file does not contain version header</source>
        <translation>Chyba zpracování: soubor neobsahuje hlavičku s verzí</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="94"/>
        <source>Parsing error: format version is not an integer</source>
        <translation>Chyba zpracování: formát verze není celé číslo</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="98"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation>Chyba zpracování: verze %1 není podporována</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="258"/>
        <source>Parsing error: file contains empty line %1, line skipped</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="291"/>
        <source>Parsing error: sequence in FASTA sequence has whitespaces at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="298"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation>Chyba zpracování: příliš mnoho polí na řádku %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="304"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation>Chyba zpracování: začáteční poloha na řádku %1 není celé číslo</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="310"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation>Chyba zpracování: koncová poloha na řádku %1 není celé číslo</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="315"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation>Chyba zpracování: nesprávná oblast s poznámkou na řádku %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="366"/>
        <source>Parsing error: incorrect attributes field %1 at line %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="405"/>
        <source>Parsing error: incorrect score parameter at line %1. Score can be a float number or &apos;.&apos; symbol</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="415"/>
        <source>Parsing error: incorrect frame parameter at line %1. Frame can be a number between 0-2 or &apos;.&apos; symbol</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="425"/>
        <source>Parsing error: incorrect strand patameter at line %1. Strand can be &apos;+&apos;,&apos;-&apos; or &apos;.&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="590"/>
        <source>Can not detect chromosome name. &apos;Chr&apos; name will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Parsing error: one of the qualifiers in attributes are incorrect</source>
        <translation type="vanished">Chyba zpracování: jeden z modifikátorů ve vlastnostech není správný</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="356"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation>Špatné umístění pro připojenou poznámku na řádku %1. Řádek byl přeskočen.</translation>
    </message>
    <message>
        <source>Fasta sequence inclusion is not supported for GFF</source>
        <translation type="vanished">Zahrnutí sekvence Fasta není podporováno pro GFF</translation>
    </message>
</context>
<context>
    <name>U2::GTFFormat</name>
    <message>
        <location filename="../src/GTFFormat.cpp" line="102"/>
        <source>The Gene transfer format (GTF) is a file format used to hold information about gene structure.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="159"/>
        <source>GTF parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="164"/>
        <source>GTF parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="169"/>
        <source>GTF parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="181"/>
        <source>GTF parsing error: unexpected value of the &quot;feature&quot; value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="199"/>
        <source>GTF parsing error: incorrect score value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="207"/>
        <source>GTF parsing error: incorrect frame value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="228"/>
        <source>GTF parsing error: invalid attributes format at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="235"/>
        <location filename="../src/GTFFormat.cpp" line="240"/>
        <source>GTF parsing error: mandatory attribute &apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="236"/>
        <location filename="../src/GTFFormat.cpp" line="241"/>
        <source>&apos; is absent at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="248"/>
        <source>GTF parsing error: incorrect strand value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="295"/>
        <source>File &quot;%1&quot; contains too many annotation tables to be displayed. However, you can process these data using pipelines built with Workflow Designer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="630"/>
        <location filename="../src/GTFFormat.cpp" line="635"/>
        <source>Can&apos;t save an annotation to a GTF file - the annotation doesn&apos;t have the &apos;%1&apos; qualifier!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="654"/>
        <source>GTF saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Genbank::LocationParser</name>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="308"/>
        <source>Ignoring remote entry: %1.%2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="339"/>
        <location filename="../src/GenbankLocationParser.cpp" line="409"/>
        <source>&apos;a single base from a range&apos; in combination with &apos;sequence span&apos; is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="342"/>
        <source>Ignoring &apos;&lt;&apos; at start position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="412"/>
        <source>Ignoring &apos;&gt;&apos; at end position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="474"/>
        <source>Wrong token after JOIN %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="479"/>
        <source>Wrong token after JOIN  - order %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="487"/>
        <source>Can&apos;t parse location on JOIN</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="493"/>
        <location filename="../src/GenbankLocationParser.cpp" line="517"/>
        <location filename="../src/GenbankLocationParser.cpp" line="537"/>
        <location filename="../src/GenbankLocationParser.cpp" line="558"/>
        <source>Must be RIGHT_PARENTHESIS instead of %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="499"/>
        <source>Wrong token after ORDER %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="511"/>
        <source>Can&apos;t parse location on ORDER</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="523"/>
        <source>Wrong token after BOND %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="531"/>
        <source>Can&apos;t parse location on BONDs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="544"/>
        <source>Must be LEFT_PARENTHESIS instead of %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="552"/>
        <source>Can&apos;t parse location on COMPLEMENT</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <source>error writing document</source>
        <translation type="obsolete">Error writing document</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="47"/>
        <source>Genbank</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="49"/>
        <source>GenBank Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="90"/>
        <source>LOCUS is not the first line</source>
        <translation>LOCUS není první řádek</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="105"/>
        <source>Error parsing LOCUS line</source>
        <translation>Chyba při zpracování řádku LOCUS</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="170"/>
        <source>incomplete SOURCE record</source>
        <translation>Neúplný SOURCE záznam</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="396"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="629"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="634"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="640"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="787"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="816"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="823"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="830"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="840"/>
        <source>Error writing document</source>
        <translation>Chyba při zápisu dokumentu</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="527"/>
        <source>There is no IOAdapter registry yet</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="529"/>
        <source>IOAdapterFactory is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="531"/>
        <source>IOAdapter is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="799"/>
        <source>Invalid annotation table!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GzipDecompressTask</name>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="120"/>
        <source>Decompression task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="125"/>
        <source>&apos;%1&apos; is not zipped file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="130"/>
        <source>Start decompression &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="132"/>
        <source>IOAdapterRegistry is NULL!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="135"/>
        <location filename="../src/tasks/BgzipTask.cpp" line="137"/>
        <source>IOAdapterFactory is NULL!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="140"/>
        <location filename="../src/tasks/BgzipTask.cpp" line="143"/>
        <source>Can not create IOAdapter!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="155"/>
        <source>Can not open output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="161"/>
        <source>Can not open input file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="176"/>
        <source>Error reading file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="182"/>
        <source>Error writing to file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="187"/>
        <source>Decompression finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="192"/>
        <source>Decompression task was finished with an error: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="194"/>
        <source>Decompression task was finished. A new decompressed file is: &lt;a href=&quot;%1&quot;&gt;%1&lt;/a&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::IndexFormat</name>
    <message>
        <source>Index</source>
        <translation type="vanished">Rejstřík</translation>
    </message>
    <message>
        <source>Read error occurred</source>
        <translation type="vanished">Vyskytla se chyba čtení</translation>
    </message>
    <message>
        <source>Write error occurred</source>
        <translation type="vanished">Vyskytla se chyba zápisu</translation>
    </message>
    <message>
        <source>Empty io identifier in item section</source>
        <translation type="vanished">Prázdný identifikátor IO v položkové části</translation>
    </message>
    <message>
        <source>Empty document format in item section </source>
        <translation type="vanished">Prázdný formát dokumentu v položkové části</translation>
    </message>
    <message>
        <source>Empty io adapter identifier in io section</source>
        <translation type="vanished">Prázdný identifikátor adaptéru IO v části IO</translation>
    </message>
    <message>
        <source>Empty url in io section</source>
        <translation type="vanished">Prázdná adresa (URL)  v části IO</translation>
    </message>
    <message>
        <source>Bad offset in item section</source>
        <translation type="vanished">Špatný posun v položkové části</translation>
    </message>
    <message>
        <source>Unrecognized io section id in item section</source>
        <translation type="vanished">Nerozpoznaný identifikátor části IO v části IO</translation>
    </message>
    <message>
        <source>Unknown index error occurred</source>
        <translation type="vanished">Vyskytla se neznámá chyba rejstříku</translation>
    </message>
    <message>
        <source>invalid header</source>
        <translation type="vanished">Neplatná hlavička</translation>
    </message>
    <message>
        <source>Unknown section found</source>
        <translation type="vanished">Nalezena neznámá část</translation>
    </message>
    <message>
        <source>no items found in index file</source>
        <translation type="vanished">V souboru s rejstříkem nebyly nalezeny žádné položky</translation>
    </message>
    <message>
        <source>Unknown error occurred</source>
        <translation type="vanished">Vyskytla se neznámá chyba</translation>
    </message>
    <message>
        <source>Bad value in item section: tag</source>
        <translation type="vanished">Špatná hodnota v položkové části: tag</translation>
    </message>
    <message>
        <source>Bad value in item section: val</source>
        <translation type="vanished">Špatná hodnota v položkové části: val</translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="64"/>
        <source>MSF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="66"/>
        <source>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="126"/>
        <source>Incorrect format</source>
        <translation>Nesprávný formát</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="145"/>
        <source>Unexpected end of file</source>
        <translation>Neočekávaný konec souboru</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="172"/>
        <location filename="../src/MSFFormat.cpp" line="213"/>
        <source>Check sum test failed</source>
        <translation>Zkouška kontrolního součtu se nezdařila</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="221"/>
        <source>Alphabet unknown</source>
        <translation>Neznámá abeceda</translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="53"/>
        <source>Mega</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="54"/>
        <source>Mega is a file format of native MEGA program</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="304"/>
        <source>Incorrect sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="316"/>
        <source>Found sequences of different sizes</source>
        <translation>Nalezeny sekvence rozdílných velikostí</translation>
    </message>
    <message>
        <source>Model is of zero size</source>
        <translation type="vanished">Model má nulovou velikost</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="149"/>
        <source>Bad name of sequence</source>
        <translation>Špatný název sekvence</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="169"/>
        <source>Unexpected # in comments</source>
        <translation>Neočekávaný # v poznámkách</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="178"/>
        <source>A comment has not end</source>
        <translation>Poznámka nemá žádný konec</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="200"/>
        <source>Unexpected symbol between comments</source>
        <translation>Neočekávaný symbol mezi poznámkami</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="261"/>
        <source>Incorrect format</source>
        <translation>Nesprávný formát</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="274"/>
        <source>Identical symbol at the first sequence</source>
        <translation>Totožný symbol u první sekvence</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="298"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation>Nesprávné pořadí názvů sekvencí</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="324"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished">Abeceda je neznámá</translation>
    </message>
    <message>
        <source>Alphabet unknown</source>
        <translation type="vanished">Neznámá abeceda</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="394"/>
        <location filename="../src/MegaFormat.cpp" line="406"/>
        <source>No header</source>
        <translation>Žádná hlavička</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="399"/>
        <source>No # before header</source>
        <translation>Žádný # před hlavičkou</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="411"/>
        <source>Not MEGA-header</source>
        <translation>Žádná MEGA hlavička</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="434"/>
        <location filename="../src/MegaFormat.cpp" line="445"/>
        <location filename="../src/MegaFormat.cpp" line="472"/>
        <source>No data in file</source>
        <translation>Žádná data v souboru</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="454"/>
        <location filename="../src/MegaFormat.cpp" line="460"/>
        <source>Incorrect title</source>
        <translation>Nesprávný název</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="504"/>
        <source>Sequence has empty part</source>
        <translation>Sekvence má prázdnou část</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="524"/>
        <source>Bad symbols in a sequence</source>
        <translation>Špatné symboly v sekvenci</translation>
    </message>
</context>
<context>
    <name>U2::MysqlUpgradeTask</name>
    <message>
        <location filename="../src/tasks/MysqlUpgradeTask.cpp" line="33"/>
        <source>Upgrade mysql database</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="50"/>
        <source>NEXUS</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="51"/>
        <source>Nexus is a multiple alignment and phylogenetic trees file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="695"/>
        <source>#NEXUS header missing</source>
        <translation>Chybí hlavička #NEXUS</translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="41"/>
        <source>Newick Standard</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="42"/>
        <source>Newick is a simple format used to write out trees in a text file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Unexpected weight: %1</source>
        <translation type="vanished">Neočekávaná váha: %1</translation>
    </message>
    <message>
        <source>Error parsing weight: %1</source>
        <translation type="vanished">Chyba při zpracování váhy: %1</translation>
    </message>
    <message>
        <source>Unexpected weight start token: %1</source>
        <translation type="vanished">Neočekávaný symbol na začátku váhy: %1</translation>
    </message>
    <message>
        <source>Unexpected new sibling %1</source>
        <translation type="vanished">Neočekávaný nový sourozenec %1</translation>
    </message>
    <message>
        <source>Unexpected closing bracket :%1</source>
        <translation type="vanished">Neočekávaná zavírající závorka:%1</translation>
    </message>
    <message>
        <source>Unexpected end of file</source>
        <translation type="vanished">Neočekávaný konec souboru</translation>
    </message>
    <message>
        <source>Empty file</source>
        <translation type="vanished">Prázdný soubor</translation>
    </message>
    <message>
        <source>weight_unexpected_%1</source>
        <translation type="obsolete">Unexpected weight: %1</translation>
    </message>
    <message>
        <source>weight_parse_error_%1</source>
        <translation type="obsolete">Error parsing weight: %1</translation>
    </message>
    <message>
        <source>unexpected_weight_start_%1</source>
        <translation type="obsolete">Unexpected weight start token: %1</translation>
    </message>
    <message>
        <source>unexpected_closing_bracket_%1</source>
        <translation type="obsolete">Unexpected closing bracket :%1</translation>
    </message>
    <message>
        <source>unexpected_eof</source>
        <translation type="obsolete">Unexpected end of file</translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <source>line_is_too_long</source>
        <translation type="obsolete">Line is too long</translation>
    </message>
    <message>
        <source>SPLIT_record_not_supported</source>
        <translation type="obsolete">SPLIT recored is not supported</translation>
    </message>
    <message>
        <source>mandatory_records_absent</source>
        <translation type="obsolete">Some mandaroty records are absent</translation>
    </message>
    <message>
        <source>invalid_header_record</source>
        <translation type="obsolete">Invalid HEADER record</translation>
    </message>
    <message>
        <source>ivalid_atom_record</source>
        <translation type="obsolete">Invalid ATOM record</translation>
    </message>
    <message>
        <source>invalid_secondary_structure_record</source>
        <translation type="obsolete">Invalid secondary structure record</translation>
    </message>
    <message>
        <source>compound_record_is_empty</source>
        <translation type="obsolete">COMPND record is empty</translation>
    </message>
    <message>
        <source>invalid_compound_record</source>
        <translation type="obsolete">Invalid COMPND record</translation>
    </message>
    <message>
        <source>invalid_seqres_less_then_24</source>
        <translation type="obsolete">Invalid SEQRES: less then 24 charachters</translation>
    </message>
    <message>
        <source>invalid_seqres_chain_index_not_found</source>
        <translation type="obsolete">Invalid SEQRES: chain index not found</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="55"/>
        <source>PDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="56"/>
        <source>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="191"/>
        <source>Line is too long</source>
        <translation>Řádek je příliš dlouhý</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="247"/>
        <source>Some mandatory records are absent</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>SPLIT recored is not supported</source>
        <translation type="vanished">SPLIT zaznamenaný není podporován</translation>
    </message>
    <message>
        <source>Some mandaroty records are absent</source>
        <translation type="vanished">Některé povinné záznamy nejsou přítomné</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="355"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation>Varování: neznámý název zbytku: %1</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="445"/>
        <source>Invalid secondary structure record</source>
        <translation>Neplatný záznam vedlejší stavby</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="482"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation>Neplatný SEQRES: méně než 24 znaků</translation>
    </message>
</context>
<context>
    <name>U2::PDWFormat</name>
    <message>
        <location filename="../src/PDWFormat.cpp" line="57"/>
        <source>pDRAW</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="58"/>
        <source>pDRAW is a sequence file format used by pDRAW software</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="102"/>
        <location filename="../src/PDWFormat.cpp" line="192"/>
        <source>Line is too long</source>
        <translation type="unfinished">Řádek je příliš dlouhý</translation>
    </message>
</context>
<context>
    <name>U2::PhylipFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="48"/>
        <source>PHYLIP multiple alignment format for phylogenetic applications.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="77"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished">Abeceda je neznámá</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="352"/>
        <source>Error parsing file</source>
        <translation type="unfinished">Chyba při zpracování souboru</translation>
    </message>
</context>
<context>
    <name>U2::PhylipInterleavedFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="240"/>
        <source>PHYLIP Interleaved</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="339"/>
        <source>Illegal line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="346"/>
        <source>Wrong header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="382"/>
        <source>Block is incomplete</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="393"/>
        <source>Block is incomlete</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="400"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhylipSequentialFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="121"/>
        <source>PHYLIP Sequential</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="202"/>
        <source>Illegal line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="209"/>
        <source>Wrong header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="212"/>
        <location filename="../src/PhylipFormat.cpp" line="350"/>
        <source>There is not enough data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="217"/>
        <location filename="../src/PhylipFormat.cpp" line="360"/>
        <source>Error parsing file</source>
        <translation type="unfinished">Chyba při zpracování souboru</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="231"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <source>plain_text_format</source>
        <translation type="obsolete">Plain text</translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="37"/>
        <source>Plain text</source>
        <translation>Prostý text</translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="39"/>
        <source>A simple plain text file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <source>raw_dna_sequence_format</source>
        <translation type="obsolete">Raw sequence</translation>
    </message>
    <message>
        <source>sequence_is_empty</source>
        <translation type="obsolete">Seqeunce is empty</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="46"/>
        <source>Raw sequence</source>
        <translation>Surová sekvence</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="49"/>
        <source>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="106"/>
        <source>Sequence is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Seqeunce is empty</source>
        <translation type="vanished">Sekvence je prázdná</translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="83"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation>Pole &quot;%1&quot; neodpovídá vzoru &quot;%2&quot;, očekáván vzor &quot;%3&quot;</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="92"/>
        <source>SAM</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="93"/>
        <source>The Sequence Alignment/Map (SAM) format is a generic alignment format forstoring read alignments against reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Unsupported file version &quot;%1&quot;</source>
        <translation type="vanished">Nepodporovaná verze souboru &quot;%1&quot;</translation>
    </message>
    <message>
        <source>Unexpected end of file</source>
        <translation type="vanished">Neočekávaný konec souboru</translation>
    </message>
    <message>
        <source>Unexpected reference sequence name value &quot;%1&quot;</source>
        <translation type="vanished">Neočekávaná hodnota názvu srovnávací sekvence &quot;%1&quot;</translation>
    </message>
    <message>
        <source>Alphabet is unknown</source>
        <translation type="vanished">Abeceda je neznámá</translation>
    </message>
    <message>
        <source>alphabet_unknown</source>
        <translation type="obsolete">Alphabet is unknown</translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="48"/>
        <source>SCF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="49"/>
        <source>It is Standard Chromatogram Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="68"/>
        <source>Failed to parse SCF file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="1232"/>
        <source>Failed to load sequence from SCF file %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Not a valid SCF file: %1</source>
        <translation type="vanished">Není platným souborem SCF: %1</translation>
    </message>
</context>
<context>
    <name>U2::SRFastaFormat</name>
    <message>
        <source>First line is not a FASTA header</source>
        <translation type="vanished">První řádek není hlavičkou FASTA</translation>
    </message>
    <message>
        <source>Incorrect sequence header: unable to parse sequence offset</source>
        <translation type="vanished">Nesprávná hlavička sekvence: nelze zpracovat posun sekvence</translation>
    </message>
    <message>
        <source>Incorrect sequence header: offset is not present</source>
        <translation type="vanished">Nesprávná hlavička sekvence: posun není přítomen</translation>
    </message>
    <message>
        <source>Alphabet is unknown</source>
        <translation type="vanished">Abeceda je neznámá</translation>
    </message>
    <message>
        <source>first_line_is_not_a_fasta_header</source>
        <translation type="obsolete">First line is not a FASTA header</translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="677"/>
        <source>Stockholm</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="678"/>
        <source>A multiple sequence alignments file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="699"/>
        <location filename="../src/StockholmFormat.cpp" line="716"/>
        <source>unknown error occurred</source>
        <translation>Vyskytla se neznámá chyba</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="451"/>
        <source>invalid file: bad header line</source>
        <translation>Neplatný soubor: špatný řádek v hlavičce</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="483"/>
        <source>invalid file: empty sequence name</source>
        <translation>Neplatný soubor: prázdný název sekvence</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="486"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation>Neplatný soubor: stejné názvy sekvencí v jednom bloku</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="499"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation>Neplatný soubor: názvy sekvencí nejsou v blocích stejné</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="493"/>
        <location filename="../src/StockholmFormat.cpp" line="505"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation>Neplatný soubor: sekvence v bloku nemají stejnou velikost</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="519"/>
        <source>invalid file: empty sequence alignment</source>
        <translation>Neplatný soubor: prázdné zarovnání sekvence</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="523"/>
        <source>invalid file: unknown alphabet</source>
        <translation>Neplatný soubor: neznámá abeceda</translation>
    </message>
</context>
<context>
    <name>U2::SwissProtPlainTextFormat</name>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="45"/>
        <source>Swiss-Prot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="47"/>
        <source>SwissProt is a format of the UniProtKB/Swiss-prot database used for storing annotated protein sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation type="unfinished">ID není na prvním řádku</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation type="unfinished">Chyba při zpracování řádku s ID</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="101"/>
        <source>Error parsing ID line. Not found sequence length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="217"/>
        <source>Reading sequence %1</source>
        <translation type="unfinished">Čte se sekvence %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="237"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation type="unfinished">Chyba při zpracování sekvence: neočekávaný prázdný řádek</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="260"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation type="unfinished">Chyba při čtení sekvence: přidělení paměti selhalo</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="275"/>
        <source>Sequence is truncated</source>
        <translation type="unfinished">Sekvence je zkrácena</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="282"/>
        <source>Reading annotations %1</source>
        <translation type="unfinished">Čtou se poznámky %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="289"/>
        <source>Invalid format of feature table</source>
        <translation type="unfinished">Neplatný formát nebo tabulka funkcí</translation>
    </message>
</context>
<context>
    <name>U2::U2DbiL10n</name>
    <message>
        <location filename="../src/mysql_dbi/MysqlAssemblyDbi.cpp" line="84"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="80"/>
        <source>There is no assembly object with the specified id.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlAttributeDbi.cpp" line="209"/>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="223"/>
        <source>Unsupported attribute type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlAttributeDbi.cpp" line="341"/>
        <source>Required attribute is not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="262"/>
        <source>Database url is incorrect</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="268"/>
        <source>User login is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="274"/>
        <source>Host is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="280"/>
        <source>Database name is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="296"/>
        <source>Error opening MySQL database: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="380"/>
        <source>Not a %1 MySQL database: %2, %3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="386"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="273"/>
        <source>Warning! The database was created with a newer %1 version: %2. Not all database features may be supported! Current %1 version: %3.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="521"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="328"/>
        <source>Illegal database state: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="552"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="410"/>
        <source>Can&apos;t synchronize database state</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlFeatureDbi.cpp" line="148"/>
        <source>Annotation table object is not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlFeatureDbi.cpp" line="177"/>
        <source>Feature is not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="134"/>
        <source>An object single modification step not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="170"/>
        <source>Failed to find user step ID</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="378"/>
        <source>Not main thread</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="390"/>
        <source>Can&apos;t create a common user modifications step, previous one is not complete</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="453"/>
        <source>Can&apos;t create a common multiple modifications step, previous one is not complete</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="489"/>
        <source>Failed to create a common user modifications step</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="510"/>
        <source>Failed to create a common multiple modifications step</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="102"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="118"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="634"/>
        <source>Msa object not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="177"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="781"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="821"/>
        <source>Msa row not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="586"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="619"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1100"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1109"/>
        <source>Unexpected modification type &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="656"/>
        <source>Invalid row position: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1030"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1161"/>
        <source>An error occurred during updating an alignment alphabet</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1048"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1176"/>
        <source>An error occurred during reverting adding of rows</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1066"/>
        <source>An error occurred during reverting addition of a row</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1079"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1196"/>
        <source>An error occurred during reverting removing of rows</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1092"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1211"/>
        <source>An error occurred during reverting removing of a row</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1106"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1222"/>
        <source>An error occurred during updating an alignment gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1119"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1232"/>
        <source>An error occurred during updating an alignment row order</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1133"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1242"/>
        <source>An error occurred during updating a row info</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1148"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1257"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="1272"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="1282"/>
        <source>An error occurred during updating an msa length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1186"/>
        <source>An error occurred during addition of a row</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="586"/>
        <source>Can&apos;t undo an operation for the object</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="649"/>
        <source>Can&apos;t redo an operation for the object</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="791"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="883"/>
        <source>Object not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="832"/>
        <source>Folder not found: %1 (canonical: %2)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1013"/>
        <source>Not an object, id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1036"/>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="289"/>
        <source>Unknown object type! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1078"/>
        <source>Can&apos;t undo an unknown operation: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1091"/>
        <source>Can&apos;t redo an unknown operation: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1122"/>
        <source>An error occurred during updating an object name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1140"/>
        <source>An error occurred during updating an object name!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="73"/>
        <source>Sequence object not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="125"/>
        <source>Internal error occurred during the sequence processing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="126"/>
        <source>An exception was thrown during reading sequence data from dbi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="387"/>
        <source>An error occurred during reverting replacing sequence data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="402"/>
        <source>An error occurred during replacing sequence data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="158"/>
        <source>Invalid variant track type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="186"/>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="219"/>
        <source>Sequence name is not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="321"/>
        <source>New variant public ID is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="337"/>
        <source>New variant track ID is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="124"/>
        <source>Packed data is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="131"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="468"/>
        <source>Packing method prefix is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="139"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="476"/>
        <source>Data is corrupted, no name end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="148"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="485"/>
        <source>Data is corrupted, no sequence end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="157"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="494"/>
        <source>Data is corrupted, no CIGAR end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="176"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="513"/>
        <source>Data is corrupted, no rnext end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="191"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="528"/>
        <source>Can not convert pnext to a number: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlHelpers.cpp" line="399"/>
        <source>Bound values: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlHelpers.cpp" line="421"/>
        <source>Cannot start a transaction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="505"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="127"/>
        <source>Failed to detect assembly storage format: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="517"/>
        <source>Failed to parse range: %1, full: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="531"/>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="536"/>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="542"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="152"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="156"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="161"/>
        <source>Failed to parse packed row range info %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="96"/>
        <source>Unsupported reads storage type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="373"/>
        <source>Packing method is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="461"/>
        <source>Packed data is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="222"/>
        <source>Error checking SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="235"/>
        <source>Error creating table: %1, error: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="267"/>
        <source>Not a %1 SQLite database: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="324"/>
        <source>Database is already opened!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="334"/>
        <source>URL is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="348"/>
        <source>Error opening SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="400"/>
        <source>Database is already closed!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="404"/>
        <source>Illegal database state %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="429"/>
        <source>Failed to close database: %1, err: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteFeatureDbi.cpp" line="144"/>
        <source>Annotation table object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteModDbi.cpp" line="139"/>
        <source>An object single modification step not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="505"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="520"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="712"/>
        <source>Msa object not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="602"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="742"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="780"/>
        <source>Msa row not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="266"/>
        <source>Not an object! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="600"/>
        <source>Can&apos;t undo an operation for the object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="685"/>
        <source>Can&apos;t redo an operation for the object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="857"/>
        <source>Object not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="918"/>
        <source>Object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="954"/>
        <source>Folder not found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSequenceDbi.cpp" line="65"/>
        <source>Sequence object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="96"/>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="130"/>
        <source>Sequence name is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/RTreeAssemblyAdapter.cpp" line="68"/>
        <source>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::VectorNtiSequenceFormat</name>
    <message>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="46"/>
        <source>Vector NTI sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="47"/>
        <source>Vector NTI sequence format is a rich format based on NCBI GenBank format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="300"/>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="316"/>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="320"/>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="325"/>
        <location filename="../src/VectorNtiSequenceFormat.cpp" line="331"/>
        <source>Error writing document</source>
        <translation type="unfinished">Chyba při zápisu dokumentu</translation>
    </message>
</context>
</TS>
