<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>Document</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="123"/>
        <source>Document is empty.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="27"/>
        <source>ABIF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="72"/>
        <source>Not a valid ABIF file: %1</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ACEFormat.cpp" line="22"/>
        <source>ACE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="206"/>
        <location filename="../src/ACEFormat.cpp" line="229"/>
        <location filename="../src/ACEFormat.cpp" line="245"/>
        <location filename="../src/ACEFormat.cpp" line="310"/>
        <location filename="../src/ACEFormat.cpp" line="360"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="211"/>
        <source>First line is not an ace header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="216"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="234"/>
        <source>Must be CO keyword</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="241"/>
        <location filename="../src/ACEFormat.cpp" line="306"/>
        <location filename="../src/ACEFormat.cpp" line="356"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="253"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="262"/>
        <location filename="../src/ACEFormat.cpp" line="318"/>
        <source>There is no AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="266"/>
        <location filename="../src/ACEFormat.cpp" line="338"/>
        <source>A name is duplicated</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="274"/>
        <source>No consensus</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="285"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="290"/>
        <source>Bad consensus data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="325"/>
        <location filename="../src/ACEFormat.cpp" line="331"/>
        <source>Bad AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="368"/>
        <source>There is no read note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="376"/>
        <source>No sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="387"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="395"/>
        <location filename="../src/ACEFormat.cpp" line="401"/>
        <source>QA error no clear range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="407"/>
        <source>QA error bad range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="413"/>
        <source>Bad sequence data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="418"/>
        <source>A name is not mutch with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="445"/>
        <source>Alphabet unknown</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="29"/>
        <source>MMDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="53"/>
        <source>Standard residue dictionary not found</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="195"/>
        <location filename="../src/ASNFormat.cpp" line="594"/>
        <source>Unknown error occurred</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="578"/>
        <source>no root element</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="584"/>
        <source>states stack is not empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="611"/>
        <source>First line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="125"/>
        <source>read error occurred</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="133"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="29"/>
        <source>CLUSTALW</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="51"/>
        <source>Illegal header line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="71"/>
        <source>Error parsing file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="83"/>
        <source>Invalid alignment format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="124"/>
        <source>Incorrect number of sequences in block</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="130"/>
        <source>Sequence names are not matched</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="154"/>
        <source>Alphabet is unknown</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="175"/>
        <source>Document is empty.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="95"/>
        <source>Reading entry header</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="142"/>
        <source>Merge error: found annotations without sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="317"/>
        <source>Annotation name is empty</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="325"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="331"/>
        <source>Error parsing location</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="355"/>
        <source>Unexpected line format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="400"/>
        <source>Reading sequence %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="422"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="449"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="464"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="471"/>
        <source>Sequence is truncated</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="478"/>
        <source>Reading annotations %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="488"/>
        <source>Invalid format of feature table</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="530"/>
        <source>Line is too long.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="532"/>
        <source>IO error.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="194"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="174"/>
        <source>Record is truncated.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="22"/>
        <source>EMBL</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="58"/>
        <source>ID is not the first line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="65"/>
        <source>Error parsing ID line</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="24"/>
        <source>FASTA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="74"/>
        <source>Line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="78"/>
        <source>First line is not a FASTA header</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <location filename="../src/FastqFormat.cpp" line="19"/>
        <source>FASTQ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="64"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="137"/>
        <source>Not a valid FASTQ file: %1. The @ identifier is not found.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="161"/>
        <source>Not a valid FASTQ file: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="173"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::GBFeatureUtils</name>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="46"/>
        <source>Sequence related to transcription termination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="47"/>
        <source>Span of the C immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="48"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="49"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="51"/>
        <source>Independent sequence determinations differ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="52"/>
        <source>Displacement loop</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="53"/>
        <source>Span of the D immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="54"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="55"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="56"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="58"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="59"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="60"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="61"/>
        <source>Span of the J immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="62"/>
        <source>Long terminal repeat</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="63"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="64"/>
        <source>Miscellaneous binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="65"/>
        <source>Miscellaneous difference feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="66"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="68"/>
        <source>Miscellaneous, recombination feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="69"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="70"/>
        <source>Miscellaneous signal</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="71"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="72"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="73"/>
        <source>Messenger RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="74"/>
        <source>Span of the N immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="75"/>
        <source>Presented sequence revises a previous version</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="76"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="77"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="78"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="79"/>
        <source>Primary (unprocessed) transcript</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="80"/>
        <source>Primer binding region used with PCR</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="81"/>
        <source>Non-covalent primer binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="82"/>
        <source>A region involved in transcription initiation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="83"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="84"/>
        <source>Ribosome binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="85"/>
        <source>Replication origin for duplex DNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="86"/>
        <source>Sequence containing repeated subsequences</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="87"/>
        <source>One repeated unit of a repeat_region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="88"/>
        <source>Ribosomal RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="89"/>
        <source>Span of the S immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="90"/>
        <source>Satellite repeated sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="91"/>
        <source>Small cytoplasmic RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="92"/>
        <source>Signal peptide coding region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="93"/>
        <source>Small nuclear RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="94"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="95"/>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="96"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="97"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="98"/>
        <source>Sequence causing transcription termination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="99"/>
        <source>Transit peptide coding region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="100"/>
        <source>Transposable element (TN)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="101"/>
        <source>Transfer RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="102"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="103"/>
        <source>Span of the V immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="104"/>
        <source>A related population contains stable mutation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="105"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="106"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="107"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="108"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="109"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="110"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="111"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="112"/>
        <source>&apos;Region&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="113"/>
        <source>&apos;Site&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="130"/>
        <source>Genes</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="145"/>
        <source>Signals</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="164"/>
        <source>Binding</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="169"/>
        <source>Variation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="178"/>
        <source>Repeats</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="185"/>
        <source>RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="193"/>
        <source>Misc</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="207"/>
        <source>Spans</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="18"/>
        <source>GFF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="50"/>
        <source>Parsing error: invalid header</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="54"/>
        <source>Parsing error: file does not contain version header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="58"/>
        <source>Parsing error: format version is not an integer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="62"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="76"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="82"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="88"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="93"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="112"/>
        <source>Parsing error: one of the qualifiers in attributes are incorrect</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="127"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="186"/>
        <source>Fasta sequence inclusion is not supported for GFF</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="30"/>
        <source>Genbank</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="52"/>
        <source>LOCUS is not the first line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="59"/>
        <source>Error parsing LOCUS line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="118"/>
        <source>incomplete SOURCE record</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="216"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="436"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="449"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="460"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="487"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="499"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="506"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="515"/>
        <source>Error writing document</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::IndexFormat</name>
    <message>
        <location filename="../src/IndexFormat.cpp" line="106"/>
        <source>Empty io identifier in item section</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="108"/>
        <source>Empty document format in item section </source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="110"/>
        <source>Empty io adapter identifier in io section</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="112"/>
        <source>Empty url in io section</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="114"/>
        <source>Bad offset in item section</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="116"/>
        <source>Unrecognized io section id in item section</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="119"/>
        <source>Unknown index error occurred</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="235"/>
        <location filename="../src/IndexFormat.cpp" line="241"/>
        <source>Bad value in item section: tag</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="247"/>
        <location filename="../src/IndexFormat.cpp" line="251"/>
        <source>Bad value in item section: val</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="329"/>
        <source>invalid header</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="374"/>
        <source>Unknown section found</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="383"/>
        <source>no items found in index file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="463"/>
        <source>Index</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.cpp" line="483"/>
        <location filename="../src/IndexFormat.cpp" line="501"/>
        <source>Unknown error occurred</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.h" line="45"/>
        <source>Read error occurred</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/IndexFormat.h" line="53"/>
        <source>Write error occurred</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="35"/>
        <source>MSF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="96"/>
        <source>Incorrect format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="114"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="139"/>
        <location filename="../src/MSFFormat.cpp" line="178"/>
        <source>Check sum test failed</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="187"/>
        <source>Alphabet unknown</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="23"/>
        <source>Mega</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="74"/>
        <source>Found sequences of different sizes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="80"/>
        <source>Model is of zero size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="138"/>
        <source>Bad name of sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="158"/>
        <source>Unexpected # in comments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="167"/>
        <source>A comment has not end</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="189"/>
        <source>Unexpected symbol between comments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="249"/>
        <source>Incorrect format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="265"/>
        <source>Identical symbol at the first sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="287"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="306"/>
        <source>Alphabet unknown</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="370"/>
        <location filename="../src/MegaFormat.cpp" line="382"/>
        <source>No header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="375"/>
        <source>No # before header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="387"/>
        <source>Not MEGA-header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="410"/>
        <location filename="../src/MegaFormat.cpp" line="421"/>
        <location filename="../src/MegaFormat.cpp" line="448"/>
        <source>No data in file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="430"/>
        <location filename="../src/MegaFormat.cpp" line="436"/>
        <source>Incorrect title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="480"/>
        <source>Sequence has empty part</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="500"/>
        <source>Bad symbols in a sequence</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="22"/>
        <source>NEXUS</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="606"/>
        <source>#NEXUS header missing</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="16"/>
        <source>Newick Standard</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="168"/>
        <source>Unexpected weight: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="173"/>
        <source>Error parsing weight: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="189"/>
        <source>Unexpected weight start token: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="197"/>
        <source>Unexpected new sibling %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="210"/>
        <source>Unexpected closing bracket :%1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="218"/>
        <location filename="../src/NewickFormat.cpp" line="240"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="252"/>
        <source>Empty file</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <location filename="../src/PDBFormat.cpp" line="32"/>
        <source>PDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="167"/>
        <source>Line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="203"/>
        <source>SPLIT recored is not supported</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="225"/>
        <source>Some mandaroty records are absent</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="334"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="428"/>
        <source>Invalid secondary structure record</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="569"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="17"/>
        <source>Plain text</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="21"/>
        <source>Raw sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="63"/>
        <source>Seqeunce is empty</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="41"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="49"/>
        <source>SAM</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="119"/>
        <source>Unsupported file version &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="173"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="187"/>
        <source>Unexpected reference sequence name value &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="234"/>
        <location filename="../src/SAMFormat.cpp" line="244"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="23"/>
        <source>SCF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="62"/>
        <source>Not a valid SCF file: %1</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SRFastaFormat</name>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="20"/>
        <source>ShortReadsFasta</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="58"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="62"/>
        <source>First line is not a FASTA header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="90"/>
        <source>Incorrect sequence header: unable to parse sequence offset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="94"/>
        <source>Incorrect sequence header: offset is not present</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SRFastaFormat.cpp" line="108"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="441"/>
        <source>invalid file: bad header line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="471"/>
        <source>invalid file: empty sequence name</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="474"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="482"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="493"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="500"/>
        <source>invalid file: empty sequence alignment</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="504"/>
        <source>invalid file: unknown alphabet</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="653"/>
        <source>Stockholm</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="678"/>
        <location filename="../src/StockholmFormat.cpp" line="697"/>
        <source>unknown error occurred</source>
        <translation></translation>
    </message>
</context>
</TS>
