<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="259"/>
        <location filename="../src/FastaFormat.cpp" line="263"/>
        <location filename="../src/FastqFormat.cpp" line="342"/>
        <location filename="../src/PDWFormat.cpp" line="156"/>
        <source>Document is empty.</source>
        <translation>Document is empty.</translation>
    </message>
</context>
<context>
    <name>DocumentFormat</name>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="50"/>
        <source>DatabaseConnection</source>
        <translation>DatabaseConnection</translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="51"/>
        <source>A fake format that was added to implement shared database connection within existing document model.</source>
        <translation>A fake format that was added to implement shared database connection within existing document model.</translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="90"/>
        <source>You have no permissions to modify the content of this database</source>
        <translation>You have no permissions to modify the content of this database</translation>
    </message>
    <message>
        <location filename="../src/DatabaseConnectionFormat.cpp" line="110"/>
        <source>Empty object name</source>
        <translation>Empty object name</translation>
    </message>
</context>
<context>
    <name>L10N</name>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="96"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="121"/>
        <source>Incorrect number of columns in the file</source>
        <translation>Incorrect number of columns in the file</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/BAMUtils.cpp" line="82"/>
        <source>Fail to open &quot;%1&quot; for reading</source>
        <translation>Fail to open &quot;%1&quot; for reading</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="86"/>
        <source>Fail to read the header from the file: &quot;%1&quot;</source>
        <translation>Fail to read the header from the file: &quot;%1&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="90"/>
        <source>Can not build the fasta index for the file: &quot;%1&quot;</source>
        <translation>Can not build the fasta index for the file: &quot;%1&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="94"/>
        <source>Error parsing the reads from the file: &quot;%1&quot;</source>
        <translation>Error parsing the reads from the file: &quot;%1&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="98"/>
        <source>Truncated file: &quot;%1&quot;</source>
        <translation>Truncated file: &quot;%1&quot;</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="460"/>
        <source>Invalid database user permissions set, so UGENE unable to use this database. Connect to your system administrator to fix the issue.</source>
        <translation>Invalid database user permissions set, so UGENE unable to use this database. Connect to your system administrator to fix the issue.</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteBlobInputStream.cpp" line="100"/>
        <source>Can not read data. The database is closed or the data were changed.</source>
        <translation>Can not read data. The database is closed or the data were changed.</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteBlobOutputStream.cpp" line="50"/>
        <source>Can not write data. The database is closed or the data were changed.</source>
        <translation>Can not write data. The database is closed or the data were changed.</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="258"/>
        <source>Incompatible database version. Try to use %1 or higher UGENE version for opening it.</source>
        <translation>Incompatible database version. Try to use %1 or higher UGENE version for opening it.</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="84"/>
        <source>BED</source>
        <translation>BED</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="76"/>
        <source>FPKM Tracking Format</source>
        <translation>FPKM Tracking Format</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="75"/>
        <source>Gap between two components of a genome or transcriptome assembly</source>
        <translation>Gap between two components of a genome or transcriptome assembly</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="76"/>
        <source>Sequence related to transcription termination</source>
        <translation>Sequence related to transcription termination</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="77"/>
        <source>Describes disulfide bonds (for protein files)</source>
        <translation>Describes disulfide bonds (for protein files)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="78"/>
        <source>Span of the C immunological feature</source>
        <translation>Span of the C immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="79"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation>`CAAT box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="80"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation>Sequence coding for amino acids in protein (includes stop codon)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="82"/>
        <source>Independent sequence determinations differ</source>
        <translation>Independent sequence determinations differ</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="83"/>
        <source>Region of biological interest identified as a centromere and which has been experimentally characterized</source>
        <translation>Region of biological interest identified as a centromere and which has been experimentally characterized</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="84"/>
        <source>Displacement loop</source>
        <translation>Displacement loop</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="85"/>
        <source>Span of the D immunological feature</source>
        <translation>Span of the D immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="86"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation>Cis-acting enhancer of promoter function</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="87"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation>Region that codes for part of spliced mRNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="88"/>
        <source>Gap in the sequence</source>
        <translation>Gap in the sequence</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="89"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="91"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation>`GC box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="92"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation>Intervening DNA eliminated by recombination</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="93"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation>Transcribed region excised by mRNA splicing</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="94"/>
        <source>Span of the J immunological feature</source>
        <translation>Span of the J immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="95"/>
        <source>Long terminal repeat</source>
        <translation>Long terminal repeat</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="96"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation>Mature peptide coding region (does not include stop codon)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="97"/>
        <source>Miscellaneous binding site</source>
        <translation>Miscellaneous binding site</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="98"/>
        <source>Miscellaneous difference feature</source>
        <translation>Miscellaneous difference feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="99"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation>Region of biological significance that cannot be described by any other feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="101"/>
        <source>Miscellaneous, recombination feature</source>
        <translation>Miscellaneous, recombination feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="102"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation>Miscellaneous transcript feature not defined by other RNA keys</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="103"/>
        <source>Miscellaneous signal</source>
        <translation>Miscellaneous signal</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="104"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation>Miscellaneous DNA or RNA structure</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="105"/>
        <source>Region of genome containing mobile elements</source>
        <translation>Region of genome containing mobile elements</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="106"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation>The indicated base is a modified nucleotide</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="107"/>
        <source>Messenger RNA</source>
        <translation>Messenger RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="108"/>
        <source>A non-protein-coding gene, other than ribosomal RNA and transfer RNA, the functional molecule of which is the RNA transcrip</source>
        <translation>A non-protein-coding gene, other than ribosomal RNA and transfer RNA, the functional molecule of which is the RNA transcription</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="109"/>
        <source>Span of the N immunological feature</source>
        <translation>Span of the N immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="110"/>
        <source>Presented sequence revises a previous version</source>
        <translation>Presented sequence revises a previous version</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="111"/>
        <source>Region containing polycistronic transcript including a cluster of genes that are under the control of the same regulatory sequences/promotor and in the same biological pathway</source>
        <translation>Region containing polycistronic transcript including a cluster of genes that are under the control of the same regulatory sequences/promotor and in the same biological pathway</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="112"/>
        <source>Origin of transfer; region of a DNA molecule where transfer is initiated during the process of conjugation or mobilization</source>
        <translation>Origin of transfer; region of a DNA molecule where transfer is initiated during the process of conjugation or mobilization</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="113"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation>Signal for cleavage &amp; polyadenylation</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="114"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation>Site at which polyadenine is added to mRNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="115"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation>Any RNA species that is not yet the mature RNA product</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="116"/>
        <source>Primary (unprocessed) transcript</source>
        <translation>Primary (unprocessed) transcript</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="117"/>
        <source>Primer binding region used with PCR</source>
        <translation>Primer binding region used with PCR</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="118"/>
        <source>Non-covalent primer binding site</source>
        <translation>Non-covalent primer binding site</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="119"/>
        <source>A region involved in transcription initiation</source>
        <translation>A region involved in transcription initiation</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="120"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation>Non-covalent protein binding site on DNA or RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="121"/>
        <source>Ribosome binding site</source>
        <translation>Ribosome binding site</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="122"/>
        <source>Replication origin for duplex DNA</source>
        <translation>Replication origin for duplex DNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="123"/>
        <source>Sequence containing repeated subsequences</source>
        <translation>Sequence containing repeated subsequences</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="124"/>
        <source>One repeated unit of a repeat_region</source>
        <translation>One repeated unit of a repeat_region</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="125"/>
        <source>Ribosomal RNA</source>
        <translation>Ribosomal RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="126"/>
        <source>Span of the S immunological feature</source>
        <translation>Span of the S immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="127"/>
        <source>Satellite repeated sequence</source>
        <translation>Satellite repeated sequence</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="128"/>
        <source>Small cytoplasmic RNA</source>
        <translation>Small cytoplasmic RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="129"/>
        <source>Signal peptide coding region</source>
        <translation>Signal peptide coding region</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="130"/>
        <source>Small nuclear RNA</source>
        <translation>Small nuclear RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="131"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation>Identifies the biological source of the specified span of the sequence</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="132"/>
        <source>Hair-pin loop sQObject::tructure in DNA or RNA</source>
        <translation>Hair-pin loop structure in DNA or RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="133"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="134"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation>`TATA box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="135"/>
        <source>Region of biological interest identified as a telomere and which has been experimentally characterized</source>
        <translation>Region of biological interest identified as a telomere and which has been experimentally characterized</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="136"/>
        <source>Sequence causing transcription termination</source>
        <translation>Sequence causing transcription termination</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="137"/>
        <source>Transfer messenger RNA; tmRNA acts as a tRNA first, and then as an mRNA that encodes a peptide tag; the ribosome translates this mRNA region of tmRNA and attaches the encoded peptide tag to the C-terminus of the unfinished protein; this attached tag targets the protein for destruction or proteolysis</source>
        <translation>Transfer messenger RNA; tmRNA acts as a tRNA first, and then as an mRNA that encodes a peptide tag; the ribosome translates this mRNA region of tmRNA and attaches the encoded peptide tag to the C-terminus of the unfinished protein; this attached tag targets the protein for destruction or proteolysis</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="138"/>
        <source>Transit peptide coding region</source>
        <translation>Transit peptide coding region</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="139"/>
        <source>Transposable element (TN)</source>
        <translation>Transposable element (TN)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="140"/>
        <source>Transfer RNA</source>
        <translation>Transfer RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="141"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation>Authors are unsure about the sequence in this region</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="142"/>
        <source>Span of the V immunological feature</source>
        <translation>Span of the V immunological feature</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="143"/>
        <source>Variable segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains; codes for most of the variable region (V_region) and the last few amino acids of the leader peptide</source>
        <translation>Variable segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains; codes for most of the variable region (V_region) and the last few amino acids of the leader peptide</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="144"/>
        <source>A related population contains stable mutation</source>
        <translation>A related population contains stable mutation</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="145"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation>`Pribnow box&apos; in prokaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="146"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation>`-35 box&apos; in prokaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="147"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation>3&apos;-most region of a precursor transcript removed in processing</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="148"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation>3&apos; untranslated region (trailer)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="149"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation>5&apos;-most region of a precursor transcript removed in processing</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="150"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation>5&apos; untranslated region (leader)</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="151"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation>&apos;Protein&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="152"/>
        <source>&apos;Region&apos; feature key</source>
        <translation>&apos;Region&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="153"/>
        <source>&apos;Site&apos; feature key</source>
        <translation>&apos;Site&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="172"/>
        <source>Genes</source>
        <translation>Genes</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="187"/>
        <source>Signals</source>
        <translation>Signals</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="207"/>
        <source>Binding</source>
        <translation>Binding</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="213"/>
        <source>Variation</source>
        <translation>Variation</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="222"/>
        <source>Repeats</source>
        <translation>Repeats</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="229"/>
        <source>RNA</source>
        <translation>RNA</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="239"/>
        <source>Misc</source>
        <translation>Misc</translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="259"/>
        <source>Spans</source>
        <translation>Spans</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="84"/>
        <source>GTF</source>
        <translation>GTF</translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="50"/>
        <source>ABIF</source>
        <translation>ABIF</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="51"/>
        <source>A chromatogram file format</source>
        <translation>A chromatogram file format</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="90"/>
        <source>Not a valid ABIF file: %1</source>
        <translation>Not a valid ABIF file: %1</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="118"/>
        <source>Failed to load sequence from ABI file %1</source>
        <translation>Failed to load sequence from ABI file %1</translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="46"/>
        <source>ACE</source>
        <translation>ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="213"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="378"/>
        <source>A name is not match with AF names</source>
        <translation>A name is not match with AF names</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="423"/>
        <source>First line is not an ace header</source>
        <translation>First line is not an ace header</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="428"/>
        <source>No contig count tag in the header line</source>
        <translation>No contig count tag in the header line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="442"/>
        <source>Must be CO keyword</source>
        <translation>Must be CO keyword</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="209"/>
        <source>Unexpected end of file</source>
        <translation>Unexpected end of file</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="456"/>
        <source>There is no note about reads count</source>
        <translation>There is no note about reads count</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="225"/>
        <location filename="../src/ace/AceFormat.cpp" line="278"/>
        <source>There is no AF note</source>
        <translation>There is no AF note</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="47"/>
        <source>ACE is a format used for storing information about genomic confgurations</source>
        <translation>ACE is a format used for storing information about genomic confgurations</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="229"/>
        <location filename="../src/ace/AceFormat.cpp" line="298"/>
        <source>A name is duplicated</source>
        <translation>A name is duplicated</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="237"/>
        <source>No consensus</source>
        <translation>No consensus</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="248"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation>BQ keyword hasn&apos;t been found</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="253"/>
        <source>Bad consensus data</source>
        <translation>Bad consensus data</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="285"/>
        <location filename="../src/ace/AceFormat.cpp" line="291"/>
        <source>Bad AF note</source>
        <translation>Bad AF note</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="328"/>
        <source>There is no read note</source>
        <translation>There is no read note</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="336"/>
        <source>No sequence</source>
        <translation>No sequence</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="347"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation>QA keyword hasn&apos;t been found</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="355"/>
        <location filename="../src/ace/AceFormat.cpp" line="361"/>
        <source>QA error no clear range</source>
        <translation>QA error no clear range</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="367"/>
        <source>QA error bad range</source>
        <translation>QA error bad range</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="373"/>
        <source>Bad sequence data</source>
        <translation>Bad sequence data</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="403"/>
        <source>Not implemented</source>
        <translation>Not implemented</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="509"/>
        <source>Alphabet unknown</source>
        <translation>Alphabet unknown</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="535"/>
        <source>File doesn&apos;t contain any msa objects</source>
        <translation>File doesn&apos;t contain any msa objects</translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="50"/>
        <source>MMDB</source>
        <translation>MMDB</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="51"/>
        <source>ASN is a format used my the Molecular Modeling Database (MMDB)</source>
        <translation>ASN is a format used my the Molecular Modeling Database (MMDB)</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="74"/>
        <source>Standard residue dictionary not found</source>
        <translation>Standard residue dictionary not found</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="206"/>
        <location filename="../src/ASNFormat.cpp" line="605"/>
        <source>Unknown error occurred</source>
        <translation>Unknown error occurred</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="589"/>
        <source>no root element</source>
        <translation>no root element</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="595"/>
        <source>states stack is not empty</source>
        <translation>states stack is not empty</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="622"/>
        <source>First line is too long</source>
        <translation>First line is too long</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="148"/>
        <source>read error occurred</source>
        <translation>read error occurred</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="156"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation>biostruct3d obj loading error: %1</translation>
    </message>
</context>
<context>
    <name>U2::AbstractVariationFormat</name>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="61"/>
        <source>SNP formats are used to store single-nucleotide polymorphism data</source>
        <translation>SNP formats are used to store single-nucleotide polymorphism data</translation>
    </message>
</context>
<context>
    <name>U2::AceImporter</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="159"/>
        <source>ACE file importer</source>
        <translation>ACE file importer</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="163"/>
        <source>ACE files importer is used to convert conventional ACE files into UGENE database format.Having ACE file converted into UGENE DB format you get an fast and efficient interface to your data with an option to change the content</source>
        <translation>ACE files importer is used to convert conventional ACE files into UGENE database format.Having ACE file converted into UGENE DB format you get an fast and efficient interface to your data with an option to change the content</translation>
    </message>
</context>
<context>
    <name>U2::AceImporterTask</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="49"/>
        <source>ACE file import: %1</source>
        <translation>ACE file import: %1</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="84"/>
        <source>Can&apos;t create a temporary database</source>
        <translation>Can&apos;t create a temporary database</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="144"/>
        <source>Failed to get load task for : %1</source>
        <translation>Failed to get load task for : %1</translation>
    </message>
</context>
<context>
    <name>U2::BAMUtils</name>
    <message>
        <location filename="../src/BAMUtils.cpp" line="312"/>
        <source>Sort bam file: &quot;%1&quot; using %2 Mb of memory. Result sorted file is: &quot;%3&quot;</source>
        <translation>Sort bam file: &quot;%1&quot; using %2 Mb of memory. Result sorted file is: &quot;%3&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="323"/>
        <source>Merging BAM files: &quot;%1&quot;. Resulting merged file is: &quot;%2&quot;</source>
        <translation>Merging BAM files: &quot;%1&quot;. Resulting merged file is: &quot;%2&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="340"/>
        <source>Remove PCR duplicate in BAM file: &quot;%1&quot;. Resulting  file is: &quot;%2&quot;</source>
        <translation>Remove PCR duplicate in BAM file: &quot;%1&quot;. Resulting  file is: &quot;%2&quot;</translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="421"/>
        <source>Build index for bam file: &quot;%1&quot;</source>
        <translation>Build index for bam file: &quot;%1&quot;</translation>
    </message>
</context>
<context>
    <name>U2::BedFormat</name>
    <message>
        <location filename="../src/BedFormat.cpp" line="103"/>
        <source>The BED (Browser Extensible Data) format was developed by UCSC for displaying transcript structures in the genome browser.</source>
        <translation>The BED (Browser Extensible Data) format was developed by UCSC for displaying transcript structures in the genome browser.</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="607"/>
        <source>BED parsing error: incorrect format of the &apos;track&apos; header line!</source>
        <translation>BED parsing error: incorrect format of the &apos;track&apos; header line!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="640"/>
        <source>BED parsing error: unexpected number of fields in the first annotations line!</source>
        <translation>BED parsing error: unexpected number of fields in the first annotations line!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="648"/>
        <source>BED parsing error: incorrect number of fields at line %1!</source>
        <translation>BED parsing error: incorrect number of fields at line %1!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="654"/>
        <source>BED parsing error: a field at line %1 is empty!</source>
        <translation>BED parsing error: a field at line %1 is empty!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="660"/>
        <source>BED parsing error: incorrect coordinates at line %1!</source>
        <translation>BED parsing error: incorrect coordinates at line %1!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="713"/>
        <source>BED parsing error: incorrect score value &apos;%1&apos; at line %2!</source>
        <translation>BED parsing error: incorrect score value &apos;%1&apos; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="718"/>
        <source>BED parsing error: incorrect strand value &apos;%1&apos; at line %2!</source>
        <translation>BED parsing error: incorrect strand value &apos;%1&apos; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="723"/>
        <source>BED parsing error: incorrect thick coordinates at line %1!</source>
        <translation>BED parsing error: incorrect thick coordinates at line %1!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="727"/>
        <source>BED parsing error: incorrect itemRgb value &apos;%1&apos; at line %2!</source>
        <translation>BED parsing error: incorrect itemRgb value &apos;%1&apos; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="732"/>
        <source>BED parsing error: incorrect value of the block parameters at line %1!</source>
        <translation>BED parsing error: incorrect value of the block parameters at line %1!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="761"/>
        <source>Starting BED saving: &apos;%1&apos;</source>
        <translation>Starting BED saving: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="772"/>
        <source>Can not convert GObject to AnnotationTableObject</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="786"/>
        <source>Can not detect chromosome name. &apos;Chr&apos; name will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="803"/>
        <source>You are trying to save joined annotation to BED format! The joining will be lost</source>
        <translation>You are trying to save joined annotation to BED format! The joining will be lost</translation>
    </message>
    <message>
        <source>BED saving error: can&apos;t save an annotation to a BED file - the annotation doesn&apos;t have the &apos;chrom&apos; qualifier!</source>
        <translation type="obsolete">BED saving error: can&apos;t save an annotation to a BED file - the annotation doesn&apos;t have the &apos;chrom&apos; qualifier!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="867"/>
        <source>BED saving error: incorrect thick coordinates in the first annotation!</source>
        <translation>BED saving error: incorrect thick coordinates in the first annotation!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="881"/>
        <source>BED saving error: incorrect block fields in the first annotation!</source>
        <translation>BED saving error: incorrect block fields in the first annotation!</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="890"/>
        <source>BED saving: detected %1 fields per line for file &apos;%2&apos;</source>
        <translation>BED saving: detected %1 fields per line for file &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="897"/>
        <source>BED saving error: an annotation is expected to have &apos;%1&apos; qualifier, but it is absent! Skipping the annotation.</source>
        <translation>BED saving error: an annotation is expected to have &apos;%1&apos; qualifier, but it is absent! Skipping the annotation.</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="957"/>
        <source>BED saving error: an annotation is expected to have the block qualifiers! Skipping the annotation.</source>
        <translation>BED saving error: an annotation is expected to have the block qualifiers! Skipping the annotation.</translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="980"/>
        <source>Finished BED saving: &apos;%1&apos;</source>
        <translation>Finished BED saving: &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::BgzipTask</name>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="43"/>
        <source>Bgzip Compression task</source>
        <translation>Bgzip Compression task</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="50"/>
        <source>Start bgzip compression &apos;%1&apos;</source>
        <translation>Start bgzip compression &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="52"/>
        <source>IOAdapterRegistry is NULL!</source>
        <translation>IOAdapterRegistry is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="55"/>
        <source>IOAdapterFactory is NULL!</source>
        <translation>IOAdapterFactory is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="57"/>
        <source>Can not create IOAdapter!</source>
        <translation>Can not create IOAdapter!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="61"/>
        <source>Can not open input file &apos;%1&apos;</source>
        <translation>Can not open input file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="71"/>
        <source>Can not open output file &apos;%2&apos;</source>
        <translation>Can not open output file &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="85"/>
        <source>Error reading file</source>
        <translation>Error reading file</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="90"/>
        <source>Error writing to file</source>
        <translation>Error writing to file</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="97"/>
        <source>Bgzip compression finished</source>
        <translation>Bgzip compression finished</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="102"/>
        <source>Bgzip compression task was finished with an error: %1</source>
        <translation>Bgzip compression task was finished with an error: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="104"/>
        <source>Bgzip compression task was finished. A new bgzf file is: &lt;a href=&quot;%1&quot;&gt;%2&lt;/a&gt;</source>
        <translation>Bgzip compression task was finished. A new bgzf file is: &lt;a href=&quot;%1&quot;&gt;%2&lt;/a&gt;</translation>
    </message>
</context>
<context>
    <name>U2::CloneAssemblyWithReferenceToDbiTask</name>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="43"/>
        <source>Clone assembly object to the destination database</source>
        <translation>Clone assembly object to the destination database</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="52"/>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="53"/>
        <source>Invalid assembly ID</source>
        <translation>Invalid assembly ID</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="54"/>
        <source>Invalid source database reference</source>
        <translation>Invalid source database reference</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="55"/>
        <source>Invalid destination database reference</source>
        <translation>Invalid destination database reference</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="73"/>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="79"/>
        <source>Can&apos;t get the cloned object</source>
        <translation>Can&apos;t get the cloned object</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="75"/>
        <source>Unexpected result object: expect AssemblyObject, got %1 object</source>
        <translation>Unexpected result object: expect AssemblyObject, got %1 object</translation>
    </message>
    <message>
        <location filename="../src/ace/CloneAssemblyWithReferenceToDbiTask.cpp" line="81"/>
        <source>Unexpected result object: expect U2SequenceObject, got %1 object</source>
        <translation>Unexpected result object: expect U2SequenceObject, got %1 object</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="59"/>
        <source>CLUSTALW</source>
        <translation>CLUSTALW</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="60"/>
        <source>Clustalw is a format for storing multiple sequence alignments</source>
        <translation>Clustalw is a format for storing multiple sequence alignments</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="83"/>
        <source>Illegal header line</source>
        <translation>Illegal header line</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="103"/>
        <source>Error parsing file</source>
        <translation>Error parsing file</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="115"/>
        <source>Invalid alignment format</source>
        <translation>Invalid alignment format</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="156"/>
        <source>Incorrect number of sequences in block</source>
        <translation>Incorrect number of sequences in block</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="162"/>
        <source>Sequence names are not matched</source>
        <translation>Sequence names are not matched</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="185"/>
        <source>Alphabet is unknown</source>
        <translation>Alphabet is unknown</translation>
    </message>
</context>
<context>
    <name>U2::ConvertAceToSqliteTask</name>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="46"/>
        <source>Convert ACE to UGENE database (%1)</source>
        <translation>Convert ACE to UGENE database (%1)</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="58"/>
        <source>Converting assembly from %1 to %2 started</source>
        <translation>Converting assembly from %1 to %2 started</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="66"/>
        <source>IOAdapterFactory is NULL</source>
        <translation>IOAdapterFactory is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="70"/>
        <source>Can&apos;t open file &apos;%1&apos;</source>
        <translation>Can&apos;t open file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="83"/>
        <source>DBI is NULL</source>
        <translation>DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="85"/>
        <source>Object DBI is NULL</source>
        <translation>Object DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="88"/>
        <source>Importing</source>
        <translation>Importing</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="136"/>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="216"/>
        <source>Assembly DBI is NULL</source>
        <translation>Assembly DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="138"/>
        <source>Cross DBI is NULL</source>
        <translation>Cross DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="140"/>
        <source>Sequence DBI is NULL</source>
        <translation>Sequence DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="161"/>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="200"/>
        <source>Invalid source file</source>
        <translation>Invalid source file</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="222"/>
        <source>Packing reads for assembly &apos;%1&apos; (%2 of %3)</source>
        <translation>Packing reads for assembly &apos;%1&apos; (%2 of %3)</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="248"/>
        <source>Attribute DBI is NULL</source>
        <translation>Attribute DBI is NULL</translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="277"/>
        <source>Warning: incorrect maxProw == %1, probably packing was not done! Attribute was not set</source>
        <translation>Warning: incorrect maxProw == %1, probably packing was not done! Attribute was not set</translation>
    </message>
</context>
<context>
    <name>U2::ConvertAssemblyToSamTask</name>
    <message>
        <location filename="../src/tasks/ConvertAssemblyToSamTask.cpp" line="94"/>
        <source>Given file is not valid UGENE database file</source>
        <translation>Given file is not valid UGENE database file</translation>
    </message>
</context>
<context>
    <name>U2::DNAQualityIOUtils</name>
    <message>
        <location filename="../src/DNAQualityIOUtils.cpp" line="67"/>
        <source>No IO adapter found for URL: %1</source>
        <translation>No IO adapter found for URL: %1</translation>
    </message>
</context>
<context>
    <name>U2::Database</name>
    <message>
        <location filename="../src/Database.cpp" line="48"/>
        <source>Not a valid S3-database file: %1</source>
        <translation>Not a valid S3-database file: %1</translation>
    </message>
    <message>
        <location filename="../src/Database.cpp" line="58"/>
        <source>File already exists: %1</source>
        <translation>File already exists: %1</translation>
    </message>
</context>
<context>
    <name>U2::DefaultConvertFileTask</name>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="121"/>
        <source>The formats are not compatible: %1 and %2</source>
        <translation>The formats are not compatible: %1 and %2</translation>
    </message>
</context>
<context>
    <name>U2::DifferentialFormat</name>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="43"/>
        <source>Differential</source>
        <translation>Differential</translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="45"/>
        <source>Differential format is a text-based format for representing Cuffdiff differential output files: expression, splicing, promoters and cds.</source>
        <translation>Differential format is a text-based format for representing Cuffdiff differential output files: expression, splicing, promoters and cds.</translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="302"/>
        <source>Required column is missed: %1</source>
        <translation>Required column is missed: %1</translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormatUtils</name>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="229"/>
        <source>Undefined sequence alphabet</source>
        <translation>Undefined sequence alphabet</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="136"/>
        <source>First line is not an ace header</source>
        <translation>First line is not an ace header</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="140"/>
        <source>There is no assemblies in input file</source>
        <translation>There is no assemblies in input file</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="156"/>
        <source>There are not enough assemblies</source>
        <translation>There are not enough assemblies</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="212"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="464"/>
        <source>Unexpected end of file</source>
        <translation>Unexpected end of file</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="213"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="220"/>
        <source>No contig count tag in the header line</source>
        <translation>No contig count tag in the header line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="232"/>
        <source>Not enough parameters in current line</source>
        <translation>Not enough parameters in current line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="243"/>
        <source>Parameter is not a digit</source>
        <translation>Parameter is not a digit</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="251"/>
        <source>There is no note about reads count</source>
        <translation>There is no note about reads count</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="263"/>
        <source>A name is duplicated</source>
        <translation>A name is duplicated</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="270"/>
        <source>No consensus</source>
        <translation>No consensus</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="280"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation>BQ keyword hasn&apos;t been found</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="283"/>
        <source>Unexpected symbols in consensus data</source>
        <translation>Unexpected symbols in consensus data</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="294"/>
        <source>Can&apos;t find a sequence name in current line</source>
        <translation>Can&apos;t find a sequence name in current line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="304"/>
        <source>An empty sequence name</source>
        <translation>An empty sequence name</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="354"/>
        <source>Invalid AF tag</source>
        <translation>Invalid AF tag</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="371"/>
        <source>A name is duplicated: %1</source>
        <translation>A name is duplicated: %1</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="377"/>
        <source>Not all reads were found</source>
        <translation>Not all reads were found</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="386"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="390"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="422"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="434"/>
        <source>Bad AF note</source>
        <translation>Bad AF note</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="460"/>
        <source>There is no read note</source>
        <translation>There is no read note</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="473"/>
        <source>Invalid RD part</source>
        <translation>Invalid RD part</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="474"/>
        <source>Can&apos;t find the RD tag</source>
        <translation>Can&apos;t find the RD tag</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="483"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation>QA keyword hasn&apos;t been found</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="492"/>
        <source>QA error bad range</source>
        <translation>QA error bad range</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="495"/>
        <source>Unexpected symbols in sequence data</source>
        <translation>Unexpected symbols in sequence data</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="497"/>
        <source>A name is not match with AF names</source>
        <translation>A name is not match with AF names</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="504"/>
        <source>Can&apos;t find clear range start in current line</source>
        <translation>Can&apos;t find clear range start in current line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="505"/>
        <source>Clear range start is invalid</source>
        <translation>Clear range start is invalid</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="512"/>
        <source>Can&apos;t find clear range end in current line</source>
        <translation>Can&apos;t find clear range end in current line</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="513"/>
        <source>Clear range end is invalid</source>
        <translation>Clear range end is invalid</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="537"/>
        <source>There is no next element</source>
        <translation>There is no next element</translation>
    </message>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="56"/>
        <source>Conversion file from %1 to %2</source>
        <translation>Conversion file from %1 to %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/MergeBamTask.cpp" line="42"/>
        <source>Merge BAM files with SAMTools merge</source>
        <translation>Merge BAM files with SAMTools merge</translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="134"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="309"/>
        <source>Reading entry header</source>
        <translation>Reading entry header</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="214"/>
        <source>Merge error: found annotations without sequence</source>
        <translation>Merge error: found annotations without sequence</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="465"/>
        <source>The file contains an incorrect data that describes a qualifier value. </source>
        <translation>The file contains an incorrect data that describes a qualifier value. </translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="466"/>
        <source>The value cannot contain a single quote character. The qualifier is &apos;%1&apos;</source>
        <translation>The value cannot contain a single quote character. The qualifier is &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="534"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="316"/>
        <source>Annotation name is empty</source>
        <translation>Annotation name is empty</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="542"/>
        <source>Error parsing location</source>
        <translation>Error parsing location</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="574"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="370"/>
        <source>Unexpected line format</source>
        <translation>Unexpected line format</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="628"/>
        <source>Reading sequence %1</source>
        <translation>Reading sequence %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="649"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>Error parsing sequence: unexpected empty line</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="676"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation>Error reading sequence: invalid sequence format</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="694"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>Error reading sequence: memory allocation failed</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="709"/>
        <source>Sequence is truncated</source>
        <translation>Sequence is truncated</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="716"/>
        <source>Reading annotations %1</source>
        <translation>Reading annotations %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="726"/>
        <source>Invalid format of feature table</source>
        <translation>Invalid format of feature table</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="765"/>
        <source>Line is too long.</source>
        <translation>Line is too long.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="767"/>
        <source>IO error.</source>
        <translation>IO error.</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="226"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="361"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="207"/>
        <source>Record is truncated.</source>
        <translation>Record is truncated.</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="325"/>
        <source>Annotation start position is empty</source>
        <translation>Annotation start position is empty</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="330"/>
        <source>Annotation end position is empty</source>
        <translation>Annotation end position is empty</translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="43"/>
        <source>EMBL</source>
        <translation>EMBL</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="46"/>
        <source>EMBL Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation>EMBL Flat File Format is a rich format for storing sequences and associated annotations</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="83"/>
        <source>ID is not the first line</source>
        <translation>ID is not the first line</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="90"/>
        <source>Error parsing ID line</source>
        <translation>Error parsing ID line</translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="57"/>
        <source>FASTA</source>
        <translation>FASTA</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="60"/>
        <source>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</source>
        <translation>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="177"/>
        <location filename="../src/FastaFormat.cpp" line="392"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="180"/>
        <location filename="../src/FastaFormat.cpp" line="394"/>
        <source>First line is not a FASTA header</source>
        <translation>First line is not a FASTA header</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="460"/>
        <source>Unreferenced sequence in the beginning of patterns: %1</source>
        <translation>Unreferenced sequence in the beginning of patterns: %1</translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <location filename="../src/FastqFormat.cpp" line="50"/>
        <source>FASTQ</source>
        <translation>FASTQ</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="53"/>
        <source>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data, but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</source>
        <translation>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data, but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="135"/>
        <source>Error while trying to find sequence name start</source>
        <translation>Error while trying to find sequence name start</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="173"/>
        <location filename="../src/FastqFormat.cpp" line="199"/>
        <source>Error while reading sequence</source>
        <translation>Error while reading sequence</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="295"/>
        <source>Not a valid FASTQ file: %1, sequence name differs from quality scores name: %2 and %3</source>
        <translation>Not a valid FASTQ file: %1, sequence name differs from quality scores name: %2 and %3</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="486"/>
        <source>Not a valid FASTQ file, sequence name differs from quality scores name</source>
        <translation>Not a valid FASTQ file, sequence name differs from quality scores name</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="495"/>
        <source>Not a valid FASTQ file. Bad quality scores: inconsistent size.</source>
        <translation>Not a valid FASTQ file. Bad quality scores: inconsistent size.</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="305"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</translation>
    </message>
</context>
<context>
    <name>U2::FpkmTrackingFormat</name>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="94"/>
        <source>The FPKM (fragments per kilobase of exon model per million mapped fragments) Tracking Format is a native Cufflinks format to output estimated expression values.</source>
        <translation>The FPKM (fragments per kilobase of exon model per million mapped fragments) Tracking Format is a native Cufflinks format to output estimated expression values.</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="131"/>
        <source>Internal error: qualifier with name &apos;%1&apos; and &apos;%2&apos; can&apos;t be added</source>
        <translation>Internal error: qualifier with name &apos;%1&apos; and &apos;%2&apos; can&apos;t be added</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="174"/>
        <source>FPKM Tracking Format parsing error: incorrect number of fields at line %1!</source>
        <translation>FPKM Tracking Format parsing error: incorrect number of fields at line %1!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="179"/>
        <source>FPKM Tracking Format parsing error: a field at line %1 is empty!</source>
        <translation>FPKM Tracking Format parsing error: a field at line %1 is empty!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="184"/>
        <source>FPKM Tracking Format parsing error: incorrect coordinates at line %1!</source>
        <translation>FPKM Tracking Format parsing error: incorrect coordinates at line %1!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="196"/>
        <source>FPKM Tracking Format parsing error: different sequence names were detected in an input file. Sequence name &apos;%1&apos; is used.</source>
        <translation>FPKM Tracking Format parsing error: different sequence names were detected in an input file. Sequence name &apos;%1&apos; is used.</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="212"/>
        <source>FPKM Tracking Format parsing error: tracking ID value is empty at line %1!</source>
        <translation>FPKM Tracking Format parsing error: tracking ID value is empty at line %1!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="235"/>
        <source>FPKM Tracking Format parsing error: incorrect length value at line %1!</source>
        <translation>FPKM Tracking Format parsing error: incorrect length value at line %1!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="240"/>
        <source>FPKM Tracking Format parsing error: incorrect coverage value at line %1!</source>
        <translation>FPKM Tracking Format parsing error: incorrect coverage value at line %1!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="525"/>
        <source>Skipped qualifier &apos;%1&apos; while saving a FPKM header.</source>
        <translation>Skipped qualifier &apos;%1&apos; while saving a FPKM header.</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="595"/>
        <source>FPKM Tracking Format saving error: tracking ID shouldn&apos;t be empty!</source>
        <translation>FPKM Tracking Format saving error: tracking ID shouldn&apos;t be empty!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="620"/>
        <source>FPKM Tracking Format saving error: failed to parse locus qualifier &apos;%1&apos;, writing it to the output file anyway!</source>
        <translation>FPKM Tracking Format saving error: failed to parse locus qualifier &apos;%1&apos;, writing it to the output file anyway!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="627"/>
        <source>FPKM Tracking Format saving error: an annotation region (%1, %2) differs from the information stored in the &apos;locus&apos; qualifier (%3, %4). Writing the &apos;locus&apos; qualifier to output!</source>
        <translation>FPKM Tracking Format saving error: an annotation region (%1, %2) differs from the information stored in the &apos;locus&apos; qualifier (%3, %4). Writing the &apos;locus&apos; qualifier to output!</translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="661"/>
        <source>FPKM Tracking Format saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation>FPKM Tracking Format saving error: one or more errors occurred while saving a file, see TRACE log for details!</translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="50"/>
        <source>GFF</source>
        <translation>GFF</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="51"/>
        <source>GFF is a format used for storing features and annotations</source>
        <translation>GFF is a format used for storing features and annotations</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="83"/>
        <source>Parsing error: invalid header</source>
        <translation>Parsing error: invalid header</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="87"/>
        <source>Parsing error: file does not contain version header</source>
        <translation>Parsing error: file does not contain version header</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="91"/>
        <source>Parsing error: format version is not an integer</source>
        <translation>Parsing error: format version is not an integer</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="95"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation>Parsing error: GFF version %1 is not supported</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="243"/>
        <source>Parsing error: file contains empty line %1, line skipped</source>
        <translation>Parsing error: file contains empty line %1, line skipped</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="275"/>
        <source>Parsing error: sequence in FASTA sequence has whitespaces at line %1</source>
        <translation>Parsing error: sequence in FASTA sequence has whitespaces at line %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="282"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation>Parsing error: too few fields at line %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="288"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation>Parsing error: start position at line %1 is not integer</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="294"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation>Parsing error: end position at line %1 is not integer</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="299"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation>Parsing error: incorrect annotation region at line %1</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="351"/>
        <source>Parsing error: incorrect attributes field %1 at line %2</source>
        <translation>Parsing error: incorrect attributes field %1 at line %2</translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="389"/>
        <source>Parsing error: incorrect score parameter at line %1. Score can be a float number or &apos;.&apos; symbol</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="399"/>
        <source>Parsing error: incorrect frame parameter at line %1. Frame can be a number between 0-2 or &apos;.&apos; symbol</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="409"/>
        <source>Parsing error: incorrect strand patameter at line %1. Strand can be &apos;+&apos;,&apos;-&apos; or &apos;.&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="577"/>
        <source>Can not detect chromosome name. &apos;Chr&apos; name will be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="341"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation>Wrong location for joined annotation at line %1. Line was skipped.</translation>
    </message>
</context>
<context>
    <name>U2::GTFFormat</name>
    <message>
        <location filename="../src/GTFFormat.cpp" line="102"/>
        <source>The Gene transfer format (GTF) is a file format used to hold information about gene structure.</source>
        <translation>The Gene transfer format (GTF) is a file format used to hold information about gene structure.</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="172"/>
        <source>GTF parsing error: incorrect number of fields at line %1!</source>
        <translation>GTF parsing error: incorrect number of fields at line %1!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="177"/>
        <source>GTF parsing error: a field at line %1 is empty!</source>
        <translation>GTF parsing error: a field at line %1 is empty!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="182"/>
        <source>GTF parsing error: incorrect coordinates at line %1!</source>
        <translation>GTF parsing error: incorrect coordinates at line %1!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="194"/>
        <source>GTF parsing error: unexpected value of the &quot;feature&quot; value &quot;%1&quot; at line %2!</source>
        <translation>GTF parsing error: unexpected value of the &quot;feature&quot; value &quot;%1&quot; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="213"/>
        <source>GTF parsing error: incorrect score value &quot;%1&quot; at line %2!</source>
        <translation>GTF parsing error: incorrect score value &quot;%1&quot; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="221"/>
        <source>GTF parsing error: incorrect frame value &quot;%1&quot; at line %2!</source>
        <translation>GTF parsing error: incorrect frame value &quot;%1&quot; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="242"/>
        <source>GTF parsing error: invalid attributes format at line %1!</source>
        <translation>GTF parsing error: invalid attributes format at line %1!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="249"/>
        <location filename="../src/GTFFormat.cpp" line="254"/>
        <source>GTF parsing error: mandatory attribute &apos;</source>
        <translation>GTF parsing error: mandatory attribute &apos;</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="250"/>
        <location filename="../src/GTFFormat.cpp" line="255"/>
        <source>&apos; is absent at line %1!</source>
        <translation>&apos; is absent at line %1!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="262"/>
        <source>GTF parsing error: incorrect strand value &quot;%1&quot; at line %2!</source>
        <translation>GTF parsing error: incorrect strand value &quot;%1&quot; at line %2!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="631"/>
        <location filename="../src/GTFFormat.cpp" line="636"/>
        <source>Can&apos;t save an annotation to a GTF file - the annotation doesn&apos;t have the &apos;%1&apos; qualifier!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>GTF saving error: unknown value of the strand qualifier &quot;%1&quot; of an annotation &quot;%2&quot; was skipped!</source>
        <translation type="obsolete">GTF saving error: unknown value of the strand qualifier &quot;%1&quot; of an annotation &quot;%2&quot; was skipped!</translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="655"/>
        <source>GTF saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation>GTF saving error: one or more errors occurred while saving a file, see TRACE log for details!</translation>
    </message>
</context>
<context>
    <name>U2::Genbank::LocationParser</name>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="307"/>
        <source>Ignoring remote entry: %1.%2</source>
        <translation>Ignoring remote entry: %1.%2</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="338"/>
        <location filename="../src/GenbankLocationParser.cpp" line="408"/>
        <source>&apos;a single base from a range&apos; in combination with &apos;sequence span&apos; is not supported</source>
        <translation>&apos;a single base from a range&apos; in combination with &apos;sequence span&apos; is not supported</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="341"/>
        <source>Ignoring &apos;&lt;&apos; at start position</source>
        <translation>Ignoring &apos;&lt;&apos; at start position</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="411"/>
        <source>Ignoring &apos;&gt;&apos; at end position</source>
        <translation>Ignoring &apos;&gt;&apos; at end position</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="473"/>
        <source>Wrong token after JOIN %1</source>
        <translation>Wrong token after JOIN %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="478"/>
        <source>Wrong token after JOIN  - order %1</source>
        <translation>Wrong token after JOIN  - order %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="486"/>
        <source>Can&apos;t parse location on JOIN</source>
        <translation>Can&apos;t parse location on JOIN</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="492"/>
        <location filename="../src/GenbankLocationParser.cpp" line="516"/>
        <location filename="../src/GenbankLocationParser.cpp" line="536"/>
        <location filename="../src/GenbankLocationParser.cpp" line="557"/>
        <source>Must be RIGHT_PARENTHESIS instead of %1</source>
        <translation>Must be RIGHT_PARENTHESIS instead of %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="498"/>
        <source>Wrong token after ORDER %1</source>
        <translation>Wrong token after ORDER %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="510"/>
        <source>Can&apos;t parse location on ORDER</source>
        <translation>Can&apos;t parse location on ORDER</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="522"/>
        <source>Wrong token after BOND %1</source>
        <translation>Wrong token after BOND %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="530"/>
        <source>Can&apos;t parse location on BONDs</source>
        <translation>Can&apos;t parse location on BONDs</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="543"/>
        <source>Must be LEFT_PARENTHESIS instead of %1</source>
        <translation>Must be LEFT_PARENTHESIS instead of %1</translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="551"/>
        <source>Can&apos;t parse location on COMPLEMENT</source>
        <translation>Can&apos;t parse location on COMPLEMENT</translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="165"/>
        <source>Genbank</source>
        <translation>Genbank</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="167"/>
        <source>GenBank Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation>GenBank Flat File Format is a rich format for storing sequences and associated annotations</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="204"/>
        <source>LOCUS is not the first line</source>
        <translation>LOCUS is not the first line</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="214"/>
        <source>Error parsing LOCUS line</source>
        <translation>Error parsing LOCUS line</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="279"/>
        <source>incomplete SOURCE record</source>
        <translation>incomplete SOURCE record</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="459"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="674"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="697"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="708"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="740"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="748"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="755"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="765"/>
        <source>Error writing document</source>
        <translation>Error writing document</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="586"/>
        <source>There is no IOAdapter registry yet</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="588"/>
        <source>IOAdapterFactory is NULL</source>
        <translation type="unfinished">IOAdapterFactory is NULL</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="590"/>
        <source>IOAdapter is NULL</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GzipDecompressTask</name>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="120"/>
        <source>Decompression task</source>
        <translation>Decompression task</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="125"/>
        <source>&apos;%1&apos; is not zipped file</source>
        <translation>&apos;%1&apos; is not zipped file</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="130"/>
        <source>Start decompression &apos;%1&apos;</source>
        <translation>Start decompression &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="132"/>
        <source>IOAdapterRegistry is NULL!</source>
        <translation>IOAdapterRegistry is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="135"/>
        <location filename="../src/tasks/BgzipTask.cpp" line="137"/>
        <source>IOAdapterFactory is NULL!</source>
        <translation>IOAdapterFactory is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="140"/>
        <location filename="../src/tasks/BgzipTask.cpp" line="143"/>
        <source>Can not create IOAdapter!</source>
        <translation>Can not create IOAdapter!</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="155"/>
        <source>Can not open output file &apos;%1&apos;</source>
        <translation>Can not open output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="161"/>
        <source>Can not open input file &apos;%1&apos;</source>
        <translation>Can not open input file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="176"/>
        <source>Error reading file</source>
        <translation>Error reading file</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="182"/>
        <source>Error writing to file</source>
        <translation>Error writing to file</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="187"/>
        <source>Decompression finished</source>
        <translation>Decompression finished</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="192"/>
        <source>Decompression task was finished with an error: %1</source>
        <translation>Decompression task was finished with an error: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/BgzipTask.cpp" line="194"/>
        <source>Decompression task was finished. A new decompressed file is: &lt;a href=&quot;%1&quot;&gt;%1&lt;/a&gt;</source>
        <translation>Decompression task was finished. A new decompressed file is: &lt;a href=&quot;%1&quot;&gt;%1&lt;/a&gt;</translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="64"/>
        <source>MSF</source>
        <translation>MSF</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="66"/>
        <source>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</source>
        <translation>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="126"/>
        <source>Incorrect format</source>
        <translation>Incorrect format</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="144"/>
        <source>Unexpected end of file</source>
        <translation>Unexpected end of file</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="170"/>
        <location filename="../src/MSFFormat.cpp" line="210"/>
        <source>Check sum test failed</source>
        <translation>Check sum test failed</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="218"/>
        <source>Alphabet unknown</source>
        <translation>Alphabet unknown</translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="53"/>
        <source>Mega</source>
        <translation>Mega</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="54"/>
        <source>Mega is a file format of native MEGA program</source>
        <translation>Mega is a file format of native MEGA program</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="107"/>
        <source>Found sequences of different sizes</source>
        <translation>Found sequences of different sizes</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="113"/>
        <source>Model is of zero size</source>
        <translation>Model is of zero size</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="171"/>
        <source>Bad name of sequence</source>
        <translation>Bad name of sequence</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="191"/>
        <source>Unexpected # in comments</source>
        <translation>Unexpected # in comments</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="200"/>
        <source>A comment has not end</source>
        <translation>A comment has not end</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="222"/>
        <source>Unexpected symbol between comments</source>
        <translation>Unexpected symbol between comments</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="282"/>
        <source>Incorrect format</source>
        <translation>Incorrect format</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="295"/>
        <source>Identical symbol at the first sequence</source>
        <translation>Identical symbol at the first sequence</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="319"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation>Incorrect order of sequences&apos; names</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="335"/>
        <source>Alphabet is unknown</source>
        <translation>Alphabet is unknown</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="407"/>
        <location filename="../src/MegaFormat.cpp" line="419"/>
        <source>No header</source>
        <translation>No header</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="412"/>
        <source>No # before header</source>
        <translation>No # before header</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="424"/>
        <source>Not MEGA-header</source>
        <translation>Not MEGA-header</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="447"/>
        <location filename="../src/MegaFormat.cpp" line="458"/>
        <location filename="../src/MegaFormat.cpp" line="485"/>
        <source>No data in file</source>
        <translation>No data in file</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="467"/>
        <location filename="../src/MegaFormat.cpp" line="473"/>
        <source>Incorrect title</source>
        <translation>Incorrect title</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="517"/>
        <source>Sequence has empty part</source>
        <translation>Sequence has empty part</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="537"/>
        <source>Bad symbols in a sequence</source>
        <translation>Bad symbols in a sequence</translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="49"/>
        <source>NEXUS</source>
        <translation>NEXUS</translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="50"/>
        <source>Nexus is a multiple alignment and phylogenetic trees file format</source>
        <translation>Nexus is a multiple alignment and phylogenetic trees file format</translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="696"/>
        <source>#NEXUS header missing</source>
        <translation>#NEXUS header missing</translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="41"/>
        <source>Newick Standard</source>
        <translation>Newick Standard</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="42"/>
        <source>Newick is a simple format used to write out trees in a text file</source>
        <translation>Newick is a simple format used to write out trees in a text file</translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <location filename="../src/PDBFormat.cpp" line="55"/>
        <source>PDB</source>
        <translation>PDB</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="56"/>
        <source>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</source>
        <translation>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="187"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="243"/>
        <source>Some mandatory records are absent</source>
        <translation>Some mandatory records are absent</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="351"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation>PDB warning: unknown residue name: %1</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="441"/>
        <source>Invalid secondary structure record</source>
        <translation>Invalid secondary structure record</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="478"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation>Invalid SEQRES: less then 24 charachters</translation>
    </message>
</context>
<context>
    <name>U2::PDWFormat</name>
    <message>
        <location filename="../src/PDWFormat.cpp" line="57"/>
        <source>pDRAW</source>
        <translation>pDRAW</translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="58"/>
        <source>pDRAW is a sequence file format used by pDRAW software</source>
        <translation>pDRAW is a sequence file format used by pDRAW software</translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="102"/>
        <location filename="../src/PDWFormat.cpp" line="195"/>
        <source>Line is too long</source>
        <translation>Line is too long</translation>
    </message>
</context>
<context>
    <name>U2::PhylipFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="48"/>
        <source>PHYLIP multiple alignment format for phylogenetic applications.</source>
        <translation>PHYLIP multiple alignment format for phylogenetic applications.</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="77"/>
        <source>Alphabet is unknown</source>
        <translation>Alphabet is unknown</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="354"/>
        <source>Error parsing file</source>
        <translation>Error parsing file</translation>
    </message>
</context>
<context>
    <name>U2::PhylipInterleavedFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="242"/>
        <source>PHYLIP Interleaved</source>
        <translation>PHYLIP Interleaved</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="341"/>
        <source>Illegal line</source>
        <translation>Illegal line</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="348"/>
        <source>Wrong header</source>
        <translation>Wrong header</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="383"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation>Number of characters does not correspond to the stated number</translation>
    </message>
</context>
<context>
    <name>U2::PhylipSequentialFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="123"/>
        <source>PHYLIP Sequantial</source>
        <translation>PHYLIP Sequential</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="204"/>
        <source>Illegal line</source>
        <translation>Illegal line</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="211"/>
        <source>Wrong header</source>
        <translation>Wrong header</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="214"/>
        <location filename="../src/PhylipFormat.cpp" line="352"/>
        <source>There is not enough data</source>
        <translation>There is not enough data</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="219"/>
        <location filename="../src/PhylipFormat.cpp" line="360"/>
        <source>Error parsing file</source>
        <translation>Error parsing file</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="233"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation>Number of characters does not correspond to the stated number</translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="38"/>
        <source>Plain text</source>
        <translation>Plain text</translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="40"/>
        <source>A simple plain text file.</source>
        <translation>A simple plain text file.</translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="45"/>
        <source>Raw sequence</source>
        <translation>Raw sequence</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="48"/>
        <source>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</source>
        <translation>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="105"/>
        <source>Sequence is empty</source>
        <translation>Sequence is empty</translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="83"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="92"/>
        <source>SAM</source>
        <translation>SAM</translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="93"/>
        <source>The Sequence Alignment/Map (SAM) format is a generic alignment format forstoring read alignments against reference sequence</source>
        <translation>The Sequence Alignment/Map (SAM) format is a generic alignment format for storing read alignments against reference sequence</translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="48"/>
        <source>SCF</source>
        <translation>SCF</translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="49"/>
        <source>It is Standard Chromatogram Format</source>
        <translation>It is Standard Chromatogram Format</translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="68"/>
        <source>Failed to parse SCF file: %1</source>
        <translation>Failed to parse SCF file: %1</translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="1230"/>
        <source>Failed to load sequence from SCF file %1</source>
        <translation>Failed to load sequence from SCF file %1</translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="467"/>
        <source>invalid file: bad header line</source>
        <translation>invalid file: bad header line</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="497"/>
        <source>invalid file: empty sequence name</source>
        <translation>invalid file: empty sequence name</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="500"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation>invalid file: equal sequence names in one block</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="508"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation>invalid file: sequence names are not equal in blocks</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="519"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation>invalid file: sequences in block are not of equal size</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="526"/>
        <source>invalid file: empty sequence alignment</source>
        <translation>invalid file: empty sequence alignment</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="530"/>
        <source>invalid file: unknown alphabet</source>
        <translation>invalid file: unknown alphabet</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="686"/>
        <source>Stockholm</source>
        <translation>Stockholm</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="687"/>
        <source>A multiple sequence alignments file format</source>
        <translation>A multiple sequence alignments file format</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="708"/>
        <location filename="../src/StockholmFormat.cpp" line="725"/>
        <source>unknown error occurred</source>
        <translation>unknown error occurred</translation>
    </message>
</context>
<context>
    <name>U2::SwissProtPlainTextFormat</name>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="45"/>
        <source>Swiss-Prot</source>
        <translation>Swiss-Prot</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="47"/>
        <source>SwissProt is a format of the UniProtKB/Swiss-prot database used for storing annotated protein sequence</source>
        <translation>SwissProt is a format of the UniProtKB/Swiss-prot database used for storing annotated protein sequence</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation>ID is not the first line</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation>Error parsing ID line</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="101"/>
        <source>Error parsing ID line. Not found sequence length</source>
        <translation>Error parsing ID line. Not found sequence length</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="217"/>
        <source>Reading sequence %1</source>
        <translation>Reading sequence %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="237"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>Error parsing sequence: unexpected empty line</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="260"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>Error reading sequence: memory allocation failed</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="275"/>
        <source>Sequence is truncated</source>
        <translation>Sequence is truncated</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="282"/>
        <source>Reading annotations %1</source>
        <translation>Reading annotations %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="289"/>
        <source>Invalid format of feature table</source>
        <translation>Invalid format of feature table</translation>
    </message>
</context>
<context>
    <name>U2DbiL10n</name>
    <message>
        <location filename="../src/mysql_dbi/MysqlAssemblyDbi.cpp" line="118"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="142"/>
        <source>There is no assembly object with the specified id.</source>
        <translation>There is no assembly object with the specified id.</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlAttributeDbi.cpp" line="209"/>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="223"/>
        <source>Unsupported attribute type: %1</source>
        <translation>Unsupported attribute type: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlAttributeDbi.cpp" line="341"/>
        <source>Required attribute is not found</source>
        <translation>Required attribute is not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="249"/>
        <source>Database url is incorrect</source>
        <translation>Database url is incorrect</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="255"/>
        <source>User login is not specified</source>
        <translation>User login is not specified</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="261"/>
        <source>Host is not specified</source>
        <translation>Host is not specified</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="267"/>
        <source>Database name is not specified</source>
        <translation>Database name is not specified</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="283"/>
        <source>Error opening MySQL database: %1</source>
        <translation>Error opening MySQL database: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="367"/>
        <source>Not a %1 MySQL database: %2, %3</source>
        <translation>Not a %1 MySQL database: %2, %3</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="373"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="285"/>
        <source>Warning! The database was created with a newer %1 version: %2. Not all database features may be supported! Current %1 version: %3.</source>
        <translation>Warning! The database was created with a newer %1 version: %2. Not all database features may be supported! Current %1 version: %3.</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="508"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="340"/>
        <source>Illegal database state: %1</source>
        <translation>Illegal database state: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlDbi.cpp" line="539"/>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="422"/>
        <source>Can&apos;t synchronize database state</source>
        <translation>Can&apos;t synchronize database state</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlFeatureDbi.cpp" line="147"/>
        <source>Annotation table object is not found.</source>
        <translation>Annotation table object is not found.</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlFeatureDbi.cpp" line="189"/>
        <source>Feature is not found.</source>
        <translation>Feature is not found.</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="134"/>
        <source>An object single modification step not found</source>
        <translation>An object single modification step not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="170"/>
        <source>Failed to find user step ID</source>
        <translation>Failed to find user step ID</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="378"/>
        <source>Not main thread</source>
        <translation>Not main thread</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="390"/>
        <source>Can&apos;t create a common user modifications step, previous one is not complete</source>
        <translation>Can&apos;t create a common user modifications step, previous one is not complete</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="453"/>
        <source>Can&apos;t create a common multiple modifications step, previous one is not complete</source>
        <translation>Can&apos;t create a common multiple modifications step, previous one is not complete</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="489"/>
        <source>Failed to create a common user modifications step</source>
        <translation>Failed to create a common user modifications step</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlModDbi.cpp" line="510"/>
        <source>Failed to create a common multiple modifications step</source>
        <translation>Failed to create a common multiple modifications step</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="102"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="118"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="585"/>
        <source>Msa object not found</source>
        <translation>Msa object not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="177"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="751"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="791"/>
        <source>Msa row not found</source>
        <translation>Msa row not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="539"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="570"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1076"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1085"/>
        <source>Unexpected modification type &apos;%1&apos;</source>
        <translation>Unexpected modification type &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="607"/>
        <source>Invalid row position: %1</source>
        <translation>Invalid row position: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1018"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1136"/>
        <source>An error occurred during updating an alignment alphabet</source>
        <translation>An error occurred during updating an alignment alphabet</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1036"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1151"/>
        <source>An error occurred during reverting adding of rows</source>
        <translation>An error occurred during reverting adding of rows</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1054"/>
        <source>An error occurred during reverting addition of a row</source>
        <translation>An error occurred during reverting addition of a row</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1067"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1171"/>
        <source>An error occurred during reverting removing of rows</source>
        <translation>An error occurred during reverting removing of rows</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1080"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1186"/>
        <source>An error occurred during reverting removing of a row</source>
        <translation>An error occurred during reverting removing of a row</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1094"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1197"/>
        <source>An error occurred during updating an alignment gaps</source>
        <translation>An error occurred during updating an alignment gaps</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1107"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1207"/>
        <source>An error occurred during updating an alignment row order</source>
        <translation>An error occurred during updating an alignment row order</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1121"/>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1217"/>
        <source>An error occurred during updating a row info</source>
        <translation>An error occurred during updating a row info</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlMsaDbi.cpp" line="1161"/>
        <source>An error occurred during addition of a row</source>
        <translation>An error occurred during addition of a row</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="574"/>
        <source>Can&apos;t undo an operation for the object</source>
        <translation>Can&apos;t undo an operation for the object</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="637"/>
        <source>Can&apos;t redo an operation for the object</source>
        <translation>Can&apos;t redo an operation for the object</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="767"/>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="859"/>
        <source>Object not found</source>
        <translation>Object not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="808"/>
        <source>Folder not found: %1 (canonical: %2)</source>
        <translation>Folder not found: %1 (canonical: %2)</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="989"/>
        <source>Not an object, id: %1, type: %2</source>
        <translation>Not an object, id: %1, type: %2</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1012"/>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="308"/>
        <source>Unknown object type! Id: %1, type: %2</source>
        <translation>Unknown object type! Id: %1, type: %2</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1054"/>
        <source>Can&apos;t undo an unknown operation: &apos;%1&apos;</source>
        <translation>Can&apos;t undo an unknown operation: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1067"/>
        <source>Can&apos;t redo an unknown operation: &apos;%1&apos;</source>
        <translation>Can&apos;t redo an unknown operation: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1098"/>
        <source>An error occurred during updating an object name</source>
        <translation>An error occurred during updating an object name</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlObjectDbi.cpp" line="1116"/>
        <source>An error occurred during updating an object name!</source>
        <translation>An error occurred during updating an object name!</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="73"/>
        <source>Sequence object not found</source>
        <translation>Sequence object not found</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="118"/>
        <source>Exception was thrown</source>
        <translation>Exception was thrown</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="119"/>
        <source>An exception was thrown during reading sequence data from dbi</source>
        <translation>An exception was thrown during reading sequence data from dbi</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="380"/>
        <source>An error occurred during reverting replacing sequence data</source>
        <translation>An error occurred during reverting replacing sequence data</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlSequenceDbi.cpp" line="395"/>
        <source>An error occurred during replacing sequence data</source>
        <translation>An error occurred during replacing sequence data</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="158"/>
        <source>Invalid variant track type: %1</source>
        <translation>Invalid variant track type: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="186"/>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="219"/>
        <source>Sequence name is not set</source>
        <translation>Sequence name is not set</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="321"/>
        <source>New variant public ID is empty</source>
        <translation>New variant public ID is empty</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/MysqlVariantDbi.cpp" line="337"/>
        <source>New variant track ID is empty</source>
        <translation>New variant track ID is empty</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="124"/>
        <source>Packed data is empty</source>
        <translation>Packed data is empty</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="131"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="515"/>
        <source>Packing method prefix is not supported: %1</source>
        <translation>Packing method prefix is not supported: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="139"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="523"/>
        <source>Data is corrupted, no name end marker found: %1</source>
        <translation>Data is corrupted, no name end marker found: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="148"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="532"/>
        <source>Data is corrupted, no sequence end marker found: %1</source>
        <translation>Data is corrupted, no sequence end marker found: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="157"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="541"/>
        <source>Data is corrupted, no CIGAR end marker found: %1</source>
        <translation>Data is corrupted, no CIGAR end marker found: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="176"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="560"/>
        <source>Data is corrupted, no rnext end marker found: %1</source>
        <translation>Data is corrupted, no rnext end marker found: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlAssemblyUtils.cpp" line="191"/>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="575"/>
        <source>Can not convert pnext to a number: %1</source>
        <translation>Can not convert pnext to a number: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlHelpers.cpp" line="399"/>
        <source>Bound values: </source>
        <translation>Bound values: </translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlHelpers.cpp" line="421"/>
        <source>Cannot start a transaction</source>
        <translation>Cannot start a transaction</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="505"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="127"/>
        <source>Failed to detect assembly storage format: %1</source>
        <translation>Failed to detect assembly storage format: %1</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="517"/>
        <source>Failed to parse range: %1, full: %2</source>
        <translation>Failed to parse range: %1, full: %2</translation>
    </message>
    <message>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="531"/>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="536"/>
        <location filename="../src/mysql_dbi/util/MysqlMultiTableAssemblyAdapter.cpp" line="542"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="152"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="156"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="161"/>
        <source>Failed to parse packed row range info %1</source>
        <translation>Failed to parse packed row range info %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/RTreeAssemblyAdapter.cpp" line="68"/>
        <source>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</source>
        <translation>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="158"/>
        <source>Unsupported reads storage type: %1</source>
        <translation>Unsupported reads storage type: %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="420"/>
        <source>Packing method is not supported: %1</source>
        <translation>Packing method is not supported: %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="508"/>
        <source>Packed data is empty!</source>
        <translation>Packed data is empty!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="212"/>
        <source>Error checking SQLite database: %1!</source>
        <translation>Error checking SQLite database: %1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="225"/>
        <source>Error creating table: %1, error: %2</source>
        <translation>Error creating table: %1, error: %2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="279"/>
        <source>Not a %1 SQLite database: %2</source>
        <translation>Not a %1 SQLite database: %2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="336"/>
        <source>Database is already opened!</source>
        <translation>Database is already opened!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="346"/>
        <source>URL is not specified</source>
        <translation>URL is not specified</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="360"/>
        <source>Error opening SQLite database: %1!</source>
        <translation>Error opening SQLite database: %1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="412"/>
        <source>Database is already closed!</source>
        <translation>Database is already closed!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="416"/>
        <source>Illegal database state %1!</source>
        <translation>Illegal database state %1!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="441"/>
        <source>Failed to close database: %1, err: %2</source>
        <translation>Failed to close database: %1, err: %2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteFeatureDbi.cpp" line="144"/>
        <source>Annotation table object not found.</source>
        <translation>Annotation table object not found.</translation>
    </message>
    <message>
        <source>chrNumber is not set!</source>
        <translation type="obsolete">chrNumber is not set!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteModDbi.cpp" line="139"/>
        <source>An object single modification step not found!</source>
        <translation>An object single modification step not found!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="493"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="508"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="660"/>
        <source>Msa object not found!</source>
        <translation>Msa object not found!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="590"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="707"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="735"/>
        <source>Msa row not found!</source>
        <translation>Msa row not found!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="285"/>
        <source>Not an object! Id: %1, type: %2</source>
        <translation>Not an object! Id: %1, type: %2</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="614"/>
        <source>Can&apos;t undo an operation for the object!</source>
        <translation>Can&apos;t undo an operation for the object!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="699"/>
        <source>Can&apos;t redo an operation for the object!</source>
        <translation>Can&apos;t redo an operation for the object!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="871"/>
        <source>Object not found!</source>
        <translation>Object not found!</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="932"/>
        <source>Object not found.</source>
        <translation>Object not found.</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="968"/>
        <source>Folder not found: %1</source>
        <translation>Folder not found: %1</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSequenceDbi.cpp" line="65"/>
        <source>Sequence object not found.</source>
        <translation>Sequence object not found.</translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="96"/>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="130"/>
        <source>Sequence name is not set!</source>
        <translation>Sequence name is not set!</translation>
    </message>
</context>
</TS>
