<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="209"/>
        <location filename="../src/FastaFormat.cpp" line="204"/>
        <location filename="../src/FastqFormat.cpp" line="248"/>
        <location filename="../src/PDWFormat.cpp" line="130"/>
        <source>Document is empty.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SQLiteL10n</name>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="155"/>
        <source>Attribute not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="210"/>
        <source>Unsupported attribute type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="185"/>
        <source>Not a %1 SQLite database: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="191"/>
        <source>Warning! Database of version %1 was created with a newer %2 version: %3. Not all database features are supported!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="127"/>
        <source>Failed to detect assembly storage format: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="152"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="156"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="161"/>
        <source>Failed to parse packed row range info %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="48"/>
        <source>ABIF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="49"/>
        <source>A chromatogram file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="88"/>
        <source>Not a valid ABIF file: %1</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ACEFormat.cpp" line="45"/>
        <source>ACE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="212"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="377"/>
        <source>A name is not match with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="410"/>
        <source>First line is not an ace header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="415"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="429"/>
        <source>Must be CO keyword</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="208"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="443"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="224"/>
        <location filename="../src/ACEFormat.cpp" line="277"/>
        <source>There is no AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="46"/>
        <source>ACE is a format used for storing information about genomic confgurations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="228"/>
        <location filename="../src/ACEFormat.cpp" line="297"/>
        <source>A name is duplicated</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="236"/>
        <source>No consensus</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="247"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="252"/>
        <source>Bad consensus data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="284"/>
        <location filename="../src/ACEFormat.cpp" line="290"/>
        <source>Bad AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="327"/>
        <source>There is no read note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="335"/>
        <source>No sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="346"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="354"/>
        <location filename="../src/ACEFormat.cpp" line="360"/>
        <source>QA error no clear range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="366"/>
        <source>QA error bad range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="372"/>
        <source>Bad sequence data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ACEFormat.cpp" line="485"/>
        <source>Alphabet unknown</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="51"/>
        <source>MMDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="52"/>
        <source>ASN is a format used my the Molecular Modeling Database (MMDB)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="75"/>
        <source>Standard residue dictionary not found</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="208"/>
        <location filename="../src/ASNFormat.cpp" line="607"/>
        <source>Unknown error occurred</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="591"/>
        <source>no root element</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="597"/>
        <source>states stack is not empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="624"/>
        <source>First line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="149"/>
        <source>read error occurred</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="157"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="53"/>
        <source>CLUSTALW</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="54"/>
        <source>Clustalw is a format for storing multiple sequence alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="77"/>
        <source>Illegal header line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="97"/>
        <source>Error parsing file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="109"/>
        <source>Invalid alignment format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="150"/>
        <source>Incorrect number of sequences in block</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="156"/>
        <source>Sequence names are not matched</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="179"/>
        <source>Alphabet is unknown</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormatUtils</name>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="206"/>
        <source>Undefined sequence alphabet</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="122"/>
        <source>Reading entry header</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="175"/>
        <source>Merge error: found annotations without sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="375"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="311"/>
        <source>Annotation name is empty</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="383"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="389"/>
        <source>Error parsing location</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="415"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="365"/>
        <source>Unexpected line format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="461"/>
        <source>Reading sequence %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="483"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="510"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="528"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="541"/>
        <source>Sequence is truncated</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="548"/>
        <source>Reading annotations %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="558"/>
        <source>Invalid format of feature table</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="597"/>
        <source>Line is too long.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="599"/>
        <source>IO error.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="227"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="215"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="205"/>
        <source>Record is truncated.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="320"/>
        <source>Annotation start position is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="325"/>
        <source>Annotation end position is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="44"/>
        <source>EMBL</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="47"/>
        <source>EMBL Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="51"/>
        <source>FASTA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="54"/>
        <source>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="152"/>
        <location filename="../src/FastaFormat.cpp" line="298"/>
        <source>Line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="155"/>
        <location filename="../src/FastaFormat.cpp" line="300"/>
        <source>First line is not a FASTA header</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <location filename="../src/FastqFormat.cpp" line="42"/>
        <source>FASTQ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="45"/>
        <source>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data, but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="102"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="182"/>
        <source>Not a valid FASTQ file: %1. The @ identifier is not found.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="217"/>
        <source>Not a valid FASTQ file: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="229"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::GBFeatureUtils</name>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="75"/>
        <source>Sequence related to transcription termination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="76"/>
        <source>Span of the C immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="77"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="78"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="80"/>
        <source>Independent sequence determinations differ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="81"/>
        <source>Displacement loop</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="82"/>
        <source>Span of the D immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="83"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="84"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="85"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="87"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="88"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="89"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="90"/>
        <source>Span of the J immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="91"/>
        <source>Long terminal repeat</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="92"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="93"/>
        <source>Miscellaneous binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="94"/>
        <source>Miscellaneous difference feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="95"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="97"/>
        <source>Miscellaneous, recombination feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="98"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="99"/>
        <source>Miscellaneous signal</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="100"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="101"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="102"/>
        <source>Messenger RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="103"/>
        <source>Span of the N immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="104"/>
        <source>Presented sequence revises a previous version</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="105"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="106"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="107"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="108"/>
        <source>Primary (unprocessed) transcript</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="109"/>
        <source>Primer binding region used with PCR</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="110"/>
        <source>Non-covalent primer binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="111"/>
        <source>A region involved in transcription initiation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="112"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="113"/>
        <source>Ribosome binding site</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="114"/>
        <source>Replication origin for duplex DNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="115"/>
        <source>Sequence containing repeated subsequences</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="116"/>
        <source>One repeated unit of a repeat_region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="117"/>
        <source>Ribosomal RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="118"/>
        <source>Span of the S immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="119"/>
        <source>Satellite repeated sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="120"/>
        <source>Small cytoplasmic RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="121"/>
        <source>Signal peptide coding region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="122"/>
        <source>Small nuclear RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="123"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="124"/>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="125"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="126"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="127"/>
        <source>Sequence causing transcription termination</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="128"/>
        <source>Transit peptide coding region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="129"/>
        <source>Transposable element (TN)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="130"/>
        <source>Transfer RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="131"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="132"/>
        <source>Span of the V immunological feature</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="133"/>
        <source>A related population contains stable mutation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="134"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="135"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="136"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="137"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="138"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="139"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="140"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="141"/>
        <source>&apos;Region&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="142"/>
        <source>&apos;Site&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="161"/>
        <source>Genes</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="176"/>
        <source>Signals</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="195"/>
        <source>Binding</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="200"/>
        <source>Variation</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="209"/>
        <source>Repeats</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="216"/>
        <source>RNA</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="224"/>
        <source>Misc</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="238"/>
        <source>Spans</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="45"/>
        <source>GFF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="46"/>
        <source>GFF is a format used for storing features and annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="78"/>
        <source>Parsing error: invalid header</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="82"/>
        <source>Parsing error: file does not contain version header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="86"/>
        <source>Parsing error: format version is not an integer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="90"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="162"/>
        <source>Parsing error: sequence in FASTA sequence has whitespaces at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="169"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="175"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="181"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="186"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="205"/>
        <source>Parsing error: one of the qualifiers in attributes are incorrect at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="220"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="51"/>
        <source>Genbank</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="53"/>
        <source>GenBank Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="86"/>
        <source>LOCUS is not the first line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="93"/>
        <source>Error parsing LOCUS line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="155"/>
        <source>incomplete SOURCE record</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="257"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="479"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="492"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="503"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="530"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="539"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="546"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="555"/>
        <source>Error writing document</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="59"/>
        <source>MSF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="61"/>
        <source>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="121"/>
        <source>Incorrect format</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="139"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="164"/>
        <location filename="../src/MSFFormat.cpp" line="203"/>
        <source>Check sum test failed</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="211"/>
        <source>Alphabet unknown</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="47"/>
        <source>Mega</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="48"/>
        <source>Mega is a file format of native MEGA program</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="95"/>
        <source>Found sequences of different sizes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="101"/>
        <source>Model is of zero size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="159"/>
        <source>Bad name of sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="179"/>
        <source>Unexpected # in comments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="188"/>
        <source>A comment has not end</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="210"/>
        <source>Unexpected symbol between comments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="270"/>
        <source>Incorrect format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="283"/>
        <source>Identical symbol at the first sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="305"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="321"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="384"/>
        <location filename="../src/MegaFormat.cpp" line="396"/>
        <source>No header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="389"/>
        <source>No # before header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="401"/>
        <source>Not MEGA-header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="424"/>
        <location filename="../src/MegaFormat.cpp" line="435"/>
        <location filename="../src/MegaFormat.cpp" line="462"/>
        <source>No data in file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="444"/>
        <location filename="../src/MegaFormat.cpp" line="450"/>
        <source>Incorrect title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="494"/>
        <source>Sequence has empty part</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="514"/>
        <source>Bad symbols in a sequence</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="44"/>
        <source>NEXUS</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="45"/>
        <source>Nexus is a multiple alignment and phylogenetic trees file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="627"/>
        <source>#NEXUS header missing</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="39"/>
        <source>Newick Standard</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="40"/>
        <source>Newick is a simple format used to write out trees in a text file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="197"/>
        <source>Unexpected weight: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="202"/>
        <source>Error parsing weight: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="218"/>
        <source>Unexpected weight start token: %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="226"/>
        <source>Unexpected new sibling %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="239"/>
        <source>Unexpected closing bracket :%1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="247"/>
        <location filename="../src/NewickFormat.cpp" line="269"/>
        <source>Unexpected end of file</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="281"/>
        <source>Empty file</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <location filename="../src/PDBFormat.cpp" line="57"/>
        <source>PDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="58"/>
        <source>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="192"/>
        <source>Line is too long</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="228"/>
        <source>SPLIT recored is not supported</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="248"/>
        <source>Some mandatory records are absent</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="357"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="454"/>
        <source>Invalid secondary structure record</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="595"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PDWFormat</name>
    <message>
        <location filename="../src/PDWFormat.cpp" line="55"/>
        <source>pDRAW</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="56"/>
        <source>pDRAW is a sequence file format used by pDRAW software</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="93"/>
        <location filename="../src/PDWFormat.cpp" line="168"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="37"/>
        <source>Plain text</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="39"/>
        <source>A simple plain text file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="42"/>
        <source>Raw sequence</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="45"/>
        <source>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="82"/>
        <source>Sequence is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="82"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="91"/>
        <source>SAM</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="92"/>
        <source>The Sequence Alignment/Map (SAM) format is a generic alignment format forstoring read alignments against reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="209"/>
        <source>Unsupported file version &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="263"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="293"/>
        <source>Can&apos;t find alphabet for sequence &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="295"/>
        <source>Can&apos;t translation for alphabet &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="325"/>
        <location filename="../src/SAMFormat.cpp" line="331"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="45"/>
        <source>SCF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="46"/>
        <source>It is Standard Chromatogram Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="83"/>
        <source>Failed to parse SCF file: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SQLiteL10N</name>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="82"/>
        <source>There is no assembly object with the specified id.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="98"/>
        <source>Unsupported reads storage type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="121"/>
        <source>Assembly object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="332"/>
        <source>Packing method is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="383"/>
        <source>Packed data is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="388"/>
        <source>Packing method prefix is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="394"/>
        <source>Data is corrupted, no name end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="402"/>
        <source>Data is corrupted, no sequence end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="410"/>
        <source>Data is corrupted, no CIGAR end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="144"/>
        <source>Error checking SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="157"/>
        <source>Error creating table: %1, error: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="235"/>
        <source>Database is already opened!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="239"/>
        <source>Illegal database state: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="245"/>
        <source>URL is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="259"/>
        <source>Error opening SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="304"/>
        <source>Database is already closed!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="308"/>
        <source>Illegal database state %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="325"/>
        <source>Failed to close database: %1, err: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="127"/>
        <source>Not an object! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="192"/>
        <source>Unknown object type! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="433"/>
        <source>Object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="456"/>
        <source>Folder not found :%1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSequenceDbi.cpp" line="64"/>
        <source>Sequence object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="74"/>
        <source>Sequence object is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/RTreeAssemblyAdapter.cpp" line="66"/>
        <source>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="463"/>
        <source>invalid file: bad header line</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="493"/>
        <source>invalid file: empty sequence name</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="496"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="504"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="515"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="522"/>
        <source>invalid file: empty sequence alignment</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="526"/>
        <source>invalid file: unknown alphabet</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="670"/>
        <source>Stockholm</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="671"/>
        <source>A multiple sequence alignments file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="692"/>
        <location filename="../src/StockholmFormat.cpp" line="709"/>
        <source>unknown error occurred</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::SwissProtPlainTextFormat</name>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="45"/>
        <source>Swiss-Prot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="47"/>
        <source>SwissProt is a format of the UniProtKB/Swiss-prot database used for storing annotated protein sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="101"/>
        <source>Error parsing ID line. Not found sequence length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="215"/>
        <source>Reading sequence %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="235"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="258"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="271"/>
        <source>Sequence is truncated</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="278"/>
        <source>Reading annotations %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="285"/>
        <source>Invalid format of feature table</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
