<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="ru">
<context>
    <name>ConvertAceToSqliteDialog</name>
    <message>
        <location filename="../src/ace/ui/ConvertAceToSqliteDialog.ui" line="14"/>
        <source>Import ACE File</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ui/ConvertAceToSqliteDialog.ui" line="28"/>
        <source>Source URL:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ui/ConvertAceToSqliteDialog.ui" line="52"/>
        <source>Destination URL:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ui/ConvertAceToSqliteDialog.ui" line="62"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ui/ConvertAceToSqliteDialog.ui" line="73"/>
        <source>Add to project</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>Document</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="239"/>
        <location filename="../src/FastaFormat.cpp" line="255"/>
        <location filename="../src/FastqFormat.cpp" line="329"/>
        <location filename="../src/PDWFormat.cpp" line="150"/>
        <source>Document is empty.</source>
        <translation type="unfinished">Файл не содержит данных.</translation>
    </message>
</context>
<context>
    <name>L10N</name>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="96"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="487"/>
        <source>There is no next element</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SQLiteL10N</name>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSNPTablesDbi.cpp" line="110"/>
        <source>Variant is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSNPTablesDbi.cpp" line="297"/>
        <source>New filter name is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SQLiteL10n</name>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAttributeDbi.cpp" line="227"/>
        <source>Unsupported attribute type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="247"/>
        <source>Not a %1 SQLite database: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="252"/>
        <source>Warning! Database of version %1 was created with a newer %2 version: %3. Not all database features are supported!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="127"/>
        <source>Failed to detect assembly storage format: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="152"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="156"/>
        <location filename="../src/sqlite_dbi/assembly/MultiTableAssemblyAdapter.cpp" line="161"/>
        <source>Failed to parse packed row range info %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ABIFormat</name>
    <message>
        <location filename="../src/ABIFormat.cpp" line="48"/>
        <source>ABIF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="49"/>
        <source>A chromatogram file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="88"/>
        <source>Not a valid ABIF file: %1</source>
        <translation>Некорректный ABIF файл: %1</translation>
    </message>
    <message>
        <location filename="../src/ABIFormat.cpp" line="116"/>
        <source>Failed to load sequence from ABI file %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ACEFormat</name>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="46"/>
        <source>ACE</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="213"/>
        <source>Line is too long</source>
        <translation type="unfinished">Строка слишком длинная</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="378"/>
        <source>A name is not match with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="423"/>
        <source>First line is not an ace header</source>
        <translation type="unfinished">Первая строка не является заголовком ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="428"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished">Отсутствует тег числа контигов в строке заголовка</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="442"/>
        <source>Must be CO keyword</source>
        <translation type="unfinished">Ожидается ключевое слово CO</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="209"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished">Неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="456"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished">Отсуствтует информация о числе считываний</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="225"/>
        <location filename="../src/ace/AceFormat.cpp" line="278"/>
        <source>There is no AF note</source>
        <translation type="unfinished">Отсутствует </translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="47"/>
        <source>ACE is a format used for storing information about genomic confgurations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="229"/>
        <location filename="../src/ace/AceFormat.cpp" line="298"/>
        <source>A name is duplicated</source>
        <translation type="unfinished">Повторные вхождения имени</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="237"/>
        <source>No consensus</source>
        <translation type="unfinished">Отсутствует консенсунс</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="248"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation type="unfinished">Не найдено ключевое слово BQ</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="253"/>
        <source>Bad consensus data</source>
        <translation type="unfinished">Плохие данные в консенсусе</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="285"/>
        <location filename="../src/ace/AceFormat.cpp" line="291"/>
        <source>Bad AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="328"/>
        <source>There is no read note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="336"/>
        <source>No sequence</source>
        <translation type="unfinished">Нет последовательности</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="347"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation type="unfinished">Ключевое слово QA не было найдено</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="355"/>
        <location filename="../src/ace/AceFormat.cpp" line="361"/>
        <source>QA error no clear range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="367"/>
        <source>QA error bad range</source>
        <translation type="unfinished">Ошибка QA: плохой регион</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="373"/>
        <source>Bad sequence data</source>
        <translation type="unfinished">Некорректные данные в последовательности</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="403"/>
        <source>Not implemented</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="509"/>
        <source>Alphabet unknown</source>
        <translation type="unfinished">Неизвестный алфавит</translation>
    </message>
    <message>
        <location filename="../src/ace/AceFormat.cpp" line="535"/>
        <source>File doesn&apos;t contain any msa objects</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ASNFormat</name>
    <message>
        <location filename="../src/ASNFormat.cpp" line="51"/>
        <source>MMDB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="149"/>
        <source>read error occurred</source>
        <translation>Ошибка чтения</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.h" line="157"/>
        <source>biostruct3d obj loading error: %1</source>
        <translation>Ошибка загрузки трехмерной структуры: %1</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="52"/>
        <source>ASN is a format used my the Molecular Modeling Database (MMDB)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="208"/>
        <location filename="../src/ASNFormat.cpp" line="607"/>
        <source>Unknown error occurred</source>
        <translation>Неизвестная ошибка</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="591"/>
        <source>no root element</source>
        <translation>Отсутствует корневой элемент</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="597"/>
        <source>states stack is not empty</source>
        <translation>Стек состояний не пуст</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="624"/>
        <source>First line is too long</source>
        <translation>Первая строка слишком длинная</translation>
    </message>
    <message>
        <location filename="../src/ASNFormat.cpp" line="75"/>
        <source>Standard residue dictionary not found</source>
        <translation>Стандартный словарь остатков не найден</translation>
    </message>
</context>
<context>
    <name>U2::AbstractVariationFormat</name>
    <message>
        <location filename="../src/AbstractVariationFormat.cpp" line="61"/>
        <source>SNP formats are used to store single-nucleotide polymorphism data</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AceImporter</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="107"/>
        <source>ACE file importer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="110"/>
        <source>ACE files importer is used to convert conventional ACE files into UGENE database format.Having ACE file converted into UGENE DB format you get an fast and efficient interface to your data with an option to change the content</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AceImporterTask</name>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="39"/>
        <source>ACE file import: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImporter.cpp" line="82"/>
        <source>Failed to get load task for : %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AceReader</name>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="132"/>
        <source>First line is not an ace header</source>
        <translation type="unfinished">Первая строка не является заголовком ACE</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="151"/>
        <source>There are not enough assemblies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="207"/>
        <source>Unexpected end of file</source>
        <translation type="unfinished">Неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="208"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="215"/>
        <source>No contig count tag in the header line</source>
        <translation type="unfinished">Отсутствует тег числа контигов в строке заголовка</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="227"/>
        <source>Not enough parameters in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="238"/>
        <source>Parameter is not a digit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="246"/>
        <source>There is no note about reads count</source>
        <translation type="unfinished">Отсуствтует информация о числе считываний</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="258"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="338"/>
        <source>A name is duplicated</source>
        <translation type="unfinished">Повторные вхождения имени</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="265"/>
        <source>No consensus</source>
        <translation type="unfinished">Отсутствует консенсунс</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="275"/>
        <source>BQ keyword hasn&apos;t been found</source>
        <translation type="unfinished">Не найдено ключевое слово BQ</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="278"/>
        <source>Bad consensus data</source>
        <translation type="unfinished">Плохие данные в консенсусе</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="288"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="293"/>
        <source>Can&apos;t find a sequence name in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="296"/>
        <source>An empty sequence name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="323"/>
        <source>There is no AF note</source>
        <translation type="unfinished">Отсутствует </translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="351"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="355"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="387"/>
        <location filename="../src/ace/AceImportUtils.cpp" line="399"/>
        <source>Bad AF note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="425"/>
        <source>There is no read note</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="432"/>
        <source>No sequence</source>
        <translation type="unfinished">Нет последовательности</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="441"/>
        <source>QA keyword hasn&apos;t been found</source>
        <translation type="unfinished">Ключевое слово QA не было найдено</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="448"/>
        <source>QA error bad range</source>
        <translation type="unfinished">Ошибка QA: плохой регион</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="451"/>
        <source>Bad sequence data</source>
        <translation type="unfinished">Некорректные данные в последовательности</translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="453"/>
        <source>A name is not match with AF names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="460"/>
        <source>Can&apos;t find clear range start in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="461"/>
        <source>Clear range start is invalid</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="468"/>
        <source>Can&apos;t find clear range end in current line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/AceImportUtils.cpp" line="469"/>
        <source>Clear range end is invalid</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BAMUtils</name>
    <message>
        <location filename="../src/BAMUtils.cpp" line="247"/>
        <source>Sort bam file: &quot;%1&quot; using %2 Mb of memory. Result sorted file is: &quot;%3&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BAMUtils.cpp" line="273"/>
        <source>Build index for bam file: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BedFormat</name>
    <message>
        <location filename="../src/BedFormat.cpp" line="82"/>
        <source>BED</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="101"/>
        <source>The BED (Browser Extensible Data) format was developed by UCSC for displaying transcript structures in the genome browser.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="622"/>
        <source>BED parsing error: incorrect format of the &apos;track&apos; header line!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="655"/>
        <source>BED parsing error: unexpected number of fields in the first annotations line!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="662"/>
        <source>BED parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="668"/>
        <source>BED parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="674"/>
        <source>BED parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="727"/>
        <source>BED parsing error: incorrect score value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="732"/>
        <source>BED parsing error: incorrect strand value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="737"/>
        <source>BED parsing error: incorrect thick coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="741"/>
        <source>BED parsing error: incorrect itemRgb value &apos;%1&apos; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="746"/>
        <source>BED parsing error: incorrect value of the block parameters at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="776"/>
        <source>Starting BED saving: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="804"/>
        <source>BED saving error: can&apos;t save an annotation to a BED file - the annotation doesn&apos;t have the &apos;chrom&apos; qualifier!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="877"/>
        <source>BED saving error: incorrect thick coordinates in the first annotation!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="891"/>
        <source>BED saving error: incorrect block fields in the first annotation!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="900"/>
        <source>BED saving: detected %1 fields per line for file &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="907"/>
        <source>BED saving error: an annotation is expected to have &apos;%1&apos; qualifier, but it is absent! Skipping the annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="968"/>
        <source>BED saving error: an annotation is expected to have the block qualifiers! Skipping the annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="993"/>
        <source>BED saving error: one or more errors occurred while saving file &apos;%1&apos;, see TRACE log for details!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/BedFormat.cpp" line="997"/>
        <source>Finished BED saving: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWAlnFormat</name>
    <message>
        <source>found_sequences_of_different_size</source>
        <translation type="obsolete">Обнаружены последовательности разной длины</translation>
    </message>
    <message>
        <source>model_of_0_size</source>
        <translation type="obsolete">Модель нулевой длины</translation>
    </message>
    <message>
        <source>illegal_header_line</source>
        <translation type="obsolete">Неверный заголовок</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="56"/>
        <source>Clustalw is a format for storing multiple sequence alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="79"/>
        <source>Illegal header line</source>
        <translation>Неправильная строка заголовка</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="99"/>
        <source>Error parsing file</source>
        <translation>Ошибка разбора формата</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="111"/>
        <source>Invalid alignment format</source>
        <translation>Неверный формат выравнивания</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="152"/>
        <source>Incorrect number of sequences in block</source>
        <translation>Неверное количество последовательностей в блоке</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="158"/>
        <source>Sequence names are not matched</source>
        <translation>Встретилось непарное имя последовательности</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="181"/>
        <source>Alphabet is unknown</source>
        <translation>Неизвестный алфавит</translation>
    </message>
    <message>
        <location filename="../src/ClustalWAlnFormat.cpp" line="55"/>
        <source>CLUSTALW</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ConvertAceToSqliteDialog</name>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="58"/>
        <source>Destination UGENEDB file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="74"/>
        <source>Destination URL is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="78"/>
        <source>Destination URL must point to a local file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="86"/>
        <source>Cannot write to &apos;%1&apos;. You may change permissions or a file path. Destination is changed to default UGENE data path: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="112"/>
        <source>Destination directory &apos;%1&apos; is not writable, please choose different destination URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="120"/>
        <source>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To append data to existing file press &apos;Append&apos;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="123"/>
        <source>Replace</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="124"/>
        <source>Append</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="125"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteDialog.cpp" line="131"/>
        <source>Destination file &apos;%1&apos; cannot be removed</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConvertAceToSqliteTask</name>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="45"/>
        <source>Convert ACE to UGENE database (%1)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="54"/>
        <source>Converting assembly from %1 to %2 started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="60"/>
        <source>Non-local files are not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="64"/>
        <source>IOAdapterFactory is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="68"/>
        <source>Can&apos;t open file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="78"/>
        <source>DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="80"/>
        <source>Object DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="91"/>
        <source>Importing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="129"/>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="208"/>
        <source>Assembly DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="131"/>
        <source>Cross DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="133"/>
        <source>Sequence DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="214"/>
        <source>Packing reads for assembly &apos;%1&apos; (%2 of %3)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="240"/>
        <source>Attribute DBI is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ace/ConvertAceToSqliteTask.cpp" line="269"/>
        <source>Warning: incorrect maxProw == %1, probably packing was not done! Attribute was not set</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConvertAssemblyToSamTask</name>
    <message>
        <location filename="../src/tasks/ConvertAssemblyToSamTask.cpp" line="96"/>
        <source>Given file is not valid UGENE database file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConvertFileTask</name>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="35"/>
        <source>Conversion file to %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="47"/>
        <source>Can not load document from url: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/ConvertFileTask.cpp" line="80"/>
        <source>The formats are not compatible: %1 and %2</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DNAQualityIOUtils</name>
    <message>
        <location filename="../src/DNAQualityIOUtils.cpp" line="72"/>
        <source>No IO adapter found for URL: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Database</name>
    <message>
        <location filename="../src/Database.cpp" line="48"/>
        <source>Not a valid S3-database file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/Database.cpp" line="58"/>
        <source>File already exists: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DifferentialFormat</name>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="42"/>
        <source>Differential</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="44"/>
        <source>Differential format is a text-based format for representing Cuffdiff differential output files: expression, splicing, promoters and cds.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/DifferentialFormat.cpp" line="294"/>
        <source>Required column is missed: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Document</name>
    <message>
        <source>Document is empty.</source>
        <translation type="obsolete">Файл не содержит данных.</translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormatUtils</name>
    <message>
        <location filename="../src/DocumentFormatUtils.cpp" line="221"/>
        <source>Undefined sequence alphabet</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLGenbankAbstractDocument</name>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="500"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="316"/>
        <source>Annotation name is empty</source>
        <translation>Не указано имя аннотации</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="508"/>
        <source>Error parsing location</source>
        <translation>Не указан регион аннотации</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="615"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation>Ошибка чтения последовательности: пустая строка</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="642"/>
        <source>Error reading sequence: invalid sequence format</source>
        <translation>Ошибка чтения последовательности: неверный формат</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="660"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation>Ошибка чтения последовательности: не удалось выделить память</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="540"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="370"/>
        <source>Unexpected line format</source>
        <translation>Слишком длинная строка или неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="131"/>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="295"/>
        <source>Reading entry header</source>
        <translation>Чтение заголовка</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="188"/>
        <source>Merge error: found annotations without sequence</source>
        <translation>Ошибка слияния: обнаружена таблица аннотаций без соотв последовательности.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="594"/>
        <source>Reading sequence %1</source>
        <translation>Чтение последовательности: %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="675"/>
        <source>Sequence is truncated</source>
        <translation>Последовательность повреждена.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="682"/>
        <source>Reading annotations %1</source>
        <translation>Чтение аннотаций: %1</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="692"/>
        <source>Invalid format of feature table</source>
        <translation>Таблица аннотаций повреждена.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="731"/>
        <source>Line is too long.</source>
        <translation>Слишком длинная строка.</translation>
    </message>
    <message>
        <location filename="../src/EMBLGenbankAbstractDocument.cpp" line="733"/>
        <source>IO error.</source>
        <translation>Ошибка чтения</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="227"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="336"/>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="207"/>
        <source>Record is truncated.</source>
        <translation>Данные повреждены.</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="325"/>
        <source>Annotation start position is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="330"/>
        <source>Annotation end position is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::EMBLPlainTextFormat</name>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="44"/>
        <source>EMBL</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="47"/>
        <source>EMBL Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation>Строка идентификатора должна идти первой</translation>
    </message>
    <message>
        <location filename="../src/EMBLPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation>Неверный заголовок</translation>
    </message>
</context>
<context>
    <name>U2::FastaFormat</name>
    <message>
        <location filename="../src/FastaFormat.cpp" line="59"/>
        <source>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="174"/>
        <location filename="../src/FastaFormat.cpp" line="361"/>
        <source>Line is too long</source>
        <translation>Слишком длинная строка</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="177"/>
        <location filename="../src/FastaFormat.cpp" line="363"/>
        <source>First line is not a FASTA header</source>
        <translation>Неправильный заголовок FASTA</translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="425"/>
        <source>Unreferenced sequence in the beginning of patterns: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastaFormat.cpp" line="56"/>
        <source>FASTA</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::FastqFormat</name>
    <message>
        <source>Unexpected end of file</source>
        <translation type="obsolete">Неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="46"/>
        <source>FASTQ</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="49"/>
        <source>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. It was originally developed at the Wellcome Trust Sanger Institute to bundle a FASTA sequence and its quality data, but has recently become the de facto standard for storing the output of high throughput sequencing instruments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="128"/>
        <source>Error while trying to find sequence name start</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="161"/>
        <location filename="../src/FastqFormat.cpp" line="189"/>
        <source>Error while reading sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="279"/>
        <source>Not a valid FASTQ file: %1, sequence name differs from quality scores name: %2 and %3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="467"/>
        <source>Not a valid FASTQ file, sequence name differs from quality scores name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="476"/>
        <source>Not a valid FASTQ file. Bad quality scores: inconsistent size.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Not a valid FASTQ file: %1. The @ identifier is not found.</source>
        <translation type="obsolete">Некорректный FASTQ файл: %1. Идентификатор @ не найден.</translation>
    </message>
    <message>
        <source>Not a valid FASTQ file: %1</source>
        <translation type="obsolete">Некорректный FASTQ файл: %1</translation>
    </message>
    <message>
        <location filename="../src/FastqFormat.cpp" line="290"/>
        <source>Not a valid FASTQ file: %1. Bad quality scores: inconsistent size.</source>
        <translation>Некорректный FASTQ файл: %1. Ошибка вычисления оценки точности: некорректный размер.</translation>
    </message>
</context>
<context>
    <name>U2::FpkmTrackingFormat</name>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="76"/>
        <source>FPKM Tracking Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="94"/>
        <source>The FPKM (fragments per kilobase of exon model per million mapped fragments) Tracking Format is a native Cufflinks format to output estimated expression values.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="130"/>
        <source>Internal error: qualifier with name &apos;%1&apos; and &apos;%2&apos; can&apos;t be added</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="175"/>
        <source>FPKM Tracking Format parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="180"/>
        <source>FPKM Tracking Format parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="185"/>
        <source>FPKM Tracking Format parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="197"/>
        <source>FPKM Tracking Format parsing error: different sequence names were detected in an input file. Sequence name &apos;%1&apos; is used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="213"/>
        <source>FPKM Tracking Format parsing error: tracking ID value is empty at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="236"/>
        <source>FPKM Tracking Format parsing error: incorrect length value at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="241"/>
        <source>FPKM Tracking Format parsing error: incorrect coverage value at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="527"/>
        <source>Skipped qualifier &apos;%1&apos; while saving a FPKM header.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="597"/>
        <source>FPKM Tracking Format saving error: tracking ID shouldn&apos;t be empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="622"/>
        <source>FPKM Tracking Format saving error: failed to parse locus qualifier &apos;%1&apos;, writing it to the output file anyway!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="629"/>
        <source>FPKM Tracking Format saving error: an annotation region (%1, %2) differs from the information stored in the &apos;locus&apos; qualifier (%3, %4). Writing the &apos;locus&apos; qualifier to output!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/FpkmTrackingFormat.cpp" line="663"/>
        <source>FPKM Tracking Format saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GBFeatureUtils</name>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="75"/>
        <source>Sequence related to transcription termination</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="76"/>
        <source>Span of the C immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="77"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="78"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="80"/>
        <source>Independent sequence determinations differ</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="81"/>
        <source>Displacement loop</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="82"/>
        <source>Span of the D immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="83"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="84"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="85"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="87"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="88"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="89"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="90"/>
        <source>Span of the J immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="91"/>
        <source>Long terminal repeat</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="92"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="93"/>
        <source>Miscellaneous binding site</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="94"/>
        <source>Miscellaneous difference feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="95"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="97"/>
        <source>Miscellaneous, recombination feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="98"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="99"/>
        <source>Miscellaneous signal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="100"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="101"/>
        <source>Region of genome containing mobile elements</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="102"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="103"/>
        <source>Messenger RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="104"/>
        <source>Span of the N immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="105"/>
        <source>Presented sequence revises a previous version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="106"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="107"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="108"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="109"/>
        <source>Primary (unprocessed) transcript</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="110"/>
        <source>Primer binding region used with PCR</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="111"/>
        <source>Non-covalent primer binding site</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="112"/>
        <source>A region involved in transcription initiation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="113"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="114"/>
        <source>Ribosome binding site</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="115"/>
        <source>Replication origin for duplex DNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="116"/>
        <source>Sequence containing repeated subsequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="117"/>
        <source>One repeated unit of a repeat_region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="118"/>
        <source>Ribosomal RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="119"/>
        <source>Span of the S immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="120"/>
        <source>Satellite repeated sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="121"/>
        <source>Small cytoplasmic RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="122"/>
        <source>Signal peptide coding region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="123"/>
        <source>Small nuclear RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="124"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="125"/>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="126"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="127"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="128"/>
        <source>Sequence causing transcription termination</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="129"/>
        <source>Transit peptide coding region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="130"/>
        <source>Transposable element (TN)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="131"/>
        <source>Transfer RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="132"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="133"/>
        <source>Span of the V immunological feature</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="134"/>
        <source>A related population contains stable mutation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="135"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="136"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="137"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="138"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="139"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="140"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="141"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="142"/>
        <source>&apos;Region&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="143"/>
        <source>&apos;Site&apos; feature key</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="162"/>
        <source>Genes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="177"/>
        <source>Signals</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="196"/>
        <source>Binding</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="201"/>
        <source>Variation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="210"/>
        <source>Repeats</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="217"/>
        <source>RNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="225"/>
        <source>Misc</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankFeatures.cpp" line="240"/>
        <source>Spans</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GFFFormat</name>
    <message>
        <location filename="../src/GFFFormat.cpp" line="45"/>
        <source>GFF</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="46"/>
        <source>GFF is a format used for storing features and annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="78"/>
        <source>Parsing error: invalid header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="82"/>
        <source>Parsing error: file does not contain version header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="86"/>
        <source>Parsing error: format version is not an integer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="90"/>
        <source>Parsing error: GFF version %1 is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="221"/>
        <source>Parsing error: file contains empty line %1, line skipped</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="247"/>
        <source>Parsing error: sequence in FASTA sequence has whitespaces at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="254"/>
        <source>Parsing error: too few fields at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="260"/>
        <source>Parsing error: start position at line %1 is not integer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="266"/>
        <source>Parsing error: end position at line %1 is not integer</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="271"/>
        <source>Parsing error: incorrect annotation region at line %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="315"/>
        <source>Parsing error: incorrect attributes field %1 at line %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GFFFormat.cpp" line="305"/>
        <source>Wrong location for joined annotation at line %1. Line was skipped.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GTFFormat</name>
    <message>
        <location filename="../src/GTFFormat.cpp" line="85"/>
        <source>GTF</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="103"/>
        <source>The Gene transfer format (GTF) is a file format used to hold information about gene structure.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="177"/>
        <source>GTF parsing error: incorrect number of fields at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="182"/>
        <source>GTF parsing error: a field at line %1 is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="187"/>
        <source>GTF parsing error: incorrect coordinates at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="199"/>
        <source>GTF parsing error: unexpected value of the &quot;feature&quot; value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="220"/>
        <source>GTF parsing error: incorrect score value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="229"/>
        <source>GTF parsing error: incorrect frame value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="250"/>
        <source>GTF parsing error: invalid attributes format at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="257"/>
        <location filename="../src/GTFFormat.cpp" line="262"/>
        <source>GTF parsing error: mandatory attribute &apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="258"/>
        <location filename="../src/GTFFormat.cpp" line="263"/>
        <source>&apos; is absent at line %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="270"/>
        <source>GTF parsing error: incorrect strand value &quot;%1&quot; at line %2!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="628"/>
        <source>GTF saving error: unknown value of the strand qualifier &quot;%1&quot; of an annotation &quot;%2&quot; was skipped!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GTFFormat.cpp" line="669"/>
        <source>GTF saving error: one or more errors occurred while saving a file, see TRACE log for details!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Genbank::LocationParser</name>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="300"/>
        <source>Ignoring remote entry: %1.%2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="331"/>
        <location filename="../src/GenbankLocationParser.cpp" line="401"/>
        <source>&apos;a single base from a range&apos; in combination with &apos;sequence span&apos; is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="334"/>
        <source>Ignoring &apos;&lt;&apos; at start position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="404"/>
        <source>Ignoring &apos;&gt;&apos; at end position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="466"/>
        <source>Wrong token after JOIN %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="471"/>
        <source>Wrong token after JOIN  - order %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="479"/>
        <source>Can&apos;t parse location on JOIN</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="488"/>
        <location filename="../src/GenbankLocationParser.cpp" line="515"/>
        <location filename="../src/GenbankLocationParser.cpp" line="538"/>
        <source>Must be RIGHT_PARENTHESIS instead of %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="494"/>
        <source>Wrong token after ORDER %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="506"/>
        <source>Can&apos;t parse location on ORDER</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="521"/>
        <source>Must be LEFT_PARENTHESIS instead of %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankLocationParser.cpp" line="529"/>
        <source>Can&apos;t parse location on COMPLEMENT</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenbankPlainTextFormat</name>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="417"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="614"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="634"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="645"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="678"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="686"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="693"/>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="702"/>
        <source>Error writing document</source>
        <translation>Ошибка записи</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="158"/>
        <source>Genbank</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="160"/>
        <source>GenBank Flat File Format is a rich format for storing sequences and associated annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="193"/>
        <source>LOCUS is not the first line</source>
        <translation>Строка локуса должна идти первой</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="200"/>
        <source>Error parsing LOCUS line</source>
        <translation>Ошибка чтения локуса</translation>
    </message>
    <message>
        <location filename="../src/GenbankPlainTextFormat.cpp" line="262"/>
        <source>incomplete SOURCE record</source>
        <translation>Данные повреждены: запись SOURCE</translation>
    </message>
</context>
<context>
    <name>U2::IndexFormat</name>
    <message>
        <source>Index</source>
        <translation type="obsolete">Индекс</translation>
    </message>
    <message>
        <source>Read error occurred</source>
        <translation type="obsolete">Ошибка чтения</translation>
    </message>
    <message>
        <source>Write error occurred</source>
        <translation type="obsolete">Ошибка записи</translation>
    </message>
    <message>
        <source>Empty io identifier in item section</source>
        <translation type="obsolete">Пустой идентификатор в секции &apos;item&apos;</translation>
    </message>
    <message>
        <source>Empty document format in item section </source>
        <translation type="obsolete">Не задан формат документа в секции &apos;item&apos;</translation>
    </message>
    <message>
        <source>Empty io adapter identifier in io section</source>
        <translation type="obsolete">Не задан идентификатор адаптера файла в секции &apos;io&apos;</translation>
    </message>
    <message>
        <source>Empty url in io section</source>
        <translation type="obsolete">Не задан путь к файлу в секции &apos;io&apos;</translation>
    </message>
    <message>
        <source>Bad offset in item section</source>
        <translation type="obsolete">Неверное смещение в секции &apos;io&apos;</translation>
    </message>
    <message>
        <source>Unrecognized io section id in item section</source>
        <translation type="obsolete">Неверный идентификатор адаптера файла в сеции &apos;item&apos;</translation>
    </message>
    <message>
        <source>Unknown index error occurred</source>
        <translation type="obsolete">Произошла неизвестная ошибка</translation>
    </message>
    <message>
        <source>invalid header</source>
        <translation type="obsolete">Неверный заголовок файла</translation>
    </message>
    <message>
        <source>Unknown section found</source>
        <translation type="obsolete">Неизвестная секция файла</translation>
    </message>
    <message>
        <source>no items found in index file</source>
        <translation type="obsolete">Не найдено ни одной секции &apos;item&apos;</translation>
    </message>
    <message>
        <source>Unknown error occurred</source>
        <translation type="obsolete">Неизвестная ошибка</translation>
    </message>
</context>
<context>
    <name>U2::MSFFormat</name>
    <message>
        <location filename="../src/MSFFormat.cpp" line="63"/>
        <source>MSF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="65"/>
        <source>MSF format is used to store multiple aligned sequences. Files include the sequence name and the sequence itself, which is usually aligned with other sequences in the file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="125"/>
        <source>Incorrect format</source>
        <translation>Неверный формат</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="143"/>
        <source>Unexpected end of file</source>
        <translation>Неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="169"/>
        <location filename="../src/MSFFormat.cpp" line="209"/>
        <source>Check sum test failed</source>
        <translation>Проверка контрольной суммы не удалась</translation>
    </message>
    <message>
        <location filename="../src/MSFFormat.cpp" line="217"/>
        <source>Alphabet unknown</source>
        <translation>Неизвестный алфавит</translation>
    </message>
</context>
<context>
    <name>U2::MegaFormat</name>
    <message>
        <location filename="../src/MegaFormat.cpp" line="52"/>
        <source>Mega</source>
        <translation>MEGA</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="53"/>
        <source>Mega is a file format of native MEGA program</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="106"/>
        <source>Found sequences of different sizes</source>
        <translation>Обнаружены последовательности разной длины</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="112"/>
        <source>Model is of zero size</source>
        <translation>Модель нулевой длины</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="170"/>
        <source>Bad name of sequence</source>
        <translation>Неверное имя последовательности</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="190"/>
        <source>Unexpected # in comments</source>
        <translation>Неожиданный символ &quot;#&quot; в комментариях</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="199"/>
        <source>A comment has not end</source>
        <translation>У комментария отсутствует окончание</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="221"/>
        <source>Unexpected symbol between comments</source>
        <translation>Неоижданные символы между комментариями</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="281"/>
        <source>Incorrect format</source>
        <translation>Неверный формат</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="294"/>
        <source>Identical symbol at the first sequence</source>
        <translation>Идентичный символ в первой последовательности</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="318"/>
        <source>Incorrect order of sequences&apos; names</source>
        <translation>Неправильный порядок имён последовательностей</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="334"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished">Неизвестный алфавит</translation>
    </message>
    <message>
        <source>Alphabet unknown</source>
        <translation type="obsolete">Неизвестный алфавит</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="406"/>
        <location filename="../src/MegaFormat.cpp" line="418"/>
        <source>No header</source>
        <translation type="unfinished">Отсутствует заголовок</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="411"/>
        <source>No # before header</source>
        <translation type="unfinished">Отсутствует символ &quot;#&quot; перед заголовком</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="423"/>
        <source>Not MEGA-header</source>
        <translation type="unfinished">Не является заголовком MEGA</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="446"/>
        <location filename="../src/MegaFormat.cpp" line="457"/>
        <location filename="../src/MegaFormat.cpp" line="484"/>
        <source>No data in file</source>
        <translation type="unfinished">Отсутствуют данные в файле</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="466"/>
        <location filename="../src/MegaFormat.cpp" line="472"/>
        <source>Incorrect title</source>
        <translation type="unfinished">Неправильный заголовок</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="516"/>
        <source>Sequence has empty part</source>
        <translation type="unfinished">Часть последовательности пустая</translation>
    </message>
    <message>
        <location filename="../src/MegaFormat.cpp" line="536"/>
        <source>Bad symbols in a sequence</source>
        <translation type="unfinished">Некорректные символы в последовательности</translation>
    </message>
</context>
<context>
    <name>U2::NEXUSFormat</name>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="47"/>
        <source>NEXUS</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="48"/>
        <source>Nexus is a multiple alignment and phylogenetic trees file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NEXUSFormat.cpp" line="686"/>
        <source>#NEXUS header missing</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::NewickFormat</name>
    <message>
        <location filename="../src/NewickFormat.cpp" line="39"/>
        <source>Newick Standard</source>
        <translation>Стандарт Newick</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="40"/>
        <source>Newick is a simple format used to write out trees in a text file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="198"/>
        <source>Unexpected weight: %1</source>
        <translation>Неожиданный вес: %1</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="203"/>
        <source>Error parsing weight: %1</source>
        <translation>Ошибка обработки веса: %1</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="219"/>
        <source>Unexpected weight start token: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="227"/>
        <source>Unexpected new sibling %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="240"/>
        <source>Unexpected closing bracket :%1</source>
        <translation>Неожиданная закрывающая скобка: %1</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="248"/>
        <location filename="../src/NewickFormat.cpp" line="270"/>
        <source>Unexpected end of file</source>
        <translation>Неожиданный конец файла</translation>
    </message>
    <message>
        <location filename="../src/NewickFormat.cpp" line="282"/>
        <source>Empty file</source>
        <translation>Файл пуст</translation>
    </message>
</context>
<context>
    <name>U2::PDBFormat</name>
    <message>
        <location filename="../src/PDBFormat.cpp" line="194"/>
        <source>Line is too long</source>
        <translation>Слишком длинная строка</translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="230"/>
        <source>SPLIT recored is not supported</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="58"/>
        <source>The Protein Data Bank (PDB) format provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="250"/>
        <source>Some mandatory records are absent</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="359"/>
        <source>PDB warning: unknown residue name: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="456"/>
        <source>Invalid secondary structure record</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="597"/>
        <source>Invalid SEQRES: less then 24 charachters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDBFormat.cpp" line="57"/>
        <source>PDB</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::PDWFormat</name>
    <message>
        <location filename="../src/PDWFormat.cpp" line="57"/>
        <source>pDRAW</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="58"/>
        <source>pDRAW is a sequence file format used by pDRAW software</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PDWFormat.cpp" line="99"/>
        <location filename="../src/PDWFormat.cpp" line="188"/>
        <source>Line is too long</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhylipFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="47"/>
        <source>PHYLIP multiple alignment format for phylogenetic applications.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="76"/>
        <source>Alphabet is unknown</source>
        <translation type="unfinished">Неизвестный алфавит</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="352"/>
        <source>Error parsing file</source>
        <translation type="unfinished">Ошибка разбора формата</translation>
    </message>
</context>
<context>
    <name>U2::PhylipInterleavedFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="240"/>
        <source>PHYLIP Interleaved</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="339"/>
        <source>Illegal line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="346"/>
        <source>Wrong header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="381"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhylipSequentialFormat</name>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="121"/>
        <source>PHYLIP Sequantial</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="202"/>
        <source>Illegal line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="209"/>
        <source>Wrong header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="212"/>
        <location filename="../src/PhylipFormat.cpp" line="350"/>
        <source>There is not enough data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="217"/>
        <location filename="../src/PhylipFormat.cpp" line="358"/>
        <source>Error parsing file</source>
        <translation type="unfinished">Ошибка разбора формата</translation>
    </message>
    <message>
        <location filename="../src/PhylipFormat.cpp" line="231"/>
        <source>Number of characters does not correspond to the stated number</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PlainTextFormat</name>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="37"/>
        <source>Plain text</source>
        <translation>Текст</translation>
    </message>
    <message>
        <location filename="../src/PlainTextFormat.cpp" line="39"/>
        <source>A simple plain text file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RawDNASequenceFormat</name>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="44"/>
        <source>Raw sequence</source>
        <translation>Произвольная последовательность</translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="47"/>
        <source>Raw sequence file - a whole content of the file is treated either as a single nucleotide or peptide sequence UGENE will remove all non-alphabetic chars from the result sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/RawDNASequenceFormat.cpp" line="102"/>
        <source>Sequence is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Seqeunce is empty</source>
        <translation type="obsolete">Последовательность пуста</translation>
    </message>
</context>
<context>
    <name>U2::SAMFormat</name>
    <message>
        <location filename="../src/SAMFormat.cpp" line="84"/>
        <source>Field &quot;%1&quot; not matched pattern &quot;%2&quot;, expected pattern &quot;%3&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="93"/>
        <source>SAM</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SAMFormat.cpp" line="94"/>
        <source>The Sequence Alignment/Map (SAM) format is a generic alignment format forstoring read alignments against reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Unsupported file version &quot;%1&quot;</source>
        <translation type="obsolete">Неподдерживаемая версия файла &quot;%1&quot;</translation>
    </message>
    <message>
        <source>Unexpected end of file</source>
        <translation type="obsolete">Неожиданный конец файла</translation>
    </message>
    <message>
        <source>Unexpected reference sequence name value &quot;%1&quot;</source>
        <translation type="obsolete">Неожиданное значение имени референсной последовательности &quot;%1&quot;</translation>
    </message>
    <message>
        <source>Alphabet is unknown</source>
        <translation type="obsolete">Неизвестный алфавит</translation>
    </message>
</context>
<context>
    <name>U2::SCFFormat</name>
    <message>
        <location filename="../src/SCFFormat.cpp" line="45"/>
        <source>SCF</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="46"/>
        <source>It is Standard Chromatogram Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="66"/>
        <source>Failed to parse SCF file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SCFFormat.cpp" line="1220"/>
        <source>Failed to load sequence from SCF file %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Not a valid SCF file: %1</source>
        <translation type="obsolete">Некорректный SCF файл: %1</translation>
    </message>
</context>
<context>
    <name>U2::SQLiteL10N</name>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="86"/>
        <source>There is no assembly object with the specified id.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="102"/>
        <source>Unsupported reads storage type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="125"/>
        <source>Assembly object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="352"/>
        <source>Packing method is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="440"/>
        <source>Packed data is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="447"/>
        <source>Packing method prefix is not supported: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="455"/>
        <source>Data is corrupted, no name end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="464"/>
        <source>Data is corrupted, no sequence end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="473"/>
        <source>Data is corrupted, no CIGAR end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="492"/>
        <source>Data is corrupted, no rnext end marker found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteAssemblyDbi.cpp" line="507"/>
        <source>Can not convert pnext to a number: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="197"/>
        <source>Error checking SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="210"/>
        <source>Error creating table: %1, error: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="298"/>
        <source>Database is already opened!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="302"/>
        <source>Illegal database state: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="308"/>
        <source>URL is not specified</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="322"/>
        <source>Error opening SQLite database: %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="369"/>
        <source>Database is already closed!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="373"/>
        <source>Illegal database state %1!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteDbi.cpp" line="393"/>
        <source>Failed to close database: %1, err: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="167"/>
        <source>Not an object! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="251"/>
        <source>Unknown object type! Id: %1, type: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="440"/>
        <source>Can&apos;t undo an operation for the object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="525"/>
        <source>Can&apos;t redo an operation for the object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="714"/>
        <source>Object not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="776"/>
        <source>Object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteObjectDbi.cpp" line="796"/>
        <source>Folder not found :%1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteSequenceDbi.cpp" line="70"/>
        <source>Sequence object not found.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/assembly/RTreeAssemblyAdapter.cpp" line="67"/>
        <source>Error during RTree index creation: %1! Check if SQLite library has RTree index support!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteKnownMutationsDbi.cpp" line="51"/>
        <source>chrNumber is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteModDbi.cpp" line="127"/>
        <source>An object single modification step not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="507"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="522"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="674"/>
        <source>Msa object not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="604"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="721"/>
        <location filename="../src/sqlite_dbi/SQLiteMsaDbi.cpp" line="749"/>
        <source>Msa row not found!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="96"/>
        <location filename="../src/sqlite_dbi/SQLiteVariantDbi.cpp" line="130"/>
        <source>Sequence name is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SRFastaFormat</name>
    <message>
        <source>Line is too long</source>
        <translation type="obsolete">Строка слишком длинная</translation>
    </message>
    <message>
        <source>first_line_is_not_a_fasta_header</source>
        <translation type="obsolete">Неправильный заголовок FASTA</translation>
    </message>
    <message>
        <source>Incorrect sequence header: offset is not present</source>
        <translation type="obsolete">Неправильный заголовок последовательности: отсутствует смещение</translation>
    </message>
    <message>
        <source>Incorrect sequence header: unable to parse sequence offset</source>
        <translation type="obsolete">Неправильный заголовок последовательности: невозможно обработать смещение последовательности</translation>
    </message>
    <message>
        <source>First line is not a FASTA header</source>
        <translation type="obsolete">Первая строка не является заголовком FASTA</translation>
    </message>
    <message>
        <source>Alphabet is unknown</source>
        <translation type="obsolete">Неизвестный алфавит</translation>
    </message>
</context>
<context>
    <name>U2::StockholmFormat</name>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="682"/>
        <source>Stockholm</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="683"/>
        <source>A multiple sequence alignments file format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="704"/>
        <location filename="../src/StockholmFormat.cpp" line="721"/>
        <source>unknown error occurred</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="467"/>
        <source>invalid file: bad header line</source>
        <translation>Неверный заголовок</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="497"/>
        <source>invalid file: empty sequence name</source>
        <translation>Не указано имя последовательности</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="500"/>
        <source>invalid file: equal sequence names in one block</source>
        <translation>одинаковые имена последовательностей в блоке</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="508"/>
        <source>invalid file: sequence names are not equal in blocks</source>
        <translation>Непарное имя последовательности в блоке</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="519"/>
        <source>invalid file: sequences in block are not of equal size</source>
        <translation>Блок содержит последовательности разной длины</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="526"/>
        <source>invalid file: empty sequence alignment</source>
        <translation>Выравнивание не содержит последовательностей</translation>
    </message>
    <message>
        <location filename="../src/StockholmFormat.cpp" line="530"/>
        <source>invalid file: unknown alphabet</source>
        <translation>Не удалось установить алфавит</translation>
    </message>
</context>
<context>
    <name>U2::SwissProtPlainTextFormat</name>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="45"/>
        <source>Swiss-Prot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="47"/>
        <source>SwissProt is a format of the UniProtKB/Swiss-prot database used for storing annotated protein sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="84"/>
        <source>ID is not the first line</source>
        <translation type="unfinished">Строка идентификатора должна идти первой</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="91"/>
        <source>Error parsing ID line</source>
        <translation type="unfinished">Неверный заголовок</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="101"/>
        <source>Error parsing ID line. Not found sequence length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="217"/>
        <source>Reading sequence %1</source>
        <translation type="unfinished">Чтение последовательности: %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="237"/>
        <source>Error parsing sequence: unexpected empty line</source>
        <translation type="unfinished">Ошибка чтения последовательности: пустая строка</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="260"/>
        <source>Error reading sequence: memory allocation failed</source>
        <translation type="unfinished">Ошибка чтения последовательности: не удалось выделить память</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="275"/>
        <source>Sequence is truncated</source>
        <translation type="unfinished">Последовательность повреждена.</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="282"/>
        <source>Reading annotations %1</source>
        <translation type="unfinished">Чтение аннотаций: %1</translation>
    </message>
    <message>
        <location filename="../src/SwissProtPlainTextFormat.cpp" line="289"/>
        <source>Invalid format of feature table</source>
        <translation type="unfinished">Таблица аннотаций повреждена.</translation>
    </message>
</context>
</TS>
