<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en_US">
<context>
    <name>GObject</name>
    <message>
        <source>Unknown</source>
        <translation type="vanished">Unknown</translation>
    </message>
    <message>
        <source>Unloaded</source>
        <translation type="vanished">Unloaded</translation>
    </message>
    <message>
        <source>Text</source>
        <translation type="vanished">Text</translation>
    </message>
    <message>
        <source>Sequence</source>
        <translation type="vanished">Sequence</translation>
    </message>
    <message>
        <source>Sequences</source>
        <translation type="vanished">Sequences</translation>
    </message>
    <message>
        <source>Annotation</source>
        <translation type="vanished">Annotation</translation>
    </message>
    <message>
        <source>Annotations</source>
        <translation type="vanished">Annotations</translation>
    </message>
    <message>
        <source>Variation</source>
        <translation type="vanished">Variation</translation>
    </message>
    <message>
        <source>Variations</source>
        <translation type="vanished">Variations</translation>
    </message>
    <message>
        <source>Chromatogram</source>
        <translation type="vanished">Chromatogram</translation>
    </message>
    <message>
        <source>Chromatograms</source>
        <translation type="vanished">Chromatograms</translation>
    </message>
    <message>
        <source>Alignment</source>
        <translation type="vanished">Alignment</translation>
    </message>
    <message>
        <source>Alignments</source>
        <translation type="vanished">Alignments</translation>
    </message>
    <message>
        <source>Tree</source>
        <translation type="vanished">Tree</translation>
    </message>
    <message>
        <source>Trees</source>
        <translation type="vanished">Trees</translation>
    </message>
    <message>
        <source>3D model</source>
        <translation type="vanished">3D model</translation>
    </message>
    <message>
        <source>3D models</source>
        <translation type="vanished">3D models</translation>
    </message>
    <message>
        <source>Assembly</source>
        <translation type="vanished">Assembly</translation>
    </message>
    <message>
        <source>Assemblies</source>
        <translation type="vanished">Assemblies</translation>
    </message>
</context>
<context>
    <name>L10N</name>
    <message>
        <source>Failed to detect IO adapter for %1</source>
        <translation type="vanished">Failed to detect IO adapter for %1</translation>
    </message>
</context>
<context>
    <name>MAlignmentObject</name>
    <message>
        <source>Unexpected gap start</source>
        <translation type="vanished">Unexpected gap start</translation>
    </message>
    <message>
        <source>Unexpected gap start (too small)</source>
        <translation type="vanished">Unexpected gap start (too small)</translation>
    </message>
    <message>
        <source>Unexpected gap start (too big)</source>
        <translation type="vanished">Unexpected gap start (too big)</translation>
    </message>
    <message>
        <source> Incorrect gap splitting</source>
        <translation type="vanished"> Incorrect gap splitting</translation>
    </message>
    <message>
        <source> Alignment walker is ended</source>
        <translation type="vanished"> Alignment walker is ended</translation>
    </message>
    <message>
        <source> Alignment changed</source>
        <translation type="vanished"> Alignment changed</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="525"/>
        <source>This is not ESearch result!</source>
        <translation>This is not ESearch result!</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="570"/>
        <source>This is not a ESummary result!</source>
        <translation>This is not a ESummary result!</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="653"/>
        <source>Use Ensembl ID. For example: %1 or %2</source>
        <translation>Use Ensembl ID. For example: %1 or %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="654"/>
        <source>Use Genbank DNA accession number. For example: %1 or %2</source>
        <translation>Use Genbank DNA accession number. For example: %1 or %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="655"/>
        <source>Use Genbank protein accession number. For example: %1</source>
        <translation>Use Genbank protein accession number. For example: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="656"/>
        <source>Use PDB molecule four-letter identifier. For example: %1 or %2</source>
        <translation>Use PDB molecule four-letter identifier. For example: %1 or %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="657"/>
        <source>Use SWISS-PROT accession number. For example: %1 or %2</source>
        <translation>Use SWISS-PROT accession number. For example: %1 or %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="658"/>
        <source>Use UniProtKB/Swiss-Prot accession number. For example: %1</source>
        <translation>Use UniProtKB/Swiss-Prot accession number. For example: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="659"/>
        <source>Use UniProtKB/TrEMBL accession number. For example: %1</source>
        <translation>Use UniProtKB/TrEMBL accession number. For example: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="688"/>
        <source>Use %1 unique identifier.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/util/FormatUtils.cpp" line="139"/>
        <location filename="../src/util/FormatUtils.cpp" line="178"/>
        <location filename="../src/util/FormatUtils.cpp" line="197"/>
        <source>All files</source>
        <translation>All files</translation>
    </message>
    <message>
        <location filename="../src/util/SequenceUtils.cpp" line="279"/>
        <source>Loaded sequences have different alphabets. That&apos;s why several sequence objects are created, each for specified alphabet. All sequences at one object have the same alphabet and sequences from different objects have different alphabets.</source>
        <translation>Loaded sequences have different alphabets. That&apos;s why several sequence objects are created, each for specified alphabet. All sequences at one object have the same alphabet and sequences from different objects have different alphabets.</translation>
    </message>
    <message>
        <location filename="../src/datatype/Annotation.cpp" line="593"/>
        <location filename="../src/datatype/Annotation.cpp" line="596"/>
        <source>Sequence</source>
        <translation>Sequence</translation>
    </message>
    <message>
        <location filename="../src/datatype/Annotation.cpp" line="604"/>
        <location filename="../src/datatype/Annotation.cpp" line="607"/>
        <source>Translation</source>
        <translation>Translation</translation>
    </message>
    <message>
        <location filename="../src/datatype/PhyTree.cpp" line="73"/>
        <source>UGENE internal error</source>
        <translation>UGENE internal error</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="509"/>
        <source>Invalid database type: %1</source>
        <translation>Invalid database type: %1</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectUtils.cpp" line="427"/>
        <source>Unsupported object type: %1</source>
        <translation>Unsupported object type: %1</translation>
    </message>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="405"/>
        <source>Unable to detect sequence alphabet. Probably, this is because some of merged sequences are empty.</source>
        <translation>Unable to detect sequence alphabet. Probably, this is because some of merged sequences are empty.</translation>
    </message>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="560"/>
        <source>Invalid number of symbols: %1</source>
        <translation>Invalid number of symbols: %1</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="52"/>
        <source>The document is created not by UGENE</source>
        <translation>The document is created not by UGENE</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="53"/>
        <source>Document sequences were merged</source>
        <translation>Document sequences were merged</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2DbiUtils.cpp" line="43"/>
        <source>UGENE public database</source>
        <translation>UGENE public database</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="76"/>
        <source>Gap between two components of a genome or transcriptome assembly</source>
        <translation>Gap between two components of a genome or transcriptome assembly</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="77"/>
        <source>Sequence related to transcription termination</source>
        <translation>Sequence related to transcription termination</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="78"/>
        <source>Describes disulfide bonds (for protein files)</source>
        <translation>Describes disulfide bonds (for protein files)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="79"/>
        <source>Span of the C immunological feature</source>
        <translation>Span of the C immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="80"/>
        <source>`CAAT box&apos; in eukaryotic promoters</source>
        <translation>`CAAT box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="81"/>
        <source>Sequence coding for amino acids in protein (includes stop codon)</source>
        <translation>Sequence coding for amino acids in protein (includes stop codon)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="83"/>
        <source>Independent sequence determinations differ</source>
        <translation>Independent sequence determinations differ</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="84"/>
        <source>Region of biological interest identified as a centromere and which has been experimentally characterized</source>
        <translation>Region of biological interest identified as a centromere and which has been experimentally characterized</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="85"/>
        <source>Displacement loop</source>
        <translation>Displacement loop</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="86"/>
        <source>Span of the D immunological feature</source>
        <translation>Span of the D immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="87"/>
        <source>Cis-acting enhancer of promoter function</source>
        <translation>Cis-acting enhancer of promoter function</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="88"/>
        <source>Region that codes for part of spliced mRNA</source>
        <translation>Region that codes for part of spliced mRNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="89"/>
        <source>Gap in the sequence</source>
        <translation>Gap in the sequence</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="90"/>
        <source>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</source>
        <translation>Region that defines a functional gene, possibly including upstream (promotor, enhancer, etc) and downstream control elements, and for which a name has been assigned.</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="92"/>
        <source>`GC box&apos; in eukaryotic promoters</source>
        <translation>`GC box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="93"/>
        <source>Intervening DNA eliminated by recombination</source>
        <translation>Intervening DNA eliminated by recombination</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="94"/>
        <source>Transcribed region excised by mRNA splicing</source>
        <translation>Transcribed region excised by mRNA splicing</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="95"/>
        <source>Span of the J immunological feature</source>
        <translation>Span of the J immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="96"/>
        <source>Joining segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains</source>
        <translation>Joining segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="97"/>
        <source>Long terminal repeat</source>
        <translation>Long terminal repeat</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="98"/>
        <source>Mature peptide coding region (does not include stop codon)</source>
        <translation>Mature peptide coding region (does not include stop codon)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="99"/>
        <source>Miscellaneous binding site</source>
        <translation>Miscellaneous binding site</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="100"/>
        <source>Miscellaneous difference feature</source>
        <translation>Miscellaneous difference feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="101"/>
        <source>Region of biological significance that cannot be described by any other feature</source>
        <translation>Region of biological significance that cannot be described by any other feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="103"/>
        <source>Miscellaneous, recombination feature</source>
        <translation>Miscellaneous, recombination feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="104"/>
        <source>Miscellaneous transcript feature not defined by other RNA keys</source>
        <translation>Miscellaneous transcript feature not defined by other RNA keys</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="105"/>
        <source>Miscellaneous signal</source>
        <translation>Miscellaneous signal</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="106"/>
        <source>Miscellaneous DNA or RNA structure</source>
        <translation>Miscellaneous DNA or RNA structure</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="107"/>
        <source>Region of genome containing mobile elements</source>
        <translation>Region of genome containing mobile elements</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="108"/>
        <source>The indicated base is a modified nucleotide</source>
        <translation>The indicated base is a modified nucleotide</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="109"/>
        <source>Messenger RNA</source>
        <translation>Messenger RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="110"/>
        <source>A non-protein-coding gene, other than ribosomal RNA and transfer RNA, the functional molecule of which is the RNA transcrip</source>
        <translation>A non-protein-coding gene, other than ribosomal RNA and transfer RNA, the functional molecule of which is the RNA transcrip</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="111"/>
        <source>Span of the N immunological feature</source>
        <translation>Span of the N immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="112"/>
        <source>Presented sequence revises a previous version</source>
        <translation>Presented sequence revises a previous version</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="113"/>
        <source>Region containing polycistronic transcript including a cluster of genes that are under the control of the same regulatory sequences/promotor and in the same biological pathway</source>
        <translation>Region containing polycistronic transcript including a cluster of genes that are under the control of the same regulatory sequences/promotor and in the same biological pathway</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="114"/>
        <source>Origin of transfer; region of a DNA molecule where transfer is initiated during the process of conjugation or mobilization</source>
        <translation>Origin of transfer; region of a DNA molecule where transfer is initiated during the process of conjugation or mobilization</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="115"/>
        <source>Signal for cleavage &amp; polyadenylation</source>
        <translation>Signal for cleavage &amp; polyadenylation</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="116"/>
        <source>Site at which polyadenine is added to mRNA</source>
        <translation>Site at which polyadenine is added to mRNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="117"/>
        <source>Any RNA species that is not yet the mature RNA product</source>
        <translation>Any RNA species that is not yet the mature RNA product</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="118"/>
        <source>Primary (unprocessed) transcript</source>
        <translation>Primary (unprocessed) transcript</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="119"/>
        <source>Primer binding region used with PCR</source>
        <translation>Primer binding region used with PCR</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="120"/>
        <source>Non-covalent primer binding site</source>
        <translation>Non-covalent primer binding site</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="121"/>
        <source>A region involved in transcription initiation</source>
        <translation>A region involved in transcription initiation</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="122"/>
        <source>Non-covalent protein binding site on DNA or RNA</source>
        <translation>Non-covalent protein binding site on DNA or RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="123"/>
        <source>Ribosome binding site</source>
        <translation>Ribosome binding site</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="124"/>
        <source>Replication origin for duplex DNA</source>
        <translation>Replication origin for duplex DNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="125"/>
        <source>Sequence containing repeated subsequences</source>
        <translation>Sequence containing repeated subsequences</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="126"/>
        <source>One repeated unit of a repeat_region</source>
        <translation>One repeated unit of a repeat_region</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="127"/>
        <source>Ribosomal RNA</source>
        <translation>Ribosomal RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="128"/>
        <source>Span of the S immunological feature</source>
        <translation>Span of the S immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="129"/>
        <source>Satellite repeated sequence</source>
        <translation>Satellite repeated sequence</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="130"/>
        <source>Small cytoplasmic RNA</source>
        <translation>Small cytoplasmic RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="131"/>
        <source>Signal peptide coding region</source>
        <translation>Signal peptide coding region</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="132"/>
        <source>Small nuclear RNA</source>
        <translation>Small nuclear RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="133"/>
        <source>Identifies the biological source of the specified span of the sequence</source>
        <translation>Identifies the biological source of the specified span of the sequence</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="134"/>
        <source>Hair-pin loop structure in DNA or RNA</source>
        <translation>Hair-pin loop structure in DNA or RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="135"/>
        <source>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</source>
        <translation>Sequence Tagged Site; operationally unique sequence that identifies the combination of primer spans used in a PCR assay</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="136"/>
        <source>`TATA box&apos; in eukaryotic promoters</source>
        <translation>`TATA box&apos; in eukaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="137"/>
        <source>Region of biological interest identified as a telomere and which has been experimentally characterized</source>
        <translation>Region of biological interest identified as a telomere and which has been experimentally characterized</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="138"/>
        <source>Sequence causing transcription termination</source>
        <translation>Sequence causing transcription termination</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="139"/>
        <source>Transfer messenger RNA; tmRNA acts as a tRNA first, and then as an mRNA that encodes a peptide tag; the ribosome translates this mRNA region of tmRNA and attaches the encoded peptide tag to the C-terminus of the unfinished protein; this attached tag targets the protein for destruction or proteolysis</source>
        <translation>Transfer messenger RNA; tmRNA acts as a tRNA first, and then as an mRNA that encodes a peptide tag; the ribosome translates this mRNA region of tmRNA and attaches the encoded peptide tag to the C-terminus of the unfinished protein; this attached tag targets the protein for destruction or proteolysis</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="140"/>
        <source>Transit peptide coding region</source>
        <translation>Transit peptide coding region</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="141"/>
        <source>Transposable element (TN)</source>
        <translation>Transposable element (TN)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="142"/>
        <source>Transfer RNA</source>
        <translation>Transfer RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="143"/>
        <source>Authors are unsure about the sequence in this region</source>
        <translation>Authors are unsure about the sequence in this region</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="144"/>
        <source>Span of the V immunological feature</source>
        <translation>Span of the V immunological feature</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="145"/>
        <source>Variable segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains; codes for most of the variable region (V_region) and the last few amino acids of the leader peptide</source>
        <translation>Variable segment of immunoglobulin light and heavy chains, and T-cell receptor alpha, beta, and gamma chains; codes for most of the variable region (V_region) and the last few amino acids of the leader peptide</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="146"/>
        <source>A related population contains stable mutation</source>
        <translation>A related population contains stable mutation</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="147"/>
        <source>`Pribnow box&apos; in prokaryotic promoters</source>
        <translation>`Pribnow box&apos; in prokaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="148"/>
        <source>`-35 box&apos; in prokaryotic promoters</source>
        <translation>`-35 box&apos; in prokaryotic promoters</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="149"/>
        <source>3&apos;-most region of a precursor transcript removed in processing</source>
        <translation>3&apos;-most region of a precursor transcript removed in processing</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="150"/>
        <source>3&apos; untranslated region (trailer)</source>
        <translation>3&apos; untranslated region (trailer)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="151"/>
        <source>5&apos;-most region of a precursor transcript removed in processing</source>
        <translation>5&apos;-most region of a precursor transcript removed in processing</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="152"/>
        <source>5&apos; untranslated region (leader)</source>
        <translation>5&apos; untranslated region (leader)</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="153"/>
        <source>&apos;Protein&apos; feature key</source>
        <translation>&apos;Protein&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="154"/>
        <source>&apos;Region&apos; feature key</source>
        <translation>&apos;Region&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="155"/>
        <source>&apos;Site&apos; feature key</source>
        <translation>&apos;Site&apos; feature key</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="156"/>
        <source>Any region of sequence that functions in the regulation of transcription or translation</source>
        <translation>Any region of sequence that functions in the regulation of transcription or translation</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="175"/>
        <source>Genes</source>
        <translation>Genes</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="190"/>
        <source>Signals</source>
        <translation>Signals</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="211"/>
        <source>Binding</source>
        <translation>Binding</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="217"/>
        <source>Variation</source>
        <translation>Variation</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="226"/>
        <source>Repeats</source>
        <translation>Repeats</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="233"/>
        <source>RNA</source>
        <translation>RNA</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="243"/>
        <source>Misc</source>
        <translation>Misc</translation>
    </message>
    <message>
        <location filename="../src/util/GenbankFeatures.cpp" line="263"/>
        <source>Spans</source>
        <translation>Spans</translation>
    </message>
    <message>
        <location filename="../src/util/U2FeatureUtils.cpp" line="133"/>
        <source>Unexpected feature operator value detected.</source>
        <translation>Unexpected feature operator value detected.</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentImporter.cpp" line="45"/>
        <source>Unexpected number of sequences in a multiple alignment</source>
        <translation>Unexpected number of sequences in a multiple alignment</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentImporter.cpp" line="82"/>
        <source>Unexpected error on MSA rows import</source>
        <translation>Unexpected error on MSA rows import</translation>
    </message>
</context>
<context>
    <name>U2::AbstractProjectFilterTask</name>
    <message>
        <location filename="../src/tasks/AbstractProjectFilterTask.cpp" line="37"/>
        <source>Filtering project content by the &quot;%1&quot; criterion</source>
        <translation>Filtering project content by the &quot;%1&quot; criterion</translation>
    </message>
</context>
<context>
    <name>U2::AddDocumentTask</name>
    <message>
        <location filename="../src/tasks/AddDocumentTask.cpp" line="38"/>
        <location filename="../src/tasks/AddDocumentTask.cpp" line="50"/>
        <source>Adding document to project: %1</source>
        <translation>Adding document to project: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/AddDocumentTask.cpp" line="81"/>
        <source>No project is opened</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/tasks/AddDocumentTask.cpp" line="91"/>
        <source>Document is already added to the project %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/tasks/AddDocumentTask.cpp" line="96"/>
        <source>Document was removed</source>
        <translation>Document was removed</translation>
    </message>
</context>
<context>
    <name>U2::AddObjectsToDocumentTask</name>
    <message>
        <location filename="../src/tasks/AddObjectsToDocumentTask.cpp" line="35"/>
        <source>Add objects to document</source>
        <translation>Add objects to document</translation>
    </message>
</context>
<context>
    <name>U2::AddSequenceObjectsToAlignmentTask</name>
    <message>
        <location filename="../src/tasks/AddSequencesToAlignmentTask.cpp" line="45"/>
        <source>Object is empty.</source>
        <translation>Object is empty.</translation>
    </message>
    <message>
        <location filename="../src/tasks/AddSequencesToAlignmentTask.cpp" line="50"/>
        <source>Object is locked for modifications.</source>
        <translation>Object is locked for modifications.</translation>
    </message>
    <message>
        <location filename="../src/tasks/AddSequencesToAlignmentTask.cpp" line="136"/>
        <source>Some sequences have wrong alphabet: </source>
        <translation>Some sequences have wrong alphabet: </translation>
    </message>
    <message>
        <location filename="../src/tasks/AddSequencesToAlignmentTask.cpp" line="139"/>
        <source> and others</source>
        <translation> and others</translation>
    </message>
</context>
<context>
    <name>U2::AddSequencesToAlignmentTask</name>
    <message>
        <source>Object is empty.</source>
        <translation type="vanished">Object is empty.</translation>
    </message>
    <message>
        <source>Object is locked for modifications.</source>
        <translation type="vanished">Object is locked for modifications.</translation>
    </message>
    <message>
        <source>Some sequences have wrong alphabet: </source>
        <translation type="vanished">Some sequences have wrong alphabet: </translation>
    </message>
    <message>
        <source> and others</source>
        <translation type="vanished"> and others</translation>
    </message>
</context>
<context>
    <name>U2::AppResourcePool</name>
    <message>
        <location filename="../src/globals/AppResources.cpp" line="56"/>
        <source>Threads</source>
        <translation>Threads</translation>
    </message>
    <message>
        <location filename="../src/globals/AppResources.cpp" line="67"/>
        <source>Memory</source>
        <translation>Memory</translation>
    </message>
    <message>
        <location filename="../src/globals/AppResources.cpp" line="67"/>
        <source>Mb</source>
        <translation>Mb</translation>
    </message>
    <message>
        <location filename="../src/globals/AppResources.cpp" line="70"/>
        <source>Project</source>
        <translation>Project</translation>
    </message>
</context>
<context>
    <name>U2::AutoAnnotationsSupport</name>
    <message>
        <location filename="../src/globals/AutoAnnotationsSupport.cpp" line="106"/>
        <source>Auto-annotations [%1 | %2]</source>
        <translation>Auto-annotations [%1 | %2]</translation>
    </message>
</context>
<context>
    <name>U2::AutoAnnotationsUpdateTask</name>
    <message>
        <location filename="../src/globals/AutoAnnotationsSupport.cpp" line="284"/>
        <source>Empty auto-annotation object</source>
        <translation>Empty auto-annotation object</translation>
    </message>
</context>
<context>
    <name>U2::BaseLoadRemoteDocumentTask</name>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="69"/>
        <source>Load remote document</source>
        <translation>Load remote document</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="145"/>
        <source>Unknown file format!</source>
        <translation>Unknown file format!</translation>
    </message>
</context>
<context>
    <name>U2::CMDLineCoreOptions</name>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="59"/>
        <source>Shows help information.</source>
        <translation>Shows help information.</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="66"/>
        <source>Loads UGENE configuration.</source>
        <translation>Loads UGENE configuration.</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="67"/>
        <source>Loads configuration from the specified .ini file. By default the UGENE.ini file is used.</source>
        <translation>Loads configuration from the specified .ini file. By default the UGENE.ini file is used.</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="68"/>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="80"/>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="89"/>
        <source>&lt;path_to_file&gt;</source>
        <translation>&lt;path_to_file&gt;</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="72"/>
        <source>Specifies the language to use.</source>
        <translation>Specifies the language to use.</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="73"/>
        <source>Specifies the language to use. The following values are available: CS, EN, RU, ZH.</source>
        <translation>Specifies the language to use. The following values are available: CS, EN, RU, ZH.</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="74"/>
        <source>&lt;language_code&gt;</source>
        <translation>&lt;language_code&gt;</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="84"/>
        <source>Path to the session database file</source>
        <translation>Path to the session database file</translation>
    </message>
    <message>
        <location filename="../src/cmdline/CMDLineCoreOptions.cpp" line="85"/>
        <source>Session database is stored in the temporary file that is created for every UGENE run.
But it can be supplied with the command line argument.
If the supplied file does not exist it will be created.
The session database file is removed after closing of UGENE.</source>
        <translation>Session database is stored in the temporary file that is created for every UGENE run.
But it can be supplied with the command line argument.
If the supplied file does not exist it will be created.
The session database file is removed after closing of UGENE.</translation>
    </message>
</context>
<context>
    <name>U2::CleanupFileStorageTask</name>
    <message>
        <location filename="../src/tasks/CleanupFileStorageTask.cpp" line="31"/>
        <source>Cleanup file storage</source>
        <translation>Cleanup file storage</translation>
    </message>
</context>
<context>
    <name>U2::CloneObjectTask</name>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="34"/>
        <source>Copy object</source>
        <translation>Copy object</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="41"/>
        <source>Invalid source object</source>
        <translation>Invalid source object</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="42"/>
        <source>Invalid destination document</source>
        <translation>Invalid destination document</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="48"/>
        <source>Copy object %1</source>
        <translation>Copy object %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="54"/>
        <source>Invalid destination database reference</source>
        <translation>Invalid destination database reference</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="65"/>
        <source>Error! No DBI</source>
        <translation>Error! No DBI</translation>
    </message>
    <message>
        <location filename="../src/tasks/CloneObjectTask.cpp" line="70"/>
        <source>The object has been removed</source>
        <translation>The object has been removed</translation>
    </message>
</context>
<context>
    <name>U2::ComplementSequenceTask</name>
    <message>
        <location filename="../src/tasks/ReverseSequenceTask.cpp" line="142"/>
        <source>Complement Sequence Task</source>
        <translation>Complement Sequence Task</translation>
    </message>
</context>
<context>
    <name>U2::ConnectSharedDatabaseTask</name>
    <message>
        <location filename="../src/tasks/ConnectSharedDatabaseTask.cpp" line="35"/>
        <source>Connecting to database: </source>
        <translation>Connecting to database:</translation>
    </message>
</context>
<context>
    <name>U2::ConsoleShutdownTask</name>
    <message>
        <location filename="../src/tasks/ConsoleShutdownTask.cpp" line="33"/>
        <source>Shutdown</source>
        <translation>Shutdown</translation>
    </message>
    <message>
        <location filename="../src/tasks/ConsoleShutdownTask.cpp" line="83"/>
        <source>Cancel active tasks</source>
        <translation>Cancel active tasks</translation>
    </message>
    <message>
        <location filename="../src/tasks/ConsoleShutdownTask.cpp" line="106"/>
        <source>Starting shutdown process...</source>
        <translation>Starting shutdown process...</translation>
    </message>
    <message>
        <location filename="../src/tasks/ConsoleShutdownTask.cpp" line="128"/>
        <source>Shutdown was canceled</source>
        <translation>Shutdown was canceled</translation>
    </message>
    <message>
        <location filename="../src/tasks/ConsoleShutdownTask.cpp" line="132"/>
        <source>Shutdown failed, error: %1</source>
        <translation>Shutdown failed, error: %1</translation>
    </message>
</context>
<context>
    <name>U2::ConvertDasIdTask</name>
    <message>
        <location filename="../src/tasks/PicrApiTask.cpp" line="279"/>
        <source>Convert resource id: %1</source>
        <translation>Convert resource id: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/PicrApiTask.cpp" line="361"/>
        <source>undefined error (code %1)</source>
        <translation>undefined error (code %1)</translation>
    </message>
    <message>
        <location filename="../src/tasks/PicrApiTask.cpp" line="383"/>
        <source>Remote server does not respond</source>
        <translation>Remote server does not respond</translation>
    </message>
</context>
<context>
    <name>U2::ConvertIdAndLoadDasFeaturesTask</name>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="783"/>
        <source>Convert ID and load DAS features for: %1</source>
        <translation>Convert ID and load DAS features for: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="799"/>
        <source>Convert IDs task</source>
        <translation>Convert IDs task</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="819"/>
        <source>&quot;%1&quot; was converted into &quot;%2&quot;</source>
        <translation>&quot;%1&quot; was converted into &quot;%2&quot;</translation>
    </message>
</context>
<context>
    <name>U2::CopyDataTask</name>
    <message>
        <location filename="../src/tasks/CopyDataTask.cpp" line="33"/>
        <source>Copy Data Task</source>
        <translation>Copy Data Task</translation>
    </message>
    <message>
        <location filename="../src/tasks/CopyDataTask.cpp" line="55"/>
        <source>Cannot get data from: &apos;%1&apos;</source>
        <translation>Cannot get data from: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/CopyDataTask.cpp" line="74"/>
        <source>IO adapter error. %1</source>
        <translation>IO adapter error. %1</translation>
    </message>
</context>
<context>
    <name>U2::CreateAnnotationsTask</name>
    <message>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="37"/>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="47"/>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="59"/>
        <source>Create annotations</source>
        <translation>Create annotations</translation>
    </message>
    <message>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="68"/>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="74"/>
        <source>Annotation table has been removed unexpectedly</source>
        <translation>Annotation table has been removed unexpectedly</translation>
    </message>
    <message>
        <location filename="../src/tasks/CreateAnnotationTask.cpp" line="110"/>
        <source>Annotation object &apos;%1&apos; not found in active project: %2</source>
        <translation>Annotation object &apos;%1&apos; not found in active project: %2</translation>
    </message>
</context>
<context>
    <name>U2::DASSourceRegistry</name>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="113"/>
        <source>UniProt (DAS)</source>
        <translation>UniProt (DAS)</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="114"/>
        <location filename="../src/datatype/DASSource.cpp" line="123"/>
        <source>http://www.ebi.ac.uk/das-srv/uniprot/das/uniprot/</source>
        <translation>http://www.ebi.ac.uk/das-srv/uniprot/das/uniprot/</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="116"/>
        <source>Use Swiss-Prot, TrEMBL, UniProt, IPI or UniParc accession number. For example: %1</source>
        <translation>Use Swiss-Prot, TrEMBL, UniProt, IPI or UniParc accession number. For example: %1</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="117"/>
        <source>UniProt database of protein sequences and features.</source>
        <translation>UniProt database of protein sequences and features.</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="122"/>
        <source>UniProt</source>
        <translation>UniProt</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="125"/>
        <source>DAS 1 reference server for sequence and features from UniProt.
UniProt (Universal Protein Resource) is the worlds most comprehensive catalog of information on proteins.
It is a central repository of protein sequence and function created by joining the information contained
in Swiss-Prot, TrEMBL, and PIR.</source>
        <translation>DAS 1 reference server for sequence and features from UniProt.
UniProt (Universal Protein Resource) is the worlds most comprehensive catalog of information on proteins.
It is a central repository of protein sequence and function created by joining the information contained
in Swiss-Prot, TrEMBL, and PIR.</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="134"/>
        <source>Pride DAS 1.6</source>
        <translation>Pride DAS 1.6</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="135"/>
        <source>http://www.ebi.ac.uk/pride-das/das/PrideDataSource/</source>
        <translation>http://www.ebi.ac.uk/pride-das/das/PrideDataSource/</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="137"/>
        <source>The PRIDE database is a centralized, public data repository for proteomics data. 
It has been developed to provide the proteomics community with a public repository 
for protein and peptide identifications together with the evidence supporting these 
identifications. In addition to identifications, PRIDE is able to capture details of 
post-translational modifications coordinated relative to the peptides in which they have been found.</source>
        <translation>The PRIDE database is a centralized, public data repository for proteomics data. 
It has been developed to provide the proteomics community with a public repository 
for protein and peptide identifications together with the evidence supporting these 
identifications. In addition to identifications, PRIDE is able to capture details of 
post-translational modifications coordinated relative to the peptides in which they have been found.</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="147"/>
        <source>cbs_sort</source>
        <translation>cbs_sort</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="148"/>
        <source>http://das.cbs.dtu.dk:9000/das/cbs_sort/</source>
        <translation>http://das.cbs.dtu.dk:9000/das/cbs_sort/</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="150"/>
        <source>The combined result of the signalp, secretomep, targetp, netnes &amp; lipop server for the given id.</source>
        <translation>The combined result of the signalp, secretomep, targetp, netnes &amp; lipop server for the given id.</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="156"/>
        <source>signalp</source>
        <translation>signalp</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="157"/>
        <source>http://das.cbs.dtu.dk:9000/das/signalp/</source>
        <translation>http://das.cbs.dtu.dk:9000/das/signalp/</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="159"/>
        <source>SignalP (http://www.cbs.dtu.dk/services/SignalP) predictions for UniProt.</source>
        <translation>SignalP (http://www.cbs.dtu.dk/services/SignalP) predictions for UniProt.</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="165"/>
        <source>InterPro-Matches-Overview</source>
        <translation>InterPro-Matches-Overview</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="166"/>
        <source>http://www.ebi.ac.uk/das-srv/interpro/das/InterPro-matches-overview/</source>
        <translation>http://www.ebi.ac.uk/das-srv/interpro/das/InterPro-matches-overview/</translation>
    </message>
    <message>
        <location filename="../src/datatype/DASSource.cpp" line="168"/>
        <source>Shows the maximum extent of the matches from all signatures that are 
integrated into a single InterPro entry against UniProtKB protein sequences 
(i.e. if an InterPro entry contains 2 signatures, A and B, and both of 
these match a protein, A between residue 3 and 125 and B between residue 
5 and 127, the overview supermatch of these signatures is between 3 and 127).</source>
        <translation>Shows the maximum extent of the matches from all signatures that are 
integrated into a single InterPro entry against UniProtKB protein sequences 
(i.e. if an InterPro entry contains 2 signatures, A and B, and both of 
these match a protein, A between residue 3 and 125 and B between residue 
5 and 127, the overview supermatch of these signatures is between 3 and 127).</translation>
    </message>
</context>
<context>
    <name>U2::DBXRefRegistry</name>
    <message>
        <location filename="../src/globals/DBXRefRegistry.cpp" line="42"/>
        <source>File with db_xref mappings not found: %1</source>
        <translation>File with db_xref mappings not found: %1</translation>
    </message>
    <message>
        <location filename="../src/globals/DBXRefRegistry.cpp" line="53"/>
        <source>Illegal db_xref file entry: %1</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::DNAAlphabetRegistryImpl</name>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="63"/>
        <source>All symbols</source>
        <translation>All symbols</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="71"/>
        <source>Standard DNA</source>
        <translation>Standard DNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="80"/>
        <source>Standard RNA</source>
        <translation>Standard RNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="90"/>
        <source>Extended DNA</source>
        <translation>Extended DNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="99"/>
        <source>Extended RNA</source>
        <translation>Extended RNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="114"/>
        <source>Standard amino</source>
        <translation>Standard amino</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseAlphabets.cpp" line="133"/>
        <source>Extended amino</source>
        <translation>Extended amino</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="270"/>
        <source>Complement for standard DNA</source>
        <translation>Complement for standard DNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="288"/>
        <source>Complement for standard RNA</source>
        <translation>Complement for standard RNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="316"/>
        <source>Complement for extended DNA</source>
        <translation>Complement for extended DNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="345"/>
        <source>Complement for extended RNA</source>
        <translation>Complement for extended RNA</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="359"/>
        <source>1. The Standard Genetic Code</source>
        <translation>1. The Standard Genetic Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="369"/>
        <source>2. The Vertebrate Mitochondrial Code</source>
        <translation>2. The Vertebrate Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="379"/>
        <source>3. The Yeast Mitochondrial Code</source>
        <translation>3. The Yeast Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="389"/>
        <source>4. The Mold, Protozoan, and Coelenterate Mitochondria and the Mycoplasma Code</source>
        <translation>4. The Mold, Protozoan, and Coelenterate Mitochondria and the Mycoplasma Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="399"/>
        <source>5. The Invertebrate Mitochondrial Code</source>
        <translation>5. The Invertebrate Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="409"/>
        <source>6. The Ciliate, Dasycladacean and Hexamita Nuclear Code</source>
        <translation>6. The Ciliate, Dasycladacean and Hexamita Nuclear Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="419"/>
        <source>9. The Echinoderm and Flatworm Mitochondrial Code</source>
        <translation>9. The Echinoderm and Flatworm Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="429"/>
        <source>10. The Euplotid Nuclear Code</source>
        <translation>10. The Euplotid Nuclear Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="439"/>
        <source>11. The Bacterial and Plant Plastid Code</source>
        <translation>11. The Bacterial and Plant Plastid Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="449"/>
        <source>12. The Alternative Yeast Nuclear Code</source>
        <translation>12. The Alternative Yeast Nuclear Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="459"/>
        <source>13. The Ascidian Mitochondrial Code</source>
        <translation>13. The Ascidian Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="469"/>
        <source>14. The Alternative Flatworm Mitochondrial Code</source>
        <translation>14. The Alternative Flatworm Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="479"/>
        <source>15. Blepharisma Nuclear Code</source>
        <translation>15. Blepharisma Nuclear Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="490"/>
        <source>16. Chlorophycean Mitochondrial Code</source>
        <translation>16. Chlorophycean Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="500"/>
        <source>21. Trematode Mitochondrial Code</source>
        <translation>21. Trematode Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="510"/>
        <source>22. Scenedesmus obliquus Mitochondrial Code</source>
        <translation>22. Scenedesmus obliquus Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="521"/>
        <source>23. Thraustochytrium Mitochondrial Code</source>
        <translation>23. Thraustochytrium Mitochondrial Code</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="572"/>
        <source>Alanine</source>
        <translation>Alanine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="573"/>
        <source>Cysteine</source>
        <translation>Cysteine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="574"/>
        <source>Aspartic acid</source>
        <translation>Aspartic acid</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="575"/>
        <source>Glutamic acid</source>
        <translation>Glutamic acid</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="576"/>
        <source>Phenylalanine</source>
        <translation>Phenylalanine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="577"/>
        <source>Glycine</source>
        <translation>Glycine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="578"/>
        <source>Histidine</source>
        <translation>Histidine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="579"/>
        <source>Isoleucine</source>
        <translation>Isoleucine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="580"/>
        <source>Lysine</source>
        <translation>Lysine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="581"/>
        <source>Leucine</source>
        <translation>Leucine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="582"/>
        <source>Methionine</source>
        <translation>Methionine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="583"/>
        <source>Asparagine</source>
        <translation>Asparagine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="584"/>
        <source>Proline</source>
        <translation>Proline</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="585"/>
        <source>Glutamine</source>
        <translation>Glutamine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="586"/>
        <source>Arginine</source>
        <translation>Arginine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="587"/>
        <source>Serine</source>
        <translation>Serine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="588"/>
        <source>Threonine</source>
        <translation>Threonine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="589"/>
        <source>Valine</source>
        <translation>Valine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="590"/>
        <source>Tryptophan</source>
        <translation>Tryptophan</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="591"/>
        <source>Tyrosine</source>
        <translation>Tyrosine</translation>
    </message>
    <message>
        <location filename="../src/datatype/BaseTranslations.cpp" line="592"/>
        <source>Stop codon</source>
        <translation>Stop codon</translation>
    </message>
</context>
<context>
    <name>U2::DatabaseConnectionAdapterFactory</name>
    <message>
        <location filename="../src/io/DatabaseConnectionAdapter.cpp" line="33"/>
        <source>Database connection</source>
        <translation>Database connection</translation>
    </message>
</context>
<context>
    <name>U2::DatatypeSerializers</name>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="310"/>
        <source>Incorrect tree parsing state</source>
        <translation>Incorrect tree parsing state</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="313"/>
        <source>Unexpected weight: %1</source>
        <translation>Unexpected weight: %1</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="317"/>
        <source>Error parsing weight: %1</source>
        <translation>Error parsing weight: %1</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="324"/>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="345"/>
        <source>Tree node stack is empty</source>
        <translation>Tree node stack is empty</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="336"/>
        <source>Error parsing nodeValue: %1</source>
        <translation>Error parsing nodeValue: %1</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="346"/>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="364"/>
        <source>Branch node stack is empty</source>
        <translation>Branch node stack is empty</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="348"/>
        <source>Unexpected new sibling %1</source>
        <translation>Unexpected new sibling %1</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="361"/>
        <source>Unexpected closing bracket :%1</source>
        <translation>Unexpected closing bracket :%1</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="369"/>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="391"/>
        <source>Unexpected end of file</source>
        <translation>Unexpected end of file</translation>
    </message>
    <message>
        <location filename="../src/util/DatatypeSerializeUtils.cpp" line="403"/>
        <source>Empty file</source>
        <translation>Empty file</translation>
    </message>
</context>
<context>
    <name>U2::DbiDocumentFormat</name>
    <message>
        <location filename="../src/dbi/DbiDocumentFormat.cpp" line="46"/>
        <source>ugenedb is a internal UGENE database file format</source>
        <translation>ugenedb is a internal UGENE database file format</translation>
    </message>
    <message>
        <location filename="../src/dbi/DbiDocumentFormat.cpp" line="129"/>
        <source>Invalid destination database reference</source>
        <translation>Invalid destination database reference</translation>
    </message>
</context>
<context>
    <name>U2::DeleteFoldersTask</name>
    <message>
        <location filename="../src/tasks/DeleteObjectsTask.cpp" line="97"/>
        <source>Delete folders</source>
        <translation>Delete folders</translation>
    </message>
</context>
<context>
    <name>U2::DeleteObjectsTask</name>
    <message>
        <location filename="../src/tasks/DeleteObjectsTask.cpp" line="64"/>
        <source>Delete objects</source>
        <translation>Delete objects</translation>
    </message>
</context>
<context>
    <name>U2::Document</name>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="542"/>
        <source>Document is not loaded</source>
        <translation>Document is not loaded</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="555"/>
        <source>IO adapter does not support write operation</source>
        <translation>IO adapter does not support write operation</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="561"/>
        <source>No write support for document format</source>
        <translation>No write support for document format</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="631"/>
        <source>Locked by user</source>
        <translation>Locked by user</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="655"/>
        <source>Lock is NULL</source>
        <translation>Lock is NULL</translation>
    </message>
</context>
<context>
    <name>U2::DocumentFormat</name>
    <message>
        <location filename="../src/models/DocumentModel.cpp" line="123"/>
        <location filename="../src/models/DocumentModel.cpp" line="128"/>
        <source>Writing is not supported for this format (%1). Feel free to send a feature request though.</source>
        <translation>Writing is not supported for this format (%1). Feel free to send a feature request though.</translation>
    </message>
</context>
<context>
    <name>U2::DocumentProviderTask</name>
    <message>
        <location filename="../src/tasks/DocumentProviderTask.cpp" line="33"/>
        <source>[unknown]</source>
        <translation>[unknown]</translation>
    </message>
</context>
<context>
    <name>U2::DocumentUtils</name>
    <message>
        <location filename="../src/models/DocumentUtils.cpp" line="260"/>
        <source>Format does not support writing of alignments</source>
        <translation>Format does not support writing of alignments</translation>
    </message>
    <message>
        <location filename="../src/models/DocumentUtils.cpp" line="271"/>
        <source>File content was merged</source>
        <translation>File content was merged</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolRunTask</name>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="59"/>
        <source> tool</source>
        <translation> tool</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="114"/>
        <source>Can not run %1 tool.</source>
        <translation>Can not run %1 tool.</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="116"/>
        <source>Can not run %1 tool. May be tool path &apos;%2&apos; not valid?</source>
        <translation>Can not run %1 tool. May be tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="125"/>
        <source>Tool %1 is cancelled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="133"/>
        <source>%1 tool exited with code %2</source>
        <translation>%1 tool exited with code %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="135"/>
        <source>Tool %1 finished successfully</source>
        <translation>Tool %1 finished successfully</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportUtils</name>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="329"/>
        <source>Can not remove temporary directory: path is empty.</source>
        <translation>Can not remove temporary directory: path is empty.</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="335"/>
        <source>Can not remove files from temporary directory.</source>
        <translation>Can not remove files from temporary directory.</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="340"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="353"/>
        <source>Can not create directory for temporary files: %1</source>
        <translation>Can not create directory for temporary files: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="384"/>
        <source>Can&apos;t run an executable file &quot;%1&quot; as it is. Try to run it as a cmd line command: &quot;%2&quot;</source>
        <translation>Can&apos;t run an executable file &quot;%1&quot; as it is. Try to run it as a cmd line command: &quot;%2&quot;</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="400"/>
        <source>Undefined tool: &apos;%1&apos;</source>
        <translation>Undefined tool: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="404"/>
        <source>Path for &apos;%1&apos; tool not set</source>
        <translation>Path for &apos;%1&apos; tool not set</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="448"/>
        <source>Working directory is &quot;%1&quot;</source>
        <translation>Working directory is &quot;%1&quot;</translation>
    </message>
    <message>
        <location filename="../src/tasks/ExternalToolRunTask.cpp" line="457"/>
        <source>Launching %1 tool: %2 %3</source>
        <translation>Launching %1 tool: %2 %3</translation>
    </message>
</context>
<context>
    <name>U2::ExtractAnnotatedRegionTask</name>
    <message>
        <location filename="../src/tasks/ExtractAnnotatedRegionTask.cpp" line="34"/>
        <source>Extract annotated regions</source>
        <translation>Extract annotated regions</translation>
    </message>
</context>
<context>
    <name>U2::FailTask</name>
    <message>
        <location filename="../src/tasks/FailTask.h" line="32"/>
        <source>Failure</source>
        <translation>Failure</translation>
    </message>
</context>
<context>
    <name>U2::GObject</name>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="42"/>
        <source>Unknown</source>
        <translation>Unknown</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="43"/>
        <source>Unloaded</source>
        <translation>Unloaded</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="44"/>
        <source>Text</source>
        <translation>Text</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="45"/>
        <source>Sequence</source>
        <translation>Sequence</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="45"/>
        <source>Sequences</source>
        <translation>Sequences</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="46"/>
        <source>Annotation</source>
        <translation>Annotation</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="46"/>
        <source>Annotations</source>
        <translation>Annotations</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="47"/>
        <source>Variation</source>
        <translation>Variation</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="47"/>
        <source>Variations</source>
        <translation>Variations</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="48"/>
        <source>Chromatogram</source>
        <translation>Chromatogram</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="48"/>
        <source>Chromatograms</source>
        <translation>Chromatograms</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="49"/>
        <source>Alignment</source>
        <translation>Alignment</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="49"/>
        <source>Alignments</source>
        <translation>Alignments</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="50"/>
        <source>Tree</source>
        <translation>Tree</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="50"/>
        <source>Trees</source>
        <translation>Trees</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="51"/>
        <source>3D model</source>
        <translation>3D model</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="51"/>
        <source>3D models</source>
        <translation>3D models</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="52"/>
        <source>Assembly</source>
        <translation>Assembly</translation>
    </message>
    <message>
        <location filename="../src/gobjects/GObjectTypes.cpp" line="52"/>
        <source>Assemblies</source>
        <translation>Assemblies</translation>
    </message>
</context>
<context>
    <name>U2::GUrlUtils</name>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="142"/>
        <source>Renamed %1 to %2</source>
        <translation>Renamed %1 to %2</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="146"/>
        <source>Failed to rename %1 to %2</source>
        <translation>Failed to rename %1 to %2</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="290"/>
        <source>Directory is not specified</source>
        <translation>Directory is not specified</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="295"/>
        <source>Directory can&apos;t be created: %1</source>
        <translation>Directory can&apos;t be created: %1</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="300"/>
        <source>Directory can&apos;t be read: %1</source>
        <translation>Directory can&apos;t be read: %1</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="342"/>
        <source>File path is not specified</source>
        <translation>File path is not specified</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="345"/>
        <source>Directory path instead of file path</source>
        <translation>Directory path instead of file path</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="403"/>
        <source>Can not create a directory: %1</source>
        <translation>Can not create a directory: %1</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="453"/>
        <source>%1 is not a local file [%2].</source>
        <translation>%1 is not a local file [%2].</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="462"/>
        <source>Can not create a directory [%1].</source>
        <translation>Can not create a directory [%1].</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.cpp" line="467"/>
        <source>%1 is a directory [%2].</source>
        <translation>%1 is a directory [%2].</translation>
    </message>
    <message>
        <location filename="../src/util/GUrlUtils.h" line="133"/>
        <source>Output URL</source>
        <translation>Output URL</translation>
    </message>
</context>
<context>
    <name>U2::GetDasIdsBySequenceTask</name>
    <message>
        <location filename="../src/tasks/PicrApiTask.cpp" line="423"/>
        <source>Get an ID for the sequence</source>
        <translation>Get an ID for the sequence</translation>
    </message>
    <message>
        <location filename="../src/tasks/PicrApiTask.cpp" line="494"/>
        <source>undefined error (code %1)</source>
        <translation>undefined error (code %1)</translation>
    </message>
</context>
<context>
    <name>U2::GzippedHttpFileAdapterFactory</name>
    <message>
        <location filename="../src/io/HttpFileAdapter.cpp" line="52"/>
        <source>HTTP GZIP adaptor</source>
        <translation>HTTP GZIP adaptor</translation>
    </message>
</context>
<context>
    <name>U2::GzippedLocalFileAdapterFactory</name>
    <message>
        <location filename="../src/io/LocalFileAdapter.cpp" line="38"/>
        <source>GZIP file</source>
        <translation>GZIP file</translation>
    </message>
</context>
<context>
    <name>U2::HttpFileAdapter</name>
    <message>
        <location filename="../src/io/HttpFileAdapter.cpp" line="110"/>
        <source>Incorrect url string has been passed to HttpFileAdapter::open()</source>
        <translation>Incorrect url string has been passed to HttpFileAdapter::open()</translation>
    </message>
</context>
<context>
    <name>U2::HttpFileAdapterFactory</name>
    <message>
        <location filename="../src/io/HttpFileAdapter.cpp" line="43"/>
        <source>HTTP file</source>
        <translation>HTTP file</translation>
    </message>
</context>
<context>
    <name>U2::ImportDirToDatabaseTask</name>
    <message>
        <location filename="../src/tasks/shared_db/ImportDirToDatabaseTask.cpp" line="36"/>
        <source>Import directory %1 to the database</source>
        <translation>Import directory %1 to the database</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDirToDatabaseTask.cpp" line="43"/>
        <source>It is not a directory: </source>
        <translation>It is not a directory:</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDirToDatabaseTask.cpp" line="44"/>
        <source>Invalid database reference</source>
        <translation>Invalid database reference</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDirToDatabaseTask.cpp" line="92"/>
        <source>Import was cancelled</source>
        <translation>Import was cancelled</translation>
    </message>
</context>
<context>
    <name>U2::ImportDocumentToDatabaseTask</name>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="35"/>
        <source>Import document %1 to the database</source>
        <translation>Import document %1 to the database</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="42"/>
        <source>Invalid document to import</source>
        <translation>Invalid document to import</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="43"/>
        <source>Invalid database reference</source>
        <translation>Invalid database reference</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="74"/>
        <source>Source document was removed, can&apos;t set object relations</source>
        <translation>Source document was removed, can&apos;t set object relations</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="163"/>
        <source>Can&apos;t set object relation: target object is not found in the source document (%1)</source>
        <translation>Can&apos;t set object relation: target object is not found in the source document (%1)</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportDocumentToDatabaseTask.cpp" line="167"/>
        <source>Can&apos;t set object relation: target object is not imported (%1)</source>
        <translation>Can&apos;t set object relation: target object is not imported (%1)</translation>
    </message>
</context>
<context>
    <name>U2::ImportFileToDatabaseTask</name>
    <message>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="44"/>
        <source>Import file %1 to the database</source>
        <translation>Import file %1 to the database</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="52"/>
        <source>It is not a file: </source>
        <translation>It is not a file:</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="53"/>
        <source>Invalid database reference</source>
        <translation>Invalid database reference</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="58"/>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="101"/>
        <source>File format is not recognized</source>
        <translation>File format is not recognized</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportFileToDatabaseTask.cpp" line="74"/>
        <source>Unrecognized url: </source>
        <translation>Unrecognized url:</translation>
    </message>
</context>
<context>
    <name>U2::ImportObjectToDatabaseTask</name>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="35"/>
        <source>Import object %1 to database</source>
        <translation>Import object %1 to database</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="42"/>
        <source>Invalid object to import</source>
        <translation>Invalid object to import</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="43"/>
        <source>Invalid database reference</source>
        <translation>Invalid database reference</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="50"/>
        <source>Error! No DBI</source>
        <translation>Error! No DBI</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="52"/>
        <source>Error! No object DBI</source>
        <translation>Error! No object DBI</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportObjectToDatabaseTask.cpp" line="57"/>
        <source>The object has been removed</source>
        <translation>The object has been removed</translation>
    </message>
</context>
<context>
    <name>U2::ImportToDatabaseTask</name>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="35"/>
        <source>Import to the database</source>
        <translation>Import to the database</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="52"/>
        <source>Import complete: %1 seconds</source>
        <translation>Import complete: %1 seconds</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="79"/>
        <source>The import task was cancelled.</source>
        <translation>The import task was cancelled.</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="81"/>
        <source>The import task has failed.</source>
        <translation>The import task has failed.</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="83"/>
        <source>The import task has finished.</source>
        <translation>The import task has finished.</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="124"/>
        <source>Successfully imported files:&lt;br&gt;&lt;br&gt;</source>
        <translation>Successfully imported files:&lt;br&gt;&lt;br&gt;</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="161"/>
        <source>Successfully imported objects:&lt;br&gt;&lt;br&gt;</source>
        <translation>Successfully imported objects:&lt;br&gt;&lt;br&gt;</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="178"/>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="269"/>
        <source>Document </source>
        <translation>Document </translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="205"/>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="296"/>
        <source>Just objects without a document:</source>
        <translation>Just objects without a document:</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="215"/>
        <source>Not imported files:&lt;br&gt;&lt;br&gt;</source>
        <translation>Not imported files:&lt;br&gt;&lt;br&gt;</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="236"/>
        <source>Import was cancelled</source>
        <translation>Import was cancelled</translation>
    </message>
    <message>
        <location filename="../src/tasks/shared_db/ImportToDatabaseTask.cpp" line="252"/>
        <source>Not imported objects:&lt;br&gt;&lt;br&gt;</source>
        <translation>Not imported objects:&lt;br&gt;&lt;br&gt;</translation>
    </message>
</context>
<context>
    <name>U2::L10N</name>
    <message>
        <location filename="../src/globals/L10n.h" line="37"/>
        <source>Error!</source>
        <translation>Error!</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="38"/>
        <source>Warning!</source>
        <translation>Warning!</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="39"/>
        <source>Internal error, bad argument: %1</source>
        <translation>Internal error, bad argument: %1</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="40"/>
        <source>Out of memory!</source>
        <translation>Out of memory!</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="41"/>
        <source>Internal error!</source>
        <translation>Internal error!</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="43"/>
        <source>Internal error! The object unexpectedly does not exist: </source>
        <translation>Internal error! The object unexpectedly does not exist: </translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="46"/>
        <source>Error opening file for reading: &apos;%1&apos;</source>
        <translation>Error opening file for reading: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="47"/>
        <source>Error opening file for writing: &apos;%1&apos;</source>
        <translation>Error opening file for writing: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="51"/>
        <source>File not found: &apos;%1&apos;</source>
        <translation>File not found: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="52"/>
        <source>&apos;%1&apos; is not a file</source>
        <translation>&apos;%1&apos; is not a file</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="53"/>
        <source>Directory not found: &apos;%1&apos;</source>
        <translation>Directory not found: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="54"/>
        <source>Error opening directory: &apos;%1&apos;</source>
        <translation>Error opening directory: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="55"/>
        <source>&apos;%1&apos; is not a directory</source>
        <translation>&apos;%1&apos; is not a directory</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="58"/>
        <source>&apos;%1&apos; does not match database object URL format</source>
        <translation>&apos;%1&apos; does not match database object URL format</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="59"/>
        <source>&apos;%1&apos; does not match database folder URL format</source>
        <translation>&apos;%1&apos; does not match database folder URL format</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="96"/>
        <source>Direct</source>
        <translation>Direct</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="97"/>
        <source>Reverse-complementary</source>
        <translation>Reverse-complementary</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="99"/>
        <source>Required</source>
        <translation>Required</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="48"/>
        <source>Write error: &apos;%1&apos;</source>
        <translation>Write error: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="49"/>
        <source>Read error: %1</source>
        <translation>Read error: %1</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="50"/>
        <source>File is too large: &apos;%1&apos;</source>
        <translation>File is too large: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="60"/>
        <source>Unable to connect to the database &apos;%1&apos;. Check your connection settings in File -&gt; Connect to shared database</source>
        <translation>Unable to connect to the database &apos;%1&apos;. Check your connection settings in File -&gt; Connect to shared database</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="62"/>
        <source>Unable to find the object &apos;%1&apos; in the database &apos;%2&apos;</source>
        <translation>Unable to find the object &apos;%1&apos; in the database &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="65"/>
        <source>Unable to find the folder &apos;%1&apos; with the specified path in the database &apos;%2&apos;</source>
        <translation>Unable to find the folder &apos;%1&apos; with the specified path in the database &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="67"/>
        <source>You do not have write permissions to the database &apos;%1&apos;</source>
        <translation>You do not have write permissions to the database &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="69"/>
        <source>Invalid file format. Expected file format: %1,  File: %2</source>
        <translation>Invalid file format. Expected file format: %1,  File: %2</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="70"/>
        <source>Invalid file format! File: %1</source>
        <translation>Invalid file format! File: %1</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="74"/>
        <source>Can&apos;t save image to file %1, format %2</source>
        <translation>Can&apos;t save image to file %1, format %2</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="76"/>
        <source>Document not found: %1</source>
        <translation>Document not found: %1</translation>
    </message>
    <message>
        <location filename="../src/globals/L10n.h" line="77"/>
        <source> bp</source>
        <translation> bp</translation>
    </message>
    <message>
        <location filename="../src/util/IOAdapterUtils.cpp" line="97"/>
        <location filename="../src/util/IOAdapterUtils.cpp" line="105"/>
        <source>Failed to detect IO adapter for %1</source>
        <translation>Failed to detect IO adapter for %1</translation>
    </message>
</context>
<context>
    <name>U2::LoadDasDocumentTask</name>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="110"/>
        <source>Cannot load cached document: %1</source>
        <translation>Cannot load cached document: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="116"/>
        <source>Cannot save document: %1</source>
        <translation>Cannot save document: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="128"/>
        <source>Cannot find DAS reference sequence: %1</source>
        <translation>Cannot find DAS reference sequence: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="147"/>
        <source>Cannot find DAS features for &apos;%1&apos; on %2</source>
        <translation>Cannot find DAS features for &apos;%1&apos; on %2</translation>
    </message>
</context>
<context>
    <name>U2::LoadDasFeaturesTask</name>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="391"/>
        <source>Load DAS annotations for current sequence</source>
        <translation>Load DAS annotations for current sequence</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="416"/>
        <source>Can not receive response from the server &quot;</source>
        <translation>Can not receive response from the server</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="424"/>
        <source>&lt;font size=&quot;5&quot; color=&quot;green&quot;&gt;Received %1 annotations from the server &quot;%2&quot;&lt;/font&gt;&lt;br&gt;</source>
        <translation>&lt;font size=&quot;5&quot; color=&quot;green&quot;&gt;Received %1 annotations from the server &quot;%2&quot;&lt;/font&gt;&lt;br&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LoadDasObjectTask</name>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="242"/>
        <source>Load DAS data for &apos;%1&apos; from %2</source>
        <translation>Load DAS data for &apos;%1&apos; from %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="301"/>
        <source>Sequence with ID=%1 is not found.</source>
        <translation>Sequence with ID=%1 is not found.</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="353"/>
        <source>undefined error (code %1)</source>
        <translation>undefined error (code %1)</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDASDocumentTask.cpp" line="375"/>
        <source>Remote server does not respond</source>
        <translation>Remote server does not respond</translation>
    </message>
</context>
<context>
    <name>U2::LoadDataFromEntrezTask</name>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="424"/>
        <source>Sequence with ID=%1 is not found.</source>
        <translation>Sequence with ID=%1 is not found.</translation>
    </message>
</context>
<context>
    <name>U2::LoadDocumentTask</name>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="227"/>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="242"/>
        <source>Read document: &apos;%1&apos;</source>
        <translation>Read document: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="263"/>
        <source>Document format is NULL!</source>
        <translation>Document format is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="264"/>
        <source>IO adapter factory is NULL!</source>
        <translation>IO adapter factory is NULL!</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="444"/>
        <source>Document not found %1</source>
        <translation>Document not found %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="464"/>
        <source>Not enough memory to load document %1</source>
        <translation>Not enough memory to load document %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="480"/>
        <source>Maximum number of objects per document limit reached for %1. Try different options for opening the document!</source>
        <translation>Maximum number of objects per document limit reached for %1. Try different options for opening the document!</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="510"/>
        <source>Object not found: %1</source>
        <translation>Object not found: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="517"/>
        <source>Can&apos;t add object. Document format constraints check failed: %1</source>
        <translation>Can&apos;t add object. Document format constraints check failed: %1</translation>
    </message>
</context>
<context>
    <name>U2::LoadRemoteDocumentTask</name>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="212"/>
        <source>Undefined database: &apos;%1&apos;</source>
        <translation>Undefined database: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadRemoteDocumentTask.cpp" line="264"/>
        <source>Cannot find %1 in %2 database</source>
        <translation>Cannot find %1 in %2 database</translation>
    </message>
</context>
<context>
    <name>U2::LoadUnloadedDocumentTask</name>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="89"/>
        <source>Load &apos;%1&apos;</source>
        <translation>Load &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="97"/>
        <source>Document not found</source>
        <translation>Document not found</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="104"/>
        <source>Starting load document from %1, document format %2</source>
        <translation>Starting load document from %1, document format %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="130"/>
        <source>Document was removed</source>
        <translation>Document was removed</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="163"/>
        <source>Document is locked</source>
        <translation>Document is locked</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="179"/>
        <source>Project Document:</source>
        <translation>Project Document:</translation>
    </message>
    <message>
        <location filename="../src/tasks/LoadDocumentTask.cpp" line="197"/>
        <source>Annotation object not found</source>
        <translation>Annotation object not found</translation>
    </message>
</context>
<context>
    <name>U2::LocalFileAdapterFactory</name>
    <message>
        <location filename="../src/io/LocalFileAdapter.cpp" line="29"/>
        <source>Local file</source>
        <translation>Local file</translation>
    </message>
</context>
<context>
    <name>U2::MAlignmentObject</name>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="77"/>
        <source>Unexpected gap start</source>
        <translation>Unexpected gap start</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="86"/>
        <source>Unexpected gap start (too small)</source>
        <translation>Unexpected gap start (too small)</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="87"/>
        <source>Unexpected gap start (too big)</source>
        <translation>Unexpected gap start (too big)</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="101"/>
        <source> Incorrect gap splitting</source>
        <translation> Incorrect gap splitting</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="136"/>
        <source> Alignment walker is ended</source>
        <translation> Alignment walker is ended</translation>
    </message>
    <message>
        <location filename="../src/util/MAlignmentWalker.cpp" line="137"/>
        <source> Alignment changed</source>
        <translation> Alignment changed</translation>
    </message>
</context>
<context>
    <name>U2::MSAUtils</name>
    <message>
        <location filename="../src/util/MSAUtils.cpp" line="181"/>
        <source>Alphabets of the alignment and the sequence cannot be derived</source>
        <translation>Alphabets of the alignment and the sequence cannot be derived</translation>
    </message>
    <message>
        <location filename="../src/util/MSAUtils.cpp" line="189"/>
        <source>Sequences have different alphabets.</source>
        <translation>Sequences have different alphabets.</translation>
    </message>
    <message>
        <location filename="../src/util/MSAUtils.cpp" line="217"/>
        <location filename="../src/util/MSAUtils.cpp" line="222"/>
        <source>Alignment is empty!</source>
        <translation>Alignment is empty!</translation>
    </message>
    <message>
        <location filename="../src/util/MSAUtils.cpp" line="229"/>
        <source>Sequences in alignment have different sizes!</source>
        <translation>Sequences in alignment have different sizes!</translation>
    </message>
    <message>
        <location filename="../src/util/MSAUtils.cpp" line="338"/>
        <source>Can&apos;t find a row in an alignment!</source>
        <translation>Can&apos;t find a row in an alignment!</translation>
    </message>
</context>
<context>
    <name>U2::ModifySequenceContentTask</name>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="54"/>
        <source>Modify sequence task</source>
        <translation>Modify sequence task</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="64"/>
        <source>Document is locked</source>
        <translation>Document is locked</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="68"/>
        <source>Region to delete is larger than the whole sequence</source>
        <translation>Region to delete is larger than the whole sequence</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="125"/>
        <source>Some annotations have qualifiers referring a sequence region that has been removed during the sequence editing. You might want to change the qualifiers manually. Find them in the table below</source>
        <translation>Some annotations have qualifiers referring a sequence region that has been removed during the sequence editing. You might want to change the qualifiers manually. Find them in the table below</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="129"/>
        <source>Annotation Name</source>
        <translation>Annotation Name</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="131"/>
        <source>Annotation Location</source>
        <translation>Annotation Location</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="133"/>
        <source>Qualifier Name</source>
        <translation>Qualifier Name</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="135"/>
        <source>Referenced Region</source>
        <translation>Referenced Region</translation>
    </message>
    <message>
        <location filename="../src/tasks/ModifySequenceObjectTask.cpp" line="140"/>
        <source>Unexpected qualifiers count</source>
        <translation>Unexpected qualifiers count</translation>
    </message>
</context>
<context>
    <name>U2::MsaDbiUtils</name>
    <message>
        <location filename="../src/util/MsaDbiUtils.cpp" line="865"/>
        <source>Can&apos;t rename an alignment to an empty name!</source>
        <translation>Can&apos;t rename an alignment to an empty name!</translation>
    </message>
    <message>
        <location filename="../src/util/MsaDbiUtils.cpp" line="894"/>
        <location filename="../src/util/MsaDbiUtils.cpp" line="901"/>
        <source>Failed to insert gaps into an alignment!</source>
        <translation>Failed to insert gaps into an alignment!</translation>
    </message>
    <message>
        <location filename="../src/util/MsaDbiUtils.cpp" line="1070"/>
        <source>Failed to crop an alignment!</source>
        <translation>Failed to crop an alignment!</translation>
    </message>
    <message>
        <source>Failed to add a row to an alignment!</source>
        <translation type="obsolete">Failed to add a row to an alignment!</translation>
    </message>
    <message>
        <location filename="../src/util/MsaDbiUtils.cpp" line="1195"/>
        <source>Can&apos;t rename a row to an empty name!</source>
        <translation>Can&apos;t rename a row to an empty name!</translation>
    </message>
</context>
<context>
    <name>U2::PasteTask</name>
    <message>
        <location filename="../src/globals/ClipboardController.cpp" line="36"/>
        <source>Paste data</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RelocateDocumentTask</name>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="348"/>
        <source>Relocate document %1 -&gt; %2</source>
        <translation>Relocate document %1 -&gt; %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="355"/>
        <source>No active project found</source>
        <translation>No active project found</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="359"/>
        <source>Project is locked</source>
        <translation>Project is locked</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="368"/>
        <source>Only unloaded objects can be relocated</source>
        <translation>Only unloaded objects can be relocated</translation>
    </message>
</context>
<context>
    <name>U2::RemoveAnnotationsTask</name>
    <message>
        <location filename="../src/tasks/RemoveAnnotationsTask.cpp" line="46"/>
        <source>Waiting for object lock released</source>
        <translation>Waiting for object lock released</translation>
    </message>
</context>
<context>
    <name>U2::RemoveMultipleDocumentsTask</name>
    <message>
        <location filename="../src/tasks/RemoveDocumentTask.cpp" line="34"/>
        <source>Remove document</source>
        <translation>Remove document</translation>
    </message>
    <message>
        <location filename="../src/tasks/RemoveDocumentTask.cpp" line="94"/>
        <source>Cannot remove document %1, since it is locked by some task.</source>
        <translation>Cannot remove document %1, since it is locked by some task.</translation>
    </message>
</context>
<context>
    <name>U2::ReplyHandler</name>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="90"/>
        <source>Can&apos;t receive result from the server: nothing to download</source>
        <translation>Can&apos;t receive result from the server: nothing to download</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="139"/>
        <source>Unexpected server response</source>
        <translation>Unexpected server response</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="169"/>
        <source>undefined error (code %1)</source>
        <translation>undefined error (code %1)</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="171"/>
        <source>Network error: %1</source>
        <translation>Network error: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="189"/>
        <source>Remote server doesn&apos;t respond</source>
        <translation>Remote server doesn&apos;t respond</translation>
    </message>
</context>
<context>
    <name>U2::ResourceTracker</name>
    <message>
        <location filename="../src/globals/ResourceTracker.cpp" line="39"/>
        <source>resource &apos;%1&apos; is used by &apos;%2&apos;</source>
        <translation>resource &apos;%1&apos; is used by &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/globals/ResourceTracker.cpp" line="55"/>
        <source>resource &apos;%1&apos; is released by &apos;%2&apos;</source>
        <translation>resource &apos;%1&apos; is released by &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::ReverseComplementSequenceTask</name>
    <message>
        <location filename="../src/tasks/ReverseSequenceTask.cpp" line="40"/>
        <source>Reverse Complement Sequence Task</source>
        <translation>Reverse Complement Sequence Task</translation>
    </message>
</context>
<context>
    <name>U2::ReverseSequenceTask</name>
    <message>
        <location filename="../src/tasks/ReverseSequenceTask.cpp" line="52"/>
        <source>Reverse Sequence Task</source>
        <translation>Reverse Sequence Task</translation>
    </message>
</context>
<context>
    <name>U2::SaveCopyAndAddToProjectTask</name>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="308"/>
        <source>Save a copy %1</source>
        <translation>Save a copy %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="326"/>
        <source>No active project found</source>
        <translation>No active project found</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="327"/>
        <source>Project is locked</source>
        <translation>Project is locked</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="331"/>
        <source>Document is already added to the project %1</source>
        <translation>Document is already added to the project %1</translation>
    </message>
</context>
<context>
    <name>U2::SaveDocumentStreamingTask</name>
    <message>
        <location filename="../src/tasks/SaveDocumentStreamingTask.cpp" line="32"/>
        <source>Save document</source>
        <translation>Save document</translation>
    </message>
</context>
<context>
    <name>U2::SaveDocumentTask</name>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="55"/>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="72"/>
        <source>Save document</source>
        <translation>Save document</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="65"/>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="78"/>
        <source>No permission to write to &apos;%1&apos; file.</source>
        <translation>No permission to write to &apos;%1&apos; file.</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="100"/>
        <source>Saving document %1
</source>
        <translation>Saving document %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="121"/>
        <source>Can&apos;t create tmp file</source>
        <translation>Can&apos;t create tmp file</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="126"/>
        <source>Can&apos;t copy file to tmp file while trying to save document by append</source>
        <translation>Can&apos;t copy file to tmp file while trying to save document by append</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="143"/>
        <source>Can&apos;t remove original file to place tmp file instead</source>
        <translation>Can&apos;t remove original file to place tmp file instead</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="146"/>
        <source>Can&apos;t rename saved tmp file to original file</source>
        <translation>Can&apos;t rename saved tmp file to original file</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="183"/>
        <source>Document &apos;%1&apos; can&apos;t be unloaded: </source>
        <translation>Document &apos;%1&apos; can&apos;t be unloaded: </translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="183"/>
        <source>unexpected error</source>
        <translation>unexpected error</translation>
    </message>
</context>
<context>
    <name>U2::SaveMultipleDocuments</name>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="201"/>
        <source>Save multiple documents</source>
        <translation>Save multiple documents</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="213"/>
        <source>Question?</source>
        <translation>Question?</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="213"/>
        <source>Save document: %1</source>
        <translation>Save document: %1</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="264"/>
        <source>You have no permission to write to &apos;%1&apos; file.
UGENE contains unsaved modifications.</source>
        <translation>You have no permission to write to &apos;%1&apos; file.
UGENE contains unsaved modifications.</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="265"/>
        <source>Do you want to save changes to another file?</source>
        <translation>Do you want to save changes to another file?</translation>
    </message>
    <message>
        <location filename="../src/tasks/SaveDocumentTask.cpp" line="286"/>
        <source>Save as</source>
        <translation>Save as</translation>
    </message>
</context>
<context>
    <name>U2::ScriptTask</name>
    <message>
        <location filename="../src/tasks/ScriptTask.cpp" line="64"/>
        <source>Script syntax check failed! Line: %1, error: %2</source>
        <translation>Script syntax check failed! Line: %1, error: %2</translation>
    </message>
    <message>
        <location filename="../src/tasks/ScriptTask.cpp" line="72"/>
        <source>Exception during script execution! Line: %1, error: %2</source>
        <translation>Exception during script execution! Line: %1, error: %2</translation>
    </message>
</context>
<context>
    <name>U2::SequenceDbiWalkerSubtask</name>
    <message>
        <location filename="../src/tasks/SequenceDbiWalkerTask.cpp" line="135"/>
        <source>Sequence walker subtask</source>
        <translation>Sequence walker subtask</translation>
    </message>
</context>
<context>
    <name>U2::SequenceWalkerSubtask</name>
    <message>
        <location filename="../src/tasks/SequenceWalkerTask.cpp" line="160"/>
        <source>Sequence walker subtask</source>
        <translation>Sequence walker subtask</translation>
    </message>
</context>
<context>
    <name>U2::StringAdapterFactory</name>
    <message>
        <location filename="../src/io/StringAdapter.cpp" line="27"/>
        <source>String buffer</source>
        <translation>String buffer</translation>
    </message>
</context>
<context>
    <name>U2::TripleStoreL10N</name>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="91"/>
        <source>Database is already opened!</source>
        <translation>Database is already opened!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="95"/>
        <source>Illegal database state: %1</source>
        <translation>Illegal database state: %1</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="100"/>
        <source>URL is not specified</source>
        <translation>URL is not specified</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="111"/>
        <source>Error opening SQLite database: %1!</source>
        <translation>Error opening SQLite database: %1!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="157"/>
        <source>Error checking SQLite database: %1!</source>
        <translation>Error checking SQLite database: %1!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="172"/>
        <source>Database is already closed!</source>
        <translation>Database is already closed!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="176"/>
        <source>Illegal database state %1!</source>
        <translation>Illegal database state %1!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SQLiteTripleStore.cpp" line="186"/>
        <source>Failed to close triple store database: %1</source>
        <translation>Failed to close triple store database: %1</translation>
    </message>
</context>
<context>
    <name>U2::U2AssemblyUtils</name>
    <message>
        <location filename="../src/util/U2AssemblyUtils.cpp" line="55"/>
        <source>Invalid CIGAR op: &apos;%1&apos;!</source>
        <translation>Invalid CIGAR op: &apos;%1&apos;!</translation>
    </message>
</context>
<context>
    <name>U2::U2BitCompression</name>
    <message>
        <location filename="../src/util/U2Bits.cpp" line="89"/>
        <source>Bit compression: illegal character in text &apos;%1&apos;</source>
        <translation>Bit compression: illegal character in text &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::U2DbiL10n</name>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="63"/>
        <source>Error querying database: %1</source>
        <translation>Error querying database: %1</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="67"/>
        <source>Found more results than expected!</source>
        <translation>Found more results than expected!</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="175"/>
        <source>Unexpected query result code: %1 (%2)</source>
        <translation>Unexpected query result code: %1 (%2)</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="193"/>
        <source>Error binding NULL value! Query: &apos;%1&apos;, idx: %2</source>
        <translation>Error binding NULL value! Query: &apos;%1&apos;, idx: %2</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="305"/>
        <source>Error binding text value! Query: &apos;%1&apos;, idx: %2, value: &apos;%3&apos;</source>
        <translation>Error binding text value! Query: &apos;%1&apos;, idx: %2, value: &apos;%3&apos;</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="317"/>
        <source>Error binding int32 value! Query: &apos;%1&apos;, idx: %2, value: %3</source>
        <translation>Error binding int32 value! Query: &apos;%1&apos;, idx: %2, value: %3</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="329"/>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="341"/>
        <source>Error binding int64 value! Query: &apos;%1&apos;, idx: %2, value: %3</source>
        <translation>Error binding int64 value! Query: &apos;%1&apos;, idx: %2, value: %3</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="354"/>
        <source>Error binding boolean value! Query: &apos;%1&apos;, idx: %2, value: %3</source>
        <translation>Error binding boolean value! Query: &apos;%1&apos;, idx: %2, value: %3</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="371"/>
        <source>Error binding blob value! Query: &apos;%1&apos;, idx: %2, size: %3</source>
        <translation>Error binding blob value! Query: &apos;%1&apos;, idx: %2, size: %3</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="383"/>
        <source>Error binding blob value! Query: &apos;%1&apos;, idx: %2</source>
        <translation>Error binding blob value! Query: &apos;%1&apos;, idx: %2</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2SqlHelpers.cpp" line="396"/>
        <source>Unexpected row count! Query: &apos;%1&apos;, rows: %2</source>
        <translation>Unexpected row count! Query: &apos;%1&apos;, rows: %2</translation>
    </message>
</context>
<context>
    <name>U2::U2DbiPool</name>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="291"/>
        <source>Invalid database type: %1</source>
        <translation>Invalid database type: %1</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="317"/>
        <source>Invalid database id</source>
        <translation>Invalid database id</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="352"/>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="368"/>
        <source>DbiPool: DBI not found! Dbi ID: %1</source>
        <translation>DbiPool: DBI not found! Dbi ID: %1</translation>
    </message>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="517"/>
        <source>Invalid dbi reference</source>
        <translation>Invalid dbi reference</translation>
    </message>
</context>
<context>
    <name>U2::U2DbiRegistry</name>
    <message>
        <location filename="../src/dbi/U2DbiRegistry.cpp" line="134"/>
        <source>The tmp dbi is not found: %1</source>
        <translation>The tmp dbi is not found: %1</translation>
    </message>
</context>
<context>
    <name>U2::U2DbiUtils</name>
    <message>
        <location filename="../src/dbi/U2DbiUtils.cpp" line="49"/>
        <source>Feature is not supported: %1, dbi: %2</source>
        <translation>Feature is not supported: %1, dbi: %2</translation>
    </message>
</context>
<context>
    <name>U2::U2SequenceObject</name>
    <message>
        <location filename="../src/gobjects/DNASequenceObject.cpp" line="200"/>
        <source>Internal error, sequence alphabet id &apos;%1&apos; is invalid</source>
        <translation>Internal error, sequence alphabet id &apos;%1&apos; is invalid</translation>
    </message>
    <message>
        <location filename="../src/gobjects/DNASequenceObject.cpp" line="209"/>
        <source>Modified sequence &amp; region have different alphabet</source>
        <translation>Modified sequence &amp; region have different alphabet</translation>
    </message>
</context>
<context>
    <name>U2::U2SequenceUtils</name>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="55"/>
        <source>Alphabet is not found!</source>
        <translation>Alphabet is not found!</translation>
    </message>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="98"/>
        <location filename="../src/util/U2SequenceUtils.cpp" line="111"/>
        <source>Invalid sequence DBI</source>
        <translation>Invalid sequence DBI</translation>
    </message>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="370"/>
        <location filename="../src/util/U2SequenceUtils.cpp" line="549"/>
        <source>Failed to derive sequence alphabet!</source>
        <translation>Failed to derive sequence alphabet!</translation>
    </message>
</context>
<context>
    <name>U2::UniprotBlastAndLoadDASAnnotations</name>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="602"/>
        <source>BLAST IDs and DAS annotations</source>
        <translation>BLAST IDs and DAS annotations</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="657"/>
        <source>Server &quot;%1&quot; is not available!</source>
        <translation>Server &quot;%1&quot; is not available!</translation>
    </message>
</context>
<context>
    <name>U2::UniprotBlastTask</name>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="504"/>
        <source>Get an ID for the sequence</source>
        <translation>Get an ID for the sequence</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="544"/>
        <source>Reply from the BLAST server is empty. Similar sequences not found</source>
        <translation>Reply from the BLAST server is empty. Similar sequences not found</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="586"/>
        <source>An empty sequence is searched</source>
        <translation>An empty sequence is searched</translation>
    </message>
    <message>
        <location filename="../src/tasks/UniprotBlastTask.cpp" line="587"/>
        <source>Settings are invalid</source>
        <translation>Settings are invalid</translation>
    </message>
</context>
<context>
    <name>U2::UserAppsSettings</name>
    <message>
        <location filename="../src/globals/UserApplicationsSettings.cpp" line="79"/>
        <source>Cleaning temp dir: %1</source>
        <translation>Cleaning temp dir: %1</translation>
    </message>
</context>
<context>
    <name>U2::VFSAdapterFactory</name>
    <message>
        <location filename="../src/io/VFSAdapter.cpp" line="32"/>
        <source>Memory buffer</source>
        <translation>Memory buffer</translation>
    </message>
</context>
<context>
    <name>U2SequenceImporter</name>
    <message>
        <location filename="../src/util/U2SequenceUtils.cpp" line="292"/>
        <source>Sequence was not imported. Probably, this is because the sequence is empty.</source>
        <translation>Sequence was not imported. Probably, this is because the sequence is empty.</translation>
    </message>
</context>
</TS>
