<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0">
<context>
    <name>BuildSuffixArraySettings</name>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="20"/>
        <source>Reference fragmentation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="28"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences on the amount of parts the reference will be divided. It is better to make it bigger, but it influences on the amount of memory used during the aligning.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="81"/>
        <source>Total memory usage:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="124"/>
        <source>System memory size:</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>GenomeAlignerSettings</name>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="14"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="30"/>
        <source>Chech this box if mismatches between the reference sequence and the reads are allowed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="33"/>
        <source>Mismatches allowed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="47"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="73"/>
        <source>Select the number of mismatched nucleotides allowed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="86"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="112"/>
        <source>Select the percentage of mismatches allowed. Note, that absolute number of mismatches can vary for different reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="50"/>
        <source>Mismatches number</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="24"/>
        <source>Common parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="89"/>
        <source>Percentage of mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="136"/>
        <source>Set short reads aligning options.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="150"/>
        <source>Use both the read and its reverse complement during aligning.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="153"/>
        <source>Align reverse complement reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="163"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Report only about best alignments (in terms of mismatches).&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="170"/>
        <source>Use &quot;best&quot;-mode during the aligning</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="225"/>
        <source>Use an openCL-enabled GPU during the aligning.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="228"/>
        <source>Use GPU-optimization</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="254"/>
        <source>Advanced parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="260"/>
        <source>Maximum memory for short reads:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="268"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;You can change the amount of memory that will be used for keeping short-reads during the aligning. It is better to make it bigger.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="321"/>
        <source>Total memory usage:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="364"/>
        <source>System memory size:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="406"/>
        <source>Index parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="412"/>
        <source>Reference fragmentation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="420"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences on the amount of parts the reference will be divided. It is better to make it bigger, but it influences on the amount of memory used during the aligning.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="470"/>
        <source>Index memory usage size:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="501"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p align=&quot;justify&quot; style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;You can choose a temporary directory for saving index files for the reference that will be built during the aligning. If you need to run this algorithm another time with the &lt;/span&gt;&lt;span style=&quot; font-size:8pt; font-weight:600;&quot;&gt;equal reference&lt;/span&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt; and with the &lt;/span&gt;&lt;span style=&quot; font-size:8pt; font-weight:600;&quot;&gt;equal reference fragmentation parameter &lt;/span&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;you can use this prebuilt index that will be located at this temporary directory.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="540"/>
        <source>Directory for index files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="180"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="206"/>
        <source>Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="530"/>
        <source>Specify an index to build during execution of the selected short reads aligning algorithm. If this option is selected, the value is &lt;b&gt;required&lt;/b&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="523"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="139"/>
        <source>Align options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="183"/>
        <source>Omit reads with qualities lower than</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2</name>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="53"/>
        <source>UGENE genome aligner</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerCMDLineTask</name>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="50"/>
        <source>Run genome aligner from command line</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="122"/>
        <source>Finished parsing genome aligner options.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="134"/>
        <source>Path to reference sequence is not set.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="140"/>
        <source>Short reads list is empty.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="144"/>
        <source>Reference (index or sequence) is not set.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="187"/>
        <source>  --%1    Use this flag to only build index for reference sequence.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="188"/>
        <source>  --%1    Path to reference genome sequence

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="189"/>
        <source>  --%1    Path to short-reads data in FASTA or FASTQ format

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="190"/>
        <source>  --%1    Path to prebuilt index (base file name or with .idx extension). If not set, index is searched in system temporary directory. If --build-index option is applied, index will be saved to specified path.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="191"/>
        <source>  --%1    Path to output alignment in UGENEDB or SAM format (see --%2)

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="192"/>
        <source>  --%1    Memory size (in Mbs) reserved for short-reads. The bigger value the faster algorithm works. Default value depends on available system memory.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="193"/>
        <source>  --%1    Index fragmentation size (in Mbs). Small fragments better fit into RAM, allowing to load more short reads. Default value is 10.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="194"/>
        <source>  --%1    Absolute amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="195"/>
        <source>  --%1    Percentage amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="196"/>
        <source>  --%1    Use both the read and its reverse complement during the aligning.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="197"/>
        <source>  --%1    Report only about best alignments (in terms of mismatches).

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="198"/>
        <source>  --%1    Omit reads with qualities lower than the specified value. Reads which have no qualities are not omitted. Default value is 0.

</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="199"/>
        <source>  --%1    Output aligned reads in SAM format. Default value is false.

</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerFindTask</name>
    <message>
        <location filename="../src/SuffixSearchCUDA.cpp" line="170"/>
        <source>The CUDA optimization is not included into this bundle.        Please make sure to set UGENE_CUDA_DECTECTED=1 when building UGENE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SuffixSearchCUDA.cpp" line="221"/>
        <source>The CUDA optimization is not included into this bundle.                                            Please make sure to set UGENE_CUDA_DECTECTED=1 when building UGENE</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerIndexTask</name>
    <message>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="90"/>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="136"/>
        <source>File %1 is not found. Try to create index another time.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="181"/>
        <source>Can not init short reads loader. %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerPlugin</name>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="71"/>
        <source>UGENE genome aligner</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="71"/>
        <source>Assembly DNA to reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="111"/>
        <source>UGENE Short Reads Aligner</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="112"/>
        <source>UGENE Genome Aligner is an efficient and fast tool for short read alignment.It has 2 work modes: build index and align short reads (default mode).
If there is no index available for reference sequence it will be built on the fly.

Usage: ugene --genome-aligner { --option[=argument] }

Options
--------

%1
Examples
--------

Build index for reference sequence:
ugene --genome-aligner --build-index --reference=/path/to/ref

Align short reads using existing index:
ugene --genome-aligner --reference=/path/to/ref --short-reads=/path/to/reads --result=/path/to/result
</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerSettingsWidget</name>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="102"/>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="163"/>
        <source>This index file is corrupted. Please, load a valid index file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="157"/>
        <source>The index directory has already contain the prebuilt index. But its reference fragmentation parameter is %1 and it doesn&apos;t equal to the parameter you have chosen (%2).

Press &quot;Ok&quot; to delete this index file and create a new during the aligning.
Press &quot;Cancel&quot; to change this parameter or the index directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="173"/>
        <source>Set index files directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerTask</name>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="112"/>
        <source>Genome Aligner settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="113"/>
        <source>Index file name: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="114"/>
        <source>Use prebuilt index: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="166"/>
        <source>Can not init short reads loader.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="293"/>
        <source>The aligning is finished.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="294"/>
        <source>Whole working time = %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="296"/>
        <location filename="../src/GenomeAlignerTask.cpp" line="300"/>
        <source>%1% reads aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="297"/>
        <source>Short-reads loading time = %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="298"/>
        <source>Results writing time = %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="302"/>
        <source>Index loading time = %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="303"/>
        <source>Short-reads IO time = %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerBuildPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="317"/>
        <source>Build genome aligner index from %1 and send it url to output.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerBuildWorker</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="236"/>
        <source>Reference</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="237"/>
        <source>Reference sequence url. The short reads will be aligned to this reference genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="238"/>
        <source>Genome aligner index builder</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="239"/>
        <source>GenomeAlignerBuild builds an index from a set of DNA sequences. GenomeAlignerBuild outputs a set of 3 files with suffixes .idx, .ref, .sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="240"/>
        <source>Index</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="241"/>
        <source>Output index url.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="242"/>
        <source>Reference fragmentation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="243"/>
        <source>Reference fragmentation size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="279"/>
        <source>Reference sequence URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="283"/>
        <source>Result index URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="306"/>
        <source>Genome aligner index building finished. Result name is %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerIndexReaderPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="391"/>
        <source>Read genome aligner index from %1 and send it url to output.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerIndexReaderWorker</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="327"/>
        <source>Genome aligner index</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="327"/>
        <source>Result of genome aligner index builder.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="333"/>
        <source>Genome aligner index reader</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="334"/>
        <source>Read a set of several files with extensions .idx, .ref, .X.sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="335"/>
        <source>Index</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="336"/>
        <source>Select an index file with the .idx extension</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="367"/>
        <source>Index URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="380"/>
        <source>Reading genome aligner index finished. Result name is %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="216"/>
        <location filename="../src/GenomeAlignerWorker.cpp" line="217"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="219"/>
        <source>Align short reads %1 to the reference genome %2 and send it to output.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerWorker</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="74"/>
        <source>Short read sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="74"/>
        <source>Short reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="76"/>
        <source>Short reads alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="76"/>
        <source>Result of alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="88"/>
        <source>Is absolute mismatches values?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="89"/>
        <source>&lt;html&gt;&lt;body&gt;&lt;p&gt;&lt;b&gt;true&lt;/b&gt; - absolute mismatches mode is used&lt;/p&gt;&lt;p&gt;&lt;b&gt;false&lt;/b&gt; - percentage mismatches mode is used&lt;/p&gt;                                You can choose absolute or percentage mismatches values mode.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="91"/>
        <source>Absolute mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="92"/>
        <source>&lt;html&gt;&lt;body&gt;Number of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="93"/>
        <source>Percentage mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="94"/>
        <source>&lt;html&gt;&lt;body&gt;Percentage of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="95"/>
        <source>Align reverse complement reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="96"/>
        <source>&lt;html&gt;&lt;body&gt;Set this option to align both direct and reverse complement reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="97"/>
        <source>Use &quot;best&quot;-mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="98"/>
        <source>&lt;html&gt;&lt;body&gt;Report only the best alignment for each read (in terms of mismatches).&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="99"/>
        <source>Omit reads with qualities lower than</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="100"/>
        <source>&lt;html&gt;&lt;body&gt;Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.                                &lt;p&gt;Set &lt;b&gt;&quot;0&quot;&lt;/b&gt; to switch off this option.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="102"/>
        <source>Use GPU-optimization</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="103"/>
        <source>&lt;html&gt;&lt;body&gt;Use GPU-calculatings while aligning reads. This option requires OpenCL-enable GPU-device.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="113"/>
        <source>UGENE genome aligner</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="114"/>
        <source>Unique UGENE algorithm for aligning short reads to reference genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="172"/>
        <source>Short reads list is empty.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
