<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en">
<context>
    <name>BuildSuffixArraySettings</name>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="29"/>
        <source>Reference fragmentation</source>
        <translation>Reference fragmentation</translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="43"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences the number of parts the
             reference will be divided. It is better to make it bigger, but it influences the
             amount of memory used during the alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences the number of parts the
             reference will be divided. It is better to make it bigger, but it influences the
             amount of memory used during the alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="103"/>
        <source>Total memory usage:</source>
        <translation>Total memory usage:</translation>
    </message>
    <message>
        <location filename="../src/ui/BuildSuffixArraySettings.ui" line="130"/>
        <source>System memory size:</source>
        <translation>System memory size:</translation>
    </message>
</context>
<context>
    <name>GenomeAlignerSettings</name>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="14"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="36"/>
        <source>Chech this box if mismatches between the reference sequence and the reads are allowed.</source>
        <translation>Check this box if mismatches between the reference sequence and the reads are allowed.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="39"/>
        <source>Mismatches allowed</source>
        <translation>Mismatches allowed</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="53"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="79"/>
        <source>Select the number of mismatched nucleotides allowed.</source>
        <translation>Select the number of mismatched nucleotides allowed.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="92"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="118"/>
        <source>Select the percentage of mismatches allowed. Note, that absolute number of mismatches can vary for different reads.</source>
        <translation>Select the percentage of mismatches allowed. Note, that absolute number of mismatches can vary for different reads.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="56"/>
        <source>Mismatches number</source>
        <translation>Mismatches number</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="30"/>
        <source>Common parameters</source>
        <translation>Common parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="95"/>
        <source>Percentage of mismatches</source>
        <translation>Percentage of mismatches</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="148"/>
        <source>Set short reads aligning options.</source>
        <translation>Set short reads aligning options.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="171"/>
        <source>Use both the read and its reverse complement during aligning.</source>
        <translation>Use both the read and its reverse complement during aligning.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="174"/>
        <source>Align reverse complement reads</source>
        <translation>Align reverse complement reads</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="184"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Report only about best alignments (in terms of mismatches).&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Report only about best alignments (in terms of mismatches).&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="191"/>
        <source>Use &quot;best&quot;-mode during the alignment</source>
        <translation>Use &quot;best&quot;-mode during the alignment</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="246"/>
        <source>Use an openCL-enabled GPU during the aligning.</source>
        <translation>Use an openCL-enabled GPU during the aligning.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="249"/>
        <source>Use GPU-optimization</source>
        <translation>Use GPU-optimization</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="275"/>
        <source>Advanced parameters</source>
        <translation>Advanced parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="281"/>
        <source>Maximum memory for short reads</source>
        <translation>Maximum memory for short reads</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="331"/>
        <source>Total memory usage:</source>
        <translation>Total memory usage:</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="374"/>
        <source>System memory size:</source>
        <translation>System memory size:</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="416"/>
        <source>Index parameters</source>
        <translation>Index parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="422"/>
        <source>Reference fragmentation</source>
        <translation>Reference fragmentation</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="482"/>
        <source>Index memory usage size:</source>
        <translation>Index memory usage size:</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="545"/>
        <source>Directory for index files:</source>
        <translation>Directory for index files:</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="201"/>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="227"/>
        <source>Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.</source>
        <translation>Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="289"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Allows one to decrease the load on the computer on one side and to increase the computation speed of the task on the other side.
                 &lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Allows one to decrease the load on the computer on one side and to increase the computation speed of the task on the other side.
                 &lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="430"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences the number of parts the
                 reference will be divided. It is better to make it bigger, but it influences the
                 amount of memory used during the alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This parameter influences the number of parts the
                 reference will be divided. It is better to make it bigger, but it influences the
                 amount of memory used during the alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="513"/>
        <source>You can choose a temporary directory for saving index files for the reference that will be built during the alignment. If you need to run this algorithm one more time with the same reference and with the same reference fragmentation parameter, you can use this prebuilt index that will be located in the temporary directory.</source>
        <translation>You can choose a temporary directory for saving index files for the reference that will be built during the alignment. If you need to run this algorithm one more time with the same reference and with the same reference fragmentation parameter, you can use this prebuilt index that will be located in the temporary directory.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="535"/>
        <source>Specify an index to build during execution of the selected short reads aligning algorithm. If this option is selected, the value is &lt;b&gt;required&lt;/b&gt;.</source>
        <translation>Specify an index to build during execution of the selected short reads aligning algorithm. If this option is selected, the value is &lt;b&gt;required&lt;/b&gt;.</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="528"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="151"/>
        <source>Align options</source>
        <translation>Align options</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettings.ui" line="204"/>
        <source>Omit reads with qualities lower than</source>
        <translation>Omit reads with qualities lower than</translation>
    </message>
</context>
<context>
    <name>GenomeAlignerSettingsWidget</name>
    <message>
        <location filename="../src/ui/GenomeAlignerSettingsWidget.ui" line="14"/>
        <source>Genome aligner settings</source>
        <translation>Genome aligner settings</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettingsWidget.ui" line="20"/>
        <source>Directories</source>
        <translation>Directories</translation>
    </message>
    <message>
        <location filename="../src/ui/GenomeAlignerSettingsWidget.ui" line="26"/>
        <source>Directory for built indexes</source>
        <translation>Directory for built indexes</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="55"/>
        <source>UGENE Genome Aligner</source>
        <translation>UGENE Genome Aligner</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerCMDLineTask</name>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="49"/>
        <source>Run genome aligner from command line</source>
        <translation>Run genome aligner from command line</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="116"/>
        <source>Finished parsing genome aligner options.</source>
        <translation>Finished parsing genome aligner options.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="128"/>
        <source>Path to reference sequence is not set.</source>
        <translation>Path to reference sequence is not set.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="134"/>
        <source>Short reads list is empty.</source>
        <translation>Short reads list is empty.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="138"/>
        <source>Reference (index or sequence) is not set.</source>
        <translation>Reference (index or sequence) is not set.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="182"/>
        <source>  --%1    Use this flag to only build index for reference sequence.

</source>
        <translation>  --%1    Use this flag to only build index for reference sequence.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="183"/>
        <source>  --%1    Path to reference genome sequence

</source>
        <translation>  --%1    Path to reference genome sequence

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="184"/>
        <source>  --%1    Path to short-reads data in FASTA or FASTQ format

</source>
        <translation>  --%1    Path to short-reads data in FASTA or FASTQ format

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="185"/>
        <source>  --%1    Path to prebuilt index (base file name or with .idx extension). If not set, index is searched in system temporary directory. If --build-index option is applied, index will be saved to specified path.

</source>
        <translation>  --%1    Path to prebuilt index (base file name or with .idx extension). If not set, index is searched in system temporary directory. If --build-index option is applied, index will be saved to specified path.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="186"/>
        <source>  --%1    Path to output alignment in UGENEDB or SAM format (see --%2)

</source>
        <translation>  --%1    Path to output alignment in UGENEDB or SAM format (see --%2)

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="187"/>
        <source>  --%1    Memory size (in Mbs) reserved for short-reads. The bigger value the faster algorithm works. Default value depends on available system memory.

</source>
        <translation>  --%1    Memory size (in Mbs) reserved for short-reads. The bigger value the faster algorithm works. Default value depends on available system memory.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="188"/>
        <source>  --%1    Index fragmentation size (in Mbs). Small fragments better fit into RAM, allowing to load more short reads. Default value is 10.

</source>
        <translation>  --%1    Index fragmentation size (in Mbs). Small fragments better fit into RAM, allowing to load more short reads. Default value is 10.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="189"/>
        <source>  --%1    Absolute amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</source>
        <translation>  --%1    Absolute amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="190"/>
        <source>  --%1    Percentage amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</source>
        <translation>  --%1    Percentage amount of allowed mismatches per every short-read (mutually exclusive with --%2). Default value is 0.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="191"/>
        <source>  --%1    Use both the read and its reverse complement during the aligning.

</source>
        <translation>  --%1    Use both the read and its reverse complement during the aligning.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="192"/>
        <source>  --%1    Report only about best alignments (in terms of mismatches).

</source>
        <translation>  --%1    Report only about best alignments (in terms of mismatches).

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="193"/>
        <source>  --%1    Omit reads with qualities lower than the specified value. Reads which have no qualities are not omitted. Default value is 0.

</source>
        <translation>  --%1    Omit reads with qualities lower than the specified value. Reads which have no qualities are not omitted. Default value is 0.

</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerCMDLineTask.cpp" line="194"/>
        <source>  --%1    Output aligned reads in SAM format. Default value is false.

</source>
        <translation>  --%1    Output aligned reads in SAM format. Default value is false.

</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerIndexTask</name>
    <message>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="90"/>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="136"/>
        <source>File %1 is not found. Try to create index another time.</source>
        <translation>File %1 is not found. Try to create index another time.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexTask.cpp" line="183"/>
        <source>Can not init short reads loader. %1</source>
        <translation>Can not init short reads loader. %1</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerPlugin</name>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="72"/>
        <source>UGENE Genome Aligner</source>
        <translation>UGENE Genome Aligner</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="72"/>
        <source>Assembly DNA to reference sequence</source>
        <translation>Assembly DNA to reference sequence</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="121"/>
        <source>UGENE Short Reads Aligner</source>
        <translation>UGENE Short Reads Aligner</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerPlugin.cpp" line="122"/>
        <source>UGENE Genome Aligner is an efficient and fast tool for short read alignment.It has 2 work modes: build index and align short reads (default mode).
If there is no index available for reference sequence it will be built on the fly.

Usage: ugene --genome-aligner { --option[=argument] }

Options
--------

%1
Examples
--------

Build index for reference sequence:
ugene --genome-aligner --build-index --reference=/path/to/ref

Align short reads using existing index:
ugene --genome-aligner --reference=/path/to/ref --short-reads=/path/to/reads --result=/path/to/result
</source>
        <translation>UGENE Genome Aligner is an efficient and fast tool for short read alignment.It has 2 work modes: build index and align short reads (default mode).
If there is no index available for reference sequence it will be built on the fly.

Usage: ugene --genome-aligner { --option[=argument] }

Options
--------

%1
Examples
--------

Build index for reference sequence:
ugene --genome-aligner --build-index --reference=/path/to/ref

Align short reads using existing index:
ugene --genome-aligner --reference=/path/to/ref --short-reads=/path/to/reads --result=/path/to/result</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerSettingsPageController</name>
    <message>
        <location filename="../src/GenomeAlignerSettingsController.cpp" line="55"/>
        <source>Genome Aligner</source>
        <translation>Genome Aligner</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerSettingsPageWidget</name>
    <message>
        <location filename="../src/GenomeAlignerSettingsController.cpp" line="96"/>
        <source>Choose Directory</source>
        <translation>Choose Directory</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerSettingsWidget</name>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="112"/>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="170"/>
        <source>This index file is corrupted. Please, load a valid index file.</source>
        <translation>This index file is corrupted. Please, load a valid index file.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="164"/>
        <source>The index directory has already contain the prebuilt index. But its reference fragmentation parameter is %1 and it doesn&apos;t equal to the parameter you have chosen (%2).

Press &quot;Ok&quot; to delete this index file and create a new during the aligning.
Press &quot;Cancel&quot; to change this parameter or the index directory.</source>
        <translation>The index directory has already contain the prebuilt index. But its reference fragmentation parameter is %1 and it doesn&apos;t equal to the parameter you have chosen (%2).

Press &quot;Ok&quot; to delete this index file and create a new during the aligning.
Press &quot;Cancel&quot; to change this parameter or the index directory.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerSettingsWidget.cpp" line="180"/>
        <source>Set index files directory</source>
        <translation>Set index files directory</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAlignerTask</name>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="131"/>
        <source>Genome Aligner settings</source>
        <translation>Genome Aligner settings</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="132"/>
        <source>Index file name: %1</source>
        <translation>Index file name: %1</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="133"/>
        <source>Use prebuilt index: %2</source>
        <translation>Use prebuilt index: %2</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="206"/>
        <source>Can not init short reads loader.</source>
        <translation>Can not init short reads loader.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="333"/>
        <source>The aligning is finished.</source>
        <translation>The aligning is finished.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="334"/>
        <source>Whole working time = %1.</source>
        <translation>Whole working time = %1.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="335"/>
        <source>%1% reads aligned.</source>
        <translation>%1% reads aligned.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="337"/>
        <source>Short-reads loading time = %1</source>
        <translation>Short-reads loading time = %1</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="338"/>
        <source>Results writing time = %1</source>
        <translation>Results writing time = %1</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="340"/>
        <source>Index loading time = %1</source>
        <translation>Index loading time = %1</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerTask.cpp" line="341"/>
        <source>Short-reads IO time = %1</source>
        <translation>Short-reads IO time = %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerBuildPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="151"/>
        <source>Build genome aligner index from %1 and send it url to output.</source>
        <translation>Build genome aligner index from %1 and send it url to output.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerBuildWorker</name>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="66"/>
        <source>Reference</source>
        <translation>Reference</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="67"/>
        <source>Reference sequence url. The short reads will be aligned to this reference genome.</source>
        <translation>Reference sequence url. The short reads will be aligned to this reference genome.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="68"/>
        <source>Genome aligner index builder</source>
        <translation>Genome aligner index builder</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="69"/>
        <source>GenomeAlignerBuild builds an index from a set of DNA sequences. GenomeAlignerBuild outputs a set of 3 files with suffixes .idx, .ref, .sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</source>
        <translation>GenomeAlignerBuild builds an index from a set of DNA sequences. GenomeAlignerBuild outputs a set of 3 files with suffixes .idx, .ref, .sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="70"/>
        <source>Index</source>
        <translation>Index</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="71"/>
        <source>Output index url.</source>
        <translation>Output index url.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="72"/>
        <source>Reference fragmentation</source>
        <translation>Reference fragmentation</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="73"/>
        <source>Reference fragmentation size</source>
        <translation>Reference fragmentation size</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="109"/>
        <source>Reference sequence URL is empty</source>
        <translation>Reference sequence URL is empty</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="113"/>
        <source>Result index URL is empty</source>
        <translation>Result index URL is empty</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="136"/>
        <source>Genome aligner index building finished. Result name is %1</source>
        <translation>Genome aligner index building finished. Result name is %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerIndexReaderPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="229"/>
        <source>Read genome aligner index from %1 and send it url to output.</source>
        <translation>Read genome aligner index from %1 and send it url to output.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerIndexReaderWorker</name>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="161"/>
        <source>Genome aligner index</source>
        <translation>Genome aligner index</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="161"/>
        <source>Result of genome aligner index builder.</source>
        <translation>Result of genome aligner index builder.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="167"/>
        <source>Genome aligner index reader</source>
        <translation>Genome aligner index reader</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="168"/>
        <source>Read a set of several files with extensions .idx, .ref, .X.sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</source>
        <translation>Read a set of several files with extensions .idx, .ref, .X.sarr. These files together constitute the index: they are all that is needed to align reads to that reference.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="169"/>
        <source>Index</source>
        <translation>Index</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="170"/>
        <source>Select an index file with the .idx extension</source>
        <translation>Select an index file with the .idx extension</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="201"/>
        <source>Index URL is empty</source>
        <translation>Index URL is empty</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerIndexWorker.cpp" line="214"/>
        <source>Reading genome aligner index finished. Result name is %1</source>
        <translation>Reading genome aligner index finished. Result name is %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerPrompter</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="198"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="202"/>
        <source>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="203"/>
        <source> to reference genome &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> to reference genome &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomeAlignerWorker</name>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="305"/>
        <source>Is absolute mismatches values?</source>
        <translation>Is absolute mismatches values?</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="247"/>
        <source>Bowtie2 cannot recognize read pairs from the same file. Please, perform demultiplexing first.</source>
        <translation>Bowtie2 cannot recognize read pairs from the same file. Please, perform demultiplexing first.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="222"/>
        <source>The slot must be not empty: &apos;%1&apos;</source>
        <translation>The slot must be not empty: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="260"/>
        <source>URL of a file with reads</source>
        <translation>URL of a file with reads</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="261"/>
        <source>Input reads to be aligned.</source>
        <translation>Input reads to be aligned.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="263"/>
        <source>URL of a file with mate reads</source>
        <translation>URL of a file with mate reads</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="264"/>
        <source>Input mate reads to be aligned.</source>
        <translation>Input mate reads to be aligned.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="270"/>
        <source>Genome aligner data</source>
        <translation>Genome aligner data</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="271"/>
        <source>Input reads to be aligned with Bowtie2.</source>
        <translation>Input reads to be aligned with Bowtie2.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="278"/>
        <source>Assembly URL</source>
        <translation>Assembly URL</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="279"/>
        <source>Output assembly URL.</source>
        <translation>Output assembly URL.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="282"/>
        <source>Genome aligner output data</source>
        <translation>Genome aligner output data</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="283"/>
        <source>Output assembly files.</source>
        <translation>Output assembly files.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="294"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="295"/>
        <source>Directory to save UGENE genome aligner output files.</source>
        <translation>Directory to save UGENE genome aligner output files.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="298"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="299"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation>Base name of the output file. &apos;out.sam&apos; by default</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="302"/>
        <source>Reference genome</source>
        <translation>Reference genome</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="303"/>
        <source>Path to indexed reference genome.</source>
        <translation>Path to indexed reference genome.</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="306"/>
        <source>&lt;html&gt;&lt;body&gt;&lt;p&gt;&lt;b&gt;true&lt;/b&gt; - absolute mismatches mode is used&lt;/p&gt;&lt;p&gt;&lt;b&gt;false&lt;/b&gt; - percentage mismatches mode is used&lt;/p&gt;                                    You can choose absolute or percentage mismatches values mode.&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;&lt;p&gt;&lt;b&gt;true&lt;/b&gt; - absolute mismatches mode is used&lt;/p&gt;&lt;p&gt;&lt;b&gt;false&lt;/b&gt; - percentage mismatches mode is used&lt;/p&gt;                                    You can choose absolute or percentage mismatches values mode.&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="309"/>
        <source>Absolute mismatches</source>
        <translation>Absolute mismatches</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="310"/>
        <source>&lt;html&gt;&lt;body&gt;Number of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Number of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="312"/>
        <source>Percentage mismatches</source>
        <translation>Percentage mismatches</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="313"/>
        <source>&lt;html&gt;&lt;body&gt;Percentage of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Percentage of mismatches allowed while aligning reads.&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="315"/>
        <source>Align reverse complement reads</source>
        <translation>Align reverse complement reads</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="316"/>
        <source>&lt;html&gt;&lt;body&gt;Set this option to align both direct and reverse complement reads.&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Set this option to align both direct and reverse complement reads.&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="318"/>
        <source>Use &quot;best&quot;-mode</source>
        <translation>Use &quot;best&quot;-mode</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="319"/>
        <source>&lt;html&gt;&lt;body&gt;Report only the best alignment for each read (in terms of mismatches).&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Report only the best alignment for each read (in terms of mismatches).&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="321"/>
        <source>Omit reads with qualities lower than</source>
        <translation>Omit reads with qualities lower than</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="322"/>
        <source>&lt;html&gt;&lt;body&gt;Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.                                    &lt;p&gt;Set &lt;b&gt;&quot;0&quot;&lt;/b&gt; to switch off this option.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Omit reads with qualities lower than the specified value. Reads that have no qualities are not omited.                                    &lt;p&gt;Set &lt;b&gt;&quot;0&quot;&lt;/b&gt; to switch off this option.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="350"/>
        <source>Align Reads with UGENE Genome Aligner</source>
        <translation>Align Reads with UGENE Genome Aligner</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="329"/>
        <source>Use GPU-optimization</source>
        <translation>Use GPU-optimization</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="330"/>
        <source>&lt;html&gt;&lt;body&gt;Use GPU-calculatings while aligning reads. This option requires OpenCL-enable GPU-device.&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;body&gt;Use GPU-calculatings while aligning reads. This option requires OpenCL-enable GPU-device.&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="351"/>
        <source>Unique UGENE algorithm for aligning short reads to reference genome</source>
        <translation>Unique UGENE algorithm for aligning short reads to reference genome</translation>
    </message>
    <message>
        <location filename="../src/GenomeAlignerWorker.cpp" line="95"/>
        <source>Short reads list is empty.</source>
        <translation>Short reads list is empty.</translation>
    </message>
</context>
<context>
    <name>U2::ShortReadAlignerCPU</name>
    <message>
        <location filename="../src/GenomeAlignerFindTask.cpp" line="208"/>
        <source>[%1] Index size for part %2/%3 is zero, skipping it.</source>
        <translation>[%1] Index size for part %2/%3 is zero, skipping it.</translation>
    </message>
</context>
<context>
    <name>U2::ShortReadAlignerOpenCL</name>
    <message>
        <location filename="../src/GenomeAlignerFindTask.cpp" line="306"/>
        <source>Index size for part %1/%2 is zero, skipping it.</source>
        <translation>Index size for part %1/%2 is zero, skipping it.</translation>
    </message>
</context>
</TS>
