<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en">
<context>
    <name>BlastAllSupportDialog</name>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="20"/>
        <source>Request to Local BLAST Database</source>
        <translation>Request to Local BLAST Database</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="42"/>
        <source>General options</source>
        <translation>General options</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="53"/>
        <source>Select search:</source>
        <translation>Select search:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="61"/>
        <source>blastn</source>
        <extracomment>qwe</extracomment>
        <translation>blastn</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="66"/>
        <source>blastp</source>
        <translation>blastp</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="76"/>
        <source>blastx</source>
        <translation>blastx</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="81"/>
        <source>tblastn</source>
        <translation>tblastn</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="86"/>
        <source>tblastx</source>
        <translation>tblastx</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="94"/>
        <source>Search for short, nearly exact matches</source>
        <translation>Search for short, nearly exact matches</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="101"/>
        <source>Expectation value </source>
        <translation>Expectation value </translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="108"/>
        <source>Megablast</source>
        <translation>Megablast</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="125"/>
        <source>Off</source>
        <translation>Off</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="159"/>
        <source>Both strands</source>
        <translation>Both strands</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="169"/>
        <source>Direct</source>
        <translation>Direct</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="176"/>
        <source>Complement</source>
        <translation>Complement</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="191"/>
        <source>Base name for BLAST DB files:</source>
        <translation>Base name for BLAST DB files:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="261"/>
        <source>Number of CPUs being used</source>
        <translation>Number of CPUs being used</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="299"/>
        <source>Advanced options</source>
        <translation>Advanced options</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="422"/>
        <source>Word size</source>
        <translation>Word size</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="559"/>
        <source>Gap costs</source>
        <translation>Gap costs</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="338"/>
        <source>Match scores</source>
        <translation>Match scores</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="368"/>
        <source>1 -3</source>
        <translation>1 -3</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="373"/>
        <source>1 -4</source>
        <translation>1 -4</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="378"/>
        <source>1 -2</source>
        <translation>1 -2</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="383"/>
        <source>1 -1</source>
        <translation>1 -1</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="388"/>
        <source>2 -7</source>
        <translation>2 -7</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="393"/>
        <source>2 -5</source>
        <translation>2 -5</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="398"/>
        <source>2 -3</source>
        <translation>2 -3</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="403"/>
        <source>4 -5</source>
        <translation>4 -5</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="408"/>
        <source>5 -4</source>
        <translation>5 -4</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="435"/>
        <source>Matrix</source>
        <translation>Matrix</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="71"/>
        <source>gpu-blastp</source>
        <translation>gpu-blastp</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="205"/>
        <source>Database path:</source>
        <translation>Database path:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="229"/>
        <source>Select
a database file</source>
        <translation>Select
a database file</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="443"/>
        <source>BLOSUM62</source>
        <translation>BLOSUM62</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="448"/>
        <source>BLOSUM45</source>
        <translation>BLOSUM45</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="453"/>
        <source>BLOSUM80</source>
        <translation>BLOSUM80</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="458"/>
        <source>PAM30</source>
        <translation>PAM30</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="463"/>
        <source>PAM70</source>
        <translation>PAM70</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="351"/>
        <source>Service</source>
        <translation>Service</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="115"/>
        <source>Max hits:</source>
        <translation>Max hits:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="314"/>
        <source>plain</source>
        <translation>plain</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="319"/>
        <source>psi</source>
        <translation>psi</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="324"/>
        <source>phi</source>
        <translation>phi</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="474"/>
        <source>PHI pattern</source>
        <translation>PHI pattern</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="514"/>
        <source>Threshold for extending hits</source>
        <translation>Threshold for extending hits</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="517"/>
        <source>Threshold</source>
        <translation>Threshold</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="566"/>
        <source>Composition-based statistics</source>
        <translation>Composition-based statistics</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="574"/>
        <source>D or d: default (equivalent to 2 )</source>
        <translation>D or d: default (equivalent to 2 )</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="579"/>
        <source>0 or F or f: No composition-based statistics</source>
        <translation>0 or F or f: No composition-based statistics</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="584"/>
        <source>1: Composition-based statistics as in NAR 29:2994-3005, 2001</source>
        <translation>1: Composition-based statistics as in NAR 29:2994-3005, 2001</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="589"/>
        <source>2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</source>
        <translation>2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="594"/>
        <source>3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally</source>
        <translation>3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="606"/>
        <source>Filters</source>
        <translation>Filters</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="612"/>
        <source>Low complexity filter</source>
        <translation>Low complexity filter</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="619"/>
        <source>Human repeats filter</source>
        <translation>Human repeats filter</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="629"/>
        <source>Masks</source>
        <translation>Masks</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="635"/>
        <source>Mask for lookup table only</source>
        <translation>Mask for lookup table only</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="645"/>
        <source>Mask lower case letters</source>
        <translation>Mask lower case letters</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="671"/>
        <source>Extension options</source>
        <translation>Extension options</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="679"/>
        <source>X dropoff value (in bits)</source>
        <translation>X dropoff value (in bits)</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="685"/>
        <source>For gapped alignment</source>
        <translation>For gapped alignment</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="699"/>
        <source>For ungapped extensions</source>
        <translation>For ungapped extensions</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="706"/>
        <source>For final gapped alignment</source>
        <translation>For final gapped alignment</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="730"/>
        <source>Multiple Hits window size</source>
        <translation>Multiple Hits Window Size</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="742"/>
        <source>Perform gapped alignment</source>
        <translation>Perform gapped alignment</translation>
    </message>
</context>
<context>
    <name>BlastDBCmdDialog</name>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="26"/>
        <source>Fetch Sequence from BLAST Database</source>
        <translation>Fetch Sequence from BLAST Database</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="44"/>
        <source>Entery query ID:</source>
        <translation>Enter query ID:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="54"/>
        <source>Select database :</source>
        <translation>Select database :</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="70"/>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="124"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="79"/>
        <source>Type of file(s)</source>
        <translation>Type of file(s)</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="85"/>
        <source>nucleotide</source>
        <translation>nucleotide</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="95"/>
        <source>protein</source>
        <translation>protein</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="108"/>
        <source>Output path:</source>
        <translation>Output path:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="136"/>
        <source>Add to project</source>
        <translation>Add to project</translation>
    </message>
</context>
<context>
    <name>Bowtie2Settings</name>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="41"/>
        <source>Parameters</source>
        <translation>Parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="60"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="76"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, Bowtie2 performs end-to-end read alignment. That is, it searches for alignments involving all of the read characters. This is also called an &amp;quot;untrimmed&amp;quot; or &amp;quot;unclipped&amp;quot; alignment.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the --local option is specified, Bowtie2 performs local read alignment. In this mode, Bowtie2 might &amp;quot;trim&amp;quot; or &amp;quot;clip&amp;quot; some read characters from one or both ends of the alignment if doing so maximizes the alignment score.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, Bowtie2 performs end-to-end read alignment. That is, it searches for alignments involving all of the read characters. This is also called an &amp;quot;untrimmed&amp;quot; or &amp;quot;unclipped&amp;quot; alignment.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the --local option is specified, Bowtie2 performs local read alignment. In this mode, Bowtie2 might &amp;quot;trim&amp;quot; or &amp;quot;clip&amp;quot; some read characters from one or both ends of the alignment if doing so maximizes the alignment score.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="63"/>
        <source>Mode</source>
        <translation>Mode</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="86"/>
        <source>--end-to-end</source>
        <translation>--end-to-end</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="91"/>
        <source>--local</source>
        <translation>--local</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="106"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="174"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Pads&amp;quot; dynamic programming problems by specified number of columns on either side to allow gaps.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Pads&amp;quot; dynamic programming problems by specified number of columns on either side to allow gaps.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="109"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation>Add columns to allow gaps (--dpad)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="119"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="187"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often much slower) but increases sensitivity.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often much slower) but increases sensitivity.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="132"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="197"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="148"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="252"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Disallow gaps within specified number of positions of the beginning or end of the read.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Disallow gaps within specified number of positions of the beginning or end of the read.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="161"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="229"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Use specified value as the seed for pseudo-random number generator.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Use specified value as the seed for pseudo-random number generator.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="164"/>
        <source>Seed (--seed)</source>
        <translation>Seed (--seed)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="190"/>
        <source>Number of mismatches</source>
        <translation>Number of mismatches</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="200"/>
        <source>Threads</source>
        <translation>Threads</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="210"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="239"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="242"/>
        <source>Seed length (--L)</source>
        <translation>Seed length (--L)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="255"/>
        <source>Disallow gaps (--gbar)</source>
        <translation>Disallow gaps (--gbar)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="284"/>
        <source>Flags</source>
        <translation>Flags</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="296"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments for the constituent mates. This is called &amp;quot;mixed mode.&amp;quot; To disable mixed mode, set this option.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 runs a little faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments for the constituent mates. This is called &amp;quot;mixed mode.&amp;quot; To disable mixed mode, set this option.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 runs a little faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="299"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation>No unpaired alignments (--no-mixed)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="306"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="309"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation>No discordant alignments (--no-discordant)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="316"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the forward reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the forward reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="319"/>
        <source>No forward orientation (--nofw)</source>
        <translation>No forward orientation (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="326"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the reverse-complement reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the reverse-complement reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="329"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>No reverse-complement orientation (--norc)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="336"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="339"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation>No overlapping mates (--no-overlap)</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="346"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the other in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the other in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="349"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation>No mates containing one another (--no-contain)</translation>
    </message>
</context>
<context>
    <name>BowtieBuildSettings</name>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="29"/>
        <source>Colorspace</source>
        <translation>Colorspace</translation>
    </message>
</context>
<context>
    <name>BowtieSettings</name>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="41"/>
        <source>Parameters</source>
        <translation>Parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="110"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;not just in the &amp;quot;seed&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The default is 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;rounding can be disabled with --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;not just in the &amp;quot;seed&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The default is 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;rounding can be disabled with --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="121"/>
        <source>Maq error (--maqerr)</source>
        <translation>Maq error (--maqerr)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="144"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The &amp;quot;seed length&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;i.e., the number of bases on the high-quality end of the read to which the -n ceiling applies.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The &amp;quot;seed length&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;i.e., the number of bases on the high-quality end of the read to which the -n ceiling applies.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="176"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The maximum insert size for valid paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;E.g. if -X 100 is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;that alignment is considered valid (as long as -I is also satisfied). &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;A 61-bp gap would not be valid in that case. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If trimming options -3 or -5 are also used, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;the -X constraint is applied with respect to the untrimmed mates, not the trimmed mates.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Default: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The maximum insert size for valid paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;E.g. if -X 100 is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;that alignment is considered valid (as long as -I is also satisfied). &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;A 61-bp gap would not be valid in that case. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If trimming options -3 or -5 are also used, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;the -X constraint is applied with respect to the untrimmed mates, not the trimmed mates.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Default: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="215"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</source>
        <translation>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="246"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator</source>
        <translation>Use &lt;int&gt; as the seed for pseudo-random number generator</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="60"/>
        <source>Mode:  </source>
        <translation>Mode:  </translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="67"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the -n option is specified (which is the default), &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie determines which alignments are valid according to the following policy, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;which is similar to Maq&apos;s default policy.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the -n option is specified (which is the default), &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie determines which alignments are valid according to the following policy, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;which is similar to Maq&apos;s default policy.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="82"/>
        <source>-n mode</source>
        <translation>-n mode</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="87"/>
        <source>-v mode</source>
        <translation>-v mode</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="153"/>
        <source>Seed length (--seedlen)</source>
        <translation>Seed length (--seedlen)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="189"/>
        <source>Maximum of backtracks (--maxbts)</source>
        <translation>Maximum of backtracks (--maxbts)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="223"/>
        <source>Descriptors memory usage (--chunkmbs)</source>
        <translation>Descriptors memory usage (--chunkmbs)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="249"/>
        <source>Seed (--seed)</source>
        <translation>Seed (--seed)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="266"/>
        <source>Threads</source>
        <translation>Threads</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="300"/>
        <source>Flags</source>
        <translation>Flags</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="312"/>
        <source>Colorspace (--color)</source>
        <translation>Colorspace (--color)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="329"/>
        <source>No Maq rounding (--nomaqround)</source>
        <translation>No Maq rounding (--nomaqround)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="339"/>
        <source>No forward orientation (--nofw)</source>
        <translation>No forward orientation (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="349"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>No reverse-complement orientation (--norc)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="369"/>
        <source>Try as hard (--tryhard)</source>
        <translation>Try as hard (--tryhard)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="383"/>
        <source>Best alignments (--best)</source>
        <translation>Best alignments (--best)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="393"/>
        <source>All alignments (--all)</source>
        <translation>All alignments (--all)</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="319"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq accepts quality values in the Phred quality scale, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but internally rounds values to the nearest 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;with a maximum of 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq accepts quality values in the Phred quality scale, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but internally rounds values to the nearest 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;with a maximum of 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="336"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="346"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="356"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Try as hard as possible to find valid alignments when they exist, including paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This is equivalent to specifying very high values for the --maxbts and --pairtries options.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is generally much slower than the default settings, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but can be useful for certain problems. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is slower when (a) the reference is very repetitive, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) the reads are low quality, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;or (c) not many reads have valid alignments.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Try as hard as possible to find valid alignments when they exist, including paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This is equivalent to specifying very high values for the --maxbts and --pairtries options.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is generally much slower than the default settings, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but can be useful for certain problems. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is slower when (a) the reference is very repetitive, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) the reads are low quality, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;or (c) not many reads have valid alignments.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="376"/>
        <source>Make Bowtie guarantee that reported singleton alignments are &quot;best&quot; in terms of stratum
(i.e. number of mismatches, or mismatches in the seed in the case of -n mode)
and in terms of the quality values at the mismatched position(s).
bowtie is somewhat slower when --best is specified.
           </source>
        <translation>Make Bowtie guarantee that reported singleton alignments are &quot;best&quot; in terms of stratum
(i.e. number of mismatches, or mismatches in the seed in the case of -n mode)
and in terms of the quality values at the mismatched position(s).
bowtie is somewhat slower when --best is specified.
</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="390"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation>Report all valid alignments per read or pair.</translation>
    </message>
</context>
<context>
    <name>BwaBuildSettings</name>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="54"/>
        <source>Index algorithm (-a)</source>
        <translation>Index algorithm (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="68"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="73"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="78"/>
        <source>is</source>
        <translation>is</translation>
    </message>
</context>
<context>
    <name>BwaMemSettings</name>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="36"/>
        <source>Base Options</source>
        <translation>Base Options</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="441"/>
        <source>Penalty for a mismatch</source>
        <translation>Penalty for a mismatch</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="409"/>
        <source>Score for a sequence match</source>
        <translation>Score for a sequence match</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="393"/>
        <source>Score for a match (-A)</source>
        <translation>Score for a match (-A)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="425"/>
        <source>Mismatch penalty (-B)</source>
        <translation>Mismatch penalty (-B)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="454"/>
        <source>Gap open penalty (-O)</source>
        <translation>Gap open penalty (-O)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="557"/>
        <source>Penalty for an unpaired read pair</source>
        <translation>Penalty for an unpaired read pair</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="586"/>
        <source>Minimum score to output</source>
        <translation>Minimum score to output</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="541"/>
        <source>Penalty unpaired (-U)</source>
        <translation>Penalty unpaired (-U)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="528"/>
        <source>Penalty for clipping </source>
        <translation>Penalty for clipping </translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="470"/>
        <source>Gap open penalty</source>
        <translation>Gap open penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="499"/>
        <source>Gap extension penalty; a gap of size k cost {-O} + {-E}*k</source>
        <translation>Gap extension penalty; a gap of size k cost {-O} + {-E}*k</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="512"/>
        <source>Penalty for clipping (-L)</source>
        <translation>Penalty for clipping (-L)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="483"/>
        <source>Gap extention penalty (-E)</source>
        <translation>Gap extention penalty (-E)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="570"/>
        <source>Score threshold (-T)</source>
        <translation>Score threshold (-T)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="209"/>
        <source>Internall seeds length (-r)</source>
        <translation>Internall seeds length (-r)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="107"/>
        <source>Min seed length (-k)</source>
        <translation>Min seed length (-k)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="69"/>
        <source>Look for internal seeds inside a seed longer than {-k} * FLOAT</source>
        <translation>Look for internal seeds inside a seed longer than {-k} * FLOAT</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="56"/>
        <source>Number of threads</source>
        <translation>Number of threads</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="193"/>
        <source>Off-diagonal X-dropoff</source>
        <translation>Off-diagonal X-dropoff</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="161"/>
        <source>Band width for banded alignment</source>
        <translation>Band width for banded alignment</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="145"/>
        <source>Band width (-w)</source>
        <translation>Band width (-w)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="123"/>
        <source>Minimum seed length</source>
        <translation>Minimum seed length</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="88"/>
        <source>Number of threads (-t)</source>
        <translation>Number of threads (-t)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="177"/>
        <source>Dropoff (-d)</source>
        <translation>Dropoff (-d)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="228"/>
        <source>Skip seeds threshold (-c)</source>
        <translation>Skip seeds threshold (-c)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="244"/>
        <source>Skip seeds with more than INT occurrences </source>
        <translation>Skip seeds with more than INT occurrences </translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="266"/>
        <source>Drop chain threshold (-D)</source>
        <translation>Drop chain threshold (-D)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="282"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain</source>
        <translation>Drop chains shorter than FLOAT fraction of the longest overlapping chain</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="308"/>
        <source>Index algorithm (-a)</source>
        <translation>Index algorithm (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="325"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="330"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="335"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="340"/>
        <source>is</source>
        <translation>is</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="599"/>
        <source>Rounds of mate rescues (-m)</source>
        <translation>Rounds of mate rescues (-m)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="615"/>
        <source>Perform at most INT rounds of mate rescues for each read</source>
        <translation>Perform at most INT rounds of mate rescues for each read</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="348"/>
        <source>Skip mate rescue</source>
        <translation>Skip mate rescue</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="351"/>
        <source>Skip mate rescue (-S)</source>
        <translation>Skip mate rescue (-S)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="358"/>
        <source>Skip pairing; mate rescue performed unless -S also in use</source>
        <translation>Skip pairing; mate rescue performed unless -S also in use</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="361"/>
        <source>Skip pairing (-P)</source>
        <translation>Skip pairing (-P)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="633"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
</context>
<context>
    <name>BwaSettings</name>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="135"/>
        <source>  Best hits (-R)</source>
        <translation>Best hits (-R)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="336"/>
        <source>Max gap extensions (-e)</source>
        <translation>Max gap extensions (-e)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="346"/>
        <source>Indel offset (-i)</source>
        <translation>Indel offset (-i)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="263"/>
        <source>Max long deletion extensions (-d)</source>
        <translation>Max long deletion extensions (-d)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="194"/>
        <source>Seed length (-l)</source>
        <translation>Seed length (-l)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="482"/>
        <source>Max seed differences (-k)</source>
        <translation>Max seed differences (-k)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="283"/>
        <source>Max queue entries (-m)</source>
        <translation>Max queue entries (-m)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="356"/>
        <source>Enable long gaps</source>
        <translation>Enable long gaps</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="171"/>
        <source>Missing prob (-n)</source>
        <translation>Missing prob (-n)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="154"/>
        <source>Max #diff (-n)</source>
        <translation>Max #diff (-n)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="42"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation>Long-scaled gap penalty for long deletions (-L)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="241"/>
        <source>Advanced</source>
        <translation>Advanced</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="395"/>
        <source>Gap open penalty (-O)</source>
        <translation>Gap open penalty (-O)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="303"/>
        <source>Barcode length (-B)</source>
        <translation>Barcode length (-B)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="435"/>
        <source>Quality threshold (-q)</source>
        <translation>Quality threshold (-q)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="375"/>
        <source>Mismatch penalty (-M)</source>
        <translation>Mismatch penalty (-M)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="415"/>
        <source>Gap extension penalty (-E)</source>
        <translation>Gap extension penalty (-E)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="462"/>
        <source>Threads (-t)</source>
        <translation>Threads (-t)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="49"/>
        <source>Non-iterative mode (-N)</source>
        <translation>Non-iterative mode (-N)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="72"/>
        <source>Max gap opens (-o)</source>
        <translation>Max gap opens (-o)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Base Options</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="85"/>
        <source>Index algorithm (-a)</source>
        <translation>Index algorithm (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="102"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="107"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="112"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="117"/>
        <source>is</source>
        <translation>is</translation>
    </message>
</context>
<context>
    <name>BwaSwSettings</name>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="14"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Base Options</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="288"/>
        <source>Score threshold (divided by match score) (-T)</source>
        <translation>Score threshold (divided by match score) (-T)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="233"/>
        <source>Number of threads (-t)</source>
        <translation>Number of threads (-t)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="334"/>
        <source>Number of seeds to start rev alginment (-N)</source>
        <translation>Number of seeds to start rev alginment (-N)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="311"/>
        <source>Z-best (-z)</source>
        <translation>Z-best (-z)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="393"/>
        <source>Prefer hard clipping in SAM output (-H)</source>
        <translation>Prefer hard clipping in SAM output (-H)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="256"/>
        <source>Size of chunk of reads (-s)</source>
        <translation>Size of chunk of reads (-s)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="90"/>
        <source>Score for a match (-a)</source>
        <translation>Score for a match (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="159"/>
        <source>Gap extention penalty (-r)</source>
        <translation>Gap extention penalty (-r)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="113"/>
        <source>Mismatch penalty (-b)</source>
        <translation>Mismatch penalty (-b)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="64"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="69"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="74"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="79"/>
        <source>is</source>
        <translation>is</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="136"/>
        <source>Gap open penalty (-q)</source>
        <translation>Gap open penalty (-q)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="182"/>
        <source>Band width (-w)</source>
        <translation>Band width (-w)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="208"/>
        <source>Index algorithm (-a)</source>
        <translation>Index algorithm (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="380"/>
        <source>Mask level (-c)</source>
        <translation>Mask level (-c)</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="347"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Number of seed to trigger reverse alignment&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Number of seed to trigger reverse alignment&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="405"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
</context>
<context>
    <name>CAP3SupportDialog</name>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="20"/>
        <source>Contig Assembly with CAP3</source>
        <translation>Contig Assembly with CAP3</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="30"/>
        <source>Base</source>
        <translation>Base</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="36"/>
        <source>Input files (long DNA reads to assembly)</source>
        <translation>Input files (long DNA reads to assembly)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="45"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Sequences in FASTA or FASTQ formats are supported. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;The quality scores for FASTA sequences can be provided in an additional file. The file must be located in the same folder as the original sequences and have the same name as FASTA file, but another extention: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Sequences in FASTA or FASTQ formats are supported. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;The quality scores for FASTA sequences can be provided in an additional file. The file must be located in the same folder as the original sequences and have the same name as FASTA file, but another extention: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="57"/>
        <source>Add</source>
        <translation>Add</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="64"/>
        <source>Remove</source>
        <translation>Remove</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="71"/>
        <source>Remove All</source>
        <translation>Remove All</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="93"/>
        <source>Result contig</source>
        <translation>Result contig</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="102"/>
        <source>Path to result contig alignment in ACE format.</source>
        <translation>Path to result contig alignment in ACE format.</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="128"/>
        <source>Clipping for poor regions</source>
        <translation>Clipping for poor regions</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="134"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation>Base quality cutoff for clipping (-c)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="151"/>
        <source>Clipping range (-y)</source>
        <translation>Clipping range (-y)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="174"/>
        <source>Length and percent identity of an overlap</source>
        <translation>Length and percent identity of an overlap</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="180"/>
        <source>Overlap length cutoff (-o)</source>
        <translation>Overlap length cutoff (-o)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="200"/>
        <source>Overlap percent identity cutoff (-p)</source>
        <translation>Overlap percent identity cutoff (-p)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="223"/>
        <source>Quality difference score of an overlap</source>
        <translation>Quality difference score of an overlap</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="229"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation>Base quality cutoff for differences (-b)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="259"/>
        <source>Max qscore sum at differences (-d)</source>
        <translation>Max qscore sum at differences (-d)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="269"/>
        <source>Other parameters</source>
        <translation>Other parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="275"/>
        <source>Max number of word matches (-t)</source>
        <translation>Max number of word matches (-t)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="295"/>
        <source>Band expansion size (-a)</source>
        <translation>Band expansion size (-a)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="318"/>
        <source>Max gap length in any overlap (-f)</source>
        <translation>Max gap length in any overlap (-f)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="338"/>
        <source>Assembly reverse reads (-r)</source>
        <translation>Assembly reverse reads (-r)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="364"/>
        <source>Similarity score of an overlap</source>
        <translation>Similarity score of an overlap</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="370"/>
        <source>Mismatch score factor (-n)</source>
        <translation>Mismatch score factor (-n)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="390"/>
        <source>Overlap similarity score cutoff (-s)</source>
        <translation>Overlap similarity score cutoff (-s)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="410"/>
        <source>Gap penalty factor (-g)</source>
        <translation>Gap penalty factor (-g)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="417"/>
        <source>Match score factor (-m)</source>
        <translation>Match score factor (-m)</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="112"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="122"/>
        <source>Advanced</source>
        <translation>Advanced</translation>
    </message>
</context>
<context>
    <name>ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="26"/>
        <source>Align with Clustal Omega</source>
        <translation>Align with Clustal Omega</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="42"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="52"/>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="69"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="59"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="81"/>
        <source>Iteration</source>
        <translation>Iteration</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="91"/>
        <source>Number of iterations</source>
        <translation>Number of iterations</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="137"/>
        <source>Max number guidetree iterations</source>
        <translation>Max number guidetree iterations</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="180"/>
        <source>Max number of HMM iterations</source>
        <translation>Max number of HMM iterations</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="217"/>
        <source>Miscellaneous</source>
        <translation>Miscellaneous</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="227"/>
        <source>Number of CPUs being used</source>
        <translation>Number of CPUs being used</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="265"/>
        <source>Set options automatically</source>
        <translation>Set options automatically</translation>
    </message>
</context>
<context>
    <name>ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="26"/>
        <source>Align with ClustalW</source>
        <translation>Align with ClustalW</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="42"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="52"/>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="76"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="59"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="88"/>
        <source>Advanced options</source>
        <translation>Advanced options</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="98"/>
        <source>Gap opening penalty</source>
        <translation>Gap opening penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="141"/>
        <source>Gap extension penalty</source>
        <translation>Gap extension penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="184"/>
        <source>Weight matrix</source>
        <translation>Weight matrix</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="214"/>
        <source>IUB</source>
        <translation>IUB</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="219"/>
        <source>CLUSTALW</source>
        <translation>CLUSTALW</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="289"/>
        <source>Max iterations</source>
        <translation>Max iterations</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="261"/>
        <source>NONE</source>
        <translation>NONE</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="266"/>
        <source>TREE</source>
        <translation>TREE</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="271"/>
        <source>ALIGNMENT</source>
        <translation>ALIGNMENT</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="231"/>
        <source>Iteration type</source>
        <translation>Iteration type</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="335"/>
        <source>Out sequences order</source>
        <translation>Out sequences order</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="365"/>
        <source>Input</source>
        <translation>Input</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="370"/>
        <source>Aligned</source>
        <translation>Aligned</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="388"/>
        <source>Protein gap parameters</source>
        <translation>Protein gap parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="398"/>
        <source>Gap separation distance</source>
        <translation>Gap separation distance</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="456"/>
        <source>Residue-specific gaps off</source>
        <translation>Residue-specific gaps off</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="442"/>
        <source>Hydrophilic gaps off</source>
        <translation>Hydrophilic gaps off</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="449"/>
        <source>No end gap separation penalty</source>
        <translation>No end gap separation penalty</translation>
    </message>
</context>
<context>
    <name>ETSSettingsWidget</name>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="26"/>
        <source>Resource settings</source>
        <translation>Resource settings</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="40"/>
        <source>You can select path for</source>
        <translation>You can select path for</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="47"/>
        <source>Check lastest version of external tools package</source>
        <translation>Check latest version of external tools package</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="50"/>
        <source>&lt;a href=&quot;http://ugene.unipro.ru/external.html&quot;&gt;external tools package&lt;/a&gt;</source>
        <translation>&lt;a href=&quot;http://ugene.unipro.ru/external.html&quot;&gt;external tools package&lt;/a&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="70"/>
        <source>Browse...</source>
        <translation>Browse...</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="86"/>
        <source>Supported tools</source>
        <translation>Supported tools</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="102"/>
        <source>Name</source>
        <translation>Name</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="107"/>
        <source>Path</source>
        <translation>Path</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="142"/>
        <source>Additional information</source>
        <translation>Additional information</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="177"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Select an external tool to view more information about it.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Select an external tool to view more information about it.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="197"/>
        <source>Binary path:</source>
        <translation>Binary path:</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="207"/>
        <source>Version info</source>
        <translation>Version info</translation>
    </message>
</context>
<context>
    <name>FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="20"/>
        <source>Format Database</source>
        <translation>Format Database</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="29"/>
        <source>Input data</source>
        <translation>Input data</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="35"/>
        <source>Select input file(s) for formatting database:</source>
        <translation>Select input file(s) for formatting database:</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="53"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="84"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="190"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="62"/>
        <source>Or select directory with input files:</source>
        <translation>Or select directory with input files:</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="96"/>
        <source>File filter</source>
        <translation>File filter</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="104"/>
        <source>Include files filter</source>
        <translation>Include files filter</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="114"/>
        <source>*.fa;*.fasta</source>
        <translation>*.fa;*.fasta</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="121"/>
        <source>Exclude files filter</source>
        <translation>Exclude files filter</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="131"/>
        <source>*.pal</source>
        <translation>*.pal</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="143"/>
        <source>Type of file(s)</source>
        <translation>Type of file(s)</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="149"/>
        <source>protein</source>
        <translation>protein</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="159"/>
        <source>nucleotide</source>
        <translation>nucleotide</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="172"/>
        <source>Output settings</source>
        <translation>Output settings</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="178"/>
        <source>Select the path to save the database into:</source>
        <translation>Select the path to save the database into:</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="201"/>
        <source>Base name for BLAST files:</source>
        <translation>Base name for BLAST files:</translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="211"/>
        <source>Title for database file:</source>
        <translation>Title for database file:</translation>
    </message>
</context>
<context>
    <name>IntegralBusPort</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="59"/>
        <source>One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="64"/>
        <source>Only one of these slots must be binded: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>Only one of these slots must be binded: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="235"/>
        <location filename="../src/spades/SpadesWorker.cpp" line="189"/>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="224"/>
        <source>The slot must be not empty: &apos;%1&apos;</source>
        <translation>The slot must be not empty: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="781"/>
        <source>Error! One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>Error! One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="787"/>
        <source>Error! You can not bind one of sequences slots and one of url slots simultaneously</source>
        <translation>Error! You can not bind one of sequences slots and one of url slots simultaneously</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="56"/>
        <source>&apos;%1&apos; slot must be not binded</source>
        <translation>&apos;%1&apos; slot must be not binded</translation>
    </message>
</context>
<context>
    <name>MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="20"/>
        <source>Align with MAFFT</source>
        <translation>Align with MAFFT</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation>Advanced options</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="85"/>
        <source>Gap opening penalty</source>
        <translation>Gap opening penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="131"/>
        <source>Offset (works like gap extension penalty)</source>
        <translation>Offset (works like gap extension penalty)</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="174"/>
        <source>Maximum number of iterative refinement</source>
        <translation>Maximum number of iterative refinement</translation>
    </message>
</context>
<context>
    <name>MrBayesDialog</name>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="59"/>
        <source>MrBayes Settings</source>
        <translation>MrBayes Settings</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="86"/>
        <source>Model</source>
        <translation>Model</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="150"/>
        <source>ModelType</source>
        <translation>ModelType</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="102"/>
        <source>Rate</source>
        <translation>Rate</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="95"/>
        <source>Gamma</source>
        <translation>Gamma</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="173"/>
        <source>MCMC</source>
        <translation>MCMC</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="191"/>
        <source>Chain length</source>
        <translation>Chain length</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="227"/>
        <source>Subsampling frequence</source>
        <translation>Subsampling frequence</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="250"/>
        <source>Burn-in length</source>
        <translation>Burn-in length</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="342"/>
        <source>Heated chains</source>
        <translation>Heated chains</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="335"/>
        <source>Heated chain temp</source>
        <translation>Heated chain temp</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="220"/>
        <source>Random seed</source>
        <translation>Random seed</translation>
    </message>
</context>
<context>
    <name>PhyMLDialog</name>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="41"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="56"/>
        <source>Substitution model</source>
        <translation>Substitution model</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="73"/>
        <source>Equilibrium frequencies</source>
        <translation>Equilibrium frequencies</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="104"/>
        <source>optimized</source>
        <translation>optimized</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="111"/>
        <source>empirical</source>
        <translation>empirical</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="124"/>
        <source>Transition / transversion ratio </source>
        <translation>Transition / transversion ratio </translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="158"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="222"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="354"/>
        <source>Check to set user value. 
Uncheck to get the maximum likelihood estimate.</source>
        <translation>Check to set user value. 
Uncheck to get the maximum likelihood estimate.</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="169"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="233"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="301"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="365"/>
        <source>estimated </source>
        <translation>estimated </translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="188"/>
        <source>Proportion of invariable sites</source>
        <translation>Proportion of invariable sites</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="255"/>
        <source>Number of substitution rate categories</source>
        <translation>Number of substitution rate categories</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="320"/>
        <source>Gamma shape parameter</source>
        <translation>Gamma shape parameter</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="387"/>
        <source>Branch Support</source>
        <translation>Branch Support</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="408"/>
        <source>Use fast likelihood-based method</source>
        <translation>Use fast likelihood-based method</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="427"/>
        <source>Perform bootstrap</source>
        <translation>Perform bootstrap</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="443"/>
        <source>It is the number of bootstrap replicates.</source>
        <translation>It is the number of bootstrap replicates.</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="456"/>
        <source>Tree Searching</source>
        <translation>Tree Searching</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="536"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="546"/>
        <source>Type of tree improvement</source>
        <translation>Type of tree improvement</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="589"/>
        <source>Set number of random starting tree</source>
        <translation>Set number of random starting tree</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="566"/>
        <source>Optimise topology</source>
        <translation>Optimise topology</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="576"/>
        <source>Optimise branch lengths</source>
        <translation>Optimise branch lengths</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="122"/>
        <source>Path for temporary files</source>
        <translation>Path for temporary files</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="123"/>
        <source>Path for temporary files not selected.</source>
        <translation>Path for temporary files not selected.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="124"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="176"/>
        <source>Genome sequence slot is not binded : &apos;%1&apos;</source>
        <translation>Genome sequence slot is not binded : &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="179"/>
        <source>Transcripts slot is not binded : &apos;%1&apos;</source>
        <translation>Transcripts slot is not binded : &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="192"/>
        <source>Input transcripts</source>
        <translation>Input transcripts</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url</source>
        <translation>Genomic sequence url</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url [FASTA]</source>
        <translation>Genomic sequence url [FASTA]</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url</source>
        <translation>Transcripts url</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url [GTF]</source>
        <translation>Transcripts url [GTF]</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="200"/>
        <source>Extracted sequences url</source>
        <translation>Extracted sequences url</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>Output sequences</source>
        <translation>Output sequences</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>The url to the output file with the extracted sequences.</source>
        <translation>The url to the output file with the extracted sequences.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="218"/>
        <source>Extract Transcript Sequences with gffread</source>
        <translation>Extract Transcript Sequences with gffread</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="219"/>
        <source>Extract transcript sequences from the genomic sequence(s) with gffread.</source>
        <translation>Extract transcript sequences from the genomic sequence(s) with gffread.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="76"/>
        <source>Installed</source>
        <translation>Installed</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="77"/>
        <source>Not installed</source>
        <translation>Not installed</translation>
    </message>
    <message>
        <location filename="../src/RnaSeqCommon.cpp" line="62"/>
        <source>Unexpected value &apos;%1&apos; of the library type</source>
        <translation>Unexpected value &apos;%1&apos; of the library type</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="813"/>
        <source>%1 tool&apos;s version is undefined, this may cause some compatibility issues</source>
        <translation>%1 tool&apos;s version is undefined, this may cause some compatibility issues</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="822"/>
        <source>Bowtie and TopHat tools have incompatible versions. Your TopHat&apos;s version is %1, Bowtie&apos;s one is %2. The following are considered to be compatible: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</source>
        <translation>Bowtie and TopHat tools have incompatible versions. Your TopHat&apos;s version is %1, Bowtie&apos;s one is %2. The following are considered to be compatible: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="858"/>
        <source>At least two samples are required</source>
        <translation>At least two samples are required</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="865"/>
        <source>Duplicate sample name: </source>
        <translation>Duplicate sample name: </translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="870"/>
        <source>No datasets in the sample: </source>
        <translation>No datasets in the sample: </translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="40"/>
        <source>NOTE: &quot;is&quot; index algorithm is not supposed to work with reference sequences having size larger than 2 GB. In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;bwtsw&quot;</source>
        <translation>NOTE: &quot;is&quot; index algorithm is not supposed to work with reference sequences having size larger than 2 GB. In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;bwtsw&quot;</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="44"/>
        <source>NOTE: &quot;bwtsw&quot; index algorithm is not supposed to work with reference sequences having size smaller than 10 MB. In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;is&quot;</source>
        <translation>NOTE: &quot;bwtsw&quot; index algorithm is not supposed to work with reference sequences having size smaller than 10 MB. In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;is&quot;</translation>
    </message>
</context>
<context>
    <name>SpadesSettings</name>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Form</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Base Options</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="45"/>
        <source>Running mode</source>
        <translation>Running mode</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="59"/>
        <source>Error Correction and Assembly</source>
        <translation>Error Correction and Assembly</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="64"/>
        <source>Assembly only</source>
        <translation>Assembly only</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="69"/>
        <source>Error correction only</source>
        <translation>Error correction only</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="80"/>
        <source>k-mer sizes (-k)</source>
        <translation>k-mer sizes (-k)</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="99"/>
        <source>Dataset type</source>
        <translation>Dataset type</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="112"/>
        <source>auto</source>
        <translation>auto</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="120"/>
        <source>Multi Cell</source>
        <translation>Multi Cell</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="125"/>
        <source>Single Cell</source>
        <translation>Single Cell</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="142"/>
        <source>Memory limit GB (-m)</source>
        <translation>Memory limit GB (-m)</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="158"/>
        <location filename="../src/ui/SpadesSettings.ui" line="190"/>
        <source>Number of threads</source>
        <translation>Number of threads</translation>
    </message>
    <message>
        <location filename="../src/ui/SpadesSettings.ui" line="174"/>
        <source>Number of threads (-t)</source>
        <translation>Number of threads (-t)</translation>
    </message>
</context>
<context>
    <name>TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="20"/>
        <source>Align with T-Coffee</source>
        <translation>Align with T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation>Advanced options</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="83"/>
        <source>Gap opening penalty</source>
        <translation>Gap opening penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="122"/>
        <source>Gap extension penalty</source>
        <translation>Gap extension penalty</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="158"/>
        <source>Number of iterations</source>
        <translation>Number of iterations</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsIntersectAnnotationsByEntityTask</name>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="114"/>
        <source>Intersect annotations task</source>
        <translation>Intersect annotations task</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="173"/>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="193"/>
        <source>Failed to get IOAdapterFactory</source>
        <translation>Failed to get IOAdapterFactory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="180"/>
        <source>Result document is NULL</source>
        <translation>Result document is NULL</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="190"/>
        <source>Failed to get BED format</source>
        <translation>Failed to get BED format</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsIntersectTask</name>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="64"/>
        <source>BedtoolsIntersect task</source>
        <translation>BedtoolsIntersect task</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="69"/>
        <source>No input A URL</source>
        <translation>No input A URL</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="70"/>
        <source>No input B URL(s)</source>
        <translation>No input B URL(s)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="71"/>
        <source>Output URL is not set</source>
        <translation>Output URL is not set</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsSupport</name>
    <message>
        <location filename="../src/bedtools/BedtoolsSupport.cpp" line="52"/>
        <source>&lt;i&gt;Bedtools&lt;/i&gt;: flexible tools for genome arithmetic and DNA sequence analysis.</source>
        <translation>&lt;i&gt;Bedtools&lt;/i&gt;: flexible tools for genome arithmetic and DNA sequence analysis.</translation>
    </message>
</context>
<context>
    <name>U2::BigWigSupport</name>
    <message>
        <location filename="../src/bigWigTools/BigWigSupport.cpp" line="50"/>
        <source>&lt;i&gt;bedGraphToBigWig&lt;/i&gt;: converts bedGrapth to bigWig.</source>
        <translation>&lt;i&gt;bedGraphToBigWig&lt;/i&gt;: converts bedGrapth to bigWig.</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupport</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="71"/>
        <source>The &lt;i&gt;blastall&lt;/i&gt; is the old program developed and distributed by the NCBI for running BLAST searches. The NCBI recommends that people start using the programs of the BLAST+ package instead.</source>
        <translation>The &lt;i&gt;blastall&lt;/i&gt; is the old program developed and distributed by the NCBI for running BLAST searches. The NCBI recommends that people start using the programs of the BLAST+ package instead.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="85"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation>Path for BLAST %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="86"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportContext</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="135"/>
        <source>Query with BLAST...</source>
        <translation>Query with BLAST...</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="155"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation>Path for BLAST %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="156"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="204"/>
        <source>Sequence object is NULL</source>
        <translation>Sequence object is NULL</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportMultiTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="655"/>
        <source>Source file</source>
        <translation>Source file</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="656"/>
        <source>Used databse</source>
        <translation>Used database</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="658"/>
        <source>No any results found</source>
        <translation>No any results found</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="98"/>
        <source>Database path contains space characters.</source>
        <translation>Database path contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="103"/>
        <source>Database name contains space characters.</source>
        <translation>Database name contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="116"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Wrong parameters for creating annotations</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="130"/>
        <source>Error</source>
        <translation>Error</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="130"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Cannot create an annotation object. Please check settings</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="76"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation>Subdirectory for temporary files exists. Can not remove this directory.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="81"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="94"/>
        <source>Can not create fake NCBI ini file</source>
        <translation>Can not create fake NCBI ini file</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="205"/>
        <source>Output file not found</source>
        <translation>Output file not found</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="207"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="253"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="288"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation>Incorrect number of fields in line: %1</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="293"/>
        <source>Can&apos;t get location. Start position is absent in [%1]</source>
        <translation>Can&apos;t get location. Start position is absent in [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="298"/>
        <source>Can&apos;t get location. End position is absent in [%1]</source>
        <translation>Can&apos;t get location. End position is absent in [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="317"/>
        <source>Can&apos;t get hit start location from [%1]</source>
        <translation>Can&apos;t get hit start location from [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="322"/>
        <source>Can&apos;t get hit end location from [%1]</source>
        <translation>Can&apos;t get hit end location from [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="352"/>
        <source>Can&apos;t get align length from [%1]</source>
        <translation>Can&apos;t get align length from [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="357"/>
        <source>Can&apos;t get gaps from [%1]</source>
        <translation>Can&apos;t get gaps from [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="362"/>
        <source>Can&apos;t get identity from [%1]</source>
        <translation>Can&apos;t get identity from [%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="391"/>
        <source>Can&apos;t open output file</source>
        <translation>Can&apos;t open output file</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="453"/>
        <source>Can&apos;t get Hsp_query-from element</source>
        <translation>Can&apos;t get Hsp_query-from element</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="460"/>
        <source>Can&apos;t get Hsp_query-to element</source>
        <translation>Can&apos;t get Hsp_query-to element</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="483"/>
        <source>Can&apos;t get location. Hsp_query-frame[%1]</source>
        <translation>Can&apos;t get location. Hsp_query-frame[%1]</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="493"/>
        <source>Can&apos;t get identity</source>
        <translation>Can&apos;t get identity</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="501"/>
        <source>Can&apos;t get gaps</source>
        <translation>Can&apos;t get gaps</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="509"/>
        <source>Can&apos;t get align length</source>
        <translation>Can&apos;t get align length</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="522"/>
        <source>Can&apos;t evaluate location</source>
        <translation>Can&apos;t evaluate location</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="171"/>
        <source>Select input file</source>
        <translation>Select input file</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="191"/>
        <source>Database path contains space characters.</source>
        <translation>Database path contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="196"/>
        <source>Database name contains space characters.</source>
        <translation>Database name contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="231"/>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation>This file has the incompatible format for the BLAST+ search.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="276"/>
        <source>This file does not contain any sequence.</source>
        <translation>This file does not contain any sequence.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="231"/>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="276"/>
        <source>Wrong input file</source>
        <translation>Wrong input file</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="323"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Wrong parameters for creating annotations</translation>
    </message>
</context>
<context>
    <name>U2::BlastDBCmdDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="42"/>
        <source>Fetch</source>
        <translation>Fetch</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="43"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="74"/>
        <source>Select a database file</source>
        <translation>Select a database file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="88"/>
        <source>Set a result FASTA file name</source>
        <translation>Set a result FASTA file name</translation>
    </message>
</context>
<context>
    <name>U2::BlastDbCmdSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="64"/>
        <source>The &lt;i&gt;BlastDBCmd&lt;/i&gt; fetches protein or nucleotide sequences from BLAST+ database based on a query.</source>
        <translation>The &lt;i&gt;BlastDBCmd&lt;/i&gt; fetches protein or nucleotide sequences from BLAST+ database based on a query.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="77"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Path for BLAST+ %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="179"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation>Path for BLAST+ tools not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="180"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportCommonTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="89"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation>Subdirectory for temporary files exists. Can not remove this directory.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="94"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="107"/>
        <source>Can not create fake NCBI ini file</source>
        <translation>Can not create fake NCBI ini file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="151"/>
        <source>Output file not found</source>
        <translation>Output file not found</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="153"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="182"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="207"/>
        <source>There were no hits found for your BLAST search.</source>
        <translation>There were no hits found for your BLAST search.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="199"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="240"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation>Incorrect number of fields in line: %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="245"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="250"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="405"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="412"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="429"/>
        <source>Can&apos;t get location</source>
        <translation>Can&apos;t get location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="262"/>
        <source>Can&apos;t evaluate location</source>
        <translation>Can&apos;t evaluate location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="268"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="273"/>
        <source>Can&apos;t get hit location</source>
        <translation>Can&apos;t get hit location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="289"/>
        <source>Can&apos;t evaluate hit location</source>
        <translation>Can&apos;t evaluate hit location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="303"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="455"/>
        <source>Can&apos;t get align length</source>
        <translation>Can&apos;t get align length</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="308"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="447"/>
        <source>Can&apos;t get gaps</source>
        <translation>Can&apos;t get gaps</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="313"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="439"/>
        <source>Can&apos;t get identity</source>
        <translation>Can&apos;t get identity</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="342"/>
        <source>Can&apos;t open output file</source>
        <translation>Can&apos;t open output file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="346"/>
        <source>Can&apos;t read output file</source>
        <translation>Can&apos;t read output file</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportContext</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="234"/>
        <source>Fetch sequences by &apos;id&apos;</source>
        <translation>Fetch sequences by &apos;id&apos;</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="245"/>
        <source>Query with BLAST+...</source>
        <translation>Query with BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="289"/>
        <source>Fetch sequences from local BLAST database</source>
        <translation>Fetch sequences from local BLAST database</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="294"/>
        <source>Fetch sequences by &apos;id&apos; %1</source>
        <translation>Fetch sequences by &apos;id&apos; %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="312"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation>Path for BLAST+ tools not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="313"/>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="397"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="369"/>
        <source>Sequence object is NULL</source>
        <translation>Sequence object is NULL</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="396"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Path for BLAST+ %1 tool not selected.</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportMultiTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="609"/>
        <source>Source file</source>
        <translation>Source file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="610"/>
        <source>Used databse</source>
        <translation>Used database</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="612"/>
        <source>No any results found</source>
        <translation>No any results found</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="104"/>
        <source>Database path contains space characters.</source>
        <translation>Database path contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="109"/>
        <source>Database name contains space characters.</source>
        <translation>Database name contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="124"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="159"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="193"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Wrong parameters for creating annotations</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="207"/>
        <source>Error</source>
        <translation>Error</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="207"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Cannot create an annotation object. Please check settings</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="245"/>
        <source>Select input file</source>
        <translation>Select input file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="319"/>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="365"/>
        <source>Wrong input file</source>
        <translation>Wrong input file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="272"/>
        <source>Database path contains space characters.</source>
        <translation>Database path contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="284"/>
        <source>Database name contains space characters.</source>
        <translation>Database name contains space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="319"/>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation>This file has the incompatible format for the BLAST+ search.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="365"/>
        <source>This file does not contain sequences.</source>
        <translation>This file does not contain sequences.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="417"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="452"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="486"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Wrong parameters for creating annotations</translation>
    </message>
</context>
<context>
    <name>U2::BlastRunCommonDialog</name>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="66"/>
        <source>Restore to default</source>
        <translation>Restore to default</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="67"/>
        <source>Search</source>
        <translation>Search</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="68"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="297"/>
        <source>Select a database file</source>
        <translation>Select a database file</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="309"/>
        <source>Direct nucleotide alignment</source>
        <translation>Direct nucleotide alignment</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="312"/>
        <source>Direct protein alignment</source>
        <translation>Direct protein alignment</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="316"/>
        <source>Direct protein alignment (on GPU)</source>
        <translation>Direct protein alignment (on GPU)</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="320"/>
        <source>Protein alignment, input nucleotide is translated input protein before the search</source>
        <translation>Protein alignment, input nucleotide is translated input protein before the search</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="324"/>
        <source>Protein alignment, nucleotide database is translated input protein before the search</source>
        <translation>Protein alignment, nucleotide database is translated input protein before the search</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="328"/>
        <source>Protein alignment, both input query and database are translated before the search</source>
        <translation>Protein alignment, both input query and database are translated before the search</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2AlignTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="147"/>
        <source>Unknown short reads format %1</source>
        <translation>Unknown short reads format %1</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2BuildIndexTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="46"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation>Reference file &quot;%1&quot; does not exist</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2Support</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="54"/>
        <source>&lt;i&gt;Bowtie 2 aligner&lt;/i&gt; takes a Bowtie 2 index and a set of sequencing read files and outputs a set of alignments.</source>
        <translation>&lt;i&gt;Bowtie 2 aligner&lt;/i&gt; takes a Bowtie 2 index and a set of sequencing read files and outputs a set of alignments.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="70"/>
        <source>&lt;i&gt;Bowtie 2 build indexer&lt;/i&gt;  builds a Bowtie index from a set of DNA sequences. It outputs a set of 6 files with suffixes .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, and .rev.2.bt2. These files together constitute the index: they are all that is needed to align reads to that reference. The original sequence files are no longer used by &lt;i&gt;Bowtie 2&lt;/i&gt; once the index is built.</source>
        <translation>&lt;i&gt;Bowtie 2 build indexer&lt;/i&gt;  builds a Bowtie index from a set of DNA sequences. It outputs a set of 6 files with suffixes .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, and .rev.2.bt2. These files together constitute the index: they are all that is needed to align reads to that reference. The original sequence files are no longer used by &lt;i&gt;Bowtie 2&lt;/i&gt; once the index is built.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="91"/>
        <source>&lt;i&gt;Bowtie 2 index inspector&lt;/i&gt; extracts information from a Bowtie index about what kind of index it is and what reference sequence were used to build it.</source>
        <translation>&lt;i&gt;Bowtie 2 index inspector&lt;/i&gt; extracts information from a Bowtie index about what kind of index it is and what reference sequence were used to build it.</translation>
    </message>
</context>
<context>
    <name>U2::BowtieAssembleTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="160"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation>Reference index file &quot;%1&quot; does not exist</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="241"/>
        <source>Unknown short reads format %1</source>
        <translation>Unknown short reads format %1</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="315"/>
        <source>There is not enough memory on the computer!</source>
        <translation>There is not enough memory on the computer!</translation>
    </message>
</context>
<context>
    <name>U2::BowtieBuildIndexTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="47"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation>Reference file &quot;%1&quot; does not exist</translation>
    </message>
</context>
<context>
    <name>U2::BowtieSupport</name>
    <message>
        <location filename="../src/bowtie/BowtieSupport.cpp" line="57"/>
        <source>&lt;i&gt;Bowtie&lt;/i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</source>
        <translation>&lt;i&gt;Bowtie&lt;/i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</translation>
    </message>
</context>
<context>
    <name>U2::BwaAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="92"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="99"/>
        <source>In paired-end mode it possible to analyze only 2 read sets using BWA</source>
        <translation>In paired-end mode it possible to analyze only 2 read sets using BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="217"/>
        <source>Log is incomplete</source>
        <translation>Log is incomplete</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="220"/>
        <source>Log is incorrect</source>
        <translation>Log is incorrect</translation>
    </message>
</context>
<context>
    <name>U2::BwaMemAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="242"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="251"/>
        <source>Wrong settings of paired reads. For paired-read alignment by BWA MEM only a single pair of reads is acceptable.</source>
        <translation>Wrong settings of paired reads. For paired-read alignment by BWA MEM only a single pair of reads is acceptable.</translation>
    </message>
</context>
<context>
    <name>U2::BwaMemSettingsWidget</name>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="338"/>
        <source>NOTE: bwa mem accepts reads only in FASTA or FASTQ format. Reads should be compiled into a single file for each mate end.</source>
        <translation>NOTE: bwa mem accepts reads only in FASTA or FASTQ format. Reads should be compiled into a single file for each mate end.</translation>
    </message>
</context>
<context>
    <name>U2::BwaSupport</name>
    <message>
        <location filename="../src/bwa/BwaSupport.cpp" line="46"/>
        <source>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; is an efficient program that aligns relatively short nucleotide sequences against a long reference sequence such as the human genome.</source>
        <translation>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; is an efficient program that aligns relatively short nucleotide sequences against a long reference sequence such as the human genome.</translation>
    </message>
</context>
<context>
    <name>U2::BwaSwAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="338"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="348"/>
        <source>BWA SW can not align paired reads</source>
        <translation>BWA SW can not align paired reads</translation>
    </message>
</context>
<context>
    <name>U2::BwaSwSettingsWidget</name>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="270"/>
        <source>NOTE: bwa-sw performs alignment of long sequencing reads (Sanger or 454). It accepts reads only in FASTA or FASTQ format. Reads should be compiled into single file.</source>
        <translation>NOTE: bwa-sw performs alignment of long sequencing reads (Sanger or 454). It accepts reads only in FASTA or FASTQ format. Reads should be compiled into single file.</translation>
    </message>
</context>
<context>
    <name>U2::BwaTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="479"/>
        <source>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</source>
        <translation>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="485"/>
        <source>Multiple read files are not supported by bwa-mem. Please combine your reads into single FASTA file.</source>
        <translation>Multiple read files are not supported by bwa-mem. Please combine your reads into single FASTA file.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="488"/>
        <source>Please, provide two files with paired reads.</source>
        <translation>Please, provide two files with paired reads.</translation>
    </message>
</context>
<context>
    <name>U2::CAP3Support</name>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="65"/>
        <source>&lt;i&gt;CAP3&lt;/i&gt; is a contig assembly program.                    &lt;br&gt;It allows to assembly long DNA reads (up to 1000 bp).                    &lt;br&gt;Binaries can be downloaded from http://seq.cs.iastate.edu/cap3.html</source>
        <translation>&lt;i&gt;CAP3&lt;/i&gt; is a contig assembly program.                    &lt;br&gt;It allows to assembly long DNA reads (up to 1000 bp).                    &lt;br&gt;Binaries can be downloaded from http://seq.cs.iastate.edu/cap3.html</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="68"/>
        <source>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</source>
        <translation>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="69"/>
        <source>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</source>
        <translation>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="70"/>
        <source>&lt;br&gt;Genome Research, 9: 868-877.</source>
        <translation>&lt;br&gt;Genome Research, 9: 868-877.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="80"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="81"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportDialog</name>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="47"/>
        <source>Run</source>
        <translation>Run</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="89"/>
        <source>List of input files is empty!</source>
        <translation>List of input files is empty!</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="101"/>
        <source>Result contig file name is not set!</source>
        <translation>Result contig file name is not set!</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="107"/>
        <source>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To save under other name press &apos;Cancel&apos; and change name in &apos;Result contig&apos; field.</source>
        <translation>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To save under other name press &apos;Cancel&apos; and change name in &apos;Result contig&apos; field.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="110"/>
        <source>Replace</source>
        <translation>Replace</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="139"/>
        <source>Add Sequences to Assembly</source>
        <translation>Add Sequences to Assembly</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="170"/>
        <source>Set Result Contig File Name</source>
        <translation>Set Result Contig File Name</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="48"/>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="111"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="170"/>
        <source>ACE format (*.ace)</source>
        <translation>ACE format (*.ace)</translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="100"/>
        <source>Output file not found</source>
        <translation>Output file not found</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="102"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="116"/>
        <source>Output file not found: copy from tmp dir failed.</source>
        <translation>Output file not found: copy from tmp dir failed.</translation>
    </message>
</context>
<context>
    <name>U2::CEASSupport</name>
    <message>
        <location filename="../src/ceas/CEASSupport.cpp" line="46"/>
        <source>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - helps to characterize genome-wide protein-DNA interaction patterns from ChIP-chip and ChIP-Seq of both sharp and broad binding factors. It provides statistics on ChIP enrichment at important genome features such as specific chromosome, promoters, gene bodies, or exons, and infers genes most likely to be regulated by a binding factor.</source>
        <translation>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - helps to characterize genome-wide protein-DNA interaction patterns from ChIP-chip and ChIP-Seq of both sharp and broad binding factors. It provides statistics on ChIP enrichment at important genome features such as specific chromosome, promoters, gene bodies, or exons, and infers genes most likely to be regulated by a binding factor.</translation>
    </message>
</context>
<context>
    <name>U2::CEASSupportTask</name>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="202"/>
        <source>CEAS error: Can not find a required output file %1.</source>
        <translation>CEAS error: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="213"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupport</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="67"/>
        <source>&lt;i&gt;Clustal Omega&lt;/i&gt; is a free sequence alignment software for proteins.</source>
        <translation>&lt;i&gt;Clustal Omega&lt;/i&gt; is a free sequence alignment software for proteins.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="77"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportContext</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="133"/>
        <source>Align with ClustalO...</source>
        <translation>Align with ClustalO...</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="157"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="158"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="55"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="56"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="97"/>
        <source>ClustalO alignment started</source>
        <translation>ClustalO alignment started</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="118"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="127"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation>Subdirectory for temporary files exists. Can not remove this directory.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="132"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="179"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="181"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="189"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="275"/>
        <source>Failed to apply the result of ClustalO: alignment object is not available!</source>
        <translation>Failed to apply the result of ClustalO: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="280"/>
        <source>ClustalO alignment successfully finished</source>
        <translation>ClustalO alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="293"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="91"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="92"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="108"/>
        <source>Open an alignment file</source>
        <translation>Open an alignment file</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="118"/>
        <source>Save an multiple alignment file</source>
        <translation>Save an multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="143"/>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="146"/>
        <source>Kalign with Align</source>
        <translation>Align with Kalign</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="144"/>
        <source>Input file is not set!</source>
        <translation>Input file is not set!</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="147"/>
        <source>Output file is not set!</source>
        <translation>Output file is not set!</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="330"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupport</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="67"/>
        <source>&lt;i&gt;ClustalW&lt;/i&gt; is a free sequence alignment software for DNA or proteins.</source>
        <translation>&lt;i&gt;ClustalW&lt;/i&gt; is a free sequence alignment software for DNA or proteins.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="77"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportContext</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="133"/>
        <source>Align with ClustalW...</source>
        <translation>Align with ClustalW...</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="158"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="159"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="52"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="53"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="105"/>
        <source>Unsupported alphabet: %1</source>
        <translation>Unsupported alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="108"/>
        <source>ClustalW alignment started</source>
        <translation>ClustalW alignment started</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="130"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="204"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="206"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="214"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="292"/>
        <source>Failed to apply the result of ClustalW: alignment object is not available!</source>
        <translation>Failed to apply the result of ClustalW: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="297"/>
        <source>ClustalW alignment successfully finished</source>
        <translation>ClustalW alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="310"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="131"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="132"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="154"/>
        <source>Open an alignment file</source>
        <translation>Open an alignment file</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="164"/>
        <source>Save an multiple alignment file</source>
        <translation>Save an multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="215"/>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="218"/>
        <source>Kalign with Align</source>
        <translation>Align with Kalign</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="216"/>
        <source>Input file is not set!</source>
        <translation>Input file is not set!</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="219"/>
        <source>Output file is not set!</source>
        <translation>Output file is not set!</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="346"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::ConductGOSupport</name>
    <message>
        <location filename="../src/conduct_go/ConductGOSupport.cpp" line="45"/>
        <source>&lt;i&gt;Conduct GO&lt;/i&gt; - For a list of input genes, this tool uses R/BioC packages (GO, GOstats) to identify over represented GO terms.</source>
        <translation>&lt;i&gt;Conduct GO&lt;/i&gt; - For a list of input genes, this tool uses R/BioC packages (GO, GOstats) to identify over represented GO terms.</translation>
    </message>
</context>
<context>
    <name>U2::ConductGOTask</name>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="154"/>
        <source>Conduct GO warning: Can not find a required output file %1.</source>
        <translation>Conduct GO warning: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="165"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotSupport</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotSupport.cpp" line="46"/>
        <source>&lt;i&gt;ConservationPlot&lt;/i&gt; - Draw conservation plot for many bed files.</source>
        <translation>&lt;i&gt;ConservationPlot&lt;/i&gt; - Draw conservation plot for many bed files.</translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotTask</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="128"/>
        <source>An annotation table object wasn&apos;t found in the workflow data storage</source>
        <translation>An annotation table object wasn&apos;t found in the workflow data storage</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="171"/>
        <source>Conservation Plot error: Can not find a required output file %1.</source>
        <translation>Conservation Plot error: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="182"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::CopyFileTask</name>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="101"/>
        <source>Copy file task</source>
        <translation>Copy file task</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="109"/>
        <source>Input file is not set</source>
        <translation>Input file is not set</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="113"/>
        <source>Output file is not set</source>
        <translation>Output file is not set</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="117"/>
        <source>Copy %1 to %2</source>
        <translation>Copy %1 to %2</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="121"/>
        <source>Can not rename existing file &apos;%1&apos;</source>
        <translation>Can not rename existing file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="127"/>
        <source>Error copying file</source>
        <translation>Error copying file</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="130"/>
        <source>File copying finished</source>
        <translation>File copying finished</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="135"/>
        <source>File copying task was finished with an error: %1</source>
        <translation>File copying task was finished with an error: %1</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="137"/>
        <source>File copy was finished. Copy of &apos;%1&apos; is &apos;%2&apos;</source>
        <translation>File copy was finished. Copy of &apos;%1&apos; is &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::CuffdiffSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="42"/>
        <source>Running Cuffdiff task</source>
        <translation>Running Cuffdiff task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="46"/>
        <source>Workflow data storage is NULL</source>
        <translation>Workflow data storage is NULL</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="83"/>
        <source>At least 2 sets of assemblies are required for Cuffdiff</source>
        <translation>At least 2 sets of assemblies are required for Cuffdiff</translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupport</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="54"/>
        <source>&lt;i&gt;Cuffcompare&lt;/i&gt; helps comparing assembled transcripts to a reference annotation, and also tracking transcripts across multiple experiments.</source>
        <translation>&lt;i&gt;Cuffcompare&lt;/i&gt; helps comparing assembled transcripts to a reference annotation, and also tracking transcripts across multiple experiments.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="70"/>
        <source>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;tests for differential expression and regulation in RNA-Seq samples.</source>
        <translation>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;tests for differential expression and regulation in RNA-Seq samples.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="85"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts and estimates their abundances.</source>
        <translation>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts and estimates their abundances.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="101"/>
        <source>&lt;i&gt;Cuffmerge&lt;/i&gt; merges together several assemblies.</source>
        <translation>&lt;i&gt;Cuffmerge&lt;/i&gt; merges together several assemblies.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="114"/>
        <source>&lt;i&gt;Gffread&lt;/i&gt; is used to verify or perform various operations on GFF files.</source>
        <translation>&lt;i&gt;Gffread&lt;/i&gt; is used to verify or perform various operations on GFF files.</translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="55"/>
        <source>Running Cufflinks task</source>
        <translation>Running Cufflinks task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="125"/>
        <source>Unable to get an assembly object.</source>
        <translation>Unable to get an assembly object.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="233"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation>An internal error occurred during getting annotations from a %1 output file!</translation>
    </message>
</context>
<context>
    <name>U2::CuffmergeSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="48"/>
        <source>Running Cuffmerge task</source>
        <translation>Running Cuffmerge task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="53"/>
        <source>Workflow data storage is NULL</source>
        <translation>Workflow data storage is NULL</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="54"/>
        <source>There are no annotations to process</source>
        <translation>There are no annotations to process</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="140"/>
        <source>Can not create a file: %1</source>
        <translation>Can not create a file: %1</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="192"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation>An internal error occurred during getting annotations from a %1 output file!</translation>
    </message>
</context>
<context>
    <name>U2::CutadaptSupport</name>
    <message>
        <location filename="../src/cutadapt/CutadaptSupport.cpp" line="43"/>
        <source>&lt;i&gt;cutadapt&lt;/i&gt; removes adapter sequences from high-throughput sequencing data. This is necessary when the reads are longer than the molecule that is sequenced, such as in microRNA data.</source>
        <translation>&lt;i&gt;cutadapt&lt;/i&gt; removes adapter sequences from high-throughput sequencing data. This is necessary when the reads are longer than the molecule that is sequenced, such as in microRNA data.</translation>
    </message>
</context>
<context>
    <name>U2::ETSProjectViewItemsContoller</name>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="52"/>
        <source>FormatDB...</source>
        <translation>FormatDB...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="53"/>
        <source>BLAST+ make DB...</source>
        <translation>BLAST+ make DB...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="79"/>
        <source>BLAST</source>
        <translation>BLAST</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="96"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation>Path for BLAST %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="99"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Path for BLAST+ %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="101"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolJustValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="51"/>
        <source>Tool&apos;s path is empty</source>
        <translation>Tool&apos;s path is empty</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="66"/>
        <source>Tool&apos;s executable isn&apos;t exists</source>
        <translation>Tool&apos;s executable isn&apos;t exists</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="74"/>
        <source>Scripting tool registry is NULL</source>
        <translation>Scripting tool registry is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="76"/>
        <source>Scripting tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation>Scripting tool &apos;%1&apos; isn&apos;t found in the registry</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="125"/>
        <source>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="156"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchAndValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="347"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="38"/>
        <source>&apos;%1&apos; external tool search task</source>
        <translation>&apos;%1&apos; external tool search task</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="46"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="127"/>
        <source>An external tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation>An external tool &apos;%1&apos; isn&apos;t found in the registry</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="115"/>
        <source>Tool pointer is NULL</source>
        <translation>Tool pointer is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="123"/>
        <source>External tool module hasn&apos;t any dependencies: it hasn&apos;t master tool</source>
        <translation>External tool module hasn&apos;t any dependencies: it hasn&apos;t master tool</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="125"/>
        <source>External tool registry is NULL</source>
        <translation>External tool registry is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="131"/>
        <source>Tool&apos;s executable name is not set</source>
        <translation>Tool&apos;s executable name is not set</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportPlugin</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="177"/>
        <source>Search tools in PATH</source>
        <translation>Search tools in PATH</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="224"/>
        <source>External tool support</source>
        <translation>External tool support</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="224"/>
        <source>Runs other external tools</source>
        <translation>Runs other external tools</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="287"/>
        <source>Align with ClustalW...</source>
        <translation>Align with ClustalW...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="295"/>
        <source>Align with ClustalO...</source>
        <translation>Align with ClustalO...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="303"/>
        <source>Align with MAFFT...</source>
        <translation>Align with MAFFT...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="311"/>
        <source>Align with T-Coffee...</source>
        <translation>Align with T-Coffee...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="464"/>
        <source>Blast format database...</source>
        <translation>BLAST format database...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="468"/>
        <source>BLAST+ make database...</source>
        <translation>BLAST+ make database...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="486"/>
        <source>BLAST+ search...</source>
        <translation>BLAST+ search...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="490"/>
        <source>BLAST+ query database...</source>
        <translation>BLAST+ query database...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="476"/>
        <source>BLAST Search...</source>
        <translation>BLAST Search...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="38"/>
        <source>The subdirectory for temporary files &apos;%1&apos; already exists. Can not remove this directory!</source>
        <translation>The subdirectory for temporary files &apos;%1&apos; already exists. Can not remove this directory!</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="43"/>
        <source>Can not create a subdirectory for temporary files &apos;%1&apos;!</source>
        <translation>Can not create a subdirectory for temporary files &apos;%1&apos;!</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportService</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="653"/>
        <source>External tools support</source>
        <translation>External tools support</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="653"/>
        <source>Provides support to run external tools from UGENE</source>
        <translation>Provides support to run external tools from UGENE</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageController</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="52"/>
        <source>External Tools</source>
        <translation>External Tools</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageWidget</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="469"/>
        <source>Select an external tool to view more information about it.</source>
        <translation>Select an external tool to view more information about it.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="476"/>
        <source>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches. BLAST can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.</source>
        <translation>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches. BLAST can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="482"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="485"/>
        <source>&lt;i&gt;BLAST+&lt;/i&gt; is a new version of the BLAST package from the NCBI.</source>
        <translation>&lt;i&gt;BLAST+&lt;/i&gt; is a new version of the BLAST package from the NCBI.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="496"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. It also estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols. </source>
        <translation>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. It also estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="504"/>
        <source>&lt;i&gt;Bowtie 2&lt;/i&gt; is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. &lt;br/&gt;&lt;br/&gt;It indexes the genome with an FM index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2Gb. &lt;br/&gt;&lt;br/&gt;&lt;i&gt;Bowtie 2&lt;/i&gt; supports gapped, local, and paired-end alignment modes.</source>
        <translation>&lt;i&gt;Bowtie 2&lt;/i&gt; is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. &lt;br/&gt;&lt;br/&gt;It indexes the genome with an FM index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2Gb. &lt;br/&gt;&lt;br/&gt;&lt;i&gt;Bowtie 2&lt;/i&gt; supports gapped, local, and paired-end alignment modes.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="513"/>
        <source>&lt;i&gt;Cistrome&lt;/i&gt; is a UGENE version of Cistrome pipeline which also includes some tools useful for ChIP-seq analysisThis pipeline is aimed to provide the following analysis steps: peak calling and annotating, motif search and gene ontology.</source>
        <translation>&lt;i&gt;Cistrome&lt;/i&gt; is a UGENE version of Cistrome pipeline which also includes some tools useful for ChIP-seq analysis. This pipeline is aimed to provide the following analysis steps: peak calling and annotating, motif search and gene ontology.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="364"/>
        <source>No description</source>
        <translation>No description</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="327"/>
        <source>External tool &apos;%1&apos; cannot be validated as it depends on other tools, some of which are not valid. The list of tools is the following: </source>
        <translation>External tool &apos;%1&apos; cannot be validated as it depends on other tools, some of which are not valid. The list of tools is the following: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="351"/>
        <source>&apos;%1&apos; is %2 module and it is not installed. Install it and restart UGENE or set another %2 with already installed &apos;%1&apos; module.</source>
        <translation>&apos;%1&apos; is %2 module and it is not installed. Install it and restart UGENE or set another %2 with already installed &apos;%1&apos; module.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="374"/>
        <source>&lt;br&gt;&lt;br&gt;Version: </source>
        <translation>&lt;br&gt;&lt;br&gt;Version: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="378"/>
        <source>unknown</source>
        <translation>unknown</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="383"/>
        <source>&lt;br&gt;&lt;br&gt;Binary path: </source>
        <translation>&lt;br&gt;&lt;br&gt;Binary path: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="488"/>
        <source>&lt;i&gt;Bowtie&lt;i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end). &lt;a href=&apos;http://qt-project.org/doc/qt-4.8/qtextbrowser.html#anchorClicked&apos;&gt;Link text&lt;/a&gt; </source>
        <translation>&lt;i&gt;Bowtie&lt;i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end). &lt;a href=&apos;http://qt-project.org/doc/qt-4.8/qtextbrowser.html#anchorClicked&apos;&gt;Link text&lt;/a&gt;</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="543"/>
        <source>Choose Directory With Executables</source>
        <translation>Choose Directory With Executables</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="598"/>
        <source>Choose Directory With External Tools Pack</source>
        <translation>Choose Directory With External Tools Pack</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="651"/>
        <source>Not a valid external tools directory</source>
        <translation>Not a valid external tools directory</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="41"/>
        <source>%1 validate task</source>
        <translation>%1 validate task</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsInstallTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="391"/>
        <source>Installing external tools</source>
        <translation>Installing external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="396"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="136"/>
        <source>Searching external tools</source>
        <translation>Searching external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="154"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="156"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="362"/>
        <source>Checking external tools</source>
        <translation>Checking external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="380"/>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="382"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::FastQCSupport</name>
    <message>
        <location filename="../src/fastqc/FastqcSupport.cpp" line="50"/>
        <source>&lt;i&gt;FastQC&lt;/i&gt;: A quality control tool for high throughput sequence data.</source>
        <translation>&lt;i&gt;FastQC&lt;/i&gt;: A quality control tool for high throughput sequence data.</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupport</name>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="66"/>
        <source>The &lt;i&gt;formatdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by &lt;i&gt;blastall&lt;/i&gt;.</source>
        <translation>The &lt;i&gt;formatdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by &lt;i&gt;blastall&lt;/i&gt;.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="82"/>
        <location filename="../src/blast/FormatDBSupport.cpp" line="97"/>
        <source>The &lt;i&gt;makeblastdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by other BLAST+ tools.</source>
        <translation>The &lt;i&gt;makeblastdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by other BLAST+ tools.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="111"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation>Path for BLAST %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="114"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Path for BLAST+ %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="116"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="48"/>
        <source>Format</source>
        <translation>Format</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="49"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="80"/>
        <source>Select file(s)</source>
        <translation>Select file(s)</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="103"/>
        <source>Select a directory with input files</source>
        <translation>Select a directory with input files</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="114"/>
        <source>Select a directory to save database files</source>
        <translation>Select a directory to save database files</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="126"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="132"/>
        <source>Input files paths contain space characters.</source>
        <translation>Input files paths contain space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="140"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="145"/>
        <source>Output database path contain space characters.</source>
        <translation>Output database path contain space characters.</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportTask</name>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="47"/>
        <source>Run NCBI FormatDB task</source>
        <translation>Run NCBI FormatDB task</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="60"/>
        <source>Input files paths contain space characters.</source>
        <translation>Input files paths contain space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="77"/>
        <source>Output database path contain space characters.</source>
        <translation>Output database path contain space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="95"/>
        <source>Blast database creation has been cancelled</source>
        <translation>Blast database creation has been cancelled</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="102"/>
        <source>Blast database has been successfully created</source>
        <translation>Blast database has been successfully created</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="103"/>
        <source>Source sequences: </source>
        <translation>Source sequences: </translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="111"/>
        <source>Database file path: %1</source>
        <translation>Database file path: %1</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="113"/>
        <source>Type: %1</source>
        <translation>Type: %1</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="115"/>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="121"/>
        <source>Formatdb log file path: </source>
        <translation>Formatdb log file path: </translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="119"/>
        <source>Blast database creation has been failed</source>
        <translation>Blast database creation has been failed</translation>
    </message>
</context>
<context>
    <name>U2::GffreadSupportTask</name>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="34"/>
        <source>Running Gffread task</source>
        <translation>Running Gffread task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="59"/>
        <source>Unknown file format: %1</source>
        <translation>Unknown file format: %1</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="68"/>
        <source>The file format is not [%1]: %2</source>
        <translation>The file format is not [%1]: %2</translation>
    </message>
</context>
<context>
    <name>U2::JavaSupport</name>
    <message>
        <location filename="../src/java/JavaSupport.cpp" line="47"/>
        <source>Java Platform lets you develop and deploy Java applications on desktops and servers.&lt;br&gt;&lt;i&gt;(Requires Java 1.7 or higher)&lt;/i&gt;.&lt;br&gt;Java can be freely downloaded on the official web-site: https://www.java.com/en/download/</source>
        <translation>Java Platform lets you develop and deploy Java applications on desktops and servers.&lt;br&gt;&lt;i&gt;(Requires Java 1.7 or higher)&lt;/i&gt;.&lt;br&gt;Java can be freely downloaded on the official web-site: https://www.java.com/en/download/</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BaseShortReadsAlignerWorker</name>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="247"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="248"/>
        <source>Directory to save output files.</source>
        <translation>Directory to save output files.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="251"/>
        <source>Reference genome</source>
        <translation>Reference genome</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="252"/>
        <source>Path to indexed reference genome.</source>
        <translation>Path to indexed reference genome.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="255"/>
        <source>Library</source>
        <translation>Library</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="256"/>
        <source>Is this library mate-paired?</source>
        <translation>Is this library mate-paired?</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="259"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="260"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation>Base name of the output file. &apos;out.sam&apos; by default</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="292"/>
        <source>URL of a file with reads</source>
        <translation>URL of a file with reads</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="293"/>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="303"/>
        <source>Input reads to be aligned.</source>
        <translation>Input reads to be aligned.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="295"/>
        <source>URL of a file with mate reads</source>
        <translation>URL of a file with mate reads</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="296"/>
        <source>Input mate reads to be aligned.</source>
        <translation>Input mate reads to be aligned.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="302"/>
        <source>Input data</source>
        <translation>Input data</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="306"/>
        <source>Reverse FASTQ file</source>
        <translation>Reverse FASTQ file</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="307"/>
        <source>Reverse paired reads to be aligned.</source>
        <translation>Reverse paired reads to be aligned.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="319"/>
        <source>Assembly URL</source>
        <translation>Assembly URL</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="320"/>
        <source>Output assembly URL.</source>
        <translation>Output assembly URL.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="323"/>
        <source>Output data</source>
        <translation>Output data</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="324"/>
        <source>Output assembly files.</source>
        <translation>Output assembly files.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigPrompter</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="75"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="76"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="78"/>
        <source>Converts bedGraph files to bigWig %1 with bedGraphToBigWig.</source>
        <translation>Converts bedGraph files to bigWig %1 with bedGraphToBigWig.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigWorker</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="95"/>
        <source>Convert bedGraph Files to bigWig</source>
        <translation>Convert bedGraph Files to bigWig</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="96"/>
        <source>Convert bedGraph Files to bigWig.</source>
        <translation>Convert bedGraph Files to bigWig.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="100"/>
        <source>BedGrapgh files</source>
        <translation>BedGrapgh files</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="101"/>
        <source>Set of bedGraph files</source>
        <translation>Set of bedGraph files</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="102"/>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="103"/>
        <source>BigWig files</source>
        <translation>BigWig files</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="117"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="118"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="122"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="123"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="125"/>
        <source>Output name</source>
        <translation>Output name</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="126"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first BAM file with an extention.</source>
        <translation>A name of an output file. If default of empty value is provided the output name is the name of the first BAM file with an extention.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="128"/>
        <source>Block size</source>
        <translation>Block size</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="129"/>
        <source>Number of items to bundle in r-tree (-blockSize).</source>
        <translation>Number of items to bundle in r-tree (-blockSize).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="131"/>
        <source>Items per slot</source>
        <translation>Items per slot</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="132"/>
        <source>Number of data points bundled at lowest level (-itemsPerSlot).</source>
        <translation>Number of data points bundled at lowest level (-itemsPerSlot).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="134"/>
        <source>Uncompressed</source>
        <translation>Uncompressed</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="135"/>
        <source>If set, do not use compression (-unc).</source>
        <translation>If set, do not use compression (-unc).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="137"/>
        <source>Genome</source>
        <translation>Genome</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="138"/>
        <source>File with genome length.</source>
        <translation>File with genome length.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="170"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="168"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="169"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedtoolsIntersectPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="992"/>
        <source>Intersect annotations from &lt;u&gt;%1&lt;/u&gt; (&lt;b&gt;set A&lt;/b&gt;) with annotations from &lt;u&gt;%2&lt;/u&gt; (&lt;b&gt;set B&lt;/b&gt;). Report </source>
        <translation>Intersect annotations from &lt;u&gt;%1&lt;/u&gt; (&lt;b&gt;set A&lt;/b&gt;) with annotations from &lt;u&gt;%2&lt;/u&gt; (&lt;b&gt;set B&lt;/b&gt;). Report </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedtoolsIntersectWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="881"/>
        <source>Data not found by %1 id</source>
        <translation>Data not found by %1 id</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="885"/>
        <source>Can not get annotation table object</source>
        <translation>Can not get annotation table object</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="899"/>
        <source>Annotations A</source>
        <translation>Annotations A</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="904"/>
        <source>Annotations B</source>
        <translation>Annotations B</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="909"/>
        <source>Annotations</source>
        <translation>Annotations</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="909"/>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="925"/>
        <source>Result annotations</source>
        <translation>Result annotations</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="918"/>
        <source>Minimum overlap</source>
        <translation>Minimum overlap</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="919"/>
        <source>Minimum overlap required as a fraction of an annotation from set A.&lt;br/&gt;By default, even 1 bp overlap between annotations from set A and set B is taken into account. Yet sometimes you may want to restrict reported overlaps to cases where the annotations in B overlaps at least X% (e.g. 50%) of the A annotation. </source>
        <translation>Minimum overlap required as a fraction of an annotation from set A.&lt;br/&gt;By default, even 1 bp overlap between annotations from set A and set B is taken into account. Yet sometimes you may want to restrict reported overlaps to cases where the annotations in B overlaps at least X% (e.g. 50%) of the A annotation. </translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="926"/>
        <source>Select one of the following:&lt;ul&gt;&lt;li&gt;&lt;i&gt;Shared interval&lt;/i&gt; to report intervals shared between overlapped annotations from set A and set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Overlapped annotations from A&lt;/i&gt; to report annotations from set A that have an overlap with annotations from set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Non-overlapped annotations from A&lt;/i&gt; to report annotations from set A that have NO overlap with annotations from set B.&lt;/li&gt;&lt;/ul&gt;</source>
        <translation>Select one of the following:&lt;ul&gt;&lt;li&gt;&lt;i&gt;Shared interval&lt;/i&gt; to report intervals shared between overlapped annotations from set A and set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Overlapped annotations from A&lt;/i&gt; to report annotations from set A that have an overlap with annotations from set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Non-overlapped annotations from A&lt;/i&gt; to report annotations from set A that have NO overlap with annotations from set B.&lt;/li&gt;&lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="938"/>
        <source>Unique overlaps</source>
        <translation>Unique overlaps</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="939"/>
        <source>If the parameter value is &quot;True&quot;, write original A entry once if any overlaps found in B. In other words, just report the fact at least one overlap was found in B.&lt;br/&gt;The minimum overlap number is ignored in this case.&lt;br/&gt;&lt;br/&gt;If the parameter value is &quot;False&quot;, the A annotation is reported for every overlap found.</source>
        <translation>If the parameter value is &quot;True&quot;, write original A entry once if any overlaps found in B. In other words, just report the fact at least one overlap was found in B.&lt;br/&gt;The minimum overlap number is ignored in this case.&lt;br/&gt;&lt;br/&gt;If the parameter value is &quot;False&quot;, the A annotation is reported for every overlap found.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="976"/>
        <source>Intersect Annotations</source>
        <translation>Intersect Annotations</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="977"/>
        <source>Intersects two sets of annotations denoted as A and B.</source>
        <translation>Intersects two sets of annotations denoted as A and B.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllPrompter</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="341"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="342"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="343"/>
        <source>For sequence %1 find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation>For sequence %1 find annotations in database &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllWorker</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="57"/>
        <source>Best hits limit</source>
        <translation>Best hits limit</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="61"/>
        <source>Number of best hits from a region to keep. 0 turns it off. If used, 100 is recommended.</source>
        <translation>Number of best hits from a region to keep. 0 turns it off. If used, 100 is recommended.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="88"/>
        <source>Input sequence</source>
        <translation>Input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="89"/>
        <source>Sequence for which annotations is searched.</source>
        <translation>Sequence for which annotations is searched.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="90"/>
        <source>Annotations</source>
        <translation>Annotations</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="90"/>
        <source>Found annotations.</source>
        <translation>Found annotations.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="99"/>
        <source>Search type</source>
        <translation>Search type</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="100"/>
        <source>Select type of BLAST searches.</source>
        <translation>Select type of BLAST searches.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="101"/>
        <source>Database Path</source>
        <translation>Database Path</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="102"/>
        <source>Path with database files.</source>
        <translation>Path with database files.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="103"/>
        <source>Database Name</source>
        <translation>Database Name</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="104"/>
        <source>Base name for BLAST DB files.</source>
        <translation>Base name for BLAST DB files.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="105"/>
        <source>Expected value</source>
        <translation>Expected value</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="106"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation>This setting specifies the statistical significance threshold for reporting matches against database sequences.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="108"/>
        <source>Annotate as</source>
        <translation>Annotate as</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="109"/>
        <source>Name for annotations.</source>
        <translation>Name for annotations.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="110"/>
        <source>Tool Path</source>
        <translation>Tool Path</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="111"/>
        <source>External tool path.</source>
        <translation>External tool path.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="112"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="113"/>
        <source>Directory for temporary files.</source>
        <translation>Directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="115"/>
        <source>BLAST output</source>
        <translation>BLAST output</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="116"/>
        <source>Location of BLAST output file.</source>
        <translation>Location of BLAST output file.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="117"/>
        <source>BLAST output type</source>
        <translation>BLAST output type</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="118"/>
        <source>Type of BLAST output file.</source>
        <translation>Type of BLAST output file.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="119"/>
        <source>Gapped alignment</source>
        <translation>Gapped alignment</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="120"/>
        <source>Perform gapped alignment.</source>
        <translation>Perform gapped alignment.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="122"/>
        <source>Gap costs</source>
        <translation>Gap costs</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="123"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation>Cost to create and extend a gap in an alignment.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="124"/>
        <source>Match scores</source>
        <translation>Match scores</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="125"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation>Reward and penalty for matching and mismatching bases.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="180"/>
        <source>Local BLAST Search</source>
        <translation>Local BLAST Search</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="181"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation>Finds annotations for DNA sequence in local database.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="400"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation>Empty sequence supplied to BLAST</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="410"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation>Selected BLAST search with nucleotide input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="415"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation>Selected BLAST search with amino acid input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="422"/>
        <source>Not selected BLAST output file</source>
        <translation>Not selected BLAST output file</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusPrompter</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="252"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="253"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="254"/>
        <source>For sequence &lt;u&gt;%1&lt;/u&gt; find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation>For sequence &lt;u&gt;%1&lt;/u&gt; find annotations in database &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusWorker</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="59"/>
        <source>Culling limit</source>
        <translation>Culling limit</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="63"/>
        <source>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit</source>
        <translation>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="85"/>
        <source>Input sequence</source>
        <translation>Input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="86"/>
        <source>Sequence for which annotations is searched.</source>
        <translation>Sequence for which annotations is searched.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="87"/>
        <source>Annotations</source>
        <translation>Annotations</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="87"/>
        <source>Found annotations.</source>
        <translation>Found annotations.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="96"/>
        <source>Search type</source>
        <translation>Search type</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="97"/>
        <source>Select type of BLAST+ searches.</source>
        <translation>Select type of BLAST+ searches.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="98"/>
        <source>Database Path</source>
        <translation>Database Path</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="99"/>
        <source>Path with database files.</source>
        <translation>Path with database files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="100"/>
        <source>Database Name</source>
        <translation>Database Name</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="101"/>
        <source>Base name for BLAST+ DB files.</source>
        <translation>Base name for BLAST+ DB files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="102"/>
        <source>Expected value</source>
        <translation>Expected value</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="103"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation>This setting specifies the statistical significance threshold for reporting matches against database sequences.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="104"/>
        <source>Composition-based statistics</source>
        <translation>Composition-based statistics</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="104"/>
        <source>Composition-based statistics.</source>
        <translation>Composition-based statistics.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="106"/>
        <source>Annotate as</source>
        <translation>Annotate as</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="107"/>
        <source>Name for annotations.</source>
        <translation>Name for annotations.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="108"/>
        <source>Tool Path</source>
        <translation>Tool Path</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="109"/>
        <source>External tool path.</source>
        <translation>External tool path.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="110"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="111"/>
        <source>Directory for temporary files.</source>
        <translation>Directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="112"/>
        <source>BLAST output</source>
        <translation>BLAST output</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="113"/>
        <source>Location of BLAST output file.</source>
        <translation>Location of BLAST output file.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="114"/>
        <source>BLAST output type</source>
        <translation>BLAST output type</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="115"/>
        <source>Type of BLAST output file.</source>
        <translation>Type of BLAST output file.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="116"/>
        <source>Gapped alignment</source>
        <translation>Gapped alignment</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="117"/>
        <source>Perform gapped alignment.</source>
        <translation>Perform gapped alignment.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="119"/>
        <source>Gap costs</source>
        <translation>Gap costs</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="120"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation>Cost to create and extend a gap in an alignment.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="121"/>
        <source>Match scores</source>
        <translation>Match scores</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="122"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation>Reward and penalty for matching and mismatching bases.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="154"/>
        <source>Local BLAST+ Search</source>
        <translation>Local BLAST+ Search</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="155"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation>Finds annotations for DNA sequence in local database.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="323"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation>Empty sequence supplied to BLAST</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="333"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation>Selected BLAST search with nucleotide input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="338"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation>Selected BLAST search with amino acid input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="345"/>
        <source>Not selected BLAST output file</source>
        <translation>Not selected BLAST output file</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Bowtie2Worker</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="116"/>
        <source>Mode</source>
        <translation>Mode</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="117"/>
        <source>When the -n option is specified (which is the default), bowtie determines which alignments                               
are valid according to the following policy, which is similar to Maq&apos;s default policy.                               
In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               
through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</source>
        <translation>When the -n option is specified (which is the default), bowtie determines which alignments                               
are valid according to the following policy, which is similar to Maq&apos;s default policy.                               
In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               
through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="123"/>
        <source>Number of mismatches</source>
        <translation>Number of mismatches</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="124"/>
        <source>Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1.                               
Setting this higher makes alignment slower (often much slower) but increases sensitivity.</source>
        <translation>Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1.                               
Setting this higher makes alignment slower (often much slower) but increases sensitivity.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="128"/>
        <source>Seed length (--L)</source>
        <translation>Seed length (--L)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="129"/>
        <source>Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.</source>
        <translation>Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="132"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation>Add columns to allow gaps (--dpad)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="133"/>
        <source>&quot;Pads&quot; dynamic programming problems by specified number of columns on either side to allow gaps.</source>
        <translation>&quot;Pads&quot; dynamic programming problems by specified number of columns on either side to allow gaps.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="136"/>
        <source>Disallow gaps (--gbar)</source>
        <translation>Disallow gaps (--gbar)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="137"/>
        <source>Disallow gaps within specified number of positions of the beginning or end of the read.</source>
        <translation>Disallow gaps within specified number of positions of the beginning or end of the read.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="140"/>
        <source>Seed (--seed)</source>
        <translation>Seed (--seed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="141"/>
        <source>Use specified value as the seed for pseudo-random number generator.</source>
        <translation>Use specified value as the seed for pseudo-random number generator.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="144"/>
        <source>Threads</source>
        <translation>Threads</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="145"/>
        <source>Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize                               
when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.</source>
        <translation>Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize                               
when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="149"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation>No unpaired alignments (--no-mixed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="150"/>
        <source>If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments                               
for the constituent mates. This is called &quot;mixed mode.&quot; To disable mixed mode, set this option. Bowtie2 runs a little                               
faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.</source>
        <translation>If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments                               
for the constituent mates. This is called &quot;mixed mode.&quot; To disable mixed mode, set this option. Bowtie2 runs a little                               
faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="155"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation>No discordant alignments (--no-discordant)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="156"/>
        <source>By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment                               
is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.</source>
        <translation>By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment                               
is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="160"/>
        <source>No forward orientation (--nofw)</source>
        <translation>No forward orientation (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="161"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="164"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>No reverse-complement orientation (--norc)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="165"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="168"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation>No overlapping mates (--no-overlap)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="169"/>
        <source>If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in                                 
a concordant alignment.</source>
        <translation>If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in                                 
a concordant alignment.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="173"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation>No mates containing one another (--no-contain)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="174"/>
        <source>If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the                                
other in a concordant alignment.</source>
        <translation>If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the                                
other in a concordant alignment.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="224"/>
        <source>Align Reads with Bowtie2</source>
        <translation>Align Reads with Bowtie2</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="225"/>
        <source>Performs alignment of short reads with Bowtie2.</source>
        <translation>Performs alignment of short reads with Bowtie2.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BowtieWorker</name>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="142"/>
        <source>Mode:</source>
        <translation>Mode:</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="143"/>
        <source>When the -n option is specified (which is the default), bowtie determines which alignments                               are valid according to the following policy, which is similar to Maq&apos;s default policy.                               In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</source>
        <translation>When the -n option is specified (which is the default), bowtie determines which alignments are valid according to the following policy, which is similar to Maq&apos;s default policy. In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="149"/>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="153"/>
        <source>Mismatches number</source>
        <translation>Mismatches number</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="150"/>
        <source>Mismatches number.</source>
        <translation>Mismatches number.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="154"/>
        <source>Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment,                              not just in the seed. The default is 70. Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30;                              rounding can be disabled with --nomaqround.</source>
        <translation>Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, not just in the seed. The default is 70. Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; rounding can be disabled with --nomaqround.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="159"/>
        <source>Seed length</source>
        <translation>Seed length</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="160"/>
        <source>The seed length; i.e., the number of bases on the high-quality end of the read to which the                               -n ceiling applies. The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.</source>
        <translation>The seed length; i.e., the number of bases on the high-quality end of the read to which the                               -n ceiling applies. The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="164"/>
        <source>No forward orientation</source>
        <translation>No forward orientation</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="165"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="168"/>
        <source>No reverse-complement orientation</source>
        <translation>No reverse-complement orientation</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="169"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="172"/>
        <source>Maximum of backtracks</source>
        <translation>Maximum of backtracks</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="173"/>
        <source>The maximum insert size for valid paired-end alignments. E.g. if -X 100 is specified and a paired-end alignment                               consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, that alignment is                               considered valid (as long as -I is also satisfied). A 61-bp gap would not be valid in that case.                                If trimming options -3 or -5 are also used, the -X constraint is applied with respect to the untrimmed mates,                               not the trimmed mates. Default: 250.</source>
        <translation>The maximum insert size for valid paired-end alignments. E.g. if -X 100 is specified and a paired-end alignment                               consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, that alignment is                               considered valid (as long as -I is also satisfied). A 61-bp gap would not be valid in that case.                                If trimming options -3 or -5 are also used, the -X constraint is applied with respect to the untrimmed mates,                               not the trimmed mates. Default: 250.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="180"/>
        <source>Try as hard</source>
        <translation>Try as hard</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="181"/>
        <source>Try as hard as possible to find valid alignments when they exist, including paired-end alignments.                                This is equivalent to specifying very high values for the --maxbts and --pairtries options. This mode is generally much                                slower than the default settings, but can be useful for certain problems. This mode is slower when (a) the reference is                                very repetitive, (b) the reads are low quality, or (c) not many reads have valid alignments.</source>
        <translation>Try as hard as possible to find valid alignments when they exist, including paired-end alignments.                                This is equivalent to specifying very high values for the --maxbts and --pairtries options. This mode is generally much                                slower than the default settings, but can be useful for certain problems. This mode is slower when (a) the reference is                                very repetitive, (b) the reads are low quality, or (c) not many reads have valid alignments.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="187"/>
        <source>Best hits</source>
        <translation>Best hits</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="188"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. Best-first                                search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost.                                Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases.                                If you receive an error message saying that chunk memory has been exhausted in --best mode,                                try adjusting this parameter up to dedicate more memory to the descriptors. Default: 64.</source>
        <translation>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="195"/>
        <source>No Maq rounding</source>
        <translation>No Maq rounding</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="196"/>
        <source>Maq accepts quality values in the Phred quality scale, but internally rounds values to the nearest 10,              with a maximum of 30. By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.</source>
        <translation>Maq accepts quality values in the Phred quality scale, but internally rounds values to the nearest 10,              with a maximum of 30. By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="200"/>
        <source>Seed</source>
        <translation>Seed</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="201"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator.</source>
        <translation>Use &lt;int&gt; as the seed for pseudo-random number generator.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="204"/>
        <source>Best alignments</source>
        <translation>Best alignments</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="205"/>
        <source>Make Bowtie guarantee that reported singleton alignments are best in terms of stratum                              (i.e. number of mismatches, or mismatches in the seed in the case of -n mode) and in terms of                              the quality values at the mismatched position(s). bowtie is somewhat slower when --best is specified.</source>
        <translation>Make Bowtie guarantee that reported singleton alignments are best in terms of stratum                              (i.e. number of mismatches, or mismatches in the seed in the case of -n mode) and in terms of                              the quality values at the mismatched position(s). bowtie is somewhat slower when --best is specified.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="210"/>
        <source>All alignments</source>
        <translation>All alignments</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="211"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation>Report all valid alignments per read or pair.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="214"/>
        <source>Colorspace</source>
        <translation>Colorspace</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="215"/>
        <source>When -C is specified, read sequences are treated as colors. Colors may be encoded either as numbers                               (0=blue, 1=green, 2=orange, 3=red) or as characters A/C/G/T (A=blue, C=green, G=orange, T=red).</source>
        <translation>When -C is specified, read sequences are treated as colors. Colors may be encoded either as numbers                               (0=blue, 1=green, 2=orange, 3=red) or as characters A/C/G/T (A=blue, C=green, G=orange, T=red).</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="252"/>
        <source>Align Reads with Bowtie</source>
        <translation>Align Reads with Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="253"/>
        <source>Performs alignment of short reads with Bowtie.</source>
        <translation>Performs alignment of short reads with Bowtie.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BwaMemWorker</name>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="137"/>
        <source>Number of threads</source>
        <translation>Number of threads</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="138"/>
        <source>Number of threads (-t).</source>
        <translation>Number of threads (-t).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="141"/>
        <source>Min seed length</source>
        <translation>Min seed length</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="142"/>
        <source>Path to indexed reference genome (-k).</source>
        <translation>Path to indexed reference genome (-k).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="145"/>
        <source>Index algorithm</source>
        <translation>Index algorithm</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="146"/>
        <source>Index algorithm (-a).</source>
        <translation>Index algorithm (-a).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="149"/>
        <source>Band width</source>
        <translation>Band width</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="150"/>
        <source>Band width for banded alignment (-w).</source>
        <translation>Band width for banded alignment (-w).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="153"/>
        <source>Dropoff</source>
        <translation>Dropoff</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="154"/>
        <source>Off-diagonal X-dropoff (-d).</source>
        <translation>Off-diagonal X-dropoff (-d).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="157"/>
        <source>Internal seed length</source>
        <translation>Internal seed length</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="158"/>
        <source>Look for internal seeds inside a seed longer than {-k} (-r).</source>
        <translation>Look for internal seeds inside a seed longer than {-k} (-r).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="161"/>
        <source>Skip seed threshold</source>
        <translation>Skip seed threshold</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="162"/>
        <source>Skip seeds with more than INT occurrences (-c).</source>
        <translation>Skip seeds with more than INT occurrences (-c).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="165"/>
        <source>Drop chain threshold</source>
        <translation>Drop chain threshold</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="166"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain (-D).</source>
        <translation>Drop chains shorter than FLOAT fraction of the longest overlapping chain (-D).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="169"/>
        <source>Rounds of mate rescues</source>
        <translation>Rounds of mate rescues</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="170"/>
        <source>Perform at most INT rounds of mate rescues for each read (-m).</source>
        <translation>Perform at most INT rounds of mate rescues for each read (-m).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="173"/>
        <source>Skip mate rescue</source>
        <translation>Skip mate rescue</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="174"/>
        <source>Skip mate rescue (-S).</source>
        <translation>Skip mate rescue (-S).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="177"/>
        <source>Skip pairing</source>
        <translation>Skip pairing</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="178"/>
        <source>Skip pairing; mate rescue performed unless -S also in use (-P).</source>
        <translation>Skip pairing; mate rescue performed unless -S also in use (-P).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="181"/>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="185"/>
        <source>Mismatch penalty</source>
        <translation>Mismatch penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="182"/>
        <source>Score for a sequence match (-A).</source>
        <translation>Score for a sequence match (-A).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="186"/>
        <source>Penalty for a mismatch (-B).</source>
        <translation>Penalty for a mismatch (-B).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="189"/>
        <source>Gap open penalty</source>
        <translation>Gap open penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="190"/>
        <source>Gap open penalty (-O).</source>
        <translation>Gap open penalty (-O).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="193"/>
        <source>Gap extension penalty</source>
        <translation>Gap extension penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="194"/>
        <source>Gap extension penalty; a gap of size k cost {-O} (-E).</source>
        <translation>Gap extension penalty; a gap of size k cost {-O} (-E).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="197"/>
        <source>Penalty for clipping</source>
        <translation>Penalty for clipping</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="198"/>
        <source>Penalty for clipping (-L).</source>
        <translation>Penalty for clipping (-L).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="201"/>
        <source>Penalty unpaired</source>
        <translation>Penalty unpaired</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="202"/>
        <source>Penalty for an unpaired read pair (-U).</source>
        <translation>Penalty for an unpaired read pair (-U).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="205"/>
        <source>Score threshold</source>
        <translation>Score threshold</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="206"/>
        <source>Minimum score to output (-T).</source>
        <translation>Minimum score to output (-T).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="256"/>
        <source>Align Reads with BWA MEM</source>
        <translation>Align Reads with BWA MEM</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="257"/>
        <source>Performs alignment of short reads with BWA MEM.</source>
        <translation>Performs alignment of short reads with BWA MEM.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BwaWorker</name>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="139"/>
        <source>Use missing prob</source>
        <translation>Use missing prob</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="140"/>
        <source>Use missing prob instead maximum edit distance.</source>
        <translation>Use missing prob instead maximum edit distance.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="143"/>
        <source>Max #diff</source>
        <translation>Max #diff</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="144"/>
        <source>Max #diff (-n)</source>
        <translation>Max #diff (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="147"/>
        <source>Missing prob</source>
        <translation>Missing prob</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="148"/>
        <source>Missing prob (-n)</source>
        <translation>Missing prob (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="151"/>
        <source>Seed length</source>
        <translation>Seed length</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="152"/>
        <source>Seed length (-l).</source>
        <translation>Seed length (-l).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="155"/>
        <source>Max gap opens</source>
        <translation>Max gap opens</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="156"/>
        <source>Max gap opens (-o).</source>
        <translation>Max gap opens (-o).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="159"/>
        <source>Index algorithm</source>
        <translation>Index algorithm</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="160"/>
        <source>Index algorithm (-a).</source>
        <translation>Index algorithm (-a).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="163"/>
        <source>Best hits</source>
        <translation>Best hits</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="164"/>
        <source>Best hits (-R).</source>
        <translation>Best hits (-R).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="167"/>
        <source>Long-scaled gap penalty for long deletions</source>
        <translation>Long-scaled gap penalty for long deletions</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="168"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation>Long-scaled gap penalty for long deletions (-L)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="171"/>
        <source>Non iterative mode</source>
        <translation>Non iterative mode</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="172"/>
        <source>Non iterative mode (-N).</source>
        <translation>Non iterative mode (-N).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="175"/>
        <location filename="../src/bwa/BwaWorker.cpp" line="176"/>
        <source>Enable long gaps</source>
        <translation>Enable long gaps</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="179"/>
        <source>Max gap extensions</source>
        <translation>Max gap extensions</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="180"/>
        <source>Max gap extensions (-e)</source>
        <translation>Max gap extensions (-e)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="183"/>
        <source>Indel offset</source>
        <translation>Indel offset</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="184"/>
        <source>Indel offset (-i)</source>
        <translation>Indel offset (-i)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="187"/>
        <source>Max long deletions extensions</source>
        <translation>Max long deletions extensions</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="188"/>
        <source>Max long deletions extensions(-d)</source>
        <translation>Max long deletions extensions(-d)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="191"/>
        <source>Max queue entries</source>
        <translation>Max queue entries</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="192"/>
        <source>Max queue entries (-m)</source>
        <translation>Max queue entries (-m)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="195"/>
        <source>Barcode length</source>
        <translation>Barcode length</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="196"/>
        <source>Barcode length (-B)</source>
        <translation>Barcode length (-B)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="199"/>
        <source>Threads</source>
        <translation>Threads</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="200"/>
        <source>Threads (-t)</source>
        <translation>Threads (-t)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="203"/>
        <source>Max seed differencies</source>
        <translation>Max seed differencies</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="204"/>
        <source>Max seed differencies (-k)</source>
        <translation>Max seed differencies (-k)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="207"/>
        <source>Mismatch penalty</source>
        <translation>Mismatch penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="208"/>
        <source>Mismatch penalty (-M)</source>
        <translation>Mismatch penalty (-M)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="211"/>
        <source>Gap open penalty</source>
        <translation>Gap open penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="212"/>
        <source>Gap open penalty (-O).</source>
        <translation>Gap open penalty (-O).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="215"/>
        <source>Gap extension penalty</source>
        <translation>Gap extension penalty</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="216"/>
        <source>Gap extension penalty; a gap of size k cost (-E).</source>
        <translation>Gap extension penalty; a gap of size k cost (-E).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="219"/>
        <source>Quolity threshold</source>
        <translation>Quality threshold</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="220"/>
        <source>Quolity threshold (-q).</source>
        <translation>Quality threshold (-q).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="286"/>
        <source>Align Reads with BWA</source>
        <translation>Align Reads with BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="287"/>
        <source>Performs alignment of short reads with BWA.</source>
        <translation>Performs alignment of short reads with BWA.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Prompter</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="377"/>
        <source> from %1</source>
        <translation> from %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Worker</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="89"/>
        <source>Input sequences</source>
        <translation>Input sequences</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="90"/>
        <source>DNA sequences that need to be assembled</source>
        <translation>DNA sequences that need to be assembled</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="94"/>
        <source>Input URL(s)</source>
        <translation>Input URL(s)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="95"/>
        <source>URL(s) of input file(s) in FASTA format with DNA sequences that need to be assembled</source>
        <translation>URL(s) of input file(s) in FASTA format with DNA sequences that need to be assembled</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="106"/>
        <source>Assembly Sequences with CAP3</source>
        <translation>Assembly Sequences with CAP3</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="111"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="112"/>
        <source>Write assembly results to this output file in ACE format.</source>
        <translation>Write assembly results to this output file in ACE format.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="115"/>
        <source>Quality cutoff for clipping</source>
        <translation>Quality cutoff for clipping</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="116"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation>Base quality cutoff for clipping (-c)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="119"/>
        <source>Clipping range</source>
        <translation>Clipping range</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="120"/>
        <source>One of the parameters to control clipping of a poor end region of a read (-y)&lt;p&gt;If there are quality values, CAP3 computes two positions qualpos5 and qualpos3 of read f such that the region of read f from position qualpos5 to position qualpos3 consists mostly of quality values greater than &lt;i&gt;Quality cutoff for clipping&lt;/i&gt;.&lt;/p&gt; &lt;p&gt;If there are no quality values, then qualpos5 is set to 1 and qualpos3 is set the length of read f. &lt;p&gt;The range for the left clipping position of read f is from 1 to qualpos5 + &lt;i&gt;Clipping range&lt;/i&gt;. &lt;br/&gt;The range for the right clipping position of read f is from qualpos3 - &lt;i&gt;Clipping range&lt;/i&gt; to the end of read f.&lt;/p&gt;</source>
        <translation>One of the parameters to control clipping of a poor end region of a read (-y)&lt;p&gt;If there are quality values, CAP3 computes two positions qualpos5 and qualpos3 of read f such that the region of read f from position qualpos5 to position qualpos3 consists mostly of quality values greater than &lt;i&gt;Quality cutoff for clipping&lt;/i&gt;.&lt;/p&gt; &lt;p&gt;If there are no quality values, then qualpos5 is set to 1 and qualpos3 is set the length of read f. &lt;p&gt;The range for the left clipping position of read f is from 1 to qualpos5 + &lt;i&gt;Clipping range&lt;/i&gt;. &lt;br/&gt;The range for the right clipping position of read f is from qualpos3 - &lt;i&gt;Clipping range&lt;/i&gt; to the end of read f.&lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="133"/>
        <source>Quality cutoff for differences</source>
        <translation>Quality cutoff for differences</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="134"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation>Base quality cutoff for differences (-b)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="137"/>
        <source>Maximum difference score</source>
        <translation>Maximum difference score</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="138"/>
        <source>Max qscore sum at differences (-d)&lt;p&gt;If an overlap contains lots of differences at bases of high quality, then the overlap is removed.&lt;/p&gt;&lt;p&gt;The difference score is calculated as follows. If the overlap contains a difference at bases of quality values q1 and q2, then the score at the difference is max(0, min(q1, q2) - b), where b is &lt;i&gt;Quality cutoff for differences&lt;/i&gt;. The difference score of an overlap is the sum of scores at each difference.&lt;/p&gt;</source>
        <translation>Max qscore sum at differences (-d)&lt;p&gt;If an overlap contains lots of differences at bases of high quality, then the overlap is removed.&lt;/p&gt;&lt;p&gt;The difference score is calculated as follows. If the overlap contains a difference at bases of quality values q1 and q2, then the score at the difference is max(0, min(q1, q2) - b), where b is &lt;i&gt;Quality cutoff for differences&lt;/i&gt;. The difference score of an overlap is the sum of scores at each difference.&lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="149"/>
        <source>Match score factor</source>
        <translation>Match score factor</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="150"/>
        <source>Match score factor (-m) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Match score factor (-m) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="155"/>
        <source>Mismatch score factor</source>
        <translation>Mismatch score factor</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="156"/>
        <source>Mismatch score factor (-n) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Mismatch score factor (-n) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="161"/>
        <source>Gap penalty factor</source>
        <translation>Gap penalty factor</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="162"/>
        <source>Gap penalty factor (-g) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Gap penalty factor (-g) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="167"/>
        <source>Overlap similarity score cutoff</source>
        <translation>Overlap similarity score cutoff</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="168"/>
        <source>If the similarity score of an overlap is less than the overlap similarity score cutoff (-s), then the overlap is removed.&lt;p&gt;The similarity score of an overlapping alignment is defined using base quality values as follows.&lt;p&gt;A match at bases of quality values q1 and q2 is given a score of m * min(q1,q2), where m is &lt;i&gt;Match score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A mismatch at bases of quality values q1 and q2 is given a score of n * min(q1,q2), where n is &lt;i&gt;Mismatch score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A base of quality value q1 in a gap is given a score of -g * min(q1,q2), where q2 is the quality value of the base in the other sequence right before the gap and g is &lt;i&gt;Gap penalty factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;The score of a gap is the sum of scores of each base in the gap minus a gap open penalty.&lt;/p&gt;&lt;p&gt;The similarity score of an overlapping alignment is the sum of scores of each match, each mismatch, and each gap. &lt;/p&gt;</source>
        <translation>If the similarity score of an overlap is less than the overlap similarity score cutoff (-s), then the overlap is removed.&lt;p&gt;The similarity score of an overlapping alignment is defined using base quality values as follows.&lt;p&gt;A match at bases of quality values q1 and q2 is given a score of m * min(q1,q2), where m is &lt;i&gt;Match score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A mismatch at bases of quality values q1 and q2 is given a score of n * min(q1,q2), where n is &lt;i&gt;Mismatch score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A base of quality value q1 in a gap is given a score of -g * min(q1,q2), where q2 is the quality value of the base in the other sequence right before the gap and g is &lt;i&gt;Gap penalty factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;The score of a gap is the sum of scores of each base in the gap minus a gap open penalty.&lt;/p&gt;&lt;p&gt;The similarity score of an overlapping alignment is the sum of scores of each match, each mismatch, and each gap. &lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="186"/>
        <source>Overlap length cutoff</source>
        <translation>Overlap length cutoff</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="187"/>
        <source>An overlap is taken into account only if the length of the overlap in bp is no less than the specified value (parameter -o of CAP3).</source>
        <translation>An overlap is taken into account only if the length of the overlap in bp is no less than the specified value (parameter -o of CAP3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="191"/>
        <source>Overlap percent identity cutoff</source>
        <translation>Overlap percent identity cutoff</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="192"/>
        <source>An overlap is taken into account only if the percent identity of the overlap is no less than the specified value (parameter -p of CAP3).</source>
        <translation>An overlap is taken into account only if the percent identity of the overlap is no less than the specified value (parameter -p of CAP3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="196"/>
        <source>Max number of word matches</source>
        <translation>Max number of word matches</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="197"/>
        <source>This parameter allows one to trade off the efficiency of the program for its accuracy (parameter -t of CAP3).&lt;p&gt;For a read f, CAP3 computes overlaps between read f and other reads by considering short word matches between read f and other reads. A word match is examined to see if it can be extended into a long overlap. If read f has overlaps with many other reads, then read f has many short word matches with many other reads.&lt;/p&gt;&lt;p&gt;This parameter gives an upper limit, for any word, on the number of word matches between read f and other reads that are considered by CAP3.&lt;/p&gt;&lt;p&gt;Using a large value for this parameter allows CAP3 to consider more word matches between read f and other reads, which can find more overlaps for read f, but slows down the program.&lt;/p&gt;&lt;p&gt;Using a small value for this parameter has the opposite effect.&lt;/p&gt;</source>
        <translation>This parameter allows one to trade off the efficiency of the program for its accuracy (parameter -t of CAP3).&lt;p&gt;For a read f, CAP3 computes overlaps between read f and other reads by considering short word matches between read f and other reads. A word match is examined to see if it can be extended into a long overlap. If read f has overlaps with many other reads, then read f has many short word matches with many other reads.&lt;/p&gt;&lt;p&gt;This parameter gives an upper limit, for any word, on the number of word matches between read f and other reads that are considered by CAP3.&lt;/p&gt;&lt;p&gt;Using a large value for this parameter allows CAP3 to consider more word matches between read f and other reads, which can find more overlaps for read f, but slows down the program.&lt;/p&gt;&lt;p&gt;Using a small value for this parameter has the opposite effect.&lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="212"/>
        <source>Band expansion size</source>
        <translation>Band expansion size</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="213"/>
        <source>CAP3 determines a minimum band of diagonals for an overlapping alignment between two sequence reads. The band is expanded by a number of bases specified by this value (parameter -a of CAP3).</source>
        <translation>CAP3 determines a minimum band of diagonals for an overlapping alignment between two sequence reads. The band is expanded by a number of bases specified by this value (parameter -a of CAP3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="218"/>
        <source>Max gap length in an overlap</source>
        <translation>Max gap length in an overlap</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="219"/>
        <source>The maximum length of gaps allowed in any overlap (-f)&lt;p&gt;I.e. overlaps with longer gaps are rejected.&lt;/p&gt;&lt;p&gt;Note that a small value for this parameter may cause the program to remove true overlaps and to produce incorrect results.&lt;/p&gt;&lt;p&gt;The parameter may be used to split reads from alternative splicing forms into separate contigs.&lt;/p&gt;</source>
        <translation>The maximum length of gaps allowed in any overlap (-f)&lt;p&gt;I.e. overlaps with longer gaps are rejected.&lt;/p&gt;&lt;p&gt;Note that a small value for this parameter may cause the program to remove true overlaps and to produce incorrect results.&lt;/p&gt;&lt;p&gt;The parameter may be used to split reads from alternative splicing forms into separate contigs.&lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="227"/>
        <source>Assembly reverse reads</source>
        <translation>Assembly reverse reads</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="228"/>
        <source>Specifies whether to consider reads in reverse orientation for assembly (originally, parameter -r of CAP3).</source>
        <translation>Specifies whether to consider reads in reverse orientation for assembly (originally, parameter -r of CAP3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="232"/>
        <source>CAP3 tool path</source>
        <translation>CAP3 tool path</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="233"/>
        <source>The path to the CAP3 external tool in UGENE.</source>
        <translation>The path to the CAP3 external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="236"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="237"/>
        <source>The directory for temporary files.</source>
        <translation>The directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="378"/>
        <source>Aligns long DNA reads from &lt;u&gt;%1&lt;/u&gt; with CAP3.</source>
        <translation>Aligns long DNA reads from &lt;u&gt;%1&lt;/u&gt; with CAP3.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportPrompter</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="373"/>
        <source>Creates summary statistics on ChIP enrichment and saves it to %1.</source>
        <translation>Creates summary statistics on ChIP enrichment and saves it to %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportWorker</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="216"/>
        <source>Peak regions</source>
        <translation>Peak regions</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="217"/>
        <source>Peak regions.</source>
        <translation>Peak regions.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="219"/>
        <source>Enrichment signal</source>
        <translation>Enrichment signal</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="220"/>
        <source>Enrichment signal.</source>
        <translation>Enrichment signal.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="225"/>
        <source>CEAS data</source>
        <translation>CEAS data</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="226"/>
        <source>Peak regions and enrichment signal profile for generating CEAS report.</source>
        <translation>Peak regions and enrichment signal profile for generating CEAS report.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="235"/>
        <source>Output report file</source>
        <translation>Output report file</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="236"/>
        <source>Path to the report output file. Result for CEAS analysis.</source>
        <translation>Path to the report output file. Result for CEAS analysis.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="241"/>
        <source>Output annotations file</source>
        <translation>Output annotations file</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="242"/>
        <source>Name of tab-delimited output text file, containing a row of annotations for every RefSeq gene. (file is not generated if no peak location data is supplied).</source>
        <translation>Name of tab-delimited output text file, containing a row of annotations for every RefSeq gene. (file is not generated if no peak location data is supplied).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="245"/>
        <source>Gene annotations table</source>
        <translation>Gene annotations table</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="246"/>
        <source>Path to gene annotation table (e.g. a refGene table in sqlite3 db format (--gt).</source>
        <translation>Path to gene annotation table (e.g. a refGene table in sqlite3 db format (--gt).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="249"/>
        <source>Span size</source>
        <translation>Span size</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="250"/>
        <source>Span from TSS and TTS in the gene-centered annotation (base pairs). ChIP regions within this range from TSS and TTS are considered when calculating the coverage rates in promoter and downstream (--span).</source>
        <translation>Span from TSS and TTS in the gene-centered annotation (base pairs). ChIP regions within this range from TSS and TTS are considered when calculating the coverage rates in promoter and downstream (--span).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="255"/>
        <source>Wiggle profiling resolution</source>
        <translation>Wiggle profiling resolution</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="256"/>
        <source>Wiggle profiling resolution. WARNING: Value smaller than the wig interval (resolution) may cause aliasing error (--pf-res).</source>
        <translation>Wiggle profiling resolution. WARNING: Value smaller than the wig interval (resolution) may cause aliasing error (--pf-res).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="260"/>
        <source>Promoter/downstream interval</source>
        <translation>Promoter/downstream interval</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="261"/>
        <source>Promoter/downstream intervals for ChIP region annotation are three values or a single value can be given. If a single value is given, it will be segmented into three equal fractions (e.g. 3000 is equivalent to 1000,2000,3000) (--rel-dist).</source>
        <translation>Promoter/downstream intervals for ChIP region annotation are three values or a single value can be given. If a single value is given, it will be segmented into three equal fractions (e.g. 3000 is equivalent to 1000,2000,3000) (--rel-dist).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="267"/>
        <source>BiPromoter ranges</source>
        <translation>BiPromoter ranges</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="268"/>
        <source>Bidirectional-promoter sizes for ChIP region annotation. It&apos;s two values or a single value can be given. If a single value is given, it will be segmented into two equal fractions (e.g. 5000 is equivalent to 2500,5000) (--bisizes).</source>
        <translation>Bidirectional-promoter sizes for ChIP region annotation. It&apos;s two values or a single value can be given. If a single value is given, it will be segmented into two equal fractions (e.g. 5000 is equivalent to 2500,5000) (--bisizes).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="273"/>
        <source>Relative distance</source>
        <translation>Relative distance</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="274"/>
        <source>Relative distance to TSS/TTS in WIGGLE file. profiling. (--rel-dist)</source>
        <translation>Relative distance to TSS/TTS in WIGGLE file. profiling. (--rel-dist)</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="277"/>
        <source>Gene group files</source>
        <translation>Gene group files</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="278"/>
        <source>Gene groups of particular interest in wig profiling. Each gene group file must have gene names in the 1st column. The file names are separated by commas (--gn-groups).</source>
        <translation>Gene groups of particular interest in wig profiling. Each gene group file must have gene names in the 1st column. The file names are separated by commas (--gn-groups).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="282"/>
        <source>Gene group names</source>
        <translation>Gene group names</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="283"/>
        <source>&lt;i&gt;Set this parameter empty for using default values.&lt;/i&gt;&lt;br&gt;The names of the gene groups from &quot;Gene group files&quot; parameter. These names appear in the legends of the wig profiling plots.&lt;br&gt; Values range: comma-separated list of strings. Default value: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</source>
        <translation>&lt;i&gt;Set this parameter empty for using default values.&lt;/i&gt;&lt;br&gt;The names of the gene groups from &quot;Gene group files&quot; parameter. These names appear in the legends of the wig profiling plots.&lt;br&gt; Values range: comma-separated list of strings. Default value: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="323"/>
        <source>Report file</source>
        <translation>Report file</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="330"/>
        <source>Annotations file</source>
        <translation>Annotations file</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="353"/>
        <source>Create CEAS Report</source>
        <translation>Create CEAS Report</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="354"/>
        <source>Provides summary statistics on ChIP enrichment in important genomic regions such as individual chromosomes, promoters, gene bodies or exons, and infers the genes most likely to be regulated by the binding factor under study.</source>
        <translation>Provides summary statistics on ChIP enrichment in important genomic regions such as individual chromosomes, promoters, gene bodies or exons, and infers the genes most likely to be regulated by the binding factor under study.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOPrompter</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="139"/>
        <source> from %1</source>
        <translation> from %1</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="140"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</source>
        <translation>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOWorker</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA</source>
        <translation>Input MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA to process.</source>
        <translation>Input MSA to process.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="62"/>
        <source>ClustalO result MSA</source>
        <translation>ClustalO result MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="63"/>
        <source>The result of the ClustalO alignment.</source>
        <translation>The result of the ClustalO alignment.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="72"/>
        <source>Number of iterations</source>
        <translation>Number of iterations</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="73"/>
        <source>Number of (combined guide-tree/HMM) iterations.</source>
        <translation>Number of (combined guide-tree/HMM) iterations.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="74"/>
        <source>Number of guidetree iterations</source>
        <translation>Number of guidetree iterations</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="75"/>
        <source>Maximum number guidetree iterations.</source>
        <translation>Maximum number guidetree iterations.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="76"/>
        <source>Number of HMM iterations</source>
        <translation>Number of HMM iterations</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="77"/>
        <source>Maximum number of HMM iterations.</source>
        <translation>Maximum number of HMM iterations.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="78"/>
        <source>Set auto options</source>
        <translation>Set auto options</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="79"/>
        <source>Set options automatically (might overwrite some of your options).</source>
        <translation>Set options automatically (might overwrite some of your options).</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="81"/>
        <source>Tool path</source>
        <translation>Tool path</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="82"/>
        <source>Path to the ClustalO tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation>Path to the ClustalO tool.&lt;p&gt;The default path can be set in the UGENE application settings.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="85"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="86"/>
        <source>Directory to store temporary files.</source>
        <translation>Directory to store temporary files.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="96"/>
        <source>Align with ClustalO</source>
        <translation>Align with ClustalO</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="97"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalO.&lt;p&gt;ClustalO is a general purpose multiple sequence alignment program for proteins.Visit &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; to learn more about it.</source>
        <translation>Aligns multiple sequence alignments (MSAs) supplied with ClustalO.&lt;p&gt;ClustalO is a general purpose multiple sequence alignment program for proteins.Visit &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; to learn more about it.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="184"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="216"/>
        <source>Aligned %1 with ClustalO</source>
        <translation>Aligned %1 with ClustalO</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWPrompter</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="188"/>
        <source> from %1</source>
        <translation> from %1</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="189"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</source>
        <translation>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWWorker</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA</source>
        <translation>Input MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA to process.</source>
        <translation>Input MSA to process.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="65"/>
        <source>ClustalW result MSA</source>
        <translation>ClustalW result MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="66"/>
        <source>The result of the ClustalW alignment.</source>
        <translation>The result of the ClustalW alignment.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="75"/>
        <source>Gap open penalty</source>
        <translation>Gap open penalty</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="76"/>
        <source>The penalty for opening a gap.</source>
        <translation>The penalty for opening a gap.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="77"/>
        <source>Gap extension penalty</source>
        <translation>Gap extension penalty</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="78"/>
        <source>The penalty for extending a gap.</source>
        <translation>The penalty for extending a gap.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="79"/>
        <source>Gap distance</source>
        <translation>Gap distance</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="80"/>
        <source>The gap separation penalty. Tries to decrease the chances of gaps being too close to each other.</source>
        <translation>The gap separation penalty. Tries to decrease the chances of gaps being too close to each other.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="81"/>
        <source>End gaps</source>
        <translation>End gaps</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="82"/>
        <source>The penalty for closing a gap.</source>
        <translation>The penalty for closing a gap.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="83"/>
        <source>Residue-specific gaps off</source>
        <translation>Residue-specific gaps off</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="84"/>
        <source>Residue-specific penalties are amino specific gap penalties that reduce or increase the gap opening penalties at each position in the alignment.</source>
        <translation>Residue-specific penalties are amino specific gap penalties that reduce or increase the gap opening penalties at each position in the alignment.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="85"/>
        <source>Hydrophilic gaps off</source>
        <translation>Hydrophilic gaps off</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="86"/>
        <source>Hydrophilic gap penalties are used to increase the chances of a gap within a run (5 or more residues) of hydrophilic amino acids.</source>
        <translation>Hydrophilic gap penalties are used to increase the chances of a gap within a run (5 or more residues) of hydrophilic amino acids.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="87"/>
        <source>Iteration type</source>
        <translation>Iteration type</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="88"/>
        <source>Alignment improvement iteration type. Can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;None - No iteration;&lt;/li&gt;                            &lt;li&gt;Tree - Iteration at each step of alignment process;&lt;/li&gt;                            &lt;li&gt;Alignment - Iteration only on final alignment.&lt;/li&gt;                            &lt;/ul&gt;</source>
        <translation>Alignment improvement iteration type. Can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;None - No iteration;&lt;/li&gt;                            &lt;li&gt;Tree - Iteration at each step of alignment process;&lt;/li&gt;                            &lt;li&gt;Alignment - Iteration only on final alignment.&lt;/li&gt;                            &lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="94"/>
        <source>Number of iterations</source>
        <translation>Number of iterations</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="95"/>
        <source>The maximum number of iterations to perform.</source>
        <translation>The maximum number of iterations to perform.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="96"/>
        <source>Weight matrix</source>
        <translation>Weight matrix</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="110"/>
        <source>Tool path</source>
        <translation>Tool path</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="111"/>
        <source>Path to the ClustalW tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation>Path to the ClustalW tool.&lt;p&gt;The default path can be set in the UGENE application settings.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="115"/>
        <source>Directory to store temporary files.</source>
        <translation>Directory to store temporary files.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="130"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalW.&lt;p&gt;ClustalW is a general purpose multiple sequence alignment program for DNA or proteins.Visit &lt;a href=&quot;http://www.clustal.org/&quot;&gt;http://www.clustal.org/&lt;/a&gt; to learn more about it.</source>
        <translation>Aligns multiple sequence alignments (MSAs) supplied with ClustalW.&lt;p&gt;ClustalW is a general purpose multiple sequence alignment program for DNA or proteins.Visit &lt;a href=&quot;http://www.clustal.org/&quot;&gt;http://www.clustal.org/&lt;/a&gt; to learn more about it.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="260"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="114"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="97"/>
        <source>For proteins it is a scoring table which describes the similarity of each amino acid to each other and can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - Appear to be the best available for carrying out database similarity (homology searches).                            &lt;li&gt;PAM - Have been extremely widely used since the late &apos;70s.                            &lt;li&gt;Gonnet - Derived using almost the same procedure as the PAM bit are much more up to date.                            &lt;li&gt;ID - Identitical matrix gives a score of 1.0 to two identical amino acids and a score of zero otherwise.                            &lt;/ul&gt;                            &lt;p&gt;                            For DNA it is the scores assigned to matches and mismatches. The values available are:                            &lt;ul&gt;                            &lt;li&gt;IUB - The default scoring matrix used by BESTFIT for the comparison of nucleic acid sequences. All matches score 1.9; all mismatches for IUB symbols score 0.                           &lt;li&gt;ClustalW - The previous system used by Clustal W, in which matches score 1.0 and mismatches score 0. All matches for IUB symbols also score 0.                            &lt;/ul&gt;</source>
        <translation>For proteins it is a scoring table which describes the similarity of each amino acid to each other and can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - Appear to be the best available for carrying out database similarity (homology searches).                            &lt;li&gt;PAM - Have been extremely widely used since the late &apos;70s.                            &lt;li&gt;Gonnet - Derived using almost the same procedure as the PAM bit are much more up to date.                            &lt;li&gt;ID - Identitical matrix gives a score of 1.0 to two identical amino acids and a score of zero otherwise.                            &lt;/ul&gt;                            &lt;p&gt;                            For DNA it is the scores assigned to matches and mismatches. The values available are:                            &lt;ul&gt;                            &lt;li&gt;IUB - The default scoring matrix used by BESTFIT for the comparison of nucleic acid sequences. All matches score 1.9; all mismatches for IUB symbols score 0.                           &lt;li&gt;ClustalW - The previous system used by Clustal W, in which matches score 1.0 and mismatches score 0. All matches for IUB symbols also score 0.                            &lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="129"/>
        <source>Align with ClustalW</source>
        <translation>Align with ClustalW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="289"/>
        <source>Aligned %1 with ClustalW</source>
        <translation>Aligned %1 with ClustalW</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOPrompter</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="217"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="222"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as gene list to conduct GO.</source>
        <translation>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as gene list to conduct GO.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="224"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOWorker</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="141"/>
        <source>Target genes</source>
        <translation>Target genes</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="142"/>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="147"/>
        <source>Gene list to identify over represented GO terms.</source>
        <translation>Gene list to identify over represented GO terms.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="146"/>
        <source>Conduct GO data</source>
        <translation>Conduct GO data</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="155"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="156"/>
        <source>The directory to store Conduct GO results.</source>
        <translation>The directory to store Conduct GO results.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="158"/>
        <source>Title</source>
        <translation>Title</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="159"/>
        <source>Title is used to name the output files - so make it meaningful.</source>
        <translation>Title is used to name the output files - so make it meaningful.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="161"/>
        <source>Gene Universe</source>
        <translation>Gene Universe</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="162"/>
        <source>Select a gene universe.</source>
        <translation>Select a gene universe.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="197"/>
        <source>Conduct GO</source>
        <translation>Conduct GO</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="198"/>
        <source>Given a list of genes, using Bioconductor (GO, GOstats) and  DAVID at NIH.</source>
        <translation>Given a list of genes, using Bioconductor (GO, GOstats) and  DAVID at NIH.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotPrompter</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="285"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="291"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions for conservation plot.</source>
        <translation>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions for conservation plot.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="292"/>
        <source> Conservations scores from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> Conservations scores from &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="293"/>
        <source> Outputs the result to &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> Outputs the result to &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotWorker</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="185"/>
        <source>Input regions</source>
        <translation>Input regions</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="186"/>
        <source>Regions (centered at peak summits for better performance).</source>
        <translation>Regions (centered at peak summits for better performance).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="190"/>
        <source>conservation_plot data</source>
        <translation>conservation_plot data</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="191"/>
        <source>Regions to plot the PhastCons scores profiles.</source>
        <translation>Regions to plot the PhastCons scores profiles.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="199"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="200"/>
        <source>File to store phastcons results (BMP).</source>
        <translation>File to store phastcons results (BMP).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="202"/>
        <source>Title</source>
        <translation>Title</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="203"/>
        <source>Title of the figure (--title).</source>
        <translation>Title of the figure (--title).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="205"/>
        <source>Label</source>
        <translation>Label</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="206"/>
        <source>Label of data in the figure (--bed-label).</source>
        <translation>Label of data in the figure (--bed-label).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="208"/>
        <source>Assembly version</source>
        <translation>Assembly version</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="209"/>
        <source>The directory to store phastcons scores (--phasdb).</source>
        <translation>The directory to store phastcons scores (--phasdb).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="211"/>
        <source>Window width</source>
        <translation>Window width</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="212"/>
        <source>Window width centered at middle of regions (-w).</source>
        <translation>Window width centered at middle of regions (-w).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="214"/>
        <source>Height</source>
        <translation>Height</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="215"/>
        <source>Height of plot (--height).</source>
        <translation>Height of plot (--height).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="217"/>
        <source>Width</source>
        <translation>Width</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="218"/>
        <source>Width of plot (--width).</source>
        <translation>Width of plot (--width).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="243"/>
        <source>Conservation plot file</source>
        <translation>Conservation plot file</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="265"/>
        <source>Build Conservation Plot</source>
        <translation>Build Conservation Plot</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="266"/>
        <source>Plots the PhastCons scores profiles.</source>
        <translation>Plots the PhastCons scores profiles.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffdiffWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="93"/>
        <source>Test for Diff. Expression with Cuffdiff</source>
        <translation>Test for Diff. Expression with Cuffdiff</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="107"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="108"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation>The base name of output directory. It could be modified with a suffix.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="111"/>
        <source>Time series analysis</source>
        <translation>Time series analysis</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="112"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, instructs Cuffdiff to analyze the provided samples as a time series, rather than testing for differences between all pairs of samples. Samples should be provided in increasing time order.</source>
        <translation>If set to &lt;i&gt;True&lt;/i&gt;, instructs Cuffdiff to analyze the provided samples as a time series, rather than testing for differences between all pairs of samples. Samples should be provided in increasing time order.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="118"/>
        <source>Upper quartile norm</source>
        <translation>Upper quartile norm</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="119"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, normalizes by the upper quartile of the number of fragments mapping to individual loci instead of the total number of sequenced fragments. This can improve robustness of differential expression calls for less abundant genes and transcripts.</source>
        <translation>If set to &lt;i&gt;True&lt;/i&gt;, normalizes by the upper quartile of the number of fragments mapping to individual loci instead of the total number of sequenced fragments. This can improve robustness of differential expression calls for less abundant genes and transcripts.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="125"/>
        <source>Hits norm</source>
        <translation>Hits norm</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="126"/>
        <source>Instructs how to count all fragments. &lt;i&gt;Total&lt;/i&gt; specifies to count all fragments, including those not compatible with any reference transcript, towards the number of mapped fragments used in the FPKM denominator. &lt;i&gt;Compatible&lt;/i&gt; specifies to use only compatible fragments. Selecting &lt;i&gt;Compatible&lt;/i&gt; is generally recommended in Cuffdiff to reduce certain types of bias caused by differential amounts of ribosomal reads which can create the impression of falsely differentially expressed genes.</source>
        <translation>Instructs how to count all fragments. &lt;i&gt;Total&lt;/i&gt; specifies to count all fragments, including those not compatible with any reference transcript, towards the number of mapped fragments used in the FPKM denominator. &lt;i&gt;Compatible&lt;/i&gt; specifies to use only compatible fragments. Selecting &lt;i&gt;Compatible&lt;/i&gt; is generally recommended in Cuffdiff to reduce certain types of bias caused by differential amounts of ribosomal reads which can create the impression of falsely differentially expressed genes.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="135"/>
        <source>Frag bias correct</source>
        <translation>Frag bias correct</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="136"/>
        <source>Providing the sequences your reads were mapped to instructs Cuffdiff to run bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation>Providing the sequences your reads were mapped to instructs Cuffdiff to run bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="141"/>
        <source>Multi read correct</source>
        <translation>Multi read correct</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="142"/>
        <source>Do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation>Do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="146"/>
        <source>Library type</source>
        <translation>Library type</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="147"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>Specifies RNA-Seq protocol.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="150"/>
        <source>Mask file</source>
        <translation>Mask file</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="151"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="159"/>
        <source>Min alignment count</source>
        <translation>Min alignment count</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="160"/>
        <source>The minimum number of alignments in a locus for needed to conduct significance testing on changes in that locus observed between samples. If no testing is performed, changes in the locus are deemed not significant, and the locus&apos; observed changes don&apos;t contribute to correction for multiple testing.</source>
        <translation>The minimum number of alignments in a locus for needed to conduct significance testing on changes in that locus observed between samples. If no testing is performed, changes in the locus are deemed not significant, and the locus&apos; observed changes don&apos;t contribute to correction for multiple testing.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="167"/>
        <source>FDR</source>
        <translation>FDR</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="168"/>
        <source>The allowed false discovery rate used in testing.</source>
        <translation>The allowed false discovery rate used in testing.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="171"/>
        <source>Max MLE iterations</source>
        <translation>Max MLE iterations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="172"/>
        <source>Sets the number of iterations allowed during maximum likelihood estimation of abundances.</source>
        <translation>Sets the number of iterations allowed during maximum likelihood estimation of abundances.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="176"/>
        <source>Emit count tables</source>
        <translation>Emit count tables</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="177"/>
        <source>Include information about the fragment counts, fragment count variances, and fitted variance model into the report.</source>
        <translation>Include information about the fragment counts, fragment count variances, and fitted variance model into the report.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="181"/>
        <source>Cuffdiff tool path</source>
        <translation>Cuffdiff tool path</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="182"/>
        <source>The path to the Cuffdiff external tool in UGENE.</source>
        <translation>The path to the Cuffdiff external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="185"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="186"/>
        <source>The directory for temporary files.</source>
        <translation>The directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="206"/>
        <source>Assembly</source>
        <translation>Assembly</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="207"/>
        <source>RNA-Seq reads assemblies</source>
        <translation>RNA-Seq reads assemblies</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="209"/>
        <source>Annotations</source>
        <translation>Annotations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="210"/>
        <source>Transcript annotations</source>
        <translation>Transcript annotations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="212"/>
        <source>Sample</source>
        <translation>Sample</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="213"/>
        <source>Sample name of assembly file</source>
        <translation>Sample name of assembly file</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="299"/>
        <source>Searches for significant changes in transcript expression, splicing and promoter use.</source>
        <translation>Searches for significant changes in transcript expression, splicing and promoter use.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CufflinksWorker</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="98"/>
        <source>Input reads</source>
        <translation>Input reads</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="99"/>
        <source>Input RNA-Seq read alignments.</source>
        <translation>Input RNA-Seq read alignments.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="102"/>
        <source>Output annotations</source>
        <translation>Output annotations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="103"/>
        <source>Assembled isoforms, estimated isoform-level expression values, and estimated gene-level expression values, produced by Cufflinks.</source>
        <translation>Assembled isoforms, estimated isoform-level expression values, and estimated gene-level expression values, produced by Cufflinks.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="116"/>
        <source>Isoform-level expression values</source>
        <translation>Isoform-level expression values</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="117"/>
        <source>A set of annotated regions</source>
        <translation>A set of annotated regions</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="134"/>
        <source>Assembly Transcripts with Cufflinks</source>
        <translation>Assembly Transcripts with Cufflinks</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="135"/>
        <source>Cufflinks accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</source>
        <translation>Cufflinks accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="143"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="144"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation>The base name of output directory. It could be modified with a suffix.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="147"/>
        <source>Reference annotation</source>
        <translation>Reference annotation</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="148"/>
        <source>Tells Cufflinks to use the supplied reference annotation to estimate isoform expression. Cufflinks will not assemble novel transcripts and the program will ignore alignments not structurally compatible with any reference transcript.</source>
        <translation>Tells Cufflinks to use the supplied reference annotation to estimate isoform expression. Cufflinks will not assemble novel transcripts and the program will ignore alignments not structurally compatible with any reference transcript.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="154"/>
        <source>RABT annotation</source>
        <translation>RABT annotation</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="155"/>
        <source>Tells Cufflinks to use the supplied reference annotation to guide Reference Annotation Based Transcript (RABT) assembly. Reference transcripts will be tiled with faux-reads to provide additional information in assembly. Output will include all reference transcripts as well as any novel genes and isoforms that are assembled.</source>
        <translation>Tells Cufflinks to use the supplied reference annotation to guide Reference Annotation Based Transcript (RABT) assembly. Reference transcripts will be tiled with faux-reads to provide additional information in assembly. Output will include all reference transcripts as well as any novel genes and isoforms that are assembled.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="162"/>
        <source>Library type</source>
        <translation>Library type</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="163"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>Specifies RNA-Seq protocol.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="166"/>
        <source>Mask file</source>
        <translation>Mask file</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="167"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="175"/>
        <source>Multi-read correct</source>
        <translation>Multi-read correct</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="176"/>
        <source>Tells Cufflinks to do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation>Tells Cufflinks to do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="180"/>
        <source>Min isoform fraction</source>
        <translation>Min isoform fraction</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="181"/>
        <source>After calculating isoform abundance for a gene, Cufflinks filters out transcripts that it believes are very low abundance, because isoforms expressed at extremely low levels often cannot reliably be assembled, and may even be artifacts of incompletely spliced precursors of processed transcripts. This parameter is also used to filter out introns that have far fewer spliced alignments supporting them.</source>
        <translation>After calculating isoform abundance for a gene, Cufflinks filters out transcripts that it believes are very low abundance, because isoforms expressed at extremely low levels often cannot reliably be assembled, and may even be artifacts of incompletely spliced precursors of processed transcripts. This parameter is also used to filter out introns that have far fewer spliced alignments supporting them.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="189"/>
        <source>Frag bias correct</source>
        <translation>Frag bias correct</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="190"/>
        <source>Providing Cufflinks with a multifasta file via this option instructs it to run the bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation>Providing Cufflinks with a multifasta file via this option instructs it to run the bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="195"/>
        <source>Pre-mRNA fraction</source>
        <translation>Pre-mRNA fraction</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="196"/>
        <source>Some RNA-Seq protocols produce a significant amount of reads that originate from incompletely spliced transcripts, and these reads can confound the assembly of fully spliced mRNAs. Cufflinks uses this parameter to filter out alignments that lie within the intronic intervals implied by the spliced alignments. The minimum depth of coverage in the intronic region covered by the alignment is divided by the number of spliced reads, and if the result is lower than this parameter value, the intronic alignments are ignored.</source>
        <translation>Some RNA-Seq protocols produce a significant amount of reads that originate from incompletely spliced transcripts, and these reads can confound the assembly of fully spliced mRNAs. Cufflinks uses this parameter to filter out alignments that lie within the intronic intervals implied by the spliced alignments. The minimum depth of coverage in the intronic region covered by the alignment is divided by the number of spliced reads, and if the result is lower than this parameter value, the intronic alignments are ignored. </translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="206"/>
        <source>Cufflinks tool path</source>
        <translation>Cufflinks tool path</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="207"/>
        <source>The path to the Cufflinks external tool in UGENE.</source>
        <translation>The path to the Cufflinks external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="210"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="211"/>
        <source>The directory for temporary files.</source>
        <translation>The directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="339"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation>Incorrect value of the library type parameter for Cufflinks!</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffmergeWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="63"/>
        <source>Merge Assemblies with Cuffmerge</source>
        <translation>Merge Assemblies with Cuffmerge</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="64"/>
        <source>Cuffmerge merges together several assemblies. It also handles running Cuffcompare for you, and automatically filters a number of transfrags that are probably artifacts. If you have a reference file available, you can provide it to Cuffmerge in order to gracefully merge input (e.g. novel) isoforms and known isoforms and maximize overall assembly quality.</source>
        <translation>Cuffmerge merges together several assemblies. It also handles running Cuffcompare for you, and automatically filters a number of transfrags that are probably artifacts. If you have a reference file available, you can provide it to Cuffmerge in order to gracefully merge input (e.g. novel) isoforms and known isoforms and maximize overall assembly quality.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="74"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="75"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation>The base name of output directory. It could be modified with a suffix.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="78"/>
        <source>Reference annotation</source>
        <translation>Reference annotation</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="79"/>
        <source>Merge the input assemblies together with this reference annotation.</source>
        <translation>Merge the input assemblies together with this reference annotation.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="83"/>
        <source>Reference sequence</source>
        <translation>Reference sequence</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="84"/>
        <source>The genomic DNA sequences for the reference. It is used to assist in classifying transfrags and excluding artifacts (e.g. repeats). For example, transcripts consisting mostly of lower-case bases are classified as repeats.</source>
        <translation>The genomic DNA sequences for the reference. It is used to assist in classifying transfrags and excluding artifacts (e.g. repeats). For example, transcripts consisting mostly of lower-case bases are classified as repeats.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="90"/>
        <source>Minimum isoform fraction</source>
        <translation>Minimum isoform fraction</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="91"/>
        <source>Discard isoforms with abundance below this.</source>
        <translation>Discard isoforms with abundance below this.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="94"/>
        <source>Cuffcompare tool path</source>
        <translation>Cuffcompare tool path</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="95"/>
        <source>The path to the Cuffcompare external tool in UGENE.</source>
        <translation>The path to the Cuffcompare external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="98"/>
        <source>Cuffmerge tool path</source>
        <translation>Cuffmerge tool path</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="99"/>
        <source>The path to the Cuffmerge external tool in UGENE.</source>
        <translation>The path to the Cuffmerge external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="102"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="103"/>
        <source>The directory for temporary files.</source>
        <translation>The directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="117"/>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="120"/>
        <source>Set of annotations</source>
        <translation>Set of annotations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="118"/>
        <source>Annotations for merging</source>
        <translation>Annotations for merging</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="121"/>
        <source>Merged annotations</source>
        <translation>Merged annotations</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="184"/>
        <source>Merges together supplied assemblies.</source>
        <translation>Merges together supplied assemblies.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqPrompter</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="71"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="72"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="74"/>
        <source>Removes adapter sequences %1.</source>
        <translation>Removes adapter sequences %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqTask</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="244"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="279"/>
        <source>Unknown file format: </source>
        <translation>Unknown file format: </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqWorker</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="92"/>
        <source>Cut Adapter</source>
        <translation>Cut Adapter</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="93"/>
        <source>Removes adapter sequences</source>
        <translation>Removes adapter sequences</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="97"/>
        <source>Input File</source>
        <translation>Input File</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="98"/>
        <source>Set of FASTQ reads files</source>
        <translation>Set of FASTQ reads files</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="99"/>
        <source>Output File</source>
        <translation>Output File</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="100"/>
        <source>Output FASTQ file(s)</source>
        <translation>Output FASTQ file(s)</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="113"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="114"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="118"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="119"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="121"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="122"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</source>
        <translation>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="124"/>
        <source>FASTA file with adapters</source>
        <translation>FASTA file with adapters</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="125"/>
        <source>A FASTA file with one or multiple sequences of adapter that were ligated to the 3&apos; end. The adapter itself and anything that follows is trimmed. If the adapter sequence ends with the &apos;$ character, the adapter is anchored to the end of the read and only found if it is a suffix of the read.</source>
        <translation>A FASTA file with one or multiple sequences of adapter that were ligated to the 3&apos; end. The adapter itself and anything that follows is trimmed. If the adapter sequence ends with the &apos;$ character, the adapter is anchored to the end of the read and only found if it is a suffix of the read.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="131"/>
        <source>FASTA file with reverse adapters</source>
        <translation>FASTA file with reverse adapters</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="132"/>
        <source>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end. If the adapter sequence starts with the character &apos;^&apos;, the adapter is &apos;anchored&apos;. An anchored adapter must appear in its entirety at the 5&apos; end of the read (it is a prefix of the read). A non-anchored adapter may appear partially at the 5&apos; end, or it may occur within the read. If it is found within a read, the sequence preceding the adapter is also trimmed. In all cases, the adapter itself is trimmed.</source>
        <translation>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end. If the adapter sequence starts with the character &apos;^&apos;, the adapter is &apos;anchored&apos;. An anchored adapter must appear in its entirety at the 5&apos; end of the read (it is a prefix of the read). A non-anchored adapter may appear partially at the 5&apos; end, or it may occur within the read. If it is found within a read, the sequence preceding the adapter is also trimmed. In all cases, the adapter itself is trimmed.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="143"/>
        <source>FASTA File with any end adapters</source>
        <translation>FASTA File with any end adapters</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="144"/>
        <source>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end or 3&apos; end.</source>
        <translation>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end or 3&apos; end.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="173"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="174"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="175"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::FastQCPrompter</name>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="76"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="77"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="79"/>
        <source>Builds FastQC report for file(s) %1.</source>
        <translation>Builds FastQC report for file(s) %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::FastQCWorker</name>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="86"/>
        <source>FastQC Quality Control</source>
        <translation>FastQC Quality Control</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="87"/>
        <source>Builds quality control reports.</source>
        <translation>Builds quality control reports.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="91"/>
        <source>Short reads</source>
        <translation>Short reads</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="92"/>
        <source>Short read data</source>
        <translation>Short read data</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="103"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="104"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="108"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="109"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="111"/>
        <source>List of adapters</source>
        <translation>List of adapters</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="112"/>
        <source>Specifies a non-default file which contains the list of adapter sequences which will be explicity searched against the library. The file must contain sets of named adapters in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</source>
        <translation>Specifies a non-default file which contains the list of adapter sequences which will be explicity searched against the library. The file must contain sets of named adapters in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="117"/>
        <source>List of contaminants</source>
        <translation>List of contaminants</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="118"/>
        <source>Specifies a non-default file which contains the list of contaminants to screen overrepresented sequences against. The file must contain sets of named contaminants in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</source>
        <translation>Specifies a non-default file which contains the list of contaminants to screen overrepresented sequences against. The file must contain sets of named contaminants in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="138"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="139"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="140"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="456"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="457"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="460"/>
        <source>%1 from %2 with bedtool genomecov.</source>
        <translation>%1 from %2 with bedtool genomecov.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovTask</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="702"/>
        <source>Unknown file format: </source>
        <translation>Unknown file format: </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="477"/>
        <source>Genome Coverage</source>
        <translation>Genome Coverage</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="478"/>
        <source>Calculates genome coverage using bedtools genomecov.</source>
        <translation>Calculates genome coverage using bedtools genomecov.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="482"/>
        <source>Input File</source>
        <translation>Input File</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="483"/>
        <source>Set of files to NGS slop</source>
        <translation>Set of files to NGS slop</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="484"/>
        <source>Output File</source>
        <translation>Output File</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="485"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="498"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="499"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="503"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="504"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="506"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="507"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</source>
        <translation>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="509"/>
        <source>Genome</source>
        <translation>Genome</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="510"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt;. (-g)</source>
        <translation>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt;. (-g)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="512"/>
        <source>Report mode</source>
        <translation>Report mode</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="513"/>
        <source>&lt;b&gt;%1 (%2)&lt;/b&gt; - %3 
&lt;b&gt;%4 (%5)&lt;/b&gt; - %6 
&lt;b&gt;%7 (%8)&lt;/b&gt; - %9 
&lt;b&gt;%10 (%11)&lt;/b&gt; - %12 
&lt;b&gt;%13 (%14)&lt;/b&gt; - %15 
</source>
        <translation>&lt;b&gt;%1 (%2)&lt;/b&gt; - %3 
&lt;b&gt;%4 (%5)&lt;/b&gt; - %6 
&lt;b&gt;%7 (%8)&lt;/b&gt; - %9 
&lt;b&gt;%10 (%11)&lt;/b&gt; - %12 
&lt;b&gt;%13 (%14)&lt;/b&gt; - %15</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="525"/>
        <source>Split</source>
        <translation>Split</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="526"/>
        <source>Treat split BAM or BED12 entries as distinct BED intervals when computing coverage. For BAM files, this uses the CIGAR N and D operations to infer the blocks for computing coverage. For BED12 files, this uses the BlockCount, BlockStarts, and BlockEnds fields (i.e., columns 10,11,12). (-split)</source>
        <translation>Treat  BAM or BED12 entries as distinct BED intervals when computing coverage. For BAM files, this uses the CIGAR  and  operations to infer the blocks for computing coverage. For BED12 files, this uses the BlockCount, BlockStarts, and BlockEnds fields (i.e., columns 10,11,12). (-split)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="528"/>
        <source>Strand</source>
        <translation>Strand</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="529"/>
        <source>Calculate coverage of intervals from a specific strand. With BED files, requires at least 6 columns (strand is column 6). (-strand)</source>
        <translation>Calculate coverage of intervals from a specific strand. With BED files, requires at least 6 columns (strand is column 6). (-strand)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="531"/>
        <source>5 prime</source>
        <translation>5 prime</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="532"/>
        <source>Calculate coverage of 5 positions (instead of entire interval). (-5)</source>
        <translation>Calculate coverage of 5&apos; positions (instead of entire interval). (-5)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="534"/>
        <source>3 prime</source>
        <translation>3 prime</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="535"/>
        <source>Calculate coverage of 3 positions (instead of entire interval). (-3)</source>
        <translation>Calculate coverage of 3&apos; positions (instead of entire interval). (-3)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="537"/>
        <source>Max</source>
        <translation>Max</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="538"/>
        <source>Combine all positions with a depth &gt;= max into a single bin in the histogram. (-max)</source>
        <translation>Combine all positions with a depth &gt;= max into a single bin in the histogram. (-max)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="540"/>
        <source>Scale</source>
        <translation>Scale</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="541"/>
        <source>Scale the coverage by a constant factor.Each coverage value is multiplied by this factor before being reported. Useful for normalizing coverage by, e.g., reads per million (RPM). Default is 1.0; i.e., unscaled. (-scale)</source>
        <translation>Scale the coverage by a constant factor.Each coverage value is multiplied by this factor before being reported. Useful for normalizing coverage by, e.g., reads per million (RPM). Default is 1.0; i.e., unscaled. (-scale)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="543"/>
        <source>Trackline</source>
        <translation>Trackline</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="544"/>
        <source>Adds a UCSC/Genome-Browser track line definition in the first line of the output. (-trackline)</source>
        <translation>Adds a UCSC/Genome-Browser track line definition in the first line of the output. (-trackline)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="546"/>
        <source>Trackopts</source>
        <translation>Trackopts</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="547"/>
        <source>Writes additional track line definition parameters in the first line. (-trackopts)</source>
        <translation>Writes additional track line definition parameters in the first line. (-trackopts)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="584"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="582"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="583"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GffreadPrompter</name>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="245"/>
        <source>Extract transcript sequences from the genomic sequence from &lt;u&gt;%1&lt;/u&gt; using transcripts from &lt;u&gt;%2&lt;/u&gt; and save them to the file %3.</source>
        <translation>Extract transcript sequences from the genomic sequence from &lt;u&gt;%1&lt;/u&gt; using transcripts from &lt;u&gt;%2&lt;/u&gt; and save them to the file %3.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSPrompter</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="570"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="577"/>
        <source>Uses &lt;u&gt;%1&lt;/u&gt; as treatment</source>
        <translation>Uses &lt;u&gt;%1&lt;/u&gt; as treatment</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="579"/>
        <source> and &lt;u&gt;%1&lt;/u&gt; as control</source>
        <translation> and &lt;u&gt;%1&lt;/u&gt; as control</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="582"/>
        <source> to call peaks.</source>
        <translation> to call peaks.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="584"/>
        <source> Outputs all files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation> Outputs all files to &lt;u&gt;%1&lt;/u&gt; directory</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="586"/>
        <source> and pileup with &lt;u&gt;%1&lt;/u&gt; span</source>
        <translation> and pileup with &lt;u&gt;%1&lt;/u&gt; span</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSWorker</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="194"/>
        <source>Bad model fold region: %1. Default region is used</source>
        <translation>Bad model fold region: %1. Default region is used</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="261"/>
        <source>Input control and treatment annotations are the same</source>
        <translation>Input control and treatment annotations are the same</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="275"/>
        <source>Treatment features</source>
        <translation>Treatment features</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="276"/>
        <source>ChIP-seq treatment features.</source>
        <translation>ChIP-seq treatment features.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="278"/>
        <source>Control features</source>
        <translation>Control features</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="279"/>
        <source>Control features (Optional).</source>
        <translation>Control features (Optional).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="284"/>
        <source>MACS data</source>
        <translation>MACS data</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="285"/>
        <source>ChIP-seq treatment features and control features (optional) to call peaks with MACS.</source>
        <translation>ChIP-seq treatment features and control features (optional) to call peaks with MACS.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="293"/>
        <source>Peak regions</source>
        <translation>Peak regions</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="294"/>
        <source>Peak locations. Typically used in gene association study like CEAS, or correlation calculation.</source>
        <translation>Peak locations. Typically used in gene association study like CEAS, or correlation calculation.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="296"/>
        <source>Peak summits</source>
        <translation>Peak summits</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="297"/>
        <source>Peak summits locations for every peaks. Typically used in DNA motif analysis or conservation check.</source>
        <translation>Peak summits locations for every peaks. Typically used in DNA motif analysis or conservation check.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="299"/>
        <source>Treatment fragments pileup (wiggle)</source>
        <translation>Treatment fragments pileup (wiggle)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="300"/>
        <source>Wiggle format files which can be imported to UCSC genome browser/GMOD/Affy IGB.</source>
        <translation>Wiggle format files which can be imported to UCSC genome browser/GMOD/Affy IGB.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="303"/>
        <source>MACS output data</source>
        <translation>MACS output data</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="304"/>
        <source>ChIP-seq peaks and summits. Pileup data (optional)</source>
        <translation>ChIP-seq peaks and summits. Pileup data (optional)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="317"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="318"/>
        <source>Directory to save MACS output files.</source>
        <translation>Directory to save MACS output files.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="320"/>
        <source>Name</source>
        <translation>Name</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="321"/>
        <source>The name string of the experiment. MACS will use this string NAME to create output files like &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; and so on. So please avoid any confliction between these filenames and your existing files (--name).</source>
        <translation>The name string of the experiment. MACS will use this string NAME to create output files like &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; and so on. So please avoid any confliction between these filenames and your existing files (--name).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="325"/>
        <source>Wiggle output</source>
        <translation>Wiggle output</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="326"/>
        <source>If this flag is on, MACS will store the fragment pileup in wiggle format for the whole genome data instead of for every chromosomes (--wig) (--single-profile).</source>
        <translation>If this flag is on, MACS will store the fragment pileup in wiggle format for the whole genome data instead of for every chromosomes (--wig) (--single-profile).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="328"/>
        <source>Wiggle space</source>
        <translation>Wiggle space</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="329"/>
        <source>By default, the resolution for saving wiggle files is 10 bps,i.e., MACS will save the raw tag count every 10 bps. You can change it along with &apos;--wig&apos; option (--space).</source>
        <translation>By default, the resolution for saving wiggle files is 10 bps,i.e., MACS will save the raw tag count every 10 bps. You can change it along with &apos;--wig&apos; option (--space).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="332"/>
        <source>Genome size (Mbp)</source>
        <translation>Genome size (Mbp)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="333"/>
        <source>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt; It&apos;s the mappable genome size or effective genome size which is defined as the genome size which can be sequenced. Because of the repetitive features on the chromosomes, the actual mappable genome size will be smaller than the original size, about 90% or 70% of the genome size (--gsize).</source>
        <translation>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt; It&apos;s the mappable genome size or effective genome size which is defined as the genome size which can be sequenced. Because of the repetitive features on the chromosomes, the actual mappable genome size will be smaller than the original size, about 90% or 70% of the genome size (--gsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="342"/>
        <source>P-value</source>
        <translation>P-value</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="343"/>
        <source>P-value cutoff. Default is 0.00001, for looser results, try 0.001 instead (--pvalue).</source>
        <translation>P-value cutoff. Default is 0.00001, for looser results, try 0.001 instead (--pvalue).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="345"/>
        <source>Q-value</source>
        <translation>Q-value</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="346"/>
        <source>Minimum FDR (q-value) cutoff for peak detection.</source>
        <translation>Minimum FDR (q-value) cutoff for peak detection.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="348"/>
        <source>Use model</source>
        <translation>Use model</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="349"/>
        <source>Whether or not to use MACS paired peaks model (--nomodel).</source>
        <translation>Whether or not to use MACS paired peaks model (--nomodel).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="351"/>
        <source>Model fold</source>
        <translation>Model fold</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="352"/>
        <source>Select the regions within MFOLD range of high-confidence enrichment ratio against. &lt;b&gt;Model fold&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is true, which is the foldchange to chose paired peaks to build paired peaks model. Users need to set a lower(smaller) and upper(larger) number for fold change so that MACS will only use the peaks within these foldchange range to build model (--mfold).</source>
        <translation>Select the regions within MFOLD range of high-confidence enrichment ratio against. &lt;b&gt;Model fold&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is true, which is the foldchange to chose paired peaks to build paired peaks model. Users need to set a lower(smaller) and upper(larger) number for fold change so that MACS will only use the peaks within these foldchange range to build model (--mfold).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="357"/>
        <source>Shift size</source>
        <translation>Shift size</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="358"/>
        <source>An arbitrary shift value used as a half of the fragment size when model is not built. &lt;b&gt;Shift size&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is false, which will represent the HALF of the fragment size of your sample. If your sonication and size selection size is 300 bps, after you trim out nearly 100 bps adapters, the fragment size is about 200 bps, so you can specify 100 here (--shiftsize).</source>
        <translation>An arbitrary shift value used as a half of the fragment size when model is not built. &lt;b&gt;Shift size&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is false, which will represent the HALF of the fragment size of your sample. If your sonication and size selection size is 300 bps, after you trim out nearly 100 bps adapters, the fragment size is about 200 bps, so you can specify 100 here (--shiftsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="363"/>
        <source>Keep duplicates</source>
        <translation>Keep duplicates</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="364"/>
        <source>It controls the MACS behavior towards duplicate tags at the exact same location -- the same coordination and the same strand. The default &lt;b&gt;auto&lt;/b&gt; option makes MACS calculate the maximum tags at the exact same location based on binomal distribution using 1e-5 as pvalue cutoff; and the &lt;b&gt;all&lt;/b&gt; option keeps every tags. If an &lt;b&gt;integer&lt;/b&gt; is given, at most this number of tags will be kept at the same location (--keep-dup).</source>
        <translation>It controls the MACS behavior towards duplicate tags at the exact same location -- the same coordination and the same strand. The default &lt;b&gt;auto&lt;/b&gt; option makes MACS calculate the maximum tags at the exact same location based on binomal distribution using 1e-5 as pvalue cutoff; and the &lt;b&gt;all&lt;/b&gt; option keeps every tags. If an &lt;b&gt;integer&lt;/b&gt; is given, at most this number of tags will be kept at the same location (--keep-dup).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="368"/>
        <source>Band width</source>
        <translation>Band width</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="369"/>
        <source>The band width which is used to scan the genome for model building. You can set this parameter as the sonication fragment size expected from wet experiment. Used only while building the shifting model (--bw).</source>
        <translation>The band width which is used to scan the genome for model building. You can set this parameter as the sonication fragment size expected from wet experiment. Used only while building the shifting model (--bw).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="373"/>
        <source>Extended fragment pileup</source>
        <translation>Extended fragment pileup</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="374"/>
        <source>Whether or not to generate extended fragment pileup, local lambda and score tracks at every bp.</source>
        <translation>Whether or not to generate extended fragment pileup, local lambda and score tracks at every bp.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="378"/>
        <source>Tag size (optional)</source>
        <translation>Tag size (optional)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="379"/>
        <source>Length of reads. Determined from first 10 reads if not specified (input &lt;b&gt;0&lt;/b&gt;) (--tsize).</source>
        <translation>Length of reads. Determined from first 10 reads if not specified (input &lt;b&gt;0&lt;/b&gt;) (--tsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="383"/>
        <source>Use lambda</source>
        <translation>Use lambda</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="384"/>
        <source>Whether to use local lambda model which can use the local bias at peak regions to throw out false positives (--nolambda).</source>
        <translation>Whether to use local lambda model which can use the local bias at peak regions to throw out false positives (--nolambda).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="386"/>
        <source>Small nearby region</source>
        <translation>Small nearby region</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="387"/>
        <source>The small nearby region in basepairs to calculate dynamic lambda. This is used to capture the bias near the peak summit region. Invalid if there is no control data (--slocal).</source>
        <translation>The small nearby region in basepairs to calculate dynamic lambda. This is used to capture the bias near the peak summit region. Invalid if there is no control data (--slocal).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="390"/>
        <source>Large nearby region</source>
        <translation>Large nearby region</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="391"/>
        <source>The large nearby region in basepairs to calculate dynamic lambda.  This is used to capture the surround bias (--llocal).</source>
        <translation>The large nearby region in basepairs to calculate dynamic lambda.  This is used to capture the surround bias (--llocal).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="394"/>
        <source>Auto bimodal</source>
        <translation>Auto bimodal</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="395"/>
        <source>Whether turn on the auto pair model process.If set, when MACS failed to build paired model, it will use the nomodelsettings, the Shift size parameter to shift and extend each tags (--on-auto).</source>
        <translation variants="yes">
            <lengthvariant>Whether turn on the auto pair model process.If set, when MACS failed to build paired model, it will use the nomodelsettings, the Shift size parameter to shift and extend each tags (--on-auto).</lengthvariant>
            <lengthvariant></lengthvariant>
        </translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="415"/>
        <source>If set, MACS will try to call broad peaks by linking nearby highly enriched regions.The linking region is controlled by another cutoff through Broad cutoff.The maximum linking region length is 4 times of d from MACS.</source>
        <translation>If set, MACS will try to call broad peaks by linking nearby highly enriched regions.The linking region is controlled by another cutoff through Broad cutoff. The maximum linking region length is 4 times of d from MACS. </translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="420"/>
        <source>Cutoff for broad region. This option is not available unless Broad is set.If P-value is set, this is a pvalue cutoff, otherwise, it&apos;s a qvalue cutoff.</source>
        <translation variants="yes">
            <lengthvariant>Cutoff for broad region. This option is not available unless Broad is set.If P-value is set, this is a pvalue cutoff, otherwise, it&apos;s a qvalue cutoff.</lengthvariant>
            <lengthvariant></lengthvariant>
            <lengthvariant></lengthvariant>
        </translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="399"/>
        <source>Scale to large</source>
        <translation>Scale to large</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="400"/>
        <source> When set, scale the small sample up to the bigger sample.By default, the bigger dataset will be scaled down towards the smaller dataset,which will lead to smaller p/qvalues and more specific results.Keep in mind that scaling down will bring down background noise more (--to-large).</source>
        <translation> When set, scale the small sample up to the bigger sample.By default, the bigger dataset will be scaled down towards the smaller dataset,which will lead to smaller p/qvalues and more specific results.Keep in mind that scaling down will bring down background noise more (--to-large).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="405"/>
        <source>Shift control</source>
        <translation>Shift control</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="406"/>
        <source>When set, control tags will be shifted just as ChIP tags according to their strand before the extension of d, slocal and llocal. By default, control tags are extended centeredat their current positions regardless of strand. You may consider to turn this option on whilecomparing two ChIP datasets of different condition but the same factor.</source>
        <translation>When set, control tags will be shifted just as ChIP tags according to their strand before the extension of d, slocal and llocal. By default, control tags are extended centeredat their current positions regardless of strand. You may consider to turn this option on whilecomparing two ChIP datasets of different condition but the same factor.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="411"/>
        <source>Half-extend</source>
        <translation>Half-extend</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="412"/>
        <source>When set, MACS extends 1/2 d size for each fragment centered at its middle point.</source>
        <translation>When set, MACS extends 1/2 d size for each fragment centered at its middle point.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="414"/>
        <source>Broad</source>
        <translation>Broad</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="419"/>
        <source>Broad cutoff</source>
        <translation>Broad cutoff</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="548"/>
        <source>Find Peaks with MACS</source>
        <translation>Find Peaks with MACS</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="549"/>
        <source>Performs peak calling for ChIP-Seq data.</source>
        <translation>Performs peak calling for ChIP-Seq data.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTPrompter</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="129"/>
        <source> from %1</source>
        <translation> from %1</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="131"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &lt;u&gt;&quot;MAFFT&quot;&lt;/u&gt;.</source>
        <translation>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &lt;u&gt;&quot;MAFFT&quot;&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTWorker</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="60"/>
        <source>Input MSA</source>
        <translation>Input MSA</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="61"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation>Multiple sequence alignment to be processed.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Multiple sequence alignment</source>
        <translation>Multiple sequence alignment</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Result of alignment.</source>
        <translation>Result of alignment.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="71"/>
        <source>Gap Open Penalty</source>
        <translation>Gap Open Penalty</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="72"/>
        <source>Gap Open Penalty.</source>
        <translation>Gap Open Penalty.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="73"/>
        <source>Offset</source>
        <translation>Offset</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="74"/>
        <source>Works like gap extension penalty.</source>
        <translation>Works like gap extension penalty.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="76"/>
        <source>Maximum number of iterative refinement.</source>
        <translation>Maximum number of iterative refinement.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="78"/>
        <source>External tool path.</source>
        <translation>External tool path.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="80"/>
        <source>Directory for temporary files.</source>
        <translation>Directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="173"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="75"/>
        <source>Max Iteration</source>
        <translation>Max Iteration</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="77"/>
        <source>Tool Path</source>
        <translation>Tool Path</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="79"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="88"/>
        <source>Align with MAFFT</source>
        <translation>Align with MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="202"/>
        <source>Aligned %1 with MAFFT</source>
        <translation>Aligned %1 with MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="89"/>
        <source>MAFFT is a multiple sequence alignment program for unix-like operating systems. &lt;p&gt;&lt;dfn&gt;It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of &amp;lt;&amp;#126;200 sequences), FFT-NS-2 (fast; for alignment of &amp;lt;&amp;#126;10,000 sequences), etc. &lt;/dfn&gt;&lt;/p&gt;</source>
        <translation>MAFFT is a multiple sequence alignment program for unix-like operating systems. &lt;p&gt;&lt;dfn&gt;It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of &amp;lt;&amp;#126;200 sequences), FFT-NS-2 (fast; for alignment of &amp;lt;&amp;#126;10,000 sequences), etc. &lt;/dfn&gt;&lt;/p&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GenePrompter</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="316"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="321"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions </source>
        <translation>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions </translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="322"/>
        <source> to annotate with genes nearby from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> to annotate with genes nearby from &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GeneWorker</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="194"/>
        <source>Treatment features</source>
        <translation>Treatment features</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="195"/>
        <source>Result peaks of MACS.</source>
        <translation>Result peaks of MACS.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="199"/>
        <source>Peak2gene data</source>
        <translation>Peak2gene data</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="200"/>
        <source>MACS peaks to get the refgenes near the peak summit/center.</source>
        <translation>MACS peaks to get the refgenes near the peak summit/center.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="208"/>
        <source>Gene regions</source>
        <translation>Gene regions</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="209"/>
        <source>Annotation for each gene, containing all the peaks nearby.</source>
        <translation>Annotation for each gene, containing all the peaks nearby.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="211"/>
        <source>Peak regions</source>
        <translation>Peak regions</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="212"/>
        <source>Annotation for each peak, containing all the genes nearby.</source>
        <translation>Annotation for each peak, containing all the genes nearby.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="215"/>
        <source>Peak2gene output data</source>
        <translation>Peak2gene output data</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="216"/>
        <source>Genes containing all the peaks nearby and peaks containing all the genes nearby.</source>
        <translation>Genes containing all the peaks nearby and peaks containing all the genes nearby.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="228"/>
        <source>Output type</source>
        <translation>Output type</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="229"/>
        <source>Select which type of genes need to output. &lt;b&gt;up&lt;/b&gt; for genes upstream to peak summit, &lt;b&gt;down&lt;/b&gt; for genes downstream to peak summit, &lt;b&gt;all&lt;/b&gt; for both &lt;b&gt;up&lt;/b&gt; and &lt;b&gt;down&lt;/b&gt; (--op).</source>
        <translation>Select which type of genes need to output. &lt;b&gt;up&lt;/b&gt; for genes upstream to peak summit, &lt;b&gt;down&lt;/b&gt; for genes downstream to peak summit, &lt;b&gt;all&lt;/b&gt; for both &lt;b&gt;up&lt;/b&gt; and &lt;b&gt;down&lt;/b&gt; (--op).</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="234"/>
        <source>Official gene symbols</source>
        <translation>Official gene symbols</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="235"/>
        <source>Output &lt;b&gt;official gene symbol&lt;/b&gt; instead of &lt;b&gt;refseq name&lt;/b&gt; (--symbol).</source>
        <translation>Output &lt;b&gt;official gene symbol&lt;/b&gt; instead of &lt;b&gt;refseq name&lt;/b&gt; (--symbol).</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="237"/>
        <source>Distance</source>
        <translation>Distance</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="238"/>
        <source>Set a number which unit is base. It will get the refGenes in n bases from peak center (--distance).</source>
        <translation>Set a number which unit is base. It will get the refGenes in n bases from peak center (--distance).</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="240"/>
        <source>Genome file</source>
        <translation>Genome file</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="241"/>
        <source>Select a genome file (sqlite3 file) to search refGenes (--genome).</source>
        <translation>Select a genome file (sqlite3 file) to search refGenes (--genome).</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="296"/>
        <source>Annotate Peaks with peak2gene</source>
        <translation>Annotate Peaks with peak2gene</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="297"/>
        <source>Gets refGenes near the ChIP regions identified by a peak-caller.</source>
        <translation>Gets refGenes near the ChIP regions identified by a peak-caller.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosComboBoxWithChecksDelegate</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="147"/>
        <source>Hint:</source>
        <translation>Hint:</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="147"/>
        <source>Use &apos;cistrome.xml&apos; to descrease the computation time. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation>Use &apos;cistrome.xml&apos; to descrease the computation time. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosPrompter</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="350"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="358"/>
        <source>Uses regions from &lt;u&gt;%1&lt;/u&gt; to find motifs enriched in them.</source>
        <translation>Uses regions from &lt;u&gt;%1&lt;/u&gt; to find motifs enriched in them.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="359"/>
        <source> Genome assembly: &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> Genome assembly: &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="361"/>
        <source> Finds de novo motifs.</source>
        <translation> Finds de novo motifs.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="364"/>
        <source> Uses &lt;u&gt;%1&lt;/u&gt; known motifs.</source>
        <translation> Uses &lt;u&gt;%1&lt;/u&gt; known motifs.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="367"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosWorker</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="233"/>
        <source>Input regions</source>
        <translation>Input regions</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="234"/>
        <source>Peak summits.</source>
        <translation>Peak summits.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="238"/>
        <source>SeqPos data</source>
        <translation>SeqPos data</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="239"/>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="331"/>
        <source>Finds motifs enriched in a set of regions.</source>
        <translation>Finds motifs enriched in a set of regions.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="247"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="248"/>
        <source>The directory to store seqpos results.</source>
        <translation>The directory to store seqpos results.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="250"/>
        <source>Genome assembly version</source>
        <translation>Genome assembly version</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="251"/>
        <source>UCSC database version (GENOME).</source>
        <translation>UCSC database version (GENOME).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="253"/>
        <source>De novo motifs</source>
        <translation>De novo motifs</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="254"/>
        <source>Run de novo motif search (-d).</source>
        <translation>Run de novo motif search (-d).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="256"/>
        <source>Motif database</source>
        <translation>Motif database</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="257"/>
        <source>Known motif collections (-m). Warning: computation time increases with selecting additional databases. It is recommended to use cistrome.xml. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation>Known motif collections (-m). Warning: computation time increases with selecting additional databases. It is recommended to use cistrome.xml. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="260"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="261"/>
        <source>Name of the output file which stores new motifs found during a de novo search (-n).</source>
        <translation>Name of the output file which stores new motifs found during a de novo search (-n).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="263"/>
        <source>Region width</source>
        <translation>Region width</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="264"/>
        <source>Width of the region to be scanned for motifs; depends on a resolution of assay (-w).</source>
        <translation>Width of the region to be scanned for motifs; depends on a resolution of assay (-w).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="266"/>
        <source>Pvalue cutoff</source>
        <translation>Pvalue cutoff</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="267"/>
        <source>Pvalue cutoff for the motif significance (-p).</source>
        <translation>Pvalue cutoff for the motif significance (-p).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="330"/>
        <source>Collect Motifs with SeqPos</source>
        <translation>Collect Motifs with SeqPos</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ShortReadsAlignerPrompter</name>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="340"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="346"/>
        <source>Aligns upstream oriented reads from &lt;u&gt;%1&lt;/u&gt; and downstream oriented reads from &lt;u&gt;%2&lt;/u&gt; </source>
        <translation>Aligns upstream oriented reads from &lt;u&gt;%1&lt;/u&gt; and downstream oriented reads from &lt;u&gt;%2&lt;/u&gt; </translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="348"/>
        <source>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="352"/>
        <source> to reference genome &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> to reference genome &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="69"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="70"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="72"/>
        <source>Increases the size of each feature in files from %1 with bedtool slop.</source>
        <translation>Increases the size of each feature in files from %1 with bedtool slop.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="89"/>
        <source>Slopbed</source>
        <translation>Slopbed</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="90"/>
        <source>Increases the size of each feature in files using bedtools slop.</source>
        <translation>Increases the size of each feature in files using bedtools slop.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="94"/>
        <source>Input File</source>
        <translation>Input File</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="95"/>
        <source>Set of files to bedtools slop</source>
        <translation>Set of files to bedtools slop</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="96"/>
        <source>Output File</source>
        <translation>Output File</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="97"/>
        <source>Output file</source>
        <translation>Output file</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="110"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="111"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="115"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="116"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="118"/>
        <source>Output file name</source>
        <translation>Output file name</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="119"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</source>
        <translation>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="121"/>
        <source>Genome</source>
        <translation>Genome</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="122"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt; (-g).</source>
        <translation>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt; (-g).</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="124"/>
        <source>Each direction increase</source>
        <translation>Each direction increase</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="125"/>
        <source>Increase the BED/GFF/VCF entry by the same number base pairs in each direction. If this parameter is used -l and -l are ignored. Enter 0 to disable. (-b)</source>
        <translation>Increase the BED/GFF/VCF entry by the same number base pairs in each direction. If this parameter is used -l and -l are ignored. Enter 0 to disable. (-b)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="127"/>
        <source>Substract from start</source>
        <translation>Substract from start</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="128"/>
        <source>The number of base pairs to subtract from the start coordinate. Enter 0 to disable. (-l)</source>
        <translation>The number of base pairs to subtract from the start coordinate. Enter 0 to disable. (-l)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="130"/>
        <source>Add to end</source>
        <translation>Add to end</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="131"/>
        <source>The number of base pairs to add to the end coordinate. Enter 0 to disable. (-r)</source>
        <translation>The number of base pairs to add to the end coordinate. Enter 0 to disable. (-r)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="133"/>
        <source>Strand-based</source>
        <translation>Strand-based</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="134"/>
        <source>Define -l and -r based on strand. For example. if used, -l 500 for a negative-stranded feature, it will add 500 bp to the end coordinate. (-s)</source>
        <translation>Define -l and -r based on strand. For example. if used, -l 500 for a negative-stranded feature, it will add 500 bp to the end coordinate. (-s)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="136"/>
        <source>As fraction</source>
        <translation>As fraction</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="137"/>
        <source>Define -l and -r as a fraction of the features length. E.g. if used on a 1000bp feature, -l 0.50, will add 500 bp upstream. (-pct)</source>
        <translation variants="yes">
            <lengthvariant>Define -l and -r as a fraction of the features length. E.g. if used on a 1000bp feature, -l 0.50, will add 500 bp </lengthvariant>
            <lengthvariant>upstream. (-pct)</lengthvariant>
        </translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="139"/>
        <source>Print header</source>
        <translation>Print header</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="140"/>
        <source>Print the header from the input file prior to results. (-header)</source>
        <translation>Print the header from the input file prior to results. (-header)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="142"/>
        <source>Filter start&gt;end fields</source>
        <translation>Filter start&gt;end fields</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="143"/>
        <source>Remove lines with start postion greater than end position</source>
        <translation>Remove lines with start postion greater than end position</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="179"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="177"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="178"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffPrompter</name>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="83"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="84"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> from &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="86"/>
        <source>Annotates and filters variations %1 with SnpEff.</source>
        <translation>Annotates and filters variations %1 with SnpEff.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffWorker</name>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="93"/>
        <source>SnpEff annotation and filtration</source>
        <translation>SnpEff annotation and filtration</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="94"/>
        <source>Annotates and filters variations with SnpEff.</source>
        <translation>Annotates and filters variations with SnpEff.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="98"/>
        <source>Variations</source>
        <translation>Variations</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="99"/>
        <source>Set of variations</source>
        <translation>Set of variations</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="100"/>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="101"/>
        <source>Annotated variations</source>
        <translation>Annotated variations</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="115"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="116"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="120"/>
        <source>Custom directory</source>
        <translation>Custom directory</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="121"/>
        <source>Select the custom output directory.</source>
        <translation>Select the custom output directory.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="123"/>
        <source>Input format</source>
        <translation>Input format</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="124"/>
        <source>Select the input format of variations.</source>
        <translation>Select the input format of variations.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="126"/>
        <source>Output format</source>
        <translation>Output format</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="127"/>
        <source>Select the format of annotated output files.</source>
        <translation>Select the format of annotated output files.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="129"/>
        <source>Genome</source>
        <translation>Genome</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="130"/>
        <source>Select the target genome. Genome data will be downloaded if it is not found.</source>
        <translation>Select the target genome. Genome data will be downloaded if it is not found.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="132"/>
        <source>Upstream/downstream length</source>
        <translation>Upstream/downstream length</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="133"/>
        <source>Upstream and downstream interval size. Eliminate any upstream and downstream effect by using 0 length</source>
        <translation>Upstream and downstream interval size. Eliminate any upstream and downstream effect by using 0 length</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="135"/>
        <source>Cannonical transcripts</source>
        <translation>Cannonical transcripts</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="136"/>
        <source>Use only cannonical transcripts</source>
        <translation>Use only cannonical transcripts</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="138"/>
        <source>HGVS nomenclature</source>
        <translation>HGVS nomenclature</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="139"/>
        <source>Annotate using HGVS nomenclature</source>
        <translation>Annotate using HGVS nomenclature</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="141"/>
        <source>Annotate Loss of function</source>
        <translation>Annotate Loss of function</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="142"/>
        <source>Annotate Loss of function (LOF) and Nonsense mediated decay (NMD)</source>
        <translation>Annotate Loss of function (LOF) and Nonsense mediated decay (NMD)</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="144"/>
        <source>Annotate TFBSs motifs</source>
        <translation>Annotate TFBSs motifs</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="145"/>
        <source>Annotate transcription factor binding site motifs (only available for latest GRCh37)</source>
        <translation>Annotate transcription factor binding site motifs (only available for latest GRCh37)</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="167"/>
        <source>Input file</source>
        <translation>Input file</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="168"/>
        <source>Workflow</source>
        <translation>Workflow</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="169"/>
        <source>Custom</source>
        <translation>Custom</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SpadesPrompter</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="341"/>
        <source>unset</source>
        <translation>unset</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="344"/>
        <source>Assemble reads from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation>Assemble reads from &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SpadesWorker</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="215"/>
        <source>SPAdes cannot recognize read pairs from the same file. Please, perform demultiplexing first.</source>
        <translation>SPAdes cannot recognize read pairs from the same file. Please, perform demultiplexing first.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="229"/>
        <source>URL of a file with reads</source>
        <translation>URL of a file with reads</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="230"/>
        <source>Input reads to be assembled.</source>
        <translation>Input reads to be assembled.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="232"/>
        <source>URL of a file with right pair reads</source>
        <translation>URL of a file with right pair reads</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="233"/>
        <source>Input right pair reads to be assembled.</source>
        <translation>Input right pair reads to be assembled.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="239"/>
        <source>SPAdes data</source>
        <translation>SPAdes data</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="240"/>
        <source>Input reads to be assembled with Spades.</source>
        <translation>Input reads to be assembled with Spades.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="248"/>
        <source>Scaffolds URL</source>
        <translation>Scaffolds URL</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="249"/>
        <source>Output scaffolds URL.</source>
        <translation>Output scaffolds URL.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="252"/>
        <source>SPAdes output data</source>
        <translation>SPAdes output data</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="253"/>
        <source>Output assembly files.</source>
        <translation>Output assembly files.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="264"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="265"/>
        <source>Directory to save Spades output files.</source>
        <translation>Directory to save Spades output files.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="268"/>
        <source>Number of threads</source>
        <translation>Number of threads</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="269"/>
        <source>Number of threads (-t).</source>
        <translation>Number of threads (-t).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="272"/>
        <source>Memory limit (GB)</source>
        <translation>Memory limit (GB)</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="273"/>
        <source>Memory limit (-m).</source>
        <translation>Memory limit (-m).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="276"/>
        <source>Dataset type</source>
        <translation>Dataset type</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="277"/>
        <source>Input dataset type.</source>
        <translation>Input dataset type.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="280"/>
        <source>Running mode</source>
        <translation>Running mode</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="281"/>
        <source>Running mode.</source>
        <translation>Running mode.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="284"/>
        <source>K-mers</source>
        <translation>K-mers</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="285"/>
        <source>k-mer sizes (-k).</source>
        <translation>k-mer sizes (-k).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="320"/>
        <source>Assemble genomes with SPAdes</source>
        <translation>Assemble genomes with SPAdes</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="321"/>
        <source>Performes assembly of input short reads.</source>
        <translation>Performes assembly of input short reads.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeePrompter</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="125"/>
        <source> from %1</source>
        <translation> from %1</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="127"/>
        <source>For each MSA&lt;u&gt;%1&lt;/u&gt;, build the alignment using &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt; and send it to output.</source>
        <translation>For each MSA&lt;u&gt;%1&lt;/u&gt;, build the alignment using &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt; and send it to output.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeeWorker</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Input MSA</source>
        <translation>Input MSA</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation>Multiple sequence alignment to be processed.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Multiple sequence alignment</source>
        <translation>Multiple sequence alignment</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Result of alignment.</source>
        <translation>Result of alignment.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="69"/>
        <source>Gap Open Penalty</source>
        <translation>Gap Open Penalty</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="70"/>
        <source>Gap Open Penalty. Must be negative, best matches get a score of 1000.</source>
        <translation>Gap Open Penalty. Must be negative, best matches get a score of 1000.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="71"/>
        <source>Gap Extension Penalty</source>
        <translation>Gap Extension Penalty</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="72"/>
        <source>Gap Extension Penalty. Positive values give rewards to gaps and prevent the alignment of unrelated segments.</source>
        <translation>Gap Extension Penalty. Positive values give rewards to gaps and prevent the alignment of unrelated segments.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="73"/>
        <source>Max Iteration</source>
        <translation>Max Iteration</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="74"/>
        <source>Number of iteration on the progressive alignment.&lt;br&gt;0 - no iteration, -1 - Nseq iterations.</source>
        <translation>Number of iteration on the progressive alignment.&lt;br&gt;0 - no iteration, -1 - Nseq iterations.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="76"/>
        <source>Tool Path</source>
        <translation>Tool Path</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="77"/>
        <source>External tool path.</source>
        <translation>External tool path.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="79"/>
        <source>Directory for temporary file.s</source>
        <translation>Directory for temporary files</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="170"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="78"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="87"/>
        <source>Align with T-Coffee</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="88"/>
        <source>T-Coffee is a multiple sequence alignment package. </source>
        <translation>T-Coffee is a multiple sequence alignment package. </translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="202"/>
        <source>Aligned %1 with T-Coffee</source>
        <translation>Aligned %1 with T-Coffee</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TopHatWorker</name>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="111"/>
        <location filename="../src/tophat/TopHatWorker.cpp" line="122"/>
        <source>Input reads</source>
        <translation>Input reads</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="112"/>
        <source>Input RNA-Seq reads</source>
        <translation>Input RNA-Seq reads</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="116"/>
        <source>TopHat output</source>
        <translation>TopHat output</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="117"/>
        <source>Accepted hits, junctions, insertions and deletions</source>
        <translation>Accepted hits, junctions, insertions and deletions</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="123"/>
        <source>TopHat input reads. Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation>TopHat input reads. Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="128"/>
        <source>Input reads url</source>
        <translation>Input reads url</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="129"/>
        <source>TopHat input reads url. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation>TopHat input reads url. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="133"/>
        <source>Input paired reads</source>
        <translation>Input paired reads</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="134"/>
        <source> Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads. Reads MUST appear in the same order as the *_1 reads.</source>
        <translation> Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads. Reads MUST appear in the same order as the *_1 reads.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="141"/>
        <source>Input paired reads url</source>
        <translation>Input paired reads url</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="142"/>
        <source>Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads.</source>
        <translation>Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="147"/>
        <source>Dataset name</source>
        <translation>Dataset name</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="148"/>
        <source>Use it only when sequences slot(or slots) is specified. Group input reads into chunks for several Tophat runs.
Set it empty if you want to run Tophat once for all input reads</source>
        <translation>Use it only when sequences slot(or slots) is specified. Group input reads into chunks for several Tophat runs.
Set it empty if you want to run Tophat once for all input reads</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="165"/>
        <source>Accepted hits</source>
        <translation>Accepted hits</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="166"/>
        <source>Accepted hits found by TopHat</source>
        <translation>Accepted hits found by TopHat</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="168"/>
        <source>Sample name</source>
        <translation>Sample name</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="169"/>
        <source>Sample name for running Cuffdiff</source>
        <translation>Sample name for running Cuffdiff</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="171"/>
        <source>Accepted hits url</source>
        <translation>Accepted hits url</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="172"/>
        <source>The url to the assembly file with the accepted hits</source>
        <translation>The url to the assembly file with the accepted hits</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="183"/>
        <source>Find Splice Junctions with TopHat</source>
        <translation>Find Splice Junctions with TopHat</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="184"/>
        <source>TopHat is a fast splice junction mapper for RNA-Seq reads. It aligns RNA-Seq reads to mammalian-sized genomes using the ultra high-throughput short read aligner Bowtie, and then analyzes the mapping results to identify splice junctions between exons.</source>
        <translation>TopHat is a fast splice junction mapper for RNA-Seq reads. It aligns RNA-Seq reads to mammalian-sized genomes using the ultra high-throughput short read aligner Bowtie, and then analyzes the mapping results to identify splice junctions between exons.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="192"/>
        <source>Output directory</source>
        <translation>Output directory</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="193"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation>The base name of output directory. It could be modified with a suffix.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="196"/>
        <source>Samples map</source>
        <translation>Samples map</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="197"/>
        <source>The map which divide all input datasets into samples. Every sample has the unique name.</source>
        <translation>The map which divide all input datasets into samples. Every sample has the unique name.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="200"/>
        <source>Bowtie index directory</source>
        <translation>Bowtie index directory</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="201"/>
        <source>The directory with the Bowtie index for the reference sequence.</source>
        <translation>The directory with the Bowtie index for the reference sequence.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="209"/>
        <source>Bowtie index basename</source>
        <translation>Bowtie index basename</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="210"/>
        <source>The basename of the Bowtie index for the reference sequence.</source>
        <translation>The basename of the Bowtie index for the reference sequence.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="223"/>
        <source>Mate inner distance</source>
        <translation>Mate inner distance</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="224"/>
        <source>The expected (mean) inner distance between mate pairs.</source>
        <translation>The expected (mean) inner distance between mate pairs.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="227"/>
        <source>Mate standard deviation</source>
        <translation>Mate standard deviation</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="228"/>
        <source>The standard deviation for the distribution on inner distances between mate pairs.</source>
        <translation>The standard deviation for the distribution on inner distances between mate pairs.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="231"/>
        <source>Library type</source>
        <translation>Library type</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="232"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>Specifies RNA-Seq protocol.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="235"/>
        <source>No novel junctions</source>
        <translation>No novel junctions</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="236"/>
        <source>Only look for reads across junctions indicated in the supplied GFF or junctions file. This parameter is ignored if &lt;i&gt;Raw junctions&lt;/i&gt; or &lt;i&gt;Known transcript file&lt;/i&gt; is not set.</source>
        <translation>Only look for reads across junctions indicated in the supplied GFF or junctions file. This parameter is ignored if &lt;i&gt;Raw junctions&lt;/i&gt; or &lt;i&gt;Known transcript file&lt;/i&gt; is not set.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="241"/>
        <source>Raw junctions</source>
        <translation>Raw junctions</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="242"/>
        <source>The list of raw junctions.</source>
        <translation>The list of raw junctions.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="245"/>
        <source>Known transcript file</source>
        <translation>Known transcript file</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="246"/>
        <source>A set of gene model annotations and/or known transcripts.</source>
        <translation>A set of gene model annotations and/or known transcripts.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="249"/>
        <source>Max multihits</source>
        <translation>Max multihits</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="250"/>
        <source>Instructs TopHat to allow up to this many alignments to the reference for a given read, and suppresses all alignments for reads with more than this many alignments.</source>
        <translation>Instructs TopHat to allow up to this many alignments to the reference for a given read, and suppresses all alignments for reads with more than this many alignments.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="255"/>
        <source>Segment length</source>
        <translation>Segment length</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="256"/>
        <source>Each read is cut up into segments, each at least this long. These segments are mapped independently.</source>
        <translation>Each read is cut up into segments, each at least this long. These segments are mapped independently.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="268"/>
        <source>Fusion search</source>
        <translation>Fusion search</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="269"/>
        <source>Turn on fusion mapping.</source>
        <translation>Turn on fusion mapping.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="272"/>
        <source>Transcriptome only</source>
        <translation>Transcriptome only</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="273"/>
        <source>Only align the reads to the transcriptome and report only those mappings as genomic mappings.</source>
        <translation>Only align the reads to the transcriptome and report only those mappings as genomic mappings.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="277"/>
        <source>Transcriptome max hits</source>
        <translation>Transcriptome max hits</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="278"/>
        <source>Maximum number of mappings allowed for a read, when aligned to the transcriptome (any reads found with more than this number of mappings will be discarded).</source>
        <translation>Maximum number of mappings allowed for a read, when aligned to the transcriptome (any reads found with more than this number of mappings will be discarded).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="283"/>
        <source>Prefilter multihits</source>
        <translation>Prefilter multihits</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="284"/>
        <source>When mapping reads on the transcriptome, some repetitive or low complexity reads that would be discarded in the context of the genome may appear to align to the transcript sequences and thus may end up reported as mapped to those genes only. This option directs TopHat to first align the reads to the whole genome in order to determine and exclude such multi-mapped reads (according to the value of the &lt;i&gt;Max multihits&lt;/i&gt; option).</source>
        <translation>When mapping reads on the transcriptome, some repetitive or low complexity reads that would be discarded in the context of the genome may appear to align to the transcript sequences and thus may end up reported as mapped to those genes only. This option directs TopHat to first align the reads to the whole genome in order to determine and exclude such multi-mapped reads (according to the value of the &lt;i&gt;Max multihits&lt;/i&gt; option).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="293"/>
        <source>Min anchor length</source>
        <translation>Min anchor length</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="294"/>
        <source>The &lt;i&gt;anchor length&lt;/i&gt;. TopHat will report junctions spanned by reads with at least this many bases on each side of the junction. Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved in spliced alignments is supported by at least one read with this many bases on each side.</source>
        <translation>The &lt;i&gt;anchor length&lt;/i&gt;. TopHat will report junctions spanned by reads with at least this many bases on each side of the junction. Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved in spliced alignments is supported by at least one read with this many bases on each side.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="302"/>
        <source>Splice mismatches</source>
        <translation>Splice mismatches</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="303"/>
        <source>The maximum number of mismatches that may appear in the &lt;i&gt;anchor&lt;/i&gt; region of a spliced alignment.</source>
        <translation>The maximum number of mismatches that may appear in the &lt;i&gt;anchor&lt;/i&gt; region of a spliced alignment.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="307"/>
        <source>Read mismatches</source>
        <translation>Read mismatches</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="308"/>
        <source>Final read alignments having more than these many mismatches are discarded.</source>
        <translation>Final read alignments having more than these many mismatches are discarded.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="312"/>
        <source>Segment mismatches</source>
        <translation>Segment mismatches</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="313"/>
        <source>Read segments are mapped independently, allowing up to this many mismatches in each segment alignment.</source>
        <translation>Read segments are mapped independently, allowing up to this many mismatches in each segment alignment.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="318"/>
        <source>Solexa 1.3 quals</source>
        <translation>Solexa 1.3 quals</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="319"/>
        <source>As of the Illumina GA pipeline version 1.3, quality scores are encoded in Phred-scaled base-64. Use this option for FASTQ files from pipeline 1.3 or later.</source>
        <translation>As of the Illumina GA pipeline version 1.3, quality scores are encoded in Phred-scaled base-64. Use this option for FASTQ files from pipeline 1.3 or later.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="324"/>
        <source>Bowtie version</source>
        <translation>Bowtie version</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="325"/>
        <source>Specifies which Bowtie version should be used.</source>
        <translation>Specifies which Bowtie version should be used.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="328"/>
        <source>Bowtie -n mode</source>
        <translation>Bowtie -n mode</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="329"/>
        <source>TopHat uses &lt;i&gt;-v&lt;/i&gt; in Bowtie for initial read mapping (the default), but with this option, &lt;i&gt;-n&lt;/i&gt; is used instead. Read segments are always mapped using &lt;i&gt;-v&lt;/i&gt; option.</source>
        <translation>TopHat uses &lt;i&gt;-v&lt;/i&gt; in Bowtie for initial read mapping (the default), but with this option, &lt;i&gt;-n&lt;/i&gt; is used instead. Read segments are always mapped using &lt;i&gt;-v&lt;/i&gt; option.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="335"/>
        <source>Bowtie tool path</source>
        <translation>Bowtie tool path</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="336"/>
        <source>The path to the Bowtie external tool.</source>
        <translation>The path to the Bowtie external tool.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="339"/>
        <source>SAMtools tool path</source>
        <translation>SAMtools tool path</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="340"/>
        <source>The path to the SAMtools tool. Note that the tool is available in the UGENE External Tool Package.</source>
        <translation>The path to the SAMtools tool. Note that the tool is available in the UGENE External Tool Package.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="344"/>
        <source>TopHat tool path</source>
        <translation>TopHat tool path</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="345"/>
        <source>The path to the TopHat external tool in UGENE.</source>
        <translation>The path to the TopHat external tool in UGENE.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="348"/>
        <source>Temporary directory</source>
        <translation>Temporary directory</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="349"/>
        <source>The directory for temporary files.</source>
        <translation>The directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="512"/>
        <source>Aligns RNA-seq reads to a reference and finds splice junctions.</source>
        <translation>Aligns RNA-seq reads to a reference and finds splice junctions.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="576"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation>Incorrect value of the library type parameter for Cufflinks!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="605"/>
        <source>Unrecognized value of the Bowtie mode option!</source>
        <translation>Unrecognized value of the Bowtie mode option!</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusPrompter</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="179"/>
        <source>Apply VCF variants from &lt;u&gt;%1&lt;/u&gt; to fasta file &lt;u&gt;%2&lt;/u&gt; and save consensus sequence to &lt;u&gt;%3&lt;/u&gt;.</source>
        <translation>Apply VCF variants from &lt;u&gt;%1&lt;/u&gt; to fasta file &lt;u&gt;%2&lt;/u&gt; and save consensus sequence to &lt;u&gt;%3&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusWorker</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="76"/>
        <source>Input fasta slot is empty</source>
        <translation>Input fasta slot is empty</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="79"/>
        <source>Input vcf slot is empty</source>
        <translation>Input vcf slot is empty</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="127"/>
        <source>Input FASTA and VCF</source>
        <translation>Input FASTA and VCF</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="128"/>
        <source>FASTA url</source>
        <translation>FASTA url</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="129"/>
        <source>VCF url</source>
        <translation>VCF url</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="136"/>
        <source>Fasta consensus url</source>
        <translation>Fasta consensus url</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="144"/>
        <source>Output FASTA consensus</source>
        <translation>Output FASTA consensus</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="145"/>
        <source>The url to the output file with the result consensus.</source>
        <translation>The url to the output file with the result consensus.</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="155"/>
        <source>Create VCF consensus</source>
        <translation>Create VCF consensus</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="156"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation>Apply VCF variants to a fasta file to create consensus sequence.</translation>
    </message>
</context>
<context>
    <name>U2::MACSSupport</name>
    <message>
        <location filename="../src/macs/MACSSupport.cpp" line="44"/>
        <source>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, which analyzes datagenerated by short read sequencers such as Solexa&apos;s Genome Analyzer. MACS empirically models the shiftsize of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.MACS also uses a dynamic Poisson distribution to effectively capture local biases in the genome,allowing for more robust predictions. MACS compares favorably to existing ChIP-Seq peak-finding algorithms, and is freely available.</source>
        <translation>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, which analyzes datagenerated by short read sequencers such as Solexa&apos;s Genome Analyzer. MACS empirically models the shiftsize of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.MACS also uses a dynamic Poisson distribution to effectively capture local biases in the genome,allowing for more robust predictions. MACS compares favorably to existing ChIP-Seq peak-finding algorithms, and is freely available.</translation>
    </message>
</context>
<context>
    <name>U2::MACSTask</name>
    <message>
        <location filename="../src/macs/MACSTask.cpp" line="53"/>
        <source>MACS peak calling</source>
        <translation>MACS peak calling</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTLogParser</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="418"/>
        <source>MAFFT has switched to the memsave mode. UGENE is unable to track its progress.</source>
        <translation>MAFFT has switched to the memsave mode. UGENE is unable to track its progress.</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupport</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="62"/>
        <source>&lt;i&gt;MAFFT&lt;/i&gt; is a multiple sequence alignment program for unix-like operating systems. </source>
        <translation>&lt;i&gt;MAFFT&lt;/i&gt; is a multiple sequence alignment program for unix-like operating systems. </translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="74"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="75"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportContext</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="131"/>
        <source>Align with MAFFT...</source>
        <translation>Align with MAFFT...</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="156"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="157"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="52"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="53"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="97"/>
        <source>MAFFT alignment started</source>
        <translation>MAFFT alignment started</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="118"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="127"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation>Subdir for temporary files exists. Can not remove this directory.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="132"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="145"/>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="147"/>
        <source>Can not open output file: </source>
        <translation>Can not open output file: </translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="148"/>
        <source> May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation> May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="185"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation>Output file &apos;%1&apos; not found</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="187"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="195"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="272"/>
        <source>Failed to apply the result of aligning with MAFFT: alignment object is not available!</source>
        <translation>Failed to apply the result of aligning with MAFFT: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="277"/>
        <source>MAFFT alignment successfully finished</source>
        <translation>MAFFT alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="296"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="84"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="85"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="100"/>
        <source>Open an alignment file</source>
        <translation>Open an alignment file</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="110"/>
        <source>Save an multiple alignment file</source>
        <translation>Save an multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="134"/>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="137"/>
        <source>Kalign with Align</source>
        <translation>Align with Kalign</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="135"/>
        <source>Input file is not set!</source>
        <translation>Input file is not set!</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="138"/>
        <source>Output file is not set!</source>
        <translation>Output file is not set!</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="330"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::MafftAddToAlignmentTask</name>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="65"/>
        <source>Align sequences to alignment task</source>
        <translation>Align sequences to alignment task</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="101"/>
        <source>Align sequences to an existing alignment by MAFFT started</source>
        <translation>Align sequences to an existing alignment by MAFFT started</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="176"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation>Output file &apos;%1&apos; not found</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="178"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="185"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="233"/>
        <source>MAFFT alignment successfully finished</source>
        <translation>MAFFT alignment successfully finished</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="202"/>
        <source>Generating output trees from MrBayes</source>
        <translation>Generating output trees from MrBayes</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="209"/>
        <source>Output file is not found</source>
        <translation>Output file is not found</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="236"/>
        <source>No trees are found</source>
        <translation>No trees are found</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesPrepareDataForCalculation</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="42"/>
        <source>Generating input file for MrBayes</source>
        <translation>Generating input file for MrBayes</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupport</name>
    <message>
        <location filename="../src/mrbayes/MrBayesSupport.cpp" line="57"/>
        <source>&lt;i&gt;MrBayes&lt;/i&gt; is a program for the Bayesian estimation of phylogeny.Bayesian inference of phylogeny is based upon a quantity called the posterior probability distribution of trees, which is the probability of a tree conditioned on the observations. The conditioning is accomplished using Bayes&apos;s theorem. The posterior probability distribution of trees is impossible to calculate analytically; instead, MrBayes uses a simulation technique called Markov chain Monte Carlo (or MCMC) to approximate the posterior probabilities of trees.</source>
        <translation>&lt;i&gt;MrBayes&lt;/i&gt; is a program for the Bayesian estimation of phylogeny.Bayesian inference of phylogeny is based upon a quantity called the posterior probability distribution of trees, which is the probability of a tree conditioned on the observations. The conditioning is accomplished using Bayes&apos;s theorem. The posterior probability distribution of trees is impossible to calculate analytically; instead, MrBayes uses a simulation technique called Markov chain Monte Carlo (or MCMC) to approximate the posterior probabilities of trees.</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupportTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="90"/>
        <source>MrBayes tree calculation</source>
        <translation>MrBayes tree calculation</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesWidget</name>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="64"/>
        <source>Substitution model</source>
        <translation>Substitution model</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="68"/>
        <source>Rate Matrix (fixed)</source>
        <translation>Rate Matrix (fixed)</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="174"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="175"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::PathLineEdit</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="670"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="672"/>
        <source>Select a file</source>
        <translation>Select a file</translation>
    </message>
</context>
<context>
    <name>U2::Peak2GeneSupport</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneSupport.cpp" line="45"/>
        <source>&lt;i&gt;peak2gene&lt;/i&gt; - Gets refGenes near the ChIP regions identified by a peak-caller</source>
        <translation>&lt;i&gt;peak2gene&lt;/i&gt; - Gets refGenes near the ChIP regions identified by a peak-caller</translation>
    </message>
</context>
<context>
    <name>U2::PerlSupport</name>
    <message>
        <location filename="../src/perl/PerlSupport.cpp" line="45"/>
        <source>Perl scripts interpreter</source>
        <translation>Perl scripts interpreter</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="196"/>
        <source>Generating output trees from PhyML</source>
        <translation>Generating output trees from PhyML</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="206"/>
        <source>Output file is not found</source>
        <translation>Output file is not found</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="232"/>
        <source>No trees are found</source>
        <translation>No trees are found</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="237"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="243"/>
        <source>No result tree in PhyML output</source>
        <translation>No result tree in PhyML output</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLLogParser</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="174"/>
        <source>PhyML finished with error</source>
        <translation>PhyML finished with error</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="191"/>
        <source>UGENE internal error</source>
        <translation>UGENE internal error</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLPrepareDataForCalculation</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="41"/>
        <source>Generating input file for PhyML</source>
        <translation>Generating input file for PhyML</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="68"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="72"/>
        <source>Internal UGENE error</source>
        <translation>Internal UGENE error</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="76"/>
        <source>Can not open tmp file</source>
        <translation>Can not open tmp file</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupport</name>
    <message>
        <location filename="../src/phyml/PhyMLSupport.cpp" line="49"/>
        <source>&lt;i&gt;PhyML&lt;/i&gt; is a simple, fast, and accurate algorithm to estimate large phylogenies by maximum likelihood</source>
        <translation>&lt;i&gt;PhyML&lt;/i&gt; is a simple, fast, and accurate algorithm to estimate large phylogenies by maximum likelihood</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupportTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="95"/>
        <source>PhyML tree calculation</source>
        <translation>PhyML tree calculation</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="145"/>
        <source>UGENE internal error</source>
        <translation>UGENE internal error</translation>
    </message>
</context>
<context>
    <name>U2::PhyMlWidget</name>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="169"/>
        <source>Open an alignment file</source>
        <translation>Open an alignment file</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="221"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="222"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForCAP3Task</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="245"/>
        <source>Failed to initialize sequence writer.</source>
        <translation>Failed to initialize sequence writer.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="272"/>
        <source>Failed to write sequence %1</source>
        <translation>Failed to write sequence %1</translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForSpideyTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="194"/>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="207"/>
        <source>Failed to write DNA sequence  %1</source>
        <translation>Failed to write DNA sequence  %1</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleDjangoSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="106"/>
        <source>: Python module for the %1 tool</source>
        <translation>: Python module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleNumpySupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="115"/>
        <source>: Python module for the %1 tool</source>
        <translation>: Python module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="95"/>
        <source>Python module is not installed. Install module or set path to another Python scripts interpreter with installed module in the External Tools settings</source>
        <translation>Python module is not installed. Install module or set path to another Python scripts interpreter with installed module in the External Tools settings</translation>
    </message>
</context>
<context>
    <name>U2::PythonSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="60"/>
        <source>Python scripts interpreter</source>
        <translation>Python scripts interpreter</translation>
    </message>
</context>
<context>
    <name>U2::RModuleCelegansdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="151"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleDrosophila2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="157"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleGodbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="115"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleGostatsSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="109"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133adbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="121"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133bdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="127"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133plus2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="133"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu95av2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="139"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleMouse430a2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="145"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgceegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="175"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgdmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="181"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrghsegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="163"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgmmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="169"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleSeqlogoSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="187"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript module for the %1 tool</translation>
    </message>
</context>
<context>
    <name>U2::RModuleSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="95"/>
        <source>R module is not installed. Install module or set path to another R scripts interpreter with installed module in the External Tools settings</source>
        <translation>R module is not installed. Install module or set path to another R scripts interpreter with installed module in the External Tools settings</translation>
    </message>
</context>
<context>
    <name>U2::RSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="59"/>
        <source>Rscript interpreter</source>
        <translation>Rscript interpreter</translation>
    </message>
</context>
<context>
    <name>U2::RunCap3AndOpenResultTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="137"/>
        <source>CAP3 run and open result task</source>
        <translation>CAP3 run and open result task</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="146"/>
        <source>Invalid CAP3 task</source>
        <translation>Invalid CAP3 task</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="162"/>
        <source>Project loader is NULL</source>
        <translation>Project loader is NULL</translation>
    </message>
</context>
<context>
    <name>U2::SaveAlignmentTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="60"/>
        <source>Export alignment to &apos;%1&apos;</source>
        <translation>Export alignment to &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="64"/>
        <source>An alignment is empty</source>
        <translation>An alignment is empty</translation>
    </message>
</context>
<context>
    <name>U2::SaveMSA2SequencesTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="91"/>
        <source>Export alignment to sequence: %1</source>
        <translation>Export alignment to sequence: %1</translation>
    </message>
</context>
<context>
    <name>U2::SeqPosSupport</name>
    <message>
        <location filename="../src/seqpos/SeqPosSupport.cpp" line="46"/>
        <source>&lt;i&gt;SeqPos&lt;/i&gt; - Finds motifs enriched in a set of regions.</source>
        <translation>&lt;i&gt;SeqPos&lt;/i&gt; - Finds motifs enriched in a set of regions.</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffSupport</name>
    <message>
        <location filename="../src/snpeff/SnpEffSupport.cpp" line="48"/>
        <source>&lt;i&gt;SnpEff&lt;/i&gt;: Genetic variant annotation and effect prediction toolbox.</source>
        <translation>&lt;i&gt;SnpEff&lt;/i&gt;: Genetic variant annotation and effect prediction toolbox.</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffTask</name>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="171"/>
        <source>SNPEff dataDir is not initialized.</source>
        <translation>SNPEff dataDir is not initialized.</translation>
    </message>
</context>
<context>
    <name>U2::SpadesSupport</name>
    <message>
        <location filename="../src/spades/SpadesSupport.cpp" line="43"/>
        <source>&lt;i&gt;SPAdes&lt;/i&gt; - St. Petersburg genome assembler - is intended for both standard isolates and single-cell MDA bacteria assemblies. Official site: http://bioinf.spbau.ru/spades</source>
        <translation>&lt;i&gt;SPAdes&lt;/i&gt; - St. Petersburg genome assembler - is intended for both standard isolates and single-cell MDA bacteria assemblies. Official site: http://bioinf.spbau.ru/spades</translation>
    </message>
</context>
<context>
    <name>U2::SpadesTask</name>
    <message>
        <location filename="../src/spades/SpadesTask.cpp" line="61"/>
        <source>Directory does not exist: </source>
        <translation>Directory does not exist: </translation>
    </message>
</context>
<context>
    <name>U2::SpideyAlignmentTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="83"/>
        <source>Output file not found</source>
        <translation>Output file not found</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="88"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="99"/>
        <source>Failed to open result file %1</source>
        <translation>Failed to open result file %1</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupport</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="72"/>
        <source>&lt;i&gt;Spidey&lt;/i&gt; is mRNA-to-DNA alignment program.                    &lt;br&gt;Binaries can be downloaded from http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</source>
        <translation>&lt;i&gt;Spidey&lt;/i&gt; is mRNA-to-DNA alignment program.                    &lt;br&gt;Binaries can be downloaded from http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="74"/>
        <source>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</source>
        <translation>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="75"/>
        <source>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</source>
        <translation>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="76"/>
        <source>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</source>
        <translation>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportContext</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="115"/>
        <source>Align sequence to mRNA</source>
        <translation>Align sequence to mRNA</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="142"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="143"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="248"/>
        <source>Failed to align mRNA to genomic sequence: no alignment is found.</source>
        <translation>Failed to align mRNA to genomic sequence: no alignment is found.</translation>
    </message>
</context>
<context>
    <name>U2::TBlastXPlusSupportTask</name>
    <message>
        <location filename="../src/blast_plus/TBlastXPlusSupportTask.cpp" line="61"/>
        <source>Unexpected settings combination</source>
        <translation>Unexpected settings combination</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupport</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="66"/>
        <source>&lt;i&gt;T-Coffee&lt;/i&gt; is a multiple sequence alignment package.</source>
        <translation>&lt;i&gt;T-Coffee&lt;/i&gt; is a multiple sequence alignment package.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="76"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="77"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportContext</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="134"/>
        <source>Align with T-Coffee...</source>
        <translation>Align with T-Coffee...</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="159"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Path for %1 tool not selected.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="160"/>
        <source>Do you want to select it now?</source>
        <translation>Do you want to select it now?</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="52"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="53"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="98"/>
        <source>Unsupported alphabet: %1</source>
        <translation>Unsupported alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="102"/>
        <source>T-Coffee alignment started</source>
        <translation>T-Coffee alignment started</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="123"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="132"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation>Subdir for temporary files exists. Can not remove this directory.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="137"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="182"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="184"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="192"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="273"/>
        <source>Failed to apply the result of TCoffee: alignment object is not available!</source>
        <translation>Failed to apply the result of TCoffee: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="278"/>
        <source>T-Coffee alignment successfully finished</source>
        <translation>T-Coffee alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="291"/>
        <source>Can not remove directory for temporary files.</source>
        <translation>Can not remove directory for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="83"/>
        <source>Align</source>
        <translation>Align</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="84"/>
        <source>Cancel</source>
        <translation>Cancel</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="98"/>
        <source>Open an alignment file</source>
        <translation>Open an alignment file</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="108"/>
        <source>Save an multiple alignment file</source>
        <translation>Save an multiple alignment file</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="132"/>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="135"/>
        <source>Kalign with Align</source>
        <translation>Align with Kalign</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="133"/>
        <source>Input file is not set!</source>
        <translation>Input file is not set!</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="136"/>
        <source>Output file is not set!</source>
        <translation>Output file is not set!</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="325"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::TabixSupport</name>
    <message>
        <location filename="../src/samtools/TabixSupport.cpp" line="49"/>
        <source>&lt;i&gt;Tabix&lt;/i&gt; is a generic indexer for TAB-delimited genome position files</source>
        <translation>&lt;i&gt;Tabix&lt;/i&gt; is a generic indexer for TAB-delimited genome position files</translation>
    </message>
</context>
<context>
    <name>U2::TabixSupportTask</name>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="37"/>
        <source>Generate index with Tabix task</source>
        <translation>Generate index with Tabix task</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="47"/>
        <source>Tabix indexing started</source>
        <translation>Tabix indexing started</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="50"/>
        <source>Input file &apos;%1&apos; is already bgzipped</source>
        <translation>Input file &apos;%1&apos; is already bgzipped</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="61"/>
        <source>Saving data to file &apos;%1&apos;</source>
        <translation>Saving data to file &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::TopHatSupportTask</name>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="53"/>
        <source>Running TopHat task</source>
        <translation>Running TopHat task</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="143"/>
        <source>An unexpected error has occurred during preparing the TopHat task!</source>
        <translation>An unexpected error has occurred during preparing the TopHat task!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="144"/>
        <source>Preparing TopHatSupportTask internal error: unable to get a sequence object!</source>
        <translation>Preparing TopHatSupportTask internal error: unable to get a sequence object!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="280"/>
        <source>There are no accepted hits in the result</source>
        <translation>There are no accepted hits in the result</translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupport</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupport.cpp" line="45"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation>Apply VCF variants to a fasta file to create consensus sequence.</translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupportTask</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="40"/>
        <source>Create Vcf Consensus</source>
        <translation>Create Vcf Consensus</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="50"/>
        <source>VcfConsensus started</source>
        <translation>VcfConsensus started</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="52"/>
        <source>AppSettings is NULL</source>
        <translation>AppSettings is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="54"/>
        <source>UserAppsSettings is NULL</source>
        <translation>UserAppsSettings is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="56"/>
        <source>Temporary directory is not set!</source>
        <translation>Temporary directory is not set!</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="61"/>
        <source>Can not create directory for temporary files.</source>
        <translation>Can not create directory for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="65"/>
        <source>Saving temporary data to file &apos;%1&apos;</source>
        <translation>Saving temporary data to file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="84"/>
        <source>ExternalToolRegistry is NULL</source>
        <translation>ExternalToolRegistry is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="88"/>
        <source>There is no VcfConsensus external tool registered</source>
        <translation>There is no VcfConsensus external tool registered</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="89"/>
        <source>There is no Tabix external tool registered</source>
        <translation>There is no Tabix external tool registered</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="111"/>
        <source>Trying to get path of NULL external tool</source>
        <translation>Trying to get path of NULL external tool</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="115"/>
        <source>Path to %1</source>
        <translation>Path to %1</translation>
    </message>
</context>
<context>
    <name>UserAppsSettings</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="135"/>
        <source>Temporary UGENE dir is empty</source>
        <translation>Temporary UGENE dir is empty</translation>
    </message>
</context>
</TS>
