<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="ru_RU">
<context>
    <name>BlastAllSupportDialog</name>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="20"/>
        <source>Request to Local BLAST Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="39"/>
        <source>General options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="47"/>
        <source>Select search:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="55"/>
        <source>blastn</source>
        <extracomment>qwe</extracomment>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="60"/>
        <source>blastp</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="75"/>
        <source>blastx</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="80"/>
        <source>tblastn</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="85"/>
        <source>tblastx</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="93"/>
        <source>Search for short, nearly exact matches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="100"/>
        <source>Expectation value </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="107"/>
        <source>Megablast</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="114"/>
        <source>Max hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="124"/>
        <source>Off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="162"/>
        <source>Base name for BLAST DB files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="228"/>
        <source>Number of CPUs being used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="266"/>
        <source>Advanced options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="280"/>
        <source>Word size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="287"/>
        <source>Gap costs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="476"/>
        <source>Match scores</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="422"/>
        <source>1 -3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="427"/>
        <source>1 -4</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="432"/>
        <source>1 -2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="437"/>
        <source>1 -1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="442"/>
        <source>2 -7</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="447"/>
        <source>2 -5</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="452"/>
        <source>2 -3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="457"/>
        <source>4 -5</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="462"/>
        <source>5 -4</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="313"/>
        <source>Matrix</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="65"/>
        <source>gpu-blastp</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="70"/>
        <source>cuda-blastp</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="176"/>
        <source>Database path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="200"/>
        <source>Select
a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="321"/>
        <source>BLOSUM62</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="326"/>
        <source>BLOSUM45</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="331"/>
        <source>BLOSUM80</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="336"/>
        <source>PAM30</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="341"/>
        <source>PAM70</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="349"/>
        <source>Service</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="357"/>
        <source>plain</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="362"/>
        <source>psi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="367"/>
        <source>phi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="378"/>
        <source>PHI pattern</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="483"/>
        <source>Threshold for extending hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="486"/>
        <source>Threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="513"/>
        <source>Filters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="519"/>
        <source>Low complexity filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="526"/>
        <source>Human repeats filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="536"/>
        <source>Masks</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="542"/>
        <source>Mask for lookup table only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="552"/>
        <source>Mask lower case letters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="578"/>
        <source>Extension options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="586"/>
        <source>X dropoff value (in bits)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="592"/>
        <source>For gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="606"/>
        <source>For ungapped extensions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="613"/>
        <source>For final gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="637"/>
        <source>Multiple Hits window size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="649"/>
        <source>Perform gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="704"/>
        <source>Restore to Default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="714"/>
        <source>Search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="727"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BlastDBCmdDialog</name>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="20"/>
        <source>Fetch sequence from BLAST database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="32"/>
        <source>Entery query ID:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="42"/>
        <source>Select database :</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="58"/>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="112"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="67"/>
        <source>Type of file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="73"/>
        <source>nucleotide</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="83"/>
        <source>protein</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="96"/>
        <source>Output path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="124"/>
        <source>Add to project</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="152"/>
        <source>Fetch</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="159"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>Bowtie2Settings</name>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="41"/>
        <source>Parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="60"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="76"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, Bowtie2 performs end-to-end read alignment. That is, it searches for alignments involving all of the read characters. This is also called an &amp;quot;untrimmed&amp;quot; or &amp;quot;unclipped&amp;quot; alignment.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the --local option is specified, Bowtie2 performs local read alignment. In this mode, Bowtie2 might &amp;quot;trim&amp;quot; or &amp;quot;clip&amp;quot; some read characters from one or both ends of the alignment if doing so maximizes the alignment score.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="63"/>
        <source>Mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="86"/>
        <source>--end-to-end</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="91"/>
        <source>--local</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="106"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="174"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Pads&amp;quot; dynamic programming problems by specified number of columns on either side to allow gaps.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="109"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="119"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="187"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often much slower) but increases sensitivity.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="132"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="197"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="148"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="252"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Disallow gaps within specified number of positions of the beginning or end of the read.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="161"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="229"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Use specified value as the seed for pseudo-random number generator.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="164"/>
        <source>Seed (--seed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="190"/>
        <source>Number of mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="200"/>
        <source>Threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="210"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="239"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="242"/>
        <source>Seed length (--L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="255"/>
        <source>Disallow gaps (--gbar)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="284"/>
        <source>Flags</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="296"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments for the constituent mates. This is called &amp;quot;mixed mode.&amp;quot; To disable mixed mode, set this option.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 runs a little faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="299"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="306"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="309"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="316"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the forward reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="319"/>
        <source>No forward orientation (--nofw)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="326"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the reverse-complement reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="329"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="336"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="339"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="346"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the other in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="349"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BowtieBuildSettings</name>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="26"/>
        <source>Colorspace</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BowtieSettings</name>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="41"/>
        <source>Parameters</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="110"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;not just in the &amp;quot;seed&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The default is 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;rounding can be disabled with --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="121"/>
        <source>Maq error (--maqerr)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="144"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The &amp;quot;seed length&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;i.e., the number of bases on the high-quality end of the read to which the -n ceiling applies.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="176"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The maximum insert size for valid paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;E.g. if -X 100 is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;that alignment is considered valid (as long as -I is also satisfied). &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;A 61-bp gap would not be valid in that case. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If trimming options -3 or -5 are also used, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;the -X constraint is applied with respect to the untrimmed mates, not the trimmed mates.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Default: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="215"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="246"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="60"/>
        <source>Mode:  </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="67"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the -n option is specified (which is the default), &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie determines which alignments are valid according to the following policy, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;which is similar to Maq&apos;s default policy.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="82"/>
        <source>-n mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="87"/>
        <source>-v mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="153"/>
        <source>Seed length (--seedlen)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="189"/>
        <source>Maximum of backtracks (--maxbts)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="223"/>
        <source>Descriptors memory usage (--chunkmbs)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="249"/>
        <source>Seed (--seed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="266"/>
        <source>Threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="300"/>
        <source>Flags</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="312"/>
        <source>Colorspace (--color)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="329"/>
        <source>No Maq rounding (--nomaqround)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="339"/>
        <source>No forward orientation (--nofw)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="349"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="369"/>
        <source>Try as hard (--tryhard)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="383"/>
        <source>Best alignments (--best)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="393"/>
        <source>All alignments (--all)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="319"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq accepts quality values in the Phred quality scale, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but internally rounds values to the nearest 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;with a maximum of 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="336"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="346"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="356"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Try as hard as possible to find valid alignments when they exist, including paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This is equivalent to specifying very high values for the --maxbts and --pairtries options.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is generally much slower than the default settings, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but can be useful for certain problems. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is slower when (a) the reference is very repetitive, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) the reads are low quality, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;or (c) not many reads have valid alignments.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="376"/>
        <source>Make Bowtie guarantee that reported singleton alignments are &quot;best&quot; in terms of stratum
(i.e. number of mismatches, or mismatches in the seed in the case of -n mode)
and in terms of the quality values at the mismatched position(s).
bowtie is somewhat slower when --best is specified.
           </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="390"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaBuildSettings</name>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="35"/>
        <source>Index algorithm (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="49"/>
        <source>bwtsw</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="54"/>
        <source>div</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="59"/>
        <source>is</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="67"/>
        <source>Colorspace (-c)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaSettings</name>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="335"/>
        <source>Max gap extensions (-e)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="345"/>
        <source>Indel offset (-i)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="262"/>
        <source>Max long deletion extensions (-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="82"/>
        <source>Seed length (-l)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="481"/>
        <source>Max seed differences (-k)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="282"/>
        <source>Max queue entries (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="355"/>
        <source>Enable long gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="59"/>
        <source>Missing prob (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="42"/>
        <source>Max #diff (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="166"/>
        <source>Best hits (-R)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="202"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="240"/>
        <source>Advanced</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="394"/>
        <source>Gap open penalty (-O)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="302"/>
        <source>Barcode length (-B)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="434"/>
        <source>Quality threshold (-q)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="374"/>
        <source>Mismatch penalty (-M)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="414"/>
        <source>Gap extension penalty (-E)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="461"/>
        <source>Threads (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="195"/>
        <source>Colorspace (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="209"/>
        <source>Non-iterative mode (-N)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="111"/>
        <source>Max gap opens (-o)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="131"/>
        <source>Index algorithm (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="148"/>
        <source>bwtsw</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="153"/>
        <source>div</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="158"/>
        <source>is</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaSwSettings</name>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="42"/>
        <source>Score threshold (divided by match score)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="58"/>
        <source>Number of threads:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="81"/>
        <source>Number of seeds to start rev alginment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="97"/>
        <source>Z-best</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="117"/>
        <source>Prefer hard clipping in SAM output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="131"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Number of seed to trigger reverse alignment&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="144"/>
        <source>Size of chunk of reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="185"/>
        <source>Score for a match</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="208"/>
        <source>Gap extention penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="224"/>
        <source>Mismatch penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="254"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="277"/>
        <source>Band width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="293"/>
        <source>Mask level</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="323"/>
        <source>NOTE: bwa-sw performs alignment of long sequencing reads (Sanger or 454). It accepts reads only in FASTA or FASTQ format. Reads should be compiled into single file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>CAP3SupportDialog</name>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="20"/>
        <source>Contig Assembly With CAP3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="30"/>
        <source>Base</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="36"/>
        <source>Input files (long DNA reads to assembly)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="45"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Sequences in FASTA or FASTQ formats are supported. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;The quality scores for FASTA sequences can be provided in an additional file. The file must be located in the same folder as the original sequences and have the same name as FASTA file, but another extention: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="57"/>
        <source>Add</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="64"/>
        <source>Remove</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="71"/>
        <source>Remove All</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="93"/>
        <source>Result contig</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="102"/>
        <source>Path to result contig alignment in ACE format.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="128"/>
        <source>Clipping for poor regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="134"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="151"/>
        <source>Clipping range (-y)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="174"/>
        <source>Length and percent identity of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="180"/>
        <source>Overlap length cutoff (-o)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="200"/>
        <source>Overlap percent identity cutoff (-p)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="223"/>
        <source>Quality difference score of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="229"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="259"/>
        <source>Max qscore sum at differences (-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="269"/>
        <source>Other parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="275"/>
        <source>Max number of word matches (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="295"/>
        <source>Band expansion size (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="318"/>
        <source>Max gap length in any overlap (-f)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="338"/>
        <source>Assembly reverse reads (-r)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="364"/>
        <source>Similarity score of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="370"/>
        <source>Mismatch score factor (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="390"/>
        <source>Overlap similarity score cutoff (-s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="410"/>
        <source>Gap penalty factor (-g)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="417"/>
        <source>Match score factor (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="461"/>
        <source>Run</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="468"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="112"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="122"/>
        <source>Advanced</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="20"/>
        <source>Align with Clustal Omega</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="75"/>
        <source>Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="85"/>
        <source>Number of iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="131"/>
        <source>Max number guidetree iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="174"/>
        <source>Max number of HMM iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="211"/>
        <source>Miscellaneous</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="221"/>
        <source>Number of CPUs being used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="259"/>
        <source>Set options automatically</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="286"/>
        <source>Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="293"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="26"/>
        <source>Align with ClustalW</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="42"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="52"/>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="76"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="59"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="88"/>
        <source>Advanced options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="98"/>
        <source>Gap opening penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="141"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="184"/>
        <source>Weight matrix</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="214"/>
        <source>IUB</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="219"/>
        <source>CLUSTALW</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="289"/>
        <source>Max iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="261"/>
        <source>NONE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="266"/>
        <source>TREE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="271"/>
        <source>ALIGNMENT</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="231"/>
        <source>Iteration type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="335"/>
        <source>Out sequences order</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="365"/>
        <source>Input</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="370"/>
        <source>Aligned</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="388"/>
        <source>Protein gap parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="398"/>
        <source>Gap separation distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="456"/>
        <source>Residue-specific gaps off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="442"/>
        <source>Hydrophilic gaps off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="449"/>
        <source>No end gap separation penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="483"/>
        <source>Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="490"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>ETSSettingsWidget</name>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="26"/>
        <source>Resource settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="40"/>
        <source>You can select path for</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="47"/>
        <source>Check lastest version of external tools package</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="50"/>
        <source>&lt;a href=&quot;http://ugene.unipro.ru/external.html&quot;&gt;external tools package&lt;/a&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="70"/>
        <source>Browse...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="86"/>
        <source>Supported tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="102"/>
        <source>Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="107"/>
        <source>Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="142"/>
        <source>Additional information</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="177"/>
        <source>Select an external tool to view more information about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="193"/>
        <source>Binary path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="203"/>
        <source>Version info</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="20"/>
        <source>Format database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="29"/>
        <source>Input data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="35"/>
        <source>Select input file(s) for formatting database:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="50"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="78"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="184"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="59"/>
        <source>Or select directory with input files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="90"/>
        <source>File filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="98"/>
        <source>Include files filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="108"/>
        <source>*.fa;*.fasta</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="115"/>
        <source>Exclude files filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="125"/>
        <source>*.pal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="137"/>
        <source>Type of file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="143"/>
        <source>protein</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="153"/>
        <source>nucleotide</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="166"/>
        <source>Output settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="172"/>
        <source>Select the path to save the database into:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="195"/>
        <source>Base name for BLAST files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="205"/>
        <source>Title for database file:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="235"/>
        <source>Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="242"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>IntegralBusPort</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="59"/>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="58"/>
        <source>One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="64"/>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="63"/>
        <source>Only one of these slots must be binded: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="231"/>
        <source>The slot must be not empty: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="753"/>
        <source>Error! One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="759"/>
        <source>Error! You can not bind one of sequences slots and one of url slots simultaneously</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="20"/>
        <source>Align with MAFFT</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="85"/>
        <source>Gap opening penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="131"/>
        <source>Offset (works like gap extension penalty)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="174"/>
        <source>Maximum number of iterative refinement</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="238"/>
        <source>Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="245"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>MrBayesDialog</name>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="59"/>
        <source>MrBayes Settings</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="80"/>
        <source>Model</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="92"/>
        <source>ModelType</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="109"/>
        <source>Rate</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="126"/>
        <source>Gamma</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="164"/>
        <source>MCMC</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="352"/>
        <source>Chain length</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="219"/>
        <source>Subsampling frequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="251"/>
        <source>Burn-in length</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="274"/>
        <source>Heated chains</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="303"/>
        <source>Heated chain temp</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="323"/>
        <source>Random seed</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="120"/>
        <source>Path for temporary files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="121"/>
        <source>Path for temporary files not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="122"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="37"/>
        <source>NOTE: &quot;is&quot; index algorithm is not supposed to work with reference
sequences having size larger than 2 GB. In order to achieve stable
BWA performance it is strongly recommend to set the index algorithm
to &quot;bwtsw&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="41"/>
        <source>NOTE: &quot;bwtsw&quot; index algorithm is not supposed to work with reference
sequences having size smaller than 10 MB. In order to achieve stable
BWA performance it is strongly recommend to set the index algorithm
to &quot;is&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="242"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="264"/>
        <source>Internal error: unexpected format of the %1 output!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="176"/>
        <source>Genome sequence slot is not binded : &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="179"/>
        <source>Transcripts slot is not binded : &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="192"/>
        <source>Input transcripts</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url [FASTA]</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url [GTF]</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="200"/>
        <source>Extracted sequences url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>Output sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>The url to the output file with the extracted sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="218"/>
        <source>Extract Transcript Sequences with gffread</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="219"/>
        <source>Extract transcript sequences from the genomic sequence(s) with gffread.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="69"/>
        <source>Installed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="70"/>
        <source>Not installed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/RnaSeqCommon.cpp" line="62"/>
        <source>Unexpected value &apos;%1&apos; of the library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="785"/>
        <source>%1 tool&apos;s version is undefined, this may cause some compatibility issues</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="794"/>
        <source>Bowtie and TopHat tools have incompatibleversions. Your TopHat&apos;s version is %1, Bowtie&apos;s one is %2. The following are considered to be compatible: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SelectPathDialog</name>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="29"/>
        <source>Path settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="41"/>
        <source>Select executable file of </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="56"/>
        <location filename="../src/ui/SelectPathDialog.ui" line="86"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="71"/>
        <source>Select directory for tempory files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="113"/>
        <source>OK</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="120"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="20"/>
        <source>Align with T-Coffee</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="83"/>
        <source>Gap opening penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="122"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="158"/>
        <source>Number of iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="208"/>
        <source>Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="215"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupport</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="72"/>
        <location filename="../src/blast/BlastAllSupport.cpp" line="87"/>
        <source>The &lt;i&gt;blastall&lt;/i&gt; is the old program developed and distributed by the NCBI for running BLAST searches. The NCBI recommends that people start using the programs of the BLAST+ package instead.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="101"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="102"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportContext</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="146"/>
        <source>Query with BLAST...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="166"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="167"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportMultiTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="654"/>
        <source>Source file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="655"/>
        <source>Used databse</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="657"/>
        <source>No any results found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="83"/>
        <source>Database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="88"/>
        <source>Database name contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="101"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="112"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="112"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="74"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="79"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="92"/>
        <source>Can not create fake NCBI ini file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="215"/>
        <source>Output file not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="217"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="263"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="298"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="147"/>
        <source>Select input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="167"/>
        <source>Database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="172"/>
        <source>Database name contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="229"/>
        <source>Wrong input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="229"/>
        <source>This file not contain any sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="271"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastDBCmdDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="61"/>
        <source>Select a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="75"/>
        <source>Set a result FASTA file name</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastDbCmdSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="66"/>
        <source>The &lt;i&gt;BlastDBCmd&lt;/i&gt; fetches protein or nucleotide sequences from BLAST+ database based on a query.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="79"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="80"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="180"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="181"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportCommonTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="85"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="90"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="103"/>
        <source>Can not create fake NCBI ini file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="148"/>
        <source>Output file not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="150"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="195"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="229"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportContext</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="230"/>
        <source>Fetch sequences by &apos;id&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="240"/>
        <source>Query with BLAST+...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="283"/>
        <source>Fetch sequences from local BLAST database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="287"/>
        <source>Fetch sequences by &apos;id&apos; %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="304"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="305"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportMultiTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="592"/>
        <source>Source file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="593"/>
        <source>Used databse</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="595"/>
        <source>No any results found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="89"/>
        <source>Database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="94"/>
        <source>Database name contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="109"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="144"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="176"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="187"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="187"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="223"/>
        <source>Select input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="250"/>
        <source>Database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="262"/>
        <source>Database name contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="319"/>
        <source>Wrong input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="319"/>
        <source>This file not contain any sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="366"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="401"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="433"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastRunCommonDialog</name>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="239"/>
        <source>Select a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="250"/>
        <source>Direct nucleotide alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="254"/>
        <source>Direct protein alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="258"/>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="262"/>
        <source>Direct protein alignment (on GPU)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="266"/>
        <source>Protein alignment, input nucleotide is translated input protein before the search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="270"/>
        <source>Protein alignment, nucleotide database is translated input protein before the search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="274"/>
        <source>Protein alignment, both input query and database are translated before the search</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2AlignTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="84"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="152"/>
        <source>Unknown short reads format %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2BuildIndexTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="43"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2SettingsWidget</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="103"/>
        <source>DNA Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="104"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="107"/>
        <source>Path for &lt;i&gt;Bowtie2&lt;/i&gt; and &lt;i&gt;Bowtie2-build&lt;/i&gt; tools is not set.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2Support</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="53"/>
        <source>&lt;i&gt;Bowtie 2 aligner&lt;/i&gt; takes a Bowtie 2 index and a set of sequencing read files and outputs a set of alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="69"/>
        <source>&lt;i&gt;Bowtie 2 build indexer&lt;/i&gt;  builds a Bowtie index from a set of DNA sequences. It outputs a set of 6 files with suffixes .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, and .rev.2.bt2. These files together constitute the index: they are all that is needed to align reads to that reference. The original sequence files are no longer used by &lt;i&gt;Bowtie 2&lt;/i&gt; once the index is built.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="90"/>
        <source>&lt;i&gt;Bowtie 2 index inspector&lt;/i&gt; extracts information from a Bowtie index about what kind of index it is and what reference sequence were used to build it.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieAssembleTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="154"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="235"/>
        <source>Unknown short reads format %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="308"/>
        <source>There is not enough memory on the computer!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieBuildIndexTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="44"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieSettingsWidget</name>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="105"/>
        <source>DNA Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="106"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="109"/>
        <source>Path for &lt;i&gt;Bowtie&lt;/i&gt; and &lt;i&gt;Bowtie-build&lt;/i&gt; tools is not set.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieSupport</name>
    <message>
        <location filename="../src/bowtie/BowtieSupport.cpp" line="57"/>
        <source>&lt;i&gt;Bowtie&lt;/i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="89"/>
        <source>Short reads are not provided</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="96"/>
        <source>In paired-end mode it possible to analyze only 2 read sets using BWA</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaSupport</name>
    <message>
        <location filename="../src/bwa/BwaSupport.cpp" line="46"/>
        <source>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; is an efficient program that aligns relatively short nucleotide sequences against a long reference sequence such as the human genome.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaSwAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="241"/>
        <source>Short reads are not provided</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="251"/>
        <source>BWA SW can not align paired reads</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="373"/>
        <source>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3Support</name>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="65"/>
        <source>&lt;i&gt;CAP3&lt;/i&gt; is a contig assembly program.                    &lt;br&gt;It allows to assembly long DNA reads (up to 1000 bp).                    &lt;br&gt;Binaries can be downloaded from http://seq.cs.iastate.edu/cap3.html</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="68"/>
        <source>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="69"/>
        <source>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="70"/>
        <source>&lt;br&gt;Genome Research, 9: 868-877.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="80"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="81"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportDialog</name>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="55"/>
        <source>List of input files is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="67"/>
        <source>Result contig file name is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="73"/>
        <source>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To save under other name press &apos;Cancel&apos; and change name in &apos;Result contig&apos; field.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="76"/>
        <source>Replace</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="77"/>
        <source>Cancel</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="105"/>
        <source>Add sequences to assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="136"/>
        <source>Set result contig file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="136"/>
        <source>ACE format (*.ace)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="109"/>
        <source>Output file not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="111"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="125"/>
        <source>Output file not found: copy from tmp dir failed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="158"/>
        <source>No assembly is found for provided reads</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CEASSupport</name>
    <message>
        <location filename="../src/ceas/CEASSupport.cpp" line="48"/>
        <source>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - helps to characterize genome-wide protein-DNA interaction patterns from ChIP-chip and ChIP-Seq of both sharp and broad binding factors. It provides statistics on ChIP enrichment at important genome features such as specific chromosome, promoters, gene bodies, or exons, and infers genes most likely to be regulated by a binding factor.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CEASSupportTask</name>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="218"/>
        <source>CEAS error: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="229"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupport</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="67"/>
        <source>&lt;i&gt;Clustal Omega&lt;/i&gt; is a free sequence alignment software for proteins.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="77"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportContext</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="140"/>
        <source>Align with ClustalO...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="163"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="164"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="79"/>
        <source>ClustalO alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="90"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="99"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="104"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="154"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="156"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="164"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="237"/>
        <source>Failed to apply the result of ClustalO: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="242"/>
        <source>ClustalO alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="255"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="84"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="94"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="119"/>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="122"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="120"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="123"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="291"/>
        <source>input_format_error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupport</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="66"/>
        <source>&lt;i&gt;ClustalW&lt;/i&gt; is a free sequence alignment software for DNA or proteins.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="76"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="77"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportContext</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="139"/>
        <source>Align with ClustalW...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="163"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="164"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="85"/>
        <source>Unsupported alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="88"/>
        <source>ClustalW alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="99"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="108"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="113"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="176"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="178"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="186"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="255"/>
        <source>Failed to apply the result of ClustalW: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="260"/>
        <source>ClustalW alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="273"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="128"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="138"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="189"/>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="192"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="190"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="193"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="308"/>
        <source>input_format_error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConductGOSupport</name>
    <message>
        <location filename="../src/conduct_go/ConductGOSupport.cpp" line="45"/>
        <source>&lt;i&gt;Conduct GO&lt;/i&gt; - For a list of input genes, this tool uses R/BioC packages (GO, GOstats) to identify over represented GO terms.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConductGOTask</name>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="162"/>
        <source>Conduct GO warning: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="173"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotSupport</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotSupport.cpp" line="47"/>
        <source>&lt;i&gt;ConservationPlot&lt;/i&gt; - Draw conservation plot for many bed files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotTask</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="178"/>
        <source>Conservation Plot error: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="189"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CuffdiffSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="41"/>
        <source>Running Cuffdiff task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="52"/>
        <source>At least 2 assemblies are required for Cuffdiff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="112"/>
        <source>Unable to get an assembly object.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupport</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="53"/>
        <source>&lt;i&gt;Cuffcompare&lt;/i&gt; helps comparing assembled transcripts to a reference annotation, and also tracking transcripts across multiple experiments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="69"/>
        <source>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;tests for differential expression and regulation in RNA-Seq samples.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="84"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts and estimates their abundances.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="100"/>
        <source>&lt;i&gt;Cuffmerge&lt;/i&gt; merges together several assemblies.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="113"/>
        <source>&lt;i&gt;Gffread&lt;/i&gt; is used to verify or perform various operations on GFF files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="52"/>
        <source>Running Cufflinks task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="123"/>
        <source>Unable to get an assembly object.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="247"/>
        <source>%1 output file &apos;%2&apos; is not found!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CuffmergeSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="39"/>
        <source>Running Cuffmerge task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="126"/>
        <source>Can not create a file: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ETSProjectViewItemsContoller</name>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="50"/>
        <source>FormatDB...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="51"/>
        <source>BLAST+ make DB...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="76"/>
        <source>BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="92"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="95"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="97"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchAndValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="322"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportPlugin</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="188"/>
        <source>External tool support</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="188"/>
        <source>Runs other external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="243"/>
        <source>ClustalW...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="250"/>
        <source>ClustalO...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="256"/>
        <source>MAFFT...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="263"/>
        <source>T-Coffee...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="271"/>
        <source>Multiple alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="405"/>
        <source>FormatDB...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="409"/>
        <source>BLAST+ make DB...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="417"/>
        <source>BLAST Search...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="427"/>
        <source>BLAST+ Search...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="431"/>
        <source>BLAST+ query DB</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="439"/>
        <source>BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="38"/>
        <source>The subdirectory for temporary files &apos;%1&apos; already exists. Can not remove this directory!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="43"/>
        <source>Can not create a subdirectory for temporary files &apos;%1&apos;!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportService</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="562"/>
        <source>External tools support</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="562"/>
        <source>Provides support to run external tools from UGENE</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageController</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="48"/>
        <source>External Tools</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageWidget</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="279"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="417"/>
        <source>&lt;br&gt;Version: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="280"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="418"/>
        <source>&lt;br&gt;Binary path: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="358"/>
        <source>Select an external tool to view more information about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="365"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="371"/>
        <source>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches. BLAST can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="377"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="380"/>
        <source>&lt;i&gt;BLAST+&lt;/i&gt; is a new version of the BLAST package from the NCBI.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="383"/>
        <source>&lt;i&gt;Bowtie&lt;i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="391"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. It also estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="399"/>
        <source>&lt;i&gt;Bowtie 2&lt;/i&gt; is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. &lt;br/&gt;&lt;br/&gt;It indexes the genome with an FM index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2Gb. &lt;br/&gt;&lt;br/&gt;&lt;i&gt;Bowtie 2&lt;/i&gt; supports gapped, local, and paired-end alignment modes.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="408"/>
        <source>&lt;i&gt;Cistrome&lt;/i&gt; is a UGENE version of Cistrome pipeline which also includes some tools useful for ChIP-seq analysisThis pipeline is aimed to provide the following analysis steps: peak calling and annotating, motif search and gene ontology.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="413"/>
        <source>No description</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="446"/>
        <source>Choose Directory With Executables</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="506"/>
        <source>Choose Directory With External Tools Pack</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="561"/>
        <source>Not a valid external tools directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="118"/>
        <source>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="141"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsInstallTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="370"/>
        <source>Installing external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="375"/>
        <source> failed: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="349"/>
        <source>Checking external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="361"/>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="363"/>
        <source> failed: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupport</name>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="67"/>
        <location filename="../src/blast/FormatDBSupport.cpp" line="82"/>
        <source>The &lt;i&gt;formatdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by &lt;i&gt;blastall&lt;/i&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="97"/>
        <location filename="../src/blast/FormatDBSupport.cpp" line="112"/>
        <source>The &lt;i&gt;makeblastdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by other BLAST+ tools.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="126"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="129"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="131"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="71"/>
        <source>Select file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="94"/>
        <source>Select a directory with input files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="105"/>
        <source>Select a directory to save database files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="117"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="123"/>
        <source>Input files paths contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="131"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="136"/>
        <source>Output database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportTask</name>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="59"/>
        <source>Input files paths contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="74"/>
        <source>Output database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GffreadSupportTask</name>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="34"/>
        <source>Running Gffread task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="64"/>
        <source>Unknown file format: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="73"/>
        <source>The file format is not [%1]: %2</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllPrompter</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="326"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="327"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="328"/>
        <source>For sequence %1 find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllWorker</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="78"/>
        <source>Input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="79"/>
        <source>Sequence for which annotations is searched.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="80"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="80"/>
        <source>Found annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="89"/>
        <source>Search type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="90"/>
        <source>Select type of BLAST searches.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="91"/>
        <source>Database Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="92"/>
        <source>Path with database files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="93"/>
        <source>Database Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="94"/>
        <source>Base name for BLAST DB files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="95"/>
        <source>Expected value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="96"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="97"/>
        <source>Annotate as</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="98"/>
        <source>Name for annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="99"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="100"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="101"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="102"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="104"/>
        <source>BLAST output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="105"/>
        <source>Location of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="106"/>
        <source>BLAST output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="107"/>
        <source>Type of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="108"/>
        <source>Gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="109"/>
        <source>Perform gapped alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="111"/>
        <source>Gap costs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="112"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="113"/>
        <source>Match scores</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="114"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="170"/>
        <source>Local BLAST Search</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="171"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="383"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="392"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="397"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="404"/>
        <source>Not selected BLAST output file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusPrompter</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="219"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="220"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="221"/>
        <source>For sequence &lt;u&gt;%1&lt;/u&gt; find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusWorker</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="74"/>
        <source>Input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="75"/>
        <source>Sequence for which annotations is searched.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="76"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="76"/>
        <source>Found annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="85"/>
        <source>Search type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="86"/>
        <source>Select type of BLAST+ searches.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="87"/>
        <source>Database Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="88"/>
        <source>Path with database files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="89"/>
        <source>Database Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="90"/>
        <source>Base name for BLAST+ DB files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="91"/>
        <source>Expected value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="92"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="93"/>
        <source>Annotate as</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="94"/>
        <source>Name for annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="95"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="96"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="97"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="98"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="99"/>
        <source>BLAST output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="100"/>
        <source>Location of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="101"/>
        <source>BLAST output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="102"/>
        <source>Type of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="103"/>
        <source>Gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="104"/>
        <source>Perform gapped alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="106"/>
        <source>Gap costs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="107"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="108"/>
        <source>Match scores</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="109"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="136"/>
        <source>Local BLAST+ Search</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="137"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="286"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="295"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="300"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="307"/>
        <source>Not selected BLAST output file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Prompter</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="95"/>
        <source>Aligns long DNA reads from &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;CAP3&lt;/u&gt;&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Worker</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="56"/>
        <source>CAP3 result multiple sequence alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="57"/>
        <source>The result of the CAP3 contig assembly.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="63"/>
        <source>Input</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="64"/>
        <source>Path to input long DNA reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="69"/>
        <source>Align with CAP3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="144"/>
        <source>Aligned %1 with CAP3</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportPrompter</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="371"/>
        <source>Creates summary statistics on ChIP enrichment and saves it to %1.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportWorker</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="214"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="215"/>
        <source>Peak regions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="217"/>
        <source>Enrichment signal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="218"/>
        <source>Enrichment signal.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="223"/>
        <source>CEAS data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="224"/>
        <source>Peak regions and enrichment signal profile for generating CEAS report.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="233"/>
        <source>Output report file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="234"/>
        <source>Path to the report output file. Result for CEAS analysis.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="239"/>
        <source>Output annotations file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="240"/>
        <source>Name of tab-delimited output text file, containing a row of annotations for every RefSeq gene. (file is not generated if no peak location data is supplied).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="243"/>
        <source>Gene annotations table</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="244"/>
        <source>Path to gene annotation table (e.g. a refGene table in sqlite3 db format (--gt).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="247"/>
        <source>Span size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="248"/>
        <source>Span from TSS and TTS in the gene-centered annotation (base pairs). ChIP regions within this range from TSS and TTS are considered when calculating the coverage rates in promoter and downstream (--span).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="253"/>
        <source>Wiggle profiling resolution</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="254"/>
        <source>Wiggle profiling resolution. WARNING: Value smaller than the wig interval (resolution) may cause aliasing error (--pf-res).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="258"/>
        <source>Promoter/downstream interval</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="259"/>
        <source>Promoter/downstream intervals for ChIP region annotation are three values or a single value can be given. If a single value is given, it will be segmented into three equal fractions (e.g. 3000 is equivalent to 1000,2000,3000) (--rel-dist).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="265"/>
        <source>BiPromoter ranges</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="266"/>
        <source>Bidirectional-promoter sizes for ChIP region annotation. It&apos;s two values or a single value can be given. If a single value is given, it will be segmented into two equal fractions (e.g. 5000 is equivalent to 2500,5000) (--bisizes).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="271"/>
        <source>Relative distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="272"/>
        <source>Relative distance to TSS/TTS in WIGGLE file. profiling. (--rel-dist)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="275"/>
        <source>Gene group files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="276"/>
        <source>Gene groups of particular interest in wig profiling. Each gene group file must have gene names in the 1st column. The file names are separated by commas (--gn-groups).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="280"/>
        <source>Gene group names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="281"/>
        <source>&lt;i&gt;Set this parameter empty for using default values.&lt;/i&gt;&lt;br&gt;The names of the gene groups from &quot;Gene group files&quot; parameter. These names appear in the legends of the wig profiling plots.&lt;br&gt; Values range: comma-separated list of strings. Default value: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="321"/>
        <source>Report file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="328"/>
        <source>Annotations file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="351"/>
        <source>Create CEAS Report</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="352"/>
        <source>Provides summary statistics on ChIP enrichment in important genomic regions such as individual chromosomes, promoters, gene bodies or exons, and infers the genes most likely to be regulated by the binding factor under study.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOPrompter</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="139"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="140"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOWorker</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA to process.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="62"/>
        <source>ClustalO result MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="63"/>
        <source>The result of the ClustalO alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="72"/>
        <source>Number of iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="73"/>
        <source>Number of (combined guide-tree/HMM) iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="74"/>
        <source>Number of guidetree iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="75"/>
        <source>Maximum number guidetree iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="76"/>
        <source>Number of HMM iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="77"/>
        <source>Maximum number of HMM iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="78"/>
        <source>Set auto options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="79"/>
        <source>Set options automatically (might overwrite some of your options).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="81"/>
        <source>Tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="82"/>
        <source>Path to the ClustalO tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="85"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="86"/>
        <source>Directory to store temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="96"/>
        <source>Align with ClustalO</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="97"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalO.&lt;p&gt;ClustalO is a general purpose multiple sequence alignment program for proteins.Visit &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; to learn more about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="184"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="216"/>
        <source>Aligned %1 with ClustalO</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWPrompter</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="188"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="189"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWWorker</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA to process.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="65"/>
        <source>ClustalW result MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="66"/>
        <source>The result of the ClustalW alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="75"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="76"/>
        <source>The penalty for opening a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="77"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="78"/>
        <source>The penalty for extending a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="79"/>
        <source>Gap distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="80"/>
        <source>The gap separation penalty. Tries to decrease the chances of gaps being too close to each other.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="81"/>
        <source>End gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="82"/>
        <source>The penalty for closing a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="83"/>
        <source>Residue-specific gaps off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="84"/>
        <source>Residue-specific penalties are amino specific gap penalties that reduce or increase the gap opening penalties at each position in the alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="85"/>
        <source>Hydrophilic gaps off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="86"/>
        <source>Hydrophilic gap penalties are used to increase the chances of a gap within a run (5 or more residues) of hydrophilic amino acids.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="87"/>
        <source>Iteration type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="88"/>
        <source>Alignment improvement iteration type. Can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;None - No iteration;&lt;/li&gt;                            &lt;li&gt;Tree - Iteration at each step of alignment process;&lt;/li&gt;                            &lt;li&gt;Alignment - Iteration only on final alignment.&lt;/li&gt;                            &lt;/ul&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="94"/>
        <source>Number of iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="95"/>
        <source>The maximum number of iterations to perform.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="96"/>
        <source>Weight matrix</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="110"/>
        <source>Tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="111"/>
        <source>Path to the ClustalW tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="115"/>
        <source>Directory to store temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="130"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalW.&lt;p&gt;ClustalW is a general purpose multiple sequence alignment program for DNA or proteins.Visit &lt;a href=&quot;http://www.clustal.org/&quot;&gt;http://www.clustal.org/&lt;/a&gt; to learn more about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="260"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="114"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="97"/>
        <source>For proteins it is a scoring table which describes the similarity of each amino acid to each other and can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - Appear to be the best available for carrying out database similarity (homology searches).                            &lt;li&gt;PAM - Have been extremely widely used since the late &apos;70s.                            &lt;li&gt;Gonnet - Derived using almost the same procedure as the PAM bit are much more up to date.                            &lt;li&gt;ID - Identitical matrix gives a score of 1.0 to two identical amino acids and a score of zero otherwise.                            &lt;/ul&gt;                            &lt;p&gt;                            For DNA it is the scores assigned to matches and mismatches. The values available are:                            &lt;ul&gt;                            &lt;li&gt;IUB - The default scoring matrix used by BESTFIT for the comparison of nucleic acid sequences. All matches score 1.9; all mismatches for IUB symbols score 0.                           &lt;li&gt;ClustalW - The previous system used by Clustal W, in which matches score 1.0 and mismatches score 0. All matches for IUB symbols also score 0.                            &lt;/ul&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="129"/>
        <source>Align with ClustalW</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="289"/>
        <source>Aligned %1 with ClustalW</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOPrompter</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="221"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="226"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as gene list to conduct GO.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="228"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOWorker</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="145"/>
        <source>Target genes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="146"/>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="151"/>
        <source>Gene list to identify over represented GO terms.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="150"/>
        <source>Conduct GO data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="159"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="160"/>
        <source>The directory to store Conduct GO results.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="162"/>
        <source>Title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="163"/>
        <source>Title is used to name the output files - so make it meaningful.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="165"/>
        <source>Gene Universe</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="166"/>
        <source>Select a gene universe.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="201"/>
        <source>Conduct GO</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="202"/>
        <source>Given a list of genes, using Bioconductor (GO, GOstats) and  DAVID at NIH.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotPrompter</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="290"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="296"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions for conservation plot.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="297"/>
        <source> Conservations scores from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="298"/>
        <source> Outputs the result to &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotWorker</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="190"/>
        <source>Input regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="191"/>
        <source>Regions (centered at peak summits for better performance).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="195"/>
        <source>conservation_plot data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="196"/>
        <source>Regions to plot the PhastCons scores profiles.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="204"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="205"/>
        <source>File to store phastcons results (BMP).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="207"/>
        <source>Title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="208"/>
        <source>Title of the figure (--title).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="210"/>
        <source>Label</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="211"/>
        <source>Label of data in the figure (--bed-label).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="213"/>
        <source>Assembly version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="214"/>
        <source>The directory to store phastcons scores (--phasdb).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="216"/>
        <source>Window width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="217"/>
        <source>Window width centered at middle of regions (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="219"/>
        <source>Height</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="220"/>
        <source>Height of plot (--height).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="222"/>
        <source>Width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="223"/>
        <source>Width of plot (--width).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="248"/>
        <source>Conservation plot file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="270"/>
        <source>Build Conservation Plot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="271"/>
        <source>Plots the PhastCons scores profiles.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffdiffWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="98"/>
        <source>Test for Diff. Expression with Cuffdiff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="112"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="113"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="116"/>
        <source>Time series analysis</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="117"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, instructs Cuffdiff to analyze the provided samples as a time series, rather than testing for differences between all pairs of samples. Samples should be provided in increasing time order.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="123"/>
        <source>Upper quartile norm</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="124"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, normalizes by the upper quartile of the number of fragments mapping to individual loci instead of the total number of sequenced fragments. This can improve robustness of differential expression calls for less abundant genes and transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="130"/>
        <source>Hits norm</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="131"/>
        <source>Instructs how to count all fragments. &lt;i&gt;Total&lt;/i&gt; specifies to count all fragments, including those not compatible with any reference transcript, towards the number of mapped fragments used in the FPKM denominator. &lt;i&gt;Compatible&lt;/i&gt; specifies to use only compatible fragments. Selecting &lt;i&gt;Compatible&lt;/i&gt; is generally recommended in Cuffdiff to reduce certain types of bias caused by differential amounts of ribosomal reads which can create the impression of falsely differentially expressed genes.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="140"/>
        <source>Frag bias correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="141"/>
        <source>Providing the sequences your reads were mapped to instructs Cuffdiff to run bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="146"/>
        <source>Multi read correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="147"/>
        <source>Do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="151"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="152"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="155"/>
        <source>Mask file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="156"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="164"/>
        <source>Min alignment count</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="165"/>
        <source>The minimum number of alignments in a locus for needed to conduct significance testing on changes in that locus observed between samples. If no testing is performed, changes in the locus are deemed not significant, and the locus&apos; observed changes don&apos;t contribute to correction for multiple testing.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="172"/>
        <source>FDR</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="173"/>
        <source>The allowed false discovery rate used in testing.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="176"/>
        <source>Max MLE iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="177"/>
        <source>Sets the number of iterations allowed during maximum likelihood estimation of abundances.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="181"/>
        <source>Emit count tables</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="182"/>
        <source>Include information about the fragment counts, fragment count variances, and fitted variance model into the report.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="186"/>
        <source>Cuffdiff tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="187"/>
        <source>The path to the Cuffdiff external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="190"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="191"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="211"/>
        <source>Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="212"/>
        <source>RNA-Seq reads assemblies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="214"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="215"/>
        <source>Transcript annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="301"/>
        <source>Searches for significant changes in transcript expression, splicing and promoter use.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CufflinksWorker</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="97"/>
        <source>Input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="98"/>
        <source>Input RNA-Seq read alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="101"/>
        <source>Output annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="102"/>
        <source>Assembled isoforms, estimated isoform-level expression values, and estimated gene-level expression values, produced by Cufflinks.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="115"/>
        <source>Isoform-level expression values</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="116"/>
        <source>A set of annotated regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="133"/>
        <source>Assembly Transcripts with Cufflinks</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="134"/>
        <source>Cufflinks accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="142"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="143"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="146"/>
        <source>Reference annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="147"/>
        <source>Tells Cufflinks to use the supplied reference annotation to estimate isoform expression. Cufflinks will not assemble novel transcripts and the program will ignore alignments not structurally compatible with any reference transcript.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="153"/>
        <source>RABT annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="154"/>
        <source>Tells Cufflinks to use the supplied reference annotation to guide Reference Annotation Based Transcript (RABT) assembly. Reference transcripts will be tiled with faux-reads to provide additional information in assembly. Output will include all reference transcripts as well as any novel genes and isoforms that are assembled.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="161"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="162"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="165"/>
        <source>Mask file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="166"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="174"/>
        <source>Multi-read correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="175"/>
        <source>Tells Cufflinks to do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="179"/>
        <source>Min isoform fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="180"/>
        <source>After calculating isoform abundance for a gene, Cufflinks filters out transcripts that it believes are very low abundance, because isoforms expressed at extremely low levels often cannot reliably be assembled, and may even be artifacts of incompletely spliced precursors of processed transcripts. This parameter is also used to filter out introns that have far fewer spliced alignments supporting them.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="188"/>
        <source>Frag bias correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="189"/>
        <source>Providing Cufflinks with a multifasta file via this option instructs it to run the bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="194"/>
        <source>Pre-mRNA fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="195"/>
        <source>Some RNA-Seq protocols produce a significant amount of reads that originate from incompletely spliced transcripts, and these reads can confound the assembly of fully spliced mRNAs. Cufflinks uses this parameter to filter out alignments that lie within the intronic intervals implied by the spliced alignments. The minimum depth of coverage in the intronic region covered by the alignment is divided by the number of spliced reads, and if the result is lower than this parameter value, the intronic alignments are ignored.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="205"/>
        <source>Cufflinks tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="206"/>
        <source>The path to the Cufflinks external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="209"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="210"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="337"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffmergeWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="62"/>
        <source>Merge Assemblies with Cuffmerge</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="63"/>
        <source>Cuffmerge merges together several assemblies. It also handles running Cuffcompare for you, and automatically filters a number of transfrags that are probably artifacts. If you have a reference file available, you can provide it to Cuffmerge in order to gracefully merge input (e.g. novel) isoforms and known isoforms and maximize overall assembly quality.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="73"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="74"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="77"/>
        <source>Reference annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="78"/>
        <source>Merge the input assemblies together with this reference annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="82"/>
        <source>Reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="83"/>
        <source>The genomic DNA sequences for the reference. It is used to assist in classifying transfrags and excluding artifacts (e.g. repeats). For example, transcripts consisting mostly of lower-case bases are classified as repeats.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="89"/>
        <source>Minimum isoform fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="90"/>
        <source>Discard isoforms with abundance below this.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="93"/>
        <source>Cuffcompare tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="94"/>
        <source>The path to the Cuffcompare external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="97"/>
        <source>Cuffmerge tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="98"/>
        <source>The path to the Cuffmerge external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="101"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="102"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="116"/>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="119"/>
        <source>Set of annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="117"/>
        <source>Annotations for merging</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="120"/>
        <source>Merged annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="183"/>
        <source>Merges together supplied assemblies.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GffreadPrompter</name>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="245"/>
        <source>Extract transcript sequences from the genomic sequence from &lt;u&gt;%1&lt;/u&gt; using transcripts from &lt;u&gt;%2&lt;/u&gt; and save them to the file %3.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSPrompter</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="563"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="570"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as treatment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="572"/>
        <source> and annotations from &lt;u&gt;%1&lt;/u&gt; as control</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="575"/>
        <source> to call peaks.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="577"/>
        <source> Outputs all files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="579"/>
        <source> and pileup with &lt;u&gt;%1&lt;/u&gt; span</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSWorker</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="190"/>
        <source>Bad model fold region: %1. Default region is used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="257"/>
        <source>Input control and treatment annotations are the same</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="271"/>
        <source>Treatment features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="272"/>
        <source>ChIP-seq treatment features.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="274"/>
        <source>Control features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="275"/>
        <source>Control features (Optional).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="280"/>
        <source>MACS data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="281"/>
        <source>ChIP-seq treatment features and control features (optional) to call peaks with MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="289"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="290"/>
        <source>Peak locations. Typically used in gene association study like CEAS, or correlation calculation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="292"/>
        <source>Peak summits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="293"/>
        <source>Peak summits locations for every peaks. Typically used in DNA motif analysis or conservation check.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="295"/>
        <source>Treatment fragments pileup (wiggle)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="296"/>
        <source>Wiggle format files which can be imported to UCSC genome browser/GMOD/Affy IGB.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="299"/>
        <source>MACS output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="300"/>
        <source>ChIP-seq peaks and summits. Pileup data (optional)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="313"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="314"/>
        <source>Directory to save MACS output files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="316"/>
        <source>Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="317"/>
        <source>The name string of the experiment. MACS will use this string NAME to create output files like &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; and so on. So please avoid any confliction between these filenames and your existing files (--name).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="321"/>
        <source>Wiggle output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="322"/>
        <source>If this flag is on, MACS will store the fragment pileup in wiggle format for the whole genome data instead of for every chromosomes (--wig) (--single-profile).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="324"/>
        <source>Wiggle space</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="325"/>
        <source>By default, the resolution for saving wiggle files is 10 bps,i.e., MACS will save the raw tag count every 10 bps. You can change it along with &apos;--wig&apos; option (--space).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="328"/>
        <source>Genome size (Mbp)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="329"/>
        <source>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt; It&apos;s the mappable genome size or effective genome size which is defined as the genome size which can be sequenced. Because of the repetitive features on the chromosomes, the actual mappable genome size will be smaller than the original size, about 90% or 70% of the genome size (--gsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="338"/>
        <source>P-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="339"/>
        <source>P-value cutoff. Default is 0.00001, for looser results, try 0.001 instead (--pvalue).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="341"/>
        <source>Q-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="342"/>
        <source>Minimum FDR (q-value) cutoff for peak detection.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="344"/>
        <source>Use model</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="345"/>
        <source>Whether or not to use MACS paired peaks model (--nomodel).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="347"/>
        <source>Model fold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="348"/>
        <source>Select the regions within MFOLD range of high-confidence enrichment ratio against. &lt;b&gt;Model fold&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is true, which is the foldchange to chose paired peaks to build paired peaks model. Users need to set a lower(smaller) and upper(larger) number for fold change so that MACS will only use the peaks within these foldchange range to build model (--mfold).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="353"/>
        <source>Shift size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="354"/>
        <source>An arbitrary shift value used as a half of the fragment size when model is not built. &lt;b&gt;Shift size&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is false, which will represent the HALF of the fragment size of your sample. If your sonication and size selection size is 300 bps, after you trim out nearly 100 bps adapters, the fragment size is about 200 bps, so you can specify 100 here (--shiftsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="359"/>
        <source>Keep duplicates</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="360"/>
        <source>It controls the MACS behavior towards duplicate tags at the exact same location -- the same coordination and the same strand. The default &lt;b&gt;auto&lt;/b&gt; option makes MACS calculate the maximum tags at the exact same location based on binomal distribution using 1e-5 as pvalue cutoff; and the &lt;b&gt;all&lt;/b&gt; option keeps every tags. If an &lt;b&gt;integer&lt;/b&gt; is given, at most this number of tags will be kept at the same location (--keep-dup).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="364"/>
        <source>Band width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="365"/>
        <source>The band width which is used to scan the genome for model building. You can set this parameter as the sonication fragment size expected from wet experiment. Used only while building the shifting model (--bw).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="369"/>
        <source>Extended fragment pileup</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="370"/>
        <source>Whether or not to generate extended fragment pileup, local lambda and score tracks at every bp.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="374"/>
        <source>Tag size (optional)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="375"/>
        <source>Length of reads. Determined from first 10 reads if not specified (input &lt;b&gt;0&lt;/b&gt;) (--tsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="379"/>
        <source>Use lambda</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="380"/>
        <source>Whether to use local lambda model which can use the local bias at peak regions to throw out false positives (--nolambda).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="382"/>
        <source>Small nearby region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="383"/>
        <source>The small nearby region in basepairs to calculate dynamic lambda. This is used to capture the bias near the peak summit region. Invalid if there is no control data (--slocal).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="386"/>
        <source>Large nearby region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="387"/>
        <source>The large nearby region in basepairs to calculate dynamic lambda.  This is used to capture the surround bias (--llocal).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="390"/>
        <source>Auto bimodal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="391"/>
        <source>Whether turn on the auto pair model process.If set, when MACS failed to build paired model, it will use the nomodelsettings, the Shift size parameter to shift and extend each tags (--on-auto).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="395"/>
        <source>Scale to large</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="396"/>
        <source> When set, scale the small sample up to the bigger sample.By default, the bigger dataset will be scaled down towards the smaller dataset,which will lead to smaller p/qvalues and more specific results.Keep in mind that scaling down will bring down background noise more (--to-large).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="401"/>
        <source>Shift control</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="402"/>
        <source>When set, control tags will be shifted just as ChIP tags according to their strand before the extension of d, slocal and llocal. By default, control tags are extended centeredat their current positions regardless of strand. You may consider to turn this option on whilecomparing two ChIP datasets of different condition but the same factor.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="407"/>
        <source>Half-extend</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="408"/>
        <source>When set, MACS extends 1/2 d size for each fragment centered at its middle point.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="410"/>
        <source>Broad</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="411"/>
        <source>If set, MACS will try to call broad peaks by linking nearby highly enriched regions.The linking region is controlled by another cutoff through Broad cutoff.The maximum linking region length is 4 times of d from MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="415"/>
        <source>Broad cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="416"/>
        <source>Cutoff for broad region. This option is not available unless Broad is set.If P-value is set, this is a pvalue cutoff, otherwise, it&apos;s a qvalue cutoff.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="541"/>
        <source>Find Peaks with MACS</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="542"/>
        <source>Performs peak calling for ChIP-Seq data.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTPrompter</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="129"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="131"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &lt;u&gt;&quot;MAFFT&quot;&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTWorker</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="60"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="61"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Multiple sequence alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Result of alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="71"/>
        <source>Gap Open Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="72"/>
        <source>Gap Open Penalty.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="73"/>
        <source>Offset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="74"/>
        <source>Works like gap extension penalty.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="76"/>
        <source>Maximum number of iterative refinement.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="78"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="80"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="173"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="75"/>
        <source>Max Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="77"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="79"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="88"/>
        <source>Align with MAFFT</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="202"/>
        <source>Aligned %1 with MAFFT</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="89"/>
        <source>MAFFT is a multiple sequence alignment program for unix-like operating systems. &lt;p&gt;&lt;dfn&gt;It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of &amp;lt;&amp;#126;200 sequences), FFT-NS-2 (fast; for alignment of &amp;lt;&amp;#126;10,000 sequences), etc. &lt;/dfn&gt;&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GenePrompter</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="311"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="316"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="317"/>
        <source> to annotate with genes nearby from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GeneWorker</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="189"/>
        <source>Treatment features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="190"/>
        <source>Result peaks of MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="194"/>
        <source>Peak2gene data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="195"/>
        <source>MACS peaks to get the refgenes near the peak summit/center.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="203"/>
        <source>Gene regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="204"/>
        <source>Annotation for each gene, containing all the peaks nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="206"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="207"/>
        <source>Annotation for each peak, containing all the genes nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="210"/>
        <source>Peak2gene output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="211"/>
        <source>Genes containing all the peaks nearby and peaks containing all the genes nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="223"/>
        <source>Output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="224"/>
        <source>Select which type of genes need to output. &lt;b&gt;up&lt;/b&gt; for genes upstream to peak summit, &lt;b&gt;down&lt;/b&gt; for genes downstream to peak summit, &lt;b&gt;all&lt;/b&gt; for both &lt;b&gt;up&lt;/b&gt; and &lt;b&gt;down&lt;/b&gt; (--op).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="229"/>
        <source>Official gene symbols</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="230"/>
        <source>Output &lt;b&gt;official gene symbol&lt;/b&gt; instead of &lt;b&gt;refseq name&lt;/b&gt; (--symbol).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="232"/>
        <source>Distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="233"/>
        <source>Set a number which unit is base. It will get the refGenes in n bases from peak center (--distance).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="235"/>
        <source>Genome file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="236"/>
        <source>Select a genome file (sqlite3 file) to search refGenes (--genome).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="291"/>
        <source>Annotate Peaks with peak2gene</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="292"/>
        <source>Gets refGenes near the ChIP regions identified by a peak-caller.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosComboBoxWithChecksDelegate</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="149"/>
        <source>Hint:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="149"/>
        <source>Use &apos;cistrome.xml&apos; to descrease the computation time. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosPrompter</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="352"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="360"/>
        <source>Uses regions from &lt;u&gt;%1&lt;/u&gt; to find motifs enriched in them.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="361"/>
        <source> Genome assembly: &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="363"/>
        <source> Finds de novo motifs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="366"/>
        <source> Uses &lt;u&gt;%1&lt;/u&gt; known motifs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="369"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosWorker</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="235"/>
        <source>Input regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="236"/>
        <source>Peak summits.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="240"/>
        <source>SeqPos data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="241"/>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="333"/>
        <source>Finds motifs enriched in a set of regions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="249"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="250"/>
        <source>The directory to store seqpos results.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="252"/>
        <source>Genome assembly version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="253"/>
        <source>UCSC database version (GENOME).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="255"/>
        <source>De novo motifs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="256"/>
        <source>Run de novo motif search (-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="258"/>
        <source>Motif database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="259"/>
        <source>Known motif collections (-m). Warning: computation time increases with selecting additional databases. It is recommended to use cistrome.xml. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="262"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="263"/>
        <source>Name of the output file which stores new motifs found during a de novo search (-n).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="265"/>
        <source>Region width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="266"/>
        <source>Width of the region to be scanned for motifs; depends on a resolution of assay (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="268"/>
        <source>Pvalue cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="269"/>
        <source>Pvalue cutoff for the motif significance (-p).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="332"/>
        <source>Collect Motifs with SeqPos</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeePrompter</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="125"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="127"/>
        <source>For each MSA&lt;u&gt;%1&lt;/u&gt;, build the alignment using &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt; and send it to output.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeeWorker</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Multiple sequence alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Result of alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="69"/>
        <source>Gap Open Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="70"/>
        <source>Gap Open Penalty. Must be negative, best matches get a score of 1000.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="71"/>
        <source>Gap Extension Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="72"/>
        <source>Gap Extension Penalty. Positive values give rewards to gaps and prevent the alignment of unrelated segments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="73"/>
        <source>Max Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="74"/>
        <source>Number of iteration on the progressive alignment.&lt;br&gt;0 - no iteration, -1 - Nseq iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="76"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="77"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="79"/>
        <source>Directory for temporary file.s</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="170"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="78"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="87"/>
        <source>Align with T-Coffee</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="88"/>
        <source>T-Coffee is a multiple sequence alignment package. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="202"/>
        <source>Aligned %1 with T-Coffee</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TopHatWorker</name>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="107"/>
        <location filename="../src/tophat/TopHatWorker.cpp" line="118"/>
        <source>Input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="108"/>
        <source>Input RNA-Seq reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="112"/>
        <source>TopHat output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="113"/>
        <source>Accepted hits, junctions, insertions and deletions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="119"/>
        <source>TopHat input reads. Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="124"/>
        <source>Input reads url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="125"/>
        <source>TopHat input reads url. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="129"/>
        <source>Input paired reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="130"/>
        <source> Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads. Reads MUST appear in the same order as the *_1 reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="137"/>
        <source>Input paired reads url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="138"/>
        <source>Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="143"/>
        <source>Dataset name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="144"/>
        <source>Use it only when sequences slot(or slots) is specified. Group input reads into chunks for several Tophat runs.
Set it empty if you want to run Tophat once for all input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="161"/>
        <source>Accepted hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="162"/>
        <source>Accepted hits found by TopHat</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="164"/>
        <source>Accepted hits url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="165"/>
        <source>The url to the assembly file with the accepted hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="175"/>
        <source>Find Splice Junctions with TopHat</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="176"/>
        <source>TopHat is a fast splice junction mapper for RNA-Seq reads. It aligns RNA-Seq reads to mammalian-sized genomes using the ultra high-throughput short read aligner Bowtie, and then analyzes the mapping results to identify splice junctions between exons.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="184"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="185"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="188"/>
        <source>Bowtie index directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="189"/>
        <source>The directory with the Bowtie index for the reference sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="197"/>
        <source>Bowtie index basename</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="198"/>
        <source>The basename of the Bowtie index for the reference sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="211"/>
        <source>Mate inner distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="212"/>
        <source>The expected (mean) inner distance between mate pairs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="215"/>
        <source>Mate standard deviation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="216"/>
        <source>The standard deviation for the distribution on inner distances between mate pairs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="219"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="220"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="223"/>
        <source>No novel junctions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="224"/>
        <source>Only look for reads across junctions indicated in the supplied GFF or junctions file. This parameter is ignored if &lt;i&gt;Raw junctions&lt;/i&gt; or &lt;i&gt;Known transcript file&lt;/i&gt; is not set.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="229"/>
        <source>Raw junctions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="230"/>
        <source>The list of raw junctions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="233"/>
        <source>Known transcript file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="234"/>
        <source>A set of gene model annotations and/or known transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="237"/>
        <source>Max multihits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="238"/>
        <source>Instructs TopHat to allow up to this many alignments to the reference for a given read, and suppresses all alignments for reads with more than this many alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="243"/>
        <source>Segment length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="244"/>
        <source>Each read is cut up into segments, each at least this long. These segments are mapped independently.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="256"/>
        <source>Fusion search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="257"/>
        <source>Turn on fusion mapping.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="260"/>
        <source>Transcriptome only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="261"/>
        <source>Only align the reads to the transcriptome and report only those mappings as genomic mappings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="265"/>
        <source>Transcriptome max hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="266"/>
        <source>Maximum number of mappings allowed for a read, when aligned to the transcriptome (any reads found with more than this number of mappings will be discarded).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="271"/>
        <source>Prefilter multihits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="272"/>
        <source>When mapping reads on the transcriptome, some repetitive or low complexity reads that would be discarded in the context of the genome may appear to align to the transcript sequences and thus may end up reported as mapped to those genes only. This option directs TopHat to first align the reads to the whole genome in order to determine and exclude such multi-mapped reads (according to the value of the &lt;i&gt;Max multihits&lt;/i&gt; option).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="281"/>
        <source>Min anchor length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="282"/>
        <source>The &lt;i&gt;anchor length&lt;/i&gt;. TopHat will report junctions spanned by reads with at least this many bases on each side of the junction. Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved in spliced alignments is supported by at least one read with this many bases on each side.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="290"/>
        <source>Splice mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="291"/>
        <source>The maximum number of mismatches that may appear in the &lt;i&gt;anchor&lt;/i&gt; region of a spliced alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="295"/>
        <source>Read mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="296"/>
        <source>Final read alignments having more than these many mismatches are discarded.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="300"/>
        <source>Segment mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="301"/>
        <source>Read segments are mapped independently, allowing up to this many mismatches in each segment alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="306"/>
        <source>Solexa 1.3 quals</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="307"/>
        <source>As of the Illumina GA pipeline version 1.3, quality scores are encoded in Phred-scaled base-64. Use this option for FASTQ files from pipeline 1.3 or later.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="312"/>
        <source>Bowtie version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="313"/>
        <source>Specifies which Bowtie version should be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="316"/>
        <source>Bowtie -n mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="317"/>
        <source>TopHat uses &lt;i&gt;-v&lt;/i&gt; in Bowtie for initial read mapping (the default), but with this option, &lt;i&gt;-n&lt;/i&gt; is used instead. Read segments are always mapped using &lt;i&gt;-v&lt;/i&gt; option.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="323"/>
        <source>Bowtie tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="324"/>
        <source>The path to the Bowtie external tool.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="327"/>
        <source>SAMtools tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="328"/>
        <source>The path to the SAMtools tool. Note that the tool is available in the UGENE External Tool Package.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="332"/>
        <source>TopHat tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="333"/>
        <source>The path to the TopHat external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="336"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="337"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="499"/>
        <source>Aligns RNA-seq reads to a reference and finds splice junctions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="563"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="592"/>
        <source>Unrecognized value of the Bowtie mode option!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MACSSupport</name>
    <message>
        <location filename="../src/macs/MACSSupport.cpp" line="44"/>
        <source>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, which analyzes datagenerated by short read sequencers such as Solexa&apos;s Genome Analyzer. MACS empirically models the shiftsize of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.MACS also uses a dynamic Poisson distribution to effectively capture local biases in the genome,allowing for more robust predictions. MACS compares favorably to existing ChIP-Seq peak-finding algorithms, and is freely available.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupport</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="65"/>
        <source>&lt;i&gt;MAFFT&lt;/i&gt; is a multiple sequence alignment program for unix-like operating systems. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="75"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="76"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportContext</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="138"/>
        <source>Align with MAFFT...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="160"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="161"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="78"/>
        <source>MAFFT alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="89"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="98"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="103"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="116"/>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="118"/>
        <source>Can not open output file: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="119"/>
        <source> May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="155"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="157"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="165"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="228"/>
        <source>Failed to apply the result of aligning with MAFFT: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="233"/>
        <source>MAFFT alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="246"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="74"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="84"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="108"/>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="111"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="109"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="112"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="279"/>
        <source>input_format_error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="202"/>
        <source>Generating output trees from MrBayes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="209"/>
        <source>Output file is not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="236"/>
        <source>No trees are found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesPrepareDataForCalculation</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="38"/>
        <source>Generating input file for MrBayes</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupport</name>
    <message>
        <location filename="../src/mrbayes/MrBayesSupport.cpp" line="56"/>
        <source>&lt;i&gt;MrBayes&lt;/i&gt; MrBayes is a program for the Bayesian estimation of phylogeny.Bayesian inference of phylogeny is based upon a quantity called the posterior probability distribution of trees, which is the probability of a tree conditioned on the observations. The conditioning is accomplished using Bayes&apos;s theorem. The posterior probability distribution of trees is impossible to calculate analytically; instead, MrBayes uses a simulation technique called Markov chain Monte Carlo (or MCMC) to approximate the posterior probabilities of trees.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupportTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="87"/>
        <source>MrBayes tree calculation</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesWidget</name>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="64"/>
        <source>Substitution model</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="68"/>
        <source>Rate Matrix (fixed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="174"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="175"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PathLineEdit</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="584"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="586"/>
        <source>Select a file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Peak2GeneSupport</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneSupport.cpp" line="46"/>
        <source>&lt;i&gt;peak2gene&lt;/i&gt; - Gets refGenes near the ChIP regions identified by a peak-caller</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PerlSupport</name>
    <message>
        <location filename="../src/perl/PerlSupport.cpp" line="45"/>
        <source>Perl scripts interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForCAP3Task</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="260"/>
        <source>Failed to initialize sequence writer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="289"/>
        <source>Failed to write sequence %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForSpideyTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="215"/>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="228"/>
        <source>Failed to write DNA sequence  %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleDjangoSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="97"/>
        <source>: Python module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleNumpySupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="106"/>
        <source>: Python module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="57"/>
        <source>Python scripts interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleCelegansdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="142"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleDrosophila2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="148"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleGodbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="106"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleGostatsSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="100"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133adbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="112"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133bdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="118"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133plus2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="124"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu95av2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="130"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleMouse430a2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="136"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgceegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="166"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgdmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="172"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrghsegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="154"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgmmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="160"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleSeqlogoSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="178"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="56"/>
        <source>Rscript interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SaveAlignmentTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="55"/>
        <source>Export alignment to &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SaveMSA2SequencesTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="85"/>
        <source>Export alignment to sequence: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SearchToolsInPathTask</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="141"/>
        <source>Search tools in PATH</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SeqPosSupport</name>
    <message>
        <location filename="../src/seqpos/SeqPosSupport.cpp" line="47"/>
        <source>&lt;i&gt;SeqPos&lt;/i&gt; - Finds motifs enriched in a set of regions.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideyAlignmentTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="91"/>
        <source>Output file not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="93"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="105"/>
        <source>Failed to open result file %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupport</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="72"/>
        <source>&lt;i&gt;Spidey&lt;/i&gt; is mRNA-to-DNA alignment program.                    &lt;br&gt;Binaries can be downloaded from http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="74"/>
        <source>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="75"/>
        <source>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="76"/>
        <source>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportContext</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="115"/>
        <source>Align sequence to mRNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="142"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="143"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="273"/>
        <source>Failed to align mRNA to genomic sequence: no alignment is found.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupport</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="67"/>
        <source>&lt;i&gt;T-Coffee&lt;/i&gt; is a multiple sequence alignment package.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="77"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportContext</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="141"/>
        <source>Align with T-Coffee...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="163"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="164"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="79"/>
        <source>Unsupported alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="83"/>
        <source>T-Coffee alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="94"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="103"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="108"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="156"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="158"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="166"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="233"/>
        <source>Failed to apply the result of TCoffee: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="238"/>
        <source>T-Coffee alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="251"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="73"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="83"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="107"/>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="110"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="108"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="111"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="284"/>
        <source>input_format_error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TopHatSupportTask</name>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="51"/>
        <source>Running TopHat task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="142"/>
        <source>An unexpected error has occurred during preparing the TopHat task!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="143"/>
        <source>Preparing TopHatSupportTask internal error: unable to get a sequence object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="276"/>
        <source>There are no accepted hits in the result</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>UserAppsSettings</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="131"/>
        <source>Temporary UGENE dir is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
