<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="ru_RU">
<context>
    <name>BlastAllSupportDialog</name>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="20"/>
        <source>Request to Local BLAST Database</source>
        <translation>Запрос к локальной БД BLAST</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="42"/>
        <source>General options</source>
        <translation>Основные параметры</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="53"/>
        <source>Select search:</source>
        <translation>Тип поиска</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="61"/>
        <source>blastn</source>
        <extracomment>qwe</extracomment>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="66"/>
        <source>blastp</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="76"/>
        <source>blastx</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="81"/>
        <source>tblastn</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="86"/>
        <source>tblastx</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="94"/>
        <source>Search for short, nearly exact matches</source>
        <translation>Поиск коротких последовательностей</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="101"/>
        <source>Expectation value </source>
        <translation>Математическое ожидание</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="108"/>
        <source>Megablast</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="115"/>
        <source>Max hits</source>
        <translation>Макс. число совпадений:</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="125"/>
        <source>Off</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Select database path:</source>
        <translation type="obsolete">Путь к БД</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="191"/>
        <source>Base name for BLAST DB files:</source>
        <translation>Базовое имя для файлов БД</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="261"/>
        <source>Number of CPUs being used</source>
        <translation>Использовать процессоров</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="299"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="313"/>
        <source>Word size</source>
        <translation>Размер слова</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="546"/>
        <source>Gap costs</source>
        <translation>Вес пропуска</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="559"/>
        <source>Match scores</source>
        <translation>Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="479"/>
        <source>1 -3</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="159"/>
        <source>Both strands</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="169"/>
        <source>Direct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="176"/>
        <source>Complement</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="484"/>
        <source>1 -4</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="489"/>
        <source>1 -2</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="494"/>
        <source>1 -1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="499"/>
        <source>2 -7</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="504"/>
        <source>2 -5</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="509"/>
        <source>2 -3</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="514"/>
        <source>4 -5</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="519"/>
        <source>5 -4</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="326"/>
        <source>Matrix</source>
        <translation>Матрица</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="71"/>
        <source>gpu-blastp</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="205"/>
        <source>Database path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="229"/>
        <source>Select
a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="334"/>
        <source>BLOSUM62</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="339"/>
        <source>BLOSUM45</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="344"/>
        <source>BLOSUM80</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="349"/>
        <source>PAM30</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="354"/>
        <source>PAM70</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="368"/>
        <source>Service</source>
        <translation type="unfinished">Сервис</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="382"/>
        <source>plain</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="387"/>
        <source>psi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="392"/>
        <source>phi</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="403"/>
        <source>PHI pattern</source>
        <translation>Шаблон PHI</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="443"/>
        <source>Threshold for extending hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="446"/>
        <source>Threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="570"/>
        <source>Filters</source>
        <translation>Фильтры</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="576"/>
        <source>Low complexity filter</source>
        <translation>Фильтр низкой сложности</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="583"/>
        <source>Human repeats filter</source>
        <translation>Фильтр человеческих повторов</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="593"/>
        <source>Masks</source>
        <translation>Маски</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="599"/>
        <source>Mask for lookup table only</source>
        <translation>Только маска таблицы поиска</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="609"/>
        <source>Mask lower case letters</source>
        <translation>Маска символов нижн. регистра</translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="635"/>
        <source>Extension options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="643"/>
        <source>X dropoff value (in bits)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="649"/>
        <source>For gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="663"/>
        <source>For ungapped extensions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="670"/>
        <source>For final gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="694"/>
        <source>Multiple Hits window size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastAllSupportDialog.ui" line="706"/>
        <source>Perform gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Restore to Default</source>
        <translation type="obsolete">Значения по умолчанию</translation>
    </message>
    <message>
        <source>Search</source>
        <translation type="obsolete">Искать</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>BlastDBCmdDialog</name>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="26"/>
        <source>Fetch Sequence from BLAST Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="44"/>
        <source>Entery query ID:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="54"/>
        <source>Select database :</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="70"/>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="124"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="79"/>
        <source>Type of file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="85"/>
        <source>nucleotide</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="95"/>
        <source>protein</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="108"/>
        <source>Output path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BlastDBCmdDialog.ui" line="136"/>
        <source>Add to project</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>Bowtie2Settings</name>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="41"/>
        <source>Parameters</source>
        <translation type="unfinished">Параметры</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="60"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="76"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, Bowtie2 performs end-to-end read alignment. That is, it searches for alignments involving all of the read characters. This is also called an &amp;quot;untrimmed&amp;quot; or &amp;quot;unclipped&amp;quot; alignment.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the --local option is specified, Bowtie2 performs local read alignment. In this mode, Bowtie2 might &amp;quot;trim&amp;quot; or &amp;quot;clip&amp;quot; some read characters from one or both ends of the alignment if doing so maximizes the alignment score.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="63"/>
        <source>Mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="86"/>
        <source>--end-to-end</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="91"/>
        <source>--local</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="106"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="174"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Pads&amp;quot; dynamic programming problems by specified number of columns on either side to allow gaps.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="109"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="119"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="187"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often much slower) but increases sensitivity.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="132"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="197"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="148"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="252"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Disallow gaps within specified number of positions of the beginning or end of the read.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="161"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="229"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Use specified value as the seed for pseudo-random number generator.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="164"/>
        <source>Seed (--seed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="190"/>
        <source>Number of mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="200"/>
        <source>Threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="210"/>
        <location filename="../src/ui/Bowtie2Settings.ui" line="239"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="242"/>
        <source>Seed length (--L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="255"/>
        <source>Disallow gaps (--gbar)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="284"/>
        <source>Flags</source>
        <translation type="unfinished">Флаги</translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="296"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments for the constituent mates. This is called &amp;quot;mixed mode.&amp;quot; To disable mixed mode, set this option.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 runs a little faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="299"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="306"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="309"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="316"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the forward reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="319"/>
        <source>No forward orientation (--nofw)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="326"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the reverse-complement reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="329"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="336"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="339"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="346"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the other in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/Bowtie2Settings.ui" line="349"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BowtieBuildSettings</name>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieBuildSettings.ui" line="26"/>
        <source>Colorspace</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BowtieSettings</name>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="41"/>
        <source>Parameters</source>
        <translation>Параметры</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="110"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;not just in the &amp;quot;seed&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The default is 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;rounding can be disabled with --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="121"/>
        <source>Maq error (--maqerr)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="144"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The &amp;quot;seed length&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;i.e., the number of bases on the high-quality end of the read to which the -n ceiling applies.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="176"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The maximum insert size for valid paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;E.g. if -X 100 is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;that alignment is considered valid (as long as -I is also satisfied). &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;A 61-bp gap would not be valid in that case. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If trimming options -3 or -5 are also used, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;the -X constraint is applied with respect to the untrimmed mates, not the trimmed mates.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Default: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="215"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="246"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="60"/>
        <source>Mode:  </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="67"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the -n option is specified (which is the default), &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie determines which alignments are valid according to the following policy, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;which is similar to Maq&apos;s default policy.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="82"/>
        <source>-n mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="87"/>
        <source>-v mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="153"/>
        <source>Seed length (--seedlen)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="189"/>
        <source>Maximum of backtracks (--maxbts)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="223"/>
        <source>Descriptors memory usage (--chunkmbs)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="249"/>
        <source>Seed (--seed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="266"/>
        <source>Threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="300"/>
        <source>Flags</source>
        <translation>Флаги</translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="312"/>
        <source>Colorspace (--color)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="329"/>
        <source>No Maq rounding (--nomaqround)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="339"/>
        <source>No forward orientation (--nofw)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="349"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="369"/>
        <source>Try as hard (--tryhard)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="383"/>
        <source>Best alignments (--best)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="393"/>
        <source>All alignments (--all)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="319"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq accepts quality values in the Phred quality scale, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but internally rounds values to the nearest 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;with a maximum of 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="336"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="346"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="356"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Try as hard as possible to find valid alignments when they exist, including paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This is equivalent to specifying very high values for the --maxbts and --pairtries options.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is generally much slower than the default settings, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but can be useful for certain problems. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is slower when (a) the reference is very repetitive, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) the reads are low quality, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;or (c) not many reads have valid alignments.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="376"/>
        <source>Make Bowtie guarantee that reported singleton alignments are &quot;best&quot; in terms of stratum
(i.e. number of mismatches, or mismatches in the seed in the case of -n mode)
and in terms of the quality values at the mismatched position(s).
bowtie is somewhat slower when --best is specified.
           </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BowtieSettings.ui" line="390"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaBuildSettings</name>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="51"/>
        <source>Index algorithm (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="65"/>
        <source>bwtsw</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="70"/>
        <source>div</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaBuildSettings.ui" line="75"/>
        <source>is</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaMemSettings</name>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="39"/>
        <source>Penalty for a mismatch</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="49"/>
        <source>Score for a sequence match</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="65"/>
        <source>Score for a match (-A)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="84"/>
        <source>Mismatch penalty (-B)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="103"/>
        <source>Gap open penalty (-O)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="119"/>
        <source>Penalty for an unpaired read pair</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="129"/>
        <source>Minimum score to output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="142"/>
        <source>Penalty unpaired (-U)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="158"/>
        <source>Penalty for clipping </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="168"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="178"/>
        <source>Gap extension penalty; a gap of size k cost {-O} + {-E}*k</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="191"/>
        <source>Penalty for clipping (-L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="210"/>
        <source>Gap extention penalty (-E)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="229"/>
        <source>Score threshold (-T)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="257"/>
        <source>Internall seeds length (-r)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="276"/>
        <source>Min seed length (-k)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="292"/>
        <source>Look for internal seeds inside a seed longer than {-k} * FLOAT</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="308"/>
        <source>Number of threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="321"/>
        <source>Off-diagonal X-dropoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="334"/>
        <source>Band width for banded alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="350"/>
        <source>Band width (-w)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="366"/>
        <source>Minimum seed length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="388"/>
        <source>Number of threads (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="407"/>
        <source>Dropoff (-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="426"/>
        <source>Skip seeds threshold (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="442"/>
        <source>Skip seeds with more than INT occurrences </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="464"/>
        <source>Drop chain threshold (-D)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="480"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="496"/>
        <source>Rounds of mate rescues (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="512"/>
        <source>Perform at most INT rounds of mate rescues for each read</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="525"/>
        <source>Skip mate rescue</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="528"/>
        <source>Skip mate rescue (-S)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="535"/>
        <source>Skip pairing; mate rescue performed unless -S also in use</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="538"/>
        <source>Skip pairing (-P)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaMemSettings.ui" line="548"/>
        <source>NOTE: bwa mem accepts reads only in FASTA or FASTQ format. Reads should be compiled into a single file for each mate end.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaSettings</name>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="130"/>
        <source>  Best hits (-R)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="328"/>
        <source>Max gap extensions (-e)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="338"/>
        <source>Indel offset (-i)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="255"/>
        <source>Max long deletion extensions (-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="189"/>
        <source>Seed length (-l)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="474"/>
        <source>Max seed differences (-k)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="275"/>
        <source>Max queue entries (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="348"/>
        <source>Enable long gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="166"/>
        <source>Missing prob (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="149"/>
        <source>Max #diff (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="42"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="233"/>
        <source>Advanced</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="387"/>
        <source>Gap open penalty (-O)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="295"/>
        <source>Barcode length (-B)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="427"/>
        <source>Quality threshold (-q)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="367"/>
        <source>Mismatch penalty (-M)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="407"/>
        <source>Gap extension penalty (-E)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="454"/>
        <source>Threads (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="49"/>
        <source>Non-iterative mode (-N)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="72"/>
        <source>Max gap opens (-o)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="85"/>
        <source>Index algorithm (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="102"/>
        <source>bwtsw</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="107"/>
        <source>div</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSettings.ui" line="112"/>
        <source>is</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>BwaSwSettings</name>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="20"/>
        <source>Form</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="42"/>
        <source>Score threshold (divided by match score) (-T)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="58"/>
        <source>Number of threads (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="81"/>
        <source>Number of seeds to start rev alginment (-N)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="97"/>
        <source>Z-best (-z)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="117"/>
        <source>Prefer hard clipping in SAM output (-H)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="144"/>
        <source>Size of chunk of reads (-s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="185"/>
        <source>Score for a match (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="208"/>
        <source>Gap extention penalty (-r)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="224"/>
        <source>Mismatch penalty (-b)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="254"/>
        <source>Gap open penalty (-q)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="277"/>
        <source>Band width (-w)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="293"/>
        <source>Mask level (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="131"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Number of seed to trigger reverse alignment&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/BwaSwSettings.ui" line="323"/>
        <source>NOTE: bwa-sw performs alignment of long sequencing reads (Sanger or 454). It accepts reads only in FASTA or FASTQ format. Reads should be compiled into single file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>CAP3SupportDialog</name>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="20"/>
        <source>Contig Assembly with CAP3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="30"/>
        <source>Base</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="36"/>
        <source>Input files (long DNA reads to assembly)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="45"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Sequences in FASTA or FASTQ formats are supported. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;The quality scores for FASTA sequences can be provided in an additional file. The file must be located in the same folder as the original sequences and have the same name as FASTA file, but another extention: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="57"/>
        <source>Add</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="64"/>
        <source>Remove</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="71"/>
        <source>Remove All</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="93"/>
        <source>Result contig</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="102"/>
        <source>Path to result contig alignment in ACE format.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="128"/>
        <source>Clipping for poor regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="134"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="151"/>
        <source>Clipping range (-y)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="174"/>
        <source>Length and percent identity of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="180"/>
        <source>Overlap length cutoff (-o)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="200"/>
        <source>Overlap percent identity cutoff (-p)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="223"/>
        <source>Quality difference score of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="229"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="259"/>
        <source>Max qscore sum at differences (-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="269"/>
        <source>Other parameters</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="275"/>
        <source>Max number of word matches (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="295"/>
        <source>Band expansion size (-a)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="318"/>
        <source>Max gap length in any overlap (-f)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="338"/>
        <source>Assembly reverse reads (-r)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="364"/>
        <source>Similarity score of an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="370"/>
        <source>Mismatch score factor (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="390"/>
        <source>Overlap similarity score cutoff (-s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="410"/>
        <source>Gap penalty factor (-g)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="417"/>
        <source>Match score factor (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="112"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/CAP3SupportDialog.ui" line="122"/>
        <source>Advanced</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="32"/>
        <source>Align with Clustal Omega</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="48"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="58"/>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="75"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="65"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="87"/>
        <source>Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="97"/>
        <source>Number of iterations</source>
        <translation type="unfinished">Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="143"/>
        <source>Max number guidetree iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="186"/>
        <source>Max number of HMM iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="223"/>
        <source>Miscellaneous</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="233"/>
        <source>Number of CPUs being used</source>
        <translation type="unfinished">Использовать процессоров</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalOSupportRunDialog.ui" line="271"/>
        <source>Set options automatically</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Align</source>
        <translation type="obsolete">Выровнять</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="26"/>
        <source>Align with ClustalW</source>
        <translation>Выравнивание ClustalW</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="42"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="52"/>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="76"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="59"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="88"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="98"/>
        <source>Gap opening penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="141"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пропуск</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="184"/>
        <source>Weight matrix</source>
        <translation>Весовая матрица</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="214"/>
        <source>IUB</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="219"/>
        <source>CLUSTALW</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="289"/>
        <source>Max iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="261"/>
        <source>NONE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="266"/>
        <source>TREE</source>
        <translation>Дерево</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="271"/>
        <source>ALIGNMENT</source>
        <translation>Выравнивание</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="231"/>
        <source>Iteration type</source>
        <translation>Тип итераций</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="335"/>
        <source>Out sequences order</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="365"/>
        <source>Input</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="370"/>
        <source>Aligned</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="388"/>
        <source>Protein gap parameters</source>
        <translation>Параметры для белковых последовательностей</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="398"/>
        <source>Gap separation distance</source>
        <translation>Расстояние между пропусками</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="456"/>
        <source>Residue-specific gaps off</source>
        <translation>Откл. специфические для остатков пропуски</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="442"/>
        <source>Hydrophilic gaps off</source>
        <translation>Отключить гидрофильные пропуски</translation>
    </message>
    <message>
        <location filename="../src/ui/ClustalWSupportRunDialog.ui" line="449"/>
        <source>No end gap separation penalty</source>
        <translation>Не вводить штраф за разделение пропусков</translation>
    </message>
    <message>
        <source>Align</source>
        <translation type="obsolete">Выровнять</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>ETSSettingsWidget</name>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="26"/>
        <source>Resource settings</source>
        <translation>Настройки ресурсов</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="40"/>
        <source>You can select path for</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="47"/>
        <source>Check lastest version of external tools package</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="50"/>
        <source>&lt;a href=&quot;http://ugene.unipro.ru/external.html&quot;&gt;external tools package&lt;/a&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="70"/>
        <source>Browse...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="86"/>
        <source>Supported tools</source>
        <translation>Поддерживаемые программы</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="102"/>
        <source>Name</source>
        <translation>Имя</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="107"/>
        <source>Path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="142"/>
        <source>Additional information</source>
        <translation>Дополнительная информация</translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="177"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Select an external tool to view more information about it.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Sans Serif&apos;; font-size:9pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;table border=&quot;0&quot; style=&quot;-qt-table-type: root; margin-top:4px; margin-bottom:4px; margin-left:4px; margin-right:4px;&quot;&gt;
&lt;tr&gt;
&lt;td style=&quot;border: none;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Select an external tool to view more information about it.&lt;/p&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="obsolete">
        </translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="197"/>
        <source>Binary path:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/ETSSettingsWidget.ui" line="207"/>
        <source>Version info</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="20"/>
        <source>Format Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="29"/>
        <source>Input data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="35"/>
        <source>Select input file(s) for formatting database:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="50"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="78"/>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="184"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="59"/>
        <source>Or select directory with input files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="90"/>
        <source>File filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="98"/>
        <source>Include files filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="108"/>
        <source>*.fa;*.fasta</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="115"/>
        <source>Exclude files filter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="125"/>
        <source>*.pal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="137"/>
        <source>Type of file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="143"/>
        <source>protein</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="153"/>
        <source>nucleotide</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="166"/>
        <source>Output settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="172"/>
        <source>Select the path to save the database into:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="195"/>
        <source>Base name for BLAST files:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/FormatDBSupportRunDialog.ui" line="205"/>
        <source>Title for database file:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>IntegralBusPort</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="58"/>
        <source>One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="63"/>
        <source>Only one of these slots must be binded: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="225"/>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="235"/>
        <location filename="../src/bwa/BwaWorker.cpp" line="240"/>
        <location filename="../src/macs/MACSWorker.cpp" line="230"/>
        <source>The slot must be not empty: &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="780"/>
        <source>Error! One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="786"/>
        <source>Error! You can not bind one of sequences slots and one of url slots simultaneously</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="56"/>
        <source>&apos;%1&apos; slot must be not binded</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="20"/>
        <source>Align with MAFFT</source>
        <translation>Выравнивание MAFFT</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation type="unfinished">Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="85"/>
        <source>Gap opening penalty</source>
        <translation type="unfinished">Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="131"/>
        <source>Offset (works like gap extension penalty)</source>
        <translation type="unfinished">Смещение (работает как штраф за промежуточный пропуск)</translation>
    </message>
    <message>
        <location filename="../src/ui/MAFFTSupportRunDialog.ui" line="174"/>
        <source>Maximum number of iterative refinement</source>
        <translation type="unfinished">Максимальное число итераций</translation>
    </message>
    <message>
        <source>Align</source>
        <translation type="obsolete">Выровнять</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>MrBayesDialog</name>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="59"/>
        <source>MrBayes Settings</source>
        <translation>Настройки метода MrBayes</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="80"/>
        <source>Model</source>
        <translation>Настройки</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="92"/>
        <source>ModelType</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="109"/>
        <source>Rate</source>
        <translation>Показатель</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="126"/>
        <source>Gamma</source>
        <translation>Гамма</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="164"/>
        <source>MCMC</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="352"/>
        <source>Chain length</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="219"/>
        <source>Subsampling frequence</source>
        <translation>Размер выборки</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="251"/>
        <source>Burn-in length</source>
        <translation>Длина прожига</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="274"/>
        <source>Heated chains</source>
        <translation>Подогретые цепи</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="303"/>
        <source>Heated chain temp</source>
        <translation>Температура подогрева цепей</translation>
    </message>
    <message>
        <location filename="../src/ui/MrBayesDialog.ui" line="323"/>
        <source>Random seed</source>
        <translation>Затравка</translation>
    </message>
</context>
<context>
    <name>PhyMLDialog</name>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="47"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="59"/>
        <source>Substitution model</source>
        <translation type="unfinished">Модель замен</translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="76"/>
        <source>Equilibrium frequencies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="107"/>
        <source>optimized</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="114"/>
        <source>empirical</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="127"/>
        <source>Transition / transversion ratio </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="158"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="219"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="345"/>
        <source>Check to set user value. 
Uncheck to get the maximum likelihood estimate.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="169"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="230"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="295"/>
        <location filename="../src/ui/PhyMLDialog.ui" line="356"/>
        <source>estimated </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="188"/>
        <source>Proportion of invariable sites</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="252"/>
        <source>Number of substitution rate categories</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="314"/>
        <source>Gamma shape parameter</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="383"/>
        <source>Branch Support</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="401"/>
        <source>Use fast likelihood-based method</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="420"/>
        <source>Perform bootstrap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="436"/>
        <source>It is the number of bootstrap replicates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="449"/>
        <source>Tree Searching</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="501"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="511"/>
        <source>Type of tree improvement</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="527"/>
        <source>Set number of random starting tree</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="541"/>
        <source>Optimise topology</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/PhyMLDialog.ui" line="551"/>
        <source>Optimise branch lengths</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="127"/>
        <source>Path for temporary files</source>
        <translation type="unfinished">Путь к папке с временными файлами</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="128"/>
        <source>Path for temporary files not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="129"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="39"/>
        <source>NOTE: &quot;is&quot; index algorithm is not supposed to work with reference
sequences having size larger than 2 GB. In order to achieve stable
BWA performance it is strongly recommend to set the index algorithm
to &quot;bwtsw&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="43"/>
        <source>NOTE: &quot;bwtsw&quot; index algorithm is not supposed to work with reference
sequences having size smaller than 10 MB. In order to achieve stable
BWA performance it is strongly recommend to set the index algorithm
to &quot;is&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="249"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="271"/>
        <source>Internal error: unexpected format of the %1 output!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="176"/>
        <source>Genome sequence slot is not binded : &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="179"/>
        <source>Transcripts slot is not binded : &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="192"/>
        <source>Input transcripts</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="193"/>
        <source>Genomic sequence url [FASTA]</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="194"/>
        <source>Transcripts url [GTF]</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="200"/>
        <source>Extracted sequences url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>Output sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="208"/>
        <source>The url to the output file with the extracted sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="218"/>
        <source>Extract Transcript Sequences with gffread</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="219"/>
        <source>Extract transcript sequences from the genomic sequence(s) with gffread.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="75"/>
        <source>Installed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="76"/>
        <source>Not installed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/RnaSeqCommon.cpp" line="62"/>
        <source>Unexpected value &apos;%1&apos; of the library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="812"/>
        <source>%1 tool&apos;s version is undefined, this may cause some compatibility issues</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="821"/>
        <source>Bowtie and TopHat tools have incompatibleversions. Your TopHat&apos;s version is %1, Bowtie&apos;s one is %2. The following are considered to be compatible: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="857"/>
        <source>At least two samples are required</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="864"/>
        <source>Duplicate sample name: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="869"/>
        <source>No datasets in the sample: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SelectPathDialog</name>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="29"/>
        <source>Path settings</source>
        <translation type="unfinished">Установки пути</translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="41"/>
        <source>Select executable file of </source>
        <translation type="unfinished">Укажите исполнимый файл </translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="56"/>
        <location filename="../src/ui/SelectPathDialog.ui" line="86"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="71"/>
        <source>Select directory for tempory files:</source>
        <translation type="unfinished">Выберите директорию для временных файлов:</translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="113"/>
        <source>OK</source>
        <translation>Готово</translation>
    </message>
    <message>
        <location filename="../src/ui/SelectPathDialog.ui" line="120"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
</context>
<context>
    <name>TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="20"/>
        <source>Align with T-Coffee</source>
        <translation>Выравнивание T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="46"/>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="83"/>
        <source>Gap opening penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="122"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пропуск</translation>
    </message>
    <message>
        <location filename="../src/ui/TCoffeeSupportRunDialog.ui" line="158"/>
        <source>Number of iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <source>Align</source>
        <translation type="obsolete">Выровнять</translation>
    </message>
    <message>
        <source>Cancel</source>
        <translation type="obsolete">Отмена</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsSupport</name>
    <message>
        <location filename="../src/bedtools/BedtoolsSupport.cpp" line="52"/>
        <source>&lt;i&gt;Bedtools&lt;/i&gt;: flexible tools for genome arithmetic and DNA sequence analysis.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BigWigSupport</name>
    <message>
        <location filename="../src/bigWigTools/BigWigSupport.cpp" line="50"/>
        <source>&lt;i&gt;bedGraphToBigWig&lt;/i&gt;: converts bedGrapth to bigWig.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupport</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="78"/>
        <source>The &lt;i&gt;blastall&lt;/i&gt; is the old program developed and distributed by the NCBI for running BLAST searches. The NCBI recommends that people start using the programs of the BLAST+ package instead.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="92"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="93"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportContext</name>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="137"/>
        <source>Query with BLAST...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="157"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupport.cpp" line="158"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportMultiTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="647"/>
        <source>Source file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="648"/>
        <source>Used databse</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="650"/>
        <source>No any results found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="98"/>
        <source>Database path contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="103"/>
        <source>Database name contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="116"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="130"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="130"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportTask</name>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="75"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="80"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="93"/>
        <source>Can not create fake NCBI ini file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="207"/>
        <source>Output file not found</source>
        <translation type="unfinished">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="209"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="255"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportTask.cpp" line="290"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastAllWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="170"/>
        <source>Select input file</source>
        <translation type="unfinished">Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="190"/>
        <source>Database path contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="195"/>
        <source>Database name contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="230"/>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="275"/>
        <source>This file does not contain any sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="230"/>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="275"/>
        <source>Wrong input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllSupportRunDialog.cpp" line="320"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastDBCmdDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="43"/>
        <source>Fetch</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="44"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="75"/>
        <source>Select a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="89"/>
        <source>Set a result FASTA file name</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastDbCmdSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="72"/>
        <source>The &lt;i&gt;BlastDBCmd&lt;/i&gt; fetches protein or nucleotide sequences from BLAST+ database based on a query.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="85"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="86"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="185"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="186"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportCommonTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="86"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="91"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="104"/>
        <source>Can not create fake NCBI ini file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="149"/>
        <source>Output file not found</source>
        <translation type="unfinished">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="151"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="196"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="230"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportContext</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="235"/>
        <source>Fetch sequences by &apos;id&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="245"/>
        <source>Query with BLAST+...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="293"/>
        <source>Fetch sequences from local BLAST database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="297"/>
        <source>Fetch sequences by &apos;id&apos; %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="316"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="317"/>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="397"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="396"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportMultiTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="593"/>
        <source>Source file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="594"/>
        <source>Used databse</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="596"/>
        <source>No any results found</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="100"/>
        <source>Database path contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="105"/>
        <source>Database name contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="120"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="155"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="187"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="201"/>
        <source>Error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="201"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="238"/>
        <source>Select input file</source>
        <translation type="unfinished">Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="312"/>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="358"/>
        <source>Wrong input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="265"/>
        <source>Database path contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="277"/>
        <source>Database name contains space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="312"/>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="358"/>
        <source>This file does not contain sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="408"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="443"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="475"/>
        <source>Wrong parameters for creating annotations</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BlastRunCommonDialog</name>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="61"/>
        <source>Restore to default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="62"/>
        <source>Search</source>
        <translation type="unfinished">Искать</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="63"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="265"/>
        <source>Select a database file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="276"/>
        <source>Direct nucleotide alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="279"/>
        <source>Direct protein alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="283"/>
        <source>Direct protein alignment (on GPU)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="287"/>
        <source>Protein alignment, input nucleotide is translated input protein before the search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="291"/>
        <source>Protein alignment, nucleotide database is translated input protein before the search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="295"/>
        <source>Protein alignment, both input query and database are translated before the search</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2AlignTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="87"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="155"/>
        <source>Unknown short reads format %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2BuildIndexTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="46"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2SettingsWidget</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="108"/>
        <source>DNA Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="109"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2SettingsWidget.cpp" line="112"/>
        <source>Path for &lt;i&gt;Bowtie2&lt;/i&gt; and &lt;i&gt;Bowtie2-build&lt;/i&gt; tools is not set.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2Support</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="53"/>
        <source>&lt;i&gt;Bowtie 2 aligner&lt;/i&gt; takes a Bowtie 2 index and a set of sequencing read files and outputs a set of alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="69"/>
        <source>&lt;i&gt;Bowtie 2 build indexer&lt;/i&gt;  builds a Bowtie index from a set of DNA sequences. It outputs a set of 6 files with suffixes .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, and .rev.2.bt2. These files together constitute the index: they are all that is needed to align reads to that reference. The original sequence files are no longer used by &lt;i&gt;Bowtie 2&lt;/i&gt; once the index is built.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="90"/>
        <source>&lt;i&gt;Bowtie 2 index inspector&lt;/i&gt; extracts information from a Bowtie index about what kind of index it is and what reference sequence were used to build it.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieAssembleTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="158"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="239"/>
        <source>Unknown short reads format %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="312"/>
        <source>There is not enough memory on the computer!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieBuildIndexTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="47"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieSettingsWidget</name>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="110"/>
        <source>DNA Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="111"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettingsWidget.cpp" line="114"/>
        <source>Path for &lt;i&gt;Bowtie&lt;/i&gt; and &lt;i&gt;Bowtie-build&lt;/i&gt; tools is not set.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BowtieSupport</name>
    <message>
        <location filename="../src/bowtie/BowtieSupport.cpp" line="57"/>
        <source>&lt;i&gt;Bowtie&lt;/i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="89"/>
        <source>Short reads are not provided</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="96"/>
        <source>In paired-end mode it possible to analyze only 2 read sets using BWA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="214"/>
        <source>Log is incomplete</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="217"/>
        <source>Log is incorrect</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaMemAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="237"/>
        <source>Short reads are not provided</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="246"/>
        <source>Wrong settings of paired reads. For paired-read alignment by BWA MEM only a single pair of reads is acceptable.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaSupport</name>
    <message>
        <location filename="../src/bwa/BwaSupport.cpp" line="46"/>
        <source>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; is an efficient program that aligns relatively short nucleotide sequences against a long reference sequence such as the human genome.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaSwAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="333"/>
        <source>Short reads are not provided</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="343"/>
        <source>BWA SW can not align paired reads</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::BwaTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="469"/>
        <source>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="475"/>
        <source>Multiple read files are not supported by bwa-mem. Please combine your reads into single FASTA file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="478"/>
        <source>Please, provide two files with paired reads.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3Support</name>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="72"/>
        <source>&lt;i&gt;CAP3&lt;/i&gt; is a contig assembly program.                    &lt;br&gt;It allows to assembly long DNA reads (up to 1000 bp).                    &lt;br&gt;Binaries can be downloaded from http://seq.cs.iastate.edu/cap3.html</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="75"/>
        <source>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="76"/>
        <source>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="77"/>
        <source>&lt;br&gt;Genome Research, 9: 868-877.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="87"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="88"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportDialog</name>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="47"/>
        <source>Run</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="89"/>
        <source>List of input files is empty!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="101"/>
        <source>Result contig file name is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="107"/>
        <source>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To save under other name press &apos;Cancel&apos; and change name in &apos;Result contig&apos; field.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="110"/>
        <source>Replace</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="139"/>
        <source>Add Sequences to Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="170"/>
        <source>Set Result Contig File Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="48"/>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="111"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="170"/>
        <source>ACE format (*.ace)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="103"/>
        <source>Output file not found</source>
        <translation type="unfinished">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="105"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="119"/>
        <source>Output file not found: copy from tmp dir failed.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CEASSupport</name>
    <message>
        <location filename="../src/ceas/CEASSupport.cpp" line="46"/>
        <source>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - helps to characterize genome-wide protein-DNA interaction patterns from ChIP-chip and ChIP-Seq of both sharp and broad binding factors. It provides statistics on ChIP enrichment at important genome features such as specific chromosome, promoters, gene bodies, or exons, and infers genes most likely to be regulated by a binding factor.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CEASSupportTask</name>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="198"/>
        <source>CEAS error: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="209"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupport</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="73"/>
        <source>&lt;i&gt;Clustal Omega&lt;/i&gt; is a free sequence alignment software for proteins.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="83"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="84"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportContext</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="134"/>
        <source>Align with ClustalO...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="158"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="159"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="53"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="54"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="97"/>
        <source>ClustalO alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="118"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="127"/>
        <source>Subdirectory for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="132"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="182"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="184"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="192"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="278"/>
        <source>Failed to apply the result of ClustalO: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="283"/>
        <source>ClustalO alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="296"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="89"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="90"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="106"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="116"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="141"/>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="144"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="142"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="145"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="333"/>
        <source>input_format_error</source>
        <translation type="unfinished">Ошибка входного формата</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupport</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="73"/>
        <source>&lt;i&gt;ClustalW&lt;/i&gt; is a free sequence alignment software for DNA or proteins.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="83"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="84"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportContext</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="134"/>
        <source>Align with ClustalW...</source>
        <translation>Выравнивание ClustalW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="159"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="160"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="51"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="52"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="105"/>
        <source>Unsupported alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="108"/>
        <source>ClustalW alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="130"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Can not create directory for temporary files.</source>
        <translation type="obsolete">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="207"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="209"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="217"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="295"/>
        <source>Failed to apply the result of ClustalW: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="300"/>
        <source>ClustalW alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output file not found</source>
        <translation type="obsolete">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="313"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportRunDialog</name>
    <message>
        <source>Select input file</source>
        <translation type="obsolete">Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="130"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="131"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="153"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="163"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="214"/>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="217"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="215"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="218"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="349"/>
        <source>input_format_error</source>
        <translation type="unfinished">Ошибка входного формата</translation>
    </message>
</context>
<context>
    <name>U2::ConductGOSupport</name>
    <message>
        <location filename="../src/conduct_go/ConductGOSupport.cpp" line="45"/>
        <source>&lt;i&gt;Conduct GO&lt;/i&gt; - For a list of input genes, this tool uses R/BioC packages (GO, GOstats) to identify over represented GO terms.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConductGOTask</name>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="158"/>
        <source>Conduct GO warning: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="169"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotSupport</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotSupport.cpp" line="46"/>
        <source>&lt;i&gt;ConservationPlot&lt;/i&gt; - Draw conservation plot for many bed files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotTask</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="174"/>
        <source>Conservation Plot error: Can not find a required output file %1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="185"/>
        <source>Can not copy the result file to: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CopyFileTask</name>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="107"/>
        <source>Copy file task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="115"/>
        <source>Input file is not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="119"/>
        <source>Output file is not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="123"/>
        <source>Copy %1 to %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="127"/>
        <source>Can not rename existing file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="133"/>
        <source>Error copying file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="136"/>
        <source>File copying finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="141"/>
        <source>File copying task was finished with an error: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="143"/>
        <source>File copy was finished. Copy of &apos;%1&apos; is &apos;%2&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CuffdiffSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="43"/>
        <source>Running Cuffdiff task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="82"/>
        <source>At least 2 sets of assemblies are required for Cuffdiff</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupport</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="53"/>
        <source>&lt;i&gt;Cuffcompare&lt;/i&gt; helps comparing assembled transcripts to a reference annotation, and also tracking transcripts across multiple experiments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="69"/>
        <source>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;tests for differential expression and regulation in RNA-Seq samples.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="84"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts and estimates their abundances.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="100"/>
        <source>&lt;i&gt;Cuffmerge&lt;/i&gt; merges together several assemblies.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="113"/>
        <source>&lt;i&gt;Gffread&lt;/i&gt; is used to verify or perform various operations on GFF files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="53"/>
        <source>Running Cufflinks task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="124"/>
        <source>Unable to get an assembly object.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="254"/>
        <source>%1 output file &apos;%2&apos; is not found!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::CuffmergeSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="43"/>
        <source>Running Cuffmerge task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="130"/>
        <source>Can not create a file: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ETSProjectViewItemsContoller</name>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="57"/>
        <source>FormatDB...</source>
        <translation type="unfinished">Создать БД BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="58"/>
        <source>BLAST+ make DB...</source>
        <translation type="unfinished">Создать БД BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="83"/>
        <source>BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="99"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="102"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="104"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolJustValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="51"/>
        <source>Tool&apos;s path is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="66"/>
        <source>Tool&apos;s executable isn&apos;t exists</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="74"/>
        <source>Scripting tool registry is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="76"/>
        <source>Scripting tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="123"/>
        <source>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="154"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchAndValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="320"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="38"/>
        <source>&apos;%1&apos; external tool search task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="46"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="127"/>
        <source>An external tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="115"/>
        <source>Tool pointer is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="123"/>
        <source>External tool module hasn&apos;t any dependencies: it hasn&apos;t master tool</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="125"/>
        <source>External tool registry is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="131"/>
        <source>Tool&apos;s executable name is not set</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportPlugin</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="205"/>
        <source>External tool support</source>
        <translation type="unfinished">Поддержка внешних инструментов</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="205"/>
        <source>Runs other external tools</source>
        <translation type="unfinished">Модуль позволяет использовать внешние программы</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="264"/>
        <source>ClustalW...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="271"/>
        <source>ClustalO...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="277"/>
        <source>MAFFT...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="284"/>
        <source>T-Coffee...</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="292"/>
        <source>Multiple alignment</source>
        <translation>Множественное выравнивание</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="435"/>
        <source>FormatDB...</source>
        <translation>Создать БД BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="439"/>
        <source>BLAST+ make DB...</source>
        <translation>Создать БД BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="447"/>
        <source>BLAST Search...</source>
        <translation>Искать с помощью BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="457"/>
        <source>BLAST+ Search...</source>
        <translation>Искать с помощью BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="461"/>
        <source>BLAST+ query DB</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="469"/>
        <source>BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="38"/>
        <source>The subdirectory for temporary files &apos;%1&apos; already exists. Can not remove this directory!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="43"/>
        <source>Can not create a subdirectory for temporary files &apos;%1&apos;!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportService</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="618"/>
        <source>External tools support</source>
        <translation>Поддержка внешних инструментов</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="618"/>
        <source>Provides support to run external tools from UGENE</source>
        <translation>Позволяет запускать внешние инструменты из UGENE</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageController</name>
    <message>
        <source>External tools</source>
        <translation type="obsolete">Внешние инструменты</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="52"/>
        <source>External Tools</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageWidget</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="468"/>
        <source>Select an external tool to view more information about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="475"/>
        <source>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches. BLAST can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="481"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="484"/>
        <source>&lt;i&gt;BLAST+&lt;/i&gt; is a new version of the BLAST package from the NCBI.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="495"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. It also estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="503"/>
        <source>&lt;i&gt;Bowtie 2&lt;/i&gt; is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. &lt;br/&gt;&lt;br/&gt;It indexes the genome with an FM index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2Gb. &lt;br/&gt;&lt;br/&gt;&lt;i&gt;Bowtie 2&lt;/i&gt; supports gapped, local, and paired-end alignment modes.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="512"/>
        <source>&lt;i&gt;Cistrome&lt;/i&gt; is a UGENE version of Cistrome pipeline which also includes some tools useful for ChIP-seq analysisThis pipeline is aimed to provide the following analysis steps: peak calling and annotating, motif search and gene ontology.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="363"/>
        <source>No description</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="326"/>
        <source>External tool &apos;%1&apos; cannot be validated as it depends on other tools, some of which are not valid. The list of tools is the following: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="350"/>
        <source>&apos;%1&apos; is %2 module and it is not installed. Install it and restart UGENE or set another %2 with already installed &apos;%1&apos; module.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="373"/>
        <source>&lt;br&gt;&lt;br&gt;Version: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="377"/>
        <source>unknown</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="382"/>
        <source>&lt;br&gt;&lt;br&gt;Binary path: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="487"/>
        <source>&lt;i&gt;Bowtie&lt;i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end). &lt;a href=&apos;http://qt-project.org/doc/qt-4.8/qtextbrowser.html#anchorClicked&apos;&gt;Link text&lt;/a&gt; </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="542"/>
        <source>Choose Directory With Executables</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="599"/>
        <source>Choose Directory With External Tools Pack</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="654"/>
        <source>Not a valid external tools directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportUtils</name>
    <message>
        <source>Can not create directory for temporary files.</source>
        <translation type="obsolete">Не удаётся создать директорию для временных файлов.</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="41"/>
        <source>%1 validate task</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsInstallTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="364"/>
        <source>Installing external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="369"/>
        <source> failed: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="136"/>
        <source>Searching external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="154"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="156"/>
        <source> failed: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="335"/>
        <source>Checking external tools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="353"/>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="355"/>
        <source> failed: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupport</name>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="73"/>
        <source>The &lt;i&gt;formatdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by &lt;i&gt;blastall&lt;/i&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="89"/>
        <location filename="../src/blast/FormatDBSupport.cpp" line="104"/>
        <source>The &lt;i&gt;makeblastdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by other BLAST+ tools.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="118"/>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="121"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupport.cpp" line="123"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="50"/>
        <source>Format</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="51"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="87"/>
        <source>Select file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="110"/>
        <source>Select a directory with input files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="121"/>
        <source>Select a directory to save database files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="133"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="139"/>
        <source>Input files paths contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="147"/>
        <location filename="../src/blast/FormatDBSupportRunDialog.cpp" line="152"/>
        <source>Output database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportTask</name>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="47"/>
        <source>Run NCBI FormatDB task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="61"/>
        <source>Input files paths contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="78"/>
        <source>Output database path contain space characters.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="97"/>
        <source>Blast database creation has been cancelled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="104"/>
        <source>Blast database has been successfully created</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="105"/>
        <source>Source sequences: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="113"/>
        <source>Database file path: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="115"/>
        <source>Type: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="117"/>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="123"/>
        <source>Formatdb log file path: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/FormatDBSupportTask.cpp" line="121"/>
        <source>Blast database creation has been failed</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GffreadSupportTask</name>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="34"/>
        <source>Running Gffread task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="64"/>
        <source>Unknown file format: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="73"/>
        <source>The file format is not [%1]: %2</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BWAMEMPrompter</name>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="457"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="461"/>
        <source>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="462"/>
        <source> to reference genome &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BWAMEMWorker</name>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="261"/>
        <source>BWA MEM cannot recognize read pairs from the same file. Please, perform demultiplexing first.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="275"/>
        <source>URL of a file with reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="276"/>
        <source>Input reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="278"/>
        <source>URL of a file with mate reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="279"/>
        <source>Input mate reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="285"/>
        <source>BWA data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="286"/>
        <source>Input reads to be aligned with BWA MEM.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="294"/>
        <source>Assembly URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="295"/>
        <source>Output assembly URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="298"/>
        <source>BWA MEM output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="299"/>
        <source>Output assembly files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="310"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="311"/>
        <source>Directory to save BWA MEM output files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="314"/>
        <source>Reference genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="315"/>
        <source>Path to indexed reference genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="318"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="319"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="322"/>
        <source>Number of threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="323"/>
        <source>Number of threads (-t).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="326"/>
        <source>Min seed length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="327"/>
        <source>Path to indexed reference genome (-k).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="330"/>
        <source>Band width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="331"/>
        <source>Band width for banded alignment (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="334"/>
        <source>Dropoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="335"/>
        <source>Off-diagonal X-dropoff (-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="338"/>
        <source>Internal seed length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="339"/>
        <source>Look for internal seeds inside a seed longer than {-k} (-r).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="342"/>
        <source>Skip seed threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="343"/>
        <source>Skip seeds with more than INT occurrences (-c).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="346"/>
        <source>Drop chain threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="347"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain (-D).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="350"/>
        <source>Rounds of mate rescues</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="351"/>
        <source>Perform at most INT rounds of mate rescues for each read (-m).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="354"/>
        <source>Skip mate rescue</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="355"/>
        <source>Skip mate rescue (-S).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="358"/>
        <source>Skip pairing</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="359"/>
        <source>Skip pairing; mate rescue performed unless -S also in use (-P).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="362"/>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="366"/>
        <source>Mismatch penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="363"/>
        <source>Score for a sequence match (-A).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="367"/>
        <source>Penalty for a mismatch (-B).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="370"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="371"/>
        <source>Gap open penalty (-O).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="374"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished">Штраф за промежуточный пропуск</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="375"/>
        <source>Gap extension penalty; a gap of size k cost {-O} (-E).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="378"/>
        <source>Penalty for clipping</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="379"/>
        <source>Penalty for clipping (-L).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="382"/>
        <source>Penalty unpaired</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="383"/>
        <source>Penalty for an unpaired read pair (-U).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="386"/>
        <source>Score threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="387"/>
        <source>Minimum score to output (-T).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="436"/>
        <source>Align reads with BWA MEM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="437"/>
        <source>Performs alignment of short reads with BWA MEM.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BWAPrompter</name>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="499"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="503"/>
        <source>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="504"/>
        <source> to reference genome &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BWAWorker</name>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="265"/>
        <source>BWA cannot recognize read pairs from the same file. Please, perform demultiplexing first.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="279"/>
        <source>URL of a file with reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="280"/>
        <source>Input reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="282"/>
        <source>URL of a file with mate reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="283"/>
        <source>Input mate reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="289"/>
        <source>BWA data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="290"/>
        <source>Input reads to be aligned with BWA.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="298"/>
        <source>Assembly URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="299"/>
        <source>Output assembly URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="302"/>
        <source>BWA output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="303"/>
        <source>Output assembly files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="314"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="315"/>
        <source>Directory to save BWA output files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="318"/>
        <source>Reference genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="319"/>
        <source>Path to indexed reference genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="322"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="323"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="326"/>
        <source>Use missing prob</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="327"/>
        <source>Use missing prob instead maximum edit distance.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="330"/>
        <source>Max #diff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="331"/>
        <source>Max #diff (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="334"/>
        <source>Missing prob</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="335"/>
        <source>Missing prob (-n)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="338"/>
        <source>Seed length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="339"/>
        <source>Seed length (-l).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="342"/>
        <source>Max gap opens</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="343"/>
        <source>Max gap opens (-o).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="346"/>
        <source>Index algorithm</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="347"/>
        <source>Index algorithm (-a).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="350"/>
        <source>Best hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="351"/>
        <source>Best hits (-R).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="354"/>
        <source>Long-scaled gap penalty for long deletions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="355"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="358"/>
        <source>Non iterative mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="359"/>
        <source>Non iterative mode (-N).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="362"/>
        <location filename="../src/bwa/BwaWorker.cpp" line="363"/>
        <source>Enable long gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="366"/>
        <source>Max gap extensions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="367"/>
        <source>Max gap extensions (-e)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="370"/>
        <source>Indel offset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="371"/>
        <source>Indel offset (-i)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="374"/>
        <source>Max long deletions extensions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="375"/>
        <source>Max long deletions extensions(-d)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="378"/>
        <source>Max queue entries</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="379"/>
        <source>Max queue entries (-m)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="382"/>
        <source>Barcode length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="383"/>
        <source>Barcode length (-B)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="386"/>
        <source>Threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="387"/>
        <source>Threads (-t)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="390"/>
        <source>Max seed differencies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="391"/>
        <source>Max seed differencies (-k)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="394"/>
        <source>Mismatch penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="395"/>
        <source>Mismatch penalty (-M)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="398"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="399"/>
        <source>Gap open penalty (-O).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="402"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished">Штраф за промежуточный пропуск</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="403"/>
        <source>Gap extension penalty; a gap of size k cost (-E).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="406"/>
        <source>Quolity threshold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="407"/>
        <source>Quolity threshold (-q).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="478"/>
        <source>Align reads with BWA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="479"/>
        <source>Performs alignment of short reads with BWA.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigPrompter</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="75"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="76"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="78"/>
        <source>Converts bedGraph files to bigWig %1 with bedGraphToBigWig.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigWorker</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="95"/>
        <source>Convert bedGraph Files to bigWig</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="96"/>
        <source>Convert bedGraph Files to bigWig.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="100"/>
        <source>BedGrapgh files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="101"/>
        <source>Set of bedGraph files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="102"/>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="103"/>
        <source>BigWig files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="117"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="118"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="122"/>
        <source>Custom directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="123"/>
        <source>Select the custom output directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="125"/>
        <source>Output name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="126"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first BAM file with an extention.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="128"/>
        <source>Block size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="129"/>
        <source>Number of items to bundle in r-tree (-blockSize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="131"/>
        <source>Items per slot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="132"/>
        <source>Number of data points bundled at lowest level (-itemsPerSlot).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="134"/>
        <source>Uncompressed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="135"/>
        <source>If set, do not use compression (-unc).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="137"/>
        <source>Genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="138"/>
        <source>File with genome length.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="142"/>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="170"/>
        <source>Custom</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="168"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="169"/>
        <source>Workflow</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllPrompter</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="324"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="325"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="326"/>
        <source>For sequence %1 find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllWorker</name>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="78"/>
        <source>Input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="79"/>
        <source>Sequence for which annotations is searched.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="80"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="80"/>
        <source>Found annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="89"/>
        <source>Search type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="90"/>
        <source>Select type of BLAST searches.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="91"/>
        <source>Database Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="92"/>
        <source>Path with database files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="93"/>
        <source>Database Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="94"/>
        <source>Base name for BLAST DB files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="95"/>
        <source>Expected value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="96"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="97"/>
        <source>Annotate as</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="98"/>
        <source>Name for annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="99"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="100"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="101"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="102"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="104"/>
        <source>BLAST output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="105"/>
        <source>Location of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="106"/>
        <source>BLAST output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="107"/>
        <source>Type of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="108"/>
        <source>Gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="109"/>
        <source>Perform gapped alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="111"/>
        <source>Gap costs</source>
        <translation type="unfinished">Вес пропуска</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="112"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="113"/>
        <source>Match scores</source>
        <translation type="unfinished">Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="114"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="169"/>
        <source>Local BLAST Search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Local BLAST search</source>
        <translation type="obsolete">Локальный BLAST поиск</translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="170"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="381"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="390"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="395"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast/BlastAllWorker.cpp" line="402"/>
        <source>Not selected BLAST output file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusPrompter</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="219"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="220"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="221"/>
        <source>For sequence &lt;u&gt;%1&lt;/u&gt; find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusWorker</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="74"/>
        <source>Input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="75"/>
        <source>Sequence for which annotations is searched.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="76"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="76"/>
        <source>Found annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="85"/>
        <source>Search type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="87"/>
        <source>Database Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="89"/>
        <source>Database Name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="91"/>
        <source>Expected value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="92"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="93"/>
        <source>Annotate as</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="95"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="97"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="99"/>
        <source>BLAST output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="100"/>
        <source>Location of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="101"/>
        <source>BLAST output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="102"/>
        <source>Type of BLAST output file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="103"/>
        <source>Gapped alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="86"/>
        <source>Select type of BLAST+ searches.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="88"/>
        <source>Path with database files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="90"/>
        <source>Base name for BLAST+ DB files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="94"/>
        <source>Name for annotations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="96"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="98"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="104"/>
        <source>Perform gapped alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="106"/>
        <source>Gap costs</source>
        <translation type="unfinished">Вес пропуска</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="107"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="108"/>
        <source>Match scores</source>
        <translation type="unfinished">Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="109"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="136"/>
        <source>Local BLAST+ Search</source>
        <translation>Локальный BLAST+ поиск</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="137"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="286"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="295"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="300"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="307"/>
        <source>Not selected BLAST output file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BowtiePrompter</name>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="468"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="472"/>
        <source>Aligns reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="473"/>
        <source> to reference genome &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BowtieWorker</name>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="250"/>
        <source>Bowtie cannot recognize read pairs from the same file. Please, perform demultiplexing first.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="264"/>
        <source>URL of a file with reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="265"/>
        <source>Input reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="267"/>
        <source>URL of a file with mate reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="268"/>
        <source>Input mate reads to be aligned.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="274"/>
        <source>Bowtie data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="275"/>
        <source>Input reads to be aligned with Bowtie.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="283"/>
        <source>Assembly URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="284"/>
        <source>Output assembly URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="287"/>
        <source>Bowtie output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="288"/>
        <source>Output assembly files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="299"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="300"/>
        <source>Directory to save Bowtie output files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="303"/>
        <source>Reference genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="304"/>
        <source>Path to indexed reference genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="307"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="308"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="329"/>
        <source>Mode:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="330"/>
        <source>When the -n option is specified (which is the default), bowtie determines which alignments                               are valid according to the following policy, which is similar to Maq&apos;s default policy.                               In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="336"/>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="340"/>
        <source>Mismatches number</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="337"/>
        <source>Mismatches number.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="341"/>
        <source>Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment,                              not just in the seed. The default is 70. Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30;                              rounding can be disabled with --nomaqround.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="346"/>
        <source>Seed length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="347"/>
        <source>The seed length; i.e., the number of bases on the high-quality end of the read to which the                               -n ceiling applies. The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="351"/>
        <source>No forward orientation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="352"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="355"/>
        <source>No reverse-complement orientation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="356"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="359"/>
        <source>Maximum of backtracks</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="360"/>
        <source>The maximum insert size for valid paired-end alignments. E.g. if -X 100 is specified and a paired-end alignment                               consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, that alignment is                               considered valid (as long as -I is also satisfied). A 61-bp gap would not be valid in that case.                                If trimming options -3 or -5 are also used, the -X constraint is applied with respect to the untrimmed mates,                               not the trimmed mates. Default: 250.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="367"/>
        <source>Try as hard</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="368"/>
        <source>Try as hard as possible to find valid alignments when they exist, including paired-end alignments.                                This is equivalent to specifying very high values for the --maxbts and --pairtries options. This mode is generally much                                slower than the default settings, but can be useful for certain problems. This mode is slower when (a) the reference is                                very repetitive, (b) the reads are low quality, or (c) not many reads have valid alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="374"/>
        <source>Best hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="375"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. Best-first                                search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost.                                Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases.                                If you receive an error message saying that chunk memory has been exhausted in --best mode,                                try adjusting this parameter up to dedicate more memory to the descriptors. Default: 64.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="382"/>
        <source>No Maq rounding</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="383"/>
        <source>Maq accepts quality values in the Phred quality scale, but internally rounds values to the nearest 10,              with a maximum of 30. By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="387"/>
        <source>Seed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="388"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="391"/>
        <source>Best alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="392"/>
        <source>Make Bowtie guarantee that reported singleton alignments are best in terms of stratum                              (i.e. number of mismatches, or mismatches in the seed in the case of -n mode) and in terms of                              the quality values at the mismatched position(s). bowtie is somewhat slower when --best is specified.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="397"/>
        <source>All alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="398"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="401"/>
        <source>Colorspace</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="402"/>
        <source>When -C is specified, read sequences are treated as colors. Colors may be encoded either as numbers                               (0=blue, 1=green, 2=orange, 3=red) or as characters A/C/G/T (A=blue, C=green, G=orange, T=red).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="447"/>
        <source>Align reads with Bowtie</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="448"/>
        <source>Performs alignment of short reads with Bowtie.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Prompter</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="377"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Worker</name>
    <message>
        <source>Align with CAP3</source>
        <translation type="obsolete">Выравнивание CAP3</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="89"/>
        <source>Input sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="90"/>
        <source>DNA sequences that need to be assembled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="94"/>
        <source>Input URL(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="95"/>
        <source>URL(s) of input file(s) in FASTA format with DNA sequences that need to be assembled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="106"/>
        <source>Assembly Sequences with CAP3</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="111"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="112"/>
        <source>Write assembly results to this output file in ACE format.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="115"/>
        <source>Quality cutoff for clipping</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="116"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="119"/>
        <source>Clipping range</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="120"/>
        <source>One of the parameters to control clipping of a poor end region of a read (-y)&lt;p&gt;If there are quality values, CAP3 computes two positions qualpos5 and qualpos3 of read f such that the region of read f from position qualpos5 to position qualpos3 consists mostly of quality values greater than &lt;i&gt;Quality cutoff for clipping&lt;/i&gt;.&lt;/p&gt; &lt;p&gt;If there are no quality values, then qualpos5 is set to 1 and qualpos3 is set the length of read f. &lt;p&gt;The range for the left clipping position of read f is from 1 to qualpos5 + &lt;i&gt;Clipping range&lt;/i&gt;. &lt;br/&gt;The range for the right clipping position of read f is from qualpos3 - &lt;i&gt;Clipping range&lt;/i&gt; to the end of read f.&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="133"/>
        <source>Quality cutoff for differences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="134"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="137"/>
        <source>Maximum difference score</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="138"/>
        <source>Max qscore sum at differences (-d)&lt;p&gt;If an overlap contains lots of differences at bases of high quality, then the overlap is removed.&lt;/p&gt;&lt;p&gt;The difference score is calculated as follows. If the overlap contains a difference at bases of quality values q1 and q2, then the score at the difference is max(0, min(q1, q2) - b), where b is &lt;i&gt;Quality cutoff for differences&lt;/i&gt;. The difference score of an overlap is the sum of scores at each difference.&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="149"/>
        <source>Match score factor</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="150"/>
        <source>Match score factor (-m) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="155"/>
        <source>Mismatch score factor</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="156"/>
        <source>Mismatch score factor (-n) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="161"/>
        <source>Gap penalty factor</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="162"/>
        <source>Gap penalty factor (-g) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="167"/>
        <source>Overlap similarity score cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="168"/>
        <source>If the similarity score of an overlap is less than the overlap similarity score cutoff (-s), then the overlap is removed.&lt;p&gt;The similarity score of an overlapping alignment is defined using base quality values as follows.&lt;p&gt;A match at bases of quality values q1 and q2 is given a score of m * min(q1,q2), where m is &lt;i&gt;Match score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A mismatch at bases of quality values q1 and q2 is given a score of n * min(q1,q2), where n is &lt;i&gt;Mismatch score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A base of quality value q1 in a gap is given a score of -g * min(q1,q2), where q2 is the quality value of the base in the other sequence right before the gap and g is &lt;i&gt;Gap penalty factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;The score of a gap is the sum of scores of each base in the gap minus a gap open penalty.&lt;/p&gt;&lt;p&gt;The similarity score of an overlapping alignment is the sum of scores of each match, each mismatch, and each gap. &lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="186"/>
        <source>Overlap length cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="187"/>
        <source>An overlap is taken into account only if the length of the overlap in bp is no less than the specified value (parameter -o of CAP3).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="191"/>
        <source>Overlap percent identity cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="192"/>
        <source>An overlap is taken into account only if the percent identity of the overlap is no less than the specified value (parameter -p of CAP3).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="196"/>
        <source>Max number of word matches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="197"/>
        <source>This parameter allows one to trade off the efficiency of the program for its accuracy (parameter -t of CAP3).&lt;p&gt;For a read f, CAP3 computes overlaps between read f and other reads by considering short word matches between read f and other reads. A word match is examined to see if it can be extended into a long overlap. If read f has overlaps with many other reads, then read f has many short word matches with many other reads.&lt;/p&gt;&lt;p&gt;This parameter gives an upper limit, for any word, on the number of word matches between read f and other reads that are considered by CAP3.&lt;/p&gt;&lt;p&gt;Using a large value for this parameter allows CAP3 to consider more word matches between read f and other reads, which can find more overlaps for read f, but slows down the program.&lt;/p&gt;&lt;p&gt;Using a small value for this parameter has the opposite effect.&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="212"/>
        <source>Band expansion size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="213"/>
        <source>CAP3 determines a minimum band of diagonals for an overlapping alignment between two sequence reads. The band is expanded by a number of bases specified by this value (parameter -a of CAP3).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="218"/>
        <source>Max gap length in an overlap</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="219"/>
        <source>The maximum length of gaps allowed in any overlap (-f)&lt;p&gt;I.e. overlaps with longer gaps are rejected.&lt;/p&gt;&lt;p&gt;Note that a small value for this parameter may cause the program to remove true overlaps and to produce incorrect results.&lt;/p&gt;&lt;p&gt;The parameter may be used to split reads from alternative splicing forms into separate contigs.&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="227"/>
        <source>Assembly reverse reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="228"/>
        <source>Specifies whether to consider reads in reverse orientation for assembly (originally, parameter -r of CAP3).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="232"/>
        <source>CAP3 tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="233"/>
        <source>The path to the CAP3 external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="236"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="237"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="378"/>
        <source>Aligns long DNA reads from &lt;u&gt;%1&lt;/u&gt; with CAP3.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportPrompter</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="373"/>
        <source>Creates summary statistics on ChIP enrichment and saves it to %1.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportWorker</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="216"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="217"/>
        <source>Peak regions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="219"/>
        <source>Enrichment signal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="220"/>
        <source>Enrichment signal.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="225"/>
        <source>CEAS data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="226"/>
        <source>Peak regions and enrichment signal profile for generating CEAS report.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="235"/>
        <source>Output report file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="236"/>
        <source>Path to the report output file. Result for CEAS analysis.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="241"/>
        <source>Output annotations file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="242"/>
        <source>Name of tab-delimited output text file, containing a row of annotations for every RefSeq gene. (file is not generated if no peak location data is supplied).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="245"/>
        <source>Gene annotations table</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="246"/>
        <source>Path to gene annotation table (e.g. a refGene table in sqlite3 db format (--gt).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="249"/>
        <source>Span size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="250"/>
        <source>Span from TSS and TTS in the gene-centered annotation (base pairs). ChIP regions within this range from TSS and TTS are considered when calculating the coverage rates in promoter and downstream (--span).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="255"/>
        <source>Wiggle profiling resolution</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="256"/>
        <source>Wiggle profiling resolution. WARNING: Value smaller than the wig interval (resolution) may cause aliasing error (--pf-res).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="260"/>
        <source>Promoter/downstream interval</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="261"/>
        <source>Promoter/downstream intervals for ChIP region annotation are three values or a single value can be given. If a single value is given, it will be segmented into three equal fractions (e.g. 3000 is equivalent to 1000,2000,3000) (--rel-dist).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="267"/>
        <source>BiPromoter ranges</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="268"/>
        <source>Bidirectional-promoter sizes for ChIP region annotation. It&apos;s two values or a single value can be given. If a single value is given, it will be segmented into two equal fractions (e.g. 5000 is equivalent to 2500,5000) (--bisizes).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="273"/>
        <source>Relative distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="274"/>
        <source>Relative distance to TSS/TTS in WIGGLE file. profiling. (--rel-dist)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="277"/>
        <source>Gene group files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="278"/>
        <source>Gene groups of particular interest in wig profiling. Each gene group file must have gene names in the 1st column. The file names are separated by commas (--gn-groups).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="282"/>
        <source>Gene group names</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="283"/>
        <source>&lt;i&gt;Set this parameter empty for using default values.&lt;/i&gt;&lt;br&gt;The names of the gene groups from &quot;Gene group files&quot; parameter. These names appear in the legends of the wig profiling plots.&lt;br&gt; Values range: comma-separated list of strings. Default value: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="323"/>
        <source>Report file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="330"/>
        <source>Annotations file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="353"/>
        <source>Create CEAS Report</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="354"/>
        <source>Provides summary statistics on ChIP enrichment in important genomic regions such as individual chromosomes, promoters, gene bodies or exons, and infers the genes most likely to be regulated by the binding factor under study.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOPrompter</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="139"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="140"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOWorker</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="61"/>
        <source>Input MSA to process.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="62"/>
        <source>ClustalO result MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="63"/>
        <source>The result of the ClustalO alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="72"/>
        <source>Number of iterations</source>
        <translation type="unfinished">Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="73"/>
        <source>Number of (combined guide-tree/HMM) iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="74"/>
        <source>Number of guidetree iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="75"/>
        <source>Maximum number guidetree iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="76"/>
        <source>Number of HMM iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="77"/>
        <source>Maximum number of HMM iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="78"/>
        <source>Set auto options</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="79"/>
        <source>Set options automatically (might overwrite some of your options).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="81"/>
        <source>Tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="82"/>
        <source>Path to the ClustalO tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="85"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="86"/>
        <source>Directory to store temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="96"/>
        <source>Align with ClustalO</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="97"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalO.&lt;p&gt;ClustalO is a general purpose multiple sequence alignment program for proteins.Visit &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; to learn more about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="184"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="216"/>
        <source>Aligned %1 with ClustalO</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWPrompter</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="188"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="189"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWWorker</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="64"/>
        <source>Input MSA to process.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="65"/>
        <source>ClustalW result MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="66"/>
        <source>The result of the ClustalW alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="75"/>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="76"/>
        <source>The penalty for opening a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="77"/>
        <source>Gap extension penalty</source>
        <translation type="unfinished">Штраф за промежуточный пропуск</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="78"/>
        <source>The penalty for extending a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="79"/>
        <source>Gap distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="80"/>
        <source>The gap separation penalty. Tries to decrease the chances of gaps being too close to each other.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="81"/>
        <source>End gaps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="82"/>
        <source>The penalty for closing a gap.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="83"/>
        <source>Residue-specific gaps off</source>
        <translation type="unfinished">Откл. спецефические для остатков пропуски</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="84"/>
        <source>Residue-specific penalties are amino specific gap penalties that reduce or increase the gap opening penalties at each position in the alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="85"/>
        <source>Hydrophilic gaps off</source>
        <translation type="unfinished">Отключить гидрофильные пропуски</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="86"/>
        <source>Hydrophilic gap penalties are used to increase the chances of a gap within a run (5 or more residues) of hydrophilic amino acids.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="87"/>
        <source>Iteration type</source>
        <translation type="unfinished">Тип итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="88"/>
        <source>Alignment improvement iteration type. Can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;None - No iteration;&lt;/li&gt;                            &lt;li&gt;Tree - Iteration at each step of alignment process;&lt;/li&gt;                            &lt;li&gt;Alignment - Iteration only on final alignment.&lt;/li&gt;                            &lt;/ul&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="94"/>
        <source>Number of iterations</source>
        <translation type="unfinished">Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="95"/>
        <source>The maximum number of iterations to perform.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="96"/>
        <source>Weight matrix</source>
        <translation type="unfinished">Весовая матрица</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="110"/>
        <source>Tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="111"/>
        <source>Path to the ClustalW tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="115"/>
        <source>Directory to store temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="130"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalW.&lt;p&gt;ClustalW is a general purpose multiple sequence alignment program for DNA or proteins.Visit &lt;a href=&quot;http://www.clustal.org/&quot;&gt;http://www.clustal.org/&lt;/a&gt; to learn more about it.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="260"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="114"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="97"/>
        <source>For proteins it is a scoring table which describes the similarity of each amino acid to each other and can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - Appear to be the best available for carrying out database similarity (homology searches).                            &lt;li&gt;PAM - Have been extremely widely used since the late &apos;70s.                            &lt;li&gt;Gonnet - Derived using almost the same procedure as the PAM bit are much more up to date.                            &lt;li&gt;ID - Identitical matrix gives a score of 1.0 to two identical amino acids and a score of zero otherwise.                            &lt;/ul&gt;                            &lt;p&gt;                            For DNA it is the scores assigned to matches and mismatches. The values available are:                            &lt;ul&gt;                            &lt;li&gt;IUB - The default scoring matrix used by BESTFIT for the comparison of nucleic acid sequences. All matches score 1.9; all mismatches for IUB symbols score 0.                           &lt;li&gt;ClustalW - The previous system used by Clustal W, in which matches score 1.0 and mismatches score 0. All matches for IUB symbols also score 0.                            &lt;/ul&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="129"/>
        <source>Align with ClustalW</source>
        <translation>Выравнивание ClustalW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="289"/>
        <source>Aligned %1 with ClustalW</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOPrompter</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="219"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="224"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as gene list to conduct GO.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="226"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOWorker</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="143"/>
        <source>Target genes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="144"/>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="149"/>
        <source>Gene list to identify over represented GO terms.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="148"/>
        <source>Conduct GO data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="157"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="158"/>
        <source>The directory to store Conduct GO results.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="160"/>
        <source>Title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="161"/>
        <source>Title is used to name the output files - so make it meaningful.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="163"/>
        <source>Gene Universe</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="164"/>
        <source>Select a gene universe.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="199"/>
        <source>Conduct GO</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="200"/>
        <source>Given a list of genes, using Bioconductor (GO, GOstats) and  DAVID at NIH.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotPrompter</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="289"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="295"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions for conservation plot.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="296"/>
        <source> Conservations scores from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="297"/>
        <source> Outputs the result to &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotWorker</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="189"/>
        <source>Input regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="190"/>
        <source>Regions (centered at peak summits for better performance).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="194"/>
        <source>conservation_plot data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="195"/>
        <source>Regions to plot the PhastCons scores profiles.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="203"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="204"/>
        <source>File to store phastcons results (BMP).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="206"/>
        <source>Title</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="207"/>
        <source>Title of the figure (--title).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="209"/>
        <source>Label</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="210"/>
        <source>Label of data in the figure (--bed-label).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="212"/>
        <source>Assembly version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="213"/>
        <source>The directory to store phastcons scores (--phasdb).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="215"/>
        <source>Window width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="216"/>
        <source>Window width centered at middle of regions (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="218"/>
        <source>Height</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="219"/>
        <source>Height of plot (--height).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="221"/>
        <source>Width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="222"/>
        <source>Width of plot (--width).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="247"/>
        <source>Conservation plot file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="269"/>
        <source>Build Conservation Plot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="270"/>
        <source>Plots the PhastCons scores profiles.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffdiffWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="93"/>
        <source>Test for Diff. Expression with Cuffdiff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="107"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="108"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="111"/>
        <source>Time series analysis</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="112"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, instructs Cuffdiff to analyze the provided samples as a time series, rather than testing for differences between all pairs of samples. Samples should be provided in increasing time order.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="118"/>
        <source>Upper quartile norm</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="119"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, normalizes by the upper quartile of the number of fragments mapping to individual loci instead of the total number of sequenced fragments. This can improve robustness of differential expression calls for less abundant genes and transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="125"/>
        <source>Hits norm</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="126"/>
        <source>Instructs how to count all fragments. &lt;i&gt;Total&lt;/i&gt; specifies to count all fragments, including those not compatible with any reference transcript, towards the number of mapped fragments used in the FPKM denominator. &lt;i&gt;Compatible&lt;/i&gt; specifies to use only compatible fragments. Selecting &lt;i&gt;Compatible&lt;/i&gt; is generally recommended in Cuffdiff to reduce certain types of bias caused by differential amounts of ribosomal reads which can create the impression of falsely differentially expressed genes.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="135"/>
        <source>Frag bias correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="136"/>
        <source>Providing the sequences your reads were mapped to instructs Cuffdiff to run bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="141"/>
        <source>Multi read correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="142"/>
        <source>Do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="146"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="147"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="150"/>
        <source>Mask file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="151"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="159"/>
        <source>Min alignment count</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="160"/>
        <source>The minimum number of alignments in a locus for needed to conduct significance testing on changes in that locus observed between samples. If no testing is performed, changes in the locus are deemed not significant, and the locus&apos; observed changes don&apos;t contribute to correction for multiple testing.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="167"/>
        <source>FDR</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="168"/>
        <source>The allowed false discovery rate used in testing.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="171"/>
        <source>Max MLE iterations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="172"/>
        <source>Sets the number of iterations allowed during maximum likelihood estimation of abundances.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="176"/>
        <source>Emit count tables</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="177"/>
        <source>Include information about the fragment counts, fragment count variances, and fitted variance model into the report.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="181"/>
        <source>Cuffdiff tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="182"/>
        <source>The path to the Cuffdiff external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="185"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="186"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="206"/>
        <source>Assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="207"/>
        <source>RNA-Seq reads assemblies</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="209"/>
        <source>Annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="210"/>
        <source>Transcript annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="212"/>
        <source>Sample</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="213"/>
        <source>Sample name of assembly file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="299"/>
        <source>Searches for significant changes in transcript expression, splicing and promoter use.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CufflinksWorker</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="97"/>
        <source>Input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="98"/>
        <source>Input RNA-Seq read alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="101"/>
        <source>Output annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="102"/>
        <source>Assembled isoforms, estimated isoform-level expression values, and estimated gene-level expression values, produced by Cufflinks.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="115"/>
        <source>Isoform-level expression values</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="116"/>
        <source>A set of annotated regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="133"/>
        <source>Assembly Transcripts with Cufflinks</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="134"/>
        <source>Cufflinks accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="142"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="143"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="146"/>
        <source>Reference annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="147"/>
        <source>Tells Cufflinks to use the supplied reference annotation to estimate isoform expression. Cufflinks will not assemble novel transcripts and the program will ignore alignments not structurally compatible with any reference transcript.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="153"/>
        <source>RABT annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="154"/>
        <source>Tells Cufflinks to use the supplied reference annotation to guide Reference Annotation Based Transcript (RABT) assembly. Reference transcripts will be tiled with faux-reads to provide additional information in assembly. Output will include all reference transcripts as well as any novel genes and isoforms that are assembled.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="161"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="162"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="165"/>
        <source>Mask file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="166"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="174"/>
        <source>Multi-read correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="175"/>
        <source>Tells Cufflinks to do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="179"/>
        <source>Min isoform fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="180"/>
        <source>After calculating isoform abundance for a gene, Cufflinks filters out transcripts that it believes are very low abundance, because isoforms expressed at extremely low levels often cannot reliably be assembled, and may even be artifacts of incompletely spliced precursors of processed transcripts. This parameter is also used to filter out introns that have far fewer spliced alignments supporting them.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="188"/>
        <source>Frag bias correct</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="189"/>
        <source>Providing Cufflinks with a multifasta file via this option instructs it to run the bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="194"/>
        <source>Pre-mRNA fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="195"/>
        <source>Some RNA-Seq protocols produce a significant amount of reads that originate from incompletely spliced transcripts, and these reads can confound the assembly of fully spliced mRNAs. Cufflinks uses this parameter to filter out alignments that lie within the intronic intervals implied by the spliced alignments. The minimum depth of coverage in the intronic region covered by the alignment is divided by the number of spliced reads, and if the result is lower than this parameter value, the intronic alignments are ignored.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="205"/>
        <source>Cufflinks tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="206"/>
        <source>The path to the Cufflinks external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="209"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="210"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="337"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffmergeWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="63"/>
        <source>Merge Assemblies with Cuffmerge</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="64"/>
        <source>Cuffmerge merges together several assemblies. It also handles running Cuffcompare for you, and automatically filters a number of transfrags that are probably artifacts. If you have a reference file available, you can provide it to Cuffmerge in order to gracefully merge input (e.g. novel) isoforms and known isoforms and maximize overall assembly quality.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="74"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="75"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="78"/>
        <source>Reference annotation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="79"/>
        <source>Merge the input assemblies together with this reference annotation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="83"/>
        <source>Reference sequence</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="84"/>
        <source>The genomic DNA sequences for the reference. It is used to assist in classifying transfrags and excluding artifacts (e.g. repeats). For example, transcripts consisting mostly of lower-case bases are classified as repeats.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="90"/>
        <source>Minimum isoform fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="91"/>
        <source>Discard isoforms with abundance below this.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="94"/>
        <source>Cuffcompare tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="95"/>
        <source>The path to the Cuffcompare external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="98"/>
        <source>Cuffmerge tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="99"/>
        <source>The path to the Cuffmerge external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="102"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="103"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="117"/>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="120"/>
        <source>Set of annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="118"/>
        <source>Annotations for merging</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="121"/>
        <source>Merged annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="184"/>
        <source>Merges together supplied assemblies.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="448"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="449"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="452"/>
        <source>%1 from %2 with bedtool genomecov.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovTask</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="694"/>
        <source>Unknown file format: </source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="469"/>
        <source>Genome Coverage</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="470"/>
        <source>Calculates genome coverage using bedtools genomecov.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="474"/>
        <source>Input File</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="475"/>
        <source>Set of files to NGS slop</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="476"/>
        <source>Output File</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="477"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="490"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="491"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="495"/>
        <source>Custom directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="496"/>
        <source>Select the custom output directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="498"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="499"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="501"/>
        <source>Genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="502"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt;. (-g)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="504"/>
        <source>Report mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="505"/>
        <source>&lt;b&gt;%1 (%2)&lt;/b&gt; - %3 
&lt;b&gt;%4 (%5)&lt;/b&gt; - %6 
&lt;b&gt;%7 (%8)&lt;/b&gt; - %9 
&lt;b&gt;%10 (%11)&lt;/b&gt; - %12 
&lt;b&gt;%13 (%14)&lt;/b&gt; - %15 
</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="517"/>
        <source>Split</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="518"/>
        <source>Treat âsplitâ BAM or BED12 entries as distinct BED intervals when computing coverage. For BAM files, this uses the CIGAR âNâ and âDâ operations to infer the blocks for computing coverage. For BED12 files, this uses the BlockCount, BlockStarts, and BlockEnds fields (i.e., columns 10,11,12). (-split)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="520"/>
        <source>Strand</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="521"/>
        <source>Calculate coverage of intervals from a specific strand. With BED files, requires at least 6 columns (strand is column 6). (-strand)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="523"/>
        <source>5 prime</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="524"/>
        <source>Calculate coverage of 5â positions (instead of entire interval). (-5)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="526"/>
        <source>3 prime</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="527"/>
        <source>Calculate coverage of 3â positions (instead of entire interval). (-3)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="529"/>
        <source>Max</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="530"/>
        <source>Combine all positions with a depth &gt;= max into a single bin in the histogram. (-max)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="532"/>
        <source>Scale</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="533"/>
        <source>Scale the coverage by a constant factor.Each coverage value is multiplied by this factor before being reported. Useful for normalizing coverage by, e.g., reads per million (RPM). Default is 1.0; i.e., unscaled. (-scale)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="535"/>
        <source>Trackline</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="536"/>
        <source>Adds a UCSC/Genome-Browser track line definition in the first line of the output. (-trackline)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="538"/>
        <source>Trackopts</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="539"/>
        <source>Writes additional track line definition parameters in the first line. (-trackopts)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="543"/>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="576"/>
        <source>Custom</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="574"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="575"/>
        <source>Workflow</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GffreadPrompter</name>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="245"/>
        <source>Extract transcript sequences from the genomic sequence from &lt;u&gt;%1&lt;/u&gt; using transcripts from &lt;u&gt;%2&lt;/u&gt; and save them to the file %3.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSPrompter</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="562"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="569"/>
        <source>Uses &lt;u&gt;%1&lt;/u&gt; as treatment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="571"/>
        <source> and &lt;u&gt;%1&lt;/u&gt; as control</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="574"/>
        <source> to call peaks.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="576"/>
        <source> Outputs all files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="578"/>
        <source> and pileup with &lt;u&gt;%1&lt;/u&gt; span</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSWorker</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="189"/>
        <source>Bad model fold region: %1. Default region is used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="256"/>
        <source>Input control and treatment annotations are the same</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="270"/>
        <source>Treatment features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="271"/>
        <source>ChIP-seq treatment features.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="273"/>
        <source>Control features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="274"/>
        <source>Control features (Optional).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="279"/>
        <source>MACS data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="280"/>
        <source>ChIP-seq treatment features and control features (optional) to call peaks with MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="288"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="289"/>
        <source>Peak locations. Typically used in gene association study like CEAS, or correlation calculation.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="291"/>
        <source>Peak summits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="292"/>
        <source>Peak summits locations for every peaks. Typically used in DNA motif analysis or conservation check.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="294"/>
        <source>Treatment fragments pileup (wiggle)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="295"/>
        <source>Wiggle format files which can be imported to UCSC genome browser/GMOD/Affy IGB.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="298"/>
        <source>MACS output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="299"/>
        <source>ChIP-seq peaks and summits. Pileup data (optional)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="312"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="313"/>
        <source>Directory to save MACS output files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="315"/>
        <source>Name</source>
        <translation type="unfinished">Имя</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="316"/>
        <source>The name string of the experiment. MACS will use this string NAME to create output files like &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; and so on. So please avoid any confliction between these filenames and your existing files (--name).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="320"/>
        <source>Wiggle output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="321"/>
        <source>If this flag is on, MACS will store the fragment pileup in wiggle format for the whole genome data instead of for every chromosomes (--wig) (--single-profile).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="323"/>
        <source>Wiggle space</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="324"/>
        <source>By default, the resolution for saving wiggle files is 10 bps,i.e., MACS will save the raw tag count every 10 bps. You can change it along with &apos;--wig&apos; option (--space).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="327"/>
        <source>Genome size (Mbp)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="328"/>
        <source>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt; It&apos;s the mappable genome size or effective genome size which is defined as the genome size which can be sequenced. Because of the repetitive features on the chromosomes, the actual mappable genome size will be smaller than the original size, about 90% or 70% of the genome size (--gsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="337"/>
        <source>P-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="338"/>
        <source>P-value cutoff. Default is 0.00001, for looser results, try 0.001 instead (--pvalue).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="340"/>
        <source>Q-value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="341"/>
        <source>Minimum FDR (q-value) cutoff for peak detection.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="343"/>
        <source>Use model</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="344"/>
        <source>Whether or not to use MACS paired peaks model (--nomodel).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="346"/>
        <source>Model fold</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="347"/>
        <source>Select the regions within MFOLD range of high-confidence enrichment ratio against. &lt;b&gt;Model fold&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is true, which is the foldchange to chose paired peaks to build paired peaks model. Users need to set a lower(smaller) and upper(larger) number for fold change so that MACS will only use the peaks within these foldchange range to build model (--mfold).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="352"/>
        <source>Shift size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="353"/>
        <source>An arbitrary shift value used as a half of the fragment size when model is not built. &lt;b&gt;Shift size&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is false, which will represent the HALF of the fragment size of your sample. If your sonication and size selection size is 300 bps, after you trim out nearly 100 bps adapters, the fragment size is about 200 bps, so you can specify 100 here (--shiftsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="358"/>
        <source>Keep duplicates</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="359"/>
        <source>It controls the MACS behavior towards duplicate tags at the exact same location -- the same coordination and the same strand. The default &lt;b&gt;auto&lt;/b&gt; option makes MACS calculate the maximum tags at the exact same location based on binomal distribution using 1e-5 as pvalue cutoff; and the &lt;b&gt;all&lt;/b&gt; option keeps every tags. If an &lt;b&gt;integer&lt;/b&gt; is given, at most this number of tags will be kept at the same location (--keep-dup).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="363"/>
        <source>Band width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="364"/>
        <source>The band width which is used to scan the genome for model building. You can set this parameter as the sonication fragment size expected from wet experiment. Used only while building the shifting model (--bw).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="368"/>
        <source>Extended fragment pileup</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="369"/>
        <source>Whether or not to generate extended fragment pileup, local lambda and score tracks at every bp.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="373"/>
        <source>Tag size (optional)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="374"/>
        <source>Length of reads. Determined from first 10 reads if not specified (input &lt;b&gt;0&lt;/b&gt;) (--tsize).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="378"/>
        <source>Use lambda</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="379"/>
        <source>Whether to use local lambda model which can use the local bias at peak regions to throw out false positives (--nolambda).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="381"/>
        <source>Small nearby region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="382"/>
        <source>The small nearby region in basepairs to calculate dynamic lambda. This is used to capture the bias near the peak summit region. Invalid if there is no control data (--slocal).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="385"/>
        <source>Large nearby region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="386"/>
        <source>The large nearby region in basepairs to calculate dynamic lambda.  This is used to capture the surround bias (--llocal).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="389"/>
        <source>Auto bimodal</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="390"/>
        <source>Whether turn on the auto pair model process.If set, when MACS failed to build paired model, it will use the nomodelsettings, the âShift sizeâ parameter to shift and extend each tags (--on-auto).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="410"/>
        <source>If set, MACS will try to call broad peaks by linking nearby highly enriched regions.The linking region is controlled by another cutoff through âBroad cutoffâ.The maximum linking region length is 4 times of d from MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="415"/>
        <source>Cutoff for broad region. This option is not available unless âBroadâ is set.If âP-valueâ is set, this is a pvalue cutoff, otherwise, it&apos;s a qvalue cutoff.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="394"/>
        <source>Scale to large</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="395"/>
        <source> When set, scale the small sample up to the bigger sample.By default, the bigger dataset will be scaled down towards the smaller dataset,which will lead to smaller p/qvalues and more specific results.Keep in mind that scaling down will bring down background noise more (--to-large).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="400"/>
        <source>Shift control</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="401"/>
        <source>When set, control tags will be shifted just as ChIP tags according to their strand before the extension of d, slocal and llocal. By default, control tags are extended centeredat their current positions regardless of strand. You may consider to turn this option on whilecomparing two ChIP datasets of different condition but the same factor.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="406"/>
        <source>Half-extend</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="407"/>
        <source>When set, MACS extends 1/2 d size for each fragment centered at its middle point.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="409"/>
        <source>Broad</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="414"/>
        <source>Broad cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="540"/>
        <source>Find Peaks with MACS</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="541"/>
        <source>Performs peak calling for ChIP-Seq data.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTPrompter</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="129"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="131"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &lt;u&gt;&quot;MAFFT&quot;&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTWorker</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="60"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="61"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Multiple sequence alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="62"/>
        <source>Result of alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="71"/>
        <source>Gap Open Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="72"/>
        <source>Gap Open Penalty.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="73"/>
        <source>Offset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="74"/>
        <source>Works like gap extension penalty.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="76"/>
        <source>Maximum number of iterative refinement.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="78"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="80"/>
        <source>Directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="173"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="75"/>
        <source>Max Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Maximum number of iterative refinement</source>
        <translation type="obsolete">Максимальное число итераций</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="77"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="79"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="88"/>
        <source>Align with MAFFT</source>
        <translation>Выравнивание MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="202"/>
        <source>Aligned %1 with MAFFT</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="89"/>
        <source>MAFFT is a multiple sequence alignment program for unix-like operating systems. &lt;p&gt;&lt;dfn&gt;It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of &amp;lt;&amp;#126;200 sequences), FFT-NS-2 (fast; for alignment of &amp;lt;&amp;#126;10,000 sequences), etc. &lt;/dfn&gt;&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GenePrompter</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="314"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="319"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="320"/>
        <source> to annotate with genes nearby from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GeneWorker</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="192"/>
        <source>Treatment features</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="193"/>
        <source>Result peaks of MACS.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="197"/>
        <source>Peak2gene data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="198"/>
        <source>MACS peaks to get the refgenes near the peak summit/center.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="206"/>
        <source>Gene regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="207"/>
        <source>Annotation for each gene, containing all the peaks nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="209"/>
        <source>Peak regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="210"/>
        <source>Annotation for each peak, containing all the genes nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="213"/>
        <source>Peak2gene output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="214"/>
        <source>Genes containing all the peaks nearby and peaks containing all the genes nearby.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="226"/>
        <source>Output type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="227"/>
        <source>Select which type of genes need to output. &lt;b&gt;up&lt;/b&gt; for genes upstream to peak summit, &lt;b&gt;down&lt;/b&gt; for genes downstream to peak summit, &lt;b&gt;all&lt;/b&gt; for both &lt;b&gt;up&lt;/b&gt; and &lt;b&gt;down&lt;/b&gt; (--op).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="232"/>
        <source>Official gene symbols</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="233"/>
        <source>Output &lt;b&gt;official gene symbol&lt;/b&gt; instead of &lt;b&gt;refseq name&lt;/b&gt; (--symbol).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="235"/>
        <source>Distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="236"/>
        <source>Set a number which unit is base. It will get the refGenes in n bases from peak center (--distance).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="238"/>
        <source>Genome file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="239"/>
        <source>Select a genome file (sqlite3 file) to search refGenes (--genome).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="294"/>
        <source>Annotate Peaks with peak2gene</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="295"/>
        <source>Gets refGenes near the ChIP regions identified by a peak-caller.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosComboBoxWithChecksDelegate</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="150"/>
        <source>Hint:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="150"/>
        <source>Use &apos;cistrome.xml&apos; to descrease the computation time. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosPrompter</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="353"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="361"/>
        <source>Uses regions from &lt;u&gt;%1&lt;/u&gt; to find motifs enriched in them.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="362"/>
        <source> Genome assembly: &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="364"/>
        <source> Finds de novo motifs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="367"/>
        <source> Uses &lt;u&gt;%1&lt;/u&gt; known motifs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="370"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; directory</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosWorker</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="236"/>
        <source>Input regions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="237"/>
        <source>Peak summits.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="241"/>
        <source>SeqPos data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="242"/>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="334"/>
        <source>Finds motifs enriched in a set of regions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="250"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="251"/>
        <source>The directory to store seqpos results.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="253"/>
        <source>Genome assembly version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="254"/>
        <source>UCSC database version (GENOME).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="256"/>
        <source>De novo motifs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="257"/>
        <source>Run de novo motif search (-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="259"/>
        <source>Motif database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="260"/>
        <source>Known motif collections (-m). Warning: computation time increases with selecting additional databases. It is recommended to use cistrome.xml. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="263"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="264"/>
        <source>Name of the output file which stores new motifs found during a de novo search (-n).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="266"/>
        <source>Region width</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="267"/>
        <source>Width of the region to be scanned for motifs; depends on a resolution of assay (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="269"/>
        <source>Pvalue cutoff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="270"/>
        <source>Pvalue cutoff for the motif significance (-p).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="333"/>
        <source>Collect Motifs with SeqPos</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="61"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="62"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="64"/>
        <source>Increases the size of each feature in files from %1 with bedtool slop.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="81"/>
        <source>Slopbed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="82"/>
        <source>Increases the size of each feature in files using bedtools slop.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="86"/>
        <source>Input File</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="87"/>
        <source>Set of files to bedtools slop</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="88"/>
        <source>Output File</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="89"/>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="102"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="103"/>
        <source>Select an output directory. &lt;b&gt;Custom&lt;/b&gt; - specify the output directory in the &apos;Custom directory&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow directory. &lt;b&gt;Input file&lt;/b&gt; - the directory of the input file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="107"/>
        <source>Custom directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="108"/>
        <source>Select the custom output directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="110"/>
        <source>Output file name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="111"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extention.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="113"/>
        <source>Genome</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="114"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt; (-g).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="116"/>
        <source>Each direction increase</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="117"/>
        <source>Increase the BED/GFF/VCF entry by the same number base pairs in each direction. If this parameter is used -l and -l are ignored. Enter 0 to disable. (-b)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="119"/>
        <source>Substract from start</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="120"/>
        <source>The number of base pairs to subtract from the start coordinate. Enter 0 to disable. (-l)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="122"/>
        <source>Add to end</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="123"/>
        <source>The number of base pairs to add to the end coordinate. Enter 0 to disable. (-r)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="125"/>
        <source>Strand-based</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="126"/>
        <source>Define -l and -r based on strand. For example. if used, -l 500 for a negative-stranded feature, it will add 500 bp to the end coordinate. (-s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="128"/>
        <source>As fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="129"/>
        <source>Define -l and -r as a fraction of the featureâs length. E.g. if used on a 1000bp feature, -l 0.50, will add 500 bp âupstreamâ. (-pct)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="131"/>
        <source>Print header</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="132"/>
        <source>Print the header from the input file prior to results. (-header)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="134"/>
        <source>Filter start&gt;end fields</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="135"/>
        <source>Remove lines with start postion greater than end position</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="140"/>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="171"/>
        <source>Custom</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="169"/>
        <source>Input file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="170"/>
        <source>Workflow</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeePrompter</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="125"/>
        <source> from %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="127"/>
        <source>For each MSA&lt;u&gt;%1&lt;/u&gt;, build the alignment using &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt; and send it to output.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeeWorker</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Input MSA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="59"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Multiple sequence alignment</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="60"/>
        <source>Result of alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="69"/>
        <source>Gap Open Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="70"/>
        <source>Gap Open Penalty. Must be negative, best matches get a score of 1000.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="71"/>
        <source>Gap Extension Penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="72"/>
        <source>Gap Extension Penalty. Positive values give rewards to gaps and prevent the alignment of unrelated segments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="73"/>
        <source>Max Iteration</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="74"/>
        <source>Number of iteration on the progressive alignment.&lt;br&gt;0 - no iteration, -1 - Nseq iterations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="76"/>
        <source>Tool Path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="77"/>
        <source>External tool path.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="79"/>
        <source>Directory for temporary file.s</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="170"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="78"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="87"/>
        <source>Align with T-Coffee</source>
        <translation>Выравнивание T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="88"/>
        <source>T-Coffee is a multiple sequence alignment package. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="202"/>
        <source>Aligned %1 with T-Coffee</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TopHatWorker</name>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="110"/>
        <location filename="../src/tophat/TopHatWorker.cpp" line="121"/>
        <source>Input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="111"/>
        <source>Input RNA-Seq reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="115"/>
        <source>TopHat output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="116"/>
        <source>Accepted hits, junctions, insertions and deletions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="122"/>
        <source>TopHat input reads. Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="127"/>
        <source>Input reads url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="128"/>
        <source>TopHat input reads url. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="132"/>
        <source>Input paired reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="133"/>
        <source> Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads. Reads MUST appear in the same order as the *_1 reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="140"/>
        <source>Input paired reads url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="141"/>
        <source>Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="146"/>
        <source>Dataset name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="147"/>
        <source>Use it only when sequences slot(or slots) is specified. Group input reads into chunks for several Tophat runs.
Set it empty if you want to run Tophat once for all input reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="164"/>
        <source>Accepted hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="165"/>
        <source>Accepted hits found by TopHat</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="167"/>
        <source>Sample name</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="168"/>
        <source>Sample name for running Cuffdiff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="170"/>
        <source>Accepted hits url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="171"/>
        <source>The url to the assembly file with the accepted hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="182"/>
        <source>Find Splice Junctions with TopHat</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="183"/>
        <source>TopHat is a fast splice junction mapper for RNA-Seq reads. It aligns RNA-Seq reads to mammalian-sized genomes using the ultra high-throughput short read aligner Bowtie, and then analyzes the mapping results to identify splice junctions between exons.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="191"/>
        <source>Output directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="192"/>
        <source>The base name of output directory. It could be modified with a suffix.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="195"/>
        <source>Samples map</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="196"/>
        <source>The map which divide all input datasets into samples. Every sample has the unique name.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="199"/>
        <source>Bowtie index directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="200"/>
        <source>The directory with the Bowtie index for the reference sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="208"/>
        <source>Bowtie index basename</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="209"/>
        <source>The basename of the Bowtie index for the reference sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="222"/>
        <source>Mate inner distance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="223"/>
        <source>The expected (mean) inner distance between mate pairs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="226"/>
        <source>Mate standard deviation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="227"/>
        <source>The standard deviation for the distribution on inner distances between mate pairs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="230"/>
        <source>Library type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="231"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="234"/>
        <source>No novel junctions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="235"/>
        <source>Only look for reads across junctions indicated in the supplied GFF or junctions file. This parameter is ignored if &lt;i&gt;Raw junctions&lt;/i&gt; or &lt;i&gt;Known transcript file&lt;/i&gt; is not set.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="240"/>
        <source>Raw junctions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="241"/>
        <source>The list of raw junctions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="244"/>
        <source>Known transcript file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="245"/>
        <source>A set of gene model annotations and/or known transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="248"/>
        <source>Max multihits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="249"/>
        <source>Instructs TopHat to allow up to this many alignments to the reference for a given read, and suppresses all alignments for reads with more than this many alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="254"/>
        <source>Segment length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="255"/>
        <source>Each read is cut up into segments, each at least this long. These segments are mapped independently.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="267"/>
        <source>Fusion search</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="268"/>
        <source>Turn on fusion mapping.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="271"/>
        <source>Transcriptome only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="272"/>
        <source>Only align the reads to the transcriptome and report only those mappings as genomic mappings.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="276"/>
        <source>Transcriptome max hits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="277"/>
        <source>Maximum number of mappings allowed for a read, when aligned to the transcriptome (any reads found with more than this number of mappings will be discarded).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="282"/>
        <source>Prefilter multihits</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="283"/>
        <source>When mapping reads on the transcriptome, some repetitive or low complexity reads that would be discarded in the context of the genome may appear to align to the transcript sequences and thus may end up reported as mapped to those genes only. This option directs TopHat to first align the reads to the whole genome in order to determine and exclude such multi-mapped reads (according to the value of the &lt;i&gt;Max multihits&lt;/i&gt; option).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="292"/>
        <source>Min anchor length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="293"/>
        <source>The &lt;i&gt;anchor length&lt;/i&gt;. TopHat will report junctions spanned by reads with at least this many bases on each side of the junction. Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved in spliced alignments is supported by at least one read with this many bases on each side.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="301"/>
        <source>Splice mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="302"/>
        <source>The maximum number of mismatches that may appear in the &lt;i&gt;anchor&lt;/i&gt; region of a spliced alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="306"/>
        <source>Read mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="307"/>
        <source>Final read alignments having more than these many mismatches are discarded.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="311"/>
        <source>Segment mismatches</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="312"/>
        <source>Read segments are mapped independently, allowing up to this many mismatches in each segment alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="317"/>
        <source>Solexa 1.3 quals</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="318"/>
        <source>As of the Illumina GA pipeline version 1.3, quality scores are encoded in Phred-scaled base-64. Use this option for FASTQ files from pipeline 1.3 or later.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="323"/>
        <source>Bowtie version</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="324"/>
        <source>Specifies which Bowtie version should be used.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="327"/>
        <source>Bowtie -n mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="328"/>
        <source>TopHat uses &lt;i&gt;-v&lt;/i&gt; in Bowtie for initial read mapping (the default), but with this option, &lt;i&gt;-n&lt;/i&gt; is used instead. Read segments are always mapped using &lt;i&gt;-v&lt;/i&gt; option.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="334"/>
        <source>Bowtie tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="335"/>
        <source>The path to the Bowtie external tool.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="338"/>
        <source>SAMtools tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="339"/>
        <source>The path to the SAMtools tool. Note that the tool is available in the UGENE External Tool Package.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="343"/>
        <source>TopHat tool path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="344"/>
        <source>The path to the TopHat external tool in UGENE.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="347"/>
        <source>Temporary directory</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="348"/>
        <source>The directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="511"/>
        <source>Aligns RNA-seq reads to a reference and finds splice junctions.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="575"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="604"/>
        <source>Unrecognized value of the Bowtie mode option!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusPrompter</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="179"/>
        <source>Apply VCF variants from &lt;u&gt;%1&lt;/u&gt; to fasta file &lt;u&gt;%2&lt;/u&gt; and save consensus sequence to &lt;u&gt;%3&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusWorker</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="76"/>
        <source>Input fasta slot is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="79"/>
        <source>Input vcf slot is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="127"/>
        <source>Input FASTA and VCF</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="128"/>
        <source>FASTA url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="129"/>
        <source>VCF url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="136"/>
        <source>Fasta consensus url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="144"/>
        <source>Output FASTA consensus</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="145"/>
        <source>The url to the output file with the result consensus.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="155"/>
        <source>Create VCF consensus</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="156"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MACSSupport</name>
    <message>
        <location filename="../src/macs/MACSSupport.cpp" line="44"/>
        <source>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, which analyzes datagenerated by short read sequencers such as Solexa&apos;s Genome Analyzer. MACS empirically models the shiftsize of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.MACS also uses a dynamic Poisson distribution to effectively capture local biases in the genome,allowing for more robust predictions. MACS compares favorably to existing ChIP-Seq peak-finding algorithms, and is freely available.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MACSTask</name>
    <message>
        <location filename="../src/macs/MACSTask.cpp" line="52"/>
        <source>MACS peak calling</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupport</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="67"/>
        <source>&lt;i&gt;MAFFT&lt;/i&gt; is a multiple sequence alignment program for unix-like operating systems. </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="77"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="78"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportContext</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="129"/>
        <source>Align with MAFFT...</source>
        <translation>Выравнивание MAFFT...</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="154"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="155"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="51"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="52"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="97"/>
        <source>MAFFT alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="118"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="127"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="132"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="145"/>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="147"/>
        <source>Can not open output file: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="148"/>
        <source> May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="184"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="186"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="194"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="271"/>
        <source>Failed to apply the result of aligning with MAFFT: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="276"/>
        <source>MAFFT alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output file not found</source>
        <translation type="obsolete">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="289"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportRunDialog</name>
    <message>
        <source>Select input file</source>
        <translation type="obsolete">Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="83"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="84"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="99"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="109"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="133"/>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="136"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="134"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="137"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="323"/>
        <source>input_format_error</source>
        <translation type="unfinished">Ошибка входного формата</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="203"/>
        <source>Generating output trees from MrBayes</source>
        <translation type="unfinished">Создание выходных дереьев черезе MrBayes</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="210"/>
        <source>Output file is not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output file not found</source>
        <translation type="obsolete">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="237"/>
        <source>No trees are found</source>
        <translation type="unfinished">Деревья не найдены</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesPrepareDataForCalculation</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="41"/>
        <source>Generating input file for MrBayes</source>
        <translation type="unfinished">Создание входного файла для MRBayes</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupport</name>
    <message>
        <location filename="../src/mrbayes/MrBayesSupport.cpp" line="57"/>
        <source>&lt;i&gt;MrBayes&lt;/i&gt; is a program for the Bayesian estimation of phylogeny.Bayesian inference of phylogeny is based upon a quantity called the posterior probability distribution of trees, which is the probability of a tree conditioned on the observations. The conditioning is accomplished using Bayes&apos;s theorem. The posterior probability distribution of trees is impossible to calculate analytically; instead, MrBayes uses a simulation technique called Markov chain Monte Carlo (or MCMC) to approximate the posterior probabilities of trees.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupportTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="89"/>
        <source>MrBayes tree calculation</source>
        <translation type="unfinished">MrBayes вычисление деревьев</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesWidget</name>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="70"/>
        <source>Substitution model</source>
        <translation>Модель замен</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="74"/>
        <source>Rate Matrix (fixed)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="180"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="181"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PathLineEdit</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="673"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="675"/>
        <source>Select a file</source>
        <translation>Выберите файл</translation>
    </message>
</context>
<context>
    <name>U2::Peak2GeneSupport</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneSupport.cpp" line="45"/>
        <source>&lt;i&gt;peak2gene&lt;/i&gt; - Gets refGenes near the ChIP regions identified by a peak-caller</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PerlSupport</name>
    <message>
        <location filename="../src/perl/PerlSupport.cpp" line="45"/>
        <source>Perl scripts interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMLGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="201"/>
        <source>Generating output trees from PhyML</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="211"/>
        <source>Output file is not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="237"/>
        <source>No trees are found</source>
        <translation type="unfinished">Деревья не найдены</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="242"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="248"/>
        <source>No result tree in PhyML output</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMLLogParser</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="179"/>
        <source>PhyML finished with error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="196"/>
        <source>UGENE internal error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMLPrepareDataForCalculation</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="40"/>
        <source>Generating input file for PhyML</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="67"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="71"/>
        <source>Internal UGENE error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="75"/>
        <source>Can not open tmp file</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupport</name>
    <message>
        <location filename="../src/phyml/PhyMLSupport.cpp" line="49"/>
        <source>&lt;i&gt;PhyML&lt;/i&gt; is a simple, fast, and accurate algorithm to estimate large phylogenies by maximum likelihood</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupportTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="95"/>
        <source>PhyML tree calculation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="150"/>
        <source>UGENE internal error</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PhyMlWidget</name>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="169"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="221"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="222"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForCAP3Task</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="262"/>
        <source>Failed to initialize sequence writer.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="289"/>
        <source>Failed to write sequence %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForSpideyTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="192"/>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="205"/>
        <source>Failed to write DNA sequence  %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleDjangoSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="106"/>
        <source>: Python module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleNumpySupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="115"/>
        <source>: Python module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="95"/>
        <source>Python module is not installed. Install module or set path to another Python scripts interpreter with installed module in the External Tools settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="60"/>
        <source>Python scripts interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleCelegansdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="151"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleDrosophila2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="157"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleGodbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="115"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleGostatsSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="109"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133adbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="121"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133bdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="127"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133plus2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="133"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu95av2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="139"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleMouse430a2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="145"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgceegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="175"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgdmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="181"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrghsegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="163"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgmmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="169"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleSeqlogoSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="187"/>
        <source>: Rscript module for the %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RModuleSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="95"/>
        <source>R module is not installed. Install module or set path to another R scripts interpreter with installed module in the External Tools settings</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="59"/>
        <source>Rscript interpreter</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::RunCap3AndOpenResultTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="140"/>
        <source>CAP3 run and open result task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="150"/>
        <source>Invalid CAP3 task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="165"/>
        <source>Load document task is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="173"/>
        <source>No assembly is found for provided reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="180"/>
        <source>Project loader is NULL</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SaveAlignmentTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="61"/>
        <source>Export alignment to &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="65"/>
        <source>An alignment is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SaveMSA2SequencesTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="93"/>
        <source>Export alignment to sequence: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SearchToolsInPathTask</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="158"/>
        <source>Search tools in PATH</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SelectPathDialog</name>
    <message>
        <source>Select a file</source>
        <translation type="obsolete">Выберите файл</translation>
    </message>
    <message>
        <source>Select a directory for tempory files</source>
        <translation type="obsolete">Выберите директорию для временных файлов</translation>
    </message>
    <message>
        <source>The directory %1 does not exist.</source>
        <translation type="obsolete">Директория %1 не существует.</translation>
    </message>
    <message>
        <source>Do you want to create this directory?</source>
        <translation type="obsolete">Хотите ли вы содать эту директорию?</translation>
    </message>
    <message>
        <source>Can not create directory for temporary files.</source>
        <translation type="obsolete">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <source>Try choose another directory.</source>
        <translation type="obsolete">Попробуйте выбрать другую директорию.</translation>
    </message>
</context>
<context>
    <name>U2::SeqPosSupport</name>
    <message>
        <location filename="../src/seqpos/SeqPosSupport.cpp" line="46"/>
        <source>&lt;i&gt;SeqPos&lt;/i&gt; - Finds motifs enriched in a set of regions.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideyAlignmentTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="82"/>
        <source>Output file not found</source>
        <translation type="unfinished">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="87"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="98"/>
        <source>Failed to open result file %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupport</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="78"/>
        <source>&lt;i&gt;Spidey&lt;/i&gt; is mRNA-to-DNA alignment program.                    &lt;br&gt;Binaries can be downloaded from http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="80"/>
        <source>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="81"/>
        <source>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="82"/>
        <source>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportContext</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="121"/>
        <source>Align sequence to mRNA</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="148"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="149"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="246"/>
        <source>Failed to align mRNA to genomic sequence: no alignment is found.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TBlastXPlusSupportTask</name>
    <message>
        <location filename="../src/blast_plus/TBlastXPlusSupportTask.cpp" line="61"/>
        <source>Unexpected settings combination</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupport</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="72"/>
        <source>&lt;i&gt;T-Coffee&lt;/i&gt; is a multiple sequence alignment package.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="82"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="83"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportContext</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="135"/>
        <source>Align with T-Coffee...</source>
        <translation>Выравнивание T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="160"/>
        <source>Path for %1 tool not selected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="161"/>
        <source>Do you want to select it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="50"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="51"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="98"/>
        <source>Unsupported alphabet: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="102"/>
        <source>T-Coffee alignment started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="123"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="132"/>
        <source>Subdir for temporary files exists. Can not remove this directory.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="137"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="185"/>
        <source>Output file %1 not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="187"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="195"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="276"/>
        <source>Failed to apply the result of TCoffee: alignment object is not available!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="281"/>
        <source>T-Coffee alignment successfully finished</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output file not found</source>
        <translation type="obsolete">Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="294"/>
        <source>Can not remove directory for temporary files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportRunDialog</name>
    <message>
        <source>Select input file</source>
        <translation type="obsolete">Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="81"/>
        <source>Align</source>
        <translation type="unfinished">Выровнять</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="82"/>
        <source>Cancel</source>
        <translation type="unfinished">Отмена</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="96"/>
        <source>Open an alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="106"/>
        <source>Save an multiple alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="130"/>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="133"/>
        <source>Kalign with Align</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="131"/>
        <source>Input file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="134"/>
        <source>Output file is not set!</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="328"/>
        <source>input_format_error</source>
        <translation type="unfinished">Ошибка входного формата</translation>
    </message>
</context>
<context>
    <name>U2::TabixSupport</name>
    <message>
        <location filename="../src/samtools/TabixSupport.cpp" line="49"/>
        <source>&lt;i&gt;Tabix&lt;/i&gt; is a generic indexer for TAB-delimited genome position files</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TabixSupportTask</name>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="37"/>
        <source>Generate index with Tabix task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="52"/>
        <source>Tabix indexing started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="55"/>
        <source>Input file &apos;%1&apos; is already bgzipped</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="66"/>
        <source>Saving data to file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TopHatSupportTask</name>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="52"/>
        <source>Running TopHat task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="144"/>
        <source>An unexpected error has occurred during preparing the TopHat task!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="145"/>
        <source>Preparing TopHatSupportTask internal error: unable to get a sequence object!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="279"/>
        <source>There are no accepted hits in the result</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupport</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupport.cpp" line="45"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupportTask</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="40"/>
        <source>Create Vcf Consensus</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="55"/>
        <source>VcfConsensus started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="57"/>
        <source>AppSettings is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="59"/>
        <source>UserAppsSettings is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="61"/>
        <source>Temporary directory is not set!</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="66"/>
        <source>Can not create directory for temporary files.</source>
        <translation type="unfinished">Не удаётся создать директорию для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="70"/>
        <source>Saving temporary data to file &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="89"/>
        <source>ExternalToolRegistry is NULL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="93"/>
        <source>There is no VcfConsensus external tool registered</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="94"/>
        <source>There is no Tabix external tool registered</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="117"/>
        <source>Trying to get path of NULL external tool</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="121"/>
        <source>Path to %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>UserAppsSettings</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="138"/>
        <source>Temporary UGENE dir is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
