<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0">
<context>
    <name>IntegralBusPort</name>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="131"/>
        <source>Empty input slot: %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CallVariantsPrompter</name>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="656"/>
        <source>unset</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="657"/>
        <source>For reference sequence from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="658"/>
        <source>with assembly data provided by &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="660"/>
        <source>%1 call variants %2.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CallVariantsTask</name>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="50"/>
        <source>Call variants for %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="63"/>
        <source>reference</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="65"/>
        <source>assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="73"/>
        <source>The %1 file does not exist: %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="86"/>
        <source>No assembly files</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="91"/>
        <source>No dbi storage</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="95"/>
        <source>No sequence URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="133"/>
        <source>No document loaded</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="137"/>
        <source>Incorrect variant track object in %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CallVariantsWorker</name>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="151"/>
        <source>Input sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="152"/>
        <source>A nucleotide reference sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="159"/>
        <source>Input assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="160"/>
        <source>Position sorted alignment file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="167"/>
        <source>Output variations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="168"/>
        <source>Output tracks with SNPs and short INDELs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="173"/>
        <source>Call Variants with SAMtools</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="174"/>
        <source>Calls SNPs and INDELS with SAMtools mpileup and bcftools.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="182"/>
        <source>Output variants file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="183"/>
        <source>The url to the file with the extracted variations.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="186"/>
        <source>Illumina-1.3+ encoding</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="187"/>
        <source>Assume the quality is in the Illumina 1.3+ encoding (mpileup)(-6).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="190"/>
        <source>Count anomalous read pairs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="191"/>
        <source>Do not skip anomalous read pairs in variant calling(mpileup)(-A).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="194"/>
        <source>Disable BAQ computation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="195"/>
        <source>Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments (mpileup)(-B).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="200"/>
        <source>Mapping quality downgrading coefficient</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="201"/>
        <source>Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50 (mpileup)(-C).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="206"/>
        <source>Max number of reads per input BAM</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="207"/>
        <source>At a position, read maximally the number of reads per input BAM (mpileup)(-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="210"/>
        <source>Extended BAQ computation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="211"/>
        <source>Extended BAQ computation. This option helps sensitivity especially for MNPs, but may hurt specificity a little bit (mpileup)(-E).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="215"/>
        <source>BED or position list file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="216"/>
        <source>BED or position list file containing a list of regions or sites where pileup or BCF should be generated (mpileup)(-l).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="220"/>
        <source>Pileup region</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="221"/>
        <source>Only generate pileup in region STR (mpileup)(-r).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="224"/>
        <source>Minimum mapping quality</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="225"/>
        <source>Minimum mapping quality for an alignment to be used (mpileup)(-q).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="228"/>
        <source>Minimum base quality</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="229"/>
        <source>Minimum base quality for a base to be considered (mpileup)(-Q).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="232"/>
        <source>Gap extension error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="233"/>
        <source>Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels (mpileup)(-e).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="236"/>
        <source>Homopolymer errors coefficient</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="237"/>
        <source>Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l (mpileup)(-h).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="240"/>
        <source>No INDELs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="241"/>
        <source>Do not perform INDEL calling (mpileup)(-I).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="244"/>
        <source>Max INDEL depth</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="245"/>
        <source>Skip INDEL calling if the average per-sample depth is above INT (mpileup)(-L).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="248"/>
        <source>Gap open error</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="249"/>
        <source>Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls (mpileup)(-o).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="252"/>
        <source>List of platforms for indels</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="253"/>
        <source>Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained.It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA (mpileup)(-P).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="258"/>
        <source>Retain all possible alternate</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="259"/>
        <source>Retain all possible alternate alleles at variant sites. By default, the view command discards unlikely alleles (bcf view)(-A).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="262"/>
        <source>Indicate PL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="263"/>
        <source>Indicate PL is generated by r921 or before (ordering is different) (bcf view)(-F).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="266"/>
        <source>No genotype information</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="267"/>
        <source>Suppress all individual genotype information (bcf view)(-G).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="270"/>
        <source>A/C/G/T only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="271"/>
        <source>Skip sites where the REF field is not A/C/G/T (bcf view)(-N).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="274"/>
        <source>List of sites</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="275"/>
        <source>List of sites at which information are outputted (bcf view)(-l).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="278"/>
        <source>QCALL likelihood</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="279"/>
        <source>Output the QCALL likelihood format (bcf view)(-Q).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="282"/>
        <source>List of samples</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="283"/>
        <source>List of samples to use. The first column in the input gives the sample names and the second gives the ploidy, which can only be 1 or 2. When the 2nd column is absent, the sample ploidy is assumed to be 2. In the output, the ordering of samples will be identical to the one in FILE (bcf view)(-s).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="289"/>
        <source>Min samples fraction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="290"/>
        <source>skip loci where the fraction of samples covered by reads is below FLOAT (bcf view)(-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="293"/>
        <source>Per-sample genotypes</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="294"/>
        <source>Call per-sample genotypes at variant sites (bcf view)(-g).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="297"/>
        <source>INDEL-to-SNP Ratio</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="298"/>
        <source>Ratio of INDEL-to-SNP mutation rate (bcf view)(-i).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="301"/>
        <source>Max P(ref|D)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="302"/>
        <source>A site is considered to be a variant if P(ref|D)&lt;FLOAT (bcf view)(-p).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="305"/>
        <source>Prior allele frequency spectrum</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="306"/>
        <source>If STR can be full, cond2, flat or the file consisting of error output from a previous variant calling run (bcf view)(-P).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="309"/>
        <source>Mutation rate</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="310"/>
        <source>Scaled mutation rate for variant calling (bcf view)(-t).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="313"/>
        <source>Pair/trio calling</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="314"/>
        <source>Enable pair/trio calling. For trio calling, option -s is usually needed to be applied to configure the trio members and their ordering. In the file supplied to the option -s, the first sample must be the child, the second the father and the third the mother. The valid values of STR are &apos;ÂpairÂ&apos;, &apos;ÂtrioautoÂ&apos;, &apos;ÂtrioxdÂ&apos; and &apos;ÂtrioxsÂ&apos;, where &apos;ÂpairÂ&apos; calls differences between two input samples, and &apos;ÂtrioxdÂ&apos; (&apos;ÂtrioxsÂ&apos;)specifies that the input is from the X chromosome non-PAR regions and the child is a female (male) (bcf view)(-T).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="321"/>
        <source>N group-1 samples</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="322"/>
        <source>Number of group-1 samples. This option is used for dividing the samples into two groups for contrast SNP calling or association test. When this option is in use, the followingVCF INFO will be outputted: PC2, PCHI2 and QCHI2 (bcf view)(-1).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="327"/>
        <source>N permutations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="328"/>
        <source>Number of permutations for association test (effective only with -1) (bcf view)(-U).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="331"/>
        <source>Min P(chi^2)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="332"/>
        <source>Only perform permutations for P(chi^2)&lt;FLOAT (N permutations) (bcf view)(-X).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="336"/>
        <source>Minimum RMS quality</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="337"/>
        <source>Minimum RMS mapping quality for SNPs (varFilter) (-Q).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="339"/>
        <source>Minimum read depth</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="340"/>
        <source>Minimum read depth (varFilter) (-d).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="342"/>
        <source>Maximum read depth</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="343"/>
        <source>Maximum read depth (varFilter) (-D).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="345"/>
        <source>Alternate bases</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="346"/>
        <source>Minimum number of alternate bases (varFilter) (-a).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="348"/>
        <source>Gap size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="349"/>
        <source>SNP within INT bp around a gap to be filtered (varFilter) (-w).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="351"/>
        <source>Window size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="352"/>
        <source>Window size for filtering adjacent gaps (varFilter) (-W).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="354"/>
        <source>Strand bias</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="355"/>
        <source>Minimum P-value for strand bias (given PV4) (varFilter) (-1).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="357"/>
        <source>BaseQ bias</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="358"/>
        <source>Minimum P-value for baseQ bias (varFilter) (-2).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="360"/>
        <source>MapQ bias</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="361"/>
        <source>Minimum P-value for mapQ bias (varFilter) (-3).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="363"/>
        <source>End distance bias</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="364"/>
        <source>Minimum P-value for end distance bias (varFilter) (-4).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="366"/>
        <source>HWE</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="367"/>
        <source>Minimum P-value for HWE (plus F&lt;0) (varFilter) (-e).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="369"/>
        <source>Log filtered</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="370"/>
        <source>Print filtered variants into the log (varFilter) (-p).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="781"/>
        <source>Assembly URL slot is empty. Please, specify the URL slot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="800"/>
        <source>Ref sequence URL slot is empty. Please, specify the URL slot</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="872"/>
        <source>Not enough references</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="877"/>
        <source>The dataset slot is not binded, only the first reference sequence against all assemblies was processed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolMpileupWorker.cpp" line="880"/>
        <source>Not enough assemblies</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SamtoolsMpileupTask</name>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="159"/>
        <source>Samtool mpileup for %1 </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="166"/>
        <source>No reference sequence URL to do pileup</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="171"/>
        <source>No assembly URL to do pileup</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="177"/>
        <source>There is an assembly with an empty path</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="186"/>
        <source>Can not create the directory: </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="238"/>
        <source>Can not run %1 tool</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="244"/>
        <source>%1 tool exited with code %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/AssemblySamtoolsMpileup.cpp" line="246"/>
        <source>Tool %1 finished successfully</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SamtoolsPlugin</name>
    <message>
        <location filename="../src/SamtoolsPlugin.cpp" line="36"/>
        <source>Samtools plugin</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/SamtoolsPlugin.cpp" line="36"/>
        <source>Samtools plugin for NGS data analysis</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
