<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<context>
    <name>ORFDialogBase</name>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="34"/>
        <source>find_dialog_caption</source>
        <translation>Find ORFs in sequence</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="40"/>
        <source>find_dialog_wit</source>
        <translation>Find Open Reading Frames in sequence</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="59"/>
        <source>strand_group</source>
        <translation>Strand</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="71"/>
        <source>strand_both</source>
        <translation>Both</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="81"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="84"/>
        <source>strand_direct_wit</source>
        <translation>Search frames in direct strand only</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="87"/>
        <source>strand_direct_label</source>
        <translation>Direct</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="100"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="103"/>
        <source>strand_complement_wit</source>
        <translation>Search frames in complement strand only</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="106"/>
        <source>strand_complement_label</source>
        <translation>Complement</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="125"/>
        <source>param_group</source>
        <translation>ORF Search parameters</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="139"/>
        <source>min_len_label</source>
        <translation>Min length, bp:</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="188"/>
        <source>must_fit_label</source>
        <translation>Must terminate within region</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="201"/>
        <source>must_init_label</source>
        <translation>Must start with init codon</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="220"/>
        <source>allow_alternative_label</source>
        <translation>Allow alternative init codons</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="246"/>
        <source>find_all_wit</source>
        <translation>Start searching ORFs</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="249"/>
        <source>find_all_button</source>
        <translation>Search</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="290"/>
        <source>hide_button_wit</source>
        <translation>Stop the search then leave the dialog</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="293"/>
        <source>close_button</source>
        <translation>Close</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="265"/>
        <source>save_as_features_tt</source>
        <translation>Save results as annotations</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="268"/>
        <source>save_as_features_wit</source>
        <translation>Save results as annotations</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="325"/>
        <source>clear_list_button</source>
        <translation>Clear results</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="391"/>
        <source>rstart_label</source>
        <translation>Range start:</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="427"/>
        <source>rend_label</source>
        <translation>Range end:</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="133"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="136"/>
        <location filename="../src/ORFQuery.cpp" line="220"/>
        <location filename="../src/ORFWorker.cpp" line="60"/>
        <source>min_len_tip</source>
        <translation>Ignore ORFs shorter than the specified length</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="182"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="185"/>
        <location filename="../src/ORFWorker.cpp" line="61"/>
        <source>must_fit_tip</source>
        <translation>Ignore boundary ORFs which last beyound the search region
(i.e. have no stop codon within the range).</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="195"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="198"/>
        <location filename="../src/ORFQuery.cpp" line="221"/>
        <location filename="../src/ORFWorker.cpp" line="62"/>
        <source>must_init_tip</source>
        <translation>Specifies that each ORF found must start with the start codon</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="214"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="217"/>
        <location filename="../src/ORFQuery.cpp" line="222"/>
        <location filename="../src/ORFWorker.cpp" line="63"/>
        <source>allow_alt_tip</source>
        <translation>Allow ORFs starting with alternative initiation codons, 
accordingly to the current translation table.</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="65"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="68"/>
        <source>strand_both_wit</source>
        <translation>Search in both strands</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="470"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="473"/>
        <source>sel_bt_tip</source>
        <translation>Set the range to current sequence selection</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="483"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="486"/>
        <source>pv_bt_tip</source>
        <translation>Set the range to Panoramic View range</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="496"/>
        <location filename="../src/ui/ORFDialogUI.ui" line="499"/>
        <source>seq_bt_tip</source>
        <translation>Set the range to the whole sequence</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="271"/>
        <source>save_as_annotations</source>
        <translation>Save as annotations..</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="371"/>
        <source>region</source>
        <translation>Location</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="376"/>
        <source>compl</source>
        <translation>Strand</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="381"/>
        <source>len</source>
        <translation>Length</translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="476"/>
        <source>Selection range</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="489"/>
        <source>Overview range</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ui/ORFDialogUI.ui" line="502"/>
        <source>Sequence range</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ORFPrompter</name>
    <message>
        <location filename="../src/ORFWorker.cpp" line="119"/>
        <source> from %1</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="130"/>
        <source>, allow ORFs &lt;u&gt;starting with any codon&lt;/u&gt; other than terminator</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="132"/>
        <source>, take into account &lt;u&gt;alternative start codons&lt;/u&gt;</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="135"/>
        <source>, &lt;u&gt;ignore non-terminated&lt;/u&gt; ORFs</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="152"/>
        <source>For each nucleotide sequence&lt;u&gt;%1&lt;/u&gt;, find ORFs in &lt;u&gt;%2&lt;/u&gt; using the &lt;u&gt;%3&lt;/u&gt;.&lt;br&gt;Detect only ORFs &lt;u&gt;not shorter than %4 bps&lt;/u&gt;%5.&lt;br&gt;Output the list of found regions annotated as &lt;u&gt;%6&lt;/u&gt;.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ORFWorker</name>
    <message>
        <location filename="../src/ORFWorker.cpp" line="242"/>
        <source>Found %1 ORFs</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="44"/>
        <source>Input sequences</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="46"/>
        <source>ORF annotations</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="59"/>
        <source>Genetic code</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="61"/>
        <source>Require stop codon</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="62"/>
        <source>Require init codon</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="63"/>
        <source>Allow alternative codons</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="140"/>
        <source>both strands</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="141"/>
        <source>direct strand</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="142"/>
        <source>complement strand</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="57"/>
        <source>Name of the result annotations marking found ORFs</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="45"/>
        <source>A nucleotide sequence to search ORFs in. Protein sequences are skipped.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="47"/>
        <source>A set of annotations marking ORFs found in the sequence.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="58"/>
        <source>Which strands should be searched: direct, complement or both.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="59"/>
        <source>Which genetic code should be used for translating the input nucleotide sequence.</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="74"/>
        <source>ORF marker</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="75"/>
        <source>Finds Open Reading Frames (ORFs) in each supplied nucleotide sequence, stores found regions as annotations.&lt;p&gt;Protein sequences are skipped if any supplied to input.&lt;p&gt;&lt;dfn&gt;ORFs are DNA sequence regions that could potentially encode a protein, and usually give a good indication of the presence of a gene in the surrounding sequence.&lt;/dfn&gt;&lt;/p&gt;&lt;p&gt;In the sequence, ORFs are located between a start-code sequence (initiation codon) and a stop-code sequence (termination codon), defined by the selected genetic code.&lt;/p&gt;</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="57"/>
        <source>Annotate as</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="60"/>
        <source>Min length</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="189"/>
        <source>ORF: result name is empty, default name used</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="194"/>
        <source>ORF: Incorrect value: min-length must be greater then zero</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="195"/>
        <source>Incorrect value: min-length must be greater then zero</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFWorker.cpp" line="219"/>
        <source>Bad sequence supplied to ORFWorker: %1</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::ORFDialog</name>
    <message>
        <location filename="../src/ORFDialog.cpp" line="190"/>
        <source>cancel_button</source>
        <translation>Stop search</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="190"/>
        <source>close_button</source>
        <translation>Close</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="208"/>
        <source>progress_%1%</source>
        <translation>Progress %1%. </translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="210"/>
        <source>%1_results_found.</source>
        <translation>%1 results found.</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="260"/>
        <source>warning</source>
        <translation>Warning!</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="260"/>
        <source>results_list_not_empty_ask_clear</source>
        <translation>Results list contains results from the previous search. Clear?</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="151"/>
        <source>start_codons</source>
        <translation>Start codons:</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="152"/>
        <source>alt_start_codons</source>
        <translation>Alternative start codons:</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="153"/>
        <source>stop_codons</source>
        <translation>Stop codons:</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="388"/>
        <source>complem_strand</source>
        <translation>complement</translation>
    </message>
    <message>
        <location filename="../src/ORFDialog.cpp" line="388"/>
        <source>direct_strand</source>
        <translation>direct</translation>
    </message>
</context>
<context>
    <name>U2::ORFMarkerPlugin</name>
    <message>
        <location filename="../src/ORFMarkerPlugin.cpp" line="35"/>
        <source>orfname</source>
        <translation>ORF Marker</translation>
    </message>
    <message>
        <location filename="../src/ORFMarkerPlugin.cpp" line="35"/>
        <source>orfdesc</source>
        <translation>Finds open reading frames (ORF) in DNA sequence</translation>
    </message>
</context>
<context>
    <name>U2::ORFViewContext</name>
    <message>
        <location filename="../src/ORFMarkerPlugin.cpp" line="70"/>
        <source>Find ORFs...</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::QDORFActor</name>
    <message>
        <location filename="../src/ORFQuery.cpp" line="91"/>
        <source>both strands</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="93"/>
        <source>direct strand</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="95"/>
        <source>complement strand</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="110"/>
        <source>, allow ORFs &lt;u&gt;starting with any codon&lt;/u&gt; other than terminator</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="112"/>
        <source>, take into account &lt;u&gt;alternative start codons&lt;/u&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="123"/>
        <source>Finds ORFs in &lt;u&gt;%1&lt;/u&gt; using the &lt;u&gt;%2&lt;/u&gt;.&lt;br&gt;Detects only ORFs &lt;u&gt;not shorter than %3, not longer than %4&lt;/u&gt;%5.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="164"/>
        <source>ORF find</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="209"/>
        <source>ORF</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="210"/>
        <source>Finds Open Reading Frames (ORFs) in supplied nucleotide sequence, stores found regions as annotations.&lt;p&gt;Protein sequences are skipped if any.&lt;p&gt;&lt;dfn&gt;ORFs are DNA sequence regions that could potentially encode a protein, and usually give a good indication of the presence of a gene in the surrounding sequence.&lt;/dfn&gt;&lt;/p&gt;&lt;p&gt;In the sequence, ORFs are located between a start-code sequence (initiation codon) and a stop-code sequence (termination codon), defined by the selected genetic code.&lt;/p&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="218"/>
        <source>Which strands should be searched: direct, complement or both.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="219"/>
        <source>Genetic code</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="219"/>
        <source>Which genetic code should be used for translating the input nucleotide sequence.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="220"/>
        <source>Min length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="221"/>
        <source>Require init codon</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="222"/>
        <source>Allow alternative codons</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="223"/>
        <source>Max length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/ORFQuery.cpp" line="223"/>
        <source>Maximum length of annotation allowed</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
